{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53cd05f0-8480-4ea4-a9db-0c5b4577a716",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Investigate coeGenes characteristics with Info from Mostafavi Paper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2ff525c-aede-4651-826d-e4cee84c7029",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d5157a6d-2367-435d-af60-1e3fe173d5be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "\"incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "69afac68-48a9-43c4-9f8d-a24686ab21f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS4_Plot_Config.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "150bd0a3-6d33-416a-b37a-5d324933484d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(httr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e7eaa618-1569-4198-80a5-f326e5616e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(jsonlite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74f7e5cf-2c41-49f5-9493-963448e7fd09",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5ad39aa7-e15e-438f-80e6-a66b1a6f60e1",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "id": "1f12cec2-4e9f-4ff0-904b-bf5beaf63d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl =  '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "id": "b3817590-aef0-46f8-8338-8470bdefb96e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_path = '../data/current/coeqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "id": "67f32d09-b7e9-4818-95bc-4004a4dde118",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl =  '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "b287c084-9558-48c7-8010-a7bf9b1ae86d",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results_path = 'co_qtls_decision_tree/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "id": "7aa2020f-2322-4fa2-a2ea-1bed1be80191",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data_path  = \"/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/meta-data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "id": "66aaadd8-4333-41aa-9a79-de54c8d6eb59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define cell-type for which enrichments should be calculated; alternative: All or cell-type name (CD4_T, CD8_T, Mono, NK, B)\n",
    "cell_type = 'All'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "id": "0dd6c181-54ed-4baf-86dd-7e2bc28daf9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define wheter to include or exclude ribosomal/ mitochondrial genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "id": "018b400f-38f2-4d24-b1f2-4c26055100bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm_rib_mt_egenes = FALSE\n",
    "rm_rib_mt_coegenes = FALSE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f350c99-34fb-4c37-8f92-826e124b4e32",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f695ea07-080b-4891-b37c-12ce85efcd18",
   "metadata": {},
   "source": [
    "## Load ribosomal/mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f91a3473-b75c-4092-a884-324b65db80ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the annotation of ribosomal and mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "d38eceba-6519-4454-a7a6-47393218dfa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_meta = '../data/current/meta-data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "1bf02aa9-4592-4258-b35a-524ebfa5af3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ribosomal_genes = read.csv(paste0(data_path_meta , '/ribosomal_genes.txt.gz') ,sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "55d71a00-94be-4348-a06e-64bcf5069a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "ribosomal_genes = unique(ribosomal_genes$Gene.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "05c1a14b-454e-49fd-ab20-6bd773b79739",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABCE1'</li><li>'ABCF1'</li><li>'APEX1'</li><li>'APOD'</li><li>'ASCC2'</li><li>'ASCC3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ABCE1'\n",
       "\\item 'ABCF1'\n",
       "\\item 'APEX1'\n",
       "\\item 'APOD'\n",
       "\\item 'ASCC2'\n",
       "\\item 'ASCC3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ABCE1'\n",
       "2. 'ABCF1'\n",
       "3. 'APEX1'\n",
       "4. 'APOD'\n",
       "5. 'ASCC2'\n",
       "6. 'ASCC3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ABCE1\" \"ABCF1\" \"APEX1\" \"APOD\"  \"ASCC2\" \"ASCC3\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(ribosomal_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0f28b79-2735-4902-ae0e-f1b40d7e85c8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "bbfe2e28-b43d-47e5-8204-a6d2b552f5b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "8cc9ca55-b7f3-46a5-82e5-1b7f639190b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrial_genes = read.csv(paste0(data_path_meta , '/mitochondrial_genes.txt') ,sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f5191612-40df-4260-8aa0-c9670e4a42c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gene.stable.ID</th><th scope=col>Gene.stable.ID.version</th><th scope=col>Transcript.stable.ID</th><th scope=col>Transcript.stable.ID.version</th><th scope=col>Chromosome.scaffold.name</th><th scope=col>Gene.start..bp.</th><th scope=col>Gene.end..bp.</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000198888</td><td>ENSG00000198888.2</td><td>ENST00000361390</td><td>ENST00000361390.2</td><td>MT</td><td>3307</td><td>4262</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000198763</td><td>ENSG00000198763.3</td><td>ENST00000361453</td><td>ENST00000361453.3</td><td>MT</td><td>4470</td><td>5511</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Gene.stable.ID & Gene.stable.ID.version & Transcript.stable.ID & Transcript.stable.ID.version & Chromosome.scaffold.name & Gene.start..bp. & Gene.end..bp.\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000198888 & ENSG00000198888.2 & ENST00000361390 & ENST00000361390.2 & MT & 3307 & 4262\\\\\n",
       "\t2 & ENSG00000198763 & ENSG00000198763.3 & ENST00000361453 & ENST00000361453.3 & MT & 4470 & 5511\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | Gene.stable.ID &lt;chr&gt; | Gene.stable.ID.version &lt;chr&gt; | Transcript.stable.ID &lt;chr&gt; | Transcript.stable.ID.version &lt;chr&gt; | Chromosome.scaffold.name &lt;chr&gt; | Gene.start..bp. &lt;int&gt; | Gene.end..bp. &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000198888 | ENSG00000198888.2 | ENST00000361390 | ENST00000361390.2 | MT | 3307 | 4262 |\n",
       "| 2 | ENSG00000198763 | ENSG00000198763.3 | ENST00000361453 | ENST00000361453.3 | MT | 4470 | 5511 |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene.stable.ID  Gene.stable.ID.version Transcript.stable.ID\n",
       "1 ENSG00000198888 ENSG00000198888.2      ENST00000361390     \n",
       "2 ENSG00000198763 ENSG00000198763.3      ENST00000361453     \n",
       "  Transcript.stable.ID.version Chromosome.scaffold.name Gene.start..bp.\n",
       "1 ENST00000361390.2            MT                       3307           \n",
       "2 ENST00000361453.3            MT                       4470           \n",
       "  Gene.end..bp.\n",
       "1 4262         \n",
       "2 5511         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "47030031-c668-4832-8d5e-f741d3e3ef4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### convert ENSG to SYMBOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "8d66f0ba-065d-4a49-ad6f-fa31dc7068b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "Warning message in bitr(unique(mitochondrial_genes$Gene.stable.ID), fromType = \"ENSEMBL\", :\n",
      "\"2.14% of input gene IDs are fail to map...\"\n"
     ]
    }
   ],
   "source": [
    "mitochondrial_genes_convert <- bitr(unique(mitochondrial_genes$Gene.stable.ID), \n",
    "                  fromType = \"ENSEMBL\", \n",
    "                  toType = \"SYMBOL\", \n",
    "                  OrgDb = org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "a7a49690-93bc-4fd8-9022-8d5610c57c0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ENSEMBL</th><th scope=col>SYMBOL</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000198888</td><td>ND1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000198763</td><td>ND2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ENSEMBL & SYMBOL\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000198888 & ND1\\\\\n",
       "\t2 & ENSG00000198763 & ND2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | ENSEMBL &lt;chr&gt; | SYMBOL &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ENSG00000198888 | ND1 |\n",
       "| 2 | ENSG00000198763 | ND2 |\n",
       "\n"
      ],
      "text/plain": [
       "  ENSEMBL         SYMBOL\n",
       "1 ENSG00000198888 ND1   \n",
       "2 ENSG00000198763 ND2   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes_convert,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "b10b7759-66f1-4e3b-82c3-68e8e2a783d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrial_genes  = unique(mitochondrial_genes_convert$SYMBOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "35c4592a-9433-4d4e-a140-e8868220f1a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ND1'</li><li>'ND2'</li><li>'COX1'</li><li>'COX2'</li><li>'ATP8'</li><li>'ATP6'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ND1'\n",
       "\\item 'ND2'\n",
       "\\item 'COX1'\n",
       "\\item 'COX2'\n",
       "\\item 'ATP8'\n",
       "\\item 'ATP6'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ND1'\n",
       "2. 'ND2'\n",
       "3. 'COX1'\n",
       "4. 'COX2'\n",
       "5. 'ATP8'\n",
       "6. 'ATP6'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ND1\"  \"ND2\"  \"COX1\" \"COX2\" \"ATP8\" \"ATP6\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "cddd67ec-4665-4a29-9a22-86d6d135483f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "c75ddd3b-d883-495d-afa9-a12803dcfb7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rb_mt_genes = unique(c(ribosomal_genes, mitochondrial_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d1ddf3b-6714-4c7e-947d-28d93e74538b",
   "metadata": {},
   "source": [
    "## Expression info of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7eaae1b4-6e09-4d44-82f1-6ddc74de3a77",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_stats_info = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "93294818-6359-44de-9399-e06c5819830f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types  = c('CD4_T', 'CD8_T', 'Mono', 'NK', 'B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "53d2a264-eb80-4b6f-a527-cbd4cf458bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = c('Franke_split_v2', 'Franke_split_v3', 'wijst2018', 'oneK1k', 'multiome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7414e4ad-684c-4ca1-afd0-6cf890268986",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type'. You can override using the `.groups`\n",
      "argument.\n"
     ]
    }
   ],
   "source": [
    "for(i in datasets_to_load){\n",
    "    for (j in cell_types){\n",
    "        if(file.exists(paste0( \"../data/current/summary_stats/D1_Sample_gene_statisticwg3_\", i, j, \".Qced.Normalized.SCs.Rds.csv\"))){\n",
    "          \n",
    "                result_file = fread( paste0( \"../data/current/summary_stats/D1_Sample_gene_statisticwg3_\", i, j, \".Qced.Normalized.SCs.Rds.csv\")) # F12_Tests_after_filtering_normalized.csv --> after decision tree filtering; F3_decision_tree_input.csv\n",
    "\n",
    "                result_file$dataset = i\n",
    "                result_file$cell_type = j \n",
    "            \n",
    "                ### aggregate on a gene level\n",
    "                result_file = result_file %>% group_by(dataset, cell_type, gene) %>% summarise(amount_samples_non_zero = n(),\n",
    "                                                                                 mean_amount_cells = mean(amount_cells),\n",
    "                                                                                 mean_expression = mean(mean_counts),\n",
    "                                                                                 mean_percentage_zero = mean(percentage_zero),\n",
    "                                                                                 mean_non_zero_counts = mean(amount_non_zero_counts))\n",
    "                #result_file$p_val_mt = NULL   # TBD: remove when this information has been added to all result files\n",
    "                #result_file$q_val = NULL\n",
    "                #result_file$co_egene = NULL\n",
    "                summary_stats_info  = rbind(result_file, summary_stats_info )\n",
    "    \n",
    "            }\n",
    "    }\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a4c8ab30-93a8-440e-929f-b37d25085da0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>gene</th><th scope=col>amount_samples_non_zero</th><th scope=col>mean_amount_cells</th><th scope=col>mean_expression</th><th scope=col>mean_percentage_zero</th><th scope=col>mean_non_zero_counts</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>multiome</td><td>B</td><td>A1BG    </td><td>5</td><td>219.0000</td><td>0.04113295</td><td>94.49146</td><td>10.4</td></tr>\n",
       "\t<tr><td>multiome</td><td>B</td><td>A1BG-AS1</td><td>3</td><td>154.6667</td><td>0.04690375</td><td>93.49882</td><td>10.0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " dataset & cell\\_type & gene & amount\\_samples\\_non\\_zero & mean\\_amount\\_cells & mean\\_expression & mean\\_percentage\\_zero & mean\\_non\\_zero\\_counts\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t multiome & B & A1BG     & 5 & 219.0000 & 0.04113295 & 94.49146 & 10.4\\\\\n",
       "\t multiome & B & A1BG-AS1 & 3 & 154.6667 & 0.04690375 & 93.49882 & 10.0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 8\n",
       "\n",
       "| dataset &lt;chr&gt; | cell_type &lt;chr&gt; | gene &lt;chr&gt; | amount_samples_non_zero &lt;int&gt; | mean_amount_cells &lt;dbl&gt; | mean_expression &lt;dbl&gt; | mean_percentage_zero &lt;dbl&gt; | mean_non_zero_counts &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| multiome | B | A1BG     | 5 | 219.0000 | 0.04113295 | 94.49146 | 10.4 |\n",
       "| multiome | B | A1BG-AS1 | 3 | 154.6667 | 0.04690375 | 93.49882 | 10.0 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset  cell_type gene     amount_samples_non_zero mean_amount_cells\n",
       "1 multiome B         A1BG     5                       219.0000         \n",
       "2 multiome B         A1BG-AS1 3                       154.6667         \n",
       "  mean_expression mean_percentage_zero mean_non_zero_counts\n",
       "1 0.04113295      94.49146             10.4                \n",
       "2 0.04690375      93.49882             10.0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(summary_stats_info,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "41cfb0ea-d9cf-43ef-b7ea-dd353a715802",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B'</li><li>'NK'</li><li>'Mono'</li><li>'CD8_T'</li><li>'CD4_T'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B'\n",
       "\\item 'NK'\n",
       "\\item 'Mono'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'CD4\\_T'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B'\n",
       "2. 'NK'\n",
       "3. 'Mono'\n",
       "4. 'CD8_T'\n",
       "5. 'CD4_T'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"B\"     \"NK\"    \"Mono\"  \"CD8_T\" \"CD4_T\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(summary_stats_info$cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "78f32d40-5b75-4a38-8fed-a05c646e1e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Adjust the dataset info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6f54d55a-33ab-41b0-ab0c-38e865625a2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_stats_info$dataset = str_replace(summary_stats_info$dataset, 'wijst2018', 'wijst')\n",
    "summary_stats_info$dataset = str_replace(summary_stats_info$dataset, 'Franke_split_v2', 'Franke_v2')\n",
    "summary_stats_info$dataset = str_replace(summary_stats_info$dataset, 'Franke_split_v3', 'Franke_v3')\n",
    "summary_stats_info$dataset = str_replace(summary_stats_info$dataset, 'oneK1k', 'oneK1K')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9f243963-a7d1-4a57-a329-70adc86bc759",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'multiome'</li><li>'oneK1K'</li><li>'wijst'</li><li>'Franke_v3'</li><li>'Franke_v2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'multiome'\n",
       "\\item 'oneK1K'\n",
       "\\item 'wijst'\n",
       "\\item 'Franke\\_v3'\n",
       "\\item 'Franke\\_v2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'multiome'\n",
       "2. 'oneK1K'\n",
       "3. 'wijst'\n",
       "4. 'Franke_v3'\n",
       "5. 'Franke_v2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"multiome\"  \"oneK1K\"    \"wijst\"     \"Franke_v3\" \"Franke_v2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(summary_stats_info$dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4b051e0d-b64c-4ac5-a1be-2468faefce1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Franke_v2</td><td>5</td></tr>\n",
       "\t<tr><td>Franke_v3</td><td>5</td></tr>\n",
       "\t<tr><td>multiome </td><td>5</td></tr>\n",
       "\t<tr><td>oneK1K   </td><td>5</td></tr>\n",
       "\t<tr><td>wijst    </td><td>5</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " dataset & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t Franke\\_v2 & 5\\\\\n",
       "\t Franke\\_v3 & 5\\\\\n",
       "\t multiome  & 5\\\\\n",
       "\t oneK1K    & 5\\\\\n",
       "\t wijst     & 5\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 5 × 2\n",
       "\n",
       "| dataset &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| Franke_v2 | 5 |\n",
       "| Franke_v3 | 5 |\n",
       "| multiome  | 5 |\n",
       "| oneK1K    | 5 |\n",
       "| wijst     | 5 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset   n\n",
       "1 Franke_v2 5\n",
       "2 Franke_v3 5\n",
       "3 multiome  5\n",
       "4 oneK1K    5\n",
       "5 wijst     5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(summary_stats_info[,c('dataset', 'cell_type')]) %>% group_by(dataset) %>% dplyr::count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "36a24d94-68e7-4496-99fe-44670d1d792c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>gene</th><th scope=col>amount_samples_non_zero</th><th scope=col>mean_amount_cells</th><th scope=col>mean_expression</th><th scope=col>mean_percentage_zero</th><th scope=col>mean_non_zero_counts</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>multiome</td><td>B</td><td>A1BG    </td><td>5</td><td>219.0000</td><td>0.04113295</td><td>94.49146</td><td>10.4</td></tr>\n",
       "\t<tr><td>multiome</td><td>B</td><td>A1BG-AS1</td><td>3</td><td>154.6667</td><td>0.04690375</td><td>93.49882</td><td>10.0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " dataset & cell\\_type & gene & amount\\_samples\\_non\\_zero & mean\\_amount\\_cells & mean\\_expression & mean\\_percentage\\_zero & mean\\_non\\_zero\\_counts\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t multiome & B & A1BG     & 5 & 219.0000 & 0.04113295 & 94.49146 & 10.4\\\\\n",
       "\t multiome & B & A1BG-AS1 & 3 & 154.6667 & 0.04690375 & 93.49882 & 10.0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 8\n",
       "\n",
       "| dataset &lt;chr&gt; | cell_type &lt;chr&gt; | gene &lt;chr&gt; | amount_samples_non_zero &lt;int&gt; | mean_amount_cells &lt;dbl&gt; | mean_expression &lt;dbl&gt; | mean_percentage_zero &lt;dbl&gt; | mean_non_zero_counts &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| multiome | B | A1BG     | 5 | 219.0000 | 0.04113295 | 94.49146 | 10.4 |\n",
       "| multiome | B | A1BG-AS1 | 3 | 154.6667 | 0.04690375 | 93.49882 | 10.0 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset  cell_type gene     amount_samples_non_zero mean_amount_cells\n",
       "1 multiome B         A1BG     5                       219.0000         \n",
       "2 multiome B         A1BG-AS1 3                       154.6667         \n",
       "  mean_expression mean_percentage_zero mean_non_zero_counts\n",
       "1 0.04113295      94.49146             10.4                \n",
       "2 0.04690375      93.49882             10.0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(summary_stats_info,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cb443d2a-af5b-4c90-88c2-e4e6a1484570",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 5 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>gene</th><th scope=col>amount_samples_non_zero</th><th scope=col>mean_amount_cells</th><th scope=col>mean_expression</th><th scope=col>mean_percentage_zero</th><th scope=col>mean_non_zero_counts</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>multiome </td><td>CD4_T</td><td>A1BG</td><td> 113</td><td>549.8584</td><td>0.02652148</td><td>96.37977</td><td>18.04425</td></tr>\n",
       "\t<tr><td>oneK1K   </td><td>CD4_T</td><td>A1BG</td><td>1012</td><td>474.6630</td><td>0.10414159</td><td>90.48451</td><td>44.59091</td></tr>\n",
       "\t<tr><td>wijst    </td><td>CD4_T</td><td>A1BG</td><td>  37</td><td>274.6757</td><td>0.09206075</td><td>91.33705</td><td>22.91892</td></tr>\n",
       "\t<tr><td>Franke_v3</td><td>CD4_T</td><td>A1BG</td><td>  47</td><td>357.6596</td><td>0.15977212</td><td>86.07066</td><td>51.08511</td></tr>\n",
       "\t<tr><td>Franke_v2</td><td>CD4_T</td><td>A1BG</td><td>  86</td><td>409.1395</td><td>0.08670870</td><td>91.87012</td><td>33.19767</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 5 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " dataset & cell\\_type & gene & amount\\_samples\\_non\\_zero & mean\\_amount\\_cells & mean\\_expression & mean\\_percentage\\_zero & mean\\_non\\_zero\\_counts\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t multiome  & CD4\\_T & A1BG &  113 & 549.8584 & 0.02652148 & 96.37977 & 18.04425\\\\\n",
       "\t oneK1K    & CD4\\_T & A1BG & 1012 & 474.6630 & 0.10414159 & 90.48451 & 44.59091\\\\\n",
       "\t wijst     & CD4\\_T & A1BG &   37 & 274.6757 & 0.09206075 & 91.33705 & 22.91892\\\\\n",
       "\t Franke\\_v3 & CD4\\_T & A1BG &   47 & 357.6596 & 0.15977212 & 86.07066 & 51.08511\\\\\n",
       "\t Franke\\_v2 & CD4\\_T & A1BG &   86 & 409.1395 & 0.08670870 & 91.87012 & 33.19767\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 5 × 8\n",
       "\n",
       "| dataset &lt;chr&gt; | cell_type &lt;chr&gt; | gene &lt;chr&gt; | amount_samples_non_zero &lt;int&gt; | mean_amount_cells &lt;dbl&gt; | mean_expression &lt;dbl&gt; | mean_percentage_zero &lt;dbl&gt; | mean_non_zero_counts &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| multiome  | CD4_T | A1BG |  113 | 549.8584 | 0.02652148 | 96.37977 | 18.04425 |\n",
       "| oneK1K    | CD4_T | A1BG | 1012 | 474.6630 | 0.10414159 | 90.48451 | 44.59091 |\n",
       "| wijst     | CD4_T | A1BG |   37 | 274.6757 | 0.09206075 | 91.33705 | 22.91892 |\n",
       "| Franke_v3 | CD4_T | A1BG |   47 | 357.6596 | 0.15977212 | 86.07066 | 51.08511 |\n",
       "| Franke_v2 | CD4_T | A1BG |   86 | 409.1395 | 0.08670870 | 91.87012 | 33.19767 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset   cell_type gene amount_samples_non_zero mean_amount_cells\n",
       "1 multiome  CD4_T     A1BG  113                    549.8584         \n",
       "2 oneK1K    CD4_T     A1BG 1012                    474.6630         \n",
       "3 wijst     CD4_T     A1BG   37                    274.6757         \n",
       "4 Franke_v3 CD4_T     A1BG   47                    357.6596         \n",
       "5 Franke_v2 CD4_T     A1BG   86                    409.1395         \n",
       "  mean_expression mean_percentage_zero mean_non_zero_counts\n",
       "1 0.02652148      96.37977             18.04425            \n",
       "2 0.10414159      90.48451             44.59091            \n",
       "3 0.09206075      91.33705             22.91892            \n",
       "4 0.15977212      86.07066             51.08511            \n",
       "5 0.08670870      91.87012             33.19767            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary_stats_info[(summary_stats_info$gene == 'A1BG') & (summary_stats_info$cell_type == 'CD4_T'),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "378219b6-aa0f-4b53-9fd6-083c95cd2236",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate weighted mean expression per Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "89fb8997-48dd-4eb0-afd4-0c7c529b7aee",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene'. You can override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "summary_stats_info_meta = summary_stats_info %>% group_by(gene, cell_type) %>% summarise(weighted_mean_expression = sum((mean_expression * amount_samples_non_zero))/sum(amount_samples_non_zero))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4770efb0-40a2-4e32-9da0-67421a5fbdee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "57cd538d-4cf2-4eda-b41c-39436adfaead",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(cell_type != 'All'){\n",
    "    summary_stats_info_meta = summary_stats_info_meta[summary_stats_info_meta$cell_type == cell_type ,]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5eb0caaa-8362-4247-9cc8-8098bdaa968f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(cell_type == 'All'){\n",
    "    summary_stats_info_meta =  summary_stats_info_meta %>% group_by(gene) %>% summarise(weighted_mean_expression = mean(weighted_mean_expression))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "684abb8a-d986-47a6-a9ea-6ba1ee215c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#unique(summary_stats_info_meta$cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ae967e9a-51c5-4a55-9d7c-6bca7e864727",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene</th><th scope=col>weighted_mean_expression</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG    </td><td>0.13038935</td></tr>\n",
       "\t<tr><td>A1BG-AS1</td><td>0.03941786</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " gene & weighted\\_mean\\_expression\\\\\n",
       " <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t A1BG     & 0.13038935\\\\\n",
       "\t A1BG-AS1 & 0.03941786\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| gene &lt;chr&gt; | weighted_mean_expression &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| A1BG     | 0.13038935 |\n",
       "| A1BG-AS1 | 0.03941786 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene     weighted_mean_expression\n",
       "1 A1BG     0.13038935              \n",
       "2 A1BG-AS1 0.03941786              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(summary_stats_info_meta,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e1a4db5-d06b-44f6-a95e-ca3a71c0f60d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Mostafavi Paper Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "617f1b64-3f9c-4e0c-8f50-8762c5c85dac",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Scores for genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "faa13d6d-7b7b-4d26-a037-e847e7c95a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "paper_scores = fread(\"/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/Mostafavi_gene_annots/pc_genes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ca88d58a-1af1-4d43-91d9-670dc34daa56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 22</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>hgnc_id</th><th scope=col>GeneSymbol</th><th scope=col>gene</th><th scope=col>TSSD</th><th scope=col>length</th><th scope=col>CDS_length</th><th scope=col>pLI</th><th scope=col>LOEUF</th><th scope=col>ABC_count</th><th scope=col>ABC_length_per_type</th><th scope=col>⋯</th><th scope=col>promoter_count</th><th scope=col>connect_decile</th><th scope=col>connect_quantile</th><th scope=col>connectedness</th><th scope=col>PPI_degree_decile</th><th scope=col>PPI_degree_quantile</th><th scope=col>PPI_degree_cat</th><th scope=col>TF</th><th scope=col>HI</th><th scope=col>hs</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HGNC:14825</td><td>ENSG00000186092</td><td>OR4F5 </td><td> 0</td><td>6.166</td><td>0.981</td><td>0.030354</td><td>1.84</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>4</td><td>2</td><td>1</td><td>0</td><td> 0</td><td>0.004438489</td></tr>\n",
       "\t<tr><td>HGNC:15079</td><td>ENSG00000284662</td><td>OR4F16</td><td>11</td><td>0.938</td><td>0.939</td><td>      NA</td><td>  NA</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>4</td><td>2</td><td>1</td><td>0</td><td>NA</td><td>         NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 22\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " hgnc\\_id & GeneSymbol & gene & TSSD & length & CDS\\_length & pLI & LOEUF & ABC\\_count & ABC\\_length\\_per\\_type & ⋯ & promoter\\_count & connect\\_decile & connect\\_quantile & connectedness & PPI\\_degree\\_decile & PPI\\_degree\\_quantile & PPI\\_degree\\_cat & TF & HI & hs\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & ⋯ & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t HGNC:14825 & ENSG00000186092 & OR4F5  &  0 & 6.166 & 0.981 & 0.030354 & 1.84 & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 4 & 2 & 1 & 0 &  0 & 0.004438489\\\\\n",
       "\t HGNC:15079 & ENSG00000284662 & OR4F16 & 11 & 0.938 & 0.939 &       NA &   NA & 0 & 0 & ⋯ & 0 & 0 & 0 & 0 & 4 & 2 & 1 & 0 & NA &          NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 22\n",
       "\n",
       "| hgnc_id &lt;chr&gt; | GeneSymbol &lt;chr&gt; | gene &lt;chr&gt; | TSSD &lt;int&gt; | length &lt;dbl&gt; | CDS_length &lt;dbl&gt; | pLI &lt;dbl&gt; | LOEUF &lt;dbl&gt; | ABC_count &lt;int&gt; | ABC_length_per_type &lt;dbl&gt; | ⋯ ⋯ | promoter_count &lt;int&gt; | connect_decile &lt;int&gt; | connect_quantile &lt;int&gt; | connectedness &lt;int&gt; | PPI_degree_decile &lt;int&gt; | PPI_degree_quantile &lt;int&gt; | PPI_degree_cat &lt;int&gt; | TF &lt;int&gt; | HI &lt;int&gt; | hs &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| HGNC:14825 | ENSG00000186092 | OR4F5  |  0 | 6.166 | 0.981 | 0.030354 | 1.84 | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 4 | 2 | 1 | 0 |  0 | 0.004438489 |\n",
       "| HGNC:15079 | ENSG00000284662 | OR4F16 | 11 | 0.938 | 0.939 |       NA |   NA | 0 | 0 | ⋯ | 0 | 0 | 0 | 0 | 4 | 2 | 1 | 0 | NA |          NA |\n",
       "\n"
      ],
      "text/plain": [
       "  hgnc_id    GeneSymbol      gene   TSSD length CDS_length pLI      LOEUF\n",
       "1 HGNC:14825 ENSG00000186092 OR4F5   0   6.166  0.981      0.030354 1.84 \n",
       "2 HGNC:15079 ENSG00000284662 OR4F16 11   0.938  0.939            NA   NA \n",
       "  ABC_count ABC_length_per_type ⋯ promoter_count connect_decile\n",
       "1 0         0                   ⋯ 0              0             \n",
       "2 0         0                   ⋯ 0              0             \n",
       "  connect_quantile connectedness PPI_degree_decile PPI_degree_quantile\n",
       "1 0                0             4                 2                  \n",
       "2 0                0             4                 2                  \n",
       "  PPI_degree_cat TF HI hs         \n",
       "1 1              0   0 0.004438489\n",
       "2 1              0  NA          NA"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(paper_scores,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c12af0c4-5138-4a5d-a59d-0ee216b96be7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "18328"
      ],
      "text/latex": [
       "18328"
      ],
      "text/markdown": [
       "18328"
      ],
      "text/plain": [
       "[1] 18328"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(paper_scores$gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "528d7a3b-ed19-401d-a795-23a98237272d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Scores for genes and GO term annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9651e1ae-4ef9-4077-9384-5b87208dc55c",
   "metadata": {},
   "outputs": [],
   "source": [
    "paper_scores_go = fread(\"/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/Mostafavi_gene_annots/genes_multiGO.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c7606f5f-5ca9-4a19-a59d-397652734c27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene</th><th scope=col>GO_BP_count</th><th scope=col>GO_BP_count_400_auc</th><th scope=col>GO_BP_count_400_auc_decile</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>HGNC:100  </td><td> 67</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><td>HGNC:10000</td><td>127</td><td>4</td><td>9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 4\n",
       "\\begin{tabular}{llll}\n",
       " gene & GO\\_BP\\_count & GO\\_BP\\_count\\_400\\_auc & GO\\_BP\\_count\\_400\\_auc\\_decile\\\\\n",
       " <chr> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t HGNC:100   &  67 & 0 & 1\\\\\n",
       "\t HGNC:10000 & 127 & 4 & 9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 4\n",
       "\n",
       "| gene &lt;chr&gt; | GO_BP_count &lt;int&gt; | GO_BP_count_400_auc &lt;int&gt; | GO_BP_count_400_auc_decile &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| HGNC:100   |  67 | 0 | 1 |\n",
       "| HGNC:10000 | 127 | 4 | 9 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene       GO_BP_count GO_BP_count_400_auc GO_BP_count_400_auc_decile\n",
       "1 HGNC:100    67         0                   1                         \n",
       "2 HGNC:10000 127         4                   9                         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(paper_scores_go,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81afbeb8-0681-4e40-9767-d87ed68e9dc2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "818a721d-97ab-4081-9c0f-ae8508d07a2d",
   "metadata": {},
   "source": [
    "## Load significant co-eQTL results from meta-analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "id": "6f7989d9-122a-4871-a4fc-c7880da8f4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_var = '' # Define which significant results to load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "id": "25c958bf-1fa2-45e8-b85b-543d1e85ba82",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(sign_var  != 'q_val'){\n",
    "    coeqtls = fread(paste0('/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/coeqtl_mapping/co_qtls_decision_tree/shared_results/5DS_Meta_Analysis_Sign_coeQTLs133p_val_mt_eGene', '_.csv'))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "id": "006d2ea4-fac2-4569-aea0-1dbfba690d5c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "coeqtls = as.data.frame(coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "id": "8becff6c-8388-4959-a31c-03a653a8be19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "91628"
      ],
      "text/latex": [
       "91628"
      ],
      "text/markdown": [
       "91628"
      ],
      "text/plain": [
       "[1] 91628"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "id": "04081160-b3ce-4112-b64d-dcb62280d36f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define whether to filter on specific cell-type or keep all coeqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "id": "fab43981-4636-4cdb-bbe2-41f29afb8623",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(cell_type != 'All'){\n",
    "    coeqtls = coeqtls[ (coeqtls$cell_type == cell_type),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "id": "650d6782-9218-4cef-8109-a15037b66431",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 176</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>eGene</th><th scope=col>coeGene</th><th scope=col>snp_id</th><th scope=col>cell_type</th><th scope=col>feature_id</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>p_value</th><th scope=col>z_score</th><th scope=col>⋯</th><th scope=col>eQTL_remap2022_1</th><th scope=col>eQTL_remap2022_2</th><th scope=col>eQTL_remap2022_3</th><th scope=col>eQTL_HPA</th><th scope=col>eQTL_MIRNA</th><th scope=col>eQTL_TF</th><th scope=col>coeGene_is_enriched_TF</th><th scope=col>cctqtl</th><th scope=col>simulation_score</th><th scope=col>ccInteraction</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>A2M</td><td>BACH2</td><td>12:8971336:T:C</td><td>CD4_T</td><td>A2M_BACH2</td><td>12:8971336:T:C;A2M_BACH2</td><td>977</td><td>2.687871e-11</td><td>-6.662730</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td></td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>A2M</td><td>BACH2</td><td>12:8992460:T:C</td><td>CD4_T</td><td>A2M_BACH2</td><td>12:8992460:T:C;A2M_BACH2</td><td>977</td><td>1.680350e-10</td><td>-6.388031</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td></td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 176\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & V1 & eGene & coeGene & snp\\_id & cell\\_type & feature\\_id & triplet & n\\_samples & p\\_value & z\\_score & ⋯ & eQTL\\_remap2022\\_1 & eQTL\\_remap2022\\_2 & eQTL\\_remap2022\\_3 & eQTL\\_HPA & eQTL\\_MIRNA & eQTL\\_TF & coeGene\\_is\\_enriched\\_TF & cctqtl & simulation\\_score & ccInteraction\\\\\n",
       "  & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <int> & <chr> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & A2M & BACH2 & 12:8971336:T:C & CD4\\_T & A2M\\_BACH2 & 12:8971336:T:C;A2M\\_BACH2 & 977 & 2.687871e-11 & -6.662730 & ⋯ & NA & NA & NA & NA & NA & NA & NA &  & 0 & 0\\\\\n",
       "\t2 & 2 & A2M & BACH2 & 12:8992460:T:C & CD4\\_T & A2M\\_BACH2 & 12:8992460:T:C;A2M\\_BACH2 & 977 & 1.680350e-10 & -6.388031 & ⋯ & NA & NA & NA & NA & NA & NA & NA &  & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 176\n",
       "\n",
       "| <!--/--> | V1 &lt;int&gt; | eGene &lt;chr&gt; | coeGene &lt;chr&gt; | snp_id &lt;chr&gt; | cell_type &lt;chr&gt; | feature_id &lt;chr&gt; | triplet &lt;chr&gt; | n_samples &lt;int&gt; | p_value &lt;dbl&gt; | z_score &lt;dbl&gt; | ⋯ ⋯ | eQTL_remap2022_1 &lt;lgl&gt; | eQTL_remap2022_2 &lt;lgl&gt; | eQTL_remap2022_3 &lt;lgl&gt; | eQTL_HPA &lt;lgl&gt; | eQTL_MIRNA &lt;lgl&gt; | eQTL_TF &lt;lgl&gt; | coeGene_is_enriched_TF &lt;int&gt; | cctqtl &lt;chr&gt; | simulation_score &lt;dbl&gt; | ccInteraction &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | A2M | BACH2 | 12:8971336:T:C | CD4_T | A2M_BACH2 | 12:8971336:T:C;A2M_BACH2 | 977 | 2.687871e-11 | -6.662730 | ⋯ | NA | NA | NA | NA | NA | NA | NA | <!----> | 0 | 0 |\n",
       "| 2 | 2 | A2M | BACH2 | 12:8992460:T:C | CD4_T | A2M_BACH2 | 12:8992460:T:C;A2M_BACH2 | 977 | 1.680350e-10 | -6.388031 | ⋯ | NA | NA | NA | NA | NA | NA | NA | <!----> | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 eGene coeGene snp_id         cell_type feature_id triplet                 \n",
       "1 1  A2M   BACH2   12:8971336:T:C CD4_T     A2M_BACH2  12:8971336:T:C;A2M_BACH2\n",
       "2 2  A2M   BACH2   12:8992460:T:C CD4_T     A2M_BACH2  12:8992460:T:C;A2M_BACH2\n",
       "  n_samples p_value      z_score   ⋯ eQTL_remap2022_1 eQTL_remap2022_2\n",
       "1 977       2.687871e-11 -6.662730 ⋯ NA               NA              \n",
       "2 977       1.680350e-10 -6.388031 ⋯ NA               NA              \n",
       "  eQTL_remap2022_3 eQTL_HPA eQTL_MIRNA eQTL_TF coeGene_is_enriched_TF cctqtl\n",
       "1 NA               NA       NA         NA      NA                           \n",
       "2 NA               NA       NA         NA      NA                           \n",
       "  simulation_score ccInteraction\n",
       "1 0                0            \n",
       "2 0                0            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtls,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "id": "05410d43-e970-4630-8f4d-4a6334348949",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "91628"
      ],
      "text/latex": [
       "91628"
      ],
      "text/markdown": [
       "91628"
      ],
      "text/plain": [
       "[1] 91628"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "id": "5743f335-e638-46db-bb79-23fa1af76722",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove co-eQTLs caused by ribosomal / mitochondrial genes (also co-eGenes that only form a co-eQTL with a ribosomal eGene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "id": "60ab2496-b34c-474b-8320-cc5fa787cb5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(rm_rib_mt_egenes == TRUE){\n",
    "    coeqtls = coeqtls[(coeqtls$coeGene_mt == FALSE) & (coeqtls$coeGene_rib == FALSE) & (coeqtls$eGene_rib == FALSE) & (coeqtls$eGene_mt == FALSE),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "id": "65471217-2c88-4d0c-b88c-c578dcca64e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "91628"
      ],
      "text/latex": [
       "91628"
      ],
      "text/markdown": [
       "91628"
      ],
      "text/plain": [
       "[1] 91628"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtls)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0ef8d3f-aab0-49f4-9093-c354e6777f96",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Tested eGenes and coeGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "id": "3b68b78d-5936-4bc9-b7fb-5bd81889ece8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load genes passing decision tree filter in the different datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "id": "6376b702-363c-491f-8ab7-f0a43aecff47",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = c( 'B', 'NK', 'Mono', 'CD8_T' , 'CD4_T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "id": "a8a5b4bd-9b5d-44e8-80b0-94638516a482",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "95a0ce76-29bd-4712-abf3-66878c145608",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = c('Franke_v2', 'Franke_v3', 'wijst', 'oneK1K', 'multiome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "id": "a6fef1d6-7d1d-41eb-bfda-ced502b0c3e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(j in datasets_to_load){\n",
    "    for(i in cell_types){\n",
    "        gene_list_input = fread(paste0(data_path, coeqtl_results_path , '/correlation_input/', 'F11_Decision_Tree_Geneswg3_',j,'_' , i, '.Qced.Normalized.SCs.Rds.tsv'), header = FALSE)\n",
    "        gene_list_input$cell_type = i\n",
    "        gene_list_input$dataset = j\n",
    "        gene_list = rbind(gene_list, gene_list_input)\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "id": "c47e733d-7ff7-4f27-8a76-880cc585b823",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>cell_type</th><th scope=col>dataset</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A2M    </td><td>CD4_T</td><td>multiome</td></tr>\n",
       "\t<tr><td>A2M-AS1</td><td>CD4_T</td><td>multiome</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " V1 & cell\\_type & dataset\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t A2M     & CD4\\_T & multiome\\\\\n",
       "\t A2M-AS1 & CD4\\_T & multiome\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 3\n",
       "\n",
       "| V1 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| A2M     | CD4_T | multiome |\n",
       "| A2M-AS1 | CD4_T | multiome |\n",
       "\n"
      ],
      "text/plain": [
       "  V1      cell_type dataset \n",
       "1 A2M     CD4_T     multiome\n",
       "2 A2M-AS1 CD4_T     multiome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_list_input,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "id": "0a7a6b5b-e701-4ce9-9125-494b658fa3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = data.frame(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "7dd89540-3172-4225-8860-033b88adc8af",
   "metadata": {},
   "outputs": [],
   "source": [
    "### unique gene list per cell-type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "id": "c06993e7-d1e0-4791-b7f4-400638d15d01",
   "metadata": {},
   "outputs": [],
   "source": [
    " gene_list= unique( gene_list[,c('cell_type', 'V1')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "id": "3b742942-c9a4-4a02-9d93-a24c98d6f82a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>cell_type</th><th scope=col>V1</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B</td><td>x         </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>B</td><td>AC004687.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & cell\\_type & V1\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B & x         \\\\\n",
       "\t2 & B & AC004687.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | cell_type &lt;chr&gt; | V1 &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | B | x          |\n",
       "| 2 | B | AC004687.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type V1        \n",
       "1 B         x         \n",
       "2 B         AC004687.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head( gene_list,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "id": "e41f30df-2d1c-43b6-9c01-59b05d67226d",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames( gene_list) = c('cell_type', 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "id": "2060ab85-fab6-4907-a6bb-c20feaba0da1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'All'"
      ],
      "text/latex": [
       "'All'"
      ],
      "text/markdown": [
       "'All'"
      ],
      "text/plain": [
       "[1] \"All\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "id": "eee7daa2-1de3-49bd-91b0-34c9656d367e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(cell_type != 'All'){\n",
    "    gene_list = gene_list[gene_list$cell_type == cell_type,]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "id": "cc9e919b-d74d-43b0-a38a-e9a0411e0f8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B'</li><li>'NK'</li><li>'Mono'</li><li>'CD8_T'</li><li>'CD4_T'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B'\n",
       "\\item 'NK'\n",
       "\\item 'Mono'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'CD4\\_T'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B'\n",
       "2. 'NK'\n",
       "3. 'Mono'\n",
       "4. 'CD8_T'\n",
       "5. 'CD4_T'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"B\"     \"NK\"    \"Mono\"  \"CD8_T\" \"CD4_T\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(gene_list$cell_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5acbea85-047b-4dfb-9bcf-2703225994bd",
   "metadata": {},
   "source": [
    "# Analyze"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b514f10d-76dd-4090-a6ce-6c685b77232d",
   "metadata": {},
   "source": [
    "## Define sign and non sign coegenes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "id": "56187468-72f8-44c4-bba4-87af5b6cd6cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CD4_T'</li><li>'CD8_T'</li><li>'Mono'</li><li>'NK'</li><li>'B'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CD4\\_T'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'Mono'\n",
       "\\item 'NK'\n",
       "\\item 'B'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CD4_T'\n",
       "2. 'CD8_T'\n",
       "3. 'Mono'\n",
       "4. 'NK'\n",
       "5. 'B'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"CD4_T\" \"CD8_T\" \"Mono\"  \"NK\"    \"B\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(coeqtls$cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "id": "0179ed23-6d6f-44f1-a96f-82148eba2de0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define list of significant eGenes per cell-type only on eGene level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "b60abc07-523c-41e7-9532-d4865052baed",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Any egene that is significant coeqtl in the cell-type or in any cell-type in the coeqtl data \n",
    "sign_coegenes = data.frame(coeGene = unique(coeqtls[coeqtls$type == 'unique',c('coeGene')]), sign = 1)\n",
    "### All the eGenes (that passed the decision tree filter) that are not coeqtls as background\n",
    "non_sign_coegenes = data.frame(coeGene = unique(gene_list$gene[(!gene_list$gene %in% coeqtls$coeGene) ]), sign = 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "a58a3daa-e4c9-49fa-890e-c997751869e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coeGene</th><th scope=col>sign</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>BACH2</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CCL5 </td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & coeGene & sign\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & BACH2 & 1\\\\\n",
       "\t2 & CCL5  & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | coeGene &lt;chr&gt; | sign &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1 | BACH2 | 1 |\n",
       "| 2 | CCL5  | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  coeGene sign\n",
       "1 BACH2   1   \n",
       "2 CCL5    1   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sign_coegenes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "id": "d51d321c-b585-4b9d-b8a0-94a35552904f",
   "metadata": {},
   "outputs": [],
   "source": [
    "coegene_data = rbind(sign_coegenes, non_sign_coegenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "id": "11c0aa23-aa58-450d-b093-4a49ddb907c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coeGene</th><th scope=col>sign</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>BACH2</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CCL5 </td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & coeGene & sign\\\\\n",
       "  & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & BACH2 & 1\\\\\n",
       "\t2 & CCL5  & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | coeGene &lt;chr&gt; | sign &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| 1 | BACH2 | 1 |\n",
       "| 2 | CCL5  | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  coeGene sign\n",
       "1 BACH2   1   \n",
       "2 CCL5    1   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coegene_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "id": "d7c86b5d-4598-4e86-8da9-b1f5b91f9b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Overview coeGenes with coeqtl effect vs. coeGenes without coeqtl effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "id": "4c607be8-ed45-4057-95df-af9e51e2e5e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sign</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>9810</td></tr>\n",
       "\t<tr><td>1</td><td>3870</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " sign & n\\\\\n",
       " <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 & 9810\\\\\n",
       "\t 1 & 3870\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| sign &lt;dbl&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 | 9810 |\n",
       "| 1 | 3870 |\n",
       "\n"
      ],
      "text/plain": [
       "  sign n   \n",
       "1 0    9810\n",
       "2 1    3870"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coegene_data %>% group_by(sign) %>% summarise(n = length(unique(coeGene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "id": "ac4292ea-df89-4a8e-9004-c437e7dbb572",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Opt: remove ribosomal / mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "id": "f150e67a-e9e5-4086-b76a-e10baa5f721a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13680"
      ],
      "text/latex": [
       "13680"
      ],
      "text/markdown": [
       "13680"
      ],
      "text/plain": [
       "[1] 13680"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coegene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "id": "20c32307-079b-4a25-ad73-df51705ac251",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(rm_rib_mt_coegenes == TRUE){\n",
    "    coegene_data = coegene_data[!coegene_data$coeGene %in% rb_mt_genes,]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "id": "83c9fac2-739b-468e-b399-f43af7302a46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13680"
      ],
      "text/latex": [
       "13680"
      ],
      "text/markdown": [
       "13680"
      ],
      "text/plain": [
       "[1] 13680"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coegene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "id": "378fffdc-3fca-4e01-8197-62ebb2914943",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13680"
      ],
      "text/latex": [
       "13680"
      ],
      "text/markdown": [
       "13680"
      ],
      "text/plain": [
       "[1] 13680"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(gene_list$gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4de0588f-ee9c-4f22-afe8-dd094cc6e48e",
   "metadata": {},
   "source": [
    "## Add Paper Scores to the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "id": "616a202f-424c-48e1-bd7a-618590e29cf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merge the gene level paper scores to coegene level data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "id": "b109e781-873b-4506-9408-69ac4c9ddaac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13680"
      ],
      "text/latex": [
       "13680"
      ],
      "text/markdown": [
       "13680"
      ],
      "text/plain": [
       "[1] 13680"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coegene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "id": "772d712f-1073-43f8-841a-4f053acee9e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "coegene_data = merge(coegene_data, paper_scores, by.x = 'coeGene', by.y = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "id": "e053fe0d-7465-45fe-830c-df01b8a0cf3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "coegene_data  = merge(coegene_data , paper_scores_go, by.x = 'hgnc_id', by.y = 'gene', all.x = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "id": "14a1a4fc-cb64-4676-b643-6b76f63744c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "10522"
      ],
      "text/latex": [
       "10522"
      ],
      "text/markdown": [
       "10522"
      ],
      "text/plain": [
       "[1] 10522"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coegene_data )  # some eGenes are not in score data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "id": "f77d3754-4523-459f-a75d-081a0820b714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 26</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>hgnc_id</th><th scope=col>coeGene</th><th scope=col>sign</th><th scope=col>GeneSymbol</th><th scope=col>TSSD</th><th scope=col>length</th><th scope=col>CDS_length</th><th scope=col>pLI</th><th scope=col>LOEUF</th><th scope=col>ABC_count</th><th scope=col>⋯</th><th scope=col>connectedness</th><th scope=col>PPI_degree_decile</th><th scope=col>PPI_degree_quantile</th><th scope=col>PPI_degree_cat</th><th scope=col>TF</th><th scope=col>HI</th><th scope=col>hs</th><th scope=col>GO_BP_count</th><th scope=col>GO_BP_count_400_auc</th><th scope=col>GO_BP_count_400_auc_decile</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>HGNC:10003</td><td>RGS7</td><td>1</td><td>ENSG00000182901</td><td>6</td><td>589.738</td><td>1.488</td><td>7.7696e-01</td><td>0.368</td><td> 34</td><td>⋯</td><td>1</td><td>7</td><td>4</td><td>1</td><td>0</td><td>0</td><td>0.094488616</td><td>64</td><td>1</td><td>4</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>HGNC:10004</td><td>RGS9</td><td>1</td><td>ENSG00000108370</td><td>4</td><td>126.891</td><td>2.025</td><td>1.1251e-14</td><td>0.847</td><td>125</td><td>⋯</td><td>1</td><td>7</td><td>4</td><td>1</td><td>0</td><td>0</td><td>0.003674329</td><td>30</td><td>1</td><td>4</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 26\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & hgnc\\_id & coeGene & sign & GeneSymbol & TSSD & length & CDS\\_length & pLI & LOEUF & ABC\\_count & ⋯ & connectedness & PPI\\_degree\\_decile & PPI\\_degree\\_quantile & PPI\\_degree\\_cat & TF & HI & hs & GO\\_BP\\_count & GO\\_BP\\_count\\_400\\_auc & GO\\_BP\\_count\\_400\\_auc\\_decile\\\\\n",
       "  & <chr> & <chr> & <dbl> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & ⋯ & <int> & <int> & <int> & <int> & <int> & <int> & <dbl> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & HGNC:10003 & RGS7 & 1 & ENSG00000182901 & 6 & 589.738 & 1.488 & 7.7696e-01 & 0.368 &  34 & ⋯ & 1 & 7 & 4 & 1 & 0 & 0 & 0.094488616 & 64 & 1 & 4\\\\\n",
       "\t2 & HGNC:10004 & RGS9 & 1 & ENSG00000108370 & 4 & 126.891 & 2.025 & 1.1251e-14 & 0.847 & 125 & ⋯ & 1 & 7 & 4 & 1 & 0 & 0 & 0.003674329 & 30 & 1 & 4\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 26\n",
       "\n",
       "| <!--/--> | hgnc_id &lt;chr&gt; | coeGene &lt;chr&gt; | sign &lt;dbl&gt; | GeneSymbol &lt;chr&gt; | TSSD &lt;int&gt; | length &lt;dbl&gt; | CDS_length &lt;dbl&gt; | pLI &lt;dbl&gt; | LOEUF &lt;dbl&gt; | ABC_count &lt;int&gt; | ⋯ ⋯ | connectedness &lt;int&gt; | PPI_degree_decile &lt;int&gt; | PPI_degree_quantile &lt;int&gt; | PPI_degree_cat &lt;int&gt; | TF &lt;int&gt; | HI &lt;int&gt; | hs &lt;dbl&gt; | GO_BP_count &lt;int&gt; | GO_BP_count_400_auc &lt;int&gt; | GO_BP_count_400_auc_decile &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | HGNC:10003 | RGS7 | 1 | ENSG00000182901 | 6 | 589.738 | 1.488 | 7.7696e-01 | 0.368 |  34 | ⋯ | 1 | 7 | 4 | 1 | 0 | 0 | 0.094488616 | 64 | 1 | 4 |\n",
       "| 2 | HGNC:10004 | RGS9 | 1 | ENSG00000108370 | 4 | 126.891 | 2.025 | 1.1251e-14 | 0.847 | 125 | ⋯ | 1 | 7 | 4 | 1 | 0 | 0 | 0.003674329 | 30 | 1 | 4 |\n",
       "\n"
      ],
      "text/plain": [
       "  hgnc_id    coeGene sign GeneSymbol      TSSD length  CDS_length pLI       \n",
       "1 HGNC:10003 RGS7    1    ENSG00000182901 6    589.738 1.488      7.7696e-01\n",
       "2 HGNC:10004 RGS9    1    ENSG00000108370 4    126.891 2.025      1.1251e-14\n",
       "  LOEUF ABC_count ⋯ connectedness PPI_degree_decile PPI_degree_quantile\n",
       "1 0.368  34       ⋯ 1             7                 4                  \n",
       "2 0.847 125       ⋯ 1             7                 4                  \n",
       "  PPI_degree_cat TF HI hs          GO_BP_count GO_BP_count_400_auc\n",
       "1 1              0  0  0.094488616 64          1                  \n",
       "2 1              0  0  0.003674329 30          1                  \n",
       "  GO_BP_count_400_auc_decile\n",
       "1 4                         \n",
       "2 4                         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coegene_data ,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10bb7460-b021-4f62-99fc-c9489af58356",
   "metadata": {},
   "source": [
    "## Add mean expression to the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 516,
   "id": "420142be-c1c9-4ab5-9001-0f5168a305f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "10522"
      ],
      "text/latex": [
       "10522"
      ],
      "text/markdown": [
       "10522"
      ],
      "text/plain": [
       "[1] 10522"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coegene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 517,
   "id": "643183e0-c7a3-4de3-acbe-2c21f0fe234f",
   "metadata": {},
   "outputs": [],
   "source": [
    "coegene_data = merge(coegene_data, summary_stats_info_meta, by.x = 'coeGene', by.y = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 518,
   "id": "b769be9d-2d19-4059-8328-b8c27ea71276",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "10512"
      ],
      "text/latex": [
       "10512"
      ],
      "text/markdown": [
       "10512"
      ],
      "text/plain": [
       "[1] 10512"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coegene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 519,
   "id": "3f9674ed-3bfc-4da4-9064-b9c681b6ebea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 27</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coeGene</th><th scope=col>hgnc_id</th><th scope=col>sign</th><th scope=col>GeneSymbol</th><th scope=col>TSSD</th><th scope=col>length</th><th scope=col>CDS_length</th><th scope=col>pLI</th><th scope=col>LOEUF</th><th scope=col>ABC_count</th><th scope=col>⋯</th><th scope=col>PPI_degree_decile</th><th scope=col>PPI_degree_quantile</th><th scope=col>PPI_degree_cat</th><th scope=col>TF</th><th scope=col>HI</th><th scope=col>hs</th><th scope=col>GO_BP_count</th><th scope=col>GO_BP_count_400_auc</th><th scope=col>GO_BP_count_400_auc_decile</th><th scope=col>weighted_mean_expression</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG</td><td>HGNC:5</td><td>1</td><td>ENSG00000121410</td><td>30</td><td> 8.314</td><td>1.488</td><td>4.9917e-09</td><td>1.208</td><td> 7</td><td>⋯</td><td>3</td><td>2</td><td>1</td><td>0</td><td>0</td><td>0.01563334</td><td> 1</td><td>0</td><td>4</td><td>0.1303894</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2M </td><td>HGNC:7</td><td>1</td><td>ENSG00000175899</td><td> 7</td><td>48.565</td><td>4.425</td><td>4.5229e-11</td><td>0.544</td><td>89</td><td>⋯</td><td>9</td><td>5</td><td>1</td><td>0</td><td>0</td><td>0.00945141</td><td>90</td><td>3</td><td>9</td><td>0.1624012</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 27\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & coeGene & hgnc\\_id & sign & GeneSymbol & TSSD & length & CDS\\_length & pLI & LOEUF & ABC\\_count & ⋯ & PPI\\_degree\\_decile & PPI\\_degree\\_quantile & PPI\\_degree\\_cat & TF & HI & hs & GO\\_BP\\_count & GO\\_BP\\_count\\_400\\_auc & GO\\_BP\\_count\\_400\\_auc\\_decile & weighted\\_mean\\_expression\\\\\n",
       "  & <chr> & <chr> & <dbl> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & ⋯ & <int> & <int> & <int> & <int> & <int> & <dbl> & <int> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG & HGNC:5 & 1 & ENSG00000121410 & 30 &  8.314 & 1.488 & 4.9917e-09 & 1.208 &  7 & ⋯ & 3 & 2 & 1 & 0 & 0 & 0.01563334 &  1 & 0 & 4 & 0.1303894\\\\\n",
       "\t2 & A2M  & HGNC:7 & 1 & ENSG00000175899 &  7 & 48.565 & 4.425 & 4.5229e-11 & 0.544 & 89 & ⋯ & 9 & 5 & 1 & 0 & 0 & 0.00945141 & 90 & 3 & 9 & 0.1624012\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 27\n",
       "\n",
       "| <!--/--> | coeGene &lt;chr&gt; | hgnc_id &lt;chr&gt; | sign &lt;dbl&gt; | GeneSymbol &lt;chr&gt; | TSSD &lt;int&gt; | length &lt;dbl&gt; | CDS_length &lt;dbl&gt; | pLI &lt;dbl&gt; | LOEUF &lt;dbl&gt; | ABC_count &lt;int&gt; | ⋯ ⋯ | PPI_degree_decile &lt;int&gt; | PPI_degree_quantile &lt;int&gt; | PPI_degree_cat &lt;int&gt; | TF &lt;int&gt; | HI &lt;int&gt; | hs &lt;dbl&gt; | GO_BP_count &lt;int&gt; | GO_BP_count_400_auc &lt;int&gt; | GO_BP_count_400_auc_decile &lt;int&gt; | weighted_mean_expression &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG | HGNC:5 | 1 | ENSG00000121410 | 30 |  8.314 | 1.488 | 4.9917e-09 | 1.208 |  7 | ⋯ | 3 | 2 | 1 | 0 | 0 | 0.01563334 |  1 | 0 | 4 | 0.1303894 |\n",
       "| 2 | A2M  | HGNC:7 | 1 | ENSG00000175899 |  7 | 48.565 | 4.425 | 4.5229e-11 | 0.544 | 89 | ⋯ | 9 | 5 | 1 | 0 | 0 | 0.00945141 | 90 | 3 | 9 | 0.1624012 |\n",
       "\n"
      ],
      "text/plain": [
       "  coeGene hgnc_id sign GeneSymbol      TSSD length CDS_length pLI        LOEUF\n",
       "1 A1BG    HGNC:5  1    ENSG00000121410 30    8.314 1.488      4.9917e-09 1.208\n",
       "2 A2M     HGNC:7  1    ENSG00000175899  7   48.565 4.425      4.5229e-11 0.544\n",
       "  ABC_count ⋯ PPI_degree_decile PPI_degree_quantile PPI_degree_cat TF HI\n",
       "1  7        ⋯ 3                 2                   1              0  0 \n",
       "2 89        ⋯ 9                 5                   1              0  0 \n",
       "  hs         GO_BP_count GO_BP_count_400_auc GO_BP_count_400_auc_decile\n",
       "1 0.01563334  1          0                   4                         \n",
       "2 0.00945141 90          3                   9                         \n",
       "  weighted_mean_expression\n",
       "1 0.1303894               \n",
       "2 0.1624012               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coegene_data,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ec6c2f9-78f0-4cb0-9769-c7d1671b7db7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Compare gene-characteristics: with coeqtl effect vs. without coeqtl effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcceb09b-777a-4922-a760-74da031ba654",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Logistic regression models with multiple features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 520,
   "id": "5efbc6d1-1936-4d77-9f38-8ebd61a10536",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Scale features to use in model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 521,
   "id": "422e563c-ffb4-412f-bca8-4ae24d658aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_data = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 522,
   "id": "9888be14-b7ac-4b75-81f1-97f65942437d",
   "metadata": {},
   "outputs": [],
   "source": [
    "coegene_data$sign = as.character(coegene_data$sign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 523,
   "id": "0669a683-95bf-4fbc-ad96-fa6834633e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABC_count'</li><li>'ABC_length_per_type'</li><li>'CDS_length'</li><li>'coeGene'</li><li>'connect_decile'</li><li>'connect_quantile'</li><li>'connectedness'</li><li>'GeneSymbol'</li><li>'GO_BP_count'</li><li>'GO_BP_count_400_auc'</li><li>'GO_BP_count_400_auc_decile'</li><li>'hgnc_id'</li><li>'HI'</li><li>'hs'</li><li>'length'</li><li>'LOEUF'</li><li>'pLI'</li><li>'PPI_degree_cat'</li><li>'PPI_degree_decile'</li><li>'PPI_degree_quantile'</li><li>'promoter_count'</li><li>'Roadmap_count'</li><li>'Roadmap_length_per_type'</li><li>'sign'</li><li>'TF'</li><li>'TSSD'</li><li>'weighted_mean_expression'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ABC\\_count'\n",
       "\\item 'ABC\\_length\\_per\\_type'\n",
       "\\item 'CDS\\_length'\n",
       "\\item 'coeGene'\n",
       "\\item 'connect\\_decile'\n",
       "\\item 'connect\\_quantile'\n",
       "\\item 'connectedness'\n",
       "\\item 'GeneSymbol'\n",
       "\\item 'GO\\_BP\\_count'\n",
       "\\item 'GO\\_BP\\_count\\_400\\_auc'\n",
       "\\item 'GO\\_BP\\_count\\_400\\_auc\\_decile'\n",
       "\\item 'hgnc\\_id'\n",
       "\\item 'HI'\n",
       "\\item 'hs'\n",
       "\\item 'length'\n",
       "\\item 'LOEUF'\n",
       "\\item 'pLI'\n",
       "\\item 'PPI\\_degree\\_cat'\n",
       "\\item 'PPI\\_degree\\_decile'\n",
       "\\item 'PPI\\_degree\\_quantile'\n",
       "\\item 'promoter\\_count'\n",
       "\\item 'Roadmap\\_count'\n",
       "\\item 'Roadmap\\_length\\_per\\_type'\n",
       "\\item 'sign'\n",
       "\\item 'TF'\n",
       "\\item 'TSSD'\n",
       "\\item 'weighted\\_mean\\_expression'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ABC_count'\n",
       "2. 'ABC_length_per_type'\n",
       "3. 'CDS_length'\n",
       "4. 'coeGene'\n",
       "5. 'connect_decile'\n",
       "6. 'connect_quantile'\n",
       "7. 'connectedness'\n",
       "8. 'GeneSymbol'\n",
       "9. 'GO_BP_count'\n",
       "10. 'GO_BP_count_400_auc'\n",
       "11. 'GO_BP_count_400_auc_decile'\n",
       "12. 'hgnc_id'\n",
       "13. 'HI'\n",
       "14. 'hs'\n",
       "15. 'length'\n",
       "16. 'LOEUF'\n",
       "17. 'pLI'\n",
       "18. 'PPI_degree_cat'\n",
       "19. 'PPI_degree_decile'\n",
       "20. 'PPI_degree_quantile'\n",
       "21. 'promoter_count'\n",
       "22. 'Roadmap_count'\n",
       "23. 'Roadmap_length_per_type'\n",
       "24. 'sign'\n",
       "25. 'TF'\n",
       "26. 'TSSD'\n",
       "27. 'weighted_mean_expression'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"ABC_count\"                  \"ABC_length_per_type\"       \n",
       " [3] \"CDS_length\"                 \"coeGene\"                   \n",
       " [5] \"connect_decile\"             \"connect_quantile\"          \n",
       " [7] \"connectedness\"              \"GeneSymbol\"                \n",
       " [9] \"GO_BP_count\"                \"GO_BP_count_400_auc\"       \n",
       "[11] \"GO_BP_count_400_auc_decile\" \"hgnc_id\"                   \n",
       "[13] \"HI\"                         \"hs\"                        \n",
       "[15] \"length\"                     \"LOEUF\"                     \n",
       "[17] \"pLI\"                        \"PPI_degree_cat\"            \n",
       "[19] \"PPI_degree_decile\"          \"PPI_degree_quantile\"       \n",
       "[21] \"promoter_count\"             \"Roadmap_count\"             \n",
       "[23] \"Roadmap_length_per_type\"    \"sign\"                      \n",
       "[25] \"TF\"                         \"TSSD\"                      \n",
       "[27] \"weighted_mean_expression\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(colnames(coegene_data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "id": "af3141c7-4ddc-4afc-9bef-d795b9f6543c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the columns to scale as a character vector\n",
    "cols_to_scale <- c(\"weighted_mean_expression\" ,'CDS_length',  'connect_quantile','GO_BP_count_400_auc', 'hs', 'length', 'LOEUF', 'pLI',  'PPI_degree_quantile', 'promoter_count', 'Roadmap_count',\n",
    "                  'Roadmap_length_per_type','TSSD' )\n",
    "\n",
    "# Apply scaling to the selected columns only\n",
    "coegene_data_scaled = coegene_data\n",
    "coegene_data_scaled[cols_to_scale] <-lapply(coegene_data_scaled[cols_to_scale], function(x) as.numeric(scale(x)))\n",
    "                                            \n",
    "#coegene_data_scaled[cols_to_scale]  = lapply(coegene_data_scaled[cols_to_scale], function(x) (x - min(x, na.rm = TRUE)) / (max(x, na.rm = TRUE) - min(x, na.rm = TRUE)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24fbee07-3c95-45ef-a0ae-c1fd4167edea",
   "metadata": {},
   "source": [
    "### Residual Model for connectedness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "id": "0af958dd-75f1-40df-bef5-1e027d5a9703",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Estimate the model ( use residuals after correcting for expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 526,
   "id": "21f584dd-28f0-44ce-95bd-bfe315fd58c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_form = 'residuals ~ connect_quantile + PPI_degree_quantile'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 527,
   "id": "7fd6328c-3f75-483f-8030-ac06d5aff73f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"glm.fit: fitted probabilities numerically 0 or 1 occurred\"\n",
      "Waiting for profiling to be done...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "test  = coegene_data_scaled \n",
    "test$sign = as.numeric(test$sign)\n",
    "\n",
    "res = residuals(glm(sign ~ weighted_mean_expression, data = test, family = 'binomial'))   # select the beta of the random snp instead\n",
    "\n",
    "## Run model on residuals\n",
    "        \n",
    "test$residuals = res\n",
    "\n",
    "result_model = glm(as.formula(model_form), data = test, family = 'gaussian')\n",
    "\n",
    "#print(summary(result_model))\n",
    "## Extract model results\n",
    "confidence_int =confint(result_model)\n",
    "model_summary = summary(result_model)\n",
    "\n",
    "coef = cbind(as.data.frame(model_summary$coefficients), as.data.frame(confidence_int))\n",
    "coef$variable = rownames(coef)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 528,
   "id": "cc97cf45-4695-4a86-a1d2-bba65cb0a153",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Estimate</th><th scope=col>Std. Error</th><th scope=col>t value</th><th scope=col>Pr(&gt;|t|)</th><th scope=col>2.5 %</th><th scope=col>97.5 %</th><th scope=col>variable</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>(Intercept)</th><td>-0.11860047</td><td>0.009077136</td><td>-13.065847</td><td>1.032952e-38</td><td>-0.136391333</td><td>-0.10080961</td><td>(Intercept)        </td></tr>\n",
       "\t<tr><th scope=row>connect_quantile</th><td> 0.05980493</td><td>0.009093687</td><td>  6.576533</td><td>5.044212e-11</td><td> 0.041981633</td><td> 0.07762823</td><td>connect_quantile   </td></tr>\n",
       "\t<tr><th scope=row>PPI_degree_quantile</th><td> 0.02046636</td><td>0.009093687</td><td>  2.250612</td><td>2.443070e-02</td><td> 0.002643064</td><td> 0.03828966</td><td>PPI_degree_quantile</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Estimate & Std. Error & t value & Pr(>\\textbar{}t\\textbar{}) & 2.5 \\% & 97.5 \\% & variable\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t(Intercept) & -0.11860047 & 0.009077136 & -13.065847 & 1.032952e-38 & -0.136391333 & -0.10080961 & (Intercept)        \\\\\n",
       "\tconnect\\_quantile &  0.05980493 & 0.009093687 &   6.576533 & 5.044212e-11 &  0.041981633 &  0.07762823 & connect\\_quantile   \\\\\n",
       "\tPPI\\_degree\\_quantile &  0.02046636 & 0.009093687 &   2.250612 & 2.443070e-02 &  0.002643064 &  0.03828966 & PPI\\_degree\\_quantile\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 7\n",
       "\n",
       "| <!--/--> | Estimate &lt;dbl&gt; | Std. Error &lt;dbl&gt; | t value &lt;dbl&gt; | Pr(&gt;|t|) &lt;dbl&gt; | 2.5 % &lt;dbl&gt; | 97.5 % &lt;dbl&gt; | variable &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| (Intercept) | -0.11860047 | 0.009077136 | -13.065847 | 1.032952e-38 | -0.136391333 | -0.10080961 | (Intercept)         |\n",
       "| connect_quantile |  0.05980493 | 0.009093687 |   6.576533 | 5.044212e-11 |  0.041981633 |  0.07762823 | connect_quantile    |\n",
       "| PPI_degree_quantile |  0.02046636 | 0.009093687 |   2.250612 | 2.443070e-02 |  0.002643064 |  0.03828966 | PPI_degree_quantile |\n",
       "\n"
      ],
      "text/plain": [
       "                    Estimate    Std. Error  t value    Pr(>|t|)    \n",
       "(Intercept)         -0.11860047 0.009077136 -13.065847 1.032952e-38\n",
       "connect_quantile     0.05980493 0.009093687   6.576533 5.044212e-11\n",
       "PPI_degree_quantile  0.02046636 0.009093687   2.250612 2.443070e-02\n",
       "                    2.5 %        97.5 %      variable           \n",
       "(Intercept)         -0.136391333 -0.10080961 (Intercept)        \n",
       "connect_quantile     0.041981633  0.07762823 connect_quantile   \n",
       "PPI_degree_quantile  0.002643064  0.03828966 PPI_degree_quantile"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 529,
   "id": "e22c7238-e291-4770-a714-4b33e3c7c6d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef$sign = 0\n",
    "coef$sign[coef$`Pr(>|t|)` < 0.05] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "id": "dd4134ab-2728-419b-aa5c-8c7a59e69264",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef = coef[coef$variable != '(Intercept)',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 531,
   "id": "0aca8dd2-f7bb-4a7f-8b62-372d11e5d403",
   "metadata": {},
   "outputs": [],
   "source": [
    "###  adjust label\n",
    "coef$variable_label = coef$variable\n",
    "coef$variable_label = str_replace(coef$variable_label, 'connect_quantile', 'Connectedness in co-expression networks')\n",
    "coef$variable_label = str_replace(coef$variable_label, 'PPI_degree_quantile', 'Connectedness in InWeb PPI networks')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 532,
   "id": "62798e9b-8b34-4b77-8505-4f5e70d403f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'All'"
      ],
      "text/latex": [
       "'All'"
      ],
      "text/markdown": [
       "'All'"
      ],
      "text/plain": [
       "[1] \"All\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "id": "8586bb7c-411f-4d70-a8b1-4e9b679a63a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLAAAAJYCAIAAAD9hIhNAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nO3dd5xU5d3w4XtYyiIuS28iKFKUrkhREGk2xBawYpQoMSYqMfbHaBTTfBQ1MSav\nIiIqPhZQUMEW0MVgRCxUFaWIdKQqHXaZ9w9QWHZ2AQUGvK/rj3x2zsw587vPHif73dldEslk\nMgAAABCfYukeAAAAgPQQhAAAAJEShAAAAJEShAAAAJEShAAAAJEShAAAAJEShAAAAJEShADA\nLhpxcWa3weu3/wCAA5sgBAAKsezxU0olal07dnO6BwFgLxGEAEBqcwcPnNK1x2HPDRy1Md2j\nALB3CEIAIKUZTzz+1YU3PXL5sa8MfHl1uocBYK8QhABACsmPBw5af8kvjqvQo1eXUQOfX5Hu\neQDYGwQhAFDQ5ncef6pcr16NQyhzxqXdPhz49Px0TwTAXlA83QMAAPufDW8M/L95Hy2vl/if\nrRsmP/HF1bemdSQA9gLvEAIAO1r10uPDjrx/XnKr3DHXLHp84MR0TwXAHicIAYAdLH/+8ZEt\nL77wkO9uZ5xwac+8Jwe+45+fAPipSSSTyXTPAAAAQBp4hxAAACBSghAAACBSghAAACBSghAA\nACBSghAAACBSghAAACBSghAAACBSghAAACBSghAAACBSghAAACBSgpB95+233073CAAAwDaC\nkH3nvffeS/cIAADANoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUsXTPQDAAS8ZwsoNG0MIZUoUL1nMN9pIYeXGTclksnTxjMyM\njHTPAgDb+MIF4Mdal5tX4anhFZ4a/vJXC9I9C/up+s+/WuGp4f/8dEa6BwGAfAQhAABApAQh\nAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQh\nAABApAQhAABApAQhAABApAQhAABApA70IBzao/gFw9M9BHvU0B6JHkPTPQQAAMRgZ0G4afbL\nfziv1eEVSpcuV7t51+sHT12110cafkHxs5/N3etPszMjLs7sNnh9uqfYk/b2ivaTTxwAALCr\nig7CZcN+0a7XG9WvfW7CgqVz3h18be03+7+ybB9Nlm7dBq8fcXFmuqfYk356KwIAAH6UIoNw\n+qN9h9T50wt/v6hV7fJlytZsfHKfJx+8qOLmuSNvOa1htaysyvU7XTNkxsYQQghDeyROveWB\n85rVyM4+pM1Vw+cVtnHpf++/sNXhFcpkVWt02q2vzd/yPEveve+i1nUqli6WSCQSiXOey3vp\nwhKJROLUQatDql3yZr/Yp/1hFbKrN+neb/zqrbPu2nNtnPpIz2Nrli1ZLJFIHHbDhzvczLf4\n799PS7m0bbYf/uzB60NIeX622WGq3GkPtDv01IFfJUPY+PGdxx5+wdCvty7nlBvv+Vnj6tkV\nDm//2+FzNoehPRKn3tSvR7OaWY1vn5hidQXXsuOW71dUcMKi15j63vwDzBzULd8nbukjJ9W4\n6u0QQhh/U+1E7Zs/CCEkZ71898n9lxR2/Wy/uhBCyJ0z9PJjjr78+a/WFvFpAgAAfoRk4XKf\nOzfR8A+f7rB14cCTs5tc++qsb76dN/p/WpVtfe/MZDKZHNI9HNT29rfnrlozf+Sv6hz6u/dT\nblw2pHvTHgM+Xrh2w9qF4+/pdMh5L65JJpc8e3bFupc8NeHrtXnJZDKZHHZ+xlnPbPru6Qru\nMvOxk8q3uHnUnG+/+XLENU0yMs4ftsvPtfmVS7M73vv5yo2bk8lkcseb+b3Ss9TpT60rbGnf\n2XH4ws5P4ctZk5z75FmHtv7Lf9+4pn79X7/1zZbHDekeije/9rWvvv32qzdvaF6+65NfD+ke\nSrW6ZfSc1bkpj/PCczuupcDqvltRigmLXGPKe1MsJP8nbvpfmzX4/ZRk8ou/tGx9XKuWf52e\nTE7uUKn63WMLu36+X92Q7qH7kFUTHjit0Yl/GLN8J58m2C+s2ZQbHn0uPPrckFlz0z0L+6nK\nTw0Pjz7Xb/K0dA8CAPkUFYSbnu1RMAhzX7yo5EmPbM2WjUN7lDh54LfJZHJI97Aln5LJ5IiL\nS5/5fxtTbBw89Of5f2KxwR1TkrnDLy7V6V/Ltz1Dvq7Ie2nHXW7re2GpM57YetRlj3TcFoQ7\nf67ktH4tqzY8+6o7Hhoy9qs1BW7ms30QFlzad2djx+FTnJ9bb22xZYK2D8wtuJw7piSTyeWv\nXFo7o3SzP3y47rvDDOkeOj6ybMvHqx87rfQlLz/fPWw51alPyx1DC6xlx9VtXVGqz2BRa0x5\nBlItJH8QJt//Xa3WD8xd9q/Orfrl3Nuyy/9bNuf+muXajSv0+vl+dUO6h3otWtQ5/f99uj6Z\nciGw3xGEO/hwyfL/GT/5qnc/GvTFl5vyfDMnmRSE7M82LrhkxJQpP43/Ujct/MWIyRN+GmuB\nfaWoHxktfkyLpjNycubv+JZi/huJxNaPMzO3JkJGRrG8vLwUG3Nzk6HDQ0u2Pfu0OxvnP0YI\nISS2v5VMptol9TS78FwNrh8zvv/59TZNfvLKlk37vF0r/811hZ2JlEsrcAIKOT/1bvlwywRj\nr62Zejkb5378SV7lcqvmLV6T7zzseMisrKzCT0v3AmtpUMjqUn8GC19jqntTLSSR/0y06Nxp\nxuiRw0Yv7tj5xM4dF44eNnL08iPqHFPo9bNtdSHUbNWp4hcjc2ZtCCEUvhBgv/TgJ9NbvjTq\nr5M++9enM3qNGd/m5VHrcnd8TQH2HzM+/XRJ3XqNEzt/ZJG+vvKp5xMDnk8MeD4xaPzrBe5e\nNH/KyUNeKvvUq13eX7jw+615K2576cWSr3yxaOfHX3r1088nBjyfeOGTaUVsLF7tuuqL75q1\n6UcuBqJSvKg76/W+/WcPXtHjuvL/uPaUhuXWzhr3wr0jKt10UofxD9zx2kl3nVDqo7vv+/cx\n3e/JKuoY28toe2bXq2+65rFW9194TPWDMrZubHfGab1vv/6Z4+45p0mlzGIhZGdnz/9y5oZk\ng1KJLbv0zrdLxvFd2vd+8M7RHW9rtfmdO//5n9Dwt7v8XCGE0rXaXvDbNt3Oabi8xd8nLAkd\nd7y5q2spZPiQcfxJHS4r9PwUXE4Iq3NuPH9wi8emDJrU7bReg//78sU1t7wkv/OvP75+yl3t\nwvt9H3q/43VPJl4KRR+n4NJ22HLy1n1TTThi99adcoB8n7gQMtp3bnfJbX+sfOqAZqH5kraT\nf3n30sNaX1Pi+NpFnJ/vVOh0z7Ced5x+Zre1Q1+6vtlBKZYG+6fF6zbMWrVm54/76Zq3Zu31\n4yaFZAhhy/+Ej5auuPmDydc2rp/WudIvL7nzx0A6rBo8I+Os00vtgSMVO3RY7+POTnlX7sIb\ncxY063jKkPKr737jvzd+1XVw7eIh5I7+4NOMuodVmbUrR6/0UM/zHvp2WvN/5xW9sfFh1aaN\nn7vyiDrlftRiICJFBmGo1P2JscX63vTn7s16LU5WaXBcjxvu/UWjhic8/fkvr29TY87m6q3O\nHzC4T51df7ZK5z46YvF1N1zUrM+XS9bmhTKXvrZ60KmVzu8/fM7VvzuzXq+5Kzcmz3pq5S03\nlL2odfbvv+kwcNXrvVLs8lj/ARN/fnnTQzYc0uX6rl3CnF1/rmMf7lL516NDokRWtaM6/e7B\nM1/tktju5mU/IDN2HH7d8IsvHVDE+Skw1RODc/+Qc87Q90+ulHnSs3fknHDhfce+fcORxUMI\nnTpXf6RrvfPnZza7ZOBTPSt/8FKRx3mq37yf51/L0vyLvazWqo+37Fut4IQv7/7CC57eR3pt\n/4k7OGR16nzknDcqN6o07bNplRsd+c3grxvV3Jzq2VPKbtv3jWfuObP7KWv+fsY73W7+cZ8m\n2Feu/u9H6R5hf/SPT6b/45Pp6Z4CSGXTsnEby/XNDGHdjHavLm9fceXjX60pXbnOwx2bnVx6\n+8ct7v3kmMfy/aW8RM+Tzh1ce7sNyQWXPvFCKJF1cuNj+jetVH77xy5d9GZ2nck1MrND5k1H\nlqs/b1moXXXJ7EmPZTV9qvKsx7YFYe4Hn3509eQFUzdkHFWr0cD2RzTNCLutfPkmSxaMD3VO\n3v1dIU6JZNI3LfdDQ3sknr0gObRHuuf4kTZu3Dhx4sTvb7788st/+tOf0jgP7CVrc/PKDHoh\n3VNwAOjXutn1TRqkewrYzurPW76R+1T3Rkeum9HumZmtTjnhj1UT4z4ee+mmpnPbVt39HyPd\nvHT53Gv/PSm7/Rn/rL7d3rPHFZ9RY1OXWokQwlfvZ86ovqbN5t4Tiv21Xa1qX0+q+X7pD8+o\nXy2E8PWkBuMS/Ts3bFNq09gP/3PHwW3HNi6T7/DfTmv+77xnuzc6sqiNX//qqS9O7Nnuop39\nY9vAFkW/Qwg/SsmSJVu1avX9zVGjRqVxGNgH/rdVs841qqR7inSav3bdOaPeTW7O973Gy+of\n/puGddM20/7hpFfHrNi44z9EBOlXqlTFjeu+2fJxVo1ehxxUJoTO9Q8pO/bbJaHqdi9nu/AO\nYQghFKtUofZVh025cdn6UH27dxgzS1Vcv+HrEKqGsGLdhoNLlfpk2oRB074dNG3clvurD9/8\n5dlHHrxs5RdfL+7wzNbfByxb/9uJH75z9MRVIZS/9/yTbtjF31FKbloZSlVSg7DLBOH+qcfQ\n5IH+7iDEqE5WmRb5f04qNi1C+f9t2fSm9yclQkgkwuZkaFoh+x/HH3NQ8R/wg18/KcWL/dg/\n2QF7RYmKbUrOnrQhtA4hrFr45IK6d1VJjJ++YFX5JpXzPa7qgEvOG7Dzw21eumLuP2dvbnZC\n/j9GXqnqSSun9ltw6O3lV98zbcUpzSs2PezU5Ja/wr7dO4TJiuUaVKvYv0OD4w4uUWLrntWT\nx+7milasmFK5YqudPw7YShACsCfd0KRB26qVhn45d8WGTa2rVPhF/cNLFvO9ethvZfWsm/fb\n2RuuqBVCdrmSX/zniH+vyaxU57FOu/nzoosmVBsxfXEoVr5M9smNj/9TjS17L+795CfNz+10\ndeka95y49JKxrx+yoUTL+i0H10799WeiSqOnj/i4z8iRH6/euD4ZOrc9a9RR3/+1m/V/e+Xl\n3y0OIYSjBnzSts0ZYxuXTrnxk9mLGtRt7y/KwK4ThADsYcdVqXhclYrpngLYJfUaNqz85vQp\nh2aGxMGXdWj1lx92lGpHL+p9dIGtVQdcUnXLRzUObTrq0KYpdqzSbN4Z398o3uKoVu8elfIJ\nMq8947xrd7oxd9EDC6v8oXnJ3RkdYicIAQAiVqL6E6dXD+tmpHuOPaF4tQGnV0v3EHCAEYQA\nANErXXfsz9I9A5AOfq8DAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAgUoIQAAAg\nUoIQAAAgUsXTPQDAAa908YyZ558eQqhaulS6Z2E/9cHZJ+UlkxVLlUz3IACQjyAE+LESIdTJ\nKpPuKdiv1T74oHSPAAAp+JFRAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlCAACASAlC\nAACASAlCAACASAlCAACASAlCAACASBVP9wCw7yRD+HLVmhBClcxSB5dw8ZPCV6vX5iWTFUqV\nLFeyRLpnAQDY67xDSEQ2bd58xHMjj3hu5Ig5C9I9C/up414edcRzIx/5bGa6BwEA2BcEIQAA\nQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAA\nQKQEIQAAQKQEIQAAQKQEIQAAQKQEIQAAQKQO9CAc2qP4BcPTPcQeNbRHosfQdA/x4/wElgAA\nAFHYWRBumv3yH85rdXiF0qXL1W7e9frBU1ft9ZGGX1D87Gdz9/rT7MyIizO7DV6/04flXFnp\nyDunpr5v4m1HVvn16C0f547olZU4qu8nW26tHtSt1GmPrdj9qXKurJRIJBKJEgdXbdC5z5AZ\neam3FDrSzuwnJx8AANgXig7CZcN+0a7XG9WvfW7CgqVz3h18be03+7+ybB9Nlm7dBq8fcXHm\njzpEsw4d8saM+TSEEMKHOe81al16TM7iEELIG5sztlHHjuV/0FEb3DElmVy3cOyfa4245JrB\ny1JuAQAA2Lkig3D6o32H1PnTC3+/qFXt8mXK1mx8cp8nH7yo4ua5I285rWG1rKzK9TtdM2TG\nxhBCCEN7JE695YHzmtXIzj6kzVXD5xW2cel/77+w1eEVymRVa3Tara/N3/I8S96976LWdSqW\nLpZIJBKJc57Le+nCEolE4tRBq0OqXfJmv9in/WEVsqs36d5v/Oqts+7ac22c+kjPY2uWLVks\nkUgcdsOHO9zMt/jv3yFMubSCUjws0bZj+1ljxnwdQvgqZ0zli287ZVFOzoYQwoScnLIdOtRJ\nfTZCWPX+PT9rXD27wuHtfzt8zuaUz1Y8q16PiztnfvrprMK37Gy8As++flC3bSe/330n1bjq\n7RBCGH9T7UTtmz8IISTf+k31k/svKewCuKlfj2Y1sxrfPnHrc+bOGXr5MUdf/vxXa7ee59t+\n//sC5xkAAEibooIwb8LHk+t27Fgz/9ZFT/Tu+eqR9703f8HEh9uM733xg98FyH/GrvjNyC8W\nfvZo89f73D8+5cblQ6/89XtdHnlv/oqvJw7q9P4vrxu2NoSlz11x1sMlrnlx2vLcZDKZTA47\nP+OsZzYlk8nXex0cQsFdZj1xRe//Hv/YpHmfv3LZnNdHbZts58+VfLPfzQsveGvumrxkMjm7\nX4v8N48t9EykXNouPKx0hw4tx+eMWR++fTtn7Yknnta+5cSccSF8lZOzvGPHFinPRgghvDVq\n4RWvfjFvYv/W71z266eXpPzcrJ4xdPDodUcdVafwLUWPl+LZM3uN2Hbybzirc5XRo6eGMH10\nTvXjqr09ekYIU0ePrta50YhCLoCcMct+M+LzlVP/2DyEEFZP/NuZXR+q+be3Hjuv1ltbz/Mf\n//znIs8zALtm08JLR06dmkz3GLsud9FlI6dMPIAGBohGUUGYTKZ45c57b1ROq6v7nnZ42axD\nOvW96eSP/z1m668Vdrzy1g41Dz6oRtczjl86c+amFBunv/3yyMlDex9T/aBSB1VvddNb8ydN\nmhXy3h3xWrPr/nZx88qlUwyzeeyOu3w8eNQ77frc2fnQrLKHnX7n1e23PXbnz5Wo16T+p4/f\n/Ps//2vou3PW7nizcCmXtisPq9qx4+FjxozLfSfn83YdGme0O7He2JxPVr+dM/GEDickCixt\n0ta0OuHXt59aKyur1kl3XtPm7VHjdvgsfN63SSJRqupx/zP7tEEP/rxiyi07Hy/l+cmnbqdO\n60aNmrd81OjEuX89N4wetXzuqFEbOnf4urAL4MQrb+10aJmMEEIIk+/u0H3UmUPe6Nu+fNh2\nnt+f+lWR55l97a0FX1859qML3nqv35TP1+bmpXscYFfN+PTTJXXrNU7sgUMtmj/l5CEvlX3q\n1S7vL1y403s3rx818b3mT7+Y9fTrXT9ctHj7h+atuO2lF0u+8sWilFuKV7uu+uK7ZhX6/6AA\npEtRQVj8mBZNZ+TkzM+/NV8lJpMh8d3/H2Vmbv2Nu4yMYnl5eSk25uYmQ4eHliS/N+3OxvmP\nEUIIie1vJZOpdkk9zS48V4Prx4zvf369TZOfvLJl0z5v18p/c11hZyLl0nbpYU06dMgdM+bp\nnP8efeLxiZDVvn1mzlsv5Ixt3LFjucKXtu0EpEryBndMSSZz1yyZ/vY/L6ifkXrLzsdLfX7y\nnfwWnTvNGD1y2OjFHTuf2LnjwtHDRo6e2anTMYVeAFlZWd9vr9mqU8UvRubM2rBl5K3nedGE\nl4o8z+xTf5n4WedXc/p/PnPol/NufH9S0xdf/2ajr9XggLBq8IyMs2qX2gNHyl14Y86CZm1P\nmdujVcuFH9z4Ve5O7l0+5//yDn3iZ2fO69Ywe/r4v26Lv9zRH3yaUfewKqHQLY0PqzZt+tyV\ne2BoAPak4kXdWa/37T978Ioe15X/x7WnNCy3dta4F+4dUemmkzqMf+CO106664RSH91937+P\n6X5PVlHH2F5G2zO7Xn3TNY+1uv/CY6oftLVcMtqdcVrv269/5rh7zmlSKbNYCNnZ2fO/nLkh\n2aBUYssuvfPtknF8l/a9H7xzdMfbWm1+585//ic0/O0uP1cIoXStthf8tk23cxoub/H3CUtC\nxx1v7upadlHi+I4nzvj9n9dVuvq6MiGEuu1PmNPrzzPKnXn1YSHj4B2X9p13/vXH10+5q114\nv+9D73e87sk98R3gAgo5P9uf/Iz2ndtdctsfK586oFlovqTt5F/evfSEPz9b4vjFHS7b6QVQ\nodM9w3recfqZ3dYOfen6ZgdtPe3T5s/45PFxe+M8764py7+plLl454/76Vq0dv3tH00NISST\nIS8kQwizvl3b+z8f/urIQn/mOBIb8lL/3i7sRzYtG7exXN8t3+NbN6Pdq8vbV1z5+FdrSleu\n83DHZieX3v6hi3s/OeaxjdtvSfQ86dzBtb+7tXTRm9l1JtfIzA6ZNx1ZrvCrDgkAABDOSURB\nVP68ZaF21W2PLXhv2/oDK4UQQihdqflBJdZ916RLZk96LKvpU5VnPTar0C2hfPkmSxaMD3VO\n3nNnAoAfr8ggDJW6PzG2WN+b/ty9Wa/FySoNjutxw72/aNTwhKc//+X1bWrM2Vy91fkDBvfZ\nja8fK5376IjF191wUbM+Xy5ZmxfKXPra6kGnVjq///A5V//uzHq95q7cmDzrqZW33FD2otbZ\nv/+mw8BVr/dKsctj/QdM/PnlTQ/ZcEiX67t2CXN2/bmOfbhL5V+PDokSWdWO6vS7B898tUti\nu5uX7Y1KyezQsfWMF+d3P7FGCCGElie2mv/Aq+d2OLqwsxFCCJ06V3+ka73z52c2u2TgUz0r\n74WpCn3243ptf/I7dT7yq9drdm6bCIl2nZst6j+/U6eDQ7VLB+zSBZDdtu8bT/71jAtO+uo3\nbSbfNmBCSBTfXKx05xue3ivneTf9ZdJnf5n0Wbqn2L8kQ3Lol3OHfjk33YMAO7Nhw7KSJbK/\nv/nNivVtTpjRLjHu47GXfrx4btuq230fseqAS84bUMSh1m9Ylllxy5t45UuXWrV+w+btf3ao\n0Htzx0/4YGS1FiPKhxBCcs3sm+ZVvr9dVsbXW/cruCWEEBIlyoUNSzcf+P8EMsBPSyLlLwrC\nHpGbmzt9+vTvbz7zzDN33XVXGufZuHlzqYFD0zgAB4q7Wza9udmR6Z4CCrFp9qlDV/a9sHnr\nEMK6Ge1GrPvXuU2ahhBWftpwbImcbvW2+7nNnb1DuGhC1Q8PntytXtUQVkx7p97SBkvbbfcO\nYep7178+bty9GQ2HtKxSIYQQwuSPXm824dtte1Vq+tqhs0/Lv+XLs488LDn//MELLv95S+8Q\nAuxXin6HEH6U4sWLH3XUUd/f/P63GNPu722O7la7RrqnSKfF69Z3GPn2ps3J7b8l1Kdx/d82\nqpfGqfYHx7086ut1G9I9BRSpRMU2JWdP2hBab/mJzVULn1xQ964qifHTF6wq3yT/j5Xs7B3C\nSlVPWjm134JDby+/+p5pK05pXnEn9+atemzM+0MqHfNS0woHf/eopi1OTbYIIYTw9aSa75f+\n8Iz61cKRBbaEsGLFlMoVW+2B9QOwJwlCYlSldKk6WWXSPUU61ckq8//atvjV2I9yvwvC46pU\n/N+WTTIzCv+zRHHISOyV39qFPSqrZ928387ecEWDUiGEkF2u5Bf/OeLfazIr1XmsU9Xdu4KL\n17jnxKWXjH39kA0lWtZvObj2lq8KFvd+8pPm53a6unSBexd8cf2s5d/MGpU1PoQQzurQfXjd\nXX3R+GT2ogZ125fbrfEA2PsEIUTqsvqHH1el4rDZ85dv2NiqcoUeh9cspoXgAFGvYcPKb06f\nUr9xkxBC4uDLOrT6yw89VI1Dm446tGn+bVUHXFI19b2HtFjZu0Whx6rSbN4ZhWzJXfTAwip/\naF7yh44JwN4iCCFeR5Ure1TzsumeAth9Jao/cXr1dA+xO4pXG3B6tXQPAUAKghAADlil6479\nWbpnAOBA5m8/AwAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAA\nREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAAREoQAgAA\nREoQAgAARKp4ugeAfadksWLLf352CKFMCVc+qX3W47TNyWTp4hnpHgQAYF/wZTFxKV+qZLpH\nYL+WXbJEukcAANh3/MgoAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApBLJZDLdMxCLBx54YOPGjXvp4MuWLatYseJeOjjszzZt2rRu\n3bqyZcumexBIAy/+RGvdunXJZPKggw5K9yAcAMqVK/erX/2qsHsFIT8RAwcOvOyyy9I9BaTB\nggULpkyZcsopp6R7EEgDL/5Ea/LkyRs3bjz22GPTPQgHPD8yCgAAEClBCAAAEClBCAAAECm/\nQ8hPxMaNG0uWLJnuKSANNm/enJeXV6JEiXQPAmngxZ9o5eXlhRAyMjLSPQgHPEEIAAAQKT8y\nCgAAEClBCAAAEClByAFjw2eDerU6pEypMjVb9356+qZdubfoXeBAsfsX/4w/NU98r02/2ft8\nZNhjirz+Z9997JbLvOYN43bl8XAg2d2L34s/P4wg5EAx55+X9/n81P+buXjm06d8ctUV/Rfu\n/N6id4EDxQ+4+EMIR/x+QnKLcTccts9nhj2l6Ov/sFs+TCaTE24+YhcfDweQ3b34t/Diz24T\nhBwg5r7w3KRTb7z1xGrlqp14640nj3/mhUU7u7foXeBA8QMufvjJ2N0r3H8R/GS4mNlXBCEH\niDlz5hxSt25mCCGEzLp1D5kzZ87O7i16FzhQ/ICLP4QQ5jzY4eDS2TUad71u2MzcfTwy7DG7\n+0rulZ+fjB94MXvxZ7cJQg4ciURi60ep/rWUlPcWvQscKHb74q9728SN3678dsWsUX+sm3Np\nz799uU/GhL1id1/JvfLzk7HbF7MXf34IQcgBolatWvOmT18fQghh/cyZC2rVqrWze4veBQ4U\nP+Di36pYZsWG59zSq/HHH032ZTEHqN19JffKz0/GD7+YvfizewQhB4hDf3Zuk9fv/cuYRSsX\njfnLvW+0vOBn1XZ2b9G7wIHiB1z8W21ev+zTF+8eNLXFsU0TBY8LB4LdfSX3ys9Pxg+/mL34\ns5uScIBYN/Wxn7eoXrrEQTVa9npy2oZk8rM7GmX0HFbYvYVuhANOqit52/Wf4t7P/tgohBCK\nZZar2eyMG4fPyk3j8PAjFf3iP6xnqW1f02ScP8wrPz8hRb/4F7z4vfjzwySSfr4eAAAgSn5k\nFAAAIFKCEAAAIFKCEAAAIFKCEAAAIFKCEADYd9aM+2uXOtklEh0eXprvYwDSQhACQHpcWSmR\nSCQSxctUqXfilf83LXefPvmIizO7DV7/Q/feNPvlP5zX6vAKpUuXq9286/WDp67a1T03vHzv\nvVl3fLEumXNlpe8/vmFszcKG2d05h19Q/Oxn9+2pBDigCUIASJcGd0xJJtd/Pa5fg1G/6PPk\nPn2XrNvg9SMuzvxh+y4b9ot2vd6ofu1zExYsnfPu4Gtrv9n/lWW7uO/X8+ZVaNSsavF8Hxcx\nzI+ZE4CdE4QAkFbFy9Q564JOZT77bObWDUv/e/+FrQ6vUCarWqPTbn1tfgi5s4ZcfcJh5bOr\nN+ne76ZTi18wPIQQhvZInHpTvx7NamY1vn3ijrtsnPpIz2Nrli1ZLJFIHHbDhyEU2PL9O2+b\n54685bSG1bKyKtfvdM2QGRvDdwe/5YHzmtXIzj6kzVXD5+UbePqjfYfU+dMLf7+oVe3yZcrW\nbHxynycfvKhiiuMUWMjShzrUumn8zD8fnUg0v/babR+3afPd24BL3r3votZ1KpYulkgkzh68\nfvt3CAscreCQ6wd1O+e5vJcuLJFIJE4dtHrffP4ADmyCEADSKm/Nly89+9bGo4+uH0IIYfnQ\nK3/9XpdH3pu/4uuJgzq9/8vrhs18/JdXjG//+JR5n4+4fM4bo7btmTNm2W9GfL5y6u9m7LDL\ni8P73bzwgrfmrslLJpOz+x0bQvLNHbdsteiJ3j1fPfK+9+YvmPhwm/G9L35w1tY7/jN2xW9G\nfrHws0ebv97n/vHbzzvh48l1O3asmX8VBY9TcCFrK12d8+VfWxzx+wnJ5MS//W3bx7fV3XKQ\npc9dcdbDJa55cdry3GQyOXz7dwZTHK3gkJm9Rgw7P+OsZzYlk8nXex28xz5FAD9hxdM9AABE\n6/O+TRJ9QwiJGj2HfHRR+RBC2Dz25ZGTh75wzNDeWx/T4PinM95tf82rHWuWCqHrnVe1/9db\n3+1+4pW3djq0TNg8esddTsj4Zf1PB978+9VdOpx0Rte2tQ4KiXpNdtiyRd57o3JaXT3ktMPL\nhtCp700n9+s/ZtUNdbJCCB2vvLVDzcwQup5x/NL+MzeFViW27pJMJguupOBx3n6t6o5TzQrn\nNC7qdOS9O+K1ZtctvLh5+QJ3FTwtJ8w6t4ghAdhF3iEEgHRpcMeU5OaNSz9+6Ngxtz8wfkMI\nW3Krw0NLkt+bdmejFAEWQgghKyurkF26Xz9mfP/z622a/OSVLZv2eXtdCA0KbNkiX90lkyGR\n2PpxZubW9+cyMorl5eVte1DxY1o0nZGTMz//LAWOEwpOVWQN7vj8O96T8mgFh0wUcgAAUhOE\nAJBOiRIVj/7N4/975ONX/21aMoSMtmd2ndrvmsc+WLh2a4RlHN/x+Hce+tOY+atXzXn9rn/9\np8ARCu4SQihdq+0Fv73rvoduaf/NuAlLUm8JIWQcf1KH8Q/d8dqXq1YvyOl737+P6dI+a2cT\n1+t9+8+m39bjumc/nLNi7bfzp7754KV9nmuw43E6nppiqqJltDvjtIn3X//MpKXrN+/KGlPJ\nzs6e/+XMDYU1NAA7EIQAkHYVLvzfmxP3XvfU4hAqnfvoiOtLDbqoWeVSiUQicXCv16tf1v/h\nY966pNEh9bs+WqHdcdnZ2TvsXWCXwQ93SSQSiWKla7S5Z9Pv7r2sVlhaYMtW1S4d8HTXz65v\nU6Nasyv+22LA4D51dj5tpe5PjB3QaU6/7s2qV6h1XM8HZne54oxGBY5TcCE7P/D5/Yf/cs39\nZ9bLLv7dH5UpdI2FHO24XjeUfbR1doY/KgOwaxIpfxMAANjPJDd9+9U791x84YTLPx35i0rp\nngaAnwbvEALA/u7ZHlve3Wt91bhjHv7HJWoQgD3FO4QAAACR8g4hAABApAQhAABApAQhAABA\npAQhAABApAQhAABApAQhAABApAQhAABApAQhAABApP4/k2zfqIv6SSIAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 300,
       "width": 600
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 10,repr.plot.height =5)\n",
    "model_plot = ggplot(coef, aes(x =reorder(variable_label, Estimate), y = Estimate, color = as.character(sign))) +\n",
    "  geom_point(size = 1) +                                  # Plot points\n",
    "  geom_errorbar(aes(ymin =`2.5 %`, \n",
    "                    ymax = `97.5 %`), \n",
    "                width = 0.2, size = 0.5) + scale_color_manual(values = c(\"1\" = \"#009ba9\", \"0\"=\"#82898f\")) + ggtitle(cell_type)  + coord_flip() +plot_config + theme(axis.title.y = element_blank(), legend.position = 'None') + ylab('Regression Coefficient') + ylim(min(coef$`2.5 %`), max(coef$`97.5 %`) + 0.1) +  geom_text(aes(y = `97.5 %` +0.05\n",
    "            , label =paste0( ' (p = ',  signif(`Pr(>|t|)`, digits = 3), ')')  ), size = 2) \n",
    "model_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "id": "057cf502-d883-4e0c-834d-4effe193bd20",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/I2/I2_Residual_Model_', model_form , cell_type,rm_rib_mt_genes, '.pdf'), width =3.5, height = 1.2)\n",
    "model_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "id": "b4b3a46a-dd8f-4cfd-98ea-b1b006bcc616",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save data for plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "49e6048d-656d-46ea-b4e4-94668aae930d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Estimate</th><th scope=col>Std. Error</th><th scope=col>t value</th><th scope=col>Pr(&gt;|t|)</th><th scope=col>2.5 %</th><th scope=col>97.5 %</th><th scope=col>variable</th><th scope=col>sign</th><th scope=col>variable_label</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>connect_quantile</th><td>0.05980493</td><td>0.009093687</td><td>6.576533</td><td>5.044212e-11</td><td>0.041981633</td><td>0.07762823</td><td>connect_quantile   </td><td>1</td><td>Connectedness in co-expression networks</td></tr>\n",
       "\t<tr><th scope=row>PPI_degree_quantile</th><td>0.02046636</td><td>0.009093687</td><td>2.250612</td><td>2.443070e-02</td><td>0.002643064</td><td>0.03828966</td><td>PPI_degree_quantile</td><td>1</td><td>Connectedness in InWeb PPI networks    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & Estimate & Std. Error & t value & Pr(>\\textbar{}t\\textbar{}) & 2.5 \\% & 97.5 \\% & variable & sign & variable\\_label\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\tconnect\\_quantile & 0.05980493 & 0.009093687 & 6.576533 & 5.044212e-11 & 0.041981633 & 0.07762823 & connect\\_quantile    & 1 & Connectedness in co-expression networks\\\\\n",
       "\tPPI\\_degree\\_quantile & 0.02046636 & 0.009093687 & 2.250612 & 2.443070e-02 & 0.002643064 & 0.03828966 & PPI\\_degree\\_quantile & 1 & Connectedness in InWeb PPI networks    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| <!--/--> | Estimate &lt;dbl&gt; | Std. Error &lt;dbl&gt; | t value &lt;dbl&gt; | Pr(&gt;|t|) &lt;dbl&gt; | 2.5 % &lt;dbl&gt; | 97.5 % &lt;dbl&gt; | variable &lt;chr&gt; | sign &lt;dbl&gt; | variable_label &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| connect_quantile | 0.05980493 | 0.009093687 | 6.576533 | 5.044212e-11 | 0.041981633 | 0.07762823 | connect_quantile    | 1 | Connectedness in co-expression networks |\n",
       "| PPI_degree_quantile | 0.02046636 | 0.009093687 | 2.250612 | 2.443070e-02 | 0.002643064 | 0.03828966 | PPI_degree_quantile | 1 | Connectedness in InWeb PPI networks     |\n",
       "\n"
      ],
      "text/plain": [
       "                    Estimate   Std. Error  t value  Pr(>|t|)     2.5 %      \n",
       "connect_quantile    0.05980493 0.009093687 6.576533 5.044212e-11 0.041981633\n",
       "PPI_degree_quantile 0.02046636 0.009093687 2.250612 2.443070e-02 0.002643064\n",
       "                    97.5 %     variable            sign\n",
       "connect_quantile    0.07762823 connect_quantile    1   \n",
       "PPI_degree_quantile 0.03828966 PPI_degree_quantile 1   \n",
       "                    variable_label                         \n",
       "connect_quantile    Connectedness in co-expression networks\n",
       "PPI_degree_quantile Connectedness in InWeb PPI networks    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "id": "c9c033ab-0587-44fc-960c-ca17901d5943",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(coef, paste0('figures/I2/I2_Residual_Model_', model_form , cell_type,rm_rib_mt_genes, '_data.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "id": "b660da34-9fa1-4202-aac9-4b94683ab5a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 28</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coeGene</th><th scope=col>hgnc_id</th><th scope=col>sign</th><th scope=col>GeneSymbol</th><th scope=col>TSSD</th><th scope=col>length</th><th scope=col>CDS_length</th><th scope=col>pLI</th><th scope=col>LOEUF</th><th scope=col>ABC_count</th><th scope=col>⋯</th><th scope=col>PPI_degree_quantile</th><th scope=col>PPI_degree_cat</th><th scope=col>TF</th><th scope=col>HI</th><th scope=col>hs</th><th scope=col>GO_BP_count</th><th scope=col>GO_BP_count_400_auc</th><th scope=col>GO_BP_count_400_auc_decile</th><th scope=col>weighted_mean_expression</th><th scope=col>residuals</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG</td><td>HGNC:5</td><td>1</td><td>ENSG00000121410</td><td> 1.2613884</td><td>-0.5302553</td><td>-0.1890112</td><td>-0.7228644</td><td> 0.8427422</td><td> 7</td><td>⋯</td><td>-0.8046553</td><td>1</td><td>0</td><td>0</td><td>-0.5549247</td><td> 1</td><td>-0.7806656</td><td>4</td><td>-0.1269236</td><td>1.701606</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2M </td><td>HGNC:7</td><td>1</td><td>ENSG00000175899</td><td>-0.6591235</td><td>-0.2184283</td><td> 1.2938860</td><td>-0.7228644</td><td>-0.5493879</td><td>89</td><td>⋯</td><td> 1.2528822</td><td>1</td><td>0</td><td>0</td><td>-0.6186645</td><td>90</td><td> 0.6995575</td><td>9</td><td>-0.1045367</td><td>1.546017</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 28\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & coeGene & hgnc\\_id & sign & GeneSymbol & TSSD & length & CDS\\_length & pLI & LOEUF & ABC\\_count & ⋯ & PPI\\_degree\\_quantile & PPI\\_degree\\_cat & TF & HI & hs & GO\\_BP\\_count & GO\\_BP\\_count\\_400\\_auc & GO\\_BP\\_count\\_400\\_auc\\_decile & weighted\\_mean\\_expression & residuals\\\\\n",
       "  & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & ⋯ & <dbl> & <int> & <int> & <int> & <dbl> & <int> & <dbl> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG & HGNC:5 & 1 & ENSG00000121410 &  1.2613884 & -0.5302553 & -0.1890112 & -0.7228644 &  0.8427422 &  7 & ⋯ & -0.8046553 & 1 & 0 & 0 & -0.5549247 &  1 & -0.7806656 & 4 & -0.1269236 & 1.701606\\\\\n",
       "\t2 & A2M  & HGNC:7 & 1 & ENSG00000175899 & -0.6591235 & -0.2184283 &  1.2938860 & -0.7228644 & -0.5493879 & 89 & ⋯ &  1.2528822 & 1 & 0 & 0 & -0.6186645 & 90 &  0.6995575 & 9 & -0.1045367 & 1.546017\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 28\n",
       "\n",
       "| <!--/--> | coeGene &lt;chr&gt; | hgnc_id &lt;chr&gt; | sign &lt;dbl&gt; | GeneSymbol &lt;chr&gt; | TSSD &lt;dbl&gt; | length &lt;dbl&gt; | CDS_length &lt;dbl&gt; | pLI &lt;dbl&gt; | LOEUF &lt;dbl&gt; | ABC_count &lt;int&gt; | ⋯ ⋯ | PPI_degree_quantile &lt;dbl&gt; | PPI_degree_cat &lt;int&gt; | TF &lt;int&gt; | HI &lt;int&gt; | hs &lt;dbl&gt; | GO_BP_count &lt;int&gt; | GO_BP_count_400_auc &lt;dbl&gt; | GO_BP_count_400_auc_decile &lt;int&gt; | weighted_mean_expression &lt;dbl&gt; | residuals &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG | HGNC:5 | 1 | ENSG00000121410 |  1.2613884 | -0.5302553 | -0.1890112 | -0.7228644 |  0.8427422 |  7 | ⋯ | -0.8046553 | 1 | 0 | 0 | -0.5549247 |  1 | -0.7806656 | 4 | -0.1269236 | 1.701606 |\n",
       "| 2 | A2M  | HGNC:7 | 1 | ENSG00000175899 | -0.6591235 | -0.2184283 |  1.2938860 | -0.7228644 | -0.5493879 | 89 | ⋯ |  1.2528822 | 1 | 0 | 0 | -0.6186645 | 90 |  0.6995575 | 9 | -0.1045367 | 1.546017 |\n",
       "\n"
      ],
      "text/plain": [
       "  coeGene hgnc_id sign GeneSymbol      TSSD       length     CDS_length\n",
       "1 A1BG    HGNC:5  1    ENSG00000121410  1.2613884 -0.5302553 -0.1890112\n",
       "2 A2M     HGNC:7  1    ENSG00000175899 -0.6591235 -0.2184283  1.2938860\n",
       "  pLI        LOEUF      ABC_count ⋯ PPI_degree_quantile PPI_degree_cat TF HI\n",
       "1 -0.7228644  0.8427422  7        ⋯ -0.8046553          1              0  0 \n",
       "2 -0.7228644 -0.5493879 89        ⋯  1.2528822          1              0  0 \n",
       "  hs         GO_BP_count GO_BP_count_400_auc GO_BP_count_400_auc_decile\n",
       "1 -0.5549247  1          -0.7806656          4                         \n",
       "2 -0.6186645 90           0.6995575          9                         \n",
       "  weighted_mean_expression residuals\n",
       "1 -0.1269236               1.701606 \n",
       "2 -0.1045367               1.546017 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(test,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b832007-1fe5-4fc8-8962-e6334b2c0664",
   "metadata": {},
   "source": [
    "### Residual Model for LOEUF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "id": "4187dbd1-8644-474e-bb10-6af280eccc93",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Estimate the model ( use residuals after correcting for expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "id": "ebf244e6-b060-4e9f-85c5-44150acd74af",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_form = 'residuals ~ LOEUF'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "id": "6a502248-e3a0-4b27-b2ad-4cc98b276e22",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"glm.fit: fitted probabilities numerically 0 or 1 occurred\"\n",
      "Waiting for profiling to be done...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "test  = coegene_data_scaled \n",
    "test$sign = as.numeric(test$sign)\n",
    "\n",
    "res = residuals(glm(sign ~ weighted_mean_expression, data = test, family = 'binomial'))   # select the beta of the random snp instead\n",
    "\n",
    "## Run model on residuals\n",
    "        \n",
    "test$residuals = res\n",
    "\n",
    "result_model = glm(as.formula(model_form), data = test, family = 'gaussian')\n",
    "\n",
    "#print(summary(result_model))\n",
    "## Extract model results\n",
    "confidence_int =confint(result_model)\n",
    "model_summary = summary(result_model)\n",
    "\n",
    "coef = cbind(as.data.frame(model_summary$coefficients), as.data.frame(confidence_int))\n",
    "coef$variable = rownames(coef)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 543,
   "id": "e2cd93da-5417-4868-a5ff-96c2a62dbcaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Estimate</th><th scope=col>Std. Error</th><th scope=col>t value</th><th scope=col>Pr(&gt;|t|)</th><th scope=col>2.5 %</th><th scope=col>97.5 %</th><th scope=col>variable</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>(Intercept)</th><td>-0.11565582</td><td>0.009120638</td><td>-12.680672</td><td>1.410130e-36</td><td>-0.13353194</td><td>-0.09777969</td><td>(Intercept)</td></tr>\n",
       "\t<tr><th scope=row>LOEUF</th><td>-0.07655475</td><td>0.009121076</td><td> -8.393171</td><td>5.346048e-17</td><td>-0.09443173</td><td>-0.05867777</td><td>LOEUF      </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Estimate & Std. Error & t value & Pr(>\\textbar{}t\\textbar{}) & 2.5 \\% & 97.5 \\% & variable\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t(Intercept) & -0.11565582 & 0.009120638 & -12.680672 & 1.410130e-36 & -0.13353194 & -0.09777969 & (Intercept)\\\\\n",
       "\tLOEUF & -0.07655475 & 0.009121076 &  -8.393171 & 5.346048e-17 & -0.09443173 & -0.05867777 & LOEUF      \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | Estimate &lt;dbl&gt; | Std. Error &lt;dbl&gt; | t value &lt;dbl&gt; | Pr(&gt;|t|) &lt;dbl&gt; | 2.5 % &lt;dbl&gt; | 97.5 % &lt;dbl&gt; | variable &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| (Intercept) | -0.11565582 | 0.009120638 | -12.680672 | 1.410130e-36 | -0.13353194 | -0.09777969 | (Intercept) |\n",
       "| LOEUF | -0.07655475 | 0.009121076 |  -8.393171 | 5.346048e-17 | -0.09443173 | -0.05867777 | LOEUF       |\n",
       "\n"
      ],
      "text/plain": [
       "            Estimate    Std. Error  t value    Pr(>|t|)     2.5 %      \n",
       "(Intercept) -0.11565582 0.009120638 -12.680672 1.410130e-36 -0.13353194\n",
       "LOEUF       -0.07655475 0.009121076  -8.393171 5.346048e-17 -0.09443173\n",
       "            97.5 %      variable   \n",
       "(Intercept) -0.09777969 (Intercept)\n",
       "LOEUF       -0.05867777 LOEUF      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "id": "47963e56-6657-4942-bb41-efb630a56627",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef$sign = 0\n",
    "coef$sign[coef$`Pr(>|t|)` < 0.05] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "a90dc248-22a8-4d36-bb7f-81bd6c5e7551",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef = coef[coef$variable != '(Intercept)',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 546,
   "id": "11f7dd68-925a-4165-a58e-bebcdab79626",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLAAAAJYCAIAAAD9hIhNAAAABmJLR0QA/wD/AP+gvaeTAAAe\nTUlEQVR4nO3de6CVc7748e/TLrUl3Ta6TRptim6E3JJymzQahlzzI4RmGMcwjGNmXIYz5xz3\nMz/8TKiQcSuMYeQyJJk6ISGki0zpoouoTNfd+v2hUu3drq209vZ5vf5aPbf9eXb203p71lo7\ny+VyCQAAgHiq5XsAAAAA8kMQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEI\nAFvimTNqHTt46boPAKDKEIQAUBHzB/6oZtb8kpGr8j0IAGw5QQgAFTB98ID3evRq8eiAl5bn\nexQA2GKCEAA23+T7B/7ztCv+dO5+fx3w9OJ8DwMAW0oQAsDmyo0dMGjpmWcf1KBXnyNfGvDY\ngnzPAwBbSBACwGZaNWLgg/X69GmbUu2eZx375oCHZuR7IgDYMtXzPQAAVBHLnh/w50/f+nz3\n7N9XL3j3/okXXZXXkQBgy7hDCACbZdFfBj7Z+tZPc6utfPUXswcOGJfvqQBgSwhCANgcnz82\n8Nn9zzit6Zo/Fxx6Vu+SBwaM8OsnAKjCslwul+8ZAAAAyAN3CAEAAIIShAAAAEEJQgAAgKAE\nIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQVXSIHzllVfyPQIAAMD3\nXCUNwlGjRuV7BAAAgO+5ShqEAAAAfNcEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUI\nAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQA\nAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAA\ngKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABA\nUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAo\nQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQg\nBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShAC\nAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEA\nAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAA\nQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICg\nBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCC\nEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEI\nAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQA\nAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAA\nBCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACC\nEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJ\nQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQh\nAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAA\nACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAA\nEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAI\nShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQl\nCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKE\nAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIA\nAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAA\nQFCCEAAAIKjq+R6girlt/MRRn83vWFT/yg6t8z0LVAG3vjdx9Jz5++1U/4r2fmQAACodQVgx\no+fMf3zq9CUlJfkeBKqGUXPmDZn66bJVfmQAACojLxkFAAAIShACAAAEJQgBAACCEoQAAABB\nCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAE\nIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQ\nAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgA\nABCUIAQAAAhKEAIAAAS1iSDsV9T62vHrL1o2+fErjuvQtE7Nmjs0aX/s5Y9PXJZSSsP7FWXr\naHvDhJQWD+qedR+0eO2eQ3plne+YXWrj4wdv7ZMCAABg0yp6h3De0LO79Hut1TXPTZo3b/Jz\nV+814vwuZz8xN6WUUqtr3sutMf63rcs/zjobP3XGtxsdAACALVG9YptPvvf6ocXXT7nxhGYp\npdSh142PzRi1+/X9f3/CId/FcAAAAHx3KnaHsGTs2HeLu3Vr9s2SXbt1a/HO2LdLUkofXddu\nzatAq5/61CaO9M3GvR5ZWeGpAQAA2GIVC8JceSvXeRXoykeOTyllWZZyubX75HK5lGXZhhsP\nObWCdymhLJ8tWTp7ydJ8TwHApqyYddaz48eX+5SiKlk5+5xn3xv3vTkdIJ6KxVj1fTq2n/zc\n8Blpz6arl0wbPvyTDj/apyD9vfTWtRs12mHWrFkp1UkppTRz5swajRsXbeHAsKFRc+af99ob\n7y9YmFLas96O9xy63yG7+O8MoJKa/MEHc4s7t822/Ehz+j04/E/LUkopVW/xXJ9O3ddbW/L+\npHd+9tY/x66ouX/LdgMP+kGLrPztS5t30UMv37kkpfptPjyxTeuU0oL3Ww99/6M1qw8+8NjX\n226fqje6tPH4qz9u/UTLGlt+SgDbXgXvzu3e9zfH3/6zU65sfPclR+6aTX3x9guunXBC/yHF\n6YOytj6ge/cZ/37t/cf+T6/igmnPXv3/Pjz6ys4FW2NqWGPGV0uOGTZi0YrVLzye+OWiHsNG\nvHti91132D6/gwFQlkWDJxcc9+OaW+dg1X7wZN+Dji9zVW7O0M/q39izffvsi2ufH3HN9Gb3\nNy93+zIU3dH75DsWTtj7xZLVC+q3mdC3TUop5T6/dMi7rVuu/oembYtGE8ZM/6LlbvW27GwA\n8mKTQfjRde2y61Y/PuKeBS/1vX/Eqt9e9ruji0+elxru0eX0u0b8R6+dU/pg/S0bXvjKvDu6\n1jvlzienXnz5cXteOGdVw9bdzhl6z1k7f3ensg29NW/ByX8fle8pSCmliV8u+nL5irV/LMnl\nFq5Y2fP5ka3r1cnjVKw1es7n+R4BqExWzB+9vN51tVJKKS2Z3Plvn3dp+MXAf35VuNNud3fr\ncHTh2u0+6/vAq/ctX3fPrPdRJw3edf2j5Waedf/QVKPO0W079m9fVH+9zRtf3TmllFu2pFqN\najWLamVlbb/yjQ/euujdmeOXFezZvM2ALi3bb97/tl4yffLQBi3fXztt/frt5s4ck3Y7uiLf\nCYBKIlvnTX55NnTo0MWLV//WwjFjxtx55535nadMp7w86rGPp+d7CqhifrJrk78c1TnfUwCV\nwOKP9n9+5YNfvwJzyeTOD0/p9KNDr98lGz125Fkr2k8/ZJeKv5J01bzPp1/y4jt1u/S8s/EG\ne8849b7XH81V+2FxpxGHNW+Wldq+4N1Wo7P+R+x1YM0VI9987ZodDhnZtvaGh184Ye8XSx75\neuDVlt837IU32vW4u+naD2KYc8GDEw/r3fn0iv4yL4BKoBJ9oMuJJ5649vGMGTPyOMkmNd6+\nsLN3qVUOE79c9M7nX2ywsF39eu4QVhKj5sz/9Kt/5XsKoNKoWbPh8iVfrv1jnSZ9mm5fO6Uj\n9mi648iFc9Mua15JtHl3CFNKqVpRg10vbPHe5fOXpsaF669q+si5J92/dOHAka+dNWGnv+9Z\nuMH28wq+mDjns64PT/h66x33WDjuzRH7jFuUUv2bTjnqVxv7Z2Thx3ctan5v03XiL7fii1Sz\nSA0CVVMlCsIqZN+i+o8dcVC+pyCllGYvWdp26LAFy5avyqWUUrUsq7tdjWHHHNpk+8JN7cq2\ncNLf/zFkqiAE1qjR8MDtPnlnWTrg63cRLpr1wMzi3++cjZk0c1H9djt9s90u95558r2bdcRV\n8xZMv/OTVR0OrbXe4rmTfjm33iXFDXfK5VblSv61clXp7RtWr9eqUcP+XVsdtEONNR8I0zi3\nX/lfLjf6w48LWnXZZ91lCxa8t1PDTps1LUClIwip2hoV1nqh+2E/f/2t/537eUpp36L6dx3S\nUQ0CVFZ1eheX/Nsny85vVTOllOrW227iay1f/KpW0W73HV7B14vOfrvRM5M+S9Xq1657dNuD\nb2iSpfRZ3wfe3/ukwy8qTKlh065T3zzq0bmzssJ9d23Xf8/apbfPUpuHWo69+Nlnxy5evjSX\njjjkuJf2XPfTbpbe/tenf/lZSintee/7hxzYc2TbwrRy9p2Ta19w4g7rDvL+J7NbFXfxiTJA\nFSUIqfI6FtUffdyRC1esyOVS3e186jdApbb7Xnvt9MKk9/Zo2y6llO1wTtdOf/h2B2q0z+y+\n+6y/aJd7z9xl9cNq2x/Xqctxncrfvvq+e3Z6fc+NfYFal/Q8+ZINllVv/GDvxustWTn7tlk7\nX733dhUbHqDSEIR8T+xYQwoCVAU1Gt//48ab3qyqqN7o3h83yvcQAN+eIAQA8qGweOQJ+Z4B\nIDwfiQUAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAA\nEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAI\nShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQl\nCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKE\nAAAAQVXP9wBVzIE7N8zlUsei+vkeBKqGg3YuylK2305+ZAAAKiNBWDG/bLvHL9vmewioOi5t\nt0e+RwAAYKO8ZBQAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAA\nQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAg\nKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCU\nIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQ\nAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgB\nAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAA\nAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACA\noAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQ\nghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChB\nCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAE\nAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIA\nAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAA\nghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABB\nCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAE\nIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQ\nAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgA\nABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAA\nCEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAE\nJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIS\nhAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlC\nAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEA\nAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAA\nIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQ\nlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAgq\ny+Vy+Z6hDA8++ODMmTPLXLVkyZKUUmFh4badCAht/vz5DRo0yLIs34MAgcyfP79hw4b5ngKo\n8urVq3fBBRdsbG0lDcJyjBs3btWqVR07dsz3IEAggwYNOvPMM6tV86oKYNsZMGDAOeeck+8p\ngO85T24AAACCEoQAAABBCUIAAICgqt57CEtKSlJKBQUF+R4ECGT58uXbbbddvqcAYnHlAbaB\nqheEAAAAbBVeMgoAABCUIAQAAAiq8gbhsg8H9enUtHbN2s0O6PvQpBXrr/zkv/bLsizLsma/\nGr2ZuwBsWvmXkbLWTr5h72ytA2/+ZJuPDFRhFb3meKoDbHWVNgin3XnuxR91//OUz6Y89KP3\nLzy//6z11ra48s1cLvf2r1tu/i4Am1T+ZWRja1v+5u3c10b/qsU2nxmouip6zfFUB9j6KmsQ\nTh/66DvdL7/qsEb1Gh121eVHj3l46OzvYBeAdZV/GXGRAbauil5z3nAVAra+yhqE06ZNa1pc\nXCullFKt4uKm06ZN+w52AVhX+ZeRja6d9seuOxTWbdK2x6VPTlm5jUcGqrCKXnNGeqoDbH2V\nKAgnr3lfYOfbP00ppSzLVq/Z7F+N8S12AWKr2JWnjLXFvx23fOEXCxd8/NL1xcPP6n371G0w\nNPC9UdFrjqc6wNZWiYKw+Mo3v34TzshLmqXmzZt/OmnS0pRSSkunTJnZvHnzTe3/LXYBwqvA\nlae8tdVqNdzrp1f2aTv2rXc9RQM2U0WvOZ091QG2vkoUhOv5wQkntRt20x9enf3F7Ff/cNPz\n+596QqPvYBeAdZV/GSlv7aql8z944r8Gjd93v/ZZ6eMClKWi15z9PdUBvgO5ymrJ+Pv+z76N\nC2ts32T/Pg9MWJbL5XK5D69pU9D7yVwul3uyd81vzqHglCc3ugtABZS6jHxz2SlrbS734fVt\nUkqpWq16zTr0vPypj1fmcXigyqnoNcdTHWCry3JegQ4AABBSZX3JKAAAAN8xQQgAABCUIAQA\nAAhKEAIAAAQlCAEg/74a/Z9H7la3Rtb1lmdWP7h7Xr5nAiAAQQhA5dKvKMuyLKtee+fdD+v3\n5wkrt+kXf+aMWscOXvqtdl3xydNXn9zphw0KC+vtunePywaPX1SBnZc9fdNNda6ZuCT3fJOB\nXz8Y3q+onGEqOudTp1Y//pFt+60EoEoQhABUNq2ueS+XWzpn9M2tXjr74ge26Y2yYwcvfeaM\nWt9ix/lPnt25z/ONL3n07Znzpr0++JJdX+j/1/mbv/ucTz9t0KbDLtW/eVD+MN96TgBYjyAE\noFKqXnu34049vPaHH05JKaV5/7j1tE4/bFC7TqM2x1z13IyUUkorP378okNb1K/buN2JN1/R\nvfqpT6UhvbLuV9zcq0OzOm1/N66MvZaP/1Pv/ZrtuF21LMta/OrNlEotWXvnbdX0Z688Zq9G\nderstMfhv3h88vKUhvTKul9528kdmtSt2/TAC5/6dL1pJ91z3eO73TD0f07vtGv92js2a3v0\nxQ/88fSGZR6n9LnMu6Nr8yvGTPmPfbIsW/Ng7xsmr3MbcO7rt5x+wG4NC6tlWXb84KXr3iEs\ndbTScy4ddOxPHy35y2k1sizrPmjxNvsrBKAKEIQAVEolX039yyMvL99nnz1S+nxIv5+NOvJP\no2YsmDNu0OH/e96lT/4rpVkDzzt/TJeB73360TPnTnv+pdW7DX91/s+f+eiL8dfvXXqvJ566\n+dezTn15+lcluVzuk5v3Syn3woZLVpt9f9/ef2t9y6gZM8fdfeCYvmf88eOUUnpt5IKfPztx\n1of37D3s4lvHrDvs22PfLe7WrVmpsyh9nDLOpeii4VP/c9+Wv3k7l8uteTDut8VrDzLv0fOP\nu7vGL56Y8PnKXC731Lp3Bsv8zmw4Z60+zzx5SsFxD6/I5XLD+uywNf+WAKjqqud7AADYwEfX\ntcuuSyllTXo//tbp9dOqp59+9t0hQzsO6bt6g1aHfpx+MvGV17v84m/dmtVMqce1F3a56+WU\nUkqH9bvq8B/UTimlVSNL7VVw3h4fDPj1bxYf2fWonj0Oab59ynZvt8GSr5WMeml4p4seP+aH\nO6Z0+HVXHH1z/1cXnZdSt35XdW1WK6UePQ+e13/KitSpxurtc7lcmWdS+jivPLdL6XP5advy\nvh0lrz/zXIdLZ52xd/1Sq8o4x49PKmdOANiAO4QAVDatrnkvt2r5vLF37Pfq724bsyzlcrnU\n9Y65ubUmXNs2pbIbrE6dOqsflbHXiZe9Oqb/KbuvePeBfvu3v/iVJSm1KrVkzc7rHDOXS1mW\nUkq1aq2+OVdQUK2kpOSbLap33Lf95OHDZ5Qap9RxUpnnUr61X7+MNWUerfSc2UYOAEB0ghCA\nyiir0XCfnw/879YDL7p9QrVDftJj/M2/uO+NWf/6JsIKDu528Ig7bnh1xuJF04b9/q7XSh2h\noKy9Cpsfcuq//f6WO67s8uXot+eWvSSlVHDwUV3H3HHNc1MXLZ45/LpbXux4ZJc6pb7Aunbv\n+7sTJv2216WPvDltwb8Wzhj/wh/PuvjP88s4TrfuZUxVvoLOPY8Zd+tlD78zb+mqzTnHstSt\nW3fG1CnLyr6PCUBgghCASqvBaf/96+ymSx8sOemeZy6rOej0DjvVzLIs26HPsJRS43P6393x\n5TPbNN2jxz0NOh9Ut27dDfYu2nCvwXcfmWVZVq2wyYE3rvjlTec0T/NKLVmt0Vn3PtTjw8sO\nbNKow/n/2PfewRfvtolRi068f+S9h0+7+cQOjRs0P6j3bZ8ceX7PhmUdp9RUwzb5bSg6pf9T\n53116092r1t9zYfKbPQcN3K0g/r8asd7Dqhb4ENlAFhftrG3PQBAVZBbsfCfI24847S3z/3g\n2bOL8j0NAFQt7hACUFU90uvru3sHXDi6493/90w1CAAV5Q4hAABAUO4QAgAABCUIAQAAghKE\nAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAI6v8D0b6YnJcSZtwAAAAASUVORK5C\nYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 300,
       "width": 600
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 10,repr.plot.height =5)\n",
    "model_plot = ggplot(coef, aes(x =variable, y = Estimate, color = as.character(sign))) +\n",
    "  geom_point(size = 1) +                                  # Plot points\n",
    "  geom_errorbar(aes(ymin =`2.5 %`, \n",
    "                    ymax = `97.5 %`), \n",
    "                width = 0.2, size = 0.5) + scale_color_manual(values = c(\"1\" = \"#009ba9\", \"0\"=\"#82898f\")) + ggtitle(cell_type)  + coord_flip() +plot_config + theme(axis.title.y = element_blank(), legend.position = 'None') + ylab('Regression Coefficient') + ylim(min(coef$`2.5 %`), max(coef$`97.5 %`) + 0.1) +  geom_text(aes(y = `97.5 %` +0.05\n",
    "            , label =paste0( ' (p = ',  signif(`Pr(>|t|)`, digits = 3), ')')  ), size = 2)\n",
    "model_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "id": "68481fb9-b299-403b-9653-d5585bdd0764",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/I2/I2_Residual_Model_', model_form , cell_type, rm_rib_mt_genes,'.pdf'), width =3.5, height = 1.2)\n",
    "model_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "id": "c7aa67b1-3477-4514-bb9a-a683cc7596de",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save LOEUF data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "id": "b4beb794-abe1-4735-bf6a-6b1599d9e333",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(coef, paste0('figures/I2/I2_Residual_Model_', model_form , cell_type, rm_rib_mt_genes,'_data.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 550,
   "id": "77aec8dd-7c0a-4320-9837-791b47ac3096",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Make boxplot based on residuals (regress out expression out of LOEUFscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "id": "b6843eef-d108-48d8-a39a-e7bb83e8ca7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 28</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coeGene</th><th scope=col>hgnc_id</th><th scope=col>sign</th><th scope=col>GeneSymbol</th><th scope=col>TSSD</th><th scope=col>length</th><th scope=col>CDS_length</th><th scope=col>pLI</th><th scope=col>LOEUF</th><th scope=col>ABC_count</th><th scope=col>⋯</th><th scope=col>PPI_degree_quantile</th><th scope=col>PPI_degree_cat</th><th scope=col>TF</th><th scope=col>HI</th><th scope=col>hs</th><th scope=col>GO_BP_count</th><th scope=col>GO_BP_count_400_auc</th><th scope=col>GO_BP_count_400_auc_decile</th><th scope=col>weighted_mean_expression</th><th scope=col>residuals</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG</td><td>HGNC:5</td><td>1</td><td>ENSG00000121410</td><td> 1.2613884</td><td>-0.5302553</td><td>-0.1890112</td><td>-0.7228644</td><td> 0.8427422</td><td> 7</td><td>⋯</td><td>-0.8046553</td><td>1</td><td>0</td><td>0</td><td>-0.5549247</td><td> 1</td><td>-0.7806656</td><td>4</td><td>-0.1269236</td><td>1.701606</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2M </td><td>HGNC:7</td><td>1</td><td>ENSG00000175899</td><td>-0.6591235</td><td>-0.2184283</td><td> 1.2938860</td><td>-0.7228644</td><td>-0.5493879</td><td>89</td><td>⋯</td><td> 1.2528822</td><td>1</td><td>0</td><td>0</td><td>-0.6186645</td><td>90</td><td> 0.6995575</td><td>9</td><td>-0.1045367</td><td>1.546017</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 28\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & coeGene & hgnc\\_id & sign & GeneSymbol & TSSD & length & CDS\\_length & pLI & LOEUF & ABC\\_count & ⋯ & PPI\\_degree\\_quantile & PPI\\_degree\\_cat & TF & HI & hs & GO\\_BP\\_count & GO\\_BP\\_count\\_400\\_auc & GO\\_BP\\_count\\_400\\_auc\\_decile & weighted\\_mean\\_expression & residuals\\\\\n",
       "  & <chr> & <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & ⋯ & <dbl> & <int> & <int> & <int> & <dbl> & <int> & <dbl> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG & HGNC:5 & 1 & ENSG00000121410 &  1.2613884 & -0.5302553 & -0.1890112 & -0.7228644 &  0.8427422 &  7 & ⋯ & -0.8046553 & 1 & 0 & 0 & -0.5549247 &  1 & -0.7806656 & 4 & -0.1269236 & 1.701606\\\\\n",
       "\t2 & A2M  & HGNC:7 & 1 & ENSG00000175899 & -0.6591235 & -0.2184283 &  1.2938860 & -0.7228644 & -0.5493879 & 89 & ⋯ &  1.2528822 & 1 & 0 & 0 & -0.6186645 & 90 &  0.6995575 & 9 & -0.1045367 & 1.546017\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 28\n",
       "\n",
       "| <!--/--> | coeGene &lt;chr&gt; | hgnc_id &lt;chr&gt; | sign &lt;dbl&gt; | GeneSymbol &lt;chr&gt; | TSSD &lt;dbl&gt; | length &lt;dbl&gt; | CDS_length &lt;dbl&gt; | pLI &lt;dbl&gt; | LOEUF &lt;dbl&gt; | ABC_count &lt;int&gt; | ⋯ ⋯ | PPI_degree_quantile &lt;dbl&gt; | PPI_degree_cat &lt;int&gt; | TF &lt;int&gt; | HI &lt;int&gt; | hs &lt;dbl&gt; | GO_BP_count &lt;int&gt; | GO_BP_count_400_auc &lt;dbl&gt; | GO_BP_count_400_auc_decile &lt;int&gt; | weighted_mean_expression &lt;dbl&gt; | residuals &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG | HGNC:5 | 1 | ENSG00000121410 |  1.2613884 | -0.5302553 | -0.1890112 | -0.7228644 |  0.8427422 |  7 | ⋯ | -0.8046553 | 1 | 0 | 0 | -0.5549247 |  1 | -0.7806656 | 4 | -0.1269236 | 1.701606 |\n",
       "| 2 | A2M  | HGNC:7 | 1 | ENSG00000175899 | -0.6591235 | -0.2184283 |  1.2938860 | -0.7228644 | -0.5493879 | 89 | ⋯ |  1.2528822 | 1 | 0 | 0 | -0.6186645 | 90 |  0.6995575 | 9 | -0.1045367 | 1.546017 |\n",
       "\n"
      ],
      "text/plain": [
       "  coeGene hgnc_id sign GeneSymbol      TSSD       length     CDS_length\n",
       "1 A1BG    HGNC:5  1    ENSG00000121410  1.2613884 -0.5302553 -0.1890112\n",
       "2 A2M     HGNC:7  1    ENSG00000175899 -0.6591235 -0.2184283  1.2938860\n",
       "  pLI        LOEUF      ABC_count ⋯ PPI_degree_quantile PPI_degree_cat TF HI\n",
       "1 -0.7228644  0.8427422  7        ⋯ -0.8046553          1              0  0 \n",
       "2 -0.7228644 -0.5493879 89        ⋯  1.2528822          1              0  0 \n",
       "  hs         GO_BP_count GO_BP_count_400_auc GO_BP_count_400_auc_decile\n",
       "1 -0.5549247  1          -0.7806656          4                         \n",
       "2 -0.6186645 90           0.6995575          9                         \n",
       "  weighted_mean_expression residuals\n",
       "1 -0.1269236               1.701606 \n",
       "2 -0.1045367               1.546017 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(test,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "id": "1560224f-fac2-472c-a9f4-4387c9faeb6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "101"
      ],
      "text/latex": [
       "101"
      ],
      "text/markdown": [
       "101"
      ],
      "text/plain": [
       "[1] 101"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(test[is.na(test$LOEUF),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 553,
   "id": "cc02c586-489d-4139-8934-cfd32aa07ab3",
   "metadata": {},
   "outputs": [],
   "source": [
    "test  = coegene_data_scaled \n",
    "test$sign = as.numeric(test$sign)\n",
    "test = test[!is.na(test$LOEUF),]\n",
    "\n",
    "res = residuals(glm(LOEUF ~ weighted_mean_expression, data = test, family = 'gaussian'))   # select the beta of the random snp instead\n",
    "\n",
    "## Add residuals to data\n",
    "        \n",
    "test$residuals = res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 554,
   "id": "a4caec66-9672-44dd-b864-9c67826aa0c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlgAAAJYCAIAAAAxBA+LAAAABmJLR0QA/wD/AP+gvaeTAAAf\nY0lEQVR4nO3daZhU9Znw4VNNI80iNogKgg0qiTKIMi4RFUVcEuXFGURRYlQISYxLXmMmuCEa\n44JO0IQkxoUoCEHcR8WEuCNGg4o7oiCCyBJbREDWhl5qPkAYNdDNUl2n4bnvD15dXaf+9TRy\n9Y86deqcTDabTQAgqoK0BwCANAkhAKEJIQChCSEAoQkhAKEJIQChCSEAoQkhbKU/n1nUc0zZ\nl78AtiVCCJvp85HfaZApuejFqrQHAXJCCGHzzB0zYkqPU9vdP+KZNWmPAuSCEMJm+XDUyI+/\ne8kdPzj48RHjlqc9DJADQgibIfvGiLvLzv7+Yc1P7X/cMyMeWJz2PMDWE0LYdFUvjPxTcf/+\n+yVJ45P69XxtxD3z054I2GqFaQ8A247VT44YO+/1Rd/IXL7uG++M+uAng1IdCdhqXhHCplr2\n2MhH9v31vOw6FRP/f+nIEW+lPRWwlYQQNtGiB0b+5ZAzv9v6n7frHdnve5WjR7zgYxSwbcu4\nMC8AkXlFCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBo6YdwwoQJaY8A\nQFzph3DSpElpjwBAXOmHEABSJIQAhCaEAISWwxCWz33q+j4HtWnauHm7Q/sOm/RF7lYGgNqS\nwxBOvu/Bev3ufv2The8/dMrnVw64aWrulgaAWlKYu6UOv/iPhydJUrV6Yf0dChu1bLlT7pYG\ngFqS2/cI593UJVOvaJeDhmauve3cNjldGgBqQ25D2Gbgy9ny5XMn/rRy8H8OfqMqp2sDQC3I\n/VGjhY3bHHFR/8NmTHp1Uc7XBoAcy10IZ9x50S8emfLJ8lVLZj3333f+bc9DD2mRs7UBoJbk\n7mCZvXud2vSiHxzaf+rSJnseevLvxl190EY2LC8vHzt27PqbCxcuzNkMALCZMtlsNt0JhgwZ\nMmjQoHRnACAsZ5YBIDQhBCA0IQQgNCEEIDQh3C6UjemZ+ZrON374pQ0W3n70fldP28TVNmvj\nLVDxyTPXn9a51Y477vZvPQY/8ekm3rVhL/6kZSaTyWS+OnDFvCeuOOmANk3rZzLF5z6T4/GB\n7YwQbheKzvxzNpvNZuf+5oi2P5+UzWaz2bcua5/yUJVLpj898pof9jh80HNfvePte8euPn3U\nm/+Y97crdxrV/4YXN+2uDet6S2k2O+OGA77yzVm/6dNn3B5XPjtnZTa75PbjcvCzANsxIYxg\nye19z5s49ZcdMplMrzEVybLJt5x1aNtmjZvufvB3b3unLElKH7/wiHY7NSjIZDJH3zLtqxtv\nSOktXffrP+jsf2+1U7O9jr34qa++cKtaNvP5MTec959dDz/l6qeX7ffjO58dcsxXH37Qf424\n5pQDWu640zcP79yquLi4+rv+ZdoaTb939PsnXX/zqfsU19+sPybyacWKFe+8884bb7yxdOnS\ntGchuhxefYI6q/jc+2677+jS29+9et8kSV655MwxbYa/OLfLzl+8eEXP82878cmd/3h/u6Ez\nnz+1Rf1MkiRJ3y9tvDHT3yj7zV8+vK3ey7/8j35XPzv3tmMz6+6Y9F8djhjR7Hu/Gvn4rQc3\ny1SzQrL81et+9Jdut/95v2rveuXar0374s/a1fDjzp07t33RS33anfHc0t0O+d5No397clv/\n4KtDysvLR44ceffdd69ZsyZJkoKCgj59+px//vmNGzdOezSC8gtiu/XW4H0zmUwmc/BNs7/y\n/YVvvvHBKzceXbJjUeM2xw17a8q77zfuddVVTUafekz3Hmde/uD01Zu0+p49+x+/R+PGux/b\n/z+aTp362f/dcdiv35vx2CUd3rrmpC7HnXHp7x957R+rNvDw7KdP/NdJl6+4/NGhRzet9q5/\nnbaaH22dFi1avD+rxXWvf/75+6MOm/T9X4zfpJ+IfLnhhhuGDx9eXl6+9mZVVdX9998/cODA\n1E/uQVhCuN3qfN20bDabzb42sF2SNGzY8IsFpauySZLsfEDnfbpeOXH2kjVV2Ww2+8VdJyZN\nD77gjsefn/jELYe+/oNfPF71pY036qPxo5+dt3LlJ8+PfnxZx467fOmegqZ7dz9r0O3jXnzp\nwauObfzObT849uvvEa6eflffnsNaDn3shu7Nkxru2sC0X/vR/tV+J564R5IkSZJJkiQpKPCX\nvA756KOPHn/88SRJvpa9yZMnT5o0KaWhiM7viBgaH/O9oyb0blaY6TWm8rCr7zlj7uXd925a\nmMlkMsfdvnD2TV0ymUxmh+aH/qHg5z8+puBLG1ckU6/ssM8vpvzLgvt03uHuk/Zu1eH7rx59\n19XHbnAXaGGzfb894Bd3/fXvX3+P8G+/+fkDrz156SE7ZjKZTKbXmLIkeeaHxV1vKd3AXZl/\nmfara5UO65rJfOPyt6f+skMm03XYvCRJksIjfnnXd146s2OLnf9twGvdRl9zYo7+EMmBN954\nY2Ov/F5//fU8DwNrOdco1Vpw+38O2OH2Pw9o9eVvlt7S9YQld781OO3jUtn2jBkzZtiwYRu8\n67TTTrvkkkvyPA8kDpahBrue+9if056B7UhJScnG7mrbtm0+J4H1hJDN1/InL76V9gxsm7p0\n6bLbbrt99tlnVVVV67+ZyWSKioqOP/74FAcjMu8RAvmzww47/OpXv1r7EdGCgoJ69eolSdKo\nUaMhQ4Y0b/71Y6cgP7wiBPKqY8eOjz766AMPPPDee+9VVFTss88+ffr0UUFSJIRAvjVq1Kh/\n//5pTwHrCCGQb4sWLRo9evR77723Zs2afffd96yzzmrdunXaQxGXEAJ59eqrrw4cOHDlypUF\nBQXZbHbq1KmPPfbYVVdddeKJPvFJOhwsA+TPqlWrrrjiirKysiRJqqqqstlsNputqKi49tpr\nFyxYkPZ0BCWEQP5MmjRp8eLFX/7sRJIk2Wx2zZo1Tz/9dFpTEZwQAvkzf/78jd01b968fE4C\n6wkhkD9NmjTZgrugVgkhkD+HHHLIxq4H0qVLlzwPA2sJIZA/bdq06du3b5Ikmcy6a5as/aJ7\n9+4HHXRQmpMRmI9PAHl10UUXtWnT5tZbb12+fHmSJA0aNOjfv//ZZ5+d9lzEJYRAXhUUFJx2\n2mmnnnrqvHnzKioq2rZtu/aMo5AWIQRSUFBQUM0lmSCfvEcIQGhCCEBoQghAaEIIQGhCCEBo\nQghAaD4+AeSbC/NSpwghkFcuzEtdY9cokD8uzEsdJIRA/rgwL3WQEAL548K81EFCCOSPC/NS\nBwkhkD8uzEsdJIRA/rgwL3WQj08AeeXCvNQ1QgjklQvzUtcIIZACF+al7vAeIQChCSEAoQkh\nAKEJIQChCSEAoQkhAKEJIQChCSEAoQkhAKE5swyQjgkTJsyZM6ekpKR79+5pz0JoQhia30Sk\naPz48RMmTOjevbu/fqTLrtHQxo8f//vf/378+PFpDwKQGiEEIDQhBCA0IQQgNCEEIDQhBCA0\nIQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEE\nIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0\nIQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEILTchTC7+PU7Lzh2\nn12a7LT7/idd8cT8bM5WBoBak7sQfvTg7a/uN/jx90s/fvaSpmNP+cnY5TlbGgBqS2HOVtrr\nnD8OX/tVizP6HnXOtfMXJUmTnK0OALWiFt4jrJo78o5nDzzj5JLcLw0AOZa7V4RrZUvHndtz\n2K63PPXTb+R4ZQCoBTl9RVg554H+xw6q/OWTd/VqlcnlwgBQS3IXwoqPxpxx3DU7DHn6zl67\nqyAA24jchfDdu65+YMbUO3vtXpDJZDKZNhe9nLOlAaC25O49ws7XfZi9blM2LC8vHzt27Pqb\nCxcuzNkMALCZcn2wzCaoX79+v3791t8cMmRI/mcAgLWcYg2A0IQQgNCEEIDQhBCA0IQQgNCE\nEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA\n0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCE\nEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA\n0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCE\nEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA\n0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCE\nEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA\n0HIYwtk3HpzJZDKZTJuBL+duVQCoTTkMYbvLXstms29eunfulgSAWmbXKAChCSEAoQkhAKEJ\nIQChCSEAoRXmcK1Hzyw6+Z7VSZIkh2Vurnf6IxX39drQZtlsdtmyZetvVlZW5nAGANgsuQxh\nrzFl2TE1b1ZZWfnMM8+sv7l06dIczgAAmyWXIdzUpyws7N279/qb06ZNy/8MALCW9wgBCE0I\nAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEILYWTbgMQ2Zo1a8rL\ny+vVq1dUVJT2LEniFSEAeXbjjTd269btwgsvTHuQdYQQgNA2I4TZlaWz5iypqL1ZACDvanyP\n8K/nt3v8pNm3nrjiuQs6nfrgsnp7D3z25Uv3z8doAGyDRo0aNWXKlGo2WHs99lmzZg0cOLCa\nzTp16tSvX78cD7chNYZwxYKPF6xIklXj77h3/9/N/d2CYw67K7n0t3mYDIBt0ZQpU55//vka\nN1uyZMmmbJYHNYawadMdv/jii+zLz7+wX/ebm5S8XPLJy/mYC4BtWVFRw2Y7t9iyxy7+fGFZ\n2arczlONGkPYpWfPqQNP7tnkvX0vmNwmSZIkW1X7UwGwbWu2c4tvHdFtyx776ksTP5k/N7fz\nVKPmV4S9757c5qnJy755zLF7JEly4I9v2ykPYwFAfmz0qNHysn+q2nn/4044omSHsrKysrLd\nj+x/dB7HA4BqrH5v+Bn771a8a4eTh725YsuW2GgI+xVvTD6O4QGAmi3404U3ZK945aM3h+76\nh/Nvm7NFa2x01+jYsrItHmzbVV5eXlUV6E3QtT9sVVXV6tWr054lf+rXr19Q4FQSULsWff7Z\npBee27LHfrFk0SZumf3789N7nHN6u2ZJu3N7//QXL64eeEaDzX66ms81WrXg77feeNsz75cu\nL88mSZIkRz3zzFWb/TzbiHPOOaf6j79slyZOnHjEEUekPUX+jBgxYv/9fRYWatfqsrIFpf+o\n7WdZtmhR0xZrj01t0aLF558vSpJWm71IjSH8bNT3T3nowL6VT8w5+rEzl//3dS8dPmjzRwUg\nloYNGzXfZdcte+yizxasWrVyU7bcsXnzpa8vTJIkSRYuXNh8/+Zb8nQ1hvCNF17o/P0HjvzL\n9R917PGjvp2XfavHC8nQY7bkqbYdu7Zs1XbP9mlPQY6VlZVNeXNy2lNAFMXNdz64S9cte+yr\nL01cNX+TQpg5vNs3fzb8/tn/3vXd2x/epdv5m79fNNmEEC5btrxx48YtW7b8+OOPk6Rpw4Yf\nb/97Dhs1brL7Hm3TnoIcW75sadojALm269m/veS5voe2/bTZ0Zffc/OW/eKuMYRFzXZrVpQc\n+r0B5Sf9v2OeKXx7Zu8RW/REAJBzDfY77953z9uqJWoMYc8/lvZMkiS57uVXevxt6qq2h3X/\nt616QgCoSzb9CvVN9jz8xD1rcRIAth+LP1/46ksTt/ixuR2mejWG8JEzir77P1/5Tu+ysrG1\nNg8A24OyslX5PF/o1qgxhL3GLF++9hPmVavmT7z+R79ueUWtDwXANqtTp07VbzBt2rTS0tLi\n4uLOnTtvzTq5UmMIMwWFhetOwrFj2+MHn3HzPsOTi1yPEIANq/Fqutdcc824ceP22muvm266\nKT8jVW+zTjSVXTTplenZbG3NAgB5V+MrwscHtPj+uLVfVpatrL/PWX/4n+ofAADVKPintAdZ\np8YQnnDLhx/+OkmSJMnUa7jjjg3qyuAAbJsGDx48ePDgtKf4PzWGsH6j4uJG+ZgEAFKw0Rd4\npxduzOn5nA8AatVGQ3h/xVorX7+x+5HnPzRlwdKlC6Y8dP6R3W909QkAth81vuU3+c6bC86+\nvleH5o0aNe/Q67qzMjffmY+5ACAvagxhgwYN3p/86vK1N5ZPnjytwRZd5QIAkiRJko8//vi1\n116bPn162oOsU+PBMgf//NffPqpPh0O7Ht4umf3Si4tOuPOFfMwFwHZq1KhR48aNO/DAA4cP\nH572LEmyKSfdbtX7zqmHvz3x71M/S3YZNPzoA1rWz8NYAJAf1YSwqryssl5RvcqyNVXF+xzZ\nY58kSZKksqwsKSrSQgC2E9WE8LF+xQ/2KevzYLGrTwCwqUaNGjVlypRqNpg2bVqSJLNmzRo4\ncGA1m3Xq1KnG05bmRDUhPHls2clJkpxcVrb2dnZl6UcLi0pKivMwFgDbqClTpjz//PM1brZk\nyZJN2SwPanyP8K/nt3v8pNm3nrjiuQs6nfrgsnp7D3z25Uv3z8doAGyziuvX/8aOTbbssTOW\nLV9SXp7beapRYwhXLPh4wYokWTX+jnv3/93c3y045rC7kktdhgmA6nxjxyY//+Y3t+yxN3/w\nweRFi3M7TzVq/Bxh06Y7fvHFF9mXn39hv+5HNykpKfnkk3zMBQB5UWMIu/TsOfX6k3v+9OF9\nz+zdJkmSJFtV+1MBwCaouK9XJpPJZLrcNHuL16hx12jT3ndP3v3Rv7xd2KXvHkmSHPjj23ba\n4icDgFwq7Ptotu+8YV1PrdiKNWrc4rNnh577s5Ef/uP4nbMPnZq8M3rgw8efe/yWPyEAAby5\neMmPXnt9yx67qrIyt8NUr8YQvvf7n48/cdwdMw5+KEmSJKlYtWJVrQ8FwDauIptdVrEVL9Py\nqMYQzp49u8OJ+xfOWHtrxYoVjbbwcFgA4tijUcNuu+yyZY+d+Nlnc1fm70VXjSHs0LHj9Hff\nXXuATOXH99z/yoHfqfWhANjGtSwq6tmq1ZY9dvqyZXUqhHv+5Dd9v9Nv4AdJ1ayjO8yeuvPF\nf/5xPuYCgJq9fFGbw347P0mSPTPDfjpp3rAum79EzQfLNO569cQJfV986e1Psrse0PWoDjvX\n2/xnAYDa0GXYvOywrVuixhBOvW/wlE7X9e32H/tu3RMBQF1UYwiXTho5es2gvh0b5WMaALYP\nM5Ytv/mDD7b4sbkdpno1hvCgAT+rvPTap48cdPSeO7oMIQCbZEl5eT7PF7o1agzhuGsvfurJ\n5Km9blz/nVOy2YdqdSaI6brrrlu5cmXaU+TP1KlT1/530KBBac+SP1deeWXDhg3TnqJ2derU\nqfoNpk2bVlpaWlxc3Llz561ZJ1dqDOGpD2Wz+RgEwnv22WeXLVuW9hT5tmDBgqeeeirtKfLn\nsssu2+5DWOPVdK+55ppx48bttddeN910U35Gql7NR40C+dS6YcNdGjRIewpybGl5+awVK9Ke\ngg0TQqhbjtt11xNbtUx7CnLs9cWLh07fwiNHtj9t2rTp2LHjnnvumfYg6wghAHk1YMCAAQMG\npD3F/6nxeoQAsD0TQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IA\nQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEIT\nQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAELLZQhXv393/2+1btygcZtDf3jP\njPIcrgwAtSSHIZzzhx9cOP2EsTM/nXnPd6ZecM7wT3K3NADUktyFcO7D9799wsWDurUsbtlt\n0MXffvXeh0tztjYA1JLCnK00Z86c1u3bFyVJkiRF7du3njNnTpK0zNny+TRvzuwFpV7Qbm+y\n2aq0RwDqotyFMEmSTCaz7qtsNpvLhfOsory8otx7nAAh5G7XaElJybwZM8qSJEmSspkz/1FS\nUpKztQGgluQuhHv07tPpiaFDJpYuKZ04ZOiTh/TtvW3uFwUglBzuGm37k7uGvdPvu3vd+EWz\nzqfd8qdzdt/IdpWVlRMmTFh/c+nSpbmbITcK69ffYYcGaU9BjmWzVatWrkx7CqDOyeV7hEUd\nB4x+bcDomjbLZDIdOnRYf/Oll17K4Qw50aak3QEHHZr2FOTY8mVLn/3ruLSnAOqcnB4ss2kK\nCgpat269/mb9+vXzPwMArOUUawCEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQm\nhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQA\nhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQm\nhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhFaY9gDAVzw8f/740tK0pyDH1lRVpT0C\nGyWEULcsr6hYXlGR9hQQiF2jAIQmhACEZtco1C1NCgsb1quX9hTk2Jqqqi/Ky9Oegg0TQqhb\nTmnd+sRWLdOeghx7ffHiodM/SHsKNsyuUQBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IA\nQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEIT\nQgBCE0IAQhNCAEITQgBCK0x7gLpo7uyPSufPS3sKciybZNMeAaiLhHADKisrKisr0p4CgHwQ\nwq/o06dPt27d0p4if8aPHz9r1qy99tqrR48eac+SP61atUp7BKAOEcKvCNWDJEmmTp06a9as\ntm3b9u/fP+1ZANLhYBkAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNC\nAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCq6UQzr7x4Ewmk8lk\n2gx8uXaeAQByoZZC2O6y17LZ7JuX7l07ywNAjtg1CkBoQghAaDkL4Yf/fFew67B5uVoTAGpb\nYa4Wan/Za9nLcrUYAOSJXaMAhJazV4Rf8+iZRSffszpJkuSwzM31Tn+k4r5e/7yrvLx87Nix\n67dcuHBhLc0AADWqrRD2GlOWHbPhu+rXr9+vX7/1N4cMGVJLM8C26MWFCz9asSLtKcixReVr\n0h6BjaqtEAJbZuaKFTOFEPJICKGuOPDAA1euXJn2FPnz4YcfLl68uFmzZu3bt097lvwpLPRb\nt87xvwTqiptvvjntEfLq4osvnjBhQufOnYcOHZr2LITmqFEAQhNCAEITQgBCE0IAQhNCAEIT\nQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IA\nQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEIT\nQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IA\nQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEIT\nQgBCE0IAQhNCAEIrTHsA0tS9e/eSkpK999477UEAUiOEofXo0SPtEQBSJoRAOuyQoI4QQiAd\ndkhQRzhYBoDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCE\nEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0DLZbDbdCe699963\n33473Rki+/TTT3fbbbe0pyCiVatWrV69uri4OO1Bgmrbtu15552X9hR1QvohJEWVlZX33HPP\n2WefnfYgRDRz5sz58+cfddRRaQ9CdHaNAhCaEAIQmhCGlslkmjRpkvYUBFW/fv2ioqK0pwDv\nEQIQm1eEAIQmhACEJoQAhCaEca1+/+7+32rduEHjNof+8J4Z5WmPQyizbzw4k8lkMpk2A19O\nexbCE8Kw5vzhBxdOP2HszE9n3vOdqRecM/yTtAciknaXvZbNZt+8dO+0BwEhjGvuw/e/fcLF\ng7q1LG7ZbdDF33713odL0x4JIA1CGNWcOXNat2+/9kNcRe3bt54zZ07KEwGkQggDy2Qy677y\naVIgLiGMqqSkZN6MGWVJkiRJ2cyZ/ygpKUl5IoBUCGFUe/Tu0+mJoUMmli4pnThk6JOH9O3d\nMu2RANLgFGtxlU0dcU6/wQ+980Wzzqfd+Kc7ztpnh7QnIpBHzyw6+Z7V627UO/2Rivt6pToP\nkQkhAKHZNQpAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBo/wu5vyyxEXZqKwAA\nAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 300,
       "width": 300
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 5)\n",
    "loeuf_plot = ggplot(test, aes(x=as.character(sign), y=residuals, fill = as.character(sign))) +  coeqtl_fill + \n",
    "    geom_boxplot() + plot_config + stat_compare_means(method = \"t.test\", size = 2) + ggtitle(cell_type) + theme(axis.title.x = element_blank())\n",
    "loeuf_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "id": "2b189afd-21e6-4493-bcde-442faa71a08b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/I2/I2_LOEUF_Score_residuals', cell_type,rm_rib_mt_genes, '.pdf'), width =2, height = 1.8)\n",
    "loeuf_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef0e2552-6031-489f-bb5a-078fdf9ffa39",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "03b167c1-ebce-4098-b933-c4606d39132d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Residual Model for GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 556,
   "id": "4007197c-5ef1-4a3c-a8a7-a90ef88c0728",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Estimate the model (use residuals after correcting for expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 557,
   "id": "8ac1f18a-471f-4ee0-820d-7120b29fb1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_form = 'residuals ~ GO_BP_count_400_auc'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 558,
   "id": "bff72264-d237-45fb-b1b9-e9f5a0a800c5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"glm.fit: fitted probabilities numerically 0 or 1 occurred\"\n",
      "Waiting for profiling to be done...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "test  = coegene_data_scaled \n",
    "test$sign = as.numeric(test$sign)\n",
    "\n",
    "res = residuals(glm(sign ~ weighted_mean_expression, data = test, family = 'binomial'))   # select the beta of the random snp instead\n",
    "\n",
    "## Run model on residuals\n",
    "        \n",
    "test$residuals = res\n",
    "\n",
    "result_model = glm(as.formula(model_form), data = test, family = 'gaussian')\n",
    "\n",
    "#print(summary(result_model))\n",
    "## Extract model results\n",
    "confidence_int =confint(result_model)\n",
    "model_summary = summary(result_model)\n",
    "\n",
    "coef = cbind(as.data.frame(model_summary$coefficients), as.data.frame(confidence_int))\n",
    "coef$variable = rownames(coef)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 559,
   "id": "13111e26-d801-4eac-a72b-e5da1398048f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Estimate</th><th scope=col>Std. Error</th><th scope=col>t value</th><th scope=col>Pr(&gt;|t|)</th><th scope=col>2.5 %</th><th scope=col>97.5 %</th><th scope=col>variable</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>(Intercept)</th><td>-0.1186005</td><td>0.009090192</td><td>-13.047081</td><td>1.316347e-38</td><td>-0.1364169</td><td>-0.1007840</td><td>(Intercept)        </td></tr>\n",
       "\t<tr><th scope=row>GO_BP_count_400_auc</th><td> 0.0395686</td><td>0.009090625</td><td>  4.352682</td><td>1.357537e-05</td><td> 0.0217513</td><td> 0.0573859</td><td>GO_BP_count_400_auc</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Estimate & Std. Error & t value & Pr(>\\textbar{}t\\textbar{}) & 2.5 \\% & 97.5 \\% & variable\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t(Intercept) & -0.1186005 & 0.009090192 & -13.047081 & 1.316347e-38 & -0.1364169 & -0.1007840 & (Intercept)        \\\\\n",
       "\tGO\\_BP\\_count\\_400\\_auc &  0.0395686 & 0.009090625 &   4.352682 & 1.357537e-05 &  0.0217513 &  0.0573859 & GO\\_BP\\_count\\_400\\_auc\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | Estimate &lt;dbl&gt; | Std. Error &lt;dbl&gt; | t value &lt;dbl&gt; | Pr(&gt;|t|) &lt;dbl&gt; | 2.5 % &lt;dbl&gt; | 97.5 % &lt;dbl&gt; | variable &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| (Intercept) | -0.1186005 | 0.009090192 | -13.047081 | 1.316347e-38 | -0.1364169 | -0.1007840 | (Intercept)         |\n",
       "| GO_BP_count_400_auc |  0.0395686 | 0.009090625 |   4.352682 | 1.357537e-05 |  0.0217513 |  0.0573859 | GO_BP_count_400_auc |\n",
       "\n"
      ],
      "text/plain": [
       "                    Estimate   Std. Error  t value    Pr(>|t|)     2.5 %     \n",
       "(Intercept)         -0.1186005 0.009090192 -13.047081 1.316347e-38 -0.1364169\n",
       "GO_BP_count_400_auc  0.0395686 0.009090625   4.352682 1.357537e-05  0.0217513\n",
       "                    97.5 %     variable           \n",
       "(Intercept)         -0.1007840 (Intercept)        \n",
       "GO_BP_count_400_auc  0.0573859 GO_BP_count_400_auc"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 560,
   "id": "935808c2-5cc5-4419-bbdb-0ce0a690a8a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef$sign = 0\n",
    "coef$sign[coef$`Pr(>|t|)` < 0.05] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 561,
   "id": "d4389bf3-134a-40f0-a1bd-79c912f5a6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "coef = coef[coef$variable != '(Intercept)',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 562,
   "id": "031082ee-0270-4abb-bbff-d1c29b8d1e41",
   "metadata": {},
   "outputs": [],
   "source": [
    "###  adjust label\n",
    "coef$variable_label = coef$variable\n",
    "coef$variable_label = str_replace(coef$variable_label, 'GO_BP_count_400_auc', 'Count of BP broadly unrelated GO terms')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 563,
   "id": "7546ed96-8c3e-498b-b8ca-e2700b0789fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLAAAAJYCAIAAAD9hIhNAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nO3debyWc/748c/dOamcTjvaFK2WFkKylCJLDWPLbghNMiWN/WsNw8zQGGMaP5Ky\nZAZlspTdSEKyZBtCkdJe1vY65/79UdE51VFUx/R+Pv+6z33f1+d63/f9uOh1rvu+TyabzSYA\nAADiKVPaAwAAAFA6BCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShADAuow4\ntfzhQxavfgGALYogBABWmTf40HKZen3GFJb2IABsHoIQAFhp6pBB73XussODg55bWtqjALBZ\nCEIAYIWJ9wz+/KSL7zhrz8cHPTa/tIcBYHMQhABASill3xp09+LTztinWpeuHZ8b9NBXpT0P\nAJuBIAQAUkqFowffV6Vr12Yp5R1x+uFvDLp/WmlPBMCml1vaAwAAvwBLnh70zy/e/LJx5v9W\nXvHuPR/3uqxURwJg03OGEABI3z06ePhON3+RXWn5i+fOHDzo7dKeCoBNTRACAF8+NHjkXqee\nVGfVzzltTz+l4N5Bo/35CYAtXCabzZb2DAAAAJQCZwgBAACCEoQAAABBCUIAAICgBCEAAEBQ\nghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQjamF154obRHAAAA1pcgZGN6\n9dVXS3sEAABgfQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQA\nAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAA\ngKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABA\nUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAo\nQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQg\nBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShAC\nAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEA\nAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAA\nQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICg\nBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCC\nEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEI\nAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQA\nAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAA\nBCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACC\nEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJ\nQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQh\nAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAA\nACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAA\nEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAI\nShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQl\nCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKE\nAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIA\nAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAA\nQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAg\nqNzSHgA2pvnLlp85+vUVl1tUq3zF7ruU7jzw0/R+dfzMhYtTSnllcwa3a13a4wAAWyxByBZl\naWHh0M+mrrj8zdJlpTsM/GRPTp0x8dv5KaUqW5Ud3K60pwEAtlzeMgoAABCUIAQAAAhKEAIA\nAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAA\nghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABB\nCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAE\nIQAAQFCCEAAAIChBCAAAENQvNwgXjP1jxwaVy2ba3z735ywzrEumy7CNNdP3a+ae+EhKKY04\ntfzhQxZv5NXXYWPt6/vhAQCA6H4sCJdNfuyq41vvWK1ChSr1d+t8wZD3v/tZu3vkxNyjHli+\nPvdc8thNN+Vf/fGi7KgeNb6/clSPGplMJpPJ5G5dvWHbswZ9sHS1K8tW3K7pQb2HTiz4WRP+\nL1r/Z3Udlk1+vO+J+zSsUaFcxVpN9+5y+dCP5qeUUloycejFR7ask1+uXMXaLQ6/aOjHSzb6\nrgEAgNJTchDOG37G/l2frtXnwfHT5055eUif+s8MeHze5hls9hdfVNu15Xa5xa9vevV72Wx2\nyZx3BrR8tfvlQxf8cOWiGWOurzfitHOHbKYJtxjzhp+x/+lPbtPz/nFTvpz17sj+p1Qc2v/R\neSnNffiMdj1eanr1k5/MnTvxyat2Gd293Rn/nlPa0wIAABtNiUH4yZ3XDG3wh4f/dnLr+lXz\nKtVtdkjve289uXrh1JGXdtqlZn7+Nk0OPHfoxKUppfT1wI41eoxasdXyB44qf+qIlFIa1iVz\n2KV/Pb5l7cqV67Tp+cjEuw8/+sGCR08qm8lkDrt7/up7Krbm9P7t6108btL1u2cyu/1h4lpn\ny6SUKlWrVvaHa3LzG3c59aDyH3zwabG7fvfajcc0q1W52o7tzntkSuGqwS7u16Vl3fxmV761\nxsMZ06tmJpPJbJVfs9mvrnp+9opFCib/u3e7HapVrtX82H7jVp994p9b1+89euVpydkDO9Y8\ndcTC1W5er2fmi6IjvZ1SmvvKzSe13rFaXn7NXTtd9uS01VYsNt7iNZ/VNbZd5/Crvcp/P7VN\ng23y8rZptNfR//fQ30+pniYOvO7hRtcNvfGYFjXz82u27HLjQ30bD7tuwGovR/Fdz33l5gf/\nX7/V9lvsQQ3rkunY+7pfN69VKX/79n1feHdYr/22r5hXvdlxgz4qSEvfv+OUPetW2qpMJpPZ\n4cI31vqiAwAAG1VJQVgw/q13G3XoULfotTPv6XbKEzv95dVp09++vc24bqfeWry/inhpzFe/\nG/nxjA/v3O2p3rftMmL4CTlH/mtZNpt9qmvFEtb8Z+dRn/1xj4aXj89m376iUZEFP7qmeSaT\nya1Yt9OIZn+7utNWq407f+KwIc8v2nnnBsVm+M9zM7o/8fEXbw/Ye/SZ59y/8hTXqBfn/W7E\nR1+/3/OdNR7O/v1nZrPZ7MKZb93R6tnf/umVFRN27/bKvne988VHj5855annVlu9Ubeeuwy9\n44lFKaU0cfAdn//m3M5bl/SMrO2ZuXnc6iNdt1v6cliPc17teMer076a/fbdB7722/OH/xCZ\nxcZ7q2uxZ3XNbT9d5/BpxavcsH37Yq9ySgVvFX/163fosMM7b43/4S255YvseumwHue82vDI\ns4vOvNqDSimlsW8XXvTMxGlvXFLh5s6/frjFbe/Mmjry15+ef/3I5c/0u2TGif+ZuqAgm81O\n7rfn+jyFAADAz1NSEGaz2TWvLHj1uVGte13TacdK+XUOvObiQ9569sWSPlbYocdl7etW3Lp2\n5yP2nTtp0rK132lD1lzxltGChTNePufLC4/v93FKaWUllttun/+b3OnuW39Tvdgmbc+58rB6\n+fn1Du57bpsXnhu74kEd0OOyA7fPy1nbruc8c/WvW9apUrFSnf3/MPazDz9ckFLBK8+N3r93\n34O2z6+0w6/69mq3+vLVT+h56IsDHpqXsm8NGFTmdz33Xr8v6lnzmVk5UkqpcMxjI98d1q1V\nra3LbV2r9cX/mfbOOz9k91rGW92a2741ZN3Dp5RNKWUymRU/jL2wbiaTyWQqdh2xlte+RCv2\nO/wffyo68w8PKqWUUvvul7StlZff9IiOTfMOO7t7y2p51dr8ql2FSZPmNG7e5IPBl1x+/W3D\nXp6ysIT98IuwuKDg0+8WLCssLO1BADaZZTNOH/n++xv6v8NfpuUzzxz53ttbxmMBNraS4iW3\n1R4tJo4aNa3otUUqMZtNmUxKKScnp6Bg5amjRYsW/XCH8uXLr7iQk1OmoKDg+/JYjzVLHLtC\nzb3OOGbX11aWY9Or38tmly+Y88kL/zixSU7xO/+w2mr7yc/PX8euX/lj1wdr3zB60ldLCgse\n/U35ZcvWyNhioVy+c69TJwy4Z9Kztw/bpfeZOxS97/o+M6uNtCLF2/efk/3ehL7NVm23lvGK\nPKslbbuW4VPu7rs3//5VbtPvi2z2s5v2Tinl7t6q2Ks/ZdSoyS1b7V7k+f1h1yv2e/n11xfd\n7w8PqsijzsnJWe0JyCkoKGh6wYvjBpzQeNm79/bYq0XvFxYlfpkWLF/eY8ybeXf/u+GDIyvd\nM/yqN98vXNtvjgD+10384IM5jRo3+7F/kayHub3ufygz8KHMw/+dUPymgv9+8la7B4ZXvO+J\nDq9MnZxNKaXlC2Ze/swzde8Zmhk4vMe0NVdbi5nT3jtk6KOV7nui42szZqSU0uwe9z2UGfhQ\nZuBDmbvHPZVSyq15fq1Z1366jl/MA7Gt8aUtq2vc7cpjbu3e5fyqf+9z6C5VFn469uGbRtS4\n+OD24/569ZMHX9u23Jt/+suzrY69MT+l1KBB2T8/9vzMffZeNvaPA0anhueufcXKlStP+2zS\nkmzTcqv/BzZn34Pbn1l8zYdKmqxw8ezx9zzyVu2dr8lPP/qtoqNvu+6pQ6/dP712Tf/XOpx/\nb5H/tK9l17mTFuWWr1Rp64IZo/veNHxxuX1TSjn7dmzX7da+z3e4onXh6L7/eCntct5qi5TZ\n85zuyw/tcklBp5v7Vy628/z1fGaKyNnv1527XXzuXa1vPqlVra2LFu6iRWuMV+RZXXPbkodP\njX971bF/796lT+Vbzju0eZ28JTM/nPxNJpNJjbtdftQt55xwaa3b+3Ssn/ns2VvO7jvhmAHD\nir6D94ddr9jv41U79VxYUHzm9VOh3n4nntfm8KN3+XKPv42fkzrU+wlrFPHFgkUDJpT4jmY2\n3L2fTH551so/BbOksPC68R+899U3nerWKt2ptjzfLvXtvVC6vhsyMefIX5XbGEvV6H/K8f2/\nnbDbs2v8iyU7++FZVW88okWLzNd9nx599dS699Rb+NfnX3mseuvnT6rddKv1e8vR8hkXjZre\nssOhQ6vO/9PTr1z0eech9VMqs/3wbvsctdq9mu1Qc8K4qV83bFBlYzwkYEtSYhCmGsfeM6bM\nNRdff2zLrrOy2zbdp8uFN52x6y5t7//otxe0qT2lsFbrEwYO6d0gpZTaXXTrfsef2qRWmYa/\nuujQ/dMX61hwn64XVjp578qXf9N+0HerfYyw5ukDi685fR0rfHRN88w1qUz5qnWbH3z50L5t\nUxr1Y4/xwINq3dG58QnTyrc8bdB9p2xT9MY1d91gyi27H3fsjtuW2X7vUw8/uPr4lXcbMPDt\n35zVos6SOh0v6NwxTSm6Sv2zejS84ubW/+qwVSpuPZ+Zomocd+eIWedfeHLL3p/NWViQ8k5/\ncn6XlTd1uGCN8Yo+q2vZ9q4Sh69x7N1jcq679I8n7nnW9MUV6zRudUjPJ284JKWyx90zuvCK\nC648pNHxc1P1Ju1Ovm309V22LbptkV3fOWLWwSf0bbnNgFX7PXy9HmxKKaW5t3fc/pznU6Zs\nfs2dD/z9rWf+7BpMKX3w9Tdnj/H1NJvQineVPzJ52iOT1++X2AD/K5bNG7u0yjUr3sqyaOL+\nT3zZrvrXgz9fUGGbBrd3aHlIhe/vN6vbvS/etXT1LTOnHHzckPrrt5dMrav2TylllywqU7ZM\nuRrlM+mbKfd+Xfv6znWbFvkH2vLXP3iz17vT31+Ss3O9XQe1a9hi9d+7zp35TOUG79YuXzmV\nv3inKk2+mJfqZ1J2+un3PJzK5h/SrNWAFjWqppSqVm0+Z/q41OCQn/R8AFuwzFo/KMgGyS6Z\n8eg57W/fd/RT3bYr7VlKwbBhwxYsWPlJxtdff71///6lOMyXS5ZWv++RUhwANq4qW5X96rSj\nS3sKiGf+R3s9vfy+Y3fdKaW0aOL+/5rU+tC2122XGfvWmNOXtZi633Yb/E7Sbyfs9mzBAysW\nLGLaiXe9/GC2zI6NWo8+oF7d6W9Web3w7Aqz/jFj+XY16vVrv/vRFTNp9jtNx2YGHLRLm3LL\nxrzx0tUV9xvTLO+HBSaPzZ1Ye1nHepmU0uevlZ9Ya+FB9cqklFLh3C+n9nn2ncrtjvhHrUxK\ns8++7+MDTtn/5PU77wjEUfIZQtbD5H57Nbxy5oHXjTgtYg2mlLp06fL95WnTnCkC4H9fuXLV\nly765vsf82t3rbN1XkoHNalTacy3c9J2q94w8/POEKaUUp0HzjrunsXfDh7z0ukTtnl+261q\nfD23RttD5lUufH38S4e/OePIA2p/Oe/rj2fPav+vlZ9ArNTk27ffGL3729+lVPWmEw6+sHy5\n6ouXzE5pu5S+WrSkYrlyq4qvTI1q9Xvu8N5F8xanWhVSdtnXqVwNNQisQRD+bDtc+EbBhaU9\nBGuxS5XK5zVrXNpTbGn+OenzF2es/OMtmZTJpuxxDbbvWDvob0M2nSvfeH/24sWlPQUEVrZ6\nm60mv7Mk7b3iU4Tfzbh3eqNrt82M+2T6d1Wbr/bpk+0Gnnb8wJ+8lzmf/H5OlT6Nqm+TzRZm\nCxYuL0xVa3XKm5vSir+3nMpkMiml6tWrNK1ZfUD7pvtULLvq7y/Xyn7/B5oqbHfw1+/3m779\nlVXn3zjhq0N3+/7L1gvnfjX1H5MLW7Ytn1JKX3313jbVW//kUYEtlyBki1U3r0L3nYr/VUp+\npt80qn/Fm+/d9sGkxQUF+WVzLtttl4tb7rQRvoSPom56d8JsPQilKf+URgXnTV7SvWm5lFKq\nXGWrj19q+OyC8jUa3HXghr5fdPEtjz/2+1kppbTzwP/u1+aIMc2+7Xbvf3c77sBe1eu0/+yN\ngx+cMyNTYY/6zQfsnJfK5F3TruaZo5669rvCWtvucO8BtcqklLbd9f6Gb/UeOfKt+UsXZ9NB\n+x353M6rfdtNbu0bD5h72pin6iwpu1eTvYbUz00zx9cc8cmsVKZqXuVDmu37h9qZlNJ/J89s\n2qidb5QB1uQzhGxMN9xww2WXXVaKA6z+GcJD6tR8ulO7ku/PT1OQzc5etKTm1uWl4CbS+KEn\nJn47P/kMIZSiZTNOf2behZ2bNV88cf8nl959zC6NfnybX6rlM7s9Pbtnpxa7e8sosAZnCIEN\nlpPJ1Nq6fGlPAbApla11z6+2lL+pk1tz4K9qlvYQwC+UIAQAWLcKjcYcU9ozAGwy3joAAAAQ\nlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhK\nEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUI\nAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQA\nAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEHllvYA\nsDFtVabMcTtuv+Jyi2qVS3cY+Mk6bV9r5sLFKaW8sjmlPQsAsCUThGxRKpbNfeigfUp7Cvi5\nbt1n99IeAQAIwVtGAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShAC\nAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEA\nAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAA\nQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICg\nBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCC\nEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEI\nAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQA\nAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAA\nBCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACC\nEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJ\nQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQh\nAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAA\nACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAA\nEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAI\nShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQl\nCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKE\nAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIA\nAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAA\nQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAg\nKEEIAAAQlCAEAAAIShACAAAEJQgBAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCU\nIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQ\nAgAABCUIAQAAghKEAAAAQQlCAACAoAQhAABAUIIQAAAgKEEIAAAQlCAEAAAIShACAAAEJQgB\nAACCEoQAAABBCUIAAICgBCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAA\nAEEJQgAAgKAEIQAAQFCCEAAAIChBCAAAEJQgBAAACEoQAgAABCUIAQAAghKEAAAAQQlCAACA\noDLZbLa0Z2DLcd99902fPr20p2CzmjdvXvXq1Ut7CvglcnTAunz99dcVK1bMzc0t7UHgF2f5\n8uULFy6sVKnSRlyzSpUqZ5999rpuFYTAzzJo0KAzzzyztKeAXyJHB6zLo48+2rZt22rVqpX2\nIPCLM2vWrDfffLNz586bbY/eMgoAABCUIAQAAAhKEAIAAATlM4TAz7J06dKtttqqtKeAXyJH\nB6zL0qVLc3Nzy5RxZgKKKywsLCgoKFu27GbboyAEAAAIyi9mAAAAghKEAAAAQQlCoCRLPry7\na+s6eeXy6u7d7f5Plq33rQUf/HnfCpmafcZszmFhc9vAA6Tg06F9Dmq6bV75vG2aHthn2KcF\npTI0bBYlHh2T/7RnJpPJZDJ1Lxy76rrsV28O7HlQ020qVq7d4ojLn5rmM01ssTb46EgpLfno\n/l4dG1cvn5PJZNr0+2KjziMIgRJM+cdZvT867J+TZk26/9D/9uw+YMb63VrwQb9eLzQ/pslm\nnxc2qw08QJaPvOL0xxv1e332N7Nfv7HhY6ddMXJ5KQ0Om1zJR8cOl76RzWbHX9Jwtes+G3r7\nuGZXPP7hzM+fv7jSP4/t9c/5m3di2Fw2/OhIy17+v1+d997+/xg3c3E2mx17Yd2NOpAgBNZt\n6sMPvnPYRZcdULNKzQMuu+iQcf96eOZ63FrwwZ96/qfLbb12zJTO1LCZbOgBMqf29rVzUsqk\nTMqklFN7+zq5pTY7bFolHx1r1aD7nQPO6dCkRsUqO598YrvstGlfbo5BYUk3PawAAAh6SURB\nVLP7CUfHokduvq18nzv6HtKwyqb47lFBCKzblClT6jRqVD6llFL5Ro3qTJky5UdvXf7en3q+\n1OW2cxrIQbZ0G3qATN3zqntPHX9a/YrlKu7Q9Z3f3HfVHqUwNGwWJR8dJSucOviO51udfHS9\nTTUclKqfcHR8+sEHWzedP7ht7by86g0P+N0DH2/c95cIQqBEmcyqsFvbH6lZ49bl7/255+hj\n/t/vGspBQtigA+S7kecce/9ud330zcJvJtzZYsgx54z0lji2YCUfHeuSnflYj8Nv2bb/0PMa\nb5qx4BdgQ4+O5cuXf/XMa/l/Hj979vjb93q1W7cBG/VDhIIQWLd69ep98ckni1NKKS2eNGl6\nvXr1fuTWySOHv/RM7wZlMplM8+s/mvW3tpnD7/ZvXrZQG3qALHhxVG7ns45pUqlCpaZdunXO\neXH0h6UwNWwOJR8d61Iw5aGuB11WcM3Tdx1Vy68V2VL9hKOjXv36md2P677vdnl59Q7+7VE7\nvvvuhI05kSAE1m37Y45r/tRNN7w48+uZL95w09N7nXhMzR+5tdGlb2RXeu/yptud91J2RNeK\npTY/bFIbeoDst+uuXz4x+JGJ3y369pPhg574ulmzhutcG/63lXx0rNXyz4ac3PHarW54duBR\ntdUgW7CfcHRUPfrkwz58aMArsxYumPrsncM/b9Vql406UhZg3Ra9f9dv9qhVoezWtffqeu+E\nJdlsNpv98Opdc04Zvq5bv7cyCGELtsYh8MPRsbZbCyY93KfjTttWLF9x250OPn/4pwWlOjxs\nUiX/72P4KeV++MdozgnDs9ns+MuL/Iakznmvlub4sAlt8NGRzS7/7OFe7XaoVG7rGk06bvT/\nfWSyG/C2bgAAALYc3jIKAAAQlCAEAAAIShACAAAEJQgBAACCEoQAwCa0YOwfOzaoXDbT/va5\nRS4D8EsgCAFgM+lRI5PJZDK5eds2PqDHPycs39z7H3Fq+cOHLP5Jmy6b/NhVx7fesVqFClXq\n79b5giHvf7e+Wy557Kab8q/+eFF2VI8a31++cEzddU2yoUM+cmLuUQ9s9qcSYIshCAFgs2l6\n9XvZ7OLZY/s1fe6M3vdu7rNkhw9ZPOLU8j9hw3nDz9i/69O1+jw4fvrcKS8P6VP/mQGPz1vP\nbWd/8UW1XVtul1vkcgmT/OQhAfgpBCEAbF65eQ2OPPHAvA8/nLTyirmv3HxS6x2r5eXX3LXT\nZU9OS2n5p0N7td2hauVazY/td/FhuSc+klJKw7pkDru4X5eWdfObXfn2GpuktPT9O07Zs26l\nrcpkMpkdLnyj2I8prTr5Vjh15KWddqmZn79NkwPPHTpxaVq1+KV/Pb5l7cqV67Tp+cgXRQb+\n5M5rhjb4w8N/O7l1/ap5leo2O6T3vbeeXH2t6xR7LO/0b1/v4nGTrt89k9mtT58fLrdps+o0\n4JyX/3Ly3g2qVyiTyWSOGrJ49TOEa3mMxeecePfhRz9Y8OhJZTOZzGF3z9+0rxvAFkkQAsDm\nVbDgs0cf+M/S3XdvklJK6cthPc55teMdr077avbbdx/42m/PHz5p8G+7j2s3+L0vPhpx1pSn\nn/thy1EvzvvdiI++fv/3E4tvsjBln+l3yYwT/zN1QUE2m53cb4+iP+75/Roz7+l2yhM7/eXV\nadPfvr3NuG6n3vrpyhteGvPV70Z+POPDO3d7qvfN41afd/xb7zbq0KFu0Uex1nWKP5Yb6jzx\n2R/3aHj5+Gz27VtuGfX95SsarVhk7oPdj7y97Ln/nvDl8mw2+8jqZwbXfFoWrjnnbbuMGH5C\nzpH/WpbNZp/qWnGjvDwAseSW9gAAEMdH1zTPXJNSytQ+ZeibJ1dNKaXCMY+NfHfYw62GdVt5\nn6b73p/zcrtzn+hQt1xKnfv2bHfbf1ZtfkCPyw7cPi8VPl98k7afpqMbN2/ywaBLLp/fsf3B\nR3Ter16xH7deedeCV58b1brX0E47VkrpwGsuPqTfgBe/u7BBfkqpQ4/L2tctn1LnI/adO2DS\nstS67MpNstnsmo9krevssMZjaftpKuHtnwUvj3iy5fkzTt2t6ho3rfm0tP00Hd1szTnb/fiz\nDsC6OUMIAJtN06vfyxYunftW/z1fvPKv45aktCK32vefk/3ehL67riXAUkop5efnr2OTZik1\nveDFcQNOaLzs3Xt77NWi9wv1iv64aNUaReoum02ZzMrL5cuvTLecnDIFBQU/3Cm31R4tJo4a\nNa3oLGtbZ+2DlWD1/Re/ZV1LFZszs44FAFgvghAANqtM2eq7/27wn3ca3OuWCdmUcvb7def3\n+5171+szFq6MsJx9O+w7uv8fXpw2/7spT11720trrLDmJitUqLffiedd+5f+l7b7Zuz4OcV/\nXLX4we3H9b/6yc++mz991DV/ebZVx3b5PzZx425XHvPJFV3Of+CNKV8t/Hba+8/cenrvB5uu\nZZ11DbYuOfsf0entmy/41ztzFxeu52NcQ+XKlad9NmnJuhoagJIJQgDY/Kqd9OdLMjedf9+s\nlGocd+eIC8rdfXLLbcplMplMxa5P1TpzwO2t/nParnWadL6z2v77VK5cudjWa26S0tzbO2Yy\nmUyZCrXb3Ljs9zf9+okiP55Zb9W2NU8feH/nDy9oU7tmy+6v7DFwSO8GPz5tjWPvGTPwwCn9\njm1Zq1q9fU756+SO3Y/YdW3rrG2wEhc+YcAjv11w868bV85d9aUyJT3Gtdmn64WV7ty7co4v\nlQH4STJr/WAAAFDassu+/Xz0jaeeNP6sD0aeUaO0pwFgi+QMIQD84jzQZcXZvb17jm11+99P\nU4MAbCLOEAIAAATlDCEAAEBQghAAACAoQQgAABCUIAQAAAhKEAIAAAQlCAEAAIIShAAAAEEJ\nQgAAgKD+PwIMIflHerxUAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 300,
       "width": 600
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 10,repr.plot.height =5)\n",
    "model_plot = ggplot(coef, aes(x =reorder(variable_label, Estimate), y = Estimate, color = as.character(sign))) +\n",
    "  geom_point(size = 1) +                                  # Plot points\n",
    "  geom_errorbar(aes(ymin =`2.5 %`, \n",
    "                    ymax = `97.5 %`), \n",
    "                width = 0.2, size = 0.8) + scale_color_manual(values = c(\"1\" = \"#009ba9\", \"0\"=\"#82898f\")) + ggtitle(cell_type)  + coord_flip() +plot_config + theme(axis.title.y = element_blank(), legend.position = 'None') + ylab('Regression Coefficient') + ylim(min(coef$`2.5 %`), max(coef$`97.5 %`) + 0.1) +  geom_text(aes(y = `97.5 %` +0.05\n",
    "            , label =paste0( ' (p = ',  signif(`Pr(>|t|)`, digits = 3), ')')  ), size = 2)\n",
    "model_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "id": "d36934db-c203-49d2-b9ca-8adefbd6981a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/I2/I2_Residual_Model_', model_form , cell_type, rm_rib_mt_genes, '.pdf'), width =3.5, height = 1.2)\n",
    "model_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "id": "ebe3b5bc-b63e-44dc-86eb-9ebf848bab24",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Save data of model\n",
    "write.csv(coef, paste0('figures/I2/I2_Residual_Model_', model_form , cell_type,rm_rib_mt_genes, '_data.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d31d8e31-c8e4-44ec-ab0e-2f232139f68f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## GnomAD Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "id": "f17e7410-255e-4beb-9c95-7914a01a1df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "coegene_data$sign = as.numeric(coegene_data$sign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "id": "03b33fab-07c2-458b-84c1-d15efa72103c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### LOEUF score from gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "id": "d5088aa8-442c-43a1-9162-49d014ffacb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_result <- t.test(LOEUF ~ sign, data = coegene_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 569,
   "id": "6ef34b1a-4e75-4dad-aedf-d14c0243345e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4.84425680642375e-50"
      ],
      "text/latex": [
       "4.84425680642375e-50"
      ],
      "text/markdown": [
       "4.84425680642375e-50"
      ],
      "text/plain": [
       "[1] 4.844257e-50"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "t_result$p.value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 570,
   "id": "af30097f-c06a-4537-afa9-495ae8b25376",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 101 rows containing non-finite outside the scale range (`stat_boxplot()`).\"\n",
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 101 rows containing non-finite outside the scale range (`stat_compare_means()`).\"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlgAAAJYCAIAAAAxBA+LAAAABmJLR0QA/wD/AP+gvaeTAAAe\nDElEQVR4nO3de5hVBb3w8bUBYZCLw0VCwFFh8HKE4pimAW9IahqPFlImIgUHjTKNOh7oQphE\nRJ7Q5D3p8VKiKKjUwQt2fLQ0wVeTyDsgII7IBmFEQBAdBhjY7x9gcR2YYbPX4O/z+cNns9fa\na/2GB/my1157rUwul0sAIKp6aQ8AAGkSQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQsiLPw4s\nOn9y5bb/pj0LUBNCCDW3+s5zG2VKvv/M1rQHAQ6cEEKNLZ08cU6frx47deITm9IeBThgQgg1\n9cakO5dc8oPbLjv1kYnTP0h7GOBACSHUTO7FiXdVfuPfPtvyq4PPfmLi799Lex7gAAkh1MjW\np++8p3jw4C5J0uSCQec/P3HK22lPBByYBmkPAIeUjY9PvHfZC2s6Z368/YlXJ71+1cjjU50J\nOCDeEUINrH/4zgdP/PWy3HZVM79bfufEl9OeCjgQQgj7b83v7/zf0wZe0v6jX9f/P4Mu3XL3\nxKe3pDkUcGAybswLQGTeEQIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQ\nWl0JYTabLSsrS3sKAMKpKyFcvnz5kiVL0p4CgHDqSggBIBVCCEBoQghAaEIIQGhCCEBoQghA\naEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBo+Qth7r0XfnflWScc2fSIdp+84CeP\nvZ3bZfnG+XcN/kz7Jo2adDj98imLNudtvwBwAPIXwsV/uHV2l1GPzC9f8uQPmt/7lavu/WCn\nxdmbLxu28Lx7y94pm3LuvCuH3r4ibzsGgNrLXwg7Dv3t7Vf0Pr510+KTBvT/XO7tt9fsuHTp\ntKmvnDdiZK+2xW17jRzxhdn3TSvP257ZrnLy+ZlddLvujR1WWHXrmV1GL9jPrdVo5ZqqWvHE\nL77W7ahmzT7xL31GPfZOjZbuwTNXtc1kMpnMDgNXLXvsJxd8qkPzwzKZ4m8/kefxgY+Tg/AZ\n4dald9725CkDLizZ8clsNtu+tLQoSZIkKSotbZ/NZvO/5+iKBv4xl8vlcktv7HHMfzyXy+Vy\nuZd/VJryUFvWLvzznWMu79N95F92ePaV++7dePGkl5Yv+3/XHDFp8C+f2fk11S/dg543ledy\ni375qX8+8+aNF100/ehrnsxW5HJrbz07Pz8L8LGU7xDmyqd/+/wJbW76w/c677ook8l8tFJu\n1w8QKYC1t/a/Yua8n52UyWT6Tq5K1v/9pq+ffkyLJs3bnXrJLa9WJkn5I8N6HHtEo3qZTObM\nmxbsvPJuym/q2WXwyG/861FHtOh41og/7faubev6shmTf3nFl3t2/8roP6/v8q3fPTnu8zss\n/vTVE8d85VNtmx1xfPduRxUXF+/84j0s3W3a6i287+75F/zihq+eUHxYTX+bgGjyGsIt2d8P\nPmvklp89fkffozI7LyopKVm2aNG2v78qy8qWl5SU7GEDHEzF377/ll4nXzs/l8s9NLDB334+\ncHKHXz2zdE353/+z7e3fueWtD//026nHji/7YEsul5tx1Yk7rbzH7S18sfLr099YPu+3n35q\nyOgnd/63zXNXn9T5S//52qeueWTWk/f9ali/09o13tMmPpg99pv/2+vWq7vscQc7LN1t2n38\nrEuXLi0tevaiY5se3rJTr+8+uGTrPn93gLDyF8KqxZMHnD2m4bg//65vu8xuS4/ud1HXx8aP\nm1m+tnzmuPGPn9a/X9u87ZlqvDzqxEwmk8mcev1bOz2/6qUXX//bdWeWNCtq0uHsCS/PmTu/\nSd+f/rTp3V/9fO8+A3/8h4Ub973p484ffM7RTZq0O2vwl5rPm/fuTss+++vXFj38g5NeHnPB\nGWcP+OFvHnx++YZdX55757GrL/jxhz9+aPyZzXff+M5Ld592bz/Xdq1bt57/ZuuxL6xePX/S\nZ5/7t2sf3fePA0SVvxDOvWP07xfN+13fdvUymUwm0+H7s5IkSRaM7tJg4ENJkhxz1R0TOj96\nScc2nQY8duJNtw1tlyTZbHb2RxYuXOiA6cHQbeyCXC6Xyz0//Ngkady48bqV5RtySZK0+lS3\nE3peM/OttZu25nK53Lo7vpg0P/XK2x6ZMfOxm05/4bJrH9m6w8p7tvjRu59cVlGxYsbdj6w/\n+eQjd15Yr3mn3l8feev0Z579w0/PavLqLZedtdNnhBsX3tH//Altxz/8y94td9/ybkt3n3an\nn2t3Xb74xaOTJEmSTJIkSb16vi8L7F0uPe+8885rH5k6deoTTzyR4jAfJzucLLOL7D0DTmzR\nqF7y5Xs259Y//9+Dux/XqqhekiTJWbe8u3j86UmSJPUbtz7h3NF/Wb3zynNHnXj8T1/dYUsr\nftPj5EE/HtitbfMjju09/PHyGo74528dscMfwi/fs2Hbs5cd0eM3K/a4dNdpd97ciht7/HP9\nHjcuzeVyq58Z9+WTP9H08Jadz7r6kaVbazgfEEgmVzfeh82aNauiouLzn//8vlcl35566qls\nNltSUtK7d+89r7Hy1i8PaXjrH4cc9Y9nym/qed7au14elfZJqRzC9v0HDwpiz6dBEMqjjz76\n1FNP9e7de69/H7X59sN/LOxMBLDvP3hQEEJIrbS96pmX054BIB+cRABAaEIIQGhCCEBoQghA\naEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhC\nCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghA\naEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhC\nCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghA\naEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhC\nCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghA\naEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhC\nCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghAaEIIQGhCCEBoQghA\naEIIQGhCCEBoeQzhW9edmslkMplMh+Gzdl/6xthumX844/q38rdfAKi9PIbw2B89n8vlXvph\np72t0OknL+W2mTX82PztFwBqz6FRAEIrZAiz/3Vm08ZHtOvS5+oHy6oKuF8A2KuChbB01Mub\n3l/7/ntvPvHz0hmDLp2wuFA7BoBqFPjQaL2iVv9y4Y8Gd3nxhVdzhd0zAOxJgUO4tXL1aw9c\nd9fcT5/6yUxh9wwAe9Igj9t6aGDRhVM2JkmSfDZzQ/2LH6y6v2+yYHSXLm+MrZp84tguJ10z\nL6lXVNzuhP9zyT33fv+4JMlms+Xl5dteu3Dhwg4dOuRxGADYH/kMYd/JlbnJuzx34ui5286L\nGTU3N2qXZUVFRc2aNdv2uHHjxnmcBAD2Uz5DWFNt2rRp06bNtsfr1q2rqKhIcRgAYvI9QgBC\nE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNC\nAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBC\nE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNC\nAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBC\nE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNC\nAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQmuQ9gB10QMP\nPPDoo4+mPUXhLF68OEmSF1988fLLL097lgLp2bPn4MGD054CqBOEcA/efvvtl19+Oe0pCm3d\nunVxfuqSkpK0RwDqCiHcq8MaNmzXwV+XHzfvLH+7snJD2lMAdYgQ7lVRUeNup56R9hTk2TNP\n/UkIgR05WQaA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0IQQgNCEEIDQhBCA0FxiDYCC2rRp\n0+bNm+vXr19UVJT2LEniHSEABXbdddf16tVr2LBhaQ+ynRACEJoQAhCazwgByKdJkybNmTOn\nmhUWLFiQJMmbb745fPjwalbr2rXroEGD8jzcngghAPk0Z86cGTNm7HO1tWvX7s9qBSCEAORf\nUVHjFq1a1+61761eVcgbaAshAPnXolXrz/ToVbvXzn525oq3l+Z3nmo4WQaA0IQQgEPXxtdu\nH/DJTxS3OenCCS99WLtNVBfCX/X91XNJkiTLnr3/2WW12zwAHDwr7xn2y9xP/rb4pfFtbv7O\nLdlabaO6zwhnPzy7Y5IkyawbL7m/f/8eX63VDgAIaM3qd597+i+1e+26tWv2c83cX2cs7DP0\n4mNbJMd+u9/3rn1m4/ABjWq8OyfLAJB/GysrV5YvP9h7Wb9mTfPW285Nbd269erVa5LkqBpv\npPoQvr907ty5ydL3tz/Y5oguXY6u8X4AiKRx48NbHtmmdq9d8+7KDRsq9mfNZi1bvv/CqiRJ\nkmTVqlUtP9myNrurLoTNW734izPP/EWSJK2SF7c9SJIk+dKqVRNrsysAwihu2erUM3rW7rWz\nn5254e39CmGme6/j//32qW/9a8+5t047std3an5cNKk+hL+eO3fcP/fWsHnrlofXr80+AOCg\naPON//uDv/Q//Zh3Wpz54yk3HFOrbVQXwsuPOmrajr8+rPkJX/jeb3475pyaH4EFgIOgUZcr\n7pt7xQFtorqvT/xPbgdbN65b/NQ1n5h2yXfvP6AdAkBdst9njWYaNm9/yqXjv3dH56eSpP/B\nHAmAQ957q1fNfnZmrV+b32GqV8OvT1RVVdVzMRoA9qGyckMhrxd6IKoLYeUHH1T94xe5ze+v\nmPvItTcuOPu6gz8VAIeqrl27Vr/CggULysvLi4uLu3XrdiDbyZfqQjiwWbMdT5Zp0Oyorude\n9Yebv3KwZwLg0LXPu+mOGTNm+vTpHTt2vP766wszUvWqO845ef0OPqisfH/5i38Y2Wuv3498\n67pTM5lMJpPpMHzWHhZvnH/X4M+0b9KoSYfTL5+yaPOBjw4AB666EK5Z9n6u6UeaNKqfJEny\n7rwZ8/ay+rE/ej6Xy730w057XJq9+bJhC8+7t+ydsinnzrty6O0rDnBwAA5N9T6S9iDbVTfH\nsJOGPZ4kSfLAxQ0ufmD7czOv7X1tbfazdNrUV84bMbJX2+K2vUaO+MLs+6aV12YzABzqRo0a\nNXv27FtvvTXtQbbbnyBv3bJly9YD3E82m21fWlqUJEmSFJWWts9ma3e3DADIqwK+M81kMtsf\n5XK5wu0WAKpR07tPLH2/dvspKSlZ9tyiyqRrUZJUlpUtL/lcSe02BAD5VPO7T7T6Um32c3S/\ni7peO37czO5Xn7Dw1+MfP+3SCW1rsxkAyK/qQjhx1UcXudmyaVPSsOG+bj3x0MCiC6dsTJIk\n+WzmhvoXP1h1f99kweguXd4YWzW57zFX3THh1UGXdLxuXYtuX7vpnqHtkmTlypWrV6/e9tps\nNtuqVasD/nHg0DZ06NDFixenPUWBfPjhh0mSPPvss+ecc07asxTO6NGje/TokfYUKVuyZMm7\n777brFmzE044Ie1ZkmRfh0Y3vv4/o0f84q4/v1K+IWnc9lPn/Ns1N1zbr3Rv93vqO7kyN3mX\n504cPXf71WmKTh5y9/ND7t5hWWVl5fr167c93rBhQ23Gh4+XdevWvffee2lPUVCbNm3atGlT\n2lMUzsaNG9MeIX2TJk2aPn36Kaeccvvtt6c9S5JUH8IV91/a8zuL+46+7ckzT2yVrJ7/1B2j\nL+tRtuXVqRd9Ii/7LikpKSnZ/lHh1q1bKyr26zaM8LF3UrNmp7RokfYU5FNVbuvUpcvSnoI9\nqy6Et42dcfbvFt3eb/v/kMd/8rM92/c/fsxtyUU/LchsEFRps6YXtHPbz4+Vyi1bhLDOqi6E\nr7/+6XPO2emfpa3OPffUS584yCMBcAibNGnSnDlzqllhwYIFSZK8+eabw4cPr2a1rl277vOy\npXlRXQhbt84uWZIkXf75TG7x4rdatz7oQwFwyJozZ86MGTP2udratWv3Z7UCqC6E/Qds/uK3\nf9Jr6rW92zdMkmTTsid/+q0JuUv/VKjZADhUFR92WOdmTWv32kXrP1i7uXD3ZqguhN3HPvyr\nQf3O7/jfrUs7tk5WvfnG6rZfmfDwz88o2HAAHKI6N2v6H8cfX7vX3vD6639fU7jTp6v9+kTR\nyd+a+trXFj337EvZ95Pmx5zS84zSJhvXViSNDi/UeABwcO3zWqP1W3Tuef7XBgz42vk9Sovr\nJw9/o8U3CjEXAOxT1f19M5lMJnPG9W/Veht15XZQAFBjDfo/lMstvfGALtZT/ZVlAKA2Xnpv\n7Teff6F2r92wZUt+h6ledSFcOnfuut2eq+XdJwCIpCqXW19VlfYU+6W6EF575pnTd3+2dnef\nACCSow9v3OvII2v32pnvvru0onAXoN6/u08AQE20LSo6/6haXilw4fr1hQyhk2UAOHTN+n6H\nzNH//uzfRhyX6fD9WbXahJNlADh0nTFhWW7CgW3CO0IAQvOOEID8W7T+gxtef73Wr83vMNUT\nQgDyb+3mzYW8XuiBEEIA8qlr167Vr7BgwYLy8vLi4uJu3bodyHbyRQgByKd93k13zJgx06dP\n79ix4/XXX1+YkarnZBkAQhNCAAqqQ4cOJ5988nHHHZf2INs5NApAQQ0ZMmTIkCFpT/FP3hEC\nEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCa\nEAIQmhACEJoQAhCaEAIQmhACEFqDtAeouz78YP2f//ehtKcgzzZWbkh7BKBuEcK92rp1a8WH\nH6Q9BQAHl0OjAIQmhACE5tDoXtWrV6+o8eFpT0GebazcsGXLlrSnAOoQIdyrJk2bff68C9Ke\ngjx75qk/rX53ZdpTAHWIQ6MAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQA\nhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQm\nhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQA\nhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQm\nhACEJoQAhCaEAIQmhACE1iDtAYBdPbny3Vmr16Q9BfmUS3sAqiGEUOdUVFVVVFWlPQVE4dAo\nAKEJIQChOTQKdc7hDRo0qV8/7SnIp1ySrNq4Me0p2DMhhDrnrDZHXlpSkvYU5FPlli2D//58\n2lOwZw6NAhCaEAIQmhACEJoQAhCaEAIQmhACEJoQAhCaEAIQmhACEFo+Q7hx/l2DP9O+SaMm\nHU6/fMqizTsvfGNst8w/nHH9W3ncLwDUWh5DmL35smELz7u37J2yKefOu3Lo7St2XaHTT17K\nbTNr+LH52y8A1F7+Qrh02tRXzhsxslfb4ra9Ro74wuz7ppXnbdsAcJDkL4TZbLZ9aWlRkiRJ\nUlRa2j6bze66xn+d2bTxEe269Ln6wTI3HQWgTsjryTKZTGb7o1wut8uy0lEvb3p/7fvvvfnE\nz0tnDLp0wuJ87hgAail/ISwpKVm2aFFlkiRJUllWtrxkT7eRqVfU6l8u/NHgLi++8OqupQSA\nFOQvhEf3u6jrY+PHzSxfWz5z3PjHT+vfr+3uK22tXP3aA9fdNffTp34ys/tSACi0PB4aPeaq\nOyZ0fvSSjm06DXjsxJtuG9ouSZJkweguDQY+lCQLxnbJZDKZw1qUnjvmrbPvuff7xyXJypUr\n538km83ufjgVAA62fN6hvujkIXc/P+TunZ47cfTcbefFjJqbG7XL+pWVlevXr9/2eMOGDXmc\nBAD2Uz5DWFMlJSX/+CBx69atFRUVKQ4DQEwusQZAaEIIQGhCCEBoQghAaGmeLFPHbd68efmy\nXS8Tx6Fu06ZNaY8A1C1CuFeVGyr+/ten054CgIPLoVEAQvOOcA+uuOKKb37zm2lPUTgjR458\n+umnP/e5z40bNy7tWQqkfv36aY8A1BVCuAcNGjRo0CDQ78y2KtSvX7+oqCjtWQAKzaFRAEIT\nQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IA\nQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEIT\nQgBCE0IAQhNCAEITQgBCE0IAQhNCAEJrkPYAwK5Wbdw0Z926tKcgn6pyubRHYK+EEOqc51av\nfm716rSngCgcGgUgNO8IoQ655ZZbqqqq0p6iQMaOHfvXv/61e/fuo0aNSnuWwikuLk57BHYl\nhFCHtGzZMu0RCqdRo0bb/tumTZu0ZyE0h0YBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0I\nAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEI\nTQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0I\nAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEI\nTQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0I\nAQhNCAEITQgBCE0IAQhNCAEIrUHaA5C+3r17l5SUdOrUKe1BAFIghCR9+vRJewSA1OTz0OjG\n+XcN/kz7Jo2adDj98imLNtdoKRBN7969Bw0a1Lt377QHIbo8hjB782XDFp53b9k7ZVPOnXfl\n0NtX1GApEE6fPn2++93vOiBB6vIXwqXTpr5y3oiRvdoWt+01csQXZt83rXy/lwJASvIXwmw2\n2760tChJkiQpKi1tn81m93spAKQkr1+fyGQy2x/lcrkaLgWANOQvhCUlJcsWLapMkiRJKsvK\nlpeUlOz3UgBISf5CeHS/i7o+Nn7czPK15TPHjX/8tP792u73UgBISR4PjR5z1R0TOj96Scc2\nnQY8duJNtw1tlyRJsmB0lwYDH9rz0pUrV87/SDabdcAUgMLLpJifRYsWLVmy5B+Pjz/++LPO\nOiutYQCIKc0ry3Tu3Llz587bHjdt2rSioiLFYQCIyUW3AQhNCAEITQgBCE0IAQhNCAEITQgB\nCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCE0IAQhNCAEITQgBCC3NG/PuaN26dT/72c8aNmyY\n9iARVVRUVFVVNW/ePO1BCKe8vLxt27ZpTxHUMcccc8UVV6Q9RZ1QV0JIihYsWLBmzZru3bun\nPQjhTJw4cciQIWlPQXQOjQIQmhACEJoQkjRs2LCoqCjtKYjIJ9PUBT4jBCA07wgBCE0IAQhN\nCAEITQij2zj/rsGfad+kUZMOp18+ZdHmtMchiLeuOzWTyWQymQ7DZ6U9C+EJYXDZmy8btvC8\ne8veKZty7rwrh96+Iu2BiOHYHz2fy+Ve+mGntAcBIYxu6bSpr5w3YmSvtsVte40c8YXZ900r\nT3skgMISwtiy2Wz70tJt3yEsKi1tn81mU54IoMCEMLxMJrP9ke+UAhEJYWwlJSXLFi2qTJIk\nSSrLypaXlJSkPBFAgQlhbEf3u6jrY+PHzSxfWz5z3PjHT+vfzy1xgGBcYi26ynkThw4a9T+v\nrmvR7WvX3XPb109wS0gK4aGBRRdO2bj9F/UvfrDq/r6pzkNkQghAaA6NAhCaEAIQmhACEJoQ\nAhCaEAIQmhACEJoQAhCaEAIQmhACENr/B8flO47n569mAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 300,
       "width": 300
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 5)\n",
    "loeuf_plot = ggplot(coegene_data, aes(x=as.character(sign), y=LOEUF, fill = as.character(sign))) +  coeqtl_fill + \n",
    "    geom_boxplot() + plot_config + stat_compare_means(method = \"t.test\", size = 2) + ggtitle(cell_type) + theme(axis.title.x = element_blank())\n",
    "loeuf_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "id": "67a71132-b2c8-459b-875d-ce8d28ec82d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 101 rows containing non-finite outside the scale range (`stat_boxplot()`).\"\n",
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 101 rows containing non-finite outside the scale range (`stat_compare_means()`).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/I2/I2_LOEUF_Score_', cell_type,rm_rib_mt_genes, '.pdf'), width =2, height = 1.8)\n",
    "loeuf_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 572,
   "id": "66d6ea03-502f-4d00-be30-a3a0cbff723a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save data for LOEUF score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "id": "d8fb0cad-71f7-40b1-a1fe-1504feb4bd37",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(coegene_data[,c('coeGene', 'sign', 'LOEUF')], paste0('figures/I2/I2_LOEUF_Score_', cell_type, rm_rib_mt_genes,'_data.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca0333fc-71b8-436e-8fc6-e699e170180a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a414a35-2b82-4a9c-804c-1b3a5473a13b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
