{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "be8f073c-217f-49ef-871f-53795c7ee630",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the cleaned up files of co-eQTL mapping from the single- datasetd (see previous script)\n",
    "### Apply multiple testing to the single-dataset files and save the significant co-eQTLs and p-value thresholds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c105325a-75df-4ae6-bb71-17cd659b4bc1",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e66fd4a9-3553-4240-ac54-20ba323a46c1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: limma\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "This is mgcv 1.9-1. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "corrplot 0.95 loaded\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:genefilter’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "\n",
      "************\n",
      "Welcome to faux. For support and examples visit:\n",
      "https://debruine.github.io/faux/\n",
      "- Get and set global package options with: faux_options()\n",
      "************\n",
      "\n",
      "wCorr v1.9.8\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘sjmisc’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    is_empty\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘DT’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "This is coloc version 5.2.3\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: gsubfn\n",
      "\n",
      "Loading required package: proto\n",
      "\n",
      "Warning message in fun(libname, pkgname):\n",
      "“couldn't connect to display \"localhost:15.0\"”\n",
      "Loading required package: RSQLite\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    replace_na\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dbscan’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    as.dendrogram\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:rtracklayer’:\n",
      "\n",
      "    blocks, path\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    components\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: dynamicTreeCut\n",
      "\n",
      "Loading required package: fastcluster\n",
      "\n",
      "\n",
      "Attaching package: ‘fastcluster’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    hclust\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘WGCNA’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:qvalue’:\n",
      "\n",
      "    qvalue\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    geometry\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: MotifDb\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'XVector'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grid':\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "See system.file(\"LICENSE\", package=\"MotifDb\") for use restrictions.\n",
      "\n",
      "Warning message:\n",
      "\"replacing previous import 'S4Vectors::as.data.frame' by 'motifStack::as.data.frame' when loading 'motifbreakR'\"\n",
      "\n",
      "Attaching package: 'Gviz'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    feature\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "\n",
      "Attaching package: 'MOFA2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    predict\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f5b9c1d2-bba4-4aa1-a82d-d6047cb0258e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(stringi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "137617a6-d762-4bd9-a2fd-44edcbb3148c",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(biomaRt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7679f5e-4fdf-48b5-9995-faf1eaf6b91a",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9c7cf7d1-eedd-44e1-93a6-a81a0b10cf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Select dataset and cell-type for which to execute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4077c9cf-1b46-4e8f-8dce-87d8a29184b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = 'oneK1K'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "01a0c0d0-a71a-4e83-94f0-60fbf8d91f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_result = 'oneK1K'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "368ed744-244f-424b-bff4-7ed0bbe943f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = 'CD4_T'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f142ad47-ce8a-41bd-a8bf-91c545ddb9bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Which co-eQTL results (final-mapping: co_qtls_decision_tree; or initial mapping in evaluation phase: co_qtls_sceqtlgen )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8b4456df-82bd-45bd-a189-6f623b50fdaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results_path = 'co_qtls_decision_tree/'\n",
    "\n",
    "# co_qtls_sceqtlgen ,co_qtls_T1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3f582d6f-3645-429b-9af3-b9d1aba4672d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the co-EQTL Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "03bc0436-c1d2-4e11-b5bd-4920eb33a65e",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data/current/coeqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67d6fa19-cf33-463c-adc7-1bd208a7c76b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the eQTL mapping files (from sceQTLGen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5ab9cfad-baee-44e3-b706-7d86a0c8bf7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl = '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e5648a1f-08df-4f1d-b114-fa97685bf088",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_results = 'sc_eqtlgen_final'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a9ea751-49a5-4bb3-a1d4-4930b0afe061",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to save files to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "366914a4-6a0b-48c0-be94-4fd7b288e82a",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path = '../results/current/F3/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b9796cbb-23aa-41d7-a653-21901dd6eae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path_analysis = paste0( \"../data/current/coeqtl_mapping/\", coeqtl_results_path , \"analysis_\", dataset_result, \"/\", cell_type, \"/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6d0a3b8f-ffcf-43b9-9fc4-dcf12c10962c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/'"
      ],
      "text/latex": [
       "'../data/current/coeqtl\\_mapping/co\\_qtls\\_decision\\_tree/analysis\\_oneK1K/CD4\\_T/'"
      ],
      "text/markdown": [
       "'../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/'"
      ],
      "text/plain": [
       "[1] \"../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "result_path_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "55842c75-fae6-483e-90df-b6782098f9eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "###  Path to meta-data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "302179ae-d00e-4b22-927e-62ca8e169bcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_meta = '../data/current/meta-data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94244edd-9c23-487b-97b0-3809576b2c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for multiple testing correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4cff4d6d-7380-454d-a049-f80d97f4de82",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_sample_filter = TRUE   # define whether to filter tests based on minimum number of samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f1ff47aa-395a-412d-b45b-f363937a4f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_sample_structure = data.frame( dataset = c('oneK1K', 'multiome', 'Franke_v3', 'Franke_v2', 'wijst'),\n",
    "                                thres = c(102,12,4, 10,4))     # minimum number of sample for test to be considered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f3ee5161-434c-4581-b572-b2a9f5bb656e",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_filter = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0f62102-7845-472d-bf32-46ef09480af5",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed2279d3-f25e-4f34-85b0-d7f94bbd570d",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874bcc2a-9ebf-498a-8f3c-b05ef782ce74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4486ad70-ad3c-4ed7-b7ee-e0c6f9afbca0",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3541ade-683d-4741-b177-06aec2cf2569",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load the initial list of genes to test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d637aec1-9bcb-44a2-947f-f874786a6461",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load all the genes that passed the decision tree filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "212ced6f-84b7-4ad3-9430-12631a575793",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list_input = fread(paste0(data_path, coeqtl_results_path , '/correlation_input/', 'F11_Decision_Tree_Geneswg3_', dataset,'_' ,  cell_type, '.Qced.Normalized.SCs.Rds.tsv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "858aba4b-5db8-4847-a5ed-41e0e026bdef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>x</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG </td></tr>\n",
       "\t<tr><td>A2M  </td></tr>\n",
       "\t<tr><td>AAGAB</td></tr>\n",
       "\t<tr><td>AAK1 </td></tr>\n",
       "\t<tr><td>AAMDC</td></tr>\n",
       "\t<tr><td>AAMP </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 1\n",
       "\\begin{tabular}{l}\n",
       " x\\\\\n",
       " <chr>\\\\\n",
       "\\hline\n",
       "\t A1BG \\\\\n",
       "\t A2M  \\\\\n",
       "\t AAGAB\\\\\n",
       "\t AAK1 \\\\\n",
       "\t AAMDC\\\\\n",
       "\t AAMP \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 1\n",
       "\n",
       "| x &lt;chr&gt; |\n",
       "|---|\n",
       "| A1BG  |\n",
       "| A2M   |\n",
       "| AAGAB |\n",
       "| AAK1  |\n",
       "| AAMDC |\n",
       "| AAMP  |\n",
       "\n"
      ],
      "text/plain": [
       "  x    \n",
       "1 A1BG \n",
       "2 A2M  \n",
       "3 AAGAB\n",
       "4 AAK1 \n",
       "5 AAMDC\n",
       "6 AAMP "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_list_input)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ad9e8a77-3ac7-416d-b114-456bbafe160e",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(gene_list_input) = 'x'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "83a9995b-fabd-4bdc-9cc4-9a6183d7b9bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "7503"
      ],
      "text/latex": [
       "7503"
      ],
      "text/markdown": [
       "7503"
      ],
      "text/plain": [
       "[1] 7503"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(gene_list_input)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3c7829-cba2-4b11-9701-5261d2a521ea",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load the SNP eGene mapping (cis-eQTLs from sceQTLgen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "bbadb005-8561-4b13-8951-dd21efd068de",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Provided by Marc-Jan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f4cc3705-bed0-4427-9366-7be422c63325",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_gene_mapping = read.csv(paste0(data_path_eqtl, eqtl_results , '/' , cell_type, '_snp_egene_mapping.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "66cbe8ef-d1e0-4516-9f71-c5298d35e70b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "6795"
      ],
      "text/latex": [
       "6795"
      ],
      "text/markdown": [
       "6795"
      ],
      "text/plain": [
       "[1] 6795"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c3370982-60f9-49f1-857f-a6050f91383a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>p_value</th><th scope=col>beta</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>GSTM3</td><td>1:109740350:G:T</td><td>109733932</td><td>109741038</td><td>5.311981e-277</td><td>-1.1703583</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>GSTM3</td><td>1:109731286:G:T</td><td>109733932</td><td>109741038</td><td> 1.872347e-74</td><td>-0.7292481</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & X & feature\\_id & snp\\_id & feature\\_start & feature\\_end & p\\_value & beta\\\\\n",
       "  & <int> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 0 & GSTM3 & 1:109740350:G:T & 109733932 & 109741038 & 5.311981e-277 & -1.1703583\\\\\n",
       "\t2 & 1 & GSTM3 & 1:109731286:G:T & 109733932 & 109741038 &  1.872347e-74 & -0.7292481\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | p_value &lt;dbl&gt; | beta &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | 0 | GSTM3 | 1:109740350:G:T | 109733932 | 109741038 | 5.311981e-277 | -1.1703583 |\n",
       "| 2 | 1 | GSTM3 | 1:109731286:G:T | 109733932 | 109741038 |  1.872347e-74 | -0.7292481 |\n",
       "\n"
      ],
      "text/plain": [
       "  X feature_id snp_id          feature_start feature_end p_value      \n",
       "1 0 GSTM3      1:109740350:G:T 109733932     109741038   5.311981e-277\n",
       "2 1 GSTM3      1:109731286:G:T 109733932     109741038    1.872347e-74\n",
       "  beta      \n",
       "1 -1.1703583\n",
       "2 -0.7292481"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1769fee2-db9d-4005-92cc-f394f4b229ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_gene_mapping = unique(snp_gene_mapping[,c('feature_id', 'snp_id')])\n",
    "colnames(snp_gene_mapping) = c('gene', 'snp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "669a1570-20aa-4c75-b568-64245172c295",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on the egenes also left after decision tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "3f7caa70-2de9-435b-8c76-69c50d3774d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_gene_mapping_filtered = snp_gene_mapping[snp_gene_mapping$gene %in% gene_list_input$x,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ce5bdf13-5c8c-4b9c-8d7b-0a12fb0ef851",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4350"
      ],
      "text/latex": [
       "4350"
      ],
      "text/markdown": [
       "4350"
      ],
      "text/plain": [
       "[1] 4350"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_gene_mapping_filtered )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5b0341dd-6fbd-471b-b8f2-4fb5d14f6f43",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate amount of expected gene pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "2ef1cf1b-d44d-46cd-949d-eafd7e4cd22e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "18182337"
      ],
      "text/latex": [
       "18182337"
      ],
      "text/markdown": [
       "18182337"
      ],
      "text/plain": [
       "[1] 18182337"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "(((length(unique(snp_gene_mapping_filtered$gene)) * length(unique(snp_gene_mapping_filtered$gene))) - length(unique(snp_gene_mapping_filtered$gene)))/2) + \n",
    "((nrow(gene_list_input) -length(unique(snp_gene_mapping_filtered$gene))) * length(unique(snp_gene_mapping_filtered$gene)))\n",
    "\n",
    "\n",
    "#(((#egenes * # egenes) - #egenes)/2) + ((#genes - #egenes)* #egenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d3896c53-4ed2-405b-b99f-b61472f86beb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check the actual amount of gene-pairs"
   ]
  },
  {
   "cell_type": "raw",
   "id": "753891fb-be22-4d66-8a30-a5051dd47736",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "19767f2d-ae78-48b6-87d4-74ec9861898f",
   "metadata": {},
   "source": [
    "## Load info about ribosomal/ mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "16aae9de-67a7-42e1-b100-2d44ab24f0c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the annotation of ribosomal and mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "752cd157-3000-40ac-b29e-71ee7bd52301",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ribosomal_genes = read.csv(paste0(data_path_meta , '/ribosomal_genes.txt.gz') ,sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "465dc44d-1cd9-49db-a8bd-07d123c52f1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gene.stable.ID</th><th scope=col>Transcript.stable.ID</th><th scope=col>Gene.name</th><th scope=col>Gene.start..bp.</th><th scope=col>Gene.end..bp.</th><th scope=col>Chromosome.scaffold.name</th><th scope=col>Gene.Synonym</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000164163</td><td>ENST00000296577</td><td>ABCE1</td><td>145098288</td><td>145129524</td><td>4</td><td>OABP</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000164163</td><td>ENST00000296577</td><td>ABCE1</td><td>145098288</td><td>145129524</td><td>4</td><td>RLI </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Gene.stable.ID & Transcript.stable.ID & Gene.name & Gene.start..bp. & Gene.end..bp. & Chromosome.scaffold.name & Gene.Synonym\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000164163 & ENST00000296577 & ABCE1 & 145098288 & 145129524 & 4 & OABP\\\\\n",
       "\t2 & ENSG00000164163 & ENST00000296577 & ABCE1 & 145098288 & 145129524 & 4 & RLI \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | Gene.stable.ID &lt;chr&gt; | Transcript.stable.ID &lt;chr&gt; | Gene.name &lt;chr&gt; | Gene.start..bp. &lt;int&gt; | Gene.end..bp. &lt;int&gt; | Chromosome.scaffold.name &lt;chr&gt; | Gene.Synonym &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000164163 | ENST00000296577 | ABCE1 | 145098288 | 145129524 | 4 | OABP |\n",
       "| 2 | ENSG00000164163 | ENST00000296577 | ABCE1 | 145098288 | 145129524 | 4 | RLI  |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene.stable.ID  Transcript.stable.ID Gene.name Gene.start..bp. Gene.end..bp.\n",
       "1 ENSG00000164163 ENST00000296577      ABCE1     145098288       145129524    \n",
       "2 ENSG00000164163 ENST00000296577      ABCE1     145098288       145129524    \n",
       "  Chromosome.scaffold.name Gene.Synonym\n",
       "1 4                        OABP        \n",
       "2 4                        RLI         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(ribosomal_genes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ef3b9ad5-72b7-4cd1-867e-f50f2077067b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ribosomal_genes = unique(ribosomal_genes$Gene.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6e1e84ca-22b1-4e05-b4f2-85412b366880",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABCE1'</li><li>'ABCF1'</li><li>'APEX1'</li><li>'APOD'</li><li>'ASCC2'</li><li>'ASCC3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ABCE1'\n",
       "\\item 'ABCF1'\n",
       "\\item 'APEX1'\n",
       "\\item 'APOD'\n",
       "\\item 'ASCC2'\n",
       "\\item 'ASCC3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ABCE1'\n",
       "2. 'ABCF1'\n",
       "3. 'APEX1'\n",
       "4. 'APOD'\n",
       "5. 'ASCC2'\n",
       "6. 'ASCC3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ABCE1\" \"ABCF1\" \"APEX1\" \"APOD\"  \"ASCC2\" \"ASCC3\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(ribosomal_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "c2f4ddf8-90d5-47c6-a327-3ba540cbfe68",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Mitochondrial Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "da615909-acd6-4a77-9ed5-316b7b6e0ed3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrial_genes = read.csv(paste0(data_path_meta , '/mitochondrial_genes.txt') ,sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "03382e48-8200-4ef5-a505-e809d3a45811",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gene.stable.ID</th><th scope=col>Gene.stable.ID.version</th><th scope=col>Transcript.stable.ID</th><th scope=col>Transcript.stable.ID.version</th><th scope=col>Chromosome.scaffold.name</th><th scope=col>Gene.start..bp.</th><th scope=col>Gene.end..bp.</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000198888</td><td>ENSG00000198888.2</td><td>ENST00000361390</td><td>ENST00000361390.2</td><td>MT</td><td>3307</td><td>4262</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000198763</td><td>ENSG00000198763.3</td><td>ENST00000361453</td><td>ENST00000361453.3</td><td>MT</td><td>4470</td><td>5511</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Gene.stable.ID & Gene.stable.ID.version & Transcript.stable.ID & Transcript.stable.ID.version & Chromosome.scaffold.name & Gene.start..bp. & Gene.end..bp.\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000198888 & ENSG00000198888.2 & ENST00000361390 & ENST00000361390.2 & MT & 3307 & 4262\\\\\n",
       "\t2 & ENSG00000198763 & ENSG00000198763.3 & ENST00000361453 & ENST00000361453.3 & MT & 4470 & 5511\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | Gene.stable.ID &lt;chr&gt; | Gene.stable.ID.version &lt;chr&gt; | Transcript.stable.ID &lt;chr&gt; | Transcript.stable.ID.version &lt;chr&gt; | Chromosome.scaffold.name &lt;chr&gt; | Gene.start..bp. &lt;int&gt; | Gene.end..bp. &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000198888 | ENSG00000198888.2 | ENST00000361390 | ENST00000361390.2 | MT | 3307 | 4262 |\n",
       "| 2 | ENSG00000198763 | ENSG00000198763.3 | ENST00000361453 | ENST00000361453.3 | MT | 4470 | 5511 |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene.stable.ID  Gene.stable.ID.version Transcript.stable.ID\n",
       "1 ENSG00000198888 ENSG00000198888.2      ENST00000361390     \n",
       "2 ENSG00000198763 ENSG00000198763.3      ENST00000361453     \n",
       "  Transcript.stable.ID.version Chromosome.scaffold.name Gene.start..bp.\n",
       "1 ENST00000361390.2            MT                       3307           \n",
       "2 ENST00000361453.3            MT                       4470           \n",
       "  Gene.end..bp.\n",
       "1 4262         \n",
       "2 5511         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "1576195b-6a05-427f-b915-9010f0699f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "### convert ENSG to SYMBOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "debcfeed-f13b-4a5b-ac49-c139e716bde4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "Warning message in bitr(unique(mitochondrial_genes$Gene.stable.ID), fromType = \"ENSEMBL\", :\n",
      "\"2.14% of input gene IDs are fail to map...\"\n"
     ]
    }
   ],
   "source": [
    "mitochondrial_genes_convert <- bitr(unique(mitochondrial_genes$Gene.stable.ID), \n",
    "                  fromType = \"ENSEMBL\", \n",
    "                  toType = \"SYMBOL\", \n",
    "                  OrgDb = org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "1f3ad49b-a194-44a6-8a4d-9d318e5ce9cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ENSEMBL</th><th scope=col>SYMBOL</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000198888</td><td>ND1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000198763</td><td>ND2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ENSEMBL & SYMBOL\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000198888 & ND1\\\\\n",
       "\t2 & ENSG00000198763 & ND2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | ENSEMBL &lt;chr&gt; | SYMBOL &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ENSG00000198888 | ND1 |\n",
       "| 2 | ENSG00000198763 | ND2 |\n",
       "\n"
      ],
      "text/plain": [
       "  ENSEMBL         SYMBOL\n",
       "1 ENSG00000198888 ND1   \n",
       "2 ENSG00000198763 ND2   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes_convert,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "8928df82-58ce-4340-a81d-2bed271fe461",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrial_genes  = unique(mitochondrial_genes_convert$SYMBOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "7be462f6-abc0-4d6c-a8f7-997593468e43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ND1'</li><li>'ND2'</li><li>'COX1'</li><li>'COX2'</li><li>'ATP8'</li><li>'ATP6'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ND1'\n",
       "\\item 'ND2'\n",
       "\\item 'COX1'\n",
       "\\item 'COX2'\n",
       "\\item 'ATP8'\n",
       "\\item 'ATP6'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ND1'\n",
       "2. 'ND2'\n",
       "3. 'COX1'\n",
       "4. 'COX2'\n",
       "5. 'ATP8'\n",
       "6. 'ATP6'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ND1\"  \"ND2\"  \"COX1\" \"COX2\" \"ATP8\" \"ATP6\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5dd95a29-8206-43f2-95fc-f88304543b88",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "93e560e1-1686-4f81-adf5-09c26900efd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "rb_mt_genes = unique(c(ribosomal_genes, mitochondrial_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e242b067-d1e8-4016-9201-266d056b6e2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABCE1'</li><li>'ABCF1'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ABCE1'\n",
       "\\item 'ABCF1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ABCE1'\n",
       "2. 'ABCF1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ABCE1\" \"ABCF1\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(rb_mt_genes,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f0a2d4a-8e52-4f3a-bff6-aa3d4919ddd5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load the co-eQTL results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a759f55d-37f1-48b4-a994-b65541fbfe50",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the co-eQTL mapping file of the selected cell-type and dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "ae393ff0-aa30-4bfb-b7a6-ddd39b115d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results = fread(paste0(data_path, coeqtl_results_path , '/', dataset, '/', cell_type, '/qtl_results_all_unqid.csv.gz'), header = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "dd6ce7f1-1511-4760-8a66-a7f14456629e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "150181698"
      ],
      "text/latex": [
       "150181698"
      ],
      "text/markdown": [
       "150181698"
      ],
      "text/plain": [
       "[1] 150181698"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "117aa107-a192-4db9-a9ed-26376823b7b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 24</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>⋯</th><th scope=col>coeGene</th><th scope=col>coQTL</th><th scope=col>snp_position</th><th scope=col>closeGenes</th><th scope=col>z_score</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>n_e_samples</th><th scope=col>gene1</th><th scope=col>gene2</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:8942630:T:C;A2M_A1BG</td><td>8942630</td><td>FALSE</td><td> 2.4227677</td><td>A1BG_A2M;12:8942630:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "\t<tr><td>1</td><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:8971336:T:C;A2M_A1BG</td><td>8971336</td><td>FALSE</td><td>-3.1526976</td><td>A1BG_A2M;12:8971336:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "\t<tr><td>2</td><td>A1BG_A2M</td><td>12:8992460:T:C</td><td>0.002398787</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:8992460:T:C;A2M_A1BG</td><td>8992460</td><td>FALSE</td><td>-3.0448233</td><td>A1BG_A2M;12:8992460:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "\t<tr><td>3</td><td>A1BG_A2M</td><td>12:9027756:A:G</td><td>0.515272692</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:9027756:A:G;A2M_A1BG</td><td>9027756</td><td>FALSE</td><td>-0.6509177</td><td>A1BG_A2M;12:9027756:A:G</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "\t<tr><td>4</td><td>A1BG_A2M</td><td>12:9157980:T:G</td><td>0.492139444</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:9157980:T:G;A2M_A1BG</td><td>9157980</td><td>FALSE</td><td>-0.6872027</td><td>A1BG_A2M;12:9157980:T:G</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "\t<tr><td>5</td><td>A1BG_A2M</td><td>12:9680928:C:G</td><td>0.822040268</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:9680928:C:G;A2M_A1BG</td><td>9680928</td><td>FALSE</td><td>-0.2249900</td><td>A1BG_A2M;12:9680928:C:G</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 24\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " V1 & feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ⋯ & coeGene & coQTL & snp\\_position & closeGenes & z\\_score & triplet & n\\_samples & n\\_e\\_samples & gene1 & gene2\\\\\n",
       " <int> & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & ⋯ & <chr> & <chr> & <int> & <lgl> & <dbl> & <chr> & <int> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0 & A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:8942630:T:C;A2M\\_A1BG & 8942630 & FALSE &  2.4227677 & A1BG\\_A2M;12:8942630:T:C & 848 & 848 & A2M & A1BG\\\\\n",
       "\t 1 & A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:8971336:T:C;A2M\\_A1BG & 8971336 & FALSE & -3.1526976 & A1BG\\_A2M;12:8971336:T:C & 848 & 848 & A2M & A1BG\\\\\n",
       "\t 2 & A1BG\\_A2M & 12:8992460:T:C & 0.002398787 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:8992460:T:C;A2M\\_A1BG & 8992460 & FALSE & -3.0448233 & A1BG\\_A2M;12:8992460:T:C & 848 & 848 & A2M & A1BG\\\\\n",
       "\t 3 & A1BG\\_A2M & 12:9027756:A:G & 0.515272692 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:9027756:A:G;A2M\\_A1BG & 9027756 & FALSE & -0.6509177 & A1BG\\_A2M;12:9027756:A:G & 848 & 848 & A2M & A1BG\\\\\n",
       "\t 4 & A1BG\\_A2M & 12:9157980:T:G & 0.492139444 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:9157980:T:G;A2M\\_A1BG & 9157980 & FALSE & -0.6872027 & A1BG\\_A2M;12:9157980:T:G & 848 & 848 & A2M & A1BG\\\\\n",
       "\t 5 & A1BG\\_A2M & 12:9680928:C:G & 0.822040268 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:9680928:C:G;A2M\\_A1BG & 9680928 & FALSE & -0.2249900 & A1BG\\_A2M;12:9680928:C:G & 848 & 848 & A2M & A1BG\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 24\n",
       "\n",
       "| V1 &lt;int&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ⋯ ⋯ | coeGene &lt;chr&gt; | coQTL &lt;chr&gt; | snp_position &lt;int&gt; | closeGenes &lt;lgl&gt; | z_score &lt;dbl&gt; | triplet &lt;chr&gt; | n_samples &lt;int&gt; | n_e_samples &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0 | A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:8942630:T:C;A2M_A1BG | 8942630 | FALSE |  2.4227677 | A1BG_A2M;12:8942630:T:C | 848 | 848 | A2M | A1BG |\n",
       "| 1 | A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:8971336:T:C;A2M_A1BG | 8971336 | FALSE | -3.1526976 | A1BG_A2M;12:8971336:T:C | 848 | 848 | A2M | A1BG |\n",
       "| 2 | A1BG_A2M | 12:8992460:T:C | 0.002398787 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:8992460:T:C;A2M_A1BG | 8992460 | FALSE | -3.0448233 | A1BG_A2M;12:8992460:T:C | 848 | 848 | A2M | A1BG |\n",
       "| 3 | A1BG_A2M | 12:9027756:A:G | 0.515272692 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:9027756:A:G;A2M_A1BG | 9027756 | FALSE | -0.6509177 | A1BG_A2M;12:9027756:A:G | 848 | 848 | A2M | A1BG |\n",
       "| 4 | A1BG_A2M | 12:9157980:T:G | 0.492139444 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:9157980:T:G;A2M_A1BG | 9157980 | FALSE | -0.6872027 | A1BG_A2M;12:9157980:T:G | 848 | 848 | A2M | A1BG |\n",
       "| 5 | A1BG_A2M | 12:9680928:C:G | 0.822040268 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:9680928:C:G;A2M_A1BG | 9680928 | FALSE | -0.2249900 | A1BG_A2M;12:9680928:C:G | 848 | 848 | A2M | A1BG |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 feature_id snp_id         p_value     beta beta_se\n",
       "1 0  A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA     \n",
       "2 1  A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA     \n",
       "3 2  A1BG_A2M   12:8992460:T:C 0.002398787 NA   NA     \n",
       "4 3  A1BG_A2M   12:9027756:A:G 0.515272692 NA   NA     \n",
       "5 4  A1BG_A2M   12:9157980:T:G 0.492139444 NA   NA     \n",
       "6 5  A1BG_A2M   12:9680928:C:G 0.822040268 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start feature_end ⋯\n",
       "1 NA                        12                 9067664       9116229     ⋯\n",
       "2 NA                        12                 9067664       9116229     ⋯\n",
       "3 NA                        12                 9067664       9116229     ⋯\n",
       "4 NA                        12                 9067664       9116229     ⋯\n",
       "5 NA                        12                 9067664       9116229     ⋯\n",
       "6 NA                        12                 9067664       9116229     ⋯\n",
       "  coeGene coQTL                   snp_position closeGenes z_score   \n",
       "1 A1BG    12:8942630:T:C;A2M_A1BG 8942630      FALSE       2.4227677\n",
       "2 A1BG    12:8971336:T:C;A2M_A1BG 8971336      FALSE      -3.1526976\n",
       "3 A1BG    12:8992460:T:C;A2M_A1BG 8992460      FALSE      -3.0448233\n",
       "4 A1BG    12:9027756:A:G;A2M_A1BG 9027756      FALSE      -0.6509177\n",
       "5 A1BG    12:9157980:T:G;A2M_A1BG 9157980      FALSE      -0.6872027\n",
       "6 A1BG    12:9680928:C:G;A2M_A1BG 9680928      FALSE      -0.2249900\n",
       "  triplet                 n_samples n_e_samples gene1 gene2\n",
       "1 A1BG_A2M;12:8942630:T:C 848       848         A2M   A1BG \n",
       "2 A1BG_A2M;12:8971336:T:C 848       848         A2M   A1BG \n",
       "3 A1BG_A2M;12:8992460:T:C 848       848         A2M   A1BG \n",
       "4 A1BG_A2M;12:9027756:A:G 848       848         A2M   A1BG \n",
       "5 A1BG_A2M;12:9157980:T:G 848       848         A2M   A1BG \n",
       "6 A1BG_A2M;12:9680928:C:G 848       848         A2M   A1BG "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "e8b91581-fd5e-4ddb-8760-2389c508ea1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Only for Mono/ B: filter out eGenes that don't fall in 5% FDR thres\n",
    "coeqtl_results = coeqtl_results[coeqtl_results$eGene %in% unique(snp_gene_mapping$gene),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "40c7a901-abaa-4943-8a21-3676c430be81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "150181698"
      ],
      "text/latex": [
       "150181698"
      ],
      "text/markdown": [
       "150181698"
      ],
      "text/plain": [
       "[1] 150181698"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "b9ec232d-792e-407a-b607-a20306de939a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 24</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>⋯</th><th scope=col>coeGene</th><th scope=col>coQTL</th><th scope=col>snp_position</th><th scope=col>closeGenes</th><th scope=col>z_score</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>n_e_samples</th><th scope=col>gene1</th><th scope=col>gene2</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:8942630:T:C;A2M_A1BG</td><td>8942630</td><td>FALSE</td><td> 2.422768</td><td>A1BG_A2M;12:8942630:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "\t<tr><td>1</td><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>⋯</td><td>A1BG</td><td>12:8971336:T:C;A2M_A1BG</td><td>8971336</td><td>FALSE</td><td>-3.152698</td><td>A1BG_A2M;12:8971336:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 24\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " V1 & feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ⋯ & coeGene & coQTL & snp\\_position & closeGenes & z\\_score & triplet & n\\_samples & n\\_e\\_samples & gene1 & gene2\\\\\n",
       " <int> & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & ⋯ & <chr> & <chr> & <int> & <lgl> & <dbl> & <chr> & <int> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 0 & A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:8942630:T:C;A2M\\_A1BG & 8942630 & FALSE &  2.422768 & A1BG\\_A2M;12:8942630:T:C & 848 & 848 & A2M & A1BG\\\\\n",
       "\t 1 & A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & 9116229 & ⋯ & A1BG & 12:8971336:T:C;A2M\\_A1BG & 8971336 & FALSE & -3.152698 & A1BG\\_A2M;12:8971336:T:C & 848 & 848 & A2M & A1BG\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 24\n",
       "\n",
       "| V1 &lt;int&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ⋯ ⋯ | coeGene &lt;chr&gt; | coQTL &lt;chr&gt; | snp_position &lt;int&gt; | closeGenes &lt;lgl&gt; | z_score &lt;dbl&gt; | triplet &lt;chr&gt; | n_samples &lt;int&gt; | n_e_samples &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 0 | A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:8942630:T:C;A2M_A1BG | 8942630 | FALSE |  2.422768 | A1BG_A2M;12:8942630:T:C | 848 | 848 | A2M | A1BG |\n",
       "| 1 | A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | 9116229 | ⋯ | A1BG | 12:8971336:T:C;A2M_A1BG | 8971336 | FALSE | -3.152698 | A1BG_A2M;12:8971336:T:C | 848 | 848 | A2M | A1BG |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 feature_id snp_id         p_value     beta beta_se\n",
       "1 0  A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA     \n",
       "2 1  A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start feature_end ⋯\n",
       "1 NA                        12                 9067664       9116229     ⋯\n",
       "2 NA                        12                 9067664       9116229     ⋯\n",
       "  coeGene coQTL                   snp_position closeGenes z_score  \n",
       "1 A1BG    12:8942630:T:C;A2M_A1BG 8942630      FALSE       2.422768\n",
       "2 A1BG    12:8971336:T:C;A2M_A1BG 8971336      FALSE      -3.152698\n",
       "  triplet                 n_samples n_e_samples gene1 gene2\n",
       "1 A1BG_A2M;12:8942630:T:C 848       848         A2M   A1BG \n",
       "2 A1BG_A2M;12:8971336:T:C 848       848         A2M   A1BG "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtl_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "d9443625-edbb-4045-99da-cd4a5ea384d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1018"
      ],
      "text/latex": [
       "1018"
      ],
      "text/markdown": [
       "1018"
      ],
      "text/plain": [
       "[1] 1018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "max(coeqtl_results$n_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "3dc8943a-4dde-45c5-b3d9-d0deb7458da9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "   16.0   881.0   956.0   922.1   990.0  1018.0 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(coeqtl_results$n_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "149e8aa0-5ec7-4864-bc47-7453919937b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove dummy columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "2ba820aa-c48e-4f65-8f96-b4726c78278b",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results$V1 = NULL\n",
    "#coeqtl_results$test1 = NULL\n",
    "#coeqtl_results$test2 = NULL\n",
    "#coeqtl_results$test3 = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "042ac612-3948-475b-9f37-59f142b7a639",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add cell-type and dataset to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "52808e5c-53ac-4a5d-aaaf-3c03dae5d2f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results$cell_type = cell_type  \n",
    "coeqtl_results$dataset = dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ec33baa-c7e2-4ea3-9d70-56543e01e524",
   "metadata": {},
   "source": [
    "### Define ribosomal and mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "b2d25c12-e5df-420c-97ec-ae16cd1202f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results$is_rb_mt = 0\n",
    "coeqtl_results$is_rb_mt[(coeqtl_results$eGene %in% rb_mt_genes) |\n",
    "                            (coeqtl_results$coeGene %in% rb_mt_genes)] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "b62f8a71-4bbb-4465-9283-436697c0eda9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the unfiltered amount of tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "4acf04a8-3780-4e09-992c-fc5c8263fd81",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type', 'n_samples'. You\n",
      "can override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "amount_tests_save = coeqtl_results %>% group_by( dataset, cell_type, n_samples, is_rb_mt) %>% summarise(amount_tests_datasets = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "cca60d86-c783-4449-b2a6-89229a1ed67e",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(amount_tests_save, paste0(result_path_analysis , 'F3_1_1_Meta_Info_Tests.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b290015-e2d7-42ec-8306-a3e1ad7acf18",
   "metadata": {},
   "source": [
    "### Filter on only cis-eQTL snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "13b0811d-622c-4ad7-afc6-b2cfee54a38c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "150181698"
      ],
      "text/latex": [
       "150181698"
      ],
      "text/markdown": [
       "150181698"
      ],
      "text/plain": [
       "[1] 150181698"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "f56d9128-9a10-495f-8e89-9d977a9ea6c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 26</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>ENSG</th><th scope=col>⋯</th><th scope=col>closeGenes</th><th scope=col>z_score</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>n_e_samples</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>is_rb_mt</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>ENSG00000175899</td><td>⋯</td><td>FALSE</td><td> 2.422768</td><td>A1BG_A2M;12:8942630:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td><td>CD4_T</td><td>oneK1K</td><td>0</td></tr>\n",
       "\t<tr><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>ENSG00000175899</td><td>⋯</td><td>FALSE</td><td>-3.152698</td><td>A1BG_A2M;12:8971336:T:C</td><td>848</td><td>848</td><td>A2M</td><td>A1BG</td><td>CD4_T</td><td>oneK1K</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 26\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ENSG & ⋯ & closeGenes & z\\_score & triplet & n\\_samples & n\\_e\\_samples & gene1 & gene2 & cell\\_type & dataset & is\\_rb\\_mt\\\\\n",
       " <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & ⋯ & <lgl> & <dbl> & <chr> & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & 9116229 & ENSG00000175899 & ⋯ & FALSE &  2.422768 & A1BG\\_A2M;12:8942630:T:C & 848 & 848 & A2M & A1BG & CD4\\_T & oneK1K & 0\\\\\n",
       "\t A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & 9116229 & ENSG00000175899 & ⋯ & FALSE & -3.152698 & A1BG\\_A2M;12:8971336:T:C & 848 & 848 & A2M & A1BG & CD4\\_T & oneK1K & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 26\n",
       "\n",
       "| feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ENSG &lt;chr&gt; | ⋯ ⋯ | closeGenes &lt;lgl&gt; | z_score &lt;dbl&gt; | triplet &lt;chr&gt; | n_samples &lt;int&gt; | n_e_samples &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | is_rb_mt &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | 9116229 | ENSG00000175899 | ⋯ | FALSE |  2.422768 | A1BG_A2M;12:8942630:T:C | 848 | 848 | A2M | A1BG | CD4_T | oneK1K | 0 |\n",
       "| A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | 9116229 | ENSG00000175899 | ⋯ | FALSE | -3.152698 | A1BG_A2M;12:8971336:T:C | 848 | 848 | A2M | A1BG | CD4_T | oneK1K | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id snp_id         p_value     beta beta_se empirical_feature_p_value\n",
       "1 A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA      NA                       \n",
       "2 A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA      NA                       \n",
       "  feature_chromosome feature_start feature_end ENSG            ⋯ closeGenes\n",
       "1 12                 9067664       9116229     ENSG00000175899 ⋯ FALSE     \n",
       "2 12                 9067664       9116229     ENSG00000175899 ⋯ FALSE     \n",
       "  z_score   triplet                 n_samples n_e_samples gene1 gene2 cell_type\n",
       "1  2.422768 A1BG_A2M;12:8942630:T:C 848       848         A2M   A1BG  CD4_T    \n",
       "2 -3.152698 A1BG_A2M;12:8971336:T:C 848       848         A2M   A1BG  CD4_T    \n",
       "  dataset is_rb_mt\n",
       "1 oneK1K  0       \n",
       "2 oneK1K  0       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtl_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "4947a943-fbbc-46eb-952f-f97767d45c10",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(snp_filter == 'cis'){\n",
    "    coeqtl_results = coeqtl_results[coeqtl_results$snp_id %in% unique(snp_gene_mapping$snp),]\n",
    "    }\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "91669ce6-5ffc-47bc-9f1a-522e80fc7d1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "150181698"
      ],
      "text/latex": [
       "150181698"
      ],
      "text/markdown": [
       "150181698"
      ],
      "text/plain": [
       "[1] 150181698"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "99770e64-a978-41c6-bc97-8c9c9b708d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save amount of tests after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "e98afa62-17c0-42e4-bdeb-f7d6737e3b73",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset', 'cell_type', 'n_samples'. You\n",
      "can override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "amount_tests_save = coeqtl_results %>% group_by( dataset, cell_type, n_samples, is_rb_mt) %>% summarise(amount_tests_datasets = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "1f1f9a56-a86e-4320-a7df-69bef1302fe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(amount_tests_save, paste0(result_path_analysis , 'F3_1_1_Meta_Info_Tests', '_', snp_filter, '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d2b9701-9c5e-47e4-864d-7cbbb9bebbd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0eb93a65-f0a6-4431-b845-9be9b32467ab",
   "metadata": {},
   "source": [
    "### Filter based on minimum amount of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "52299d37-869b-49fe-941b-f29e9997af6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "150181698"
      ],
      "text/latex": [
       "150181698"
      ],
      "text/markdown": [
       "150181698"
      ],
      "text/plain": [
       "[1] 150181698"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "b9a11e0e-c8ac-4161-a7ea-8a5dfac79998",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(n_sample_filter == TRUE){\n",
    "    threshold = n_sample_structure$thres[n_sample_structure$dataset == dataset]\n",
    "    coeqtl_results = coeqtl_results[coeqtl_results$n_samples > threshold,]\n",
    "    }\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "f6d560b2-1a1f-4237-b438-4c6bd4031340",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "\"number of rows of result is not a multiple of vector length (arg 2)\"\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "\"number of rows of result is not a multiple of vector length (arg 2)\"\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "\"number of rows of result is not a multiple of vector length (arg 2)\"\n",
      "Warning message in cbind(parts$left, chars$ellip_h, parts$right, deparse.level = 0L):\n",
      "\"number of rows of result is not a multiple of vector length (arg 2)\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 26</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>ENSG</th><th scope=col>⋯</th><th scope=col>closeGenes</th><th scope=col>z_score</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>n_e_samples</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>is_rb_mt</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 0 × 26\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ENSG & ⋯ & closeGenes & z\\_score & triplet & n\\_samples & n\\_e\\_samples & gene1 & gene2 & cell\\_type & dataset & is\\_rb\\_mt\\\\\n",
       " <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & ⋯ & <lgl> & <dbl> & <chr> & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 0 × 26\n",
       "\n",
       "| feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ENSG &lt;chr&gt; | ⋯ ⋯ | closeGenes &lt;lgl&gt; | z_score &lt;dbl&gt; | triplet &lt;chr&gt; | n_samples &lt;int&gt; | n_e_samples &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | is_rb_mt &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     feature_id snp_id p_value beta beta_se empirical_feature_p_value\n",
       "     feature_chromosome feature_start feature_end ENSG ⋯ closeGenes z_score\n",
       "     triplet n_samples n_e_samples gene1 gene2 cell_type dataset is_rb_mt"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coeqtl_results[coeqtl_results$feature_id ==  'IGHG2_VCPKMT',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "2c7ce481-53da-4d50-b22a-78958878f4ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "103"
      ],
      "text/latex": [
       "103"
      ],
      "text/markdown": [
       "103"
      ],
      "text/plain": [
       "[1] 103"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "min(coeqtl_results$n_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "29909473-879c-4ca1-8121-0731005ed1ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "149928159"
      ],
      "text/latex": [
       "149928159"
      ],
      "text/markdown": [
       "149928159"
      ],
      "text/plain": [
       "[1] 149928159"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "b4319a48-9d37-43ce-9cf9-4453abf3673c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### check nas for p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "15cd59b6-f6ce-495d-9c0c-502c746ca4f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results[is.na(coeqtl_results$p_value),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "2b98913c-c526-4288-80dc-db36e128c1cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### columns within results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "eb8f3544-1d5b-4a4b-9c69-47bafa0b7c0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'alpha_param'</li><li>'beta'</li><li>'beta_param'</li><li>'beta_se'</li><li>'cell_type'</li><li>'closeGenes'</li><li>'coeGene'</li><li>'coQTL'</li><li>'dataset'</li><li>'eGene'</li><li>'empirical_feature_p_value'</li><li>'ENSG'</li><li>'feature_chromosome'</li><li>'feature_end'</li><li>'feature_id'</li><li>'feature_start'</li><li>'gene1'</li><li>'gene2'</li><li>'is_rb_mt'</li><li>'n_e_samples'</li><li>'n_samples'</li><li>'p_value'</li><li>'snp_id'</li><li>'snp_position'</li><li>'triplet'</li><li>'z_score'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'alpha\\_param'\n",
       "\\item 'beta'\n",
       "\\item 'beta\\_param'\n",
       "\\item 'beta\\_se'\n",
       "\\item 'cell\\_type'\n",
       "\\item 'closeGenes'\n",
       "\\item 'coeGene'\n",
       "\\item 'coQTL'\n",
       "\\item 'dataset'\n",
       "\\item 'eGene'\n",
       "\\item 'empirical\\_feature\\_p\\_value'\n",
       "\\item 'ENSG'\n",
       "\\item 'feature\\_chromosome'\n",
       "\\item 'feature\\_end'\n",
       "\\item 'feature\\_id'\n",
       "\\item 'feature\\_start'\n",
       "\\item 'gene1'\n",
       "\\item 'gene2'\n",
       "\\item 'is\\_rb\\_mt'\n",
       "\\item 'n\\_e\\_samples'\n",
       "\\item 'n\\_samples'\n",
       "\\item 'p\\_value'\n",
       "\\item 'snp\\_id'\n",
       "\\item 'snp\\_position'\n",
       "\\item 'triplet'\n",
       "\\item 'z\\_score'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'alpha_param'\n",
       "2. 'beta'\n",
       "3. 'beta_param'\n",
       "4. 'beta_se'\n",
       "5. 'cell_type'\n",
       "6. 'closeGenes'\n",
       "7. 'coeGene'\n",
       "8. 'coQTL'\n",
       "9. 'dataset'\n",
       "10. 'eGene'\n",
       "11. 'empirical_feature_p_value'\n",
       "12. 'ENSG'\n",
       "13. 'feature_chromosome'\n",
       "14. 'feature_end'\n",
       "15. 'feature_id'\n",
       "16. 'feature_start'\n",
       "17. 'gene1'\n",
       "18. 'gene2'\n",
       "19. 'is_rb_mt'\n",
       "20. 'n_e_samples'\n",
       "21. 'n_samples'\n",
       "22. 'p_value'\n",
       "23. 'snp_id'\n",
       "24. 'snp_position'\n",
       "25. 'triplet'\n",
       "26. 'z_score'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"alpha_param\"               \"beta\"                     \n",
       " [3] \"beta_param\"                \"beta_se\"                  \n",
       " [5] \"cell_type\"                 \"closeGenes\"               \n",
       " [7] \"coeGene\"                   \"coQTL\"                    \n",
       " [9] \"dataset\"                   \"eGene\"                    \n",
       "[11] \"empirical_feature_p_value\" \"ENSG\"                     \n",
       "[13] \"feature_chromosome\"        \"feature_end\"              \n",
       "[15] \"feature_id\"                \"feature_start\"            \n",
       "[17] \"gene1\"                     \"gene2\"                    \n",
       "[19] \"is_rb_mt\"                  \"n_e_samples\"              \n",
       "[21] \"n_samples\"                 \"p_value\"                  \n",
       "[23] \"snp_id\"                    \"snp_position\"             \n",
       "[25] \"triplet\"                   \"z_score\"                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(colnames(coeqtl_results))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b664ac00-ea28-4cae-8397-a486ba55e2ea",
   "metadata": {},
   "source": [
    "### Multiple Testing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "d5543410-cd14-4ada-b254-1e9b469beff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the amount of tests per eGene\n",
    "coeqtl_results = coeqtl_results %>% group_by(eGene) %>% mutate(amount_tests_eGene = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "14fd89c1-0993-4d88-9d09-cfe95aa480e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Correct for multiple testing per eGene\n",
    "coeqtl_results$p_val_mt_eGene = coeqtl_results$p_value * coeqtl_results$amount_tests_eGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "44d0681a-3ccf-4e62-8c86-e20dfb5f51e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the top effect per eGene\n",
    "top_eGene_effects = coeqtl_results %>% group_by(eGene) %>% dplyr::filter(p_value == min(p_value)) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "d8d8f769-90c4-4b6f-8dec-4786b92ee587",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_eGene_effects = top_eGene_effects %>% group_by(eGene) %>% slice_min(p_value, with_ties = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "07247d40-6ba9-486e-811a-015165bc79f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3037"
      ],
      "text/latex": [
       "3037"
      ],
      "text/markdown": [
       "3037"
      ],
      "text/plain": [
       "[1] 3037"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(top_eGene_effects)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "6dbcab52-3ced-4885-b76c-ac3f8d97ce46",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 3 × 28</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>ENSG</th><th scope=col>⋯</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>n_e_samples</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG_RPL8 </td><td>19:58497199:G:A</td><td>9.263029e-06</td><td>NA</td><td>NA</td><td>NA</td><td>19</td><td>58345178</td><td>58353499</td><td>ENSG00000121410</td><td>⋯</td><td>A1BG_RPL8;19:58497199:G:A</td><td>990</td><td>990</td><td>RPL8 </td><td>A1BG</td><td>CD4_T</td><td>oneK1K</td><td>1</td><td>37478</td><td>3.471598e-01</td></tr>\n",
       "\t<tr><td>A2M_PZP   </td><td>12:8992460:T:C </td><td>3.856050e-23</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td> 9067664</td><td> 9116229</td><td>ENSG00000126838</td><td>⋯</td><td>A2M_PZP;12:8992460:T:C   </td><td>837</td><td>837</td><td>PZP  </td><td>A2M </td><td>CD4_T</td><td>oneK1K</td><td>0</td><td>97356</td><td>3.754096e-18</td></tr>\n",
       "\t<tr><td>AAK1_CDIN1</td><td>2:69517668:C:T </td><td>4.342112e-05</td><td>NA</td><td>NA</td><td>NA</td><td> 2</td><td>69457997</td><td>69674349</td><td>ENSG00000115977</td><td>⋯</td><td>AAK1_CDIN1;2:69517668:C:T</td><td>852</td><td>852</td><td>CDIN1</td><td>AAK1</td><td>CD4_T</td><td>oneK1K</td><td>0</td><td>14994</td><td>6.510562e-01</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 3 × 28\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ENSG & ⋯ & triplet & n\\_samples & n\\_e\\_samples & gene1 & gene2 & cell\\_type & dataset & is\\_rb\\_mt & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene\\\\\n",
       " <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & ⋯ & <chr> & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t A1BG\\_RPL8  & 19:58497199:G:A & 9.263029e-06 & NA & NA & NA & 19 & 58345178 & 58353499 & ENSG00000121410 & ⋯ & A1BG\\_RPL8;19:58497199:G:A & 990 & 990 & RPL8  & A1BG & CD4\\_T & oneK1K & 1 & 37478 & 3.471598e-01\\\\\n",
       "\t A2M\\_PZP    & 12:8992460:T:C  & 3.856050e-23 & NA & NA & NA & 12 &  9067664 &  9116229 & ENSG00000126838 & ⋯ & A2M\\_PZP;12:8992460:T:C    & 837 & 837 & PZP   & A2M  & CD4\\_T & oneK1K & 0 & 97356 & 3.754096e-18\\\\\n",
       "\t AAK1\\_CDIN1 & 2:69517668:C:T  & 4.342112e-05 & NA & NA & NA &  2 & 69457997 & 69674349 & ENSG00000115977 & ⋯ & AAK1\\_CDIN1;2:69517668:C:T & 852 & 852 & CDIN1 & AAK1 & CD4\\_T & oneK1K & 0 & 14994 & 6.510562e-01\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 3 × 28\n",
       "\n",
       "| feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ENSG &lt;chr&gt; | ⋯ ⋯ | triplet &lt;chr&gt; | n_samples &lt;int&gt; | n_e_samples &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A1BG_RPL8  | 19:58497199:G:A | 9.263029e-06 | NA | NA | NA | 19 | 58345178 | 58353499 | ENSG00000121410 | ⋯ | A1BG_RPL8;19:58497199:G:A | 990 | 990 | RPL8  | A1BG | CD4_T | oneK1K | 1 | 37478 | 3.471598e-01 |\n",
       "| A2M_PZP    | 12:8992460:T:C  | 3.856050e-23 | NA | NA | NA | 12 |  9067664 |  9116229 | ENSG00000126838 | ⋯ | A2M_PZP;12:8992460:T:C    | 837 | 837 | PZP   | A2M  | CD4_T | oneK1K | 0 | 97356 | 3.754096e-18 |\n",
       "| AAK1_CDIN1 | 2:69517668:C:T  | 4.342112e-05 | NA | NA | NA |  2 | 69457997 | 69674349 | ENSG00000115977 | ⋯ | AAK1_CDIN1;2:69517668:C:T | 852 | 852 | CDIN1 | AAK1 | CD4_T | oneK1K | 0 | 14994 | 6.510562e-01 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id snp_id          p_value      beta beta_se\n",
       "1 A1BG_RPL8  19:58497199:G:A 9.263029e-06 NA   NA     \n",
       "2 A2M_PZP    12:8992460:T:C  3.856050e-23 NA   NA     \n",
       "3 AAK1_CDIN1 2:69517668:C:T  4.342112e-05 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start feature_end\n",
       "1 NA                        19                 58345178      58353499   \n",
       "2 NA                        12                  9067664       9116229   \n",
       "3 NA                         2                 69457997      69674349   \n",
       "  ENSG            ⋯ triplet                   n_samples n_e_samples gene1 gene2\n",
       "1 ENSG00000121410 ⋯ A1BG_RPL8;19:58497199:G:A 990       990         RPL8  A1BG \n",
       "2 ENSG00000126838 ⋯ A2M_PZP;12:8992460:T:C    837       837         PZP   A2M  \n",
       "3 ENSG00000115977 ⋯ AAK1_CDIN1;2:69517668:C:T 852       852         CDIN1 AAK1 \n",
       "  cell_type dataset is_rb_mt amount_tests_eGene p_val_mt_eGene\n",
       "1 CD4_T     oneK1K  1        37478              3.471598e-01  \n",
       "2 CD4_T     oneK1K  0        97356              3.754096e-18  \n",
       "3 CD4_T     oneK1K  0        14994              6.510562e-01  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(top_eGene_effects,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "916c4fdf-4a65-4aa0-9e3d-095e02d03084",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_eGene_effects$p_val_mt_eGene[top_eGene_effects$p_val_mt_eGene > 1] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "6bb50b8a-8e97-43f4-b38c-d166fd32cf4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Correct top-effects with q_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "e7acd88b-1d90-43a9-87b8-86772eb8a47f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#top_eGene_effects = top_eGene_effects %>% mutate(p_bh = p.adjust(p_val_mt_eGene, method = \"BH\"))\n",
    "\n",
    "top_eGene_effects$q_val = qvalue(p = top_eGene_effects$p_val_mt_eGene)$qvalues\n",
    "\n",
    "#top_eGene_effects$p_bh = qvalue(p = top_eGene_effects$p_val_mt_eGene,pi0 = 1)$qvalues\n",
    "\n",
    "top_eGene_effects$p_bh =  p.adjust(top_eGene_effects$p_val_mt_eGene, method = \"BH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "b2fa07a4-69d2-48b7-a2ae-75c591a238aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 3 × 30</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>ENSG</th><th scope=col>⋯</th><th scope=col>n_e_samples</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th><th scope=col>q_val</th><th scope=col>p_bh</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG_RPL8 </td><td>19:58497199:G:A</td><td>9.263029e-06</td><td>NA</td><td>NA</td><td>NA</td><td>19</td><td>58345178</td><td>58353499</td><td>ENSG00000121410</td><td>⋯</td><td>990</td><td>RPL8 </td><td>A1BG</td><td>CD4_T</td><td>oneK1K</td><td>1</td><td>37478</td><td>3.471598e-01</td><td>9.798553e-01</td><td>9.798553e-01</td></tr>\n",
       "\t<tr><td>A2M_PZP   </td><td>12:8992460:T:C </td><td>3.856050e-23</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td> 9067664</td><td> 9116229</td><td>ENSG00000126838</td><td>⋯</td><td>837</td><td>PZP  </td><td>A2M </td><td>CD4_T</td><td>oneK1K</td><td>0</td><td>97356</td><td>3.754096e-18</td><td>4.256168e-16</td><td>4.256168e-16</td></tr>\n",
       "\t<tr><td>AAK1_CDIN1</td><td>2:69517668:C:T </td><td>4.342112e-05</td><td>NA</td><td>NA</td><td>NA</td><td> 2</td><td>69457997</td><td>69674349</td><td>ENSG00000115977</td><td>⋯</td><td>852</td><td>CDIN1</td><td>AAK1</td><td>CD4_T</td><td>oneK1K</td><td>0</td><td>14994</td><td>6.510562e-01</td><td>1.000000e+00</td><td>1.000000e+00</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 3 × 30\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ENSG & ⋯ & n\\_e\\_samples & gene1 & gene2 & cell\\_type & dataset & is\\_rb\\_mt & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene & q\\_val & p\\_bh\\\\\n",
       " <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & ⋯ & <int> & <chr> & <chr> & <chr> & <chr> & <dbl> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t A1BG\\_RPL8  & 19:58497199:G:A & 9.263029e-06 & NA & NA & NA & 19 & 58345178 & 58353499 & ENSG00000121410 & ⋯ & 990 & RPL8  & A1BG & CD4\\_T & oneK1K & 1 & 37478 & 3.471598e-01 & 9.798553e-01 & 9.798553e-01\\\\\n",
       "\t A2M\\_PZP    & 12:8992460:T:C  & 3.856050e-23 & NA & NA & NA & 12 &  9067664 &  9116229 & ENSG00000126838 & ⋯ & 837 & PZP   & A2M  & CD4\\_T & oneK1K & 0 & 97356 & 3.754096e-18 & 4.256168e-16 & 4.256168e-16\\\\\n",
       "\t AAK1\\_CDIN1 & 2:69517668:C:T  & 4.342112e-05 & NA & NA & NA &  2 & 69457997 & 69674349 & ENSG00000115977 & ⋯ & 852 & CDIN1 & AAK1 & CD4\\_T & oneK1K & 0 & 14994 & 6.510562e-01 & 1.000000e+00 & 1.000000e+00\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 3 × 30\n",
       "\n",
       "| feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ENSG &lt;chr&gt; | ⋯ ⋯ | n_e_samples &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; | q_val &lt;dbl&gt; | p_bh &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A1BG_RPL8  | 19:58497199:G:A | 9.263029e-06 | NA | NA | NA | 19 | 58345178 | 58353499 | ENSG00000121410 | ⋯ | 990 | RPL8  | A1BG | CD4_T | oneK1K | 1 | 37478 | 3.471598e-01 | 9.798553e-01 | 9.798553e-01 |\n",
       "| A2M_PZP    | 12:8992460:T:C  | 3.856050e-23 | NA | NA | NA | 12 |  9067664 |  9116229 | ENSG00000126838 | ⋯ | 837 | PZP   | A2M  | CD4_T | oneK1K | 0 | 97356 | 3.754096e-18 | 4.256168e-16 | 4.256168e-16 |\n",
       "| AAK1_CDIN1 | 2:69517668:C:T  | 4.342112e-05 | NA | NA | NA |  2 | 69457997 | 69674349 | ENSG00000115977 | ⋯ | 852 | CDIN1 | AAK1 | CD4_T | oneK1K | 0 | 14994 | 6.510562e-01 | 1.000000e+00 | 1.000000e+00 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id snp_id          p_value      beta beta_se\n",
       "1 A1BG_RPL8  19:58497199:G:A 9.263029e-06 NA   NA     \n",
       "2 A2M_PZP    12:8992460:T:C  3.856050e-23 NA   NA     \n",
       "3 AAK1_CDIN1 2:69517668:C:T  4.342112e-05 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start feature_end\n",
       "1 NA                        19                 58345178      58353499   \n",
       "2 NA                        12                  9067664       9116229   \n",
       "3 NA                         2                 69457997      69674349   \n",
       "  ENSG            ⋯ n_e_samples gene1 gene2 cell_type dataset is_rb_mt\n",
       "1 ENSG00000121410 ⋯ 990         RPL8  A1BG  CD4_T     oneK1K  1       \n",
       "2 ENSG00000126838 ⋯ 837         PZP   A2M   CD4_T     oneK1K  0       \n",
       "3 ENSG00000115977 ⋯ 852         CDIN1 AAK1  CD4_T     oneK1K  0       \n",
       "  amount_tests_eGene p_val_mt_eGene q_val        p_bh        \n",
       "1 37478              3.471598e-01   9.798553e-01 9.798553e-01\n",
       "2 97356              3.754096e-18   4.256168e-16 4.256168e-16\n",
       "3 14994              6.510562e-01   1.000000e+00 1.000000e+00"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(top_eGene_effects,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "22c9b97f-a9e2-48a4-bc01-48fed644b7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the cutoff p-value\n",
    "p_val_cutoff_bh = head(top_eGene_effects[top_eGene_effects$p_bh < 0.05,c('p_val_mt_eGene', 'p_bh')] %>% arrange(-p_bh),1)\n",
    "p_val_cutoff_q = head(top_eGene_effects[top_eGene_effects$q_val < 0.05,c('p_val_mt_eGene', 'q_val')] %>% arrange(-q_val),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "e0ae0e06-bc83-455c-93c9-558002a1d748",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>p_val_mt_eGene</th><th scope=col>p_bh</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.00266828</td><td>0.04941199</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 2\n",
       "\\begin{tabular}{ll}\n",
       " p\\_val\\_mt\\_eGene & p\\_bh\\\\\n",
       " <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 0.00266828 & 0.04941199\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 2\n",
       "\n",
       "| p_val_mt_eGene &lt;dbl&gt; | p_bh &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| 0.00266828 | 0.04941199 |\n",
       "\n"
      ],
      "text/plain": [
       "  p_val_mt_eGene p_bh      \n",
       "1 0.00266828     0.04941199"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "p_val_cutoff_bh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "ea11930a-09f1-46c7-b128-734fe31d7940",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add the cutoff p-value to the result\n",
    "coeqtl_results$p_bh_cutoff = p_val_cutoff_bh$p_val_mt_eGene\n",
    "coeqtl_results$p_q_cutoff = p_val_cutoff_q$p_val_mt_eGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2894174-8a6c-4fed-9c92-c2cb1403bcc2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3c72cfda-5a4b-4731-9a73-1a1fc6af5362",
   "metadata": {},
   "source": [
    "###  Check for duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "92927a4b-f4ab-41ed-b777-04091e66b640",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results = setDT(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "524a36c5-5981-4cb2-8d57-a37a456fed13",
   "metadata": {},
   "outputs": [],
   "source": [
    "dupl_invest =  coeqtl_results[, .(.N), by = .(feature_id, snp_id, dataset, cell_type)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "3cfee8f4-088e-4984-b31d-a68034138bb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "dupl_invest = dupl_invest[dupl_invest$N >= 2,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "543419e3-d2f2-4a71-a50a-0621dc382a83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "71691"
      ],
      "text/latex": [
       "71691"
      ],
      "text/markdown": [
       "71691"
      ],
      "text/plain": [
       "[1] 71691"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(dupl_invest)  # amount of dupicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "5cbd90ff-3605-4da7-aa26-5e08a16f6a55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>N</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG_MZF1-AS1</td><td>19:57620738:A:C  </td><td>oneK1K</td><td>CD4_T</td><td>2</td></tr>\n",
       "\t<tr><td>A1BG_MZF1-AS1</td><td>19:57738275:G:C  </td><td>oneK1K</td><td>CD4_T</td><td>2</td></tr>\n",
       "\t<tr><td>A1BG_MZF1-AS1</td><td>19:58357814:C:A  </td><td>oneK1K</td><td>CD4_T</td><td>2</td></tr>\n",
       "\t<tr><td>A1BG_MZF1-AS1</td><td>19:58358468:AGT:A</td><td>oneK1K</td><td>CD4_T</td><td>2</td></tr>\n",
       "\t<tr><td>A1BG_MZF1-AS1</td><td>19:58497199:G:A  </td><td>oneK1K</td><td>CD4_T</td><td>2</td></tr>\n",
       "\t<tr><td>A1BG_MZF1-AS1</td><td>19:58501767:G:A  </td><td>oneK1K</td><td>CD4_T</td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " feature\\_id & snp\\_id & dataset & cell\\_type & N\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t A1BG\\_MZF1-AS1 & 19:57620738:A:C   & oneK1K & CD4\\_T & 2\\\\\n",
       "\t A1BG\\_MZF1-AS1 & 19:57738275:G:C   & oneK1K & CD4\\_T & 2\\\\\n",
       "\t A1BG\\_MZF1-AS1 & 19:58357814:C:A   & oneK1K & CD4\\_T & 2\\\\\n",
       "\t A1BG\\_MZF1-AS1 & 19:58358468:AGT:A & oneK1K & CD4\\_T & 2\\\\\n",
       "\t A1BG\\_MZF1-AS1 & 19:58497199:G:A   & oneK1K & CD4\\_T & 2\\\\\n",
       "\t A1BG\\_MZF1-AS1 & 19:58501767:G:A   & oneK1K & CD4\\_T & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 5\n",
       "\n",
       "| feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | dataset &lt;chr&gt; | cell_type &lt;chr&gt; | N &lt;int&gt; |\n",
       "|---|---|---|---|---|\n",
       "| A1BG_MZF1-AS1 | 19:57620738:A:C   | oneK1K | CD4_T | 2 |\n",
       "| A1BG_MZF1-AS1 | 19:57738275:G:C   | oneK1K | CD4_T | 2 |\n",
       "| A1BG_MZF1-AS1 | 19:58357814:C:A   | oneK1K | CD4_T | 2 |\n",
       "| A1BG_MZF1-AS1 | 19:58358468:AGT:A | oneK1K | CD4_T | 2 |\n",
       "| A1BG_MZF1-AS1 | 19:58497199:G:A   | oneK1K | CD4_T | 2 |\n",
       "| A1BG_MZF1-AS1 | 19:58501767:G:A   | oneK1K | CD4_T | 2 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id    snp_id            dataset cell_type N\n",
       "1 A1BG_MZF1-AS1 19:57620738:A:C   oneK1K  CD4_T     2\n",
       "2 A1BG_MZF1-AS1 19:57738275:G:C   oneK1K  CD4_T     2\n",
       "3 A1BG_MZF1-AS1 19:58357814:C:A   oneK1K  CD4_T     2\n",
       "4 A1BG_MZF1-AS1 19:58358468:AGT:A oneK1K  CD4_T     2\n",
       "5 A1BG_MZF1-AS1 19:58497199:G:A   oneK1K  CD4_T     2\n",
       "6 A1BG_MZF1-AS1 19:58501767:G:A   oneK1K  CD4_T     2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(dupl_invest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "bf4e4912-ac43-4ee8-9d2a-c973545568ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### put togehter identifier of duplicates\n",
    "dupl_invest$triplet = paste0(dupl_invest$snp_id, ';', dupl_invest$feature_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "69ebd599-6b01-408b-861f-f8a27e056aaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "### put together identifier for all results\n",
    "coeqtl_results$triplet = paste0(coeqtl_results$snp_id, ';', coeqtl_results$feature_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "99ec1fe4-2b21-49df-9ed9-1ea423ff19f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get only unique values (if there are duplictes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "27803b73-70bd-4ae0-98ee-376b5e284742",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(nrow(dupl_invest)  > 0){\n",
    "    ## Previous solution\n",
    "    # coeqtl_results = coeqtl_results %>% group_by(feature_id, snp_id,   cell_type, dataset, feature_chromosome, feature_end, feature_start) %>% summarise(beta_se = max(beta_se), p_value = mean(p_value),beta = mean(beta)\n",
    "    \n",
    "    coeqtl_results_un = coeqtl_results[!coeqtl_results$triplet %in% dupl_invest$triplet,]\n",
    "    coeqtl_results_un$type = 'unique'\n",
    "    \n",
    "    coeqtl_results_dupl = coeqtl_results[coeqtl_results$triplet %in% dupl_invest$triplet,]\n",
    "    coeqtl_results_dupl$eGene = coeqtl_results_dupl$feature_id\n",
    "    coeqtl_results_dupl$coeGene = coeqtl_results_dupl$feature_id\n",
    "    coeqtl_results_dupl$type = 'duplicate'\n",
    "    \n",
    "    coeqtl_results_dupl=    coeqtl_results_dupl %>% group_by(triplet) %>% mutate(rank = rank(eGene, ties.method = 'first'))\n",
    "    coeqtl_results_dupl =  coeqtl_results_dupl[ coeqtl_results_dupl$rank == 1,]\n",
    "    coeqtl_results_dupl$rank = NULL\n",
    "    \n",
    "    coeqtl_results = rbind(coeqtl_results_un, coeqtl_results_dupl)\n",
    "    \n",
    "    \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "d7071b3e-fce7-43bc-a1a9-207d6c69dcbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "149856468"
      ],
      "text/latex": [
       "149856468"
      ],
      "text/markdown": [
       "149856468"
      ],
      "text/plain": [
       "[1] 149856468"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "45719152-d094-48fe-9c02-7a2a4fcbfa69",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results = setDT(coeqtl_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7618fd86-23fd-4819-bb87-68a227ef512b",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Analyze results "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9878308-4ff0-4264-b3d3-35ad8b6b5540",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Check significant triplets with different multiple testing corrections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "b476b894-c9b5-4994-99fd-56eec4064cba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>ENSG</th><th scope=col>⋯</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th><th scope=col>p_bh_cutoff</th><th scope=col>p_q_cutoff</th><th scope=col>type</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>ENSG00000175899</td><td>⋯</td><td>A2M</td><td>A1BG</td><td>CD4_T</td><td>oneK1K</td><td>0</td><td>97356</td><td>1519.477</td><td>0.00266828</td><td>0.00266828</td><td>unique</td></tr>\n",
       "\t<tr><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>9116229</td><td>ENSG00000175899</td><td>⋯</td><td>A2M</td><td>A1BG</td><td>CD4_T</td><td>oneK1K</td><td>0</td><td>97356</td><td> 162.928</td><td>0.00266828</td><td>0.00266828</td><td>unique</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 31\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & feature\\_end & ENSG & ⋯ & gene1 & gene2 & cell\\_type & dataset & is\\_rb\\_mt & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene & p\\_bh\\_cutoff & p\\_q\\_cutoff & type\\\\\n",
       " <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & <int> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & 9116229 & ENSG00000175899 & ⋯ & A2M & A1BG & CD4\\_T & oneK1K & 0 & 97356 & 1519.477 & 0.00266828 & 0.00266828 & unique\\\\\n",
       "\t A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & 9116229 & ENSG00000175899 & ⋯ & A2M & A1BG & CD4\\_T & oneK1K & 0 & 97356 &  162.928 & 0.00266828 & 0.00266828 & unique\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 31\n",
       "\n",
       "| feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | ENSG &lt;chr&gt; | ⋯ ⋯ | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; | p_bh_cutoff &lt;dbl&gt; | p_q_cutoff &lt;dbl&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | 9116229 | ENSG00000175899 | ⋯ | A2M | A1BG | CD4_T | oneK1K | 0 | 97356 | 1519.477 | 0.00266828 | 0.00266828 | unique |\n",
       "| A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | 9116229 | ENSG00000175899 | ⋯ | A2M | A1BG | CD4_T | oneK1K | 0 | 97356 |  162.928 | 0.00266828 | 0.00266828 | unique |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id snp_id         p_value     beta beta_se empirical_feature_p_value\n",
       "1 A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA      NA                       \n",
       "2 A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA      NA                       \n",
       "  feature_chromosome feature_start feature_end ENSG            ⋯ gene1 gene2\n",
       "1 12                 9067664       9116229     ENSG00000175899 ⋯ A2M   A1BG \n",
       "2 12                 9067664       9116229     ENSG00000175899 ⋯ A2M   A1BG \n",
       "  cell_type dataset is_rb_mt amount_tests_eGene p_val_mt_eGene p_bh_cutoff\n",
       "1 CD4_T     oneK1K  0        97356              1519.477       0.00266828 \n",
       "2 CD4_T     oneK1K  0        97356               162.928       0.00266828 \n",
       "  p_q_cutoff type  \n",
       "1 0.00266828 unique\n",
       "2 0.00266828 unique"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtl_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "d637bac7-c547-4a7f-b1c3-955244a2a133",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Amount of tests per dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "21a16f75-5c03-47f3-91f9-3f0950521b92",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "amount_tests = coeqtl_results %>% group_by( dataset, cell_type) %>% summarise(amount_tests_datasets = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "5bf22e2f-ae32-4351-8872-4ad623abe2f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>amount_tests_datasets</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>oneK1K</td><td>CD4_T</td><td>149856468</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 1 × 3\n",
       "\\begin{tabular}{lll}\n",
       " dataset & cell\\_type & amount\\_tests\\_datasets\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t oneK1K & CD4\\_T & 149856468\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 1 × 3\n",
       "\n",
       "| dataset &lt;chr&gt; | cell_type &lt;chr&gt; | amount_tests_datasets &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| oneK1K | CD4_T | 149856468 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset cell_type amount_tests_datasets\n",
       "1 oneK1K  CD4_T     149856468            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "amount_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "2f39cc96-6cb5-48f5-b515-13f444651b65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "149856468"
      ],
      "text/latex": [
       "149856468"
      ],
      "text/markdown": [
       "149856468"
      ],
      "text/plain": [
       "[1] 149856468"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "f453a634-b7d9-4a56-9b9c-ed43e3e41b44",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Add to result and correct p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "d6835459-4e05-490b-8827-13e641f28aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results = merge(coeqtl_results, amount_tests[,c( 'dataset', 'cell_type', 'amount_tests_datasets')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "b15399a4-056a-457d-b7d2-bced96d926bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "149856468"
      ],
      "text/latex": [
       "149856468"
      ],
      "text/markdown": [
       "149856468"
      ],
      "text/plain": [
       "[1] 149856468"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtl_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "f9367ae7-17f2-4fb4-ad86-43920e7b0b95",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results$p_val_mt = coeqtl_results$p_value *  coeqtl_results$amount_tests_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "f503ec49-1e62-472d-ba44-82d42d301df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Correct with Benjamini hochberg\n",
    "coeqtl_results = coeqtl_results %>% group_by(dataset, cell_type) %>% mutate(p_val_corrected = p.adjust(p_value, method=\"BH\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "ac75ffcf-57be-47d9-89a6-11c4ff60b18c",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results$p_value[coeqtl_results$p_value  > 1] = 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "8d88544b-2b43-4b02-9a04-b4fd71a8298b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Alternative qvalue correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "0d792138-2706-4df8-a943-e75f7a47bd4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sum(is.na(coeqtl_results$p_value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d1bdefa-72bf-401c-b29c-0c9696acf8a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "6ea75dd3-edb4-4b58-930f-0c164ba4af73",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>amount_sign</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>oneK1K</td><td>CD4_T</td><td>96189</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 1 × 3\n",
       "\\begin{tabular}{lll}\n",
       " dataset & cell\\_type & amount\\_sign\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t oneK1K & CD4\\_T & 96189\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 1 × 3\n",
       "\n",
       "| dataset &lt;chr&gt; | cell_type &lt;chr&gt; | amount_sign &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| oneK1K | CD4_T | 96189 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset cell_type amount_sign\n",
       "1 oneK1K  CD4_T     96189      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Significant per dataset (according to BH corrections) - 0.05\n",
    "coeqtl_results %>% group_by(dataset, cell_type) %>% summarise(amount_sign = sum(p_val_corrected <= 0.05, na.rm = TRUE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "486dd72f-09b5-4c98-8a59-bdf7b3ee78b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'dataset'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>amount_sign</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>oneK1K</td><td>CD4_T</td><td>114232</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 1 × 3\n",
       "\\begin{tabular}{lll}\n",
       " dataset & cell\\_type & amount\\_sign\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t oneK1K & CD4\\_T & 114232\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 1 × 3\n",
       "\n",
       "| dataset &lt;chr&gt; | cell_type &lt;chr&gt; | amount_sign &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| oneK1K | CD4_T | 114232 |\n",
       "\n"
      ],
      "text/plain": [
       "  dataset cell_type amount_sign\n",
       "1 oneK1K  CD4_T     114232     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Significant per dataset (according to BH corrections) - 0.1\n",
    "coeqtl_results %>% group_by(dataset, cell_type) %>% summarise(amount_sign = sum(p_val_corrected <= 0.1, na.rm = TRUE))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f070a5ff-31a4-4254-91cf-a3bb5cb6afc9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Analyze structure of co-eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "99c1a4a3-3e20-438d-b2d6-b22c46d983c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## only significant results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "21677c5c-b40f-4afa-b0ca-de7dfa4f1717",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 34</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>⋯</th><th scope=col>gene2</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th><th scope=col>p_bh_cutoff</th><th scope=col>p_q_cutoff</th><th scope=col>type</th><th scope=col>amount_tests_datasets</th><th scope=col>p_val_mt</th><th scope=col>p_val_corrected</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CD4_T</td><td>oneK1K</td><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>⋯</td><td>A1BG</td><td>0</td><td>97356</td><td>1519.477</td><td>0.00266828</td><td>0.00266828</td><td>unique</td><td>149856468</td><td>2338874</td><td>0.8627429</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>oneK1K</td><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>⋯</td><td>A1BG</td><td>0</td><td>97356</td><td> 162.928</td><td>0.00266828</td><td>0.00266828</td><td>unique</td><td>149856468</td><td> 250789</td><td>0.5682995</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 34\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cell\\_type & dataset & feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & ⋯ & gene2 & is\\_rb\\_mt & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene & p\\_bh\\_cutoff & p\\_q\\_cutoff & type & amount\\_tests\\_datasets & p\\_val\\_mt & p\\_val\\_corrected\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & ⋯ & <chr> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <chr> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t CD4\\_T & oneK1K & A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & ⋯ & A1BG & 0 & 97356 & 1519.477 & 0.00266828 & 0.00266828 & unique & 149856468 & 2338874 & 0.8627429\\\\\n",
       "\t CD4\\_T & oneK1K & A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & ⋯ & A1BG & 0 & 97356 &  162.928 & 0.00266828 & 0.00266828 & unique & 149856468 &  250789 & 0.5682995\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 34\n",
       "\n",
       "| cell_type &lt;chr&gt; | dataset &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | ⋯ ⋯ | gene2 &lt;chr&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; | p_bh_cutoff &lt;dbl&gt; | p_q_cutoff &lt;dbl&gt; | type &lt;chr&gt; | amount_tests_datasets &lt;int&gt; | p_val_mt &lt;dbl&gt; | p_val_corrected &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| CD4_T | oneK1K | A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | ⋯ | A1BG | 0 | 97356 | 1519.477 | 0.00266828 | 0.00266828 | unique | 149856468 | 2338874 | 0.8627429 |\n",
       "| CD4_T | oneK1K | A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | ⋯ | A1BG | 0 | 97356 |  162.928 | 0.00266828 | 0.00266828 | unique | 149856468 |  250789 | 0.5682995 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type dataset feature_id snp_id         p_value     beta beta_se\n",
       "1 CD4_T     oneK1K  A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA     \n",
       "2 CD4_T     oneK1K  A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start ⋯ gene2 is_rb_mt\n",
       "1 NA                        12                 9067664       ⋯ A1BG  0       \n",
       "2 NA                        12                 9067664       ⋯ A1BG  0       \n",
       "  amount_tests_eGene p_val_mt_eGene p_bh_cutoff p_q_cutoff type  \n",
       "1 97356              1519.477       0.00266828  0.00266828 unique\n",
       "2 97356               162.928       0.00266828  0.00266828 unique\n",
       "  amount_tests_datasets p_val_mt p_val_corrected\n",
       "1 149856468             2338874  0.8627429      \n",
       "2 149856468              250789  0.5682995      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtl_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "7dd093a4-ed55-4b97-832f-49c27acb013a",
   "metadata": {},
   "outputs": [],
   "source": [
    "thres = 0.05\n",
    "column = 'p_value'  # p_val_mt , q_val, p_value = nominal p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "40af9197-d8af-489d-a5b2-7fc1495d3d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#coeqtl_results[,column]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "0d7d0e6a-b094-4c86-bd5c-3231daa3b615",
   "metadata": {},
   "outputs": [],
   "source": [
    "### get sign results\n",
    "sign_results = coeqtl_results[coeqtl_results[,column] < thres,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "ff642562-e7cd-415b-8667-a17be9858d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_results = sign_results[!is.na(sign_results$p_value),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "419c2b24-4250-4e85-9f3b-4fec61a67c81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "8050384"
      ],
      "text/latex": [
       "8050384"
      ],
      "text/markdown": [
       "8050384"
      ],
      "text/plain": [
       "[1] 8050384"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_results )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "a25f9a19-8e36-4923-a52f-3f55ca97f4d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3652020"
      ],
      "text/latex": [
       "3652020"
      ],
      "text/markdown": [
       "3652020"
      ],
      "text/plain": [
       "[1] 3652020"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Unique gene-pairs\n",
    "length(unique(sign_results$feature_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "46d0744f-005b-42de-a488-e7a32dd00853",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "7247"
      ],
      "text/latex": [
       "7247"
      ],
      "text/markdown": [
       "7247"
      ],
      "text/plain": [
       "[1] 7247"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Unique SNPs\n",
    "length(unique(sign_results$snp_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "65db4667-abc3-4eac-8a94-7ef240084329",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4858"
      ],
      "text/latex": [
       "4858"
      ],
      "text/markdown": [
       "4858"
      ],
      "text/plain": [
       "[1] 4858"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Unique eGenes\n",
    "length(unique(sign_results$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "8550b1e9-948f-44e2-af99-73dc5d1d77fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "9315"
      ],
      "text/latex": [
       "9315"
      ],
      "text/markdown": [
       "9315"
      ],
      "text/plain": [
       "[1] 9315"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Unique co _eGenes\n",
    "length(unique(sign_results$coeGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "4874c909-f752-4132-81c3-d3d2bf7cde52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "8050384"
      ],
      "text/latex": [
       "8050384"
      ],
      "text/markdown": [
       "8050384"
      ],
      "text/plain": [
       "[1] 8050384"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "213a0573-aa70-40b4-982f-94f27270fa85",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "4a5e47dc-7f70-4072-bbc2-1465f9cfed5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Analyze amount of co-egenes per Gene1 + SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "52436233-b28c-4106-a49c-18d34f48bfab",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'snp_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "amount_co_egenes = sign_results %>% group_by( snp_id, eGene) %>% summarise(amount_co_e_genes = n()) ## TBD add: , snp_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "5fe165c3-e1cf-492e-a3c3-2154a35b5f31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 15 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>snp_id</th><th scope=col>eGene</th><th scope=col>amount_co_e_genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>12:56042145:C:G</td><td>RPS26</td><td>5472</td></tr>\n",
       "\t<tr><td>12:56007301:G:A</td><td>RPS26</td><td>5441</td></tr>\n",
       "\t<tr><td>12:56076841:C:T</td><td>RPS26</td><td>5440</td></tr>\n",
       "\t<tr><td>12:56083910:A:T</td><td>RPS26</td><td>5415</td></tr>\n",
       "\t<tr><td>12:55991020:G:A</td><td>RPS26</td><td>5345</td></tr>\n",
       "\t<tr><td>12:55995509:T:C</td><td>RPS26</td><td>5333</td></tr>\n",
       "\t<tr><td>12:56001170:A:G</td><td>RPS26</td><td>5307</td></tr>\n",
       "\t<tr><td>12:56061510:C:T</td><td>RPS26</td><td>5149</td></tr>\n",
       "\t<tr><td>12:56051582:G:A</td><td>RPS26</td><td>5147</td></tr>\n",
       "\t<tr><td>12:56023144:A:C</td><td>RPS26</td><td>5144</td></tr>\n",
       "\t<tr><td>12:56041720:G:C</td><td>RPS26</td><td>5144</td></tr>\n",
       "\t<tr><td>12:56018703:T:G</td><td>RPS26</td><td>5143</td></tr>\n",
       "\t<tr><td>12:56088396:T:G</td><td>RPS26</td><td>5076</td></tr>\n",
       "\t<tr><td>12:55975722:A:G</td><td>RPS26</td><td>4998</td></tr>\n",
       "\t<tr><td>12:55991795:C:T</td><td>RPS26</td><td>4858</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 15 × 3\n",
       "\\begin{tabular}{lll}\n",
       " snp\\_id & eGene & amount\\_co\\_e\\_genes\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 12:56042145:C:G & RPS26 & 5472\\\\\n",
       "\t 12:56007301:G:A & RPS26 & 5441\\\\\n",
       "\t 12:56076841:C:T & RPS26 & 5440\\\\\n",
       "\t 12:56083910:A:T & RPS26 & 5415\\\\\n",
       "\t 12:55991020:G:A & RPS26 & 5345\\\\\n",
       "\t 12:55995509:T:C & RPS26 & 5333\\\\\n",
       "\t 12:56001170:A:G & RPS26 & 5307\\\\\n",
       "\t 12:56061510:C:T & RPS26 & 5149\\\\\n",
       "\t 12:56051582:G:A & RPS26 & 5147\\\\\n",
       "\t 12:56023144:A:C & RPS26 & 5144\\\\\n",
       "\t 12:56041720:G:C & RPS26 & 5144\\\\\n",
       "\t 12:56018703:T:G & RPS26 & 5143\\\\\n",
       "\t 12:56088396:T:G & RPS26 & 5076\\\\\n",
       "\t 12:55975722:A:G & RPS26 & 4998\\\\\n",
       "\t 12:55991795:C:T & RPS26 & 4858\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 15 × 3\n",
       "\n",
       "| snp_id &lt;chr&gt; | eGene &lt;chr&gt; | amount_co_e_genes &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 12:56042145:C:G | RPS26 | 5472 |\n",
       "| 12:56007301:G:A | RPS26 | 5441 |\n",
       "| 12:56076841:C:T | RPS26 | 5440 |\n",
       "| 12:56083910:A:T | RPS26 | 5415 |\n",
       "| 12:55991020:G:A | RPS26 | 5345 |\n",
       "| 12:55995509:T:C | RPS26 | 5333 |\n",
       "| 12:56001170:A:G | RPS26 | 5307 |\n",
       "| 12:56061510:C:T | RPS26 | 5149 |\n",
       "| 12:56051582:G:A | RPS26 | 5147 |\n",
       "| 12:56023144:A:C | RPS26 | 5144 |\n",
       "| 12:56041720:G:C | RPS26 | 5144 |\n",
       "| 12:56018703:T:G | RPS26 | 5143 |\n",
       "| 12:56088396:T:G | RPS26 | 5076 |\n",
       "| 12:55975722:A:G | RPS26 | 4998 |\n",
       "| 12:55991795:C:T | RPS26 | 4858 |\n",
       "\n"
      ],
      "text/plain": [
       "   snp_id          eGene amount_co_e_genes\n",
       "1  12:56042145:C:G RPS26 5472             \n",
       "2  12:56007301:G:A RPS26 5441             \n",
       "3  12:56076841:C:T RPS26 5440             \n",
       "4  12:56083910:A:T RPS26 5415             \n",
       "5  12:55991020:G:A RPS26 5345             \n",
       "6  12:55995509:T:C RPS26 5333             \n",
       "7  12:56001170:A:G RPS26 5307             \n",
       "8  12:56061510:C:T RPS26 5149             \n",
       "9  12:56051582:G:A RPS26 5147             \n",
       "10 12:56023144:A:C RPS26 5144             \n",
       "11 12:56041720:G:C RPS26 5144             \n",
       "12 12:56018703:T:G RPS26 5143             \n",
       "13 12:56088396:T:G RPS26 5076             \n",
       "14 12:55975722:A:G RPS26 4998             \n",
       "15 12:55991795:C:T RPS26 4858             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(amount_co_egenes %>% arrange(-amount_co_e_genes),15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "74e329e3-34d7-412c-86ee-2f499f03518b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 20 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>snp_id</th><th scope=col>eGene</th><th scope=col>amount_co_e_genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>17:7304878:C:T </td><td>EIF5A  </td><td>3894</td></tr>\n",
       "\t<tr><td>17:7303462:G:A </td><td>EIF5A  </td><td>3415</td></tr>\n",
       "\t<tr><td>22:42095730:G:T</td><td>SMDT1  </td><td>2162</td></tr>\n",
       "\t<tr><td>22:42074313:T:C</td><td>SMDT1  </td><td>2158</td></tr>\n",
       "\t<tr><td>6:31353491:T:G </td><td>HLA-B  </td><td>1817</td></tr>\n",
       "\t<tr><td>6:166960059:T:G</td><td>RNASET2</td><td>1698</td></tr>\n",
       "\t<tr><td>6:166993145:G:C</td><td>RNASET2</td><td>1686</td></tr>\n",
       "\t<tr><td>5:65541278:C:T </td><td>CENPK  </td><td>1685</td></tr>\n",
       "\t<tr><td>6:166978686:C:T</td><td>RNASET2</td><td>1680</td></tr>\n",
       "\t<tr><td>4:118283565:A:C</td><td>SNHG8  </td><td>1679</td></tr>\n",
       "\t<tr><td>6:166976754:A:G</td><td>RNASET2</td><td>1663</td></tr>\n",
       "\t<tr><td>6:166957622:T:C</td><td>RNASET2</td><td>1656</td></tr>\n",
       "\t<tr><td>17:7337072:C:T </td><td>EIF5A  </td><td>1625</td></tr>\n",
       "\t<tr><td>6:31353904:C:T </td><td>HLA-B  </td><td>1613</td></tr>\n",
       "\t<tr><td>17:7323638:G:A </td><td>EIF5A  </td><td>1576</td></tr>\n",
       "\t<tr><td>2:85705087:T:C </td><td>GNLY   </td><td>1528</td></tr>\n",
       "\t<tr><td>6:29952002:C:A </td><td>HLA-A  </td><td>1510</td></tr>\n",
       "\t<tr><td>6:166969587:G:A</td><td>RNASET2</td><td>1505</td></tr>\n",
       "\t<tr><td>6:166986857:T:C</td><td>RNASET2</td><td>1498</td></tr>\n",
       "\t<tr><td>6:166957044:C:G</td><td>RNASET2</td><td>1492</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 20 × 3\n",
       "\\begin{tabular}{lll}\n",
       " snp\\_id & eGene & amount\\_co\\_e\\_genes\\\\\n",
       " <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 17:7304878:C:T  & EIF5A   & 3894\\\\\n",
       "\t 17:7303462:G:A  & EIF5A   & 3415\\\\\n",
       "\t 22:42095730:G:T & SMDT1   & 2162\\\\\n",
       "\t 22:42074313:T:C & SMDT1   & 2158\\\\\n",
       "\t 6:31353491:T:G  & HLA-B   & 1817\\\\\n",
       "\t 6:166960059:T:G & RNASET2 & 1698\\\\\n",
       "\t 6:166993145:G:C & RNASET2 & 1686\\\\\n",
       "\t 5:65541278:C:T  & CENPK   & 1685\\\\\n",
       "\t 6:166978686:C:T & RNASET2 & 1680\\\\\n",
       "\t 4:118283565:A:C & SNHG8   & 1679\\\\\n",
       "\t 6:166976754:A:G & RNASET2 & 1663\\\\\n",
       "\t 6:166957622:T:C & RNASET2 & 1656\\\\\n",
       "\t 17:7337072:C:T  & EIF5A   & 1625\\\\\n",
       "\t 6:31353904:C:T  & HLA-B   & 1613\\\\\n",
       "\t 17:7323638:G:A  & EIF5A   & 1576\\\\\n",
       "\t 2:85705087:T:C  & GNLY    & 1528\\\\\n",
       "\t 6:29952002:C:A  & HLA-A   & 1510\\\\\n",
       "\t 6:166969587:G:A & RNASET2 & 1505\\\\\n",
       "\t 6:166986857:T:C & RNASET2 & 1498\\\\\n",
       "\t 6:166957044:C:G & RNASET2 & 1492\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 20 × 3\n",
       "\n",
       "| snp_id &lt;chr&gt; | eGene &lt;chr&gt; | amount_co_e_genes &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 17:7304878:C:T  | EIF5A   | 3894 |\n",
       "| 17:7303462:G:A  | EIF5A   | 3415 |\n",
       "| 22:42095730:G:T | SMDT1   | 2162 |\n",
       "| 22:42074313:T:C | SMDT1   | 2158 |\n",
       "| 6:31353491:T:G  | HLA-B   | 1817 |\n",
       "| 6:166960059:T:G | RNASET2 | 1698 |\n",
       "| 6:166993145:G:C | RNASET2 | 1686 |\n",
       "| 5:65541278:C:T  | CENPK   | 1685 |\n",
       "| 6:166978686:C:T | RNASET2 | 1680 |\n",
       "| 4:118283565:A:C | SNHG8   | 1679 |\n",
       "| 6:166976754:A:G | RNASET2 | 1663 |\n",
       "| 6:166957622:T:C | RNASET2 | 1656 |\n",
       "| 17:7337072:C:T  | EIF5A   | 1625 |\n",
       "| 6:31353904:C:T  | HLA-B   | 1613 |\n",
       "| 17:7323638:G:A  | EIF5A   | 1576 |\n",
       "| 2:85705087:T:C  | GNLY    | 1528 |\n",
       "| 6:29952002:C:A  | HLA-A   | 1510 |\n",
       "| 6:166969587:G:A | RNASET2 | 1505 |\n",
       "| 6:166986857:T:C | RNASET2 | 1498 |\n",
       "| 6:166957044:C:G | RNASET2 | 1492 |\n",
       "\n"
      ],
      "text/plain": [
       "   snp_id          eGene   amount_co_e_genes\n",
       "1  17:7304878:C:T  EIF5A   3894             \n",
       "2  17:7303462:G:A  EIF5A   3415             \n",
       "3  22:42095730:G:T SMDT1   2162             \n",
       "4  22:42074313:T:C SMDT1   2158             \n",
       "5  6:31353491:T:G  HLA-B   1817             \n",
       "6  6:166960059:T:G RNASET2 1698             \n",
       "7  6:166993145:G:C RNASET2 1686             \n",
       "8  5:65541278:C:T  CENPK   1685             \n",
       "9  6:166978686:C:T RNASET2 1680             \n",
       "10 4:118283565:A:C SNHG8   1679             \n",
       "11 6:166976754:A:G RNASET2 1663             \n",
       "12 6:166957622:T:C RNASET2 1656             \n",
       "13 17:7337072:C:T  EIF5A   1625             \n",
       "14 6:31353904:C:T  HLA-B   1613             \n",
       "15 17:7323638:G:A  EIF5A   1576             \n",
       "16 2:85705087:T:C  GNLY    1528             \n",
       "17 6:29952002:C:A  HLA-A   1510             \n",
       "18 6:166969587:G:A RNASET2 1505             \n",
       "19 6:166986857:T:C RNASET2 1498             \n",
       "20 6:166957044:C:G RNASET2 1492             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## check out non ribosomal genes\n",
    "head(amount_co_egenes[!amount_co_egenes$eGene %in% ribosomal_genes,] %>% arrange(-amount_co_e_genes),20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "3151d9c7-0008-4a71-aa1d-05124e82a844",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ribosomal_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "84cee7aa-b444-4a2d-a65d-f08daa811476",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "25077"
      ],
      "text/latex": [
       "25077"
      ],
      "text/markdown": [
       "25077"
      ],
      "text/plain": [
       "[1] 25077"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(unique(amount_co_egenes[,c('eGene', 'snp_id')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "fcdbf677-ed48-4d6c-9fc7-4f1451f33ca5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "20060"
      ],
      "text/latex": [
       "20060"
      ],
      "text/markdown": [
       "20060"
      ],
      "text/plain": [
       "[1] 20060"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(unique(amount_co_egenes[amount_co_egenes$amount_co_e_genes >= 5,c('eGene', 'snp_id')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "cc211628-12ae-4f14-adef-36539ed50e5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "      1     334     368     321     395    5472 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(amount_co_egenes$amount_co_e_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "id": "956f7363-b28e-4730-91dc-4413a67d5671",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzde2ANZ/7H8W8SJJELrUtVrgRpKxc2ro0iRYWWaonLKq2gqltLL6tFsdWiSott\nbe0P6UW1sS7V2g1dJQlxD216UFQUFUS0IvdE4vz+0MpJRDInOcmYx/v1lzPzmPmemXnmfDJX\nO7PZLAAAADA+e70LAAAAgG0Q7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQ\nBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4A\nAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEE\nOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAA\nRRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsKtQwgtN7G5o9/YJveuBtbIOr3lz\nVO8Qv3vcnWvb/7Emu3xwQe+6AIPaMq6enZ3dYx/n6V0IgJvdEcHu48fsLAxae3OLS0u7W7Rw\nfeY/Nq/h0N8Dimfg+0qizWeAsuXtmf5Qu8EzPv7fwZMXM/MKzXrXY3hsyRCzmY5UafQgVLc7\nItjhzpX22evzvue4AmBLBDvgNlZL7wJuf10+uGD+QO8iUEmJu3dfvfHBJ3J93KLHfNxq2+lY\nEWBMV8/viV7+yVdb95pOnjqTJSIxzzX1md+qpf8DbTt379Hj4YfaeLpwpADQH8EOKivKyMgp\n/tSmVz9fNzZ5wFrmcxte6Dvyn0mZJQbmXz5zZO+ZI3u3fvnRAmkwbsulpT31KhBAMfMd4KNH\nLb/xwDU3t0j7sJtFC5enN1qM2/GXe4pHhcz9qXjMtctJ/577/JNdg5rd4163Tq06ro28WgZ1\nCBsQ+cq8jzYlXci73uqLgRWsgkc/yrSsJefUtqVTnnm0o79HA3enWrXr1m/sG9ht0PNzVn93\nqfDmygtTty/5a792zRq41HF0b9S8fb8XFn17tsC8ZVwDy6J/Lu/r5Cf/d+4z3R/wuMvRXhr8\nJdZsNpsLfju+Y92Hb/xlcI/293s1dHOq7VDLyfWue5q16Tbg2Tc+3nk2v2QRsSXn9tO13/Yv\nf3lgR79Gro5O9Zve133krA3Hc6+Xey5+yV8fb9+ssWsdR9dGLToO/NtH31/Wuh6tWkT7X/Ur\nb6F3ez+torncDgtB68aQ+VFvy038i6uWI0tsgY9+lFs85qaNIffnmPnjHw1p1tC1Tp26d/m0\n7fP84u0XisqajoYtWYvCS0lr3pk49JEQv3sbuDnVdr6racvg0Mefm/P5ztPZlV4aVspL2Rk1\nY+zjoQ94NnR3ql27bv3Gzdv2Gv63Jd+eyimrYmt7XCXmYt16qfx30bYHM5vNP70X6ljBypcG\n47ZYTtvK7mOujh5kxQZDD7J2xhas2+o0fas7sZfZEMHObDZXKtgV/bSiv0d5Jx46zv/FbDZb\n15l/TZj3qOctDym5tX3ui58sN4OcpHd7Nbr5tKJ7+8nTBmkMdoe/X9SjYfEkrge7yx/2KLdi\nt4DnN6RY1FFijxw0evoTng6l/4tryBt7U3fN6nLXTeU6B7yyI0v7ytS6iKoc7HRfCNZsDCV/\nlgZFl9hLa/xZCn5u3ugHnG+akYP3U+tSb55OWaz8WcoyRUUGuN5qYsGWf0NZ3TW0upL4foTf\nrTJLHd8n/7E/w7J5ZXqc9XOxcr1Uci6a92Bm8/evtfhjoF3T/u9t++nfz9QTkb7/unDuRNLu\nrev+9cZzj3cMeWWbxdSt7T5mW/cgKzcYelDN9CAt7tBeZksEO7PZXJlgd+mTfk7lrC2pRLDL\nPfBWJ5cK2tp5DVtz/vfCrsREelfQ/I+if77F1wl4MuL+EjtPbcFORJy7/vP0jWmW2CPfin39\n+u5lj7F7YPr32takFYuo2oOdVOtCsHJjKPmzFLG6xJ+ZGn+Wbq3R2G/ySk+nLNb8LF39cckj\nd5c3McufJWu7hkYFpne73WJ13HBXz38e+2NpVq7HWTsXq9dL5eaifQ9mvho94MbPrPdL+8xm\n85Zx9aTUtlSKtd3HbNMeZP0GQw+qiR6kxZ3ay2zqDgx2FdMS7DY8dSPd1/nTxHU/nL2SW5CX\nfv7nIwcS/vPxvEnDHn6g3/tnLaZimtm6eCo+L++/ucrj80IsMpZT61FRu89k5uf9emzTW480\nsaiu4dMbM8xms/nQ3++3GOp438gVe37Jyss5d3DVuOC6Jb7PrYOdiNjd0/31fx84m5mTfvrA\nl3PGvLvLbDanRw0Pfmz87E837/rux5/PX84tyMv49ZcfvvnX+D+5Ff/HllMO/DHNUntk++ZD\nlu47l511MXFRn0YlF229h16POXEl5/LRNU/7WfxJ1mrqD1pWpLWLyGy+WmJX+njJEyzl03Uh\nWPtNS/4sDf73Nctvov1nqX7nV9abLmZfObt76VBviwLqR8bcmGLFW7IWp97vYvlXr5P/kPkb\nvvslPS8v88KJpO1fLnllwLAlyZVdGhqdXNipdvFUA5/+4Nsj5zNyc349mbBidFDx7til90cX\nzGZzZXuctXOp3Hqxfi5W7MGK1kXcmKl9i2c+P5y+ucJgZ233MduyB1Vig6EHlVogWlRm267I\nHdvLbIpgVwYNwa7o8ydv7FAaDF//27Wy51ysws6892/NLEq4b2aSxSTT//2kxd6wVt+oK6WP\nSDV/db/FnyJH3gq0PJBdTrCza/nK7lvvnG+y/a9Ni/9vl0V/bL0l98j1h2/840qDa5vHWP59\nY9fu7ZN/TOrn+R0tRgyI1hC5rF1E5ioFOz0XgtXftOTP0pCSB6W1/iw59lhW/Af7lWW9LU4g\ndFpwY+9jk5+lA1NaWsy48bB1F8tpXIn1roVlD7LvOP+U5bhria9YzLPLorPmyvY4a+diNpsr\nsV6sn4tVe7AzCzqVOJtk7+BgJyLePV9864NVmw+cy6lw/1fsFt3HbMMeVJkNhh5UaoFoUKlt\n24pp3mG9zJa4O71y7P39b1x18uuqgZ6eQV36DI6cOP2dpdHfHPwl65rVE/wlPv7n4k9tR4wM\nstiE6z359IDiKykKt2/fLZf27k0ubnD/8JHtLFbl/c88bbG7K8+Dz7/UqcxDxbknt/zr9dH9\nH2rT4t673JzrOFx/mmbXf5wrbnLu3Lmy/qfLYxF9/vgrxc7Ly8NiVIchQ25s/N6+vsUVm69c\nyaiwVmsXUdXptRBq/puKiDg+NmZE8V/v7s2bW/zMXrlyxUZzERGRlNjYn4o/+Y+d+mSjWzeu\npqWREh9f3IOu7f2br+VTzO3bLbCY5769e6WSPc7auZRBw3qpxFys2oN5jZk1ztfiGMa1oiKz\niJz5duHrLwwPD/Fo2LznC8sOppcqvPLdR6rUg/TpPkIPsn7bvsmd3Mts6Q589sPANea1g0oN\nu7S0e6Px8dZM5U/jJvf459it1zurOeecaec5087Nv4908ug8dPK78//auaHm6Z0/f774g72v\nb8mrDBx8fb1Efvz9U9a5c5mpFyzfiOXr61uifdNmzerInoIK5+rdseO9Nw8t/OmjYb3Hrf35\n6s2jLOXk5JQ1uKmXV/FPgLOzxfWo17/F7+wdLC/uu3at4m3c2kUk4iaVp+dCsP6b2uLBfJ4t\nWpQ4t+Nkkfi1rB8rlPg9dwgObn3rptW23lNSUrQVK1KQlnZFKtfjrJ2L1LtpuIb1Upm5WLUH\nq9fr/YRNHn99ccGXh9NvfjSxOefU1iXPPrj95PZ9cztcr65q3Ueq1INqdkdhgR50K7fatm9y\nR/cyG+KIXWX5jPky4dOJvVu5lbEI81J2fzyxx6PvHdX+eHZziUe525X+oTaXftJ76fZ2pUdr\nmmuDBmVcrnxi8ciK98hl1XSdu7vFaZMShbnWr2+xFy4qsnJPZ+0iqhJdF0JVv2nJH5HLly+X\n3/x39evXt/xob199O4cS3+Cmjbe8xtW93st29epVG/W4iuZSBhuvlxtzsW4P5uDRa9q6Q+fO\n7N2wdPZLz3TzufmQQL5pweufpl7/dxW7j1SpB9lgg6EH2dgttu2b3OG9zGYIdpXnFjBi0eZj\nly4eTdi46sN3pr84Zkh4O6/ivyxz930QtV/zxJo2tbjy5NrPP58pMbbo9OmzxZ9c773X7d57\nLQ+1nTlTsv25U6cqPlwnZW++p9dF77mx6ds1fXRezA8p6bnX7/6v4FbT6mXtIqrKvPRdCNZ/\n05IrsuRxkOTkk9VSZRV4eFicWitM+v5weY2rab2X6EHOEavLe9RD3AsNpXI9ztq5VE4l52L9\nHszZs8Pj46a++9Hr4S4i0uOd7xJWT3u48R9jCxMSrp+BMlr3EXrQH6qxB1k9zTu1l9kAwa6q\n6jTwD33sz8/9bdZ7y6I37U/ePMHnxqhTyck3/u6rVcviT9z8/PzSk/Hq1s3i2svvVn76g0WK\nv7L+k6+ybnyq9VDXB6VBp04tihsciv7cZNH+6MefaLykoQwlDjj7/Hn65D6BTes5OYhI0U+b\nv9VxD2ftIqoKfReC9d/U2c3N4ihGcnLxtSD5W1esOmXL4irakrXwCAuzuPT76PK3N1y6deNq\nWu9e3bs3v/Eh95vlK8+U/3dz5XqctXOpnCrNReMerAxOje4LHfzWm8Nv/N7lZWfffNLKAN1H\n6EF/1FaNPUiLO7SX2bJMESHYVdqRd3q1DX9myqIvtuz+/ujP5y/nXL2am37mwPro7cXXCLjW\nr39j+TZsaPF3Quq3X8T8nFFQYp/ZYexzFnekH503fMzHe89mFeRfPv7N7CF/XV98b0HD4c9H\nuIuEjBhpcV3F0QUjxq86cD7nau6F774YP2yOqfIbdqNGFpfhnvny/U8OnMvOy7pw5Jt3hzw6\n64Ce7/62dhFVgc4LwepvavfAAxYPCTi6ZMKbW05ezs08Ff+PoaOXWV61UnUVbsla/OnZCQ/W\nKZ7Kqj93Hf7exqSUjIKC7LRTh3d//cFf+wxe8kcAqKb13n7s+PY3fmEztkzo+eRba/Ymn8/I\nL8j+9dzJI9/t+uaLf7w+7sku/gP+L1Wksj3O2rlUTiXmYsUerGjdqJahA59/Y+n6uINHfr6Q\nnv/7+jZfzTr3/arlMTcK92ne3F7EeN1H6EG/l1atPUiLO7OXVb6mWynnaKIyquMBxSVuWS9b\n/RFfWTxt8uDU5mU3u1FPbuKbneqW3eaGEg8o3jS6qg8oDin5bPLrkma01nQt/j1/2fH7/yj9\nLqDiaf08P6R4TL0S7xxaM8RiWt0+rPANX5VZRJV/3IneC8Hqb3p4VpCmgst/IZKF3S8X/00p\n/tNMN0ZUvCVrUXD4Hz3K3Z+VeLyqtUtDm7zv3g695XP7i4W+//t0K9fjrJ2LuVLrxeq5WLEH\nu/rF4yVHlDjmdIND4Jum3x9OYX33Mdu0B1Vmg6EHVX8P0uJO7WU2xRG7alI36IVVC/pbbANt\nn5vSu/xg7hTy+n//93Z5b31pM+7zbR8N+uOmbPfwf3z9bs8yX73yt6kDLZ5JXtG1taUETV4x\nrV0ZW29tv6dmPnOfNVOyOWsXUeXpvRCs/qYPvLz01eCbd932TfsufDXcpqVVvCVrUfuBCRvi\n/m/k/RU9Dl9Eqm29O7Z59T9bFw265QuCSqtcj7N2LpVj87nctAezUFhYWHqQc8sRH61/LeD3\nHxTDdR+hB9VED9KCXmYLNo+Kt6HqOGJ3LTfVtG31+9OfG9ijU7C/b5O76tZyqO3s3sg34MF+\no6ev2HG24Oa55BxdM+OpsACvu+rWtr91Pdk/b/1wyjN9O7RqererY61azu6NvAO6Dhw/O/pg\nWhnvaS68sP2DCY+F+Nxdt7ajW6Pm7R/7y6JvzxZkLg+32OZ7Lk2v6OvcVOqxdX8f0e2+Jm51\najm6NvAO6vn0zOgjmSX//tDhiJ3Vi6hKDyi+DRaCdRvDlR8+fTWic4vGrnVqO9dv2jps+PTP\nvk+/Zs0rzC3c8niDWeuWrMXVtIOr500Y3LNt8yZ3uTrWdqrftEVQ6OPj53yecPrm92lbtzQ0\nyz+3Z+Wb4weGBfveU9+lTq06Lnff6+sf1K7rYyNfmrt8w87jv5boy1b3uErMpbLrxZq5WLMH\nK0xP3vP1irdfHNEvrFNwK+/GbnXsRETsHes1aR7c9Ylxf4/aceamS8mt6z7m6uhBVm8w9KBK\nsa4HaXEn9jJbsjPXwK3OqFnmX6LCA0b/748rJ+6dEH/uH111rQhQ2R3X4zaPce2zIvvRj3L/\n80wFr8IEbOSO62VVwKlYIzs4++GHhr36/rr4g0fPpGVfLSrITP1p9/r5fw6f8L/i62G9hg0N\n1bFGQB30uOusvLoDsAK9rMruwDdPKORa5smE6NiE6Hdu2cKh+dil0x90uOV4ANrR464j2KH6\n0MuqjCN26qp1b/fJX8Yv7VsN91IDuMmd0+MIdtDLndPLqoBr7IzMnH16z+YNG77atD3pZMq5\n86mXC2q5uN/VpNkDwe27Pjp4+KBuzTTdMQXYyNpBdhHrKmrUeqbp0N8DDFkDPQ7VS/ceVE0F\nWDNZelmVcSrWyOxcfDoPnNh54ES9CwHuCPQ4oLrRy6qMI3YAAACK4Bo7AAAARRDsAAAAFEGw\nAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBnzzRM7puLWr1sZsP2A6djot\nPTOnsHZdt3qNffwDQrr2HTQ8oruPs94VAgAA6MFgb55Ii501bMTsrSkFt2pQx6PHtJXRM8Ia\n1mRVAAAAtwMjBbtC05yO7acdzBfXFuEjIwf16hjo59HA3alWYV7GrynJpr1b1kZ9svlEtjiG\nzEncMyXAgAcjAQAAqsBAwS5ndUTjoWuzmwxYsTM6srljWU3yk6OGdhm94YLLoOiLa4bUrekK\nAQAA9GSgmyf2x8Vli7SZtOAWqU5EHP0i508MFsmOj0+s0doAAAD0Z6Bgl5GRISKenp7ltro+\n/npbAACAO4mBgp2Xl5eI7N+xI6+cRnkJCYki4u3tXUNVAQAA3C4MFOyCIwa3spPUZZGDF+9K\nLSyjQWHqrsWDI5elip1/xMCgGq8PAABAXwa6eUIka+drXXrPS8oWcWwcGNqtQ4CfR0M3R4ei\n/MxLKcmH9sXvNF3MF3Fp89o3CXNDXfSuFgAAoGYZKtiJSMYPH02bNGN57Nkyz8c6eYaNmbVo\n9qgg95quCwAAQHdGC3YiInIt69Te2O2JpuNnLqZn5RY5OLvWb+zdKrBdt7COPq4GOrkMAABg\nS4YMdgAAALgZx7cAAAAUYcD3buWcjlu7am3M9gOmY6fT0jNzCmvXdavX2Mc/IKRr30HDI7r7\nOOtdIQAAgB4Mdio2LXbWsBGzt6YU3KpBHY8e01ZGzwhrWJNVAQAA3A6MFOwKTXM6tp92MF9c\nW4SPjBzUq2Ogn0cDd6dahXkZv6Ykm/ZuWRv1yeYT2eIYMidxz5QAAx6MBAAAqAIDBbuc1RGN\nh67NbjJgxc7oW7wuNj85amiX0RsuuAyKvrhmSN2arhAAAEBPBrp5Yn9cXLZIm0kLbpHqRMTR\nL3L+xGCR7Pj4xBqtDQAAQH8GCnYZGRki4unpWW6r6+OvtwUAALiTGCjYeXl5icj+HTvKfOnE\n7/ISEhJFxNvbu4aqAgAAuF0YKNgFRwxuZSepyyIHL96VWlhGg8LUXYsHRy5LFTv/iIFBNV4f\nAACAvgx084RI1s7XuvSel5Qt4tg4MLRbhwA/j4Zujg5F+ZmXUpIP7YvfabqYL+LS5rVvEuaG\nuuhdLQAAQM0yVLATkYwfPpo2acby2LNlno918gwbM2vR7FFB7jVdFwAAgO6MFuxERORa1qm9\nsdsTTcfPXEzPyi1ycHat39i7VWC7bmEdfVwNdHIZAADAlgwZ7AAAAHAzjm8BAAAowoDv3co5\nHbd21dqY7QdMx06npWfmFNau61avsY9/QEjXvoOGR3T3cda7QgAAAD0Y7FRsWuysYSNmb00p\nuFWDOh49pq2MnhHWsCarAgAAuB0YKdgVmuZ0bD/tYL64tggfGTmoV8dAP48G7k61CvMyfk1J\nNu3dsjbqk80nssUxZE7inikBBjwYCQAAUAUGCnY5qyMaD12b3WTAip3Rt3hdbH5y1NAuozdc\ncBkUfXHNkLo1XSEAAICeDBTs4sc36r70Upu3T3z3qt+tW514u03LKUmNxsdf/GdXDRMtKiqK\niYnJyyvvNWXXrl07duzYa6+9VqdOHWuLBgAAqDEGOl+ZkZEhIp6enuW28vT0FEm63laD2NjY\n/v37a2lpb2//+uuva5sqAACADgwU7Ly8vESS9+/YkTe8p9OtGuUlJCSKiLe3t7aJhoWFff31\n1+Ufsdu9e/fChQtbtmxpXb0AAAA1y0DBLjhicKt35h5fFjn4/uhlf3nwnptKL0zdtWRs5LJU\nsfOPGBikbaIODg79+vWrsNnChQvt7XnmHwAAuK0ZKNjZhUyNmhzTe17SxkmhPnMCQ7t1CPDz\naOjm6FCUn3kpJfnQvvidpov5Ii5tXlsxNUTvagEAAGqagYKdiGvo29t3+U+bNGN57FnTtjWm\nbaUbOHmGjZm1aPaoIBc9ygMAANCVoYKdiLgHjXp/29NzT+2N3Z5oOn7mYnpWbpGDs2v9xt6t\nAtt1C+vo48oJUwAAcIcyWrATERF7V9/O/Xw7V3xpHAAAwB2E41sAAACKMOQRuxLyzu3b9N8d\nh89miJvnA1369unk4ax3SQAAAHowULA7tXX5tz9Ls55jevj+MShj37vDBk2L+SX/RqM6Xn3m\nrFv9cns3PSoEAADQk4FOxSZ+OHbs2LEfJt4YcO6zP/d5JeaXfIdGbZ8Y/cILo59o29ih4JdN\nr/QZufqijnUCAADow0BH7Eox71rw+n9/E7uWz27evbRnAzsRMf/67bhOjyw7sWHm4u+GzG6r\nd4UAAAA1ykBH7Eo5smnTaRGXx2fOvZ7qRMSuQc+5M/u7iBzbtOmkvtUBAADUOOMGu1OnTolI\nQLdud1sObdCtW2sROXHihC5FAQAA6Me4wa5WrVoi0qBBg5KDGzVqJCJXr17VoyYAAAAdGe4a\nu8xzR48eFRFxubeVyMHTp0+L3Gcx/ty5cyLi5eWlT3kAAAC6MVyw+9/E+++3+HgsLu7C6/c1\nufH56o8/JovUDQ5uUfOl1ZSioqKYmJi8vLwqTsfJyalv374ODg42qQoAAOjOQMHunqAePdJv\nGmqXFHtOhjX9/VPOVyvXZ4jb8OH96tZscTVp06ZN/fv3t8mkNm7c+Nhjj9lkUgAAQHcGCnYP\nzfj224raZHg8vvij8KahfRXOdZKbmysiI0eODA4OrvREkpKSPv300+uTAgAAajBQsNOiSedh\nz3TWu4iaERwc/Mgjj+hdBQAAuI0Y965YAAAAlKBesEve/MEHH3ywOVnvOgAAAGqYesHuu+UT\nJkyYsPw7vesAAACoYeoFOwAAgDuUgW6eKMrLyi2suFleUfWXAgAAcBsyULD78im3iHV6FwEA\nAHDb4lQsAACAIgwU7Jo18xWRtnOPXy3X6if1LRMAAEAnBgp2f+rVq6HI91u3pdcqj72d3oUC\nAADowkDBzq7rIz2cxJzwvy28BgsAAOBmBrp5Qpx6jpz0eN6P7lmnRO6/Zat245ctC5dm7Wqu\nLgAAgNuCkYKd1O87d0Pfihr59hgzpiaKAQAAuM0Y6FQsAAAAykOwAwAAUATBDgAAQBEEOwAA\nAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDs\nAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAU\nQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMA\nAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATB\nDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABA\nEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsA\nAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ\n7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAA\nFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbAD\nAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAE\nwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAA\nQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7\nAAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAA\nABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGw\nAwAAUEQtvQuwXs7puLWr1sZsP2A6djotPTOnsHZdt3qNffwDQrr2HTQ8oruPs94VAgAA6MFg\nwS4tdtawEbO3phSUGJqddTnt/KnDe75ZveSNqT2mrYyeEdZQpwIBAAB0Y063GBQAACAASURB\nVKRgV2iaE95n5sF8cW0RPjJyUK+OgX4eDdydahXmZfyakmzau2Vt1CebT2yd2Se8duKeKQFG\n+moAAABVZ6D0k7Nu1pyD+dJkwIqd0ZHNHUuM8/MP7PDwgNEvvRQ1tMvoDQdmv7Fu4pohdXUq\nFAAAQBcGunlif1xctkibSQtKp7pijn6R8ycGi2THxyfWaG0AAAD6M1Cwy8jIEBFPT89yW10f\nf70tAADAncRAwc7Ly0tE9u/YkVdOo7yEhEQR8fb2rqGqAAAAbhcGCnbBEYNb2UnqssjBi3el\nFpbRoDB11+LBkctSxc4/YmBQjdcHAACgLwPdPGEXMjVqckzveUkbJ4X6zAkM7dYhwM+joZuj\nQ1F+5qWU5EP74neaLuaLuLR5bcXUEL2rBQAAqGkGCnYirqFvb9/lP23SjOWxZ03b1pi2lW7g\n5Bk2Ztai2aOCXPQoDwAAQFeGCnYi4h406v1tT889tTd2e6Lp+JmL6Vm5RQ7OrvUbe7cKbNct\nrKOPq4FOLgMAANiS0YKdiIjYu/p27ufbuZ/edQAAANxOOL4FAACgCAMescs5Hbd21dqY7QdM\nx06npWfmFNau61avsY9/QEjXvoOGR3T3cda7QgAAAD0YLNilxc4aNmL21pSCEkOzsy6nnT91\neM83q5e8MbXHtJXRM8Ia6lQgAACAbowU7ApNc8L7zDyYL64twkdGDurVMdDPo4G7U63CvIxf\nU5JNe7esjfpk84mtM/uE107cMyXASF8NAACg6gyUfnLWzZpzMF+aDFixM7r062L9/AM7PDxg\n9EsvRQ3tMnrDgdlvrJu4ZkhdnQoFAADQhYFuntgfF5ct0mbSgtKprpijX+T8icEi2fHxiTVa\nGwAAgP4MFOwyMjJExNPTs9xW18dfbwsAAHAnMVCw8/LyEpH9O3bkldMoLyEhUUS8vb1rqCoA\nAIDbhYGCXXDE4FZ2kroscvDiXamFZTQoTN21eHDkslSx848YGFTj9QEAAOjLQDdP2IVMjZoc\n03te0sZJoT5zAkO7dQjw82jo5uhQlJ95KSX50L74naaL+SIubV5bMTVE72oBAABqmoGCnYhr\n6Nvbd/lPmzRjeexZ07Y1pm2lGzh5ho2ZtWj2qCAXPcoDAADQlaGCnYi4B416f9vTc0/tjd2e\naDp+5mJ6Vm6Rg7Nr/cberQLbdQvr6ONqoJPLAAAAtmS0YCciIvauvp37+Xbup3cdAAAAtxOO\nbwEAACjCgEfsck7HrV21Nmb7AdOx02npmTmFteu61Wvs4x8Q0rXvoOER3X2c9a4QAABADwYL\ndmmxs4aNmL01paDE0Oysy2nnTx3e883qJW9M7TFtZfSMsIY6FQgAAKAbIwW7QtOc8D4zD+aL\na4vwkZGDenUM9PNo4O5UqzAv49eUZNPeLWujPtl8YuvMPuG1E/dMCTDSVwMAAKg6A6WfnHWz\n5hzMlyYDVuyMLv26WD//wA4PDxj90ktRQ7uM3nBg9hvrJq4ZUlenQgEAAHRhoGC3Py4uW6TN\npAWlU10xR7/I+RP/sWFKUnx8ogzpqmGiRUVFMTExeXnlvaZs9+7dInLt2jXrawYAAKg5Bgp2\nGRkZIuLp6VluK09PT5Gk6201iI2N7d+/v5aWhw8f1jZJAAAAfRgo2Hl5eYkk79+xI294T6db\nNcpLSEgUEW9vb20TDQsL+/rrrys8Yrdw4cLWrVtbVy8AAEDNMlCwC44Y3OqduceXRQ6+P3rZ\nXx6856bSC1N3LRkbuSxV7PwjBgZpm6iDg0O/fhU/6XjhwoX29jzzDwAA3NYMFOzsQqZGTY7p\nPS9p46RQnzmBod06BPh5NHRzdCjKz7yUknxoX/xO08V8EZc2r62YGqJ3tQAAADXNQMFOxDX0\n7e27/KdNmrE89qxp2xrTttINnDzDxsxaNHtUkIse5QEAAOjKUMFORNyDRr2/7em5p/bGbk80\nHT9zMT0rt8jB2bV+Y+9Wge26hXX0ceWEKQAAuEMZLdiJiIi9q2/nfr6dK740DgAA4A7C8S0A\nAABFGPCIXc7puLWr1sZsP2A6djotPTOnsHZdt3qNffwDQrr2HTQ8oruPs94VAgAA6MFgwS4t\ndtawEbO3phSUGJqddTnt/KnDe75ZveSNqT2mrYyeEdZQpwIBAAB0Y6RgV2iaE95n5sF8cW0R\nPjJyUK+OgX4eDdydahXmZfyakmzau2Vt1CebT2yd2Se8duKeKQFG+moAAABVZ6D0k7Nu1pyD\n+dJkwIqd0aVfF+vnH9jh4QGjX3opamiX0RsOzH5j3cQ1Q+rqVCgAAIAuDHTzxP64uGyRNpMW\nlE51xRz9IudPDBbJjo9PrNHaAAAA9GegYJeRkSEinp6e5ba6Pv56WwAAgDuJgYKdl5eXiOzf\nsSOvnEZ5CQmJIuLt7V1DVQEAANwuDBTsgiMGt7KT1GWRgxfvSi0so0Fh6q7FgyOXpYqdf8TA\noBqvDwAAQF8GunnCLmRq1OSY3vOSNk4K9ZkTGNqtQ4CfR0M3R4ei/MxLKcmH9sXvNF3MF3Fp\n89qKqSF6VwsAAFDTDBTsRFxD396+y3/apBnLY8+atq0xbSvdwMkzbMysRbNHBbnoUR4AAICu\nDBXsRMQ9aNT7256ee2pv7PZE0/EzF9OzcoscnF3rN/ZuFdiuW1hHH1cDnVwGAACwJaMFOxER\nsXf17dzPt3M/vesAAAC4nXB8CwAAQBHqBbvkzR988MEHm5P1rgMAAKCGqRfsvls+YcKECcu/\n07sOAACAGqZesAMAALhDGejmiaK8rNyynktcSl5R9ZcCAABwGzJQsPvyKbeIdXoXAQAAcNvi\nVCwAAIAiDBTsmjXzFZG2c49fLdfqJ/UtEwAAQCcGCnZ/6tWrocj3W7el1yqPvZ3ehQIAAOjC\nQMHOrusjPZzEnPC/Lbl6lwIAAHAbMtDNE+LUc+Skx/N+dM86JXL/LVu1G79sWbg0a1dzdQEA\nANwWjBTspH7fuRv6VtTIt8eYMTVRDAAAwG3GQKdiAQAAUB6CHQAAgCIIdgAAAIog2AEAACiC\nYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAA\noAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCI3B\nLmHxhPe/OZFVvbUAAACgCjQGuws7PvhruL9n68dfWRZ3Oq96SwIAAEBlaAx2XV/858v9H3A4\n9vW7z4b5ebUdPP3T3RcKqrcyAAAAWEVjsGscOn7BV6azJ+Ojpgxua394zVtPP+jj0+npt1Z/\nf6mwegsEAACANlbdPOHs3XXUnNX7fzmz97NZT7dzSPp0+tC2Xs3Dnp3/9ZH0a9VVIQAAADSp\nzF2xdZp0GD79452nz36/bu4g77S4ZZMfb+3p/8hz8788dIV8BwAAoJPKPu7katqBdf/3j8Uf\n/vd4nohDw5YeGXH/mvxkULP2z284Q7gDAADQgdXBLv/cnpV/H9HJx6vdkKlROzNbPjF5edzP\nvxw/duanTVMeuuvywQ9H/GXlpeqoFAAAAOXSHuxyT8dHTRkc4unTeeQbn+3Pad7/pQ+3njib\ntH7e6G5eTiKOPuFzVr3eQSQrPv5gNRYMAACAstXS1mzX9NaPzTly+ZrYufv3mThh4l+f6d3c\n5aZWXl5eIvsyMzJsXCQAAAAqpjHYnfvxx4Lmjzw/YeLEUX1audndqlmnF7/4YpB4dbJVdQAA\nANBMY7ALnX04peX99So6cesZOnRolUsCAABAZWi8xu5e/1a5h+Li4hKOXS4x/PKxhLi4uB8u\nFFVDaQAAALCG1psnLq9+PjQs7Ml/Hq5dYnDtQ0ueDAsLnbA23falAQAAwBoag93FVf/3VZY0\nHj5ugGuJ4a5PPDe8sWSt/78veMQJAACAvjQGux8PH74m0jowsHR7+4CAB0SuHTlyzOalAQAA\nwBoag11mZma54y9fvlzueAAAAFQ3jcHO09NTRH48dKj068KKDh8+KiJNmza1cWEAAACwjsZg\nF9izZ2ORC1FvLUsxWww2pyx7K+qCSMOePYOrpTwAAABopTHYOfR8eUonZ8neMqHr49M/3bL/\n8I+H92/5dHr/rn/dki3OHae+0suheusEAABABTQ+oFik5cQ1XxztPexfRza+9fTGt4qH1239\n3OdrJ7W65csoAAAAUDM0Bzux83x86YGjgz/656f/3XX4l8sFde72Dnjw0ZHjR4V5O1ZjgQAA\nANBGe7ATEXHyfnj82w+Pr6ZaAAAAUAVa3zwBAACA25xVR+xyz+ze/L89R365lJVfZC45KvCp\nt4cH2LIwAAAAWEdzsMvcO3fAE9O3nS8qe/TAdgQ7AAAAXWkMdlfjp0ZM3XZenJr1fXZsnyDP\n+o6lboP16WD72gAAAGAFjcEucf36X0Tcn1ixZ/2fG1dvRQAAAKgUjTdPpKWliUiHRx8l1QEA\nANymrHpXbEFBQbUWAwAAgMrTGOzaDh1+v718v23b5eotBwAAAJWlMdjZ/Wnaqrlda6975ekl\niWQ7AACA25HGmye2vdpl8tYCZ+ezG19of88075bNGzmXjIQ93kmc93A11AcAAACNNAa735IP\nHDjw+7+vXjlz5LszpRr4/mbLqgAAAGA1jcGu3/Lz5z8or4HTXbaoBgAAAJWmMdg51m/SpHoL\nAQAAQNVovHkCAAAAtzurgl328f8sevGpx8Ie7NC+3avbREQu7P33Z599tjX5avVUBwAAAM00\nnooVkQtfPRs2ZNnR/N8/ev8mIuJy4otxIzbUGlkv9ZN+TtVRHwAAADTSesQubeW4p5YdLfQd\nsjThvT7Fg92eeKp/Xcn48ouY/Fv/XwAAANQAjcHu/Kp/fZ0l7kPeWzEu1NvFYkTdgIDmIpm7\ndpmqpTwAAABopTHY/ZCUJCJ/6trVpfQYLy8vETl//rxt6wIAAICVNAa7goICEXF1dRUROzs7\nizFZWVkiUrt2bdvXBgAAACtoDHb33nuviJw9e7b0iKIjR46LiK+vr03LAgAAgLU0Brs2PXs2\nEEn6cv1Js+URO/PFle+tTBXx6d3bv7oqBAAAgCYag12t3n+b2sHZvHfmE2P/b0+aiEjGT1s/\nf3Pgg89uzBD3ntMndbaraBIAAACoVlqfY2fn/+K66OS+f/7ninE/iIjIlqk9t4iIuARP+vdn\no72qqz4AAABopP0BxXae/ZckHhv66dJP/pNgOn0pz6G+R+vQfqNeGN3ds041FggAAABttAc7\nEZE6Hg+NefOhMdVUCwAAAKrAqnfFAgAA4Pal8YjdyS1L/5dcXgO/R57r1dwWBQEAAKByNAa7\ng/8aP35deQ0GriHYAQAA6EpjsAsY9uabbUoMuZablrw3Zu3WE3Zthk8aeF/bgGooDgAAANpp\nDHb3DXz99TIGv/3a/L7tJ8fsfX7WrPtsWhYAAACsVbWbJ5zvf3n2s56Xv3156vocGxUEAACA\nyqnqXbH2QUGtRTK//XafTcoBAABAZVX5cSdXrlwRkaysLBsUAwAAgMqrWrC79tvWef/aJyKt\nWrWyTT0AAACoJI03T+x5d9CC3SUHmQuupBzev+/kFbO493kxkmAHAACgL43B7uzudevKfI5d\nnSYdhk9fsmiUpy2LAgAAgPU0BruH5+7e/UrJQXYOdVwaeLVo1sjJzvZlAQAAwFoag93dLTt1\nalm9lQAAAKBKqnxXLAAAAG4PGo/YJW/+YNMJrdNs0eeFcL/KFgQAAIDK0Rjsvls+YUKZN0+U\nZeAagh0AAECN0xjsAp+aO9c/adWi6EN5DYIfe7xrgKdb5llT/Ib//vCbU8DQScOD3Swa3xdY\nLaUCAACgPBqDnX+f8KLZsw4VtZ+y85vZne76/T7Ya/N3TunV452vtpinJEwJqlN9VQIAAKBC\nGm+eSP982puJuR7jF755I9WJiP3dobMXjWuau//N16OvVFOBAAAA0EZjsDuwa1e+SEBwsEOp\nEbXatg0Uyd2586DNSwMAAIA1NAa7vLw8EUlNTb1pzIULF0QkNzfXpmUBAADAWhqDXUBgoIj8\n8NGHO/NKDM/ZseQjk4gEBgbYvDQAAABYQ2Ow83nmpYH15dqx9/qFjlqwLi7x6E9HE+PWzn+m\nS//Fx6/JXQNffMa7eusEAABABTTeFSuNh0Zt/OnXJ/8ed/Djvw362GKEfaPus76MGtqoGmoD\nAACAFbQGOxH3LtO3/th//Yqo9dsOHDuXfrV2/ab+IQ8/GTnmyeC7eTEZAACA7rQHOxGxbxA8\naPLiQZOrqxgAAABUHsfaAAAAFGFVsMs+/p9FLz71WNiDHdq3e3WbiMiFvf/+7LPPtiZfrZ7q\nAAAAoJn2U7EXvno2bMiyo/m/f/T+TUTE5cQX40ZsqDWyXuon/Zyqoz4AAABopPWIXdrKcU8t\nO1roO2Rpwnt9ige7PfFU/7qS8eUXMfm3/r8AAACoARqD3flV//o6S9yHvLdiXKi3i8WIugEB\nzUUyd+0yVUt5AAAA0EpjsPshKUlE/tS1q0vpMV5eXiJy/vx529YFAAAAK2kMdgUFBSLi6uoq\nInZ2dhZjsrKyRKR27dq2rw0AAABW0Bjs7r33XhE5e/Zs6RFFR44cFxFfX1+blgUAAABraQx2\nbXr2bCCS9OX6k2bLI3bmiyvfW5kq4tO7t391VQgAAABNNAa7Wr3/NrWDs3nvzCfG/t+eNBGR\njJ+2fv7mwAef3Zgh7j2nT+psV9EkAAAAUK20PsfOzv/FddHJff/8zxXjfhARkS1Te24REXEJ\nnvTvz0Z7VVd9AAAA0Ej7A4rtPPsvSTw29NOln/wnwXT6Up5DfY/Wof1GvTC6u2edaiwQAAAA\n2mgMdolLnnl/77V2Ez6dMObNh8ZUb0kAAACoDI3X2J3b+fmnK1fGn67eYgAAAFB5GoOdp6en\niFy4cKFaiwEAAEDlaQx2bZ98splI0pYtF6u3HAAAAFSWxmBn1+nvn03vUOvrl4fO+uZMQfWW\nBAAAgMrQePPEtle7T96a71o3LXZmuO/cu72a+zRyLpEJe7yTOO/h6igQAAAA2mgMdr8lHzhw\n4Pd/m/N+O3PktzMlG/j+ZtOyAAAAYC2Nwa7f8vPnPyivgdNdtqgGAAAAlaYx2DnWb9KkegsB\nAABA1Wi8eQIAAAC3u3KC3eYJLVq0aDFhc/GQyycTExMTT16u/rIAAABgrXKCXdb55OTk5PNZ\nxUO2Tm7fvn37yVurvywAAABYi1OxAAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIqfI5dTtqpU6d+\n/3daTukhN7g09m1U16aVAQAAwCoVBrtNzzdrVtEQERm4xrx2kM2qAgAAgNXKCXZ1XOrVq6d1\nOi51bFENAAAAKq2cYNf/k/T0misEAAAAVcPNEwAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYA\nAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCJq6V2A9XJO\nx61dtTZm+wHTsdNp6Zk5hbXrutVr7OMfENK176DhEd19nPWuEAAAQA8GC3ZpsbOGjZi9NaWg\nxNDsrMtp508d3vPN6iVvTO0xbWX0jLCGOhUIAACgGyMFu0LTnPA+Mw/mi2uL8JGRg3p1DPTz\naODuVKswL+PXlGTT3i1roz7ZfGLrzD7htRP3TAkw0lcDAACoOgOln5x1s+YczJcmA1bsjI5s\n7lhinJ9/YIeHB4x+6aWooV1Gbzgw+411E9cMqatToQAAALow0M0T++PiskXaTFpQOtUVc/SL\nnD8xWCQ7Pj6xRmsDAADQn4GCXUZGhoh4enqW2+r6+OttAQAA7iQGCnZeXl4isn/HjrxyGuUl\nJCSKiLe3dw1VBQAAcLswULALjhjcyk5Sl0UOXrwrtbCMBoWpuxYPjlyWKnb+EQODarw+AAAA\nfRno5gm7kKlRk2N6z0vaOCnUZ05gaLcOAX4eDd0cHYryMy+lJB/aF7/TdDFfxKXNayumhuhd\nLQAAQE0zULATcQ19e/su/2mTZiyPPWvatsa0rXQDJ8+wMbMWzR4V5KJHeQAAALoyVLATEfeg\nUe9ve3ruqb2x2xNNx89cTM/KLXJwdq3f2LtVYLtuYR19XA10chkAAMCWjBbsRETE3tW3cz/f\nzv30rgMAAOB2wvEtAAAARRjwiF3O6bi1q9bGbD9gOnY6LT0zp7B2Xbd6jX38A0K69h00PKK7\nj7PeFQIAAOjBYMEuLXbWsBGzt6YUlBianXU57fypw3u+Wb3kjak9pq2MnhHWUKcCAQAAdGOk\nYFdomhPeZ+bBfHFtET4yclCvjoF+Hg3cnWoV5mX8mpJs2rtlbdQnm09sndknvHbinikBRvpq\nAAAAVWeg9JOzbtacg/nSZMCKndGlXxfr5x/Y4eEBo196KWpol9EbDsx+Y93ENUPq6lQoAACA\nLgx088T+uLhskTaTFpROdcUc/SLnTwwWyY6PT6zR2gAAAPRnoGCXkZEhIp6enuW2uj7+elsA\nAIA7iYGCnZeXl4js37Ejr5xGeQkJiSLi7e1dQ1UBAADcLgwU7IIjBreyk9RlkYMX70otLKNB\nYequxYMjl6WKnX/EwKAarw8AAEBfBrp5wi5katTkmN7zkjZOCvWZExjarUOAn0dDN0eHovzM\nSynJh/bF7zRdzBdxafPaiqkheldbtqKiopiYmLy88g46Vmj37t22qgcAAKjEQMFOxDX07e27\n/KdNmrE89qxp2xrTttINnDzDxsxaNHtUkIse5WmwadOm/v37610FAABQk6GCnYi4B416f9vT\nc0/tjd2eaDp+5mJ6Vm6Rg7Nr/cberQLbdQvr6ON6W59czs3NFZGRI0cGBwdXeiJff/11fHy8\n7YoCAACKMFqwExERe1ffzv18O/ezwaS0nBu9furz2rVrNpifiIgEBwc/8sgjlf7vSUlJtqoE\nAACoxJDBzoZiY2M1nhs9fPhwdRcDAABQFQYMdjmn49auWhuz/YDp2Om09Mycwtp13eo19vEP\nCOnad9DwiO4+zlZMLCws7Ouvv67wiN3ChQtbt25d1coBAACqk8GCXVrsrGEjZm9NKSgxNDvr\nctr5U4f3fLN6yRtTe0xbGT0jrKHGCTo4OPTrV/E53YULF9rb39ZX7wEAABgp2BWa5oT3mXkw\nX1xbhI+MHNSrY6CfRwN3p1qFeRm/piSb9m5ZG/XJ5hNbZ/YJr524Z0qAkb4aAABA1Rko/eSs\nmzXnYL40GbBiZ3Tp18X6+Qd2eHjA6JdeihraZfSGA7PfWDdxzZC6OhUKAACgCwOdXtwfF5ct\n0mbSgtKprpijX+T8icEi2fHxiTVaGwAAgP4MFOwyMjJExNPTs9xW18dfbwsAAHAnMdCpWC8v\nL5Hk/Tt25A3v6XSrRnkJCYki4u3tXYOVGdH1x/JV/e1kTk5Offv2dXBwsEVRAACgSgwU7IIj\nBrd6Z+7xZZGD749e9pcH77mp9MLUXUvGRi5LFTv/iIFBepRoIEePHhWRhQsXLly4sIqT2rhx\n42OPPWaLogAAQJUYKNjZhUyNmhzTe17SxkmhPnMCQ7t1CPDzaOjm6FCUn3kpJfnQvvidpov5\nIi5tXlsxNUTvam93V69elSq/3CwpKenTTz+9/p40AACgOwMFOxHX0Le37/KfNmnG8tizpm1r\nTNtKN3DyDBsza9HsUUEuepRnQFV8uRkAALitGCrYiYh70Kj3tz0999Te2O2JpuNnLqZn5RY5\nOLvWb+zdKrBdt7COPq4Guh0EAADAlowW7ERExN7Vt3M/384VvzACAADgDsLxLQAAAEUY8ohd\nCXnn9m36747DZzPEzfOBLn37dPJw1rskAAAAPRgo2J3auvzbn6VZzzE9fP8YlLHv3WGDpsX8\nkn+jUR2vPnPWrX65vZseFQIAAOjJQKdiEz8cO3bs2A+L3xV27rM/93kl5pd8h0Ztnxj9wguj\nn2jb2KHgl02v9Bm5+qKOdQIAAOjDQEfsSjHvWvD6f38Tu5bPbt69tGcDOxEx//rtuE6PLDux\nYebi74bMbqt3hQAAADXKQEfsSjmyadNpEZfHZ869nupExK5Bz7kz+7uIHNu06aS+1QEAANQ4\n4wa7U6dOiUhAt253Ww5t0K1baxE5ceKELkUBAADox7jBrlatWiLSoEGDkoMbNWokf7wwCwAA\n4E5iuGvsMs9df329uNzbSuTg6dOnRe6zGH/u3DkR8fLy0qc8AAAA3Rgu2P1v4v33W3w8Fhd3\n4fX7mtz4fPXHH5NF6gYHt6j50gAAAHRloGB3T1CPHuk3DbVLij0nw5r+/innq5XrM8Rt+PB+\ndWu2OAAAAN0ZKNg9NOPbbytqk+Hx+OKPwpuG9iXXAQCAO46Bgp0WTToPe6az3kUAAADowrh3\nxQIAAKAEgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAA\noAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIId\nAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAi\nCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAA\nAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDY\nAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAo\ngmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcA\nAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAABm4rOwAAHLxJREFUiiDY\nAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAo\ngmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcA\nAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiC\nHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACA\nIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYA\nAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog\n2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAA\nKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYA8P/t3Xd8FNX6x/Fn\n0hMICYTQQjV0iEGiIgTBCNKLhaYoXAUJ2LHQLwhcBEVv7MBFpBcFlB8Iooh0EQg1QYHQISQS\nQgippOz8/ggllcwmu0n28Hn/4eu1Z86cPfNwzHx3dncWABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAA\nABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARTiU9gRgw0wm\nk4js3r27mOO4uLh069bN3t7eEpMCAODeRbBD0R07dkxEQkJCQkJCijnUunXrevToYYlJAQBw\n7yLYoejS09NFZNCgQf7+/kUe5PDhw4sWLUpJSbHcvAAAuEcR7FBc/v7+nTp1Ku1ZAAAAvjwB\nAACgCoIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCK43QlKGT9fAQCApdhgsEs+t3XV\n0lUbtu8PO34u5lpCcoajm7tHlTqNmge069ZnYN/H6riW9gxhDn6+AgAAS7GxYBezZcqzL0zb\nHJmWozUpMS4m6uzRP3/57qvJ4zqMX7xiYlDlUpogzMbPVwAAYCm2FOwywj7o0nXSgRtSvn6X\nQS/1eaKVn6+PVwUXh4zU67GRp8L2bFr17cKNJzdP6trFMfTPsc1t6dBQzJ+v4P1cAADEpoJd\n8uopHxy4IdWenLdrxUv3OefY5tvI7+HHnxzy9tvfDmg7ZM3+aZNXv7myv1spTRQlz4Lv544e\nPTogIKBo+5pMpvDw8ObNm9vZFetrSeRLAEDR2FCw27d1a5JIi7c+zp3q7nD2fWnmm5+vGXt4\n27ZQ6d/OwKCZmZkbNmxITU29S5+s60BZ14Qs4vDhw8XZ/dy5cwySy8mTJ0XE39+/atWqRR4k\nIiLizJkzH374YXFmYinFyZdioYjJIGVzGgxipUHKyDQYxHqD3CuvmXWbsfY5JxHpsST1rr1S\nFncXEeeB64wNumnTJoOFmjp1avGPYd26dVb8twQAAAXbtGlT8U/lZZwNXbGrVauWyKl9O3ak\nDuzoUlCn1J07Q0Wkdu3axgYNCgpau3bt3a/YmUym48ePjxo1yrz55qdr166FPl2hys6rHwYp\nm9NgECsNUkamwSBWGqSMTINBrDeIi4tLUFBQkXe3FZqu66U9B4P0/eMaPzT9hFar539XzH21\nTdU8mTTjnz++ennA2+su6I3G7Ts2rehvYwEAANgiGwp2Iom7xrTt/OHhJBHnKn6B7R9u7utT\n2d3ZPvNGwpXIU+F7t+0Ku3xDpFyLMb/snB5YrrRnCwAAULJsKtiJyPUj88e/NfGbLRfzfTPT\npWbQ0CmfTnvx/golPS8AAIBSZ2vBTkRETIln92zZHhp24vzla4kpmfau5T2r1G7o92D7oFZ1\nyvPrtwAA4B5lk8EOAAAAeXF9CwAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAA\nABThUNoTuLekpaXNmDGjUaNGdnZEaksymUxhYWF+fn4U1oKoqjVQVSuhsNZgMpmOHz8+ZswY\nJyen0p4LjCLYlaiPPvpo0qRJpT0LAACMsrOzmzBhQmnPAkYR7EpUgwYNRGTkyJGtW7cu7bko\nZffu3SEhIRTWsqiqNVBVK6Gw1pBV1awzF2wFwa5EZb1H0Lp16759+5b2XFQTEhJCYS2OqloD\nVbUSCmsNISEhvLttW/jXAgAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsA\nAABFEOwAAAAUQbADAABQBMGuRLm6ut7+LyyIwloDVbUGqmolFNYaqKot0nRdL+053EMyMzM3\nb97coUMHe3v70p6LUiisNVBVa6CqVkJhrYGq2iKCHQAAgCJ4KxYAAEARBDsAAABFEOwAAAAU\nQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMA\nAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMGupCT9vXx8/8CGVSu4uLhXqd+m75il4Qml\nPacyIePayZ2rv54Y/OQj9TwcNE3TuixILKCrwRpatpuNSo89+uv8KcG927ZoUMPTxcnVo0aT\nwD4jZ+2IysinM4U1yhQfsWXhf4b3autfv3oFZye3ijWbPNpn5Kxtl/IpK1UtmsTfh9XRNE3T\ntD6r8m6lqkZdm91Ry1eXb67l7ktVFaOjBMTvfK+Fa+7SOzd7+/erpT2zUpe+uHeuunSen5Bf\nR4M1tGw3m5WSp6xZtErtPzmYnKMrhTXDpiEe+da10mOfHErL3pGqFlHy1td9NTs7OxGRZ1bm\n2khVzRA3q0O+a1U6z43L0ZGqKodgVwJSt71WV0TsavWYuelEXEpKXMTmj5+sYyciPi9vTi58\nf6VlrPiXb+BTw9+f9cMfJ2b3ECkg2BmsoWW72bDU74Y06/TSxDn/t/PQici45OSr50JXT+la\n215E7P3/cyxbRwprji2j2vZ8/cNFG/ccPXclOTUx5tyBdTP7N3IUEYd2X1+63Y2qFlHqn6Ma\n2jm0eOvV9iJ5gh1VNUtWsGs29e+7d6OqCiLYWV/C8t4uIlJv5K6UO42pf46sLyKOPRfGFbzn\nPSZ9ee+Cgp3BGlq2m3JuHBjXWESk9WeRt5oobPGZQt+tJyKuL66/1UJViybt4LjmDvZNx4T+\nM7dD3mBHVc1jLNhRVRUR7KwuY+0gVxHxm3w8Z/uJD1qKiHP/lemlM6+yp+BgZ7CGlu2moo3/\nKi8iLT86ffMxhbWEw2Mbiki9sftvPqaqRZJxZGqAo139d/9I0ePyCXZU1UyGgh1VVRJfnrC6\nU2FhKSIeDz/cMGd7g1atKorcCA8/WTrzsiUGa2jZbgo6sf9AoohPYGDdmw0UtjgyU65dOrp5\n9pBBn54Q96B3h7W82U5Vi8B0PGTof/bXHPG/ya1d8u1AVYvkwqLnm1d3d3Zyq1iz2aP935uz\nKyoz+2aqqiSCndVdvnxZRHx8fHJvyGrK2oy7MlhDy3ZTjenM169+eETcO09+r412s43CFsmV\nLx/TNE1zcKvo07zjq2tMXd5buuenV+re2kxVzaaf+WLYpL3eQ+ZMD3IroAtVLZLrEfuPRiem\npadci/xr5/cfD3/0/sdn7k++vZmqKolgZ3UpKSki4uzsnHuDi4uLiCQnJ+fdBzkZrKFlu6kl\n5udXur35W2KtgQsWDal1u5XCFp8p/vxfh/Yejky93UJVzXX+f8ETtrsPnjXzCfcC+1BVM2nl\n6nd7+6s1u8LPXU1OunL+r22Lxna7z0W/sn3UM6N23FqtVFVJBDurc3V1FZEbN27k3pCamioi\nbm4FvULFbQZraNlu6jBd+r/gx56ac6LaU9/8Pv/pKtm2UNgiqfzaVl3XTWnXo8/sX/ffQb6X\nNn327CPPfnfrcgRVNc+lhcNHb3Lt98UnPSrepRdVNZPHC7PXf/JK7zbNald0dfOq1aTdCx+s\n37Okr7fIuflzNt48cKqqJIKd1VWpUkVEIiMjc2/IavL29i75OdkagzW0bDdFpJ9aPPDRPv87\nXr3fgu3fv1jfMcdGClsMmqN71bote4z89vdv+1eSq2umzf4rawNVNUfMstfe+dmu12ef9/W6\naz+qagGVn36pl6dIcnj4mawGqqokgp3V+fr5uYrE7917Imd7xJ49cSLOfn4NSmdetsRgDS3b\nTQXJhz/t1Xbwigv1Bi/fvuyFeg65t1NYS/Bo185fRCIiIrIeU1Vz7NmwIVbi1j5XLdtvI1R8\nebOIyOq+mqZp9SccEqGq1kFV1VTaX8u9B+R/a5893NonN7PvY5enhpbtZutid4x/xEPEudnw\ndVGmAvpQWAu4tmqAl4jUeWf3zQaqaoZ1A/N8Iisn3/EHdV2nqpYQ++OAKiLi9uzqW0dNVVVE\nsCsBqVtv3oy758ebTsSlpsad3PzJ7ZtxJ5X27MqOuwQ7ozW0bDcblhn50/DmriJuD7y9KeZu\nHSmsGa4sGNzm+X/P37gn/HT09dTU65fP/7Vt6ZQBfuVFRPMdszfjVkeqWjz53ceOqpolcfnL\nrQaMnbt+d/jp6OupyVcvHNu5dGLvhm4iIrWGb8n2WxFUVT0EuxIRv+M9/7w/n9d0JD+fp+tx\nczsX9FK9/RdRd/oZrKFlu9msiOkBBVVVxCN4U7auFNawmIJ+fdPO+4mP9+c4bVHV4sg/2FFV\nMyTM757vUtW82s3Ym/OlM1VVDsGupCQcXTq23yO+3uWcnMpVvu+RPqOWHLle2nMqE4wGO91w\nDS3bzTaZEex0CmtYatTeVSFv9mnfokENDxcHJzfPGo1a9Rg6edmB2Hze66aqRVZQsNOpqmE3\nokNXh7z5THv/+jU8nB2c3SvXa9Hx+XHz/ozOyKczVVWLput6wScAAAAA2Ay+FQsAAKAIgh0A\nAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCII\ndgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAA\niiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgB\nAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAUBBtg6v3PbL6NKeBQAYRrADAABQBMEOKHsSF3TR\nNK3xhPDSnsi9ynRl3zdv9wnyr9N7XuyutxrUaNq6R/D070Kj00p7YlbGwgNsn0NpTwAAyhTT\n8VldA1/9NVa/1ZAY9fef6//+c/3/9ixOX/M8fzQBlGVcsQOAbJJ/nPDer7F6pUffW/bnmTVD\nvNp8/NfZI7vWzh7TL6C6s1baswOAuyPYwdaZrkf8Nnf8C0882KRWJTcnV48ajQP7vjsv9Kqe\no1f0l201TXtwxtmk8MXv9g6o41XOxb1qo6Bhs/bGi4ikn90w7YV2japVcHHxqBXQZ+LGC5k5\nnyX94pbPX+v1sK+3u7OTq6dP08eem7A8LP7O9j/fralpWp8VuSZ37D8tNE3rOPtK7mmknlz1\n7/6P+FYu5+ziWfP+7m/MP5J4s8vFT9tq7i/+IiLHp/lpt3Scfc1wRZIi1s8c0atVg2oers7l\nq9znHzTw/eUHYu8cUGHHYvRZfpoxrFvAfd7uzk6uHtWbPDpg/JLDOUdJDF865umH61Vyc3bz\nqh3Qa+SiIwknZzyoadpjOb+PUPhQBupmzsQyo/74ZlSfNk3qVnV3dfOq1Tig46Apy0NjMkRE\nJHzPniSRdu+v+ujZVnU9HERzrljHr03P4Onfhc7qa39nEBZe3sqXkYVXKAutzJJcloBxOmDb\ndr/pk8/Cdqw/7Jer2XpFfREoIn7PvfxIuZwdy7f//Njx2V28cl6KcWj+78OZt3dOO/xpx8p5\nr9U4Nxi2NvrWLN7xEZFnluea3N9T/UWkw6yYnNN4/rV2HrkGq9Dh67O6ruv6hZDAfI6nw6w4\nY+WIXBPc2C3v/l6vbjF8LAZE/fRK03yexanhy2sv3+ySvHvSg7lqLa4PvTTAT0TafxFlzlCG\n6mZ8tKs/vVQzn5e0zoPX6bqu6/vH1hOx6zovQdd1fUuwV2C2yWbHwsuhzCy8QlluZZbksgQM\nI9jB1u19v1OPkV/+3+6w05cTU5Pjov7etvjdx7xFpMGY0Du9sv4EizjU7jJ59f4zsUmJl//+\naUI7TxHxrFbNxavNWwt2RkQnJF09tfWzHj4iUr7f99ezdjWFTWluLyLu/kO+/v2vi/GJsaf2\nrZrQsYa9iHg89V3WWdys86uIfY3Hxy7bfTI6ITnuwp4lwfe7iIjP23/c2i1hfmcRaTQ+zNxi\nnJvV3lVEnGo/8e6ibeFnryQlXTlzZOvSSQM6Tdph+FgKFTnvifIiLo36T//uj4jouJQbSTFn\n9/0wrWddB5HKg9cl6LqeeWBCQ01Eqxz4zpLdEZcTEqKP71r4eqtKWSerO6dPA0OZUTdDoyXO\n724v4uw3dN7OE/9cT72REH3y0JZlkwe2G/WLruu6nr79FR8RuzrPfLr5TMLmgoMdCy+bMrPw\nCmXJlVmSyxIwjGAHFcUv6uYg4jf5+O2Wm3+CG4/bfyNbv7DxjUVEqg9aG5+t9fKsDiJy36i9\nWQ93vOkjIhWfWhqTrY+etn90EzsRLeirKF039/zacNTelGy9Mra/UVtE2vz3/M2Gop5fQ8fU\nFxHPHvMvmfLvYOhYCnN4QmMRx46zL+Vqz9g1sqZIuYE/puv6ttdriEj1F9fnONnGrRngneP0\naWQoXTdaN2OjJc3vbify8O1i53Vt17T2VexFRBycnOwqtez3+pTZa4/EZBReGhZeaS+8Qlly\nZZbssgQM4jN2sHmZ/+yeO+rZIP+63u4uDnaapmmax6ANGSLnz5/P1bV2tydbOmV73LhpE03E\nKejJrhWytXo3beotcuHCBRERiQ4NjRRxe3Los5WzD+XYcsSQh0X0ffv2mz3j2j36POSS7bH9\n/fc3FZErV64UtIcxl3ftOilSsd8bg6rn/yF/ixxL7I6dx0TSN79ay8HBwcHe3t7e3s7Ozs5O\ncwgMuSiSdPr0ZYnav/+SiNfTg7qVz76rZ+/BvTzMHCqbQupmcDS3rs8/5SV7pw8Y8cny3w6d\njct7CxOPNuO2njr2y5xJwb1beNldPfD9FxOH97q/TpP+sw+nZOvGwrul7Cy8QllhZZbUsgQM\nItjBxl1c0e+BtsNmrth65NyVxBuZ2T+5npqamquzt7d3jscOTk52IhW9vXPewsLJyUkkPS3r\nT2t8fLyI1KxVK/cpK6sp8dq1XB93L1zuaYizs7OIZGaaPVJOcXFxIlK9Zn4f1RERCx3LzROW\nnpmZmZmZaTKZTFlXaW5tT0tLu/lEPj55PoWWs8nIUNkUUjejo1UdsGTvT9O7u+366MUnHqhX\nqXyl+m2efuPTX87cyD50+fqdhr0/e9WHvTzafHjgyM9fBQc4Rnw/oufYHbdmxMK7o+wsvEJZ\nYWWW5LIEDCDYwbb98d/RP0SZqj4+4budf124kpCanmnSdT1j7XOulnoGDw8PEbl44YKea0NW\nU3lPT3sRsbe3l3xOLTExMZaahwEVK1YUkaiLF00FdDB2LIXw9PQUEa8RmzP1fIWOqXvziSIj\nI3PvnLPJyFDGGR/N5b7uY+ZvOhJ9Pe7Moa0rZgxqeHHZyC4P9Jx9MnddREQ0t+p+XV6ZvX76\n45pc+OHH0KxWFl42ZWfhFarkV6YFlqXRpwJEhGAHW3fs2HmRWoOmT+0X2KSmV3lnBztN5Nqv\na7amFL6vMdUefNBHJHnNN8tzvF+VcWjO/H0i2oMPBoiIVKpUSUSOhoVlP7Wl7F614VJRntPR\n0VHyu/BTiCqBgfVF4r7/YlFUfgnF6LEUomqbNr4isau/Wl7As4hI9YCA6iKxPy7+OSl7c/za\nRWuz3+HByFDGmT2a5uRZ17/908MmLlg3ua3Eb/pyyfG7dNZ1ufNvkszCy6bsLLxClfzKtMCy\nLOYMcK8h2MG2VavmIXLx+w8/3XIiJjkt5eqF8F++DO743LdFOq3lL3DIsGZ2Evfj8CdenrPt\nWFRC0tUzB36c1L3nzKOZUqH38GeriYjc16ZNNZHTs0a89d3ByOspiZdP7V42utszXxTtxbZz\ntWqeIhc3r9x6PiGtoKsg+QgY8U57N7m2LviRzqMWb//r/NXk5KvnwrYtm/xc5/d3Gj6Wwjz0\n2qggN7n8w9B2vccv2nb0fEzSjZS4iycObV/92cinA4JX3hCRtv96qYEml+YP7jZ6+d7TV5IS\nY07uXjKy65AVMWYPZZyx0Y580Dnoxffn/bwn/MzlxLS0pNjTe5Z+sPCgiJhMJhHZ+e9HuwTP\nWLblyOnohHRdRE+9fGzLt2/0nLBFF9fAwJYiIuLKwsuu7Cy8QpX8yiz+sjT+XIAI97GDjUvb\nN7GZU65FrdXoG9zLS8R54J07QGV9fy1g+pmcu6/sby9S9c0dOVtzf9PwxsFPgirl/Z/Hyfel\nH+980y3i8/a5b47lGfTawKaS98uJeaaRsri3iPiO3nerIXntIK8cQ5lxO7FhjfN5N/D27cQM\nHUuhote/4Vc+7ygiIr0XZ31DMGnX+Ja5b93l+tDgvk1F5Ik5seYMZbxuRkY7ONo3v81a5Z7z\nz+q6rm959U7p7ezt73wqTPPu9NXxW1+NZeHlUmYWXqEstzJLclkChhHsYOtMMXvnvta1RT3v\nck6unj7NOw6d+duljC3Bljy/6rqedu63kBHdA+p6lXN0cK5QrXG7AWMXH8x196340Lmvdm5e\nw93J0a1S3VbPjPshIiX/u04YOBPoV/d8PrRjcx8P56xkYfj8qut6wrE104Z2eaCuVzknJ/eq\nvi0ef37y8gNX7tysw8ixFC7l7G+fv9WnbRMfT1dHp3KVazduGdRv5Gc/Hrh857Ye1w8veq93\nQB1PFyfXirUe6PHWgsPXNw6tKGLff2W6eUMZr1vho5mun9q6cOqw7q2b1/Uu7+zqWb1e83YD\nxs7befHmnDLjT/y+cGpw77YtGvp4utiJ5uzp0+zRvu/O2h6V44YnLLzcyszCK5SFVmZJLkvA\nME3XLfDZFgAoXPzGfzXtuvBSixknC7g+Ueb8NtTz/RbHdr5m6K1C2CzbW5lAgfiMHQCrOD93\naL+JC347ePKfxNSkmLNHfp01vNMLCy+JQ+sXBtjMudPOjr+RylFiZQIFcii8CwCYzxQfvnLq\nvJVTc7ZWfHTmnNfrlM6MikDT8r/jLmyYEisTKBCvRgFbEDqmrnZXPRaYeYsKqz9LnWHfrpw2\npMtDjWt5uTk6latcr2XXYR9tPLDpLT/H4s+0pATNib2334e1oYVneBAlViZQIK7YAbAKrULT\nPuO+6TOutOcB5MTKhNr48gQAAIAieCsWAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEE\nOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAA\nRRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwA\nAAAU8f9jbqy3oPIH+AAAAABJRU5ErkJggg==",
      "text/plain": [
       "Plot with title \"Histogram of amount_co_egenes$amount_co_e_genes\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "hist(amount_co_egenes$amount_co_e_genes, breaks = 25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "857c1fe6-1912-4257-baac-93504417aaa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save for usage in enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "53042ad7-3175-48c7-8ff8-9c2908154ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check co-eGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "60546428-f581-4ecc-b28a-d2ed9293d84d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 34</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>⋯</th><th scope=col>gene2</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th><th scope=col>p_bh_cutoff</th><th scope=col>p_q_cutoff</th><th scope=col>type</th><th scope=col>amount_tests_datasets</th><th scope=col>p_val_mt</th><th scope=col>p_val_corrected</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CD4_T</td><td>oneK1K</td><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>⋯</td><td>A1BG</td><td>0</td><td>97356</td><td>1519.477</td><td>0.00266828</td><td>0.00266828</td><td>unique</td><td>149856468</td><td>2338874</td><td>0.8627429</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>oneK1K</td><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>⋯</td><td>A1BG</td><td>0</td><td>97356</td><td> 162.928</td><td>0.00266828</td><td>0.00266828</td><td>unique</td><td>149856468</td><td> 250789</td><td>0.5682995</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 34\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cell\\_type & dataset & feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & ⋯ & gene2 & is\\_rb\\_mt & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene & p\\_bh\\_cutoff & p\\_q\\_cutoff & type & amount\\_tests\\_datasets & p\\_val\\_mt & p\\_val\\_corrected\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & ⋯ & <chr> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <chr> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t CD4\\_T & oneK1K & A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & ⋯ & A1BG & 0 & 97356 & 1519.477 & 0.00266828 & 0.00266828 & unique & 149856468 & 2338874 & 0.8627429\\\\\n",
       "\t CD4\\_T & oneK1K & A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & ⋯ & A1BG & 0 & 97356 &  162.928 & 0.00266828 & 0.00266828 & unique & 149856468 &  250789 & 0.5682995\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 34\n",
       "\n",
       "| cell_type &lt;chr&gt; | dataset &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | ⋯ ⋯ | gene2 &lt;chr&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; | p_bh_cutoff &lt;dbl&gt; | p_q_cutoff &lt;dbl&gt; | type &lt;chr&gt; | amount_tests_datasets &lt;int&gt; | p_val_mt &lt;dbl&gt; | p_val_corrected &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| CD4_T | oneK1K | A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | ⋯ | A1BG | 0 | 97356 | 1519.477 | 0.00266828 | 0.00266828 | unique | 149856468 | 2338874 | 0.8627429 |\n",
       "| CD4_T | oneK1K | A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | ⋯ | A1BG | 0 | 97356 |  162.928 | 0.00266828 | 0.00266828 | unique | 149856468 |  250789 | 0.5682995 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type dataset feature_id snp_id         p_value     beta beta_se\n",
       "1 CD4_T     oneK1K  A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA     \n",
       "2 CD4_T     oneK1K  A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start ⋯ gene2 is_rb_mt\n",
       "1 NA                        12                 9067664       ⋯ A1BG  0       \n",
       "2 NA                        12                 9067664       ⋯ A1BG  0       \n",
       "  amount_tests_eGene p_val_mt_eGene p_bh_cutoff p_q_cutoff type  \n",
       "1 97356              1519.477       0.00266828  0.00266828 unique\n",
       "2 97356               162.928       0.00266828  0.00266828 unique\n",
       "  amount_tests_datasets p_val_mt p_val_corrected\n",
       "1 149856468             2338874  0.8627429      \n",
       "2 149856468              250789  0.5682995      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sign_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "924c928e-dc98-418d-8059-70fa938e46d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sign_results$co_egene = sign_results$gene1\n",
    "#sign_results$co_egene[ sign_results$gene1 ==  sign_results$egene] = sign_results$gene2[ sign_results$gene1 ==  sign_results$egene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "7ee97383-5148-4f4f-bc54-511f6bb3dde5",
   "metadata": {},
   "outputs": [],
   "source": [
    "amount_egenes = sign_results %>% group_by(coeGene) %>% summarise(amount_e_genes = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "44ce40e1-7655-48ff-a880-d2c938bc39ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>979</li><li>4</li><li>1</li><li>1295</li><li>5</li><li>9</li><li>8</li><li>1030</li><li>1136</li><li>1068</li><li>941</li><li>1006</li><li>1104</li><li>1031</li><li>905</li><li>994</li><li>1123</li><li>1059</li><li>968</li><li>893</li><li>2</li><li>3</li><li>1107</li><li>1077</li><li>1067</li><li>899</li><li>930</li><li>1004</li><li>1131</li><li>985</li><li>1001</li><li>1170</li><li>1097</li><li>1185</li><li>1053</li><li>1099</li><li>1148</li><li>1023</li><li>1086</li><li>1024</li><li>977</li><li>955</li><li>1069</li><li>1138</li><li>993</li><li>1019</li><li>1152</li><li>973</li><li>980</li><li>1062</li><li>1071</li><li>1014</li><li>1005</li><li>925</li><li>944</li><li>884</li><li>950</li><li>998</li><li>1092</li><li>984</li><li>1203</li><li>987</li><li>952</li><li>1082</li><li>1095</li><li>967</li><li>1188</li><li>978</li><li>1054</li><li>1090</li><li>1040</li><li>1017</li><li>1042</li><li>1037</li><li>991</li><li>1642</li><li>1298</li><li>909</li><li>1330</li><li>1394</li><li>927</li><li>1114</li><li>1215</li><li>1021</li><li>1043</li><li>1064</li><li>1039</li><li>1110</li><li>1063</li><li>1051</li><li>1034</li><li>1016</li><li>1196</li><li>1429</li><li>1098</li><li>1044</li><li>1085</li><li>947</li><li>960</li><li>1010</li><li>896</li><li>1151</li><li>958</li><li>1032</li><li>990</li><li>1366</li><li>1055</li><li>989</li><li>1156</li><li>887</li><li>1025</li><li>1057</li><li>1362</li><li>901</li><li>942</li><li>1190</li><li>1206</li><li>836</li><li>1020</li><li>1012</li><li>914</li><li>1205</li><li>1027</li><li>1089</li><li>915</li><li>1002</li><li>854</li><li>935</li><li>1210</li><li>1079</li><li>1094</li><li>969</li><li>1580</li><li>1124</li><li>986</li><li>1375</li><li>1112</li><li>964</li><li>982</li><li>1448</li><li>857</li><li>1022</li><li>988</li><li>1382</li><li>1239</li><li>1102</li><li>1070</li><li>1165</li><li>1041</li><li>992</li><li>949</li><li>932</li><li>1153</li><li>966</li><li>983</li><li>1084</li><li>1049</li><li>1081</li><li>1045</li><li>902</li><li>956</li><li>1008</li><li>1140</li><li>1052</li><li>1276</li><li>981</li><li>1116</li><li>1091</li><li>1126</li><li>962</li><li>1046</li><li>1252</li><li>1007</li><li>936</li><li>1296</li><li>1324</li><li>974</li><li>1000</li><li>1080</li><li>929</li><li>1038</li><li>1061</li><li>965</li><li>1179</li><li>1011</li><li>1378</li><li>1111</li><li>1293</li><li>894</li><li>1228</li><li>918</li><li>1113</li><li>1076</li><li>6</li><li>945</li><li>1047</li><li>1159</li><li>1604</li><li>1679</li><li>1088</li><li>⋯</li><li>1358</li><li>1319</li><li>1731</li><li>1342</li><li>868</li><li>1701</li><li>1273</li><li>907</li><li>1202</li><li>1377</li><li>1441</li><li>1405</li><li>1796</li><li>1586</li><li>1274</li><li>1433</li><li>840</li><li>1410</li><li>846</li><li>1420</li><li>1443</li><li>1400</li><li>1402</li><li>1499</li><li>1246</li><li>1350</li><li>1453</li><li>1336</li><li>1869</li><li>1777</li><li>1424</li><li>1289</li><li>1372</li><li>1338</li><li>1853</li><li>844</li><li>861</li><li>1673</li><li>1496</li><li>1779</li><li>1249</li><li>1412</li><li>1344</li><li>1413</li><li>1509</li><li>1503</li><li>1368</li><li>880</li><li>1552</li><li>1549</li><li>1650</li><li>1445</li><li>1323</li><li>1628</li><li>1269</li><li>839</li><li>1562</li><li>1347</li><li>1738</li><li>1641</li><li>1475</li><li>871</li><li>1244</li><li>1308</li><li>1318</li><li>1383</li><li>1810</li><li>1542</li><li>1369</li><li>881</li><li>1320</li><li>1370</li><li>1245</li><li>1349</li><li>784</li><li>870</li><li>832</li><li>1491</li><li>2104</li><li>2109</li><li>2096</li><li>2035</li><li>2254</li><li>2057</li><li>1982</li><li>2047</li><li>2020</li><li>2008</li><li>2054</li><li>2088</li><li>2263</li><li>1962</li><li>1761</li><li>1825</li><li>2031</li><li>1882</li><li>1998</li><li>1925</li><li>1914</li><li>2159</li><li>2155</li><li>2258</li><li>2144</li><li>1886</li><li>2108</li><li>1980</li><li>1867</li><li>1415</li><li>2026</li><li>1967</li><li>2076</li><li>2024</li><li>2033</li><li>2081</li><li>1764</li><li>2053</li><li>1778</li><li>1879</li><li>2177</li><li>1722</li><li>1971</li><li>2136</li><li>1928</li><li>1637</li><li>2195</li><li>2218</li><li>2291</li><li>2007</li><li>2203</li><li>2106</li><li>1976</li><li>2114</li><li>2030</li><li>2015</li><li>2065</li><li>2137</li><li>2111</li><li>1427</li><li>2236</li><li>2141</li><li>2038</li><li>2190</li><li>2222</li><li>2021</li><li>2044</li><li>2205</li><li>1861</li><li>2099</li><li>1831</li><li>1724</li><li>1783</li><li>2131</li><li>1476</li><li>1461</li><li>1414</li><li>1408</li><li>1457</li><li>1248</li><li>1348</li><li>1725</li><li>1281</li><li>1260</li><li>1480</li><li>842</li><li>1332</li><li>862</li><li>820</li><li>1902</li><li>1521</li><li>1473</li><li>819</li><li>883</li><li>1240</li><li>1285</li><li>866</li><li>1409</li><li>864</li><li>1532</li><li>1572</li><li>1469</li><li>1265</li><li>1576</li><li>1310</li><li>1843</li><li>1294</li><li>1720</li><li>1603</li><li>1268</li><li>1488</li><li>1794</li><li>1430</li><li>1307</li><li>1404</li><li>1478</li><li>1334</li><li>827</li><li>1343</li><li>874</li><li>825</li><li>824</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 979\n",
       "\\item 4\n",
       "\\item 1\n",
       "\\item 1295\n",
       "\\item 5\n",
       "\\item 9\n",
       "\\item 8\n",
       "\\item 1030\n",
       "\\item 1136\n",
       "\\item 1068\n",
       "\\item 941\n",
       "\\item 1006\n",
       "\\item 1104\n",
       "\\item 1031\n",
       "\\item 905\n",
       "\\item 994\n",
       "\\item 1123\n",
       "\\item 1059\n",
       "\\item 968\n",
       "\\item 893\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 1107\n",
       "\\item 1077\n",
       "\\item 1067\n",
       "\\item 899\n",
       "\\item 930\n",
       "\\item 1004\n",
       "\\item 1131\n",
       "\\item 985\n",
       "\\item 1001\n",
       "\\item 1170\n",
       "\\item 1097\n",
       "\\item 1185\n",
       "\\item 1053\n",
       "\\item 1099\n",
       "\\item 1148\n",
       "\\item 1023\n",
       "\\item 1086\n",
       "\\item 1024\n",
       "\\item 977\n",
       "\\item 955\n",
       "\\item 1069\n",
       "\\item 1138\n",
       "\\item 993\n",
       "\\item 1019\n",
       "\\item 1152\n",
       "\\item 973\n",
       "\\item 980\n",
       "\\item 1062\n",
       "\\item 1071\n",
       "\\item 1014\n",
       "\\item 1005\n",
       "\\item 925\n",
       "\\item 944\n",
       "\\item 884\n",
       "\\item 950\n",
       "\\item 998\n",
       "\\item 1092\n",
       "\\item 984\n",
       "\\item 1203\n",
       "\\item 987\n",
       "\\item 952\n",
       "\\item 1082\n",
       "\\item 1095\n",
       "\\item 967\n",
       "\\item 1188\n",
       "\\item 978\n",
       "\\item 1054\n",
       "\\item 1090\n",
       "\\item 1040\n",
       "\\item 1017\n",
       "\\item 1042\n",
       "\\item 1037\n",
       "\\item 991\n",
       "\\item 1642\n",
       "\\item 1298\n",
       "\\item 909\n",
       "\\item 1330\n",
       "\\item 1394\n",
       "\\item 927\n",
       "\\item 1114\n",
       "\\item 1215\n",
       "\\item 1021\n",
       "\\item 1043\n",
       "\\item 1064\n",
       "\\item 1039\n",
       "\\item 1110\n",
       "\\item 1063\n",
       "\\item 1051\n",
       "\\item 1034\n",
       "\\item 1016\n",
       "\\item 1196\n",
       "\\item 1429\n",
       "\\item 1098\n",
       "\\item 1044\n",
       "\\item 1085\n",
       "\\item 947\n",
       "\\item 960\n",
       "\\item 1010\n",
       "\\item 896\n",
       "\\item 1151\n",
       "\\item 958\n",
       "\\item 1032\n",
       "\\item 990\n",
       "\\item 1366\n",
       "\\item 1055\n",
       "\\item 989\n",
       "\\item 1156\n",
       "\\item 887\n",
       "\\item 1025\n",
       "\\item 1057\n",
       "\\item 1362\n",
       "\\item 901\n",
       "\\item 942\n",
       "\\item 1190\n",
       "\\item 1206\n",
       "\\item 836\n",
       "\\item 1020\n",
       "\\item 1012\n",
       "\\item 914\n",
       "\\item 1205\n",
       "\\item 1027\n",
       "\\item 1089\n",
       "\\item 915\n",
       "\\item 1002\n",
       "\\item 854\n",
       "\\item 935\n",
       "\\item 1210\n",
       "\\item 1079\n",
       "\\item 1094\n",
       "\\item 969\n",
       "\\item 1580\n",
       "\\item 1124\n",
       "\\item 986\n",
       "\\item 1375\n",
       "\\item 1112\n",
       "\\item 964\n",
       "\\item 982\n",
       "\\item 1448\n",
       "\\item 857\n",
       "\\item 1022\n",
       "\\item 988\n",
       "\\item 1382\n",
       "\\item 1239\n",
       "\\item 1102\n",
       "\\item 1070\n",
       "\\item 1165\n",
       "\\item 1041\n",
       "\\item 992\n",
       "\\item 949\n",
       "\\item 932\n",
       "\\item 1153\n",
       "\\item 966\n",
       "\\item 983\n",
       "\\item 1084\n",
       "\\item 1049\n",
       "\\item 1081\n",
       "\\item 1045\n",
       "\\item 902\n",
       "\\item 956\n",
       "\\item 1008\n",
       "\\item 1140\n",
       "\\item 1052\n",
       "\\item 1276\n",
       "\\item 981\n",
       "\\item 1116\n",
       "\\item 1091\n",
       "\\item 1126\n",
       "\\item 962\n",
       "\\item 1046\n",
       "\\item 1252\n",
       "\\item 1007\n",
       "\\item 936\n",
       "\\item 1296\n",
       "\\item 1324\n",
       "\\item 974\n",
       "\\item 1000\n",
       "\\item 1080\n",
       "\\item 929\n",
       "\\item 1038\n",
       "\\item 1061\n",
       "\\item 965\n",
       "\\item 1179\n",
       "\\item 1011\n",
       "\\item 1378\n",
       "\\item 1111\n",
       "\\item 1293\n",
       "\\item 894\n",
       "\\item 1228\n",
       "\\item 918\n",
       "\\item 1113\n",
       "\\item 1076\n",
       "\\item 6\n",
       "\\item 945\n",
       "\\item 1047\n",
       "\\item 1159\n",
       "\\item 1604\n",
       "\\item 1679\n",
       "\\item 1088\n",
       "\\item ⋯\n",
       "\\item 1358\n",
       "\\item 1319\n",
       "\\item 1731\n",
       "\\item 1342\n",
       "\\item 868\n",
       "\\item 1701\n",
       "\\item 1273\n",
       "\\item 907\n",
       "\\item 1202\n",
       "\\item 1377\n",
       "\\item 1441\n",
       "\\item 1405\n",
       "\\item 1796\n",
       "\\item 1586\n",
       "\\item 1274\n",
       "\\item 1433\n",
       "\\item 840\n",
       "\\item 1410\n",
       "\\item 846\n",
       "\\item 1420\n",
       "\\item 1443\n",
       "\\item 1400\n",
       "\\item 1402\n",
       "\\item 1499\n",
       "\\item 1246\n",
       "\\item 1350\n",
       "\\item 1453\n",
       "\\item 1336\n",
       "\\item 1869\n",
       "\\item 1777\n",
       "\\item 1424\n",
       "\\item 1289\n",
       "\\item 1372\n",
       "\\item 1338\n",
       "\\item 1853\n",
       "\\item 844\n",
       "\\item 861\n",
       "\\item 1673\n",
       "\\item 1496\n",
       "\\item 1779\n",
       "\\item 1249\n",
       "\\item 1412\n",
       "\\item 1344\n",
       "\\item 1413\n",
       "\\item 1509\n",
       "\\item 1503\n",
       "\\item 1368\n",
       "\\item 880\n",
       "\\item 1552\n",
       "\\item 1549\n",
       "\\item 1650\n",
       "\\item 1445\n",
       "\\item 1323\n",
       "\\item 1628\n",
       "\\item 1269\n",
       "\\item 839\n",
       "\\item 1562\n",
       "\\item 1347\n",
       "\\item 1738\n",
       "\\item 1641\n",
       "\\item 1475\n",
       "\\item 871\n",
       "\\item 1244\n",
       "\\item 1308\n",
       "\\item 1318\n",
       "\\item 1383\n",
       "\\item 1810\n",
       "\\item 1542\n",
       "\\item 1369\n",
       "\\item 881\n",
       "\\item 1320\n",
       "\\item 1370\n",
       "\\item 1245\n",
       "\\item 1349\n",
       "\\item 784\n",
       "\\item 870\n",
       "\\item 832\n",
       "\\item 1491\n",
       "\\item 2104\n",
       "\\item 2109\n",
       "\\item 2096\n",
       "\\item 2035\n",
       "\\item 2254\n",
       "\\item 2057\n",
       "\\item 1982\n",
       "\\item 2047\n",
       "\\item 2020\n",
       "\\item 2008\n",
       "\\item 2054\n",
       "\\item 2088\n",
       "\\item 2263\n",
       "\\item 1962\n",
       "\\item 1761\n",
       "\\item 1825\n",
       "\\item 2031\n",
       "\\item 1882\n",
       "\\item 1998\n",
       "\\item 1925\n",
       "\\item 1914\n",
       "\\item 2159\n",
       "\\item 2155\n",
       "\\item 2258\n",
       "\\item 2144\n",
       "\\item 1886\n",
       "\\item 2108\n",
       "\\item 1980\n",
       "\\item 1867\n",
       "\\item 1415\n",
       "\\item 2026\n",
       "\\item 1967\n",
       "\\item 2076\n",
       "\\item 2024\n",
       "\\item 2033\n",
       "\\item 2081\n",
       "\\item 1764\n",
       "\\item 2053\n",
       "\\item 1778\n",
       "\\item 1879\n",
       "\\item 2177\n",
       "\\item 1722\n",
       "\\item 1971\n",
       "\\item 2136\n",
       "\\item 1928\n",
       "\\item 1637\n",
       "\\item 2195\n",
       "\\item 2218\n",
       "\\item 2291\n",
       "\\item 2007\n",
       "\\item 2203\n",
       "\\item 2106\n",
       "\\item 1976\n",
       "\\item 2114\n",
       "\\item 2030\n",
       "\\item 2015\n",
       "\\item 2065\n",
       "\\item 2137\n",
       "\\item 2111\n",
       "\\item 1427\n",
       "\\item 2236\n",
       "\\item 2141\n",
       "\\item 2038\n",
       "\\item 2190\n",
       "\\item 2222\n",
       "\\item 2021\n",
       "\\item 2044\n",
       "\\item 2205\n",
       "\\item 1861\n",
       "\\item 2099\n",
       "\\item 1831\n",
       "\\item 1724\n",
       "\\item 1783\n",
       "\\item 2131\n",
       "\\item 1476\n",
       "\\item 1461\n",
       "\\item 1414\n",
       "\\item 1408\n",
       "\\item 1457\n",
       "\\item 1248\n",
       "\\item 1348\n",
       "\\item 1725\n",
       "\\item 1281\n",
       "\\item 1260\n",
       "\\item 1480\n",
       "\\item 842\n",
       "\\item 1332\n",
       "\\item 862\n",
       "\\item 820\n",
       "\\item 1902\n",
       "\\item 1521\n",
       "\\item 1473\n",
       "\\item 819\n",
       "\\item 883\n",
       "\\item 1240\n",
       "\\item 1285\n",
       "\\item 866\n",
       "\\item 1409\n",
       "\\item 864\n",
       "\\item 1532\n",
       "\\item 1572\n",
       "\\item 1469\n",
       "\\item 1265\n",
       "\\item 1576\n",
       "\\item 1310\n",
       "\\item 1843\n",
       "\\item 1294\n",
       "\\item 1720\n",
       "\\item 1603\n",
       "\\item 1268\n",
       "\\item 1488\n",
       "\\item 1794\n",
       "\\item 1430\n",
       "\\item 1307\n",
       "\\item 1404\n",
       "\\item 1478\n",
       "\\item 1334\n",
       "\\item 827\n",
       "\\item 1343\n",
       "\\item 874\n",
       "\\item 825\n",
       "\\item 824\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 979\n",
       "2. 4\n",
       "3. 1\n",
       "4. 1295\n",
       "5. 5\n",
       "6. 9\n",
       "7. 8\n",
       "8. 1030\n",
       "9. 1136\n",
       "10. 1068\n",
       "11. 941\n",
       "12. 1006\n",
       "13. 1104\n",
       "14. 1031\n",
       "15. 905\n",
       "16. 994\n",
       "17. 1123\n",
       "18. 1059\n",
       "19. 968\n",
       "20. 893\n",
       "21. 2\n",
       "22. 3\n",
       "23. 1107\n",
       "24. 1077\n",
       "25. 1067\n",
       "26. 899\n",
       "27. 930\n",
       "28. 1004\n",
       "29. 1131\n",
       "30. 985\n",
       "31. 1001\n",
       "32. 1170\n",
       "33. 1097\n",
       "34. 1185\n",
       "35. 1053\n",
       "36. 1099\n",
       "37. 1148\n",
       "38. 1023\n",
       "39. 1086\n",
       "40. 1024\n",
       "41. 977\n",
       "42. 955\n",
       "43. 1069\n",
       "44. 1138\n",
       "45. 993\n",
       "46. 1019\n",
       "47. 1152\n",
       "48. 973\n",
       "49. 980\n",
       "50. 1062\n",
       "51. 1071\n",
       "52. 1014\n",
       "53. 1005\n",
       "54. 925\n",
       "55. 944\n",
       "56. 884\n",
       "57. 950\n",
       "58. 998\n",
       "59. 1092\n",
       "60. 984\n",
       "61. 1203\n",
       "62. 987\n",
       "63. 952\n",
       "64. 1082\n",
       "65. 1095\n",
       "66. 967\n",
       "67. 1188\n",
       "68. 978\n",
       "69. 1054\n",
       "70. 1090\n",
       "71. 1040\n",
       "72. 1017\n",
       "73. 1042\n",
       "74. 1037\n",
       "75. 991\n",
       "76. 1642\n",
       "77. 1298\n",
       "78. 909\n",
       "79. 1330\n",
       "80. 1394\n",
       "81. 927\n",
       "82. 1114\n",
       "83. 1215\n",
       "84. 1021\n",
       "85. 1043\n",
       "86. 1064\n",
       "87. 1039\n",
       "88. 1110\n",
       "89. 1063\n",
       "90. 1051\n",
       "91. 1034\n",
       "92. 1016\n",
       "93. 1196\n",
       "94. 1429\n",
       "95. 1098\n",
       "96. 1044\n",
       "97. 1085\n",
       "98. 947\n",
       "99. 960\n",
       "100. 1010\n",
       "101. 896\n",
       "102. 1151\n",
       "103. 958\n",
       "104. 1032\n",
       "105. 990\n",
       "106. 1366\n",
       "107. 1055\n",
       "108. 989\n",
       "109. 1156\n",
       "110. 887\n",
       "111. 1025\n",
       "112. 1057\n",
       "113. 1362\n",
       "114. 901\n",
       "115. 942\n",
       "116. 1190\n",
       "117. 1206\n",
       "118. 836\n",
       "119. 1020\n",
       "120. 1012\n",
       "121. 914\n",
       "122. 1205\n",
       "123. 1027\n",
       "124. 1089\n",
       "125. 915\n",
       "126. 1002\n",
       "127. 854\n",
       "128. 935\n",
       "129. 1210\n",
       "130. 1079\n",
       "131. 1094\n",
       "132. 969\n",
       "133. 1580\n",
       "134. 1124\n",
       "135. 986\n",
       "136. 1375\n",
       "137. 1112\n",
       "138. 964\n",
       "139. 982\n",
       "140. 1448\n",
       "141. 857\n",
       "142. 1022\n",
       "143. 988\n",
       "144. 1382\n",
       "145. 1239\n",
       "146. 1102\n",
       "147. 1070\n",
       "148. 1165\n",
       "149. 1041\n",
       "150. 992\n",
       "151. 949\n",
       "152. 932\n",
       "153. 1153\n",
       "154. 966\n",
       "155. 983\n",
       "156. 1084\n",
       "157. 1049\n",
       "158. 1081\n",
       "159. 1045\n",
       "160. 902\n",
       "161. 956\n",
       "162. 1008\n",
       "163. 1140\n",
       "164. 1052\n",
       "165. 1276\n",
       "166. 981\n",
       "167. 1116\n",
       "168. 1091\n",
       "169. 1126\n",
       "170. 962\n",
       "171. 1046\n",
       "172. 1252\n",
       "173. 1007\n",
       "174. 936\n",
       "175. 1296\n",
       "176. 1324\n",
       "177. 974\n",
       "178. 1000\n",
       "179. 1080\n",
       "180. 929\n",
       "181. 1038\n",
       "182. 1061\n",
       "183. 965\n",
       "184. 1179\n",
       "185. 1011\n",
       "186. 1378\n",
       "187. 1111\n",
       "188. 1293\n",
       "189. 894\n",
       "190. 1228\n",
       "191. 918\n",
       "192. 1113\n",
       "193. 1076\n",
       "194. 6\n",
       "195. 945\n",
       "196. 1047\n",
       "197. 1159\n",
       "198. 1604\n",
       "199. 1679\n",
       "200. 1088\n",
       "201. ⋯\n",
       "202. 1358\n",
       "203. 1319\n",
       "204. 1731\n",
       "205. 1342\n",
       "206. 868\n",
       "207. 1701\n",
       "208. 1273\n",
       "209. 907\n",
       "210. 1202\n",
       "211. 1377\n",
       "212. 1441\n",
       "213. 1405\n",
       "214. 1796\n",
       "215. 1586\n",
       "216. 1274\n",
       "217. 1433\n",
       "218. 840\n",
       "219. 1410\n",
       "220. 846\n",
       "221. 1420\n",
       "222. 1443\n",
       "223. 1400\n",
       "224. 1402\n",
       "225. 1499\n",
       "226. 1246\n",
       "227. 1350\n",
       "228. 1453\n",
       "229. 1336\n",
       "230. 1869\n",
       "231. 1777\n",
       "232. 1424\n",
       "233. 1289\n",
       "234. 1372\n",
       "235. 1338\n",
       "236. 1853\n",
       "237. 844\n",
       "238. 861\n",
       "239. 1673\n",
       "240. 1496\n",
       "241. 1779\n",
       "242. 1249\n",
       "243. 1412\n",
       "244. 1344\n",
       "245. 1413\n",
       "246. 1509\n",
       "247. 1503\n",
       "248. 1368\n",
       "249. 880\n",
       "250. 1552\n",
       "251. 1549\n",
       "252. 1650\n",
       "253. 1445\n",
       "254. 1323\n",
       "255. 1628\n",
       "256. 1269\n",
       "257. 839\n",
       "258. 1562\n",
       "259. 1347\n",
       "260. 1738\n",
       "261. 1641\n",
       "262. 1475\n",
       "263. 871\n",
       "264. 1244\n",
       "265. 1308\n",
       "266. 1318\n",
       "267. 1383\n",
       "268. 1810\n",
       "269. 1542\n",
       "270. 1369\n",
       "271. 881\n",
       "272. 1320\n",
       "273. 1370\n",
       "274. 1245\n",
       "275. 1349\n",
       "276. 784\n",
       "277. 870\n",
       "278. 832\n",
       "279. 1491\n",
       "280. 2104\n",
       "281. 2109\n",
       "282. 2096\n",
       "283. 2035\n",
       "284. 2254\n",
       "285. 2057\n",
       "286. 1982\n",
       "287. 2047\n",
       "288. 2020\n",
       "289. 2008\n",
       "290. 2054\n",
       "291. 2088\n",
       "292. 2263\n",
       "293. 1962\n",
       "294. 1761\n",
       "295. 1825\n",
       "296. 2031\n",
       "297. 1882\n",
       "298. 1998\n",
       "299. 1925\n",
       "300. 1914\n",
       "301. 2159\n",
       "302. 2155\n",
       "303. 2258\n",
       "304. 2144\n",
       "305. 1886\n",
       "306. 2108\n",
       "307. 1980\n",
       "308. 1867\n",
       "309. 1415\n",
       "310. 2026\n",
       "311. 1967\n",
       "312. 2076\n",
       "313. 2024\n",
       "314. 2033\n",
       "315. 2081\n",
       "316. 1764\n",
       "317. 2053\n",
       "318. 1778\n",
       "319. 1879\n",
       "320. 2177\n",
       "321. 1722\n",
       "322. 1971\n",
       "323. 2136\n",
       "324. 1928\n",
       "325. 1637\n",
       "326. 2195\n",
       "327. 2218\n",
       "328. 2291\n",
       "329. 2007\n",
       "330. 2203\n",
       "331. 2106\n",
       "332. 1976\n",
       "333. 2114\n",
       "334. 2030\n",
       "335. 2015\n",
       "336. 2065\n",
       "337. 2137\n",
       "338. 2111\n",
       "339. 1427\n",
       "340. 2236\n",
       "341. 2141\n",
       "342. 2038\n",
       "343. 2190\n",
       "344. 2222\n",
       "345. 2021\n",
       "346. 2044\n",
       "347. 2205\n",
       "348. 1861\n",
       "349. 2099\n",
       "350. 1831\n",
       "351. 1724\n",
       "352. 1783\n",
       "353. 2131\n",
       "354. 1476\n",
       "355. 1461\n",
       "356. 1414\n",
       "357. 1408\n",
       "358. 1457\n",
       "359. 1248\n",
       "360. 1348\n",
       "361. 1725\n",
       "362. 1281\n",
       "363. 1260\n",
       "364. 1480\n",
       "365. 842\n",
       "366. 1332\n",
       "367. 862\n",
       "368. 820\n",
       "369. 1902\n",
       "370. 1521\n",
       "371. 1473\n",
       "372. 819\n",
       "373. 883\n",
       "374. 1240\n",
       "375. 1285\n",
       "376. 866\n",
       "377. 1409\n",
       "378. 864\n",
       "379. 1532\n",
       "380. 1572\n",
       "381. 1469\n",
       "382. 1265\n",
       "383. 1576\n",
       "384. 1310\n",
       "385. 1843\n",
       "386. 1294\n",
       "387. 1720\n",
       "388. 1603\n",
       "389. 1268\n",
       "390. 1488\n",
       "391. 1794\n",
       "392. 1430\n",
       "393. 1307\n",
       "394. 1404\n",
       "395. 1478\n",
       "396. 1334\n",
       "397. 827\n",
       "398. 1343\n",
       "399. 874\n",
       "400. 825\n",
       "401. 824\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1]  979    4    1 1295    5    9    8 1030 1136 1068  941 1006 1104 1031  905\n",
       " [16]  994 1123 1059  968  893    2    3 1107 1077 1067  899  930 1004 1131  985\n",
       " [31] 1001 1170 1097 1185 1053 1099 1148 1023 1086 1024  977  955 1069 1138  993\n",
       " [46] 1019 1152  973  980 1062 1071 1014 1005  925  944  884  950  998 1092  984\n",
       " [61] 1203  987  952 1082 1095  967 1188  978 1054 1090 1040 1017 1042 1037  991\n",
       " [76] 1642 1298  909 1330 1394  927 1114 1215 1021 1043 1064 1039 1110 1063 1051\n",
       " [91] 1034 1016 1196 1429 1098 1044 1085  947  960 1010  896 1151  958 1032  990\n",
       "[106] 1366 1055  989 1156  887 1025 1057 1362  901  942 1190 1206  836 1020 1012\n",
       "[121]  914 1205 1027 1089  915 1002  854  935 1210 1079 1094  969 1580 1124  986\n",
       "[136] 1375 1112  964  982 1448  857 1022  988 1382 1239 1102 1070 1165 1041  992\n",
       "[151]  949  932 1153  966  983 1084 1049 1081 1045  902  956 1008 1140 1052 1276\n",
       "[166]  981 1116 1091 1126  962 1046 1252 1007  936 1296 1324  974 1000 1080  929\n",
       "[181] 1038 1061  965 1179 1011 1378 1111 1293  894 1228  918 1113 1076    6  945\n",
       "[196] 1047 1159 1604 1679 1088 1177  963 1143  999 1115 1173 1181 1139 1066 1481\n",
       "[211] 1227  891 1060 1029 1199 1033  948 1200 1015 1279 1075  895 1447 1230 1233\n",
       "[226] 1180  970 1072 1065 1363 1050 1261 1254 1158 1035 1118  919 1087 1226 1026\n",
       "[241]  838 1175  953 1169 1191 1167 1003 1133 1162 1146  959 1141 1018  934 1157\n",
       "[256] 1357 1288 1127  910 1048 1145 1128 1163 1083 1056 1101 1103  937  877  976\n",
       "[271]  853  946 1197 1105 1236 1287  939 1168 1498 1465 1223 1172 1120 1074 1474\n",
       "[286] 1547 1160 1096 1164  933 1121 1119 1255  997 1186 1058  996 1688  903 1009\n",
       "[301]  931 1797 1800 1073 1335 1161  878 1108 1251 1356  922 1411  954 1238  904\n",
       "[316] 1142 1117  908  972  995 1078 1201 1442 1421 1109 1132  975 1507 1135 1272\n",
       "[331] 1321  961 1454  851 1187 1093 1385    7  924 1212 1379 1192 1147 1309 1280\n",
       "[346] 1313 1270 1013 1391 1209 1129  885 1106  928 1122  890 1331 1130 1028 1182\n",
       "[361] 1818 1513 1371 1224   20 1625 1264 1241 1155 1183 1229  943 1297 1446 1267\n",
       "[376] 1416 1253   22 1388 1225 1675 1250  971 1150 1216 1189 1231 1381  872  951\n",
       "[391] 1036  957 1194 1312  889  940 1100 1340 1306 1214  912 1374 1609  875 1283\n",
       "[406] 1333 1220   12   10   13   23   32   92   17 1198  898 1351  803 1234  823\n",
       "[421] 1494 1380 1432  856 1176 1840  792   26   11   24   15 1530 1144 1305 1193\n",
       "[436] 1125 1207  892 1645  850  858 1137  886 1242 1219 1232 1149   16  938  926\n",
       "[451]  906 1154 1243  923 1222 1337  879 1218 1316  920 1486 1235 1178 1211 2039\n",
       "[466] 1920 1595 2198 1493 1278 1277 1584 1311 1263 1166 1271  913  869 1523 1256\n",
       "[481] 1134  897  921  804 1315 1365  888 1221  917 1257 1204 1497 1933 1329  900\n",
       "[496] 1262 1367 1830 1395   28 1392 1463 1208 1760 1317 1396 1302 1314 1171 1341\n",
       "[511] 1944 1328  873  845 1258  863  882  809   14  916  860 1247 1237 1470 1291\n",
       "[526] 1346 1286 1213 1195 1460 1303   18   29   72   27   64   88   84   71  128\n",
       "[541]   42 1561 1259   50   61  799  911 1525 1528 1174  867 1399 1301 1753  876\n",
       "[556] 1184 1290 1849 1352 1482  865 1544 1846  859 1327  835 1217 1559 1639 1282\n",
       "[571] 1304 1417 1275 1655 1345 1821 1358 1319 1731 1342  868 1701 1273  907 1202\n",
       "[586] 1377 1441 1405 1796 1586 1274 1433  840 1410  846 1420 1443 1400 1402 1499\n",
       "[601] 1246 1350 1453 1336 1869 1777 1424 1289 1372 1338 1853  844  861 1673 1496\n",
       "[616] 1779 1249 1412 1344 1413 1509 1503 1368  880 1552 1549 1650 1445 1323 1628\n",
       "[631] 1269  839 1562 1347 1738 1641 1475  871 1244 1308 1318 1383 1810 1542 1369\n",
       "[646]  881 1320 1370 1245 1349  784  870  832 1491 2104 2109 2096 2035 2254 2057\n",
       "[661] 1982 2047 2020 2008 2054 2088 2263 1962 1761 1825 2031 1882 1998 1925 1914\n",
       "[676] 2159 2155 2258 2144 1886 2108 1980 1867 1415 2026 1967 2076 2024 2033 2081\n",
       "[691] 1764 2053 1778 1879 2177 1722 1971 2136 1928 1637 2195 2218 2291 2007 2203\n",
       "[706] 2106 1976 2114 2030 2015 2065 2137 2111 1427 2236 2141 2038 2190 2222 2021\n",
       "[721] 2044 2205 1861 2099 1831 1724 1783 2131 1476 1461 1414 1408 1457 1248 1348\n",
       "[736] 1725 1281 1260 1480  842 1332  862  820 1902 1521 1473  819  883 1240 1285\n",
       "[751]  866 1409  864 1532 1572 1469 1265 1576 1310 1843 1294 1720 1603 1268 1488\n",
       "[766] 1794 1430 1307 1404 1478 1334  827 1343  874  825  824"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(amount_egenes$amount_e_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "3d75dfa5-f5e8-45ba-a5e3-db25f098abe6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 10 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>coeGene</th><th scope=col>amount_e_genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>RPS14 </td><td>2291</td></tr>\n",
       "\t<tr><td>RPL21 </td><td>2263</td></tr>\n",
       "\t<tr><td>RPL32 </td><td>2258</td></tr>\n",
       "\t<tr><td>RPL13 </td><td>2254</td></tr>\n",
       "\t<tr><td>RPL31 </td><td>2254</td></tr>\n",
       "\t<tr><td>RPS27 </td><td>2236</td></tr>\n",
       "\t<tr><td>RPS3A </td><td>2222</td></tr>\n",
       "\t<tr><td>RPS13 </td><td>2218</td></tr>\n",
       "\t<tr><td>RPS6  </td><td>2205</td></tr>\n",
       "\t<tr><td>RPS15A</td><td>2203</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 10 × 2\n",
       "\\begin{tabular}{ll}\n",
       " coeGene & amount\\_e\\_genes\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t RPS14  & 2291\\\\\n",
       "\t RPL21  & 2263\\\\\n",
       "\t RPL32  & 2258\\\\\n",
       "\t RPL13  & 2254\\\\\n",
       "\t RPL31  & 2254\\\\\n",
       "\t RPS27  & 2236\\\\\n",
       "\t RPS3A  & 2222\\\\\n",
       "\t RPS13  & 2218\\\\\n",
       "\t RPS6   & 2205\\\\\n",
       "\t RPS15A & 2203\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 10 × 2\n",
       "\n",
       "| coeGene &lt;chr&gt; | amount_e_genes &lt;int&gt; |\n",
       "|---|---|\n",
       "| RPS14  | 2291 |\n",
       "| RPL21  | 2263 |\n",
       "| RPL32  | 2258 |\n",
       "| RPL13  | 2254 |\n",
       "| RPL31  | 2254 |\n",
       "| RPS27  | 2236 |\n",
       "| RPS3A  | 2222 |\n",
       "| RPS13  | 2218 |\n",
       "| RPS6   | 2205 |\n",
       "| RPS15A | 2203 |\n",
       "\n"
      ],
      "text/plain": [
       "   coeGene amount_e_genes\n",
       "1  RPS14   2291          \n",
       "2  RPL21   2263          \n",
       "3  RPL32   2258          \n",
       "4  RPL13   2254          \n",
       "5  RPL31   2254          \n",
       "6  RPS27   2236          \n",
       "7  RPS3A   2222          \n",
       "8  RPS13   2218          \n",
       "9  RPS6    2205          \n",
       "10 RPS15A  2203          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(amount_egenes %>% arrange(-amount_e_genes),10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "f500d181-0792-4023-b3e8-e377e8e0b539",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Analyze only for egene/ co-egene pairs\n",
    "analysis = unique(sign_results[,c('gene1', 'gene2', 'eGene', 'coeGene')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "ce472cab-4518-4d3e-8fe1-1b82d8968d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "amount_egenes = analysis %>% group_by(coeGene) %>% summarise(amount_e_genes = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "8611c64e-68b1-4bad-88af-c84726a6841e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdfWCN9f/H8fc2s83ORrlJ7TaTVXbDd27ShLnJCPlmGxIy+ur2a6nEFF8rkyhU\nyveLFVIrW4l+i7RbM8amdBAyITMzZWa3bPb7Y9iN2c6OnXN2XZ6Pv+xcn3Ou17n2cc5r17mu\n65iVl5cLAAAAlM/c1AEAAADQOCh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsA\nAACVoNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACV\noNgBAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgB\nAACoBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFDsAAACVoNgBAACo\nBMUOAABAJSh2AAAAKkGxAwAAUAmKHQAAgEpQ7AAAAFSCYgcAAKASFLumL/nF9mbXdXvnqKnz\nQA/5Bza8NWmwj9td9jaW5td+mb0/OmPqXIAetk1taWZmNuyzYlMHAXAjip1xFH82zKyKgKgb\nh5xb0a/KCM3T3zdugv3/8ah8dNdX0xr30VGX4l1vPtItaM5nP+49dvZicWm5qfMoHpPZxMrL\nmcR6Y/bC0Ch2gIHlfP7Gwl/YtwH1oNgBTVgzUwdAvXp/dKb8I1OHgP7Sdu68fP0Hl+BvEpYO\nc7GzNDNhIqDhLmftily15rvYVO2x4yfzRSTm2XtcFnW6z/3Brr36DRjQ/5EujrbsKQBMj2IH\nGFZZXl5h5U9dBg13teO/HZSl/PTGF4dO+HjfxWo3lpw/eTD15MHU2G8/XSytp247t2KgqQIC\nqFQOYyj69LGqW33UhhuH5HzSt8oI24mbry3Y/sJdlbf7LPi92r2unN/39YLnn+jjde9d9i2a\nN2uuaet0n1cPv5HBry789Id9Z4rLy8u/HFXPHHjs04tVH7LweNyKWU8/1tPdobW9dTPLFq3a\nuXr2DXg+/Kufz5XW+uRKs5OW/3t4t3tb2za3sm/bofvwF5f+dOpS+bapravm/uOmT6ck4/8W\nPN3vQYc7rMyl9QvxV8dd+vvI9uhP5r0QNKD7A05t7KwtLZpZa+64694ufUf+a95nO06VVA8R\nX31tv1/5e8+qV0b1dGursbJudc/9/SaEbTxSVBH3dOLyfz/e/d52muZWmrYde4567dNfzuv6\ni2zQJtrzultd273vhzl1rqWBW8AgG6EBk+Hip4OrTvEvL1ddWG0SPvZp0dWbb5gMRX/ELHru\nMZ9722iaN29xh0vXIc8vSzpTdpPH0WEy66I4c0fEnGce933QsY29taVli1btOnQdNO615T8d\nL6xldMNmu35rafBm0fO56PDqUeH3932t6tnw0nrqtqqPfRvM3gb/pkw/e3Wk1yRX9AxXGYqd\ncRim2JX9vnqEQ10ffvRc9GfDXk3+Sl74mONN9yfZdX32y99r/F8o3PfeoLY3fqxo333G7ABd\nit2BX5YOaFN5/+vF7vwnA+oMbefx/MbMKjmqvSt4TX7zn44WNe+i8ZmXmp0S1vuOG+LaeLy6\nPV/HX2UDNtGtFbuGboFG3wgNnAzV3xoDIqu9depS7LyfXTj5QZsbVmTh/FR0dq2PU5sGvjVe\nSPsw0O1mnaW56xMf7MmrOrzBs12vtTR8s+izFt1ePcrLy8vLf5nZ8dqNZveMeD/u96+fbiki\nQ/975vTRfTtjo/8779nHe/q8Glfl0W+H2dvg35SJZ6+O9JnkSp/hakOxMw6DFLtza4Zb1zFr\npaHFrij97Yds6xlr5jR2Q1Zl6Asxwc713ONa7j9qezoeTwQ+UO3VW+diJyI2fT4+cT1JtXeF\nmzFv1cq+9iVmD775iy6/xwZtIsMWO6mxBRp3IzR8MlR/awz8qtqf27oUu5tr+8zW4loepzYN\neWu8pH2v7002xXV3DPz48LVnosds12Mt+mwWPdai46tHeXl5+eXIkdff552n7y4vL982taXU\nqDg13A6zt8G/KZPOXh3pMcmVP8NVh2JnHDWKXf10KXYbn7r+V07zf0yL/vXUhaJLxblZfxxM\nT/7+s4UhY/s/OPzDU9cGa+d2rnwUl1f23JjxyEKfKh3LuvOkiJ0nL5YU/3X4h7cfbV8lWpuJ\nm6/9XbT/Pw9UWWB1/4TVu/7MLy48vXf9VO8W1Z7PTYqdiJjd1e+Nr9NPXSzMPZH+bfiU91Iq\nhuVGjPMe9tz8tVtSfv7tj6zzRZeK8/7689et/33uH3aV971vVvq19DXeFcw7jF6x+3RB/tm0\npUPaVt+0LR95I+bohcLzhzZMdKvyd2mn0F/r/TXqsYnKyy9Xezl/vPqHPHVo6BZo1I2gxzOt\n/tYY9PWVqk9Gx2LXqter32jPFlw4tXPFGOcqAVoFx1R9uPonsy6OLXnIsvIpek786KeDWXlF\nhX8dS1492avyLcF28KdnysvL9ZvtDV+Lfpul4WtpwKtHWXTg9ZWad3z6iwO5W+otdrfD7NXr\nN2Wq2asjPSa5Cma46lDsjMMQxa7siyeuv6i1HvfN31dqX/VV9b6apL52b5X13z93X5XHy/36\niSovx82GRlwoLy+vuUeqw+t7qvxJdvBtz6p7829W7Mzue3Xnzd8dapP073sq79576bX/xdXf\nFVqN23ztiIsrW6ZU/TvPrNs7x6491B+LelZZMDKyvsqlxyYq17/YNXQLNOZG0OeZVn9rHF19\np7ROxc5qwMrKXSgXVg6u8inKQ4urvgQ3yltj1dlr3nPR8arLrqS9WmUD9F56qlzP2d7gtei3\nWRq+lga9epxc/FC1T7TMLSzMRMR54Mtvf7R+S/rpwrpfe6pR0ewt128Cm2T26vGwOk5yVcxw\nteHsdOUyd3e/fuTLX+tHOTp69R4SFDztzXdXRG7d+2f+lYY92p+JiX9U/tR1/ASvKv+JWz4x\ncaTm+k+lSUk7RUTOpaZmVI55YNyEblWm0wNPT6zykntTDz8//aGb7S8vOrbtv29MHvFIl453\n32Fn09yi4oqefT44XTnk9OnTtd3TdljgkGt/rZk5OTlUWdRj9OjrLwLOrq6VicsvXMirO6se\nm+gW6b8F5JY2gvGfqYiI1bAp4yv3p9h36FDlff7ChQuNtJZrMhMTK2fvldTXXKteQdy82+Iq\nG2B3aqq+s72ha6mFDptFj7U06NXDaUrYVNcq+1GulJWVi8jJn5a88eI4fx+HNh0Gvrhyb26N\n4LfX7BWjTuBGmFe10WOSq2OGqw3XXTCJURvKowJq3HZuRb+2zyU25FH+MXXGgI+fia14wSgv\nPK3dcVq7Y8vVhdYOvcbMeG/Rv3u10e3BsrKyKn8wd3WtfpyFhaurk8hvV3/KP336oohd9pmq\n34jl6upa7S733Htvc9l1qe61OvfseXetC0p//3Ts4KlRf1yudel1hYWFtd18j5NT5duQjU2V\n43IrnshV5hZVj++7cqWe/+t6bKK6H7BOt7YF5JY2gnGf6TWOHTtWPTDa2rpK46/3l9NgmZmZ\nug69lJNzQfSb7Q1di7S84XYdNos+a2nQq0fLQR8m/+Dw75cXf3sg98ZLE5cXHo9d/q+Hk44l\n7V7QoyLd7Td7xagTuBHmVW30mOQqmeEqwx47JXOZ8m3y2mmDO9nV8mssztz52bQBj71/SMdL\nxJdXu5a8Wc2TosprudR8zbuY1Vxc/1pbt679eOmjyybU/65wk1giYm9f5aObasE0rVpVeSco\nK2vIq60em0h/t7gF5JY2QiM80+pvZOfPn6/3HtKqVauqP5qbN5lXp8uXLzfGbNdhLbVo5M1y\nfS0Ne/WwcBg0O3r/6ZOpG1fMn/50X5cbdwmUaBe/sTa74t+34eyVJjuBbzKvamPgSd6kZ7iq\nNI2ZB33ZeYxfuuXwubOHkjev/+TdN1+eMtq/m1PlX7dFuz+K2KPbI91zT5VDX6788cfJakvL\nTpw4VfmT5u677UTk7rur7m07ebL6XU4fP17P7jq56f/hE9GRu66/BJjd89jCmF8zc4sqrkBQ\nz6mmBqTHJtKbabeAXs+0+u+y+r6YjIxjjZ/y1lSbvTaBX9V1SauEF9voOdsbuhbjPJerGv7q\nYePY4/Gpoe99+oa/rYgMePfn5K9m9293bWlpcnLFp2DMXkMz0LzSY5KrbIarBMVODZq3dvcd\n9uSzr4W9vzLyhz0ZW15yub7oeEZGxd+ezZpV+Ru7pKSk5mM49e1b5QDUn9et/bXKnzIXvlnz\nXf71n5o90udhEZHWDz3UsXLM/sgvtFXucuizNbof2lFTtR3vLk++OWOI5z0trS1EpOz3LT+Z\n6lVWj02kN9NuAb2eqY2dXZU9KRkZlcfElMSuXn+8MfPVN5l14dSvX4frPxRtXbXuZN1/u+s3\n2xu6Fv3c0lp0efWonXXb+32D3n5r3PX33OKCghs/OGP21mCK2asjPSa5qme4YlHsFOzgu4O6\n+j89a+mX23b+cuiPrPOFly8X5Z5M/yYyqfJACU2rVhW/4zZtqvyxlP3TlzF/5F2qNqV7PPNs\nlWsEHFo4bspnqafyL5WcP7J1/uh/f1N5YkGbcc8HVnxM4jN+QpXzuw4tHv/c+vSswstFZ37+\n8rmx4Vr9/4O3bVvl+gYnv/1wTfrpguL8Mwe3vjf6sbB0k+0812MT6cvEW0CfZ2r24INVrpRw\naPlLb207dr7o4vHED8ZMXln10J1bV+9k1kX3Z57rfv0dNm/bSwOfeHtDakZWXsmlgr9OHzv4\nc8rWLz94Y+oTvd1H/i9bRN/Z3tC16EePtTTg1aMsetJ9vqOen7fim4S9B/84k1tydVuXX84/\n/cv6VTHXg7t06GAuwuytm0lmr470mORqmOHqU8ceTTQeg1yguNpp87VrNf67a1e83BvaofYx\n18MUpb31UIvax1xX8wLFP0y+pQsU1/yGtGv2zel846XPa3HXC9uv3qPm9xFVPtYfi3wql7Ss\n9r1HG0ZXeay+n9T9DV96biI9L3fS8C3QuBtBn2d6IMxLp8x1fKVYFTtfqfzDWtxna6suq38y\n66L453d8NbU/TlW+H159knrMdj3Wot9mafBaGvDqcfnLx6svqLbP6ToLz7e0V6+OcXvMXv1+\nU6aavTrSY5Irf4arjjrbKkREpIXXi+sXj7j2f6Hrs7MGt6pzvLXPG//34zt1fQ9Pl6lfxH0a\nUOUSn/b+H2x6b2Ct3z/zWuioOyt/vuE43Lp5zVg9u1st/4st3Z6a+/T9DXmkxqXHJtKTqbeA\nPs/0wVdWvO594/up+T1Dl7zu36jp6p/MurDq8vr3sUsDbvolRTXpN9sbuhb9NPpaqr96VFda\nWlrzJpv7xn/6zUyPq28ozN66mGb26kiPSa7CGa50XO5EwTrPjNP2TUiIj09I+eXoqTPZZ8+e\ny7tsadvqLuf7PHsOGPn01PG9HSov1i1OU77d1eqdtz/+JvGXY2cuFF6ube//nb6vf394dNza\n/63bHJvyy9HTf+UVm7e4o63zA90fGRz4zL8Curap+fWNLbynb9F2XzH/3U83pfx2usCilcP9\nPYeMC5n1fM8tIxZXDqtx4lO9bHu+lZTe9d23P4jctvfYuUvNW97VoavfP6e8/urossUe8xr0\nUI2s4ZtIP6bfAnpMhl4LknY9GP7WJ9GJ2pPny1q07dDV74nJr73ypPcPgS83ajZdJrMuWvWY\ntuFgUOrXq9dsjEtJP3gi53z+ZUv71m3btm1zl/OD3Xz79OnT55Hu19//9ZztDVyLUZ5LA149\nmgWuybg3KSEhPmHHL7+fzDydlXUm5+KlcjG3atnOwbWT90P9hz8V/GRvpypvuczeupho9upI\nn0mu8BmuOmbljXqNBqBC+Z8R/h6Tf7x2OMvdLyWe/qCPSRMBhnLbzfYtUzRDVhc89mnR90/X\n83WcUIvbbpIrGR/F4tbsnd//kbGvfxiduPfQyZyCy2WXLmb/vvObRU/6v/Rj5UHKTmPH+Jow\nI9A4mO0VGnhkBZSESa58fBSLW3Pl4rHkyPjkyHdvOsKiwzMr3ny4cT6fBEyJ2V6BYqdiTHLl\nY48dDKnZ3f1mfJu4YuitHyoMNHW3z2yn2N22bp9JrmQcY4dbU15wYteWjRu/+yFp37HM01nZ\n5y81s7W/o/29D3p37/NY0LiAvvfamjoibitRAWaB0fUN6jxXu/8/Hg1+bGY7DMtAs7chD8sk\nVz4+isWtMbN16TVqWq9R00wdBDA4ZjtUj0mufOyxAwAAUAmOsQMAAFAJih0AAIBKUOwAAABU\ngmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIH\nAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACg\nEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7\nAAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAA\nlWhm6gANV3giIWp9VExSuvbwiZzci4Wlli3sWrZzcffw6TM0YFxgPxcbUycEAAAwBbPy8nJT\nZ2iAnPiwsePnx2ZeutmA5g4DZq+LnOPXxpipAAAAmgIlFbtSbXjP7rP3loimo/+E4IBBPT3d\nHFrbWzcrLc77KzNDm7otKmLNlqMFYuUTnrZrlocCd0YCAADcAgUVu8KvAtuNiSpoP3L1jsjg\nDla1DSnJiBjTe/LGM7YBkWc3jG5h7IQAAACmpKCTJ/YkJBSIdAlZfJNWJyJWbsGLpnmLFCQm\nphk1GwAAgOkpqNjl5eWJiKOjY52jKpZXjAUAALidKKjYOTk5icie7duL6xhUnJycJiLOzs5G\nSgUAANBUKKjYeQcGdTKT7JXBQctSsktrGVCanbIsKHhltpi5B47yMno+AAAA01LQyRMi+Ttm\n9h68cF+BiFU7T9++PTzcHNrYWVmUlVw8l5mxf3fiDu3ZEhHbLjO3Ji/wtTV1WgAAAONSVLET\nkbxfP50dMmdV/KlaP4+1dvSbErZ0/iQve2PnAgAAMDmlFTsREbmSfzw1PilNe+Tk2dz8ojIL\nG02rds6dPLv19evpolHQh8sAAACNSZHFDgAAADdi/xYAAIBKKPB7twpPJEStj4pJStcePpGT\ne7Gw1LKFXct2Lu4ePn2GBowL7OdiY+qEAAAApqCwj2Jz4sPGjp8fm3npZgOaOwyYvS5yjl8b\nY6YCAABoCpRU7Eq14T27z95bIpqO/hOCAwb19HRzaG1v3ay0OO+vzAxt6raoiDVbjhaIlU94\n2q5ZHgrcGQkAAHALFFTsCr8KbDcmqqD9yNU7Im/ydbElGRFjek/eeMY2IPLshtEtjJ0QAADA\nlBR08sSehIQCkS4hi2/S6kTEyi140TRvkYLExDSjZgMAADA9BX1emZeXJyKOjo51jnJ0dBTZ\nVzEWABpVWVlZTExMcXFd31jdWKytrYcOHWphYWGEdQFQDQUVOycnJ5GMPdu3F48baH2zQcXJ\nyWki4uzsbMRkAG4TP/zww4gRI4y2us2bNw8bNsxoqwOgAgoqdt6BQZ3eXXBkZXDQA5ErX3j4\nrhuil2anLH8meGW2mLkHjvIyRUQA6lZUVCQiEyZM8Pb2NuiK9u3bt3bt2orVAYDuFFTszHxC\nI2bEDF64b3OIr0u4p2/fHh5uDm3srCzKSi6ey8zYvztxh/ZsiYhtl5mrQ31MnRaAanl7ez/6\n6KOmTgEAtVBQsRPR+L6TlOI+O2TOqvhT2rgN2riaA6wd/aaELZ0/ycvWFPEAAABMSlHFTkTs\nvSZ9GDdxwfHU+KQ07ZGTZ3Pzi8osbDSt2jl38uzW16+ni6Zh5/nqcij0lStXDh8+PHPmzObN\nm99aegAAAANSWrETERFzjWuv4a69hjfCQ8XHx+t4KLS5ufkbb7zRCKsEAAAwDEUWu0bk5+e3\nadOmuvfY7dy5c8mSJffdd5/RUgEAAOhBgcWu8ERC1PqomKR07eETObkXC0stW9i1bOfi7uHT\nZ2jAuMB+LjYNeDALC4vhw+vf9bdkyRJzcwVdzBkAANyOFFbscuLDxo6fH5t5qdqtBfnnc7KO\nH9i19avl80IHzF4XOcevjYkCAgAAmIySil2pNtx/yNy9JaLp6D8hOGBQT083h9b21s1Ki/P+\nyszQpm6Liliz5Wjs3CH+lmm7Znko6akBAADcOgW1n8LosPC9JdJ+5OodkTW/LtbN3bNH/5GT\np0+PGNN78sb0+fOip20Y3cJEQQEAAExCQceN7UlIKBDpErK4ZqurZOUWvGiat0hBYmKaUbMB\nAACYnoKKXV5enog4OjrWOapiecVYAACA24mCip2Tk5OI7Nm+va5LkxQnJ6eJiLOzs5FSAQAA\nNBUKKnbegUGdzCR7ZXDQspTs0loGlGanLAsKXpktZu6Bo7yMng8AAMC0FHTyhJlPaMSMmMEL\n920O8XUJ9/Tt28PDzaGNnZVFWcnFc5kZ+3cn7tCeLRGx7TJzdaiPqdMCAAAYm4KKnYjG952k\nFPfZIXNWxZ/Sxm3QxtUcYO3oNyVs6fxJXramiAcAAGBSiip2ImLvNenDuIkLjqfGJ6Vpj5w8\nm5tfVGZho2nVzrmTZ7e+fj1dNAr6cBkAAKAxKa3YiYiIuca113DXXvV/ExgAAMBtRDX7ty4c\n/+WXX345fsHUOQAAAExFNcVu26tdu3bt+uo2U+cAAAAwFdUUOwAAgNudgopdVIBZHQKjRUSi\nA6/+GBBl6rgAAABGpqBiBwAAgLooqNi1d3FuLuZte4d88Wv2+RusGSEiMmLN1R/XPm7quAAA\nAEamoGLX+70DP6974b5DHzz5sN8L6367bN+qqhaWIiKWLar9CAAAcBtRULET0Tz41AfJv6V8\nNNJ88797P+D74tr9F00dCQAAoMlQVLETETFr0/OFdXsPbpnne2bVxH90fnTO93+UmDoTAABA\nU6C4YiciIpaOg9/4bv++r150/TV8uEeX0e8lZ5eZOhMAAICJKbPYiYiIrXvQ+4m/7Voxxmrr\nq30eeC7W1HkAAABMS8HFTkTE7I5uU1enHYwL79fuipWVlVVzC1MnAgAAMJVmpg7QCJrd4zfz\nm0MzTR0DAADAtBS+xw4AAADXUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAA\nACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpB\nsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMA\nAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJ\nih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0A\nAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBK\nUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwA\nAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABU\ngmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIH\nAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACg\nEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEs1M\nHaDhCk8kRK2PiklK1x4+kZN7sbDUsoVdy3Yu7h4+fYYGjAvs52Jj6oQAAACmoLBilxMfNnb8\n/NjMS9VuLcg/n5N1/MCurV8tnxc6YPa6yDl+bUwUEAAAwGSUVOxKteH+Q+buLRFNR/8JwQGD\nenq6ObS2t25WWpz3V2aGNnVbVMSaLUdj5w7xt0zbNctDSU8NAADg1imo/RRGh4XvLZH2I1fv\niAzuYFVtmZu7Z4/+IydPnx4xpvfkjenz50VP2zC6hYmCAgAAmISCTp7Yk5BQINIlZHHNVlfJ\nyi140TRvkYLExDSjZgMAADA9BRW7vLw8EXF0dKxzVMXyirEAAAC3EwUVOycnJxHZs317cR2D\nipOT00TE2dnZSKkAAACaCgUVO+/AoE5mkr0yOGhZSnZpLQNKs1OWBQWvzBYz98BRXkbPBwAA\nYFoKOnnCzCc0YkbM4IX7Nof4uoR7+vbt4eHm0MbOyqKs5OK5zIz9uxN3aM+WiNh2mbk61MfU\naQEAAIxNQcVOROP7TlKK++yQOaviT2njNmjjag6wdvSbErZ0/iQvW1PEAwAAMClFFTsRsfea\n9GHcxAXHU+OT0rRHTp7NzS8qs7DRtGrn3MmzW1+/ni4aBX24DAAA0JiUVuxERMRc49pruGuv\n4abOAQAA0JSwfwsAAEAlFLjHrvBEQtT6qJikdO3hEzm5FwtLLVvYtWzn4u7h02dowLjAfi42\npk4IAABgCgordjnxYWPHz4/NvFTt1oL88zlZxw/s2vrV8nmhA2avi5zj18ZEAQEAAExGScWu\nVBvuP2Tu3hLRdPSfEBwwqKenm0Nre+tmpcV5f2VmaFO3RUWs2XI0du4Qf8u0XbM8lPTUAAAA\nbp2C2k9hdFj43hJpP3L1jsiaXxfr5u7Zo//IydOnR4zpPXlj+vx50dM2jG5hoqAAAAAmoaCT\nJ/YkJBSIdAlZXLPVVbJyC140zVukIDExzajZAAAATE9BxS4vL09EHB0d62AgahYAACAASURB\nVBxVsbxiLAAAwO1EQcXOyclJRPZs315cx6Di5OQ0EXF2djZSKgAAgKZCQcXOOzCok5lkrwwO\nWpaSXVrLgNLslGVBwSuzxcw9cJSX0fMBAACYloJOnjDzCY2YETN44b7NIb4u4Z6+fXt4uDm0\nsbOyKCu5eC4zY//uxB3asyUitl1mrg71MXVaAAAAY1NQsRPR+L6TlOI+O2TOqvhT2rgN2ria\nA6wd/aaELZ0/ycvWFPEAAABMSlHFTkTsvSZ9GDdxwfHU+KQ07ZGTZ3Pzi8osbDSt2jl38uzW\n16+ni6ZhHy6XlZXFxMQUF9d13N7OnTtF5MqVK7eUHAAAwMCUVuxERMRc49pruGuv4Y3wUPHx\n8SNGjNBl5IEDBxphfQAAAAajyGLXiPz8/DZt2lTvHrslS5Z07tzZaKkAAAD0oMBiV3giIWp9\nVExSuvbwiZzci4Wlli3sWrZzcffw6TM0YFxgPxebBjyYhYXF8OH17/pbsmSJubmCziAGAAC3\nI4UVu5z4sLHj58dmXqp2a0H++Zys4wd2bf1q+bzQAbPXRc7xa2OigAAAACajpGJXqg33HzJ3\nb4loOvpPCA4Y1NPTzaG1vXWz0uK8vzIztKnboiLWbDkaO3eIv2XarlkeSnpqAAAAt05B7acw\nOix8b4m0H7l6R2TNr4t1c/fs0X/k5OnTI8b0nrwxff686GkbRrcwUVAAAACTUNBxY3sSEgpE\nuoQsrtnqKlm5BS+a5i1SkJiYZtRsAAAApqegYpeXlycijo6OdY6qWF4xFgAA4HaioGLn5OQk\nInu2b6/r0iTFyclpIuLs7GykVAAAAE2Fgoqdd2BQJzPJXhkctCwlu7SWAaXZKcuCgldmi5l7\n4Cgvo+cDAAAwLQWdPGHmExoxI2bwwn2bQ3xdwj19+/bwcHNoY2dlUVZy8Vxmxv7diTu0Z0tE\nbLvMXB3qY+q0AAAAxqagYiei8X0nKcV9dsicVfGntHEbtHE1B1g7+k0JWzp/kpetKeIBAACY\nlKKKnYjYe036MG7iguOp8Ulp2iMnz+bmF5VZ2GhatXPu5Nmtr19PF42CPlwGAABoTEordiIi\nYq5x7TXctVf93wQGAABwG2H/FgAAgEoob49dcda+lJ8zzdp7dv+Hk0ZERMrO7ly7YsOO38+Z\n3/Wg35jg0d3aUVcBAMBtSFnFrmD3oidGhv6YVSoize7u//am71/v8ueKIb2e/+nv8ooRK99f\nvHrxT99P72Jt0qAAAADGp6R9W2V7wsbO+DGr1KJlh64+99nnxM0a85//W/XitJ/+tvUePWvx\n8mVvjPHUlP8V+9rY8J/LTB0WAADA2BS0x65w4/srjok4Pvl12udP3GX21/eTfIavmjTnwiXn\nZzYn/+9RjYg8P6GLdAyIPPTJJ7Fz//eohakTAwAAGJOC9tgdTU/PE+n49Iwn7jITkdbDpk/s\n9HdOTpnb+Ocf1Vwd02rkc0+2FzmXnHzYlFEBAABMQEHF7tSpUyLSsWPHazdc/WeVW0Qs7r//\nPhE5efKkseMBAACYmIKKnb29vYhcunTp2g02NjYiIhqNpsqoli1bikhZGQfZAQCA242CjrFz\nc3MTyf7jjz9E7qi4pW3nvn3PSee2VUdlZmaKiLOzswkSAgAAmJKCit3dQ/y9pqf8Ghd3XP7h\nKiIifeclJNQYVHrkyB8iLby9O9a8OwAAgMop6KNYuX/0Uw+1bHl4U/Sxmw4p/v7z6FzRjBw7\nzMaIwQAAAJoCBe2xE+n02s7c1+occb71oAWf9HboO7SFkSIBgAFcuXJFRHbu3Gmc1VlbWw8d\nOtTCgotEAYqnqGJXv7sfmfjsI6YOAQC36NChQyKyZMmSJUuWGGeNmzdvHjZsmHHWBcBwVFbs\nAEANLl++LCITJkzw9vY29Lr27du3du3aoqIiQ68IgBFQ7ACgifL29n700UdNnQKAkijp5AkA\nAADUgWIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABU\ngmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIH\nAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACg\nEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7\nAAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAA\nlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDY\nAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABUQsdil7zspQ+3Hs03bBYAAADc\nAh2L3ZntH/3b392x8+Ovrkw4UWzYSAAAANCHjsWuz8sfvzLiQYvDm977l5+bU9egN9fuPHPJ\nsMkAAADQIDoWu3a+zy3+TnvqWGLErKCu5gc2vD3xYReXhya+/dUv50oNGxAAAAC6adDJEzbO\nfSaFf7Xnz5Opn4dN7Gaxb+2bY7o6dfD716JNB3OvGCohAAAAdKLPWbHN2/cY9+ZnO06c+iV6\nQYBzTsLKGY93dnR/9NlF3+6/QL8DAAAwEX0vd3I5Jz36fx8s++T/jhSLWLS5zyEv4b8znvC6\nt/vzG09S7gAAAEygwcWu5PSudf8Z/5CLU7fRoRE7Lt73zxmrEv7488jhk7//MOuRO87v/WT8\nC+vOGSIpAAAA6qR7sSs6kRgxK8jH0aXXhHmf7ynsMGL6J7FHT+37ZuHkvk7WIlYu/uHr3+gh\nkp+YuNeAgQEAAFC7ZroNS3mz87Dwg+eviJm9+5BpL03799ODO9jeMMrJyUlk98W8vEYOCQAA\ngPrpWOxO//bbpQ6PPv/StGmThnSyM7vZsIde/vLLAHF6qLHSAQAAQGc6Fjvf+Qcy73ugZX0f\n3Dr6jhlzy5EAAACgDx2PsbvbvVPR/oSEhOTD56vdfv5wckJCwq9nygwQDQAAAA2h68kT5796\n3tfP74mPD1hWu9ly//In/Px8X4rKbfxoAAAAaAgdi93Z9f/7Ll/ajZs6UlPtds0/nx3XTvK/\n+d+XXOIEAADAtHQsdr8dOHBFpLOnZ83x5h4eD4pcOXjwcKNHAwAAQEPoWOwuXrxY5/Lz58/X\nuRwAAACGpmOxc3R0FJHf9u+v+XVhZQcOHBKRe+65p5GDAQAAoGF0LHaeAwe2EzkT8fbKzPIq\nN5dnrnw74oxIm4EDvQ0SDwAAALrSsdhZDHxl1kM2UrDtpT6Pv7l2254Dvx3Ys23tmyP6/Htb\ngdj0DH11kIVhcwIAAKAeOl6gWOS+aRu+PDR47H8Pbn574ua3K29v0fnZL6JCOt30yygAAABg\nHDoXOzFzfHxF+qGgTz9e+38pB/48f6n5nc4eDz824blJfs5WBgwIAAAA3ehe7ERErJ37P/dO\n/+cMlAUAAAC3QNdvngAAAEAT16A9dkUnd275cdfBP8/ll5SVV1/k+dQ74zwaMxgAAAAaRudi\ndzF1wch/vhmXVVb74lHdKHYAAAAmpWOxu5wYGhgalyXW9w791zNDvBxbWdU4DdalR+NnAwAA\nQAPoWOzSvvnmTxH7f67e9c2T7QybCAAAAHrR8eSJnJwcEenx2GO0OgAAgCaqQd8Ve+nSJYOG\nAQAAgP50LHZdx4x7wFx+iYs7b9g4AAAA0JeOxc7sH7PXL+hjGf3qxOVpdDsAAICmSMeTJ+Je\n7z0j9pKNzanNL3a/a7bzfR3a2lSvhAPeTVvY3wD5alF4IiFqfVRMUrr28Imc3IuFpZYt7Fq2\nc3H38OkzNGBcYD8XG+PkAAAAaFp0LHZ/Z6Snp1/99+ULJw/+fLLGANe/GzPVTeXEh40dPz82\ns/qxfgX553Oyjh/YtfWr5fNCB8xeFznHr41R4gAAADQhOha74auysj6qa4D1HY2Rpm6l2nD/\nIXP3loimo/+E4IBBPT3dHFrbWzcrLc77KzNDm7otKmLNlqOxc4f4W6btmuXRsK/BBQAAUDod\n249Vq/btDRukfoXRYeF7S6T9yNU7IoM7WFVb5ubu2aP/yMnTp0eM6T15Y/r8edHTNoxuYaKg\nAAAAJqHjyRNNwZ6EhAKRLiGLa7a6SlZuwYumeYsUJCamGTUbAACA6TWo2BUc+X7py08N83u4\nR/dur8eJiJxJ/frzzz+PzbhsmHTV5OXlybUr6t1cxfKKsQAAALcT3Q9EO/Pdv/xGrzxUcvVH\n579FRGyPfjl1/MZmE1pmrxlubYh8VTg5OYlk7Nm+vXjcwJuuqzg5OU1EnJ2dDZwGAACgqdF1\nj13OuqlPrTxU6jp6RfL7QypvtvvnUyNaSN63X8aU3Py+jcQ7MKiTmWSvDA5alpJdWsuA0uyU\nZUHBK7PFzD1wlJfB8wAAADQtOu6xy1r/3035Yv/k+6un+m75qcqCFh4eHUT2p6Ro5YluBkl4\nnZlPaMSMmMEL920O8XUJ9/Tt28PDzaGNnZVFWcnFc5kZ+3cn7tCeLRGx7TJzdaiPYbMAAAA0\nPToWu1/37RORf/TpY1tziZOTk8j+rKysRg5WG43vO0kp7rND5qyKP6WN26CNqznA2tFvStjS\n+ZO8bogJAACgejoWu0uXLomIRqMRETMzM5Hya0vy8/NFxNLS0iDxbmDvNenDuIkLjqfGJ6Vp\nj5w8m5tfVGZho2nVzrmTZ7e+fj1dNAo6zxcAAKAx6Vjs7r77bpETp06dEnGrtqDs4MEjIuLq\n6tro0W7OXOPaa7hrr+FGXCUAAECTp+P+rS4DB7YW2fftN8fKK/bYVSg/u+79ddkiLoMHuxsq\nIQAAAHSiY7FrNvi10B425alz//nM/3bliIjk/R77xVujHv7X5jyxH/hmSC+z+h6i0RSeSFgb\n/uIY/16d772n3R12Grs7291zr0cv/zEvhq9NOFFktBwAAABNi67XsTNzfzk6MmPokx+vnvqr\niIhsCx24TUTE1jvk688nOxkqXw058WFjx8+PzbxU7daC/PM5WccP7Nr61fJ5oQNmr4uc49fG\nSIEAAACaDN0vUGzmOGJ52uExa1es+T5Ze+JcsUUrh86+wye9OLmfY3MDBqyiVBvuP2Tu3hLR\ndPSfEBwwqKenm0Nre+tmpcV5f2VmaFO3RUWs2XI0du4Qf8u0XbM8dH9qAAAAatCw9tPc4ZEp\nbz0yxUBZ6lEYHRa+t0Taj1y9I7Lm18W6uXv26D9y8vTpEWN6T96YPn9e9LQNo1uYJicAAIBp\nKOjiIHsSEgpEuoQsrtnqKlm5BS+a5i1SkJiYZtRsAAAApqfjHrtj21b8mFHXALdHnx3UoTEC\n3VxeXp6IODo61jnK0dFRZF/FWAAAgNuJjsVu73+fey66rgGjNhi82Dk5OYlk7Nm+vXjcQOub\nDSpOTk4TEWdnZ8OGAQAAaHJ0LHYeY996q0u1W64U5WSkxkTFHjXrMi5k1P1dPQwQrjrvwKBO\n7y44sjI46IHIlS88fNcN0UuzU5Y/E7wyW8zcA0d5GTwPAABA06Jjsbt/1Btv1HLzOzMXDe0+\nIyb1+bCw+xs1Vm3MfEIjZsQMXrhvc4ivS7inb98eHm4ObeysLMpKLp7LzNi/O3GH9myJiG2X\nmatDfQweBwAAoIm5tWuC2Dzwyvx/feC75JXQb8ZFBhj8LFSN7ztJKe6zQ+asij+ljdugjas5\nwNrRb0rY0vmTvGwNHQUAAKDJudWLvZl7eXUW2frTT7sloF9jBKqHvdekD+MmLjieGp+Upj1y\n8mxuflGZhY2mVTvnTp7d+vr1dNE07DzfsrKymJiY4uLiOsbs3LlTRK5cuXJLyQEAAAzslq/i\ne+HCBRHJz89vhDC6Mte49hru2mt4IzxUfHz8iBEjdBl54MCBRlgfAACAwdxasbvyd+zC/+4W\nkU6dOjVOHmPz8/PbtGlTvXvslixZ0rlzZ6OlAgAA0IOOxW7XewGLd1a/qfzShcwDe3Yfu1Au\n9kNeDjZesSs8kRC1PiomKV17+ERO7sXCUssWdi3bubh7+PQZGjAusJ+LTQMezMLCYvjw+nf9\nLVmyxNxcQRdzBgAAtyMdi92pndHRtV7Hrnn7HuPeXL50Ut1XDW40OfFhY8fPj828VO3Wgvzz\nOVnHD+za+tXyeaEDZq+LnOPXxjh5AAAAmg4di13/BTt3vlr9JjOL5ratnTre29barPFj1apU\nG+4/ZO7eEtF09J8QHDCop6ebQ2t762alxXl/ZWZoU7dFRazZcjR27hB/y7Rdszxu+fBBAAAA\nRdGx/dx530MP3WfYJPUqjA4L31si7Ueu3hFZ8+ti3dw9e/QfOXn69IgxvSdvTJ8/L3rahtEG\nv/wKAABAU6Kg48b2JCQUiHQJWVyz1VWycgteNM1bpCAxMc2o2QAAAExPxz12GVs++uGoro/Z\ncciL/m76Brq5vLw8EXF0rPtwPkdHR5F9FWMBAABuJzoWu59XvfRSrSdP1GbUBoMUOycnJ5GM\nPdu3F48baH2zQcXJyWki4uzs3PgBAAAAmjQdi53nUwsWuO9bvzRyf3Fr72GP9/FwtLt4Spu4\n8f9+/dvaY0zIOG+7KoPv9zRIVO/AoE7vLjiyMjjogciVLzx81w3RS7NTlj8TvDJbzNwDR3kZ\nJAMAAEDTpWOxcx/iXzY/bH9Z91k7ts5/6I6r58FeWbRj1qAB7363rXxW8iyv5oZLKSIiZj6h\nETNiBi/ctznE1yXc07dvDw83hzZ2VhZlJRfPZWbs3524Q3u2RMS2y8zVoT4GDgMAANDk6Fjs\ncr+Y/VZakUPIkreutzoRMb/Td/7SqZ/3+eCtNyKf3zShpYEyXqfxfScpxX12yJxV8ae0cRu0\ncTUHWDv6TQlbOn+Sl62howAAADQ5Oha79JSUEhEPb2+Lmvfv2tVTZOuOHXtlgl+jp7uRvdek\nD+MmLjieGp+Upj1y8mxuflGZhY2mVTvnTp7d+vr1dNEo6DxfAACAxqRjsav4MtXs7GwRTfUl\nZ86cEZGioqLGTlYHc41rr+Guver/JjAAAIDbiI77tzw8PUXk108/2VFc7fbC7cs/1YqIp6dH\no0cDAABAQ+hY7Fyenj6qlVw5/P5w30mLoxPSDv1+KC0hatHTvUcsO3JF7hj18tNGvLzI5bzT\nGUf/yCkqr3XpWe1PP/30k/as8fIAAAA0CboekdZuTMTmsH5tzc7v/ey1AL/uD3R6oLtf4Iw1\nP+eate331uaIMW0NmvK6i3s/Gdel/Z0OHe/r0O5Opz7Pr957w4WIk+YNGjRo0Lwk4wQCAABo\nMnQ8xk5E7Hu/GfvbiG9WR3wTl374dO5ly1b3uPv0fyJ4yhPedxrphIXMtWMGPh9zXkTMrDU2\nlzO3fzLl4Z/iV/3w2VNuuj8PAAAAlWpQITJv7R0wY1nADEOFqduV7e++EXNezF0D/vfNyold\nWxYd2fTuC88vWD/Rr6Q8KXK8a80TdgEAAG4vCro4yIEtW/4UuWPskojJXVs1EzO7To+/tWV3\n9NQHzkZN8pvw1Z9XTB0QAADApBpU7AqOfL/05aeG+T3co3u31+NERM6kfv3555/HZlw2TLpq\nTpw4ISJd+/at8vVlFg6Pr4iLeqbjqS+eGjDl26zaT6cAAAC4Lej+UeyZ7/7lN3rloZKrPzr/\nLSJie/TLqeM3NpvQMnvNcGtD5KvCyspKpMTcvGYXbTdsxY9rcns/9emYgdYbEz42cAoAAICm\nStc9djnrpj618lCp6+gVye8PqbzZ7p9PjWghed9+GVNy8/s2EhcXF7m23646c+cnP/9x2SC7\ng5+MevS1+FyDJwEAAGiKdCx2Wev/uylf7Ee/v3qqr3PVL2Jt4eHRQeRiSorWIPGqcuvVq63I\nsdTUc7UsbH7/S9/+MLenxS+LP441eBIAAICmSMdi9+u+fSLyjz59bGsucXJyEpGsrKzGzVUL\nC78nHr9DyuKjN56vdblt9//EfPdSZyuDBwEAAGiSdDzG7tKlSyKi0WhExMzMTOT6aQr5+fki\nYmlpaZB41TTrP3PDur5ZzTve9GPfO/sv+/GHB9amXpD7PQ2fRw9lZWUxMTEV37xrBNbW1kOH\nDrWw4EowAADcFnQsdnfffbfIiVOnTom4VVtQdvDgERFxdXVt9Gg3snQb8JRb3UPM7vF7bqaf\nEbLo54cffhgxYoQx17h58+Zhw4YZc40AAMBUdCx2XQYObP3e6n3ffnNsxmtV9tiVn133/rps\nEZfBg90NGFI9ioqKRGTChAne3t6GXte+ffvWrl1bsUYAAHA70LHYNRv8WmiPL15JnfvPZ1r6\n54iI5P0e+8Vby+e8tTlP7Ae+GdLLzJApVcbb2/vRRx81dQoAAKA2ul7Hzsz95ejIjKFPfrx6\n6q8iIrItdOA2ERFb75CvP5/sZKh8AAAA0JHuFyg2cxyxPO3wmLUr1nyfrD1xrtiilUNn3+GT\nXpzcz7G5AQMCAABANzoWu7TlT3+YeqXbS2tfmvLWI1MMGwkAAAD60PE6dqd3fLF23brEG7/0\nAQAAAE2EjsXO0dFRRM6cOWPQMAAAANCfjsWu6xNP3Cuyb9u2s4aNAwAAAH3pWOzMHvrP52/2\naLbplTFhW09eMmwkAAAA6EPHkyfiXu83I7ZE0yInfq6/64I7nTq4tLWp1gkHvJu2sL8hAgIA\nAEA3Oha7vzPS09Ov/ru8+O+TB/8+WX2A69+NGgsAAAANpWOxG74qK+ujugZY39EYaQAAAKA3\nHYudVav27Q0bBAAAALdGx5MnAAAA0NTVUey2vNSxY8eOL22pvOX8sbS0tLRj5w0fCwAAAA1V\nR7HLz8rIyMjIyq+8JXZG9+7du8+INXwsAAAANBQfxQIAAKgExQ4AAEAlKHYAAAAqQbEDAABQ\niXqvY1eYc/z48av/zimsect1tu1c27Zo1GQAAABokHqL3Q/P33tvfbeIyKgN5VEBjZYKAAAA\nDVZHsWtu27JlS10fx7Z5Y6QBAACA3uoodiPW5OYaLwgAAABuDSdPAAAAqATFDgAAQCUodgAA\nACpBsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpB\nsQMAAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMA\nAFAJih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJ\nih0AAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0A\nAIBKUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBK\nUOwAAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwA\nAABUgmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACpBsQMAAFAJih0AAIBKUOwAAABU\ngmIHAACgEhQ7AAAAlaDYAQAAqATFDgAAQCUodgAAACrRzNQBGq7wRELU+qiYpHTt4RM5uRcL\nSy1b2LVs5+Lu4dNnaMC4wH4uNqZOCAAAYAoKK3Y58WFjx8+PzbxU7daC/PM5WccP7Nr61fJ5\noQNmr4uc49fGRAEBAABMRknFrlQb7j9k7t4S0XT0nxAcMKinp5tDa3vrZqXFeX9lZmhTt0VF\nrNlyNHbuEH/LtF2zPJT01AAAAG6dgtpPYXRY+N4SaT9y9Y7I4A5W1Za5uXv26D9y8vTpEWN6\nT96YPn9e9LQNo1uYKCgAAIBJKOjkiT0JCQUiXUIW12x1lazcghdN8xYpSExMM2o2AAAA01NQ\nscvLyxMRR0fHOkdVLK8YCwAAcDtRULFzcnISkT3btxfXMag4OTlNRJydnY2UCgAAoKlQULHz\nDgzqZCbZK4ODlqVkl9YyoDQ7ZVlQ8MpsMXMPHOVl9HwAAACmpaCTJ8x8QiNmxAxeuG9ziK9L\nuKdv3x4ebg5t7KwsykounsvM2L87cYf2bImIbZeZq0N9TJ0WAADA2BRU7EQ0vu8kpbjPDpmz\nKv6UNm6DNq7mAGtHvylhS+dP8rI1RTwAAACTUlSxExF7r0kfxk1ccDw1PilNe+Tk2dz8ojIL\nG02rds6dPLv19evpolHQh8sAAACNSWnFTkREzDWuvYa79hpu6hwAAABNCfu3AAAAVEKBe+wK\nTyRErY+KSUrXHj6Rk3uxsNSyhV3Ldi7uHj59hgaMC+znYmPqhAAAAKagsGKXEx82dvz82MxL\n1W4tyD+fk3X8wK6tXy2fFzpg9rrIOX5tTBQQAADAZJRU7Eq14f5D5u4tEU1H/wnBAYN6ero5\ntLa3blZanPdXZoY2dVtUxJotR2PnDvG3TNs1y0NJTw0AAODWKaj9FEaHhe8tkfYjV++IrPl1\nsW7unj36j5w8fXrEmN6TN6bPnxc9bcPoFiYKCgAAYBIKOnliT0JCgUiXkMU1W10lK7fgRdO8\nRQoSE9OMmg0AAMD0FFTs8vLyRMTR0bHOURXLK8YCAADcThRU7JycnERkz/btxXUMKk5OThMR\nZ2dnI6UCAABoKhRU7LwDgzqZSfbK4KBlKdmltQwozU5ZFhS8MlvM3ANHeRk9HwAAgGkp6OQJ\nM5/QiBkxgxfu2xzi6xLu6du3h4ebQxs7K4uykovnMjP2707coT1bImLbZebqUB9TpwUAADA2\nBRU7EY3vO0kp7rND5qyKP6WN26CNqznA2tFvStjS+ZO8bE0RDwAAwKQUVexExN5r0odxExcc\nT41PStMeOXk2N7+ozMJG06qdcyfPbn39erpoGvbhcllZWUxMTHFxeaxrPQAAIABJREFUXcft\n7dy5U0SuXLlyS8kBAAAMTGnFTkREzDWuvYa79hreCA8VHx8/YsQIXUYeOHCgEdYHAABgMIos\ndo3Iz89v06ZN9e6xW7JkSefOnY2WCgAAQA8KLHaFJxKi1kfFJKVrD5/Iyb1YWGrZwq5lOxd3\nD58+QwPGBfZzsWnAg1lYWAwfXv+uvyVLlpibK+gMYgAAcDtSWLHLiQ8bO35+bOalarcW5J/P\nyTp+YNfWr5bPCx0we13kHL82JgoIAABgMkoqdqXacP8hc/eWiKaj/4TggEE9Pd0cWttbNyst\nzvsrM0Obui0qYs2Wo7Fzh/hbpu2a5aGkpwYAAHDrFNR+CqPDwveWSPuRq3dE1vy6WDd3zx79\nR06ePj1iTO/JG9Pnz4uetmF0CxMFBQAAMAkFHTe2JyGhQKRLyOKara6SlVvwomneIgWJiWlG\nzQYAAGB6Cip2eXl5IuLo6FjnqIrlFWMBAABuJwoqdk5OTiKyZ/v2ui5NUpycnCYizs7ORkoF\nAADQVCio2HkHBnUyk+yVwUHLUrJLaxlQmp2yLCh4ZbaYuQeO8jJ6PgAAANNS0MkTZj6hETNi\nBi/ctznE1yXc07dvDw83hzZ2VhZlJRfPZWbs3524Q3u2RMS2y8zVoT6mTgsAAGBsCip2Ihrf\nd5JS3GeHzFkVf0obt0EbV3OAtaPflLCl8yd52ZoiHgAAgEkpqtiJiL3XpA/jJi44nhqflKY9\ncvJsbn5RmYWNplU7506e3fr69XTRKOjDZQAAgMaktGInIiLmGtdew1171f9NYAAAALcR9m8B\nAACohPqKXcaWjz766KMtGabOAQAAYGTqK3Y/r3rppZdeWvWzqXMAAAAYmfqKHQAAwG1KQSdP\nlBXnF9V2XeIaissMHwUAAKAJUlCx+/Ypu8BoU4cAAABosvgoFgAAQCUUVOzuvddVRLouOHK5\nTl89YdqYAAAAJqKgYvePQYPaiPwSG5fbrC7mZqYOCgAAYBIKKnZmfR4dYC3lyT9uKzJ1FAAA\ngCZIQSdPiPXACSGPF/9mn39c5IGbjur23P+3d+dxUdX7H8c/w7ANLqCIqIhL7gth4pJoGlcj\nxa3FrSwrNbVduub+y8rMbLloy1Uzc00tLb1YbuSahgsuCCq4p6IkKiI7yJzfH7gMizKozDBf\nXs8/fDz4nu+c7/ecz8Hz5sycM3PmdJW6rSw3LwAAgFLBloKduAVNXRVUVKc6nYcOtcRkAAAA\nShkbeisWAAAAd0OwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAA\nABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGw\nAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQ\nBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4A\nAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEE\nOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAA\nRRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwA\nAAAUYW/tCQAArMloNIpIeHi4BcZydnYOCgrS6/UWGAsomwh2AFCmxcTEiEhISEhISIgFhlu9\nenWPHj0sMBBQNhHsAKBMy87OFpFBgwb5+vqW6ECRkZELFy5MT08v0VGAMo5gBwAQX1/fwMBA\na88CwP3i5gkAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAA\nQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7\nAAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEfbWngAA3JecnJw1a9ZkZGRYYKzw8HALjAIA94xgB8C2rV27\ntlevXtaeBQCUCgQ7ALYtPT1dRAYNGuTr61vSY4WGhm7durWkR1GV0WgUC171dHZ2DgoK0uv1\nlhkOKCUIdgBU4OvrGxgYWNKjREZGlvQQCouJiRGRkJCQkJAQy4y4evXqHj16WGYsoJQg2AEA\nLCE7O1ssdW01MjJy4cKFuVdzgTKFYAcAsBzLXFsFyiwedwIAAKAIgh0AAIAieCsWAIB7ZMnH\nKAq3+sIMBDsAgGos9miVvXv3Tps2raRHMTVmzBg/P7+SHoUEabtsMNil/b1lxY8r1mzbGxX7\nd8LV5LTrDi4VXKvWbtTcr2NQn4F9H69tsPYMAQBWZeFHq1jmVt/ff/9906ZNFsuRPCzGRtlY\nsEvY/NFzL07ZGJeVpzU1JTHhwulDO9f/9O2H4ztPWLTs/YAqVpogAMD6LPZoldxnVlvyMYoW\n2CgeFmPTbCnYXY/6pGu3SfsypXz9roMG93mirU89L/eKzvbXM65djjsRtStsxQ8L1h3fOKlb\nV4eIneOa29KmAQAeOAvkLcs/s9oCG8V3hNg0G0o/ab989Mm+TKn21NwdywY/5JRnWb1GPm3+\n9dSQd9/9YUCHIav2Tvnwl3eW93ex0kQBALBdfEeITbOhYLdny5ZUkRYjv8if6m5zqjf483e+\nWjUucuvWCOnf0YyVmnNDU+5fLbl/wTwQlvkL78CBA2KpP7mMRmN0dHTz5s3t7Er2AToWG8iS\nY7FR9yn3ILfMr9Xff/9tmbEsNpAlx2KjbGWs48ePi4ivr6+np2dJjxUfH3/w4MFNmzZZ5p3f\nMnF1ULMZoc87ikiPxRl37ZW+qLuIOA1cbd5Kw8LCzNxRkydPvv9tWL16dQnWEgAA3FVYWNj9\nn81LMxu6Yuft7S1yYs+ff2YM7OJ8p04Z27dHiEitWrXMW2lAQEBoaOjdr9gZjcbY2NjRo0cX\nb76F6datW5HDPShcB7KVsdgoxrLiQJYci42ylbGU3Khczs7OAQEBFhjIinSapll7DmbS9o5v\n3HrqUZ13z/8sm/OGv2eBTHr9n7++fXXAu6vPao3G74mZUuKP+QEAAChVbCjYiaTsGNvhyWmR\nqSJOVX3ad2rTvJ5XlQpO+pzM5EtxJ6J3b90RdTFTpFyLseu3T21fztqzBQAAsCybCnYicu3g\nvAkj3/9+87lC38x0rhkw9KPpU155uKKl5wUAAGB1thbsRETEmHJ61+ZtEVFHz1y8mpKeozeU\nd6taq6FPq04BbWuXt8Sb9AAAAKWQTQY7AAAAFMT1LQAAAEUQ7AAAABRBsAMAAFAEwQ4AAEAR\nBDsAAABFEOwAAAAUQbADAABQBMEOAABAEQQ7AAAARRDsAAAAFEGwAwAAUATBDgAAQBEEOwAA\nAEUQ7AAAABRBsAMAAFCEvbUnULZkZWV9+umnjRo1srMjUpd2RqMxKirKx8eHYpVyVMpWUClb\nYTQaY2Njx44d6+joaO25oNgIdhb12WefTZo0ydqzAACgCHZ2dhMnTrT2LFBsBDuLatCggYgE\nBwe3a9fO2nNBEcLDw0NCQihW6UelbAWVshW5lco9YcHmEOwsKvcNiHbt2vXt29fac0HRQkJC\nKJZNoFK2gkrZipCQEN4xt1GUDQAAQBEEOwAAAEUQ7AAAABRBsAMAAFAEwQ4AAEARBDsAAABF\nEOwAAAAUQbADAABQBMEOAABAEQQ7izIYDLf+RSlHsWwFlbIVVMpWUCmbptM0zdpzKENycnI2\nbtzYuXNnvV5v7bmgCBTLVlApW0GlbAWVsmkEOwAAAEXwViwAAIAiCHYAAACKINgBAAAogmAH\nAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAI\ngh0AAIAiCHYAAACKINgBAAAogmAHAACgCIKdpaQeWTqhf/uGnhWdnStUre/fd+yP0cnWnlNZ\nkn350IZ5Hw3v3aFFgxpuzo4G1xpN2vcJnvnnheuFdDazWNTUAlI2Daut0+l0Ol2fFQWXUqlS\nIOvcphlvP9OuQXU3g8HNu2n7Z0fN2Xo2I18nKmVlOQkRi8f379SioVclF4ObV4MWHftPWBRx\nKadARyqlAA0WkLT9vRaG/Lveqdm7m65Ye2ZlRfqi3oUe/7rKnb7cn5anq5nFoqaWkLblrXo6\nOzs7EZFnl+dbSKWszxi/Lti3XMFfrCfnJZv0olJWlnNkZqB7Yf8BegT+N/a6SUcqpQSCnQVk\nbH2zjojYeff4POxoYnp64rGNXzxV205EvF7dmFb063H/Mn4a0ixw8Puz/7f9wNG4xLS0K39H\n/PJRt1p6EdH7fhxj0tG8YlFTS8jYObqhnX2LkW90EikQ7KiU9V0/8sWjBhHRefi/M3dLzMXU\nzIyrZ/aGTh/++LBlKbd6USlrS1n+rIuI6Ov3/c/66PNJ6RlJFw6tnz6gvr2IGPosu3azH5VS\nBMGu5CUv7e0sInWDd6TfbszYGVxfRBx6Lki03szKusx94xuLiLSbEXezycxiUVMLyNo/vrm9\nvunYiH/mdC4Y7KiU9V2a280gIo5+kyLT79yLSlnd/gn1RKTii79nm7ZeXze0iojUHxtxo4FK\nqYLP2JW4nM1rN2SI+Lw8wt/5dqtT29cGtxTJ3rDmj8I+5AVLcHyk46PlRSQzM/NGi5nFoqYl\nLydq2tDPD9cJ/n6Sn2Nhi6mU1cUtmbsuXaTakM/GP+x8x15UyvpqeHmJiOh0+RdoIlKzplfu\nD1RKGQS7EnciKipdxLVNm4Z52xu0bVtJJDM6+rh15gWRo3v3pYh4tW9f50aDmcWipiXNGBsy\n9OO9NV/77sN2hUcGKmV1Gdu37dFEqj7VP6Cw6H0TlbK+qgNHveylu/bj289P33A4/lpm5rX4\nIxu+euHN+ZftvF5674Vqub2olDIIdiXu4sWLIuLl5ZV/QW5T7mJYnvHUf9+YdlAqPPnhe/43\n/5I1s1jUtGRpp74eNmm3x5DZUwNc7tCFSlnduVOnskV0rdo8Er9p2qDH6rm7OBncvJp1Gjhx\nien9kVSqFKjY8/tdaz8ZWPWvsU82q+7q7OxavemTo7Z6DPh47c65Qa43OlEpZRDsSlx6erqI\nODk55V/g7OwsImlpaZafEyRh7etB7/yR4j1w/sIh3rdazSwWNS1RZ74bPnFbhZdmfv5EhTv2\noVJWl5KSIiJVDPteaxs4dtH2k1fSszKSzh/etmTKwFatR6y7dKMblSoNtMQTB6KOnLyUebsp\n+/KpI5H7jl3WbjZQKWUQ7EqcwWAQ049x3ZKRkSEiLi53uiaBkmI8/7/hjz89+2i1p7/fNO+Z\nqiZLzCwWNS1B5xeMGBNm6Pf1lz0q3aUXlbK63H176ddvf3F5dvra6LiktNSEUzuXjX/cUzJj\nZw+eEJZl0o1KWVNm+OgOXcYujfboN2N99LnEtLTEc4c2fPuCZ+zycYEdRu+48cxBKqUMgl2J\nq1q1qojExcXlX5Db5OHhYfk5lWXZJxYNfKzPd7HV+83f9vMr9R3yLDSzWNS0xCQsefPfa+16\nzfiqb6GP3bqFSlmdp6eniGha8wkrl7zTtVmNigaXKnXa9p8S+t2LlUUurFy5U0SoVCmQuOTj\n6bHZ+semrl3wdmAzLzeDwc2r6ROvz133eRf768emT16cKCJUSiEEuxJXz8fHIJK0e/fRvO3H\ndu1KFHHy8WlgnXmVSWmR03t1eGnZ2bovLd225MW69vmXm1ksalpidq1Zc1kSQ5+vprut0qsb\nRUR+6avT6XT1Jx4QoVKlgNvDD3uLSLmWrZvoTdsrtG7dWEQS4uONIlSqFDh6+PB1kZpt2nrl\nvS22Wts2tUSuHz58TESolEIIdiVOH9At0Fkkav6sv0y+ZSdz98wf9ok4BAZ1KZAuUDKubJ/Y\nuVPwusSmI37dNq+vt76QLmYWi5paHZWyvlbdu3uKpO7bcyTP91Kl7NkTIyKeNWrYiVCpUsDV\n1VVEzu3efT5v+/nwnWduLaZSKrH2g/TKgowtN57T3fOLsKOJGRmJxzd+ees53anWnl3ZkBP3\n24jmBhGXR94NS7hbRzOLRU0tKLGwBxRTKeszxnzW1kHEqVH/Geuizyelp146vWvZ+Mc9RURq\nvbU160Y3KmVlxr3j64mIuPi+8k3Yobir6emJcYfWf/OyTzkRkXqjd9/sSKUUQbCziKQ/3/Mt\n+M16TYP5Zj1LOTbV785/3bgODzPpamaxqKnFFB7sqFQpkBU7s1vV/PtWpNwj7229atKNSlnZ\n1c2jCu5ZERGDT/Bm06+KoFJKINhZSvKhH8f1e7SeRzlHx3JVHnq0z+jFB68V/So8IMUIdprZ\nxaKmlnGnYKdRqVIg+/ymkBFdfb0ruzg4GCrV8g0cMuV/xwp+YSiVsrK02FVTh3Zr9ZBHeSd7\nvVN5j4f8ug6dsjK24AU2KmX7dJqm3fmEBwAAAJvBzRMAAACKINgBAAAogmAHAACgCIIdAACA\nIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYA\nAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAAKIJgBwAAoAiCHQAAgCIIdgAAAIog\n2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQAA\nKIJgBwAAoAiCHQDcyZYRVTp8E2/tWQCA2Qh2AAAAiiDYAWVVyvyuOp2u8cRoa0+k9DFe2vP9\nu30CfGv3nnt5x8gGNZq26zF86k8R8VnWnlgJ45AAbJ+9tScAAKWKMXZmt/ZvbLis3WxIuXBk\n5+9Hdv7+3a5F2ate4D9NAKUZV+wAwETayonvbbisVX7svSU7T60a4u7/xeHTB3eEzhrbz6+6\nk87aswOAuyPYAeYwXjv2x5wJLz7Rqol3ZRdHg2uNxu37jpobcUXL0yv+mw46na7Vp6dToxeN\n6u1X272ccwXPRgHDZu5OEhHJPr1myosdG1Wr6Ozs6u3X5/11Z3PyjpJ9bvNXb/ZqU8+jgpOj\nwc2r6ePPT1walXR7+c5RNXU6XZ9l+SYX83ELnU7XZdal/NPIOL7i//o/Wq9KOSdnt5oPd397\n3sGUG13OTe+gq/DKehGJneKju6nLrKtm75HUY799OizI7yGPCk6OBtfqTR4bMGFxZFKeLinR\nP459pk3dyi5OLu61/HoFLzyYfPzTVjqd7vG89yMUvSoztqg4E8u58Nf3o/v4N6njWcHg4u7d\n2K/LoI+WRiRcFxGR6F27UkU6frDis+fa1nG1F51Tpdo+/j2HT/0pYmZf/e2VcEjcw56/O9s8\nYIDSRgNQtPB3vAr59XGoP2z9FZNeF75uLyI+z7/6aLm8Hct3+iomdlZX97wXfOyb/19kzq0X\nZ0VO71Kl4BUhpwbDQuNvzuLfXiLy7NJ8kzsy2VdEOs9MyDuNF97s6JpvZRU7//e0pmmadjak\nfSHb03lmonm748Jvrzd1Kfh6x4avhl680SUtfFKrfHtBDK0HD/ARkU5fXyjOqszaIvPXduW3\nwTUL+ZPW6aXVmqZp2t5xdUXsus1N1jRN2zzcvb3JZE1xSORhVh3vxmYPGKCUIdgB5tj9QWCP\n4G/+Fx518mJKRlrihSNbF4163ENEGoyNuN0r93QiYl+r64e/7D11OTXl4pHfJnZ0ExG3atWc\n3f1Hzt9+LD459cqJLTN6eIlI+X4/X8t9qTHqo+Z6EangO+S/mw6fS0q5fGLPioldauhFxPXp\nn3KzQrHO4iL6Gv8atyT8eHxyWuLZXYuHP+wsIl7v/nXzZcnznhSRRhOiirsz4uY+UV7EuVH/\nqT/9dSw+MT0zNeH0nl+n9KxjL1LlpdXJmqbl7JvYUCeiq9L+34vDj11MTo6P3bHgrbaVc0+J\nt8/TZqyqGFtk1tpS5nXXizj5DJ27/eg/1zIyk+OPH9i85MOBHUev1zRN07K3ve4lYlf72ekb\nTyVvvHOw45Ao3iFxdzZ8wAClDMEOuFdJC4PsRXw+jL3VcuN00nj83kyTflETGouIVB8UmmTS\nenFmZxF5aPTu3B//fMdLRCo9/WOCSR8ta++YJnYiuoBvL2hacc/iDUfvTjfpdX3b27VExP8/\nZ2403OtZPHJiYxGHLrPO52u/viO4pki5gSuzNW3rWzVEpPorv+c5pSeuGuCR5zxtzqrM3yLz\n1pY6r7udSJtbu6GgqzumdKqqFxGxd3S0q9yy31sfzQo9mHC96F3DIZFX/jremU0fMECpwmfs\nALPk/BM+Z/RzAb51PCo429vpdDqdznXQmusiZ86cyde1VtBTLR1Nfm7ctIlOxDHgqW4VTVo9\nmjb1EDl79qyIiMRHRMSJuDw19LkqpqtyaPnakDYi2p49e4s941o9+rR2NvlZ//DDTUXk0qVL\nd3qFeS7/uT1GJHvjG9729vb2er1er7ezs7Oz09m3Dzknknry5EW5sHfveRH3ZwYFlTd9qVvv\nl3q5FnNV5m+RmWtz6fbC0+6ye+qA175c+seB04kFH2Hi6j9+y4mY9bMnDe/dwt3uyr6fv35/\nRK+HazfpPysy3aQbh8RNxaxjIWz7gAFKFYIdYIZzy/o90mHY58u2HPz7Ukpmjunn4zMyMvJ1\n9vDwyPOzvaOjnUglD4+8D8pwdHQUyc7KPU0kJSWJSE1v7/yfqMptSrl6Nd+H6ouWfxri5OQk\nIjk5xV5TXjdOi1pOTk5OTo7RaDQaNU3Tbu2TrKysG5vj5VXgU2h5m8xZlflbZO7aPAcs3v3b\n1O4uOz575YlH6lYuX7m+/zNvT19/KtN01eXrBw77YNaKab1c/aftO7j22+F+Dsd+fq3nuD9v\nzohD4rZi1rEQtn/AAKUGwQ4o2l//GfPrBaPnvyb+tP3w2UvJGdk5Rk3Troc+b3hQI7i6uorI\nubNntXwLcpvKu7npRUSv10shp8mEhIQHNQ8zuLm5iYj7axtztEJFjK1zY3Pi4uLyvzhvkzmr\nerATy+X8UPex88IOxl9LPHVgy7JPBzU8tyS46yM9Zx3Pv/dFRHQu1X26vj7r96n/0snZX1dG\n5LZySJi4/zra6gFTjNEASyHYAUWLiTkj4j1o6uR+7ZvUdC/vZG+nE7m6YdWW9KJfa55qrVp5\niaSt+n5pnnfFrh+YPW+PiK5VKz8RkcqVK4vIoagoo0mf9PAVa87fy5gODg5S2OWlInj6+9cT\nufzLt0svFJaDRESkup9fdZHLKxetTTVtTgpdGGr6HAlzVvVgJ5aHztGtjm+nZ4a9P3/1hx0k\nKeybxbF36axpcntvpXFImLj/OtrqAXP/kwAeOIIdULRq1VxFzv08bfrmowlpWelXzkav/2Z4\nl+d/uKeTZ+HaDxnWzE4SV4544tXZW2MuJKdeObVv5aTuPT8/lCMVe494rpqIyEP+/tVETs58\nbeRP++OupadcPBG+ZEzQs1/f24UDp2rV3ETObVy+5UxylrHo/je1fnN0gItc/HVox94TFm49\ndCYhNTM98dzRA9t+mRH8jN/w5Zki0uHlwQ10cn7eS0Fjlu4+eSk1JeF4+OLgbkOWJRR7VQ92\nYiIHP3ky4JUP5q7dFX3qYkpWVurlk7t+/GTBfhExGo0isv3/Hus6/NMlmw+ejE/O1kS0jIsx\nm394u+fEzZoY2rdvKSIiBg4JU/dfRxs9YIo1HGAhhV+CBmAia8/7zRzz/eroavQd3stdxGng\n7adZ5d6L5zf1VN6XL++vF/F858+8rfnvZ8zc/2VA5YK/oo71Bq+8fdfesa865X/Wl1vAmwOb\nSsFbIAtMI31RbxGpN2bPzYa00EHueVZl9kPL4n9/26e8FKr3otz7EFN3TGiZ/wFhhtYv9W0q\nIk/MvlycVZm/Reasbf+YeoUt1lXpOe+0pmna5jdu7xQ7vf72J9x0HoHfxt68NZZDIh+z6ng3\nNnvAAKUMV+yAojm0+mDL9jlvdmtR16Oco8HNq3mXoZ+HRSwdUP1BDuLY4t31+/8Iea27Xx33\ncg72ThWrNe44YNyiXXvmPnV7nPpvhW6d88aTzWtUcHRwqVyn7bPjf92z5o3GDvc0oqHn9DVf\nDe3S3MvVSV+878ryDJqxO/qPr0b26dDEy83g4FiuSq3GLQP6Bc9YuW9Ov9z7EF38P94SvvC9\n3n613ZwdDZW8H+kxcv7Ojc+5XhDRV65csTirerAT852wYcuCycO6t2tex6O8k8Gtet3mHQeM\nm/vngV9fri0i0vGT8E0LJg/v3aFFQ6+KDpqmc3LzavZY31Eztx5c83rDm189wSGRz33X0WYP\nGKCU0WnaA/ioAgAULWndy027LTjf4tPjd7gKUur8MdTtgxYx29+sZu2JlE22d8AA1scVOwAl\n4sycof3en//H/uP/pGSkJpw+uGHmiMAXF5wX+3YvDrCZk7SdHf9HWooSBwxgffZFdwGA4jMm\nRS+fPHf55LytlR77fPZbtvMOlk5XvDckce+UOGAA6+OvUQA3RYyto7urHvPNfxBG7WE/LJ8y\npGvrxt7uLg6O5arUbdlt2Gfr9oWN9Lm3j39ZRcDsy2X7fdgHckiYtxIlDhjA+rhiB6BE6Co2\n7TP++z7jrT0P2AgOGOCB4OYJAAAARfBWLAAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAi\nCHYAAACKINgBAAAogmAHAACgCIIdAACAIgh2AAAAiiDYAQC56RjXAAAAVElEQVQAKIJgBwAA\noAiCHQAAgCIIdgAAAIog2AEAACiCYAcAAKAIgh0AAIAiCHYAAACKINgBAAAogmAHAACgCIId\nAACAIgh2AAAAiiDYAQAAKOL/AeNaJYD1g8QLAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title \"Histogram of amount_egenes$amount_e_genes\""
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "hist(amount_egenes$amount_e_genes, breaks = 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "id": "8ac33f4a-d4f3-49d4-af82-59fa37e1402b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save to use as input for enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "fec7ce21-30d2-42d1-9d9b-281194467fd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 34</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>beta_se</th><th scope=col>empirical_feature_p_value</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>⋯</th><th scope=col>gene2</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th><th scope=col>p_bh_cutoff</th><th scope=col>p_q_cutoff</th><th scope=col>type</th><th scope=col>amount_tests_datasets</th><th scope=col>p_val_mt</th><th scope=col>p_val_corrected</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CD4_T</td><td>oneK1K</td><td>A1BG_A2M</td><td>12:8942630:T:C</td><td>0.015607427</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>⋯</td><td>A1BG</td><td>0</td><td>97356</td><td>1519.477</td><td>0.00266828</td><td>0.00266828</td><td>unique</td><td>149856468</td><td>2338874</td><td>0.8627429</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>oneK1K</td><td>A1BG_A2M</td><td>12:8971336:T:C</td><td>0.001673528</td><td>NA</td><td>NA</td><td>NA</td><td>12</td><td>9067664</td><td>⋯</td><td>A1BG</td><td>0</td><td>97356</td><td> 162.928</td><td>0.00266828</td><td>0.00266828</td><td>unique</td><td>149856468</td><td> 250789</td><td>0.5682995</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 34\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cell\\_type & dataset & feature\\_id & snp\\_id & p\\_value & beta & beta\\_se & empirical\\_feature\\_p\\_value & feature\\_chromosome & feature\\_start & ⋯ & gene2 & is\\_rb\\_mt & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene & p\\_bh\\_cutoff & p\\_q\\_cutoff & type & amount\\_tests\\_datasets & p\\_val\\_mt & p\\_val\\_corrected\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <dbl> & <lgl> & <lgl> & <lgl> & <int> & <int> & ⋯ & <chr> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <chr> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t CD4\\_T & oneK1K & A1BG\\_A2M & 12:8942630:T:C & 0.015607427 & NA & NA & NA & 12 & 9067664 & ⋯ & A1BG & 0 & 97356 & 1519.477 & 0.00266828 & 0.00266828 & unique & 149856468 & 2338874 & 0.8627429\\\\\n",
       "\t CD4\\_T & oneK1K & A1BG\\_A2M & 12:8971336:T:C & 0.001673528 & NA & NA & NA & 12 & 9067664 & ⋯ & A1BG & 0 & 97356 &  162.928 & 0.00266828 & 0.00266828 & unique & 149856468 &  250789 & 0.5682995\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 34\n",
       "\n",
       "| cell_type &lt;chr&gt; | dataset &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | p_value &lt;dbl&gt; | beta &lt;lgl&gt; | beta_se &lt;lgl&gt; | empirical_feature_p_value &lt;lgl&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | ⋯ ⋯ | gene2 &lt;chr&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; | p_bh_cutoff &lt;dbl&gt; | p_q_cutoff &lt;dbl&gt; | type &lt;chr&gt; | amount_tests_datasets &lt;int&gt; | p_val_mt &lt;dbl&gt; | p_val_corrected &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| CD4_T | oneK1K | A1BG_A2M | 12:8942630:T:C | 0.015607427 | NA | NA | NA | 12 | 9067664 | ⋯ | A1BG | 0 | 97356 | 1519.477 | 0.00266828 | 0.00266828 | unique | 149856468 | 2338874 | 0.8627429 |\n",
       "| CD4_T | oneK1K | A1BG_A2M | 12:8971336:T:C | 0.001673528 | NA | NA | NA | 12 | 9067664 | ⋯ | A1BG | 0 | 97356 |  162.928 | 0.00266828 | 0.00266828 | unique | 149856468 |  250789 | 0.5682995 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type dataset feature_id snp_id         p_value     beta beta_se\n",
       "1 CD4_T     oneK1K  A1BG_A2M   12:8942630:T:C 0.015607427 NA   NA     \n",
       "2 CD4_T     oneK1K  A1BG_A2M   12:8971336:T:C 0.001673528 NA   NA     \n",
       "  empirical_feature_p_value feature_chromosome feature_start ⋯ gene2 is_rb_mt\n",
       "1 NA                        12                 9067664       ⋯ A1BG  0       \n",
       "2 NA                        12                 9067664       ⋯ A1BG  0       \n",
       "  amount_tests_eGene p_val_mt_eGene p_bh_cutoff p_q_cutoff type  \n",
       "1 97356              1519.477       0.00266828  0.00266828 unique\n",
       "2 97356               162.928       0.00266828  0.00266828 unique\n",
       "  amount_tests_datasets p_val_mt p_val_corrected\n",
       "1 149856468             2338874  0.8627429      \n",
       "2 149856468              250789  0.5682995      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sign_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "id": "b8d7033d-3703-4d62-a71f-b4cb8c992f55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'cell_type'</li><li>'dataset'</li><li>'feature_id'</li><li>'snp_id'</li><li>'p_value'</li><li>'beta'</li><li>'beta_se'</li><li>'empirical_feature_p_value'</li><li>'feature_chromosome'</li><li>'feature_start'</li><li>'feature_end'</li><li>'ENSG'</li><li>'alpha_param'</li><li>'beta_param'</li><li>'eGene'</li><li>'coeGene'</li><li>'coQTL'</li><li>'snp_position'</li><li>'closeGenes'</li><li>'z_score'</li><li>'triplet'</li><li>'n_samples'</li><li>'n_e_samples'</li><li>'gene1'</li><li>'gene2'</li><li>'is_rb_mt'</li><li>'amount_tests_eGene'</li><li>'p_val_mt_eGene'</li><li>'p_bh_cutoff'</li><li>'p_q_cutoff'</li><li>'type'</li><li>'amount_tests_datasets'</li><li>'p_val_mt'</li><li>'p_val_corrected'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'cell\\_type'\n",
       "\\item 'dataset'\n",
       "\\item 'feature\\_id'\n",
       "\\item 'snp\\_id'\n",
       "\\item 'p\\_value'\n",
       "\\item 'beta'\n",
       "\\item 'beta\\_se'\n",
       "\\item 'empirical\\_feature\\_p\\_value'\n",
       "\\item 'feature\\_chromosome'\n",
       "\\item 'feature\\_start'\n",
       "\\item 'feature\\_end'\n",
       "\\item 'ENSG'\n",
       "\\item 'alpha\\_param'\n",
       "\\item 'beta\\_param'\n",
       "\\item 'eGene'\n",
       "\\item 'coeGene'\n",
       "\\item 'coQTL'\n",
       "\\item 'snp\\_position'\n",
       "\\item 'closeGenes'\n",
       "\\item 'z\\_score'\n",
       "\\item 'triplet'\n",
       "\\item 'n\\_samples'\n",
       "\\item 'n\\_e\\_samples'\n",
       "\\item 'gene1'\n",
       "\\item 'gene2'\n",
       "\\item 'is\\_rb\\_mt'\n",
       "\\item 'amount\\_tests\\_eGene'\n",
       "\\item 'p\\_val\\_mt\\_eGene'\n",
       "\\item 'p\\_bh\\_cutoff'\n",
       "\\item 'p\\_q\\_cutoff'\n",
       "\\item 'type'\n",
       "\\item 'amount\\_tests\\_datasets'\n",
       "\\item 'p\\_val\\_mt'\n",
       "\\item 'p\\_val\\_corrected'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'cell_type'\n",
       "2. 'dataset'\n",
       "3. 'feature_id'\n",
       "4. 'snp_id'\n",
       "5. 'p_value'\n",
       "6. 'beta'\n",
       "7. 'beta_se'\n",
       "8. 'empirical_feature_p_value'\n",
       "9. 'feature_chromosome'\n",
       "10. 'feature_start'\n",
       "11. 'feature_end'\n",
       "12. 'ENSG'\n",
       "13. 'alpha_param'\n",
       "14. 'beta_param'\n",
       "15. 'eGene'\n",
       "16. 'coeGene'\n",
       "17. 'coQTL'\n",
       "18. 'snp_position'\n",
       "19. 'closeGenes'\n",
       "20. 'z_score'\n",
       "21. 'triplet'\n",
       "22. 'n_samples'\n",
       "23. 'n_e_samples'\n",
       "24. 'gene1'\n",
       "25. 'gene2'\n",
       "26. 'is_rb_mt'\n",
       "27. 'amount_tests_eGene'\n",
       "28. 'p_val_mt_eGene'\n",
       "29. 'p_bh_cutoff'\n",
       "30. 'p_q_cutoff'\n",
       "31. 'type'\n",
       "32. 'amount_tests_datasets'\n",
       "33. 'p_val_mt'\n",
       "34. 'p_val_corrected'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"cell_type\"                 \"dataset\"                  \n",
       " [3] \"feature_id\"                \"snp_id\"                   \n",
       " [5] \"p_value\"                   \"beta\"                     \n",
       " [7] \"beta_se\"                   \"empirical_feature_p_value\"\n",
       " [9] \"feature_chromosome\"        \"feature_start\"            \n",
       "[11] \"feature_end\"               \"ENSG\"                     \n",
       "[13] \"alpha_param\"               \"beta_param\"               \n",
       "[15] \"eGene\"                     \"coeGene\"                  \n",
       "[17] \"coQTL\"                     \"snp_position\"             \n",
       "[19] \"closeGenes\"                \"z_score\"                  \n",
       "[21] \"triplet\"                   \"n_samples\"                \n",
       "[23] \"n_e_samples\"               \"gene1\"                    \n",
       "[25] \"gene2\"                     \"is_rb_mt\"                 \n",
       "[27] \"amount_tests_eGene\"        \"p_val_mt_eGene\"           \n",
       "[29] \"p_bh_cutoff\"               \"p_q_cutoff\"               \n",
       "[31] \"type\"                      \"amount_tests_datasets\"    \n",
       "[33] \"p_val_mt\"                  \"p_val_corrected\"          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(sign_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "e5048b6a-7239-48f1-95cb-9948296d1a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#unique(sign_results$p_bh_cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "f588c4ab-b119-423d-90f0-5057a6890b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(sign_results, paste0(result_path_analysis, 'F3_Co_E_Genes', n_sample_filter  , '_', snp_filter,  '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "9ecd9435-ab7b-48fd-8599-66c43fe1a1b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/'"
      ],
      "text/latex": [
       "'../data/current/coeqtl\\_mapping/co\\_qtls\\_decision\\_tree/analysis\\_oneK1K/CD4\\_T/'"
      ],
      "text/markdown": [
       "'../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/'"
      ],
      "text/plain": [
       "[1] \"../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "result_path_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "id": "088d31e9-15c8-438c-93c8-7c8135def835",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "8050384"
      ],
      "text/latex": [
       "8050384"
      ],
      "text/markdown": [
       "8050384"
      ],
      "text/plain": [
       "[1] 8050384"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f30da423-c0ff-46f2-9021-8fce9d3c72ea",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Save processed co-eQTL testing file for further analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "993e12f7-93d6-4ff0-b1aa-0130ab4083e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### In the initial mapping the output of this was used as decision tree input, filename is a relict of this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "id": "557a200c-ea27-4714-8e0d-3ba3f6eb1ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### save as decision tree input\n",
    "write.csv(coeqtl_results,  gzfile(paste0(data_path, coeqtl_results_path , '/', dataset, '/', cell_type, '/F3_decision_tree_input.csv.gz')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "fac81630-9f30-44d6-9c05-fa331b760609",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Prepare dummy empty correlation stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "2e67c5c1-74d3-4221-b98f-bf15528575a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file.exists(paste0(result_path_analysis , 'F6_Correlation_Summary_Stats_all.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "d525467c-3f45-4996-8a60-d5981e6b792f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_stats = data.frame(gene_pair = unique(coeqtl_results$feature_id),\n",
    "                       mean_correlation = 0,\n",
    "                       mean_abs_correlation = 0,\n",
    "                       var_correlation = 0,\n",
    "                       max_correlation = 0,\n",
    "                       n_NA = 0,\n",
    "                       n_not_NA = 0,\n",
    "                       n_significant = 0, \n",
    "                       weighted_variance = 0,\n",
    "                       perc_significant = 0,\n",
    "                       weighted_var_div_mean = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "5ff26099-178b-4073-b580-d1fe1cdbde22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/F6_Correlation_Summary_Stats_all.csv'"
      ],
      "text/latex": [
       "'../data/current/coeqtl\\_mapping/co\\_qtls\\_decision\\_tree/analysis\\_oneK1K/CD4\\_T/F6\\_Correlation\\_Summary\\_Stats\\_all.csv'"
      ],
      "text/markdown": [
       "'../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/F6_Correlation_Summary_Stats_all.csv'"
      ],
      "text/plain": [
       "[1] \"../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/CD4_T/F6_Correlation_Summary_Stats_all.csv\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "paste0(result_path_analysis , 'F6_Correlation_Summary_Stats_all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "b0315a3e-7604-4ae6-b552-4216e5541a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### save as decision tree input\n",
    "if(!file.exists(paste0(result_path_analysis , 'F6_Correlation_Summary_Stats_all', '_', snp_filter, '.csv'))){\n",
    "    write.csv(cor_stats,  paste0(result_path_analysis , 'F6_Correlation_Summary_Stats_all', '_', snp_filter, '.csv'))\n",
    "    }"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc-autonumbering": false,
  "toc-showcode": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
