{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8c6615df-7b05-4d17-830c-e3e70c3ccd07",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check Meta-Analysis results\n",
    "### Load the file of co-eQTL mapping from the meta-analysis (all datasets)\n",
    "### Apply multiple testing to the meta-analysis files and save the significant co-eQTLs and p-value thresholds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0f9cc8-9018-4ce5-bf8a-5bd63d7f7c1e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "14087201-e77e-4290-a63a-f7c7a13e9979",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "This is mgcv 1.9-1. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "corrplot 0.95 loaded\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:genefilter’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "\n",
      "************\n",
      "Welcome to faux. For support and examples visit:\n",
      "https://debruine.github.io/faux/\n",
      "- Get and set global package options with: faux_options()\n",
      "************\n",
      "\n",
      "wCorr v1.9.8\n",
      "\n",
      "\n",
      "Learn more about sjmisc with 'browseVignettes(\"sjmisc\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘sjmisc’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    is_empty\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘DT’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "This is coloc version 5.2.3\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: gsubfn\n",
      "\n",
      "Loading required package: proto\n",
      "\n",
      "Warning message in fun(libname, pkgname):\n",
      "“couldn't connect to display \"localhost:19.0\"”\n",
      "Loading required package: RSQLite\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    replace_na\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dbscan’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    as.dendrogram\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:rtracklayer’:\n",
      "\n",
      "    blocks, path\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    components\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: dynamicTreeCut\n",
      "\n",
      "Loading required package: fastcluster\n",
      "\n",
      "\n",
      "Attaching package: ‘fastcluster’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    hclust\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘WGCNA’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:qvalue’:\n",
      "\n",
      "    qvalue\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    geometry\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: MotifDb\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'XVector'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grid':\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "See system.file(\"LICENSE\", package=\"MotifDb\") for use restrictions.\n",
      "\n",
      "Warning message:\n",
      "\"replacing previous import 'S4Vectors::as.data.frame' by 'motifStack::as.data.frame' when loading 'motifbreakR'\"\n",
      "\n",
      "Attaching package: 'Gviz'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    feature\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "\n",
      "Attaching package: 'MOFA2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    predict\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d118880-0a5e-4344-80d2-95796ebb3c2f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dcf323df-2e3d-4c30-bcd8-2b791e0e3ed8",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "30e3d622-08ef-4ced-83aa-c527e3403dd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Cell-type for which to execute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "35d94547-9203-4f5a-875a-d5ffba3fc0e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = 'B'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d832b66-8beb-4ef7-9ae2-0a425b1e935a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Which co-eQTL results (final-mapping: co_qtls_decision_tree; or initial mapping in evaluation phase: co_qtls_sceqtlgen )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "860b87c8-2f9d-445f-a039-86d7addc1ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results_path = 'co_qtls_decision_tree/'   # path to new results\n",
    "coeqtl_results_path_old = 'co_qtls_sceqtlgen/'   # path to old results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a4be65e4-0693-4fff-8fbe-c81e87c869b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the co-EQTL Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7815e317-c26f-4f32-8778-29cd148830a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data/current/coeqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ed3affd3-b654-4ef4-9f88-fb1ff4654363",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the eQTL files (sceQTLGen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7608b240-f56d-4f0b-bdcf-2d3bdeca1a78",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_results = 'sc_eqtlgen_final'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d9a54106-41e0-4b73-a61a-d90b741951e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl = '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e4fabec8-710b-4833-8160-a19ad9f0411f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## which meta_analysis (which datasets where included)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "863ba71a-7f4a-44cb-a10d-fa0ce745ae42",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_analysis_type = 'meta_5ds'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a981a00e-2023-4957-abfe-bba86159c026",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = 'meta_qtl_results_all.txt.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "50dce54d-846b-4742-8acb-9eea1cb59cb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for multiple testing correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "640b737a-acd3-4ed3-adde-967e9d262e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_samples_filter =  133 # either 0 or 133; whether to filter tests that did not include a minimum number of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7122fab8-6332-4aef-930e-c45184f0b0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_filter = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "db02afa7-35c0-48d2-8f62-6609b31b3e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the significant co-eqtl results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d0f3af76-da10-40c2-9cec-22053ae0a092",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path_analysis = paste0( \"../data/current/coeqtl_mapping/\", coeqtl_results_path , \"analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "37d3f67d-24ca-49c6-a2b3-2595957eaa39",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_meta = '../data/current/meta-data'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f26dcd15-8141-47b5-ace3-f17dc4cdddb8",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36e93189-b874-4d5e-addc-fcccaf23d1de",
   "metadata": {},
   "source": [
    "## Load ribosomal/mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c5424972-23eb-4489-9ad8-a918643ee600",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the annotation of ribosomal and mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "aa563e46-dd36-410c-8180-de296eadf35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ribosomal_genes = read.csv(paste0(data_path_meta , '/ribosomal_genes.txt.gz') ,sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "81ce1a49-fa0c-456f-9e07-bd723a309c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ribosomal_genes = unique(ribosomal_genes$Gene.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "967d3602-2dd7-4d12-a8d0-99f3bc55903c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ABCE1'</li><li>'ABCF1'</li><li>'APEX1'</li><li>'APOD'</li><li>'ASCC2'</li><li>'ASCC3'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ABCE1'\n",
       "\\item 'ABCF1'\n",
       "\\item 'APEX1'\n",
       "\\item 'APOD'\n",
       "\\item 'ASCC2'\n",
       "\\item 'ASCC3'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ABCE1'\n",
       "2. 'ABCF1'\n",
       "3. 'APEX1'\n",
       "4. 'APOD'\n",
       "5. 'ASCC2'\n",
       "6. 'ASCC3'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ABCE1\" \"ABCF1\" \"APEX1\" \"APOD\"  \"ASCC2\" \"ASCC3\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(ribosomal_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aab884ed-9683-4e31-86cd-58c5c5eeb580",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1c2b70a0-3769-4862-8dcc-811d80485f93",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrial_genes = read.csv(paste0(data_path_meta , '/mitochondrial_genes.txt') ,sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "db5b383d-6430-41eb-a900-1753e004d09d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gene.stable.ID</th><th scope=col>Gene.stable.ID.version</th><th scope=col>Transcript.stable.ID</th><th scope=col>Transcript.stable.ID.version</th><th scope=col>Chromosome.scaffold.name</th><th scope=col>Gene.start..bp.</th><th scope=col>Gene.end..bp.</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000198888</td><td>ENSG00000198888.2</td><td>ENST00000361390</td><td>ENST00000361390.2</td><td>MT</td><td>3307</td><td>4262</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000198763</td><td>ENSG00000198763.3</td><td>ENST00000361453</td><td>ENST00000361453.3</td><td>MT</td><td>4470</td><td>5511</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & Gene.stable.ID & Gene.stable.ID.version & Transcript.stable.ID & Transcript.stable.ID.version & Chromosome.scaffold.name & Gene.start..bp. & Gene.end..bp.\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000198888 & ENSG00000198888.2 & ENST00000361390 & ENST00000361390.2 & MT & 3307 & 4262\\\\\n",
       "\t2 & ENSG00000198763 & ENSG00000198763.3 & ENST00000361453 & ENST00000361453.3 & MT & 4470 & 5511\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | Gene.stable.ID &lt;chr&gt; | Gene.stable.ID.version &lt;chr&gt; | Transcript.stable.ID &lt;chr&gt; | Transcript.stable.ID.version &lt;chr&gt; | Chromosome.scaffold.name &lt;chr&gt; | Gene.start..bp. &lt;int&gt; | Gene.end..bp. &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000198888 | ENSG00000198888.2 | ENST00000361390 | ENST00000361390.2 | MT | 3307 | 4262 |\n",
       "| 2 | ENSG00000198763 | ENSG00000198763.3 | ENST00000361453 | ENST00000361453.3 | MT | 4470 | 5511 |\n",
       "\n"
      ],
      "text/plain": [
       "  Gene.stable.ID  Gene.stable.ID.version Transcript.stable.ID\n",
       "1 ENSG00000198888 ENSG00000198888.2      ENST00000361390     \n",
       "2 ENSG00000198763 ENSG00000198763.3      ENST00000361453     \n",
       "  Transcript.stable.ID.version Chromosome.scaffold.name Gene.start..bp.\n",
       "1 ENST00000361390.2            MT                       3307           \n",
       "2 ENST00000361453.3            MT                       4470           \n",
       "  Gene.end..bp.\n",
       "1 4262         \n",
       "2 5511         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2afc808a-a425-4d44-ab3c-9ec6e9a3adff",
   "metadata": {},
   "outputs": [],
   "source": [
    "### convert ENSG to SYMBOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6a92b3cc-0408-4ea4-943c-a04067357e79",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "Warning message in bitr(unique(mitochondrial_genes$Gene.stable.ID), fromType = \"ENSEMBL\", :\n",
      "\"2.14% of input gene IDs are fail to map...\"\n"
     ]
    }
   ],
   "source": [
    "mitochondrial_genes_convert <- bitr(unique(mitochondrial_genes$Gene.stable.ID), \n",
    "                  fromType = \"ENSEMBL\", \n",
    "                  toType = \"SYMBOL\", \n",
    "                  OrgDb = org.Hs.eg.db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "50713925-a973-456c-b09b-d9f7955c56f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ENSEMBL</th><th scope=col>SYMBOL</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000198888</td><td>ND1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000198763</td><td>ND2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & ENSEMBL & SYMBOL\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000198888 & ND1\\\\\n",
       "\t2 & ENSG00000198763 & ND2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | ENSEMBL &lt;chr&gt; | SYMBOL &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | ENSG00000198888 | ND1 |\n",
       "| 2 | ENSG00000198763 | ND2 |\n",
       "\n"
      ],
      "text/plain": [
       "  ENSEMBL         SYMBOL\n",
       "1 ENSG00000198888 ND1   \n",
       "2 ENSG00000198763 ND2   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes_convert,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9ecd7e18-c9e4-4b4a-8b18-ba6923c0fc08",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitochondrial_genes  = unique(mitochondrial_genes_convert$SYMBOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ebae6999-20b7-4352-a456-9beef1d5f1f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ND1'</li><li>'ND2'</li><li>'COX1'</li><li>'COX2'</li><li>'ATP8'</li><li>'ATP6'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ND1'\n",
       "\\item 'ND2'\n",
       "\\item 'COX1'\n",
       "\\item 'COX2'\n",
       "\\item 'ATP8'\n",
       "\\item 'ATP6'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ND1'\n",
       "2. 'ND2'\n",
       "3. 'COX1'\n",
       "4. 'COX2'\n",
       "5. 'ATP8'\n",
       "6. 'ATP6'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ND1\"  \"ND2\"  \"COX1\" \"COX2\" \"ATP8\" \"ATP6\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mitochondrial_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a52ce4ec-94c2-4b04-93c1-583ecf825238",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f98b6751-9ec2-4be2-9ea9-41ae77161952",
   "metadata": {},
   "outputs": [],
   "source": [
    "rb_mt_genes = unique(c(ribosomal_genes, mitochondrial_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78cf6f79-5deb-4d4e-879a-6d016e8bac57",
   "metadata": {},
   "source": [
    "## Load the SNP eGene mapping (cis-eQTLs from sceQTLgen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1116274d-de75-406b-ad74-78469c04eb07",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_gene_mapping = read.csv(paste0(data_path_eqtl, eqtl_results , '/' , cell_type, '_snp_egene_mapping.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cfbdd2ce-ff01-4b7f-9934-e46515ab3c3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2487"
      ],
      "text/latex": [
       "2487"
      ],
      "text/markdown": [
       "2487"
      ],
      "text/plain": [
       "[1] 2487"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_gene_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "62a21979-9c67-4175-8d82-3ff8d7911d79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>p_value</th><th scope=col>beta</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>CHI3L2</td><td>1:111230252:A:AT</td><td>111200771</td><td>111243440</td><td>8.658187e-104</td><td> 1.0929843</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>FCRL3 </td><td>1:157699488:C:T </td><td>157674321</td><td>157700857</td><td> 7.538984e-89</td><td>-0.6075662</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & X & feature\\_id & snp\\_id & feature\\_start & feature\\_end & p\\_value & beta\\\\\n",
       "  & <int> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & CHI3L2 & 1:111230252:A:AT & 111200771 & 111243440 & 8.658187e-104 &  1.0929843\\\\\n",
       "\t2 & 2 & FCRL3  & 1:157699488:C:T  & 157674321 & 157700857 &  7.538984e-89 & -0.6075662\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | p_value &lt;dbl&gt; | beta &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | CHI3L2 | 1:111230252:A:AT | 111200771 | 111243440 | 8.658187e-104 |  1.0929843 |\n",
       "| 2 | 2 | FCRL3  | 1:157699488:C:T  | 157674321 | 157700857 |  7.538984e-89 | -0.6075662 |\n",
       "\n"
      ],
      "text/plain": [
       "  X feature_id snp_id           feature_start feature_end p_value      \n",
       "1 1 CHI3L2     1:111230252:A:AT 111200771     111243440   8.658187e-104\n",
       "2 2 FCRL3      1:157699488:C:T  157674321     157700857    7.538984e-89\n",
       "  beta      \n",
       "1  1.0929843\n",
       "2 -0.6075662"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c1247be1-4f95-4539-b3a3-54f73a73916a",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_gene_mapping = unique(snp_gene_mapping[,c('feature_id', 'snp_id')])\n",
    "colnames(snp_gene_mapping) = c('gene', 'snp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9c239b8e-8267-4722-9a2a-810337a2b4a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2017"
      ],
      "text/latex": [
       "2017"
      ],
      "text/markdown": [
       "2017"
      ],
      "text/plain": [
       "[1] 2017"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(snp_gene_mapping$gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cd7dfe4-522a-45aa-93f5-79b22eec7541",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2cb6b4-de99-46db-a133-b9bd344fed38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "294c60ad-e1de-4705-8d4f-c00e579b7a4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "03dd8cac-743e-4910-8585-27ad16d76baa",
   "metadata": {},
   "source": [
    "## Load Meta-Analysis co-eQTL mapping file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d82af6d0-c4db-4416-b026-dad05a327742",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load meta-analysis file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f6384ed0-ebe4-4506-a6d6-f4eabd5d117f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'B'"
      ],
      "text/latex": [
       "'B'"
      ],
      "text/markdown": [
       "'B'"
      ],
      "text/plain": [
       "[1] \"B\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e9753412-3e36-46ce-8533-5708a2135ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_analysis = fread(paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/', file_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6cdc6260-b586-4168-b36a-fad0153f4872",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/', file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "9f88e668-900b-43b2-bcb9-4324be95909f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] \"2025-03-03 11:54:43 CET\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file.info(path)$ctime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "8282455a-2e19-4736-adc0-8a365745173f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 8 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>coQTL</th><th scope=col>eGene</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>coeGene</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>snp_position</th><th scope=col>closeGenes</th><th scope=col>⋯</th><th scope=col>z_score;oneK1K</th><th scope=col>n_samples;Franke_V3</th><th scope=col>n_e_samples;Franke_V3</th><th scope=col>z_score;Franke_V3</th><th scope=col>n_samples;Franke_V2</th><th scope=col>n_e_samples;Franke_V2</th><th scope=col>z_score;Franke_V2</th><th scope=col>n_samples;Wijst</th><th scope=col>n_e_samples;Wijst</th><th scope=col>z_score;Wijst</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>14:50315331:G:A;VCPKMT_IGHG2 </td><td>VCPKMT</td><td>14</td><td> 50108632</td><td> 50116600</td><td>IGHG2 </td><td>IGHG2_VCPKMT</td><td>14:50315331:G:A </td><td> 50315331</td><td>FALSE</td><td>⋯</td><td>-5.74351200</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:50219111:C:CA;VCPKMT_IGHG2</td><td>VCPKMT</td><td>14</td><td> 50108632</td><td> 50116600</td><td>IGHG2 </td><td>IGHG2_VCPKMT</td><td>14:50219111:C:CA</td><td> 50219111</td><td>FALSE</td><td>⋯</td><td>-1.83053200</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:104896820:A:G;IGHG2_VCPKMT</td><td>IGHG2 </td><td>14</td><td>105639559</td><td>105644790</td><td>VCPKMT</td><td>IGHG2_VCPKMT</td><td>14:104896820:A:G</td><td>104896820</td><td>FALSE</td><td>⋯</td><td>-1.39619300</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:49843683:T:C;VCPKMT_IGHG2 </td><td>VCPKMT</td><td>14</td><td> 50108632</td><td> 50116600</td><td>IGHG2 </td><td>IGHG2_VCPKMT</td><td>14:49843683:T:C </td><td> 49843683</td><td>FALSE</td><td>⋯</td><td> 0.55119880</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:104924668:C:T;IGHG2_VCPKMT</td><td>IGHG2 </td><td>14</td><td>105639559</td><td>105644790</td><td>VCPKMT</td><td>IGHG2_VCPKMT</td><td>14:104924668:C:T</td><td>104924668</td><td>FALSE</td><td>⋯</td><td> 0.30941640</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:105057326:G:A;IGHG2_VCPKMT</td><td>IGHG2 </td><td>14</td><td>105639559</td><td>105644790</td><td>VCPKMT</td><td>IGHG2_VCPKMT</td><td>14:105057326:G:A</td><td>105057326</td><td>FALSE</td><td>⋯</td><td>-0.26904990</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:104919812:G:C;IGHG2_VCPKMT</td><td>IGHG2 </td><td>14</td><td>105639559</td><td>105644790</td><td>VCPKMT</td><td>IGHG2_VCPKMT</td><td>14:104919812:G:C</td><td>104919812</td><td>FALSE</td><td>⋯</td><td>-0.05839607</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>14:104921845:A:C;IGHG2_VCPKMT</td><td>IGHG2 </td><td>14</td><td>105639559</td><td>105644790</td><td>VCPKMT</td><td>IGHG2_VCPKMT</td><td>14:104921845:A:C</td><td>104921845</td><td>FALSE</td><td>⋯</td><td>-0.04793983</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 8 × 31\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " coQTL & eGene & feature\\_chromosome & feature\\_start & feature\\_end & coeGene & feature\\_id & snp\\_id & snp\\_position & closeGenes & ⋯ & z\\_score;oneK1K & n\\_samples;Franke\\_V3 & n\\_e\\_samples;Franke\\_V3 & z\\_score;Franke\\_V3 & n\\_samples;Franke\\_V2 & n\\_e\\_samples;Franke\\_V2 & z\\_score;Franke\\_V2 & n\\_samples;Wijst & n\\_e\\_samples;Wijst & z\\_score;Wijst\\\\\n",
       " <chr> & <chr> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <lgl> & ⋯ & <dbl> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <int> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 14:50315331:G:A;VCPKMT\\_IGHG2  & VCPKMT & 14 &  50108632 &  50116600 & IGHG2  & IGHG2\\_VCPKMT & 14:50315331:G:A  &  50315331 & FALSE & ⋯ & -5.74351200 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:50219111:C:CA;VCPKMT\\_IGHG2 & VCPKMT & 14 &  50108632 &  50116600 & IGHG2  & IGHG2\\_VCPKMT & 14:50219111:C:CA &  50219111 & FALSE & ⋯ & -1.83053200 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:104896820:A:G;IGHG2\\_VCPKMT & IGHG2  & 14 & 105639559 & 105644790 & VCPKMT & IGHG2\\_VCPKMT & 14:104896820:A:G & 104896820 & FALSE & ⋯ & -1.39619300 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:49843683:T:C;VCPKMT\\_IGHG2  & VCPKMT & 14 &  50108632 &  50116600 & IGHG2  & IGHG2\\_VCPKMT & 14:49843683:T:C  &  49843683 & FALSE & ⋯ &  0.55119880 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:104924668:C:T;IGHG2\\_VCPKMT & IGHG2  & 14 & 105639559 & 105644790 & VCPKMT & IGHG2\\_VCPKMT & 14:104924668:C:T & 104924668 & FALSE & ⋯ &  0.30941640 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:105057326:G:A;IGHG2\\_VCPKMT & IGHG2  & 14 & 105639559 & 105644790 & VCPKMT & IGHG2\\_VCPKMT & 14:105057326:G:A & 105057326 & FALSE & ⋯ & -0.26904990 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:104919812:G:C;IGHG2\\_VCPKMT & IGHG2  & 14 & 105639559 & 105644790 & VCPKMT & IGHG2\\_VCPKMT & 14:104919812:G:C & 104919812 & FALSE & ⋯ & -0.05839607 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 14:104921845:A:C;IGHG2\\_VCPKMT & IGHG2  & 14 & 105639559 & 105644790 & VCPKMT & IGHG2\\_VCPKMT & 14:104921845:A:C & 104921845 & FALSE & ⋯ & -0.04793983 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 8 × 31\n",
       "\n",
       "| coQTL &lt;chr&gt; | eGene &lt;chr&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | coeGene &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | snp_position &lt;int&gt; | closeGenes &lt;lgl&gt; | ⋯ ⋯ | z_score;oneK1K &lt;dbl&gt; | n_samples;Franke_V3 &lt;int&gt; | n_e_samples;Franke_V3 &lt;int&gt; | z_score;Franke_V3 &lt;dbl&gt; | n_samples;Franke_V2 &lt;int&gt; | n_e_samples;Franke_V2 &lt;int&gt; | z_score;Franke_V2 &lt;dbl&gt; | n_samples;Wijst &lt;int&gt; | n_e_samples;Wijst &lt;int&gt; | z_score;Wijst &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 14:50315331:G:A;VCPKMT_IGHG2  | VCPKMT | 14 |  50108632 |  50116600 | IGHG2  | IGHG2_VCPKMT | 14:50315331:G:A  |  50315331 | FALSE | ⋯ | -5.74351200 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:50219111:C:CA;VCPKMT_IGHG2 | VCPKMT | 14 |  50108632 |  50116600 | IGHG2  | IGHG2_VCPKMT | 14:50219111:C:CA |  50219111 | FALSE | ⋯ | -1.83053200 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:104896820:A:G;IGHG2_VCPKMT | IGHG2  | 14 | 105639559 | 105644790 | VCPKMT | IGHG2_VCPKMT | 14:104896820:A:G | 104896820 | FALSE | ⋯ | -1.39619300 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:49843683:T:C;VCPKMT_IGHG2  | VCPKMT | 14 |  50108632 |  50116600 | IGHG2  | IGHG2_VCPKMT | 14:49843683:T:C  |  49843683 | FALSE | ⋯ |  0.55119880 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:104924668:C:T;IGHG2_VCPKMT | IGHG2  | 14 | 105639559 | 105644790 | VCPKMT | IGHG2_VCPKMT | 14:104924668:C:T | 104924668 | FALSE | ⋯ |  0.30941640 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:105057326:G:A;IGHG2_VCPKMT | IGHG2  | 14 | 105639559 | 105644790 | VCPKMT | IGHG2_VCPKMT | 14:105057326:G:A | 105057326 | FALSE | ⋯ | -0.26904990 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:104919812:G:C;IGHG2_VCPKMT | IGHG2  | 14 | 105639559 | 105644790 | VCPKMT | IGHG2_VCPKMT | 14:104919812:G:C | 104919812 | FALSE | ⋯ | -0.05839607 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 14:104921845:A:C;IGHG2_VCPKMT | IGHG2  | 14 | 105639559 | 105644790 | VCPKMT | IGHG2_VCPKMT | 14:104921845:A:C | 104921845 | FALSE | ⋯ | -0.04793983 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  coQTL                         eGene  feature_chromosome feature_start\n",
       "1 14:50315331:G:A;VCPKMT_IGHG2  VCPKMT 14                  50108632    \n",
       "2 14:50219111:C:CA;VCPKMT_IGHG2 VCPKMT 14                  50108632    \n",
       "3 14:104896820:A:G;IGHG2_VCPKMT IGHG2  14                 105639559    \n",
       "4 14:49843683:T:C;VCPKMT_IGHG2  VCPKMT 14                  50108632    \n",
       "5 14:104924668:C:T;IGHG2_VCPKMT IGHG2  14                 105639559    \n",
       "6 14:105057326:G:A;IGHG2_VCPKMT IGHG2  14                 105639559    \n",
       "7 14:104919812:G:C;IGHG2_VCPKMT IGHG2  14                 105639559    \n",
       "8 14:104921845:A:C;IGHG2_VCPKMT IGHG2  14                 105639559    \n",
       "  feature_end coeGene feature_id   snp_id           snp_position closeGenes ⋯\n",
       "1  50116600   IGHG2   IGHG2_VCPKMT 14:50315331:G:A   50315331    FALSE      ⋯\n",
       "2  50116600   IGHG2   IGHG2_VCPKMT 14:50219111:C:CA  50219111    FALSE      ⋯\n",
       "3 105644790   VCPKMT  IGHG2_VCPKMT 14:104896820:A:G 104896820    FALSE      ⋯\n",
       "4  50116600   IGHG2   IGHG2_VCPKMT 14:49843683:T:C   49843683    FALSE      ⋯\n",
       "5 105644790   VCPKMT  IGHG2_VCPKMT 14:104924668:C:T 104924668    FALSE      ⋯\n",
       "6 105644790   VCPKMT  IGHG2_VCPKMT 14:105057326:G:A 105057326    FALSE      ⋯\n",
       "7 105644790   VCPKMT  IGHG2_VCPKMT 14:104919812:G:C 104919812    FALSE      ⋯\n",
       "8 105644790   VCPKMT  IGHG2_VCPKMT 14:104921845:A:C 104921845    FALSE      ⋯\n",
       "  z_score;oneK1K n_samples;Franke_V3 n_e_samples;Franke_V3 z_score;Franke_V3\n",
       "1 -5.74351200    NA                  NA                    NA               \n",
       "2 -1.83053200    NA                  NA                    NA               \n",
       "3 -1.39619300    NA                  NA                    NA               \n",
       "4  0.55119880    NA                  NA                    NA               \n",
       "5  0.30941640    NA                  NA                    NA               \n",
       "6 -0.26904990    NA                  NA                    NA               \n",
       "7 -0.05839607    NA                  NA                    NA               \n",
       "8 -0.04793983    NA                  NA                    NA               \n",
       "  n_samples;Franke_V2 n_e_samples;Franke_V2 z_score;Franke_V2 n_samples;Wijst\n",
       "1 NA                  NA                    NA                NA             \n",
       "2 NA                  NA                    NA                NA             \n",
       "3 NA                  NA                    NA                NA             \n",
       "4 NA                  NA                    NA                NA             \n",
       "5 NA                  NA                    NA                NA             \n",
       "6 NA                  NA                    NA                NA             \n",
       "7 NA                  NA                    NA                NA             \n",
       "8 NA                  NA                    NA                NA             \n",
       "  n_e_samples;Wijst z_score;Wijst\n",
       "1 NA                NA           \n",
       "2 NA                NA           \n",
       "3 NA                NA           \n",
       "4 NA                NA           \n",
       "5 NA                NA           \n",
       "6 NA                NA           \n",
       "7 NA                NA           \n",
       "8 NA                NA           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_analysis[meta_analysis$feature_id ==  'IGHG2_VCPKMT',1:31]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b250b44e-866e-4539-988f-238e8dcc5ad0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "8f830ccb-5872-4a58-856c-e17b1c1f10b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Only for Mono/ B: filter additionally on eGenes within 5% FDR threshold\n",
    "meta_analysis = meta_analysis[meta_analysis$eGene %in% unique(snp_gene_mapping$gene),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "aab25c46-98c2-4b54-867a-cfa16a8f77cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "120536750"
      ],
      "text/latex": [
       "120536750"
      ],
      "text/markdown": [
       "120536750"
      ],
      "text/plain": [
       "[1] 120536750"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "89f25858-08c0-4558-a06f-5077d7bb7b16",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define triplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ec51a225-070a-4e26-930b-ef12ec13fa78",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_analysis$triplet = paste0( meta_analysis$snp_id,  ';', meta_analysis$feature_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c91858fa-0a86-4348-8fc6-a72071d18dac",
   "metadata": {},
   "source": [
    "### Define ribosomal and mitochondrial genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "22d96480-98cf-4232-9cd2-776379c4a88a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define rb/mt genes\n",
    "meta_analysis$is_rb_mt = 0\n",
    "meta_analysis$is_rb_mt[(meta_analysis$eGene %in% rb_mt_genes) |\n",
    "                            (meta_analysis$coeGene %in% rb_mt_genes)] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ef7a1830-7def-4db8-8ba5-5a7412938244",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the info about the amount of tests per diff gene-type and amount samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1e71b252-29d0-49fd-a5a2-1a5a0d44cf6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(meta_analysis,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2206ed8e-b772-41cf-a0d3-13a32174c323",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'closeGenes'</li><li>'coeGene'</li><li>'coQTL'</li><li>'eGene'</li><li>'feature_chromosome'</li><li>'feature_end'</li><li>'feature_id'</li><li>'feature_start'</li><li>'I_square'</li><li>'is_rb_mt'</li><li>'n_e_samples'</li><li>'n_e_samples;Franke_V2'</li><li>'n_e_samples;Franke_V3'</li><li>'n_e_samples;Multiome'</li><li>'n_e_samples;oneK1K'</li><li>'n_e_samples;Wijst'</li><li>'n_samples'</li><li>'n_samples;Franke_V2'</li><li>'n_samples;Franke_V3'</li><li>'n_samples;Multiome'</li><li>'n_samples;oneK1K'</li><li>'n_samples;Wijst'</li><li>'p_value'</li><li>'snp_id'</li><li>'snp_position'</li><li>'triplet'</li><li>'z_score'</li><li>'z_score_dataset_dev'</li><li>'z_score;Franke_V2'</li><li>'z_score;Franke_V3'</li><li>'z_score;Multiome'</li><li>'z_score;oneK1K'</li><li>'z_score;Wijst'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'closeGenes'\n",
       "\\item 'coeGene'\n",
       "\\item 'coQTL'\n",
       "\\item 'eGene'\n",
       "\\item 'feature\\_chromosome'\n",
       "\\item 'feature\\_end'\n",
       "\\item 'feature\\_id'\n",
       "\\item 'feature\\_start'\n",
       "\\item 'I\\_square'\n",
       "\\item 'is\\_rb\\_mt'\n",
       "\\item 'n\\_e\\_samples'\n",
       "\\item 'n\\_e\\_samples;Franke\\_V2'\n",
       "\\item 'n\\_e\\_samples;Franke\\_V3'\n",
       "\\item 'n\\_e\\_samples;Multiome'\n",
       "\\item 'n\\_e\\_samples;oneK1K'\n",
       "\\item 'n\\_e\\_samples;Wijst'\n",
       "\\item 'n\\_samples'\n",
       "\\item 'n\\_samples;Franke\\_V2'\n",
       "\\item 'n\\_samples;Franke\\_V3'\n",
       "\\item 'n\\_samples;Multiome'\n",
       "\\item 'n\\_samples;oneK1K'\n",
       "\\item 'n\\_samples;Wijst'\n",
       "\\item 'p\\_value'\n",
       "\\item 'snp\\_id'\n",
       "\\item 'snp\\_position'\n",
       "\\item 'triplet'\n",
       "\\item 'z\\_score'\n",
       "\\item 'z\\_score\\_dataset\\_dev'\n",
       "\\item 'z\\_score;Franke\\_V2'\n",
       "\\item 'z\\_score;Franke\\_V3'\n",
       "\\item 'z\\_score;Multiome'\n",
       "\\item 'z\\_score;oneK1K'\n",
       "\\item 'z\\_score;Wijst'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'closeGenes'\n",
       "2. 'coeGene'\n",
       "3. 'coQTL'\n",
       "4. 'eGene'\n",
       "5. 'feature_chromosome'\n",
       "6. 'feature_end'\n",
       "7. 'feature_id'\n",
       "8. 'feature_start'\n",
       "9. 'I_square'\n",
       "10. 'is_rb_mt'\n",
       "11. 'n_e_samples'\n",
       "12. 'n_e_samples;Franke_V2'\n",
       "13. 'n_e_samples;Franke_V3'\n",
       "14. 'n_e_samples;Multiome'\n",
       "15. 'n_e_samples;oneK1K'\n",
       "16. 'n_e_samples;Wijst'\n",
       "17. 'n_samples'\n",
       "18. 'n_samples;Franke_V2'\n",
       "19. 'n_samples;Franke_V3'\n",
       "20. 'n_samples;Multiome'\n",
       "21. 'n_samples;oneK1K'\n",
       "22. 'n_samples;Wijst'\n",
       "23. 'p_value'\n",
       "24. 'snp_id'\n",
       "25. 'snp_position'\n",
       "26. 'triplet'\n",
       "27. 'z_score'\n",
       "28. 'z_score_dataset_dev'\n",
       "29. 'z_score;Franke_V2'\n",
       "30. 'z_score;Franke_V3'\n",
       "31. 'z_score;Multiome'\n",
       "32. 'z_score;oneK1K'\n",
       "33. 'z_score;Wijst'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"closeGenes\"            \"coeGene\"               \"coQTL\"                \n",
       " [4] \"eGene\"                 \"feature_chromosome\"    \"feature_end\"          \n",
       " [7] \"feature_id\"            \"feature_start\"         \"I_square\"             \n",
       "[10] \"is_rb_mt\"              \"n_e_samples\"           \"n_e_samples;Franke_V2\"\n",
       "[13] \"n_e_samples;Franke_V3\" \"n_e_samples;Multiome\"  \"n_e_samples;oneK1K\"   \n",
       "[16] \"n_e_samples;Wijst\"     \"n_samples\"             \"n_samples;Franke_V2\"  \n",
       "[19] \"n_samples;Franke_V3\"   \"n_samples;Multiome\"    \"n_samples;oneK1K\"     \n",
       "[22] \"n_samples;Wijst\"       \"p_value\"               \"snp_id\"               \n",
       "[25] \"snp_position\"          \"triplet\"               \"z_score\"              \n",
       "[28] \"z_score_dataset_dev\"   \"z_score;Franke_V2\"     \"z_score;Franke_V3\"    \n",
       "[31] \"z_score;Multiome\"      \"z_score;oneK1K\"        \"z_score;Wijst\"        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(colnames(meta_analysis))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5f60615e-f17c-4552-a5d3-2ea741b7ce1c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'n_samples'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "amount_tests_save =meta_analysis %>% group_by( n_samples, is_rb_mt) %>% summarise(amount_tests_datasets = length(unique(triplet)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "013a6375-fa33-4e96-8d0e-b1c41979e97e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>n_samples</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_datasets</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>16</td><td>0</td><td>22</td></tr>\n",
       "\t<tr><td>16</td><td>1</td><td> 9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " n\\_samples & is\\_rb\\_mt & amount\\_tests\\_datasets\\\\\n",
       " <int> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 16 & 0 & 22\\\\\n",
       "\t 16 & 1 &  9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 3\n",
       "\n",
       "| n_samples &lt;int&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_datasets &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 16 | 0 | 22 |\n",
       "| 16 | 1 |  9 |\n",
       "\n"
      ],
      "text/plain": [
       "  n_samples is_rb_mt amount_tests_datasets\n",
       "1 16        0        22                   \n",
       "2 16        1         9                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(amount_tests_save,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5fdd2ce1-72f3-414f-961e-c797cd25c33d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2048"
      ],
      "text/latex": [
       "2048"
      ],
      "text/markdown": [
       "2048"
      ],
      "text/plain": [
       "[1] 2048"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(amount_tests_save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "313bee79-2aea-4f1b-9919-92af06ed0078",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(amount_tests_save, paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/','Meta_Info_Tests.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a29df234-0949-45ed-a461-b6c9ec88b790",
   "metadata": {},
   "source": [
    "### Filter on only the cis-eQTL snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7c5c0865-f875-4fc6-8305-5ca24611629f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "120536750"
      ],
      "text/latex": [
       "120536750"
      ],
      "text/markdown": [
       "120536750"
      ],
      "text/plain": [
       "[1] 120536750"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6bbe9008-d34e-4924-9a6b-4dec6a6df3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(snp_filter == 'cis'){\n",
    "    meta_analysis = meta_analysis[meta_analysis$snp_id %in% unique(snp_gene_mapping$snp),]\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c41f5ef3-6340-407a-ac34-34d10e569648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "120536750"
      ],
      "text/latex": [
       "120536750"
      ],
      "text/markdown": [
       "120536750"
      ],
      "text/plain": [
       "[1] 120536750"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1d3cd348-b96b-44f6-a855-e7d9910141a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "120536750"
      ],
      "text/latex": [
       "120536750"
      ],
      "text/markdown": [
       "120536750"
      ],
      "text/plain": [
       "[1] 120536750"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "f61a00da-b1d8-4b40-bd3b-c0d819c081d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save amount of tests after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a1416528-f380-4726-8d29-9191caa2ed34",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'n_samples'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "amount_tests_save =meta_analysis %>% group_by( n_samples, is_rb_mt) %>% summarise(amount_tests_datasets = length(unique(triplet)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "31ae957b-0022-41e2-83c0-d38cf34e364e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>n_samples</th><th scope=col>is_rb_mt</th><th scope=col>amount_tests_datasets</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>16</td><td>0</td><td>22</td></tr>\n",
       "\t<tr><td>16</td><td>1</td><td> 9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " n\\_samples & is\\_rb\\_mt & amount\\_tests\\_datasets\\\\\n",
       " <int> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 16 & 0 & 22\\\\\n",
       "\t 16 & 1 &  9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 3\n",
       "\n",
       "| n_samples &lt;int&gt; | is_rb_mt &lt;dbl&gt; | amount_tests_datasets &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| 16 | 0 | 22 |\n",
       "| 16 | 1 |  9 |\n",
       "\n"
      ],
      "text/plain": [
       "  n_samples is_rb_mt amount_tests_datasets\n",
       "1 16        0        22                   \n",
       "2 16        1         9                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(amount_tests_save,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "7c0a85f1-c191-4274-95aa-1690fc778691",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2048"
      ],
      "text/latex": [
       "2048"
      ],
      "text/markdown": [
       "2048"
      ],
      "text/plain": [
       "[1] 2048"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(amount_tests_save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "535cbb7f-c06e-4354-8255-24822b7c2266",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "''"
      ],
      "text/latex": [
       "''"
      ],
      "text/markdown": [
       "''"
      ],
      "text/plain": [
       "[1] \"\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "be4f1c43-4e6c-4ff4-9f8d-57078d78a3e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(amount_tests_save, paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/','Meta_Info_Tests_', snp_filter, '.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e474ce29-9ee7-4cbf-ae07-7ae53a504e5c",
   "metadata": {},
   "source": [
    "### Filter on minimum number of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "fc4c8803-6990-4273-8aa6-e257a6cd9d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if (n_samples_filter > 0){\n",
    "    meta_analysis = meta_analysis[meta_analysis$n_samples > n_samples_filter,]\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f2946d0c-5571-4871-b076-84b07bac810a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67453627"
      ],
      "text/latex": [
       "67453627"
      ],
      "text/markdown": [
       "67453627"
      ],
      "text/plain": [
       "[1] 67453627"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "aaa40ac0-3950-44d2-8c49-57999de83faa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "7908909"
      ],
      "text/latex": [
       "7908909"
      ],
      "text/markdown": [
       "7908909"
      ],
      "text/plain": [
       "[1] 7908909"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(meta_analysis$feature_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "6be38e59-6617-4116-9d84-d967b0fdef52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "7908909"
      ],
      "text/latex": [
       "7908909"
      ],
      "text/markdown": [
       "7908909"
      ],
      "text/plain": [
       "[1] 7908909"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(meta_analysis$feature_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "284d4a6b-40a8-4cff-9247-320a9cf28ac5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67422435"
      ],
      "text/latex": [
       "67422435"
      ],
      "text/markdown": [
       "67422435"
      ],
      "text/plain": [
       "[1] 67422435"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(meta_analysis$triplet))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ddab226d-62e3-4a0c-82be-32dc323163f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67422435"
      ],
      "text/latex": [
       "67422435"
      ],
      "text/markdown": [
       "67422435"
      ],
      "text/plain": [
       "[1] 67422435"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(meta_analysis$triplet))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "441d8dc6-caa3-46eb-92e5-99a822949ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "amount_tests = length(unique(meta_analysis$triplet))   # get the unique amount of tests for multiple testing correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "884b0627-3c7c-4004-af1c-c5dcc56c7259",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'data.table'</li><li>'data.frame'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'data.table'\n",
       "\\item 'data.frame'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'data.table'\n",
       "2. 'data.frame'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"data.table\" \"data.frame\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "eb414819-9bb3-44ac-b1f2-37f759053e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_analysis = setDT(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "9f8acb5e-6957-401b-99d2-416647ebb7cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'data.table'</li><li>'data.frame'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'data.table'\n",
       "\\item 'data.frame'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'data.table'\n",
       "2. 'data.frame'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"data.table\" \"data.frame\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "class(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8737ff5-86eb-4f25-a666-61a3ef53444b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5ccf9da6-6a8b-475e-a7f8-6acfecd8465d",
   "metadata": {},
   "source": [
    "### Check for duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6e6deecf-11d4-41b9-92a2-a79c6622c41f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dupl_invest =  meta_analysis[, .(.N), by = .(triplet)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "591e7584-a582-4950-9039-06deeaae6744",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>triplet</th><th scope=col>N</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>12:56007301:G:A;RPL13_RPS26</td><td>1</td></tr>\n",
       "\t<tr><td>12:56007301:G:A;RPL10_RPS26</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " triplet & N\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 12:56007301:G:A;RPL13\\_RPS26 & 1\\\\\n",
       "\t 12:56007301:G:A;RPL10\\_RPS26 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 2\n",
       "\n",
       "| triplet &lt;chr&gt; | N &lt;int&gt; |\n",
       "|---|---|\n",
       "| 12:56007301:G:A;RPL13_RPS26 | 1 |\n",
       "| 12:56007301:G:A;RPL10_RPS26 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  triplet                     N\n",
       "1 12:56007301:G:A;RPL13_RPS26 1\n",
       "2 12:56007301:G:A;RPL10_RPS26 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(dupl_invest,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "e3ed5dfa-01a1-4173-ab0e-3ca6f6540078",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>triplet</th><th scope=col>N</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>6:32658626:A:G;HLA-DQA1_HLA-DRB1</td><td>2</td></tr>\n",
       "\t<tr><td>6:32632280:A:G;HLA-DQA1_HLA-DRB1</td><td>2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " triplet & N\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 6:32658626:A:G;HLA-DQA1\\_HLA-DRB1 & 2\\\\\n",
       "\t 6:32632280:A:G;HLA-DQA1\\_HLA-DRB1 & 2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 2\n",
       "\n",
       "| triplet &lt;chr&gt; | N &lt;int&gt; |\n",
       "|---|---|\n",
       "| 6:32658626:A:G;HLA-DQA1_HLA-DRB1 | 2 |\n",
       "| 6:32632280:A:G;HLA-DQA1_HLA-DRB1 | 2 |\n",
       "\n"
      ],
      "text/plain": [
       "  triplet                          N\n",
       "1 6:32658626:A:G;HLA-DQA1_HLA-DRB1 2\n",
       "2 6:32632280:A:G;HLA-DQA1_HLA-DRB1 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(dupl_invest %>% arrange(-N),2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "8e0009a4-c609-49a6-a7b5-59e532181444",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "31192"
      ],
      "text/latex": [
       "31192"
      ],
      "text/markdown": [
       "31192"
      ],
      "text/plain": [
       "[1] 31192"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(dupl_invest[dupl_invest$N == 2,])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b8ed2f16-d018-493c-8803-ac68ab60a15a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "31192"
      ],
      "text/latex": [
       "31192"
      ],
      "text/markdown": [
       "31192"
      ],
      "text/plain": [
       "[1] 31192"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(dupl_invest[dupl_invest$N == 2,])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "cc6e0f69-b124-462f-8c64-6d952c82c8f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dupl_invest =dupl_invest[dupl_invest$N == 2,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "27969e71-fe47-426f-b3c5-45037ebdbd38",
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicates = meta_analysis[meta_analysis$triplet %in% dupl_invest$triplet,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "f407dfd4-2aae-4026-b0c8-696b8f627506",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "62384"
      ],
      "text/latex": [
       "62384"
      ],
      "text/markdown": [
       "62384"
      ],
      "text/plain": [
       "[1] 62384"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(duplicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "50a46c73-36b5-43dc-846c-0bbcbb500fdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(duplicates$closeGenes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5af33211-a37a-4722-b992-8dfeb4ae324d",
   "metadata": {},
   "source": [
    "### Multiple testing correctin and get significant results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "63450626-cb9b-4343-9b9f-b7f210814fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add p-value with Bonferroni correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "bed49d78-7fd0-4452-b00f-559555b15d1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67422435"
      ],
      "text/latex": [
       "67422435"
      ],
      "text/markdown": [
       "67422435"
      ],
      "text/plain": [
       "[1] 67422435"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "amount_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "a04255c8-3dc4-4f4b-a9e0-a38b5ac652e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67422435"
      ],
      "text/latex": [
       "67422435"
      ],
      "text/markdown": [
       "67422435"
      ],
      "text/plain": [
       "[1] 67422435"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "amount_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b2d3df0-bf7f-4d38-a14e-9048c8455beb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "0d154171-9c17-4dec-b50a-e8788037e289",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_analysis$p_val_mt = as.numeric(meta_analysis$p_value) * amount_tests\n",
    "meta_analysis$q_value = qvalue(p = meta_analysis$p_value)$qvalues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d3fdde9c-1ebe-40a8-9078-854328559f56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67453627"
      ],
      "text/latex": [
       "67453627"
      ],
      "text/markdown": [
       "67453627"
      ],
      "text/plain": [
       "[1] 67453627"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "49ac5a50-4a57-454f-aee5-02225b419b16",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Step-wise multiple testing correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "5e63a461-e869-43b8-ace2-890d37307961",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the amount of tests per eGene\n",
    "meta_analysis = meta_analysis %>% group_by(eGene) %>% mutate(amount_tests_eGene = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "a1dd372e-7517-454a-bdd9-1ee3513dfbb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 36</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>coQTL</th><th scope=col>eGene</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>coeGene</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>snp_position</th><th scope=col>closeGenes</th><th scope=col>⋯</th><th scope=col>n_e_samples;Franke_V2</th><th scope=col>z_score;Franke_V2</th><th scope=col>n_samples;Wijst</th><th scope=col>n_e_samples;Wijst</th><th scope=col>z_score;Wijst</th><th scope=col>is_rb_mt</th><th scope=col>triplet</th><th scope=col>p_val_mt</th><th scope=col>q_value</th><th scope=col>amount_tests_eGene</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>12:56007301:G:A;RPS26_RPL13</td><td>RPS26</td><td>12</td><td>56041351</td><td>56044676</td><td>RPL13</td><td>RPL13_RPS26</td><td>12:56007301:G:A</td><td>56007301</td><td>FALSE</td><td>⋯</td><td>91</td><td>9.331416</td><td>40</td><td>40</td><td>5.272181</td><td>1</td><td>12:56007301:G:A;RPL13_RPS26</td><td>2.030599e-130</td><td>1.983404e-130</td><td>143330</td></tr>\n",
       "\t<tr><td>12:56007301:G:A;RPS26_RPL10</td><td>RPS26</td><td>12</td><td>56041351</td><td>56044676</td><td>RPL10</td><td>RPL10_RPS26</td><td>12:56007301:G:A</td><td>56007301</td><td>FALSE</td><td>⋯</td><td>92</td><td>9.062418</td><td>40</td><td>40</td><td>3.646652</td><td>1</td><td>12:56007301:G:A;RPL10_RPS26</td><td>1.597442e-118</td><td>7.801572e-119</td><td>143330</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 36\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " coQTL & eGene & feature\\_chromosome & feature\\_start & feature\\_end & coeGene & feature\\_id & snp\\_id & snp\\_position & closeGenes & ⋯ & n\\_e\\_samples;Franke\\_V2 & z\\_score;Franke\\_V2 & n\\_samples;Wijst & n\\_e\\_samples;Wijst & z\\_score;Wijst & is\\_rb\\_mt & triplet & p\\_val\\_mt & q\\_value & amount\\_tests\\_eGene\\\\\n",
       " <chr> & <chr> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <lgl> & ⋯ & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 12:56007301:G:A;RPS26\\_RPL13 & RPS26 & 12 & 56041351 & 56044676 & RPL13 & RPL13\\_RPS26 & 12:56007301:G:A & 56007301 & FALSE & ⋯ & 91 & 9.331416 & 40 & 40 & 5.272181 & 1 & 12:56007301:G:A;RPL13\\_RPS26 & 2.030599e-130 & 1.983404e-130 & 143330\\\\\n",
       "\t 12:56007301:G:A;RPS26\\_RPL10 & RPS26 & 12 & 56041351 & 56044676 & RPL10 & RPL10\\_RPS26 & 12:56007301:G:A & 56007301 & FALSE & ⋯ & 92 & 9.062418 & 40 & 40 & 3.646652 & 1 & 12:56007301:G:A;RPL10\\_RPS26 & 1.597442e-118 & 7.801572e-119 & 143330\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 36\n",
       "\n",
       "| coQTL &lt;chr&gt; | eGene &lt;chr&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | coeGene &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | snp_position &lt;int&gt; | closeGenes &lt;lgl&gt; | ⋯ ⋯ | n_e_samples;Franke_V2 &lt;int&gt; | z_score;Franke_V2 &lt;dbl&gt; | n_samples;Wijst &lt;int&gt; | n_e_samples;Wijst &lt;int&gt; | z_score;Wijst &lt;dbl&gt; | is_rb_mt &lt;dbl&gt; | triplet &lt;chr&gt; | p_val_mt &lt;dbl&gt; | q_value &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 12:56007301:G:A;RPS26_RPL13 | RPS26 | 12 | 56041351 | 56044676 | RPL13 | RPL13_RPS26 | 12:56007301:G:A | 56007301 | FALSE | ⋯ | 91 | 9.331416 | 40 | 40 | 5.272181 | 1 | 12:56007301:G:A;RPL13_RPS26 | 2.030599e-130 | 1.983404e-130 | 143330 |\n",
       "| 12:56007301:G:A;RPS26_RPL10 | RPS26 | 12 | 56041351 | 56044676 | RPL10 | RPL10_RPS26 | 12:56007301:G:A | 56007301 | FALSE | ⋯ | 92 | 9.062418 | 40 | 40 | 3.646652 | 1 | 12:56007301:G:A;RPL10_RPS26 | 1.597442e-118 | 7.801572e-119 | 143330 |\n",
       "\n"
      ],
      "text/plain": [
       "  coQTL                       eGene feature_chromosome feature_start\n",
       "1 12:56007301:G:A;RPS26_RPL13 RPS26 12                 56041351     \n",
       "2 12:56007301:G:A;RPS26_RPL10 RPS26 12                 56041351     \n",
       "  feature_end coeGene feature_id  snp_id          snp_position closeGenes ⋯\n",
       "1 56044676    RPL13   RPL13_RPS26 12:56007301:G:A 56007301     FALSE      ⋯\n",
       "2 56044676    RPL10   RPL10_RPS26 12:56007301:G:A 56007301     FALSE      ⋯\n",
       "  n_e_samples;Franke_V2 z_score;Franke_V2 n_samples;Wijst n_e_samples;Wijst\n",
       "1 91                    9.331416          40              40               \n",
       "2 92                    9.062418          40              40               \n",
       "  z_score;Wijst is_rb_mt triplet                     p_val_mt     \n",
       "1 5.272181      1        12:56007301:G:A;RPL13_RPS26 2.030599e-130\n",
       "2 3.646652      1        12:56007301:G:A;RPL10_RPS26 1.597442e-118\n",
       "  q_value       amount_tests_eGene\n",
       "1 1.983404e-130 143330            \n",
       "2 7.801572e-119 143330            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(meta_analysis,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "2a3e9b0a-de01-4310-8bdb-47097b4156bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Correct for multiple testing per eGene\n",
    "meta_analysis$p_val_mt_eGene = meta_analysis$p_value * meta_analysis$amount_tests_eGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "50426e16-e461-4971-abc2-639cf005674c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3.01175555847339e-138"
      ],
      "text/latex": [
       "3.01175555847339e-138"
      ],
      "text/markdown": [
       "3.01175555847339e-138"
      ],
      "text/plain": [
       "[1] 3.011756e-138"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1"
      ],
      "text/latex": [
       "1"
      ],
      "text/markdown": [
       "1"
      ],
      "text/plain": [
       "[1] 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "min(meta_analysis$p_value)\n",
    "max(meta_analysis$p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "417fb417-2393-437f-85ac-d1a20350362e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1417"
      ],
      "text/latex": [
       "1417"
      ],
      "text/markdown": [
       "1417"
      ],
      "text/plain": [
       "[1] 1417"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(meta_analysis$eGene))   # amount of eGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "7ef1d32c-fb53-49f1-85cc-455f6a5eaebe",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the top effect per eGene\n",
    "top_eGene_effects = meta_analysis %>% group_by(eGene) %>% dplyr::filter(p_value == min(p_value)) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "f1b4c0bb-af66-45ae-a744-b1eccb2ff15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_eGene_effects = top_eGene_effects %>% group_by(eGene) %>% slice_min(p_value, with_ties = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "75eed6c4-83bf-4dc2-9128-83fb79e70a7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1417"
      ],
      "text/latex": [
       "1417"
      ],
      "text/markdown": [
       "1417"
      ],
      "text/plain": [
       "[1] 1417"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(top_eGene_effects)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "17dd77a8-a82a-4238-b043-727330171f46",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set values greater 1 to 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "c45352cb-eff6-498c-9418-e6167f0bdb2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_eGene_effects$p_val_mt_eGene[top_eGene_effects$p_val_mt_eGene > 1] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "48f2feeb-1fb1-4257-bc2c-33ab45e162cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Correct top-effects with q_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "7397456d-7f44-4898-96de-f8253168f66c",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_eGene_effects$q_val = qvalue(p = top_eGene_effects$p_val_mt_eGene)$qvalues\n",
    "\n",
    "#top_eGene_effects$p_bh = qvalue(p = top_eGene_effects$p_val_mt_eGene,pi0 = 1)$qvalues\n",
    "\n",
    "top_eGene_effects$p_bh =  p.adjust(top_eGene_effects$p_val_mt_eGene, method = \"BH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "890010aa-a068-457d-8a38-16337a4671d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 39</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>coQTL</th><th scope=col>eGene</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>coeGene</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>snp_position</th><th scope=col>closeGenes</th><th scope=col>⋯</th><th scope=col>n_e_samples;Wijst</th><th scope=col>z_score;Wijst</th><th scope=col>is_rb_mt</th><th scope=col>triplet</th><th scope=col>p_val_mt</th><th scope=col>q_value</th><th scope=col>amount_tests_eGene</th><th scope=col>p_val_mt_eGene</th><th scope=col>q_val</th><th scope=col>p_bh</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>19:58358721:TG:T;A1BG_POLE3</td><td>A1BG </td><td>19</td><td>58345178</td><td>58353499</td><td>POLE3</td><td>A1BG_POLE3</td><td>19:58358721:TG:T</td><td>58358721</td><td>FALSE</td><td>⋯</td><td>NA</td><td>NA</td><td>0</td><td>19:58358721:TG:T;A1BG_POLE3</td><td> 1635.152</td><td>0.07942455</td><td>19688</td><td>0.4774802</td><td>0.8775479</td><td>0.8775479</td></tr>\n",
       "\t<tr><td>4:56348606:C:T;AASDH_PTEN  </td><td>AASDH</td><td> 4</td><td>56338287</td><td>56387508</td><td>PTEN </td><td>AASDH_PTEN</td><td>4:56348606:C:T  </td><td>56348606</td><td>FALSE</td><td>⋯</td><td>NA</td><td>NA</td><td>0</td><td>4:56348606:C:T;AASDH_PTEN  </td><td>15414.861</td><td>0.26287243</td><td> 3591</td><td>0.8210140</td><td>1.0000000</td><td>1.0000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 39\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " coQTL & eGene & feature\\_chromosome & feature\\_start & feature\\_end & coeGene & feature\\_id & snp\\_id & snp\\_position & closeGenes & ⋯ & n\\_e\\_samples;Wijst & z\\_score;Wijst & is\\_rb\\_mt & triplet & p\\_val\\_mt & q\\_value & amount\\_tests\\_eGene & p\\_val\\_mt\\_eGene & q\\_val & p\\_bh\\\\\n",
       " <chr> & <chr> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <lgl> & ⋯ & <int> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 19:58358721:TG:T;A1BG\\_POLE3 & A1BG  & 19 & 58345178 & 58353499 & POLE3 & A1BG\\_POLE3 & 19:58358721:TG:T & 58358721 & FALSE & ⋯ & NA & NA & 0 & 19:58358721:TG:T;A1BG\\_POLE3 &  1635.152 & 0.07942455 & 19688 & 0.4774802 & 0.8775479 & 0.8775479\\\\\n",
       "\t 4:56348606:C:T;AASDH\\_PTEN   & AASDH &  4 & 56338287 & 56387508 & PTEN  & AASDH\\_PTEN & 4:56348606:C:T   & 56348606 & FALSE & ⋯ & NA & NA & 0 & 4:56348606:C:T;AASDH\\_PTEN   & 15414.861 & 0.26287243 &  3591 & 0.8210140 & 1.0000000 & 1.0000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 39\n",
       "\n",
       "| coQTL &lt;chr&gt; | eGene &lt;chr&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | coeGene &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | snp_position &lt;int&gt; | closeGenes &lt;lgl&gt; | ⋯ ⋯ | n_e_samples;Wijst &lt;int&gt; | z_score;Wijst &lt;dbl&gt; | is_rb_mt &lt;dbl&gt; | triplet &lt;chr&gt; | p_val_mt &lt;dbl&gt; | q_value &lt;dbl&gt; | amount_tests_eGene &lt;int&gt; | p_val_mt_eGene &lt;dbl&gt; | q_val &lt;dbl&gt; | p_bh &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 19:58358721:TG:T;A1BG_POLE3 | A1BG  | 19 | 58345178 | 58353499 | POLE3 | A1BG_POLE3 | 19:58358721:TG:T | 58358721 | FALSE | ⋯ | NA | NA | 0 | 19:58358721:TG:T;A1BG_POLE3 |  1635.152 | 0.07942455 | 19688 | 0.4774802 | 0.8775479 | 0.8775479 |\n",
       "| 4:56348606:C:T;AASDH_PTEN   | AASDH |  4 | 56338287 | 56387508 | PTEN  | AASDH_PTEN | 4:56348606:C:T   | 56348606 | FALSE | ⋯ | NA | NA | 0 | 4:56348606:C:T;AASDH_PTEN   | 15414.861 | 0.26287243 |  3591 | 0.8210140 | 1.0000000 | 1.0000000 |\n",
       "\n"
      ],
      "text/plain": [
       "  coQTL                       eGene feature_chromosome feature_start\n",
       "1 19:58358721:TG:T;A1BG_POLE3 A1BG  19                 58345178     \n",
       "2 4:56348606:C:T;AASDH_PTEN   AASDH  4                 56338287     \n",
       "  feature_end coeGene feature_id snp_id           snp_position closeGenes ⋯\n",
       "1 58353499    POLE3   A1BG_POLE3 19:58358721:TG:T 58358721     FALSE      ⋯\n",
       "2 56387508    PTEN    AASDH_PTEN 4:56348606:C:T   56348606     FALSE      ⋯\n",
       "  n_e_samples;Wijst z_score;Wijst is_rb_mt triplet                    \n",
       "1 NA                NA            0        19:58358721:TG:T;A1BG_POLE3\n",
       "2 NA                NA            0        4:56348606:C:T;AASDH_PTEN  \n",
       "  p_val_mt  q_value    amount_tests_eGene p_val_mt_eGene q_val     p_bh     \n",
       "1  1635.152 0.07942455 19688              0.4774802      0.8775479 0.8775479\n",
       "2 15414.861 0.26287243  3591              0.8210140      1.0000000 1.0000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(top_eGene_effects,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ad3d036b-479f-4a3a-85f6-f64aac3dae7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_val_thres = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "7a4f3763-e7d6-4871-a33a-644986479704",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the cutoff p-value\n",
    "p_val_cutoff_bh = head(top_eGene_effects[top_eGene_effects$p_bh < p_val_thres,c('p_val_mt_eGene', 'p_bh')] %>% arrange(-p_bh),1)\n",
    "p_val_cutoff_q = head(top_eGene_effects[top_eGene_effects$q_val < p_val_thres,c('p_val_mt_eGene', 'q_val')] %>% arrange(-q_val),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "11d3a0ee-a98c-4d6c-854f-1b8c77bcc502",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>p_val_mt_eGene</th><th scope=col>p_bh</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.002149966</td><td>0.04406373</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 2\n",
       "\\begin{tabular}{ll}\n",
       " p\\_val\\_mt\\_eGene & p\\_bh\\\\\n",
       " <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 0.002149966 & 0.04406373\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 2\n",
       "\n",
       "| p_val_mt_eGene &lt;dbl&gt; | p_bh &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| 0.002149966 | 0.04406373 |\n",
       "\n"
      ],
      "text/plain": [
       "  p_val_mt_eGene p_bh      \n",
       "1 0.002149966    0.04406373"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "p_val_cutoff_bh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "a4314c4c-ce76-4ae1-bf97-be3d10e61ca2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>p_val_mt_eGene</th><th scope=col>q_val</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0.002149966</td><td>0.04406373</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 × 2\n",
       "\\begin{tabular}{ll}\n",
       " p\\_val\\_mt\\_eGene & q\\_val\\\\\n",
       " <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 0.002149966 & 0.04406373\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 × 2\n",
       "\n",
       "| p_val_mt_eGene &lt;dbl&gt; | q_val &lt;dbl&gt; |\n",
       "|---|---|\n",
       "| 0.002149966 | 0.04406373 |\n",
       "\n"
      ],
      "text/plain": [
       "  p_val_mt_eGene q_val     \n",
       "1 0.002149966    0.04406373"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "p_val_cutoff_q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "ad5757a3-6208-452e-b958-42eea5aaa241",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add the cutoff p-value to the result\n",
    "meta_analysis$p_bh_cutoff = p_val_cutoff_bh$p_val_mt_eGene\n",
    "meta_analysis$p_q_cutoff = p_val_cutoff_q$p_val_mt_eGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "87b4114c-dc17-490c-b493-accd15a65497",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "67453627"
      ],
      "text/latex": [
       "67453627"
      ],
      "text/markdown": [
       "67453627"
      ],
      "text/plain": [
       "[1] 67453627"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(meta_analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "569b4fff-7d34-4e23-abc1-3fa2317748b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get nominal significant results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "baa03206-38a2-4271-b2d4-d70ec1638239",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_meta = meta_analysis[meta_analysis$p_value < 0.05,]\n",
    "#sign_meta_nom =  meta_analysis[meta_analysis$p_value < 0.05,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "f9c7215a-d85c-4595-b5ec-94ac28cdf888",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3906619"
      ],
      "text/latex": [
       "3906619"
      ],
      "text/markdown": [
       "3906619"
      ],
      "text/plain": [
       "[1] 3906619"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "b3800642-ef0e-46ed-b517-8aa38408866c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Summarise and mark the duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "72dc1e33-f778-4b0f-9f4d-90e589d332bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(nrow(dupl_invest)  > 0){\n",
    "    ## Previous solution\n",
    "    # coeqtl_results = coeqtl_results %>% group_by(feature_id, snp_id,   cell_type, dataset, feature_chromosome, feature_end, feature_start) %>% summarise(beta_se = max(beta_se), p_value = mean(p_value),beta = mean(beta)\n",
    "    \n",
    "    sign_meta_un = sign_meta[!sign_meta$triplet %in% dupl_invest$triplet,]\n",
    "    sign_meta_un$type = 'unique'\n",
    "    \n",
    "    sign_meta_dupl = sign_meta[sign_meta$triplet %in% dupl_invest$triplet,]\n",
    "    sign_meta_dupl$eGene = sign_meta_dupl$feature_id\n",
    "    sign_meta_dupl$coeGene = sign_meta_dupl$feature_id\n",
    "    sign_meta_dupl$type = 'duplicate'\n",
    "    \n",
    "    sign_meta_dupl=    sign_meta_dupl %>% group_by(triplet) %>% mutate(rank = rank(eGene, ties.method = 'first'))\n",
    "    sign_meta_dupl =  sign_meta_dupl[ sign_meta_dupl$rank == 1,]\n",
    "    sign_meta_dupl$rank = NULL\n",
    "    \n",
    "    sign_meta = rbind(sign_meta_un, sign_meta_dupl)\n",
    "    \n",
    "    \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "27dc1936-4c77-4f0d-a183-d46533afce70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "7589"
      ],
      "text/latex": [
       "7589"
      ],
      "text/markdown": [
       "7589"
      ],
      "text/plain": [
       "[1] 7589"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_meta$coeGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "ebab4738-4e40-4d0c-847e-5bef1a919c4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1955"
      ],
      "text/latex": [
       "1955"
      ],
      "text/markdown": [
       "1955"
      ],
      "text/plain": [
       "[1] 1955"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_meta$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "e42fdc79-9d6f-4cb6-aa48-f1c33be72cab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1514078"
      ],
      "text/latex": [
       "1514078"
      ],
      "text/markdown": [
       "1514078"
      ],
      "text/plain": [
       "[1] 1514078"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_meta$feature_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "5984f1eb-1c86-4c0c-b5ff-c0146d7d9002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3903170"
      ],
      "text/latex": [
       "3903170"
      ],
      "text/markdown": [
       "3903170"
      ],
      "text/plain": [
       "[1] 3903170"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "d8d2557b-7e73-4324-af55-64306d708da1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.0499999969036516"
      ],
      "text/latex": [
       "0.0499999969036516"
      ],
      "text/markdown": [
       "0.0499999969036516"
      ],
      "text/plain": [
       "[1] 0.05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "max(sign_meta$p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88fd02ce-4377-4dd5-826d-f5cd0ddadaa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the nominally significant results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "6a81d080-4140-4636-90ac-a7229abc74e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(sign_meta, paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/','Meta_Analysis_Sign_results', n_samples_filter ,'_', snp_filter,p_val_thres,  '.csv'))\n",
    "#write.csv(sign_meta_nom, paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/','Meta_Analysis_Sign_results_nom', n_samples_filter ,'.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "8ba607c4-ca30-4925-81e1-686dd46af35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate info about snps giving a co-eqtl and not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "e73d616c-f965-45f9-b86e-6050b845023c",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_analysis$is_coeqtl = 0\n",
    "meta_analysis$is_coeqtl[meta_analysis$p_val_mt < 0.05] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "a1e55fdb-113a-44f8-8d76-c8b2a40be823",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_coeqtl = meta_analysis %>% group_by(snp_id) %>% summarise(is_coeqtl = max(is_coeqtl))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "ea2de349-9e5f-4b09-8a5c-d1e875318ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(snp_coeqtl, paste0(data_path,coeqtl_results_path , '/meta_analysis/', meta_analysis_type, '/', cell_type, '/', 'SNP_Class_', cell_type, '_', meta_analysis_type, '_', snp_filter,p_val_thres, '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ef07f78-991a-4191-aa79-68753268acac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
