{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "03e17e1c-954c-46f7-9f1e-b4b57c124d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load results from co-eQTL pipeline and save as .csv for further downstream analyses\n",
    "## Filter co-eQTL testing files, e.g. all tests for significant gene-pairs, eGenes etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a867967-b7e2-419a-98e7-c6c8a93d4403",
   "metadata": {},
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5170b547-4118-419e-b3dd-e4afc2986459",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from matplotlib import pyplot as plt\n",
    "from scipy.sparse import issparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8ca49572-9af1-4f10-89ed-0999daaccd21",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "efa9c540-0903-4402-9c2e-def8574cbf5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "13116102-c025-42ca-aa2b-bc5949b2231b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import mean_squared_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "54008b48-506e-4eb9-b21f-28c11e85e716",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1464b356-9bd8-4db1-b268-d8c477611d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.datasets import load_iris\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.utils import Bunch\n",
    "from sklearn.tree import plot_tree\n",
    "from sklearn.feature_selection import mutual_info_classif\n",
    "import pandas as pd \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b707005c-5ed5-4b73-9441-0481149dfb05",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b4af597d-e51f-4330-a9af-898123c7e5c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ffda82aa-7ac2-4197-b8c6-c016532a18dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the regressor \n",
    "from sklearn.tree import DecisionTreeRegressor "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "48845181-b018-494b-839e-c3d3e8bbc010",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07b96e3d-ac8b-4d30-9deb-4fe39f620c23",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "396b7c39-544a-4b4b-b7de-c67a285d6db5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e3b1713b-e8af-4e03-8f6a-715ef018f982",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "c52e1239-3f96-41aa-a8cf-2dd453546ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### which co-eqtl mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "0c2965ef-f018-4742-a2d9-cdeed74a5f95",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results_path = 'co_qtls_decision_tree/'   # path to final mapping results (based on decision tree filter)\n",
    "coeqtl_results_path_old = 'co_qtls_sceqtlgen/'   # path to initial evaluation phase co-eQTL results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4ca4e8c2-b3ce-45e3-a270-b7cbeb5bbe3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the co-EQTL Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "96795423-2c6b-4486-b6ca-9420b0bb6569",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data/current/coeqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbabfb90-f469-40d1-99ae-73e9e891e07b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the eQTL files from sceQTLGen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4e802037-3e51-4ca4-8ff7-bf6a921211eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl = '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "1424fdd8-7c73-4592-853b-fc6309a757c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_results = 'sc_eqtlgen_final'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6397c36d-a14e-42e7-9901-1f6484a0379f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to store results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a822e11c-2e37-45af-82e6-178d051a1541",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path = '../results/current/F3/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "7bfef69c-d84f-417b-a958-f6af6801dcfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path_analysis =  \"../data/current/coeqtl_mapping/\" + coeqtl_results_path + \"analysis\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48f8f682-7486-43a4-b67d-694d3eef93d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to meta-data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "84089784-827e-4e3a-a750-b98e2cafaa19",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_meta = '../data/current/meta-data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61916a50-4652-4531-b71e-83c24a96ad97",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for determination of significant co-eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "a13f279e-6dc9-4c80-b8ef-2e66af2dbc97",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_value_sign = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "6e294e28-a024-4842-96d5-8eace1314245",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_sample_filter = 133"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "75dd4859-c027-4630-92f7-231ba0d9485e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_correction = 'eGene_bf' # eGene_bf, p_val_mt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "80dc8133-80df-462d-a951-e628f40551dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_filter = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f43c2550-f421-47cb-958c-d35e0ad27bb3",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21e34509-c8de-4bfe-a0e9-5785cac7e3e8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load the significant co-eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec414394-16ad-46e3-a84d-9b7c667c1bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the nominally significant co-QTLs from each single-cell dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "dfaa27d9-a319-4ee0-a9b6-8a27a2a5b0c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define cell-types and datasets for which to execute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "c79eece1-ad31-4698-90f1-d80878c7e44f",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = ['oneK1K']\n",
    "\n",
    "#datasets_to_load = ['oneK1K', 'Franke_v2'  , 'Franke_v3', 'wijst', 'multiome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d4037dba-1959-4dac-8b1d-ab5d06be501a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = ['NK', 'CD4_T', 'CD8_T', 'Mono', 'B']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b40459e6-68be-486a-b7fb-47255ca74298",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-03-14 16:52:36\n",
      "2025-03-19 11:34:45\n",
      "2025-03-14 14:48:45\n",
      "2025-03-14 14:44:43\n",
      "2025-03-14 15:52:22\n"
     ]
    }
   ],
   "source": [
    "# Initialize an empty DataFrame to store the results\n",
    "coeqtl_results = pd.DataFrame()\n",
    "\n",
    "# Loop through each dataset and cell type\n",
    "for i in datasets_to_load:\n",
    "    for j in cell_types:\n",
    "        # Construct the file path\n",
    "        if n_sample_filter > 0:\n",
    "            n_sample_filter_param = 'TRUE'\n",
    "        if n_sample_filter == 0:\n",
    "            n_sample_filter_param = 'FALSE'\n",
    "        #file_path = f\"{result_path_analysis}_{i}/{j}/F3_Co_E_Genes\" + n_sample_filter_param + '_' + snp_filter + \".csv\"\n",
    "        file_path = f\"{result_path_analysis}_{i}/{j}/F3_Co_E_Genes\" + n_sample_filter_param +  \".csv\"\n",
    "        print(datetime.datetime.fromtimestamp(os.path.getmtime(file_path)))\n",
    "        \n",
    "        # Check if the file exists\n",
    "        if os.path.exists(file_path):\n",
    "            # Read the CSV file into a DataFrame\n",
    "            result_file = pd.read_csv(file_path)\n",
    "            \n",
    "            # Add the dataset and cell_type columns\n",
    "            result_file['dataset'] = i\n",
    "            result_file['cell_type'] = j\n",
    "            \n",
    "            # Append the result to the coeqtl_results DataFrame\n",
    "            coeqtl_results = pd.concat([coeqtl_results, result_file], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a7642d0a-b720-4d41-9642-e9561004a821",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter the file based on multiple testing correction method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ed4da171-521c-4e89-8f39-367c54c25c87",
   "metadata": {},
   "outputs": [],
   "source": [
    "if mt_correction  == 'eGene_bf':\n",
    "    coeqtl_results = coeqtl_results[coeqtl_results['p_val_mt_eGene'] < coeqtl_results['p_q_cutoff']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "91335715-697f-49f8-9112-4d27d3e17c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter only bonferroni significant results\n",
    "if mt_correction  != 'eGene_bf':\n",
    "    coeqtl_results = coeqtl_results[coeqtl_results['p_val_mt'] < 0.05]  # how to define significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ecc97c-e3c7-4cab-9a72-f6bece7ed14e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "dc41b3df-01de-4a22-93c7-9585ec006d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load the significant co-eQTLs from the meta-analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "231c6bd8-e113-4db5-a7c8-2db7b691cdf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-10-15 10:47:28\n",
      "2025-02-12 16:10:20\n",
      "2024-10-15 11:00:27\n",
      "2025-02-12 17:13:29\n",
      "2024-10-15 11:39:25\n"
     ]
    }
   ],
   "source": [
    "# Initialize an empty DataFrame to store the results\n",
    "meta_results = pd.DataFrame()\n",
    "cell_types = [ 'NK', 'B', 'CD8_T', 'Mono', 'CD4_T']\n",
    "\n",
    "# Loop through each dataset and cell type\n",
    "\n",
    "for j in cell_types:\n",
    "    # Construct the file path\n",
    "    #file_path =  \"../data/current/coeqtl_mapping/\" + coeqtl_results_path +   '/meta_analysis/' + 'meta_5ds/' + j + '/Meta_Analysis_Sign_results' +  str(n_sample_filter) + '_' + snp_filter +  '.csv' \n",
    "    file_path =  \"../data/current/coeqtl_mapping/\" + coeqtl_results_path +   '/meta_analysis/' + 'meta_5ds/' + j + '/Meta_Analysis_Sign_results' +  str(n_sample_filter) +   '.csv' \n",
    "\n",
    "    # Check if the file exists\n",
    "    if os.path.exists(file_path):\n",
    "        # Read the CSV file into a DataFrame\n",
    "        result_file = pd.read_csv(file_path)\n",
    "        print(datetime.datetime.fromtimestamp(os.path.getmtime(file_path)))\n",
    "\n",
    "        # Add the dataset and cell_type columns\n",
    "        result_file['cell_type'] = j\n",
    "\n",
    "        # Append the result to the coeqtl_results DataFrame\n",
    "        meta_results = pd.concat([meta_results, result_file], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45dd28d4-7095-445b-afbe-07b6b1fe04ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter meta-analysis results based on multiple correction p-value threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b369c877-8b9a-4b2d-91d6-a2683f4f1611",
   "metadata": {},
   "outputs": [],
   "source": [
    "if mt_correction  == 'eGene_bf':\n",
    "    meta_results = meta_results[meta_results['p_val_mt_eGene'] < meta_results['p_q_cutoff']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "f711dbe0-ac6d-4343-909c-7bcda273acd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>coQTL</th>\n",
       "      <th>eGene</th>\n",
       "      <th>feature_chromosome</th>\n",
       "      <th>feature_start</th>\n",
       "      <th>feature_end</th>\n",
       "      <th>coeGene</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>...</th>\n",
       "      <th>triplet</th>\n",
       "      <th>is_rb_mt</th>\n",
       "      <th>p_val_mt</th>\n",
       "      <th>amount_tests_eGene</th>\n",
       "      <th>p_val_mt_eGene</th>\n",
       "      <th>p_bh_cutoff</th>\n",
       "      <th>p_q_cutoff</th>\n",
       "      <th>type</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>q_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12:56007301:G:A;RPS26_RPL41</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>RPL41</td>\n",
       "      <td>RPL41_RPS26</td>\n",
       "      <td>12:56007301:G:A</td>\n",
       "      <td>56007301</td>\n",
       "      <td>...</td>\n",
       "      <td>12:56007301:G:A;RPL41_RPS26</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>172827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>unique</td>\n",
       "      <td>NK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>12:56007301:G:A;RPS26_RPS27</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>RPS27</td>\n",
       "      <td>RPS26_RPS27</td>\n",
       "      <td>12:56007301:G:A</td>\n",
       "      <td>56007301</td>\n",
       "      <td>...</td>\n",
       "      <td>12:56007301:G:A;RPS26_RPS27</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>172827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>unique</td>\n",
       "      <td>NK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>12:56007301:G:A;RPS26_RPS29</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>RPS29</td>\n",
       "      <td>RPS26_RPS29</td>\n",
       "      <td>12:56007301:G:A</td>\n",
       "      <td>56007301</td>\n",
       "      <td>...</td>\n",
       "      <td>12:56007301:G:A;RPS26_RPS29</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>172827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>unique</td>\n",
       "      <td>NK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>12:56042145:C:G;RPS26_RPL41</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>RPL41</td>\n",
       "      <td>RPL41_RPS26</td>\n",
       "      <td>12:56042145:C:G</td>\n",
       "      <td>56042145</td>\n",
       "      <td>...</td>\n",
       "      <td>12:56042145:C:G;RPL41_RPS26</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>172827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>unique</td>\n",
       "      <td>NK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>12:56042145:C:G;RPS26_RPS27</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>RPS27</td>\n",
       "      <td>RPS26_RPS27</td>\n",
       "      <td>12:56042145:C:G</td>\n",
       "      <td>56042145</td>\n",
       "      <td>...</td>\n",
       "      <td>12:56042145:C:G;RPS26_RPS27</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>172827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>0.001380</td>\n",
       "      <td>unique</td>\n",
       "      <td>NK</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23789512</th>\n",
       "      <td>8693916</td>\n",
       "      <td>11:65885771:C:T;FAU_CTSW</td>\n",
       "      <td>CTSW_FAU</td>\n",
       "      <td>11</td>\n",
       "      <td>65120628</td>\n",
       "      <td>65122473</td>\n",
       "      <td>CTSW_FAU</td>\n",
       "      <td>CTSW_FAU</td>\n",
       "      <td>11:65885771:C:T</td>\n",
       "      <td>65885771</td>\n",
       "      <td>...</td>\n",
       "      <td>11:65885771:C:T;CTSW_FAU</td>\n",
       "      <td>1</td>\n",
       "      <td>0.151367</td>\n",
       "      <td>162076</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>duplicate</td>\n",
       "      <td>CD4_T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23789513</th>\n",
       "      <td>8693917</td>\n",
       "      <td>18:24001745:T:C;TTC39C_TTC39C-AS1</td>\n",
       "      <td>TTC39C_TTC39C-AS1</td>\n",
       "      <td>18</td>\n",
       "      <td>23992773</td>\n",
       "      <td>24135610</td>\n",
       "      <td>TTC39C_TTC39C-AS1</td>\n",
       "      <td>TTC39C_TTC39C-AS1</td>\n",
       "      <td>18:24001745:T:C</td>\n",
       "      <td>24001745</td>\n",
       "      <td>...</td>\n",
       "      <td>18:24001745:T:C;TTC39C_TTC39C-AS1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.169862</td>\n",
       "      <td>8053</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>duplicate</td>\n",
       "      <td>CD4_T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23789514</th>\n",
       "      <td>8693918</td>\n",
       "      <td>6:32658495:C:A;HLA-DRB1_HLA-DQB1</td>\n",
       "      <td>HLA-DQB1_HLA-DRB1</td>\n",
       "      <td>6</td>\n",
       "      <td>32578769</td>\n",
       "      <td>32589848</td>\n",
       "      <td>HLA-DQB1_HLA-DRB1</td>\n",
       "      <td>HLA-DQB1_HLA-DRB1</td>\n",
       "      <td>6:32658495:C:A</td>\n",
       "      <td>32658495</td>\n",
       "      <td>...</td>\n",
       "      <td>6:32658495:C:A;HLA-DQB1_HLA-DRB1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.209483</td>\n",
       "      <td>1742214</td>\n",
       "      <td>0.002290</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>duplicate</td>\n",
       "      <td>CD4_T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23789533</th>\n",
       "      <td>8693937</td>\n",
       "      <td>12:8971336:T:C;A2M_KLRB1</td>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>12</td>\n",
       "      <td>9067664</td>\n",
       "      <td>9116229</td>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>12:8971336:T:C</td>\n",
       "      <td>8971336</td>\n",
       "      <td>...</td>\n",
       "      <td>12:8971336:T:C;A2M_KLRB1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.558180</td>\n",
       "      <td>97252</td>\n",
       "      <td>0.000951</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>duplicate</td>\n",
       "      <td>CD4_T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23789546</th>\n",
       "      <td>8693950</td>\n",
       "      <td>12:8992460:T:C;A2M_KLRB1</td>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>12</td>\n",
       "      <td>9067664</td>\n",
       "      <td>9116229</td>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>12:8992460:T:C</td>\n",
       "      <td>8992460</td>\n",
       "      <td>...</td>\n",
       "      <td>12:8992460:T:C;A2M_KLRB1</td>\n",
       "      <td>0</td>\n",
       "      <td>3.490753</td>\n",
       "      <td>97252</td>\n",
       "      <td>0.002130</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>0.002728</td>\n",
       "      <td>duplicate</td>\n",
       "      <td>CD4_T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91628 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Unnamed: 0                              coQTL              eGene  \\\n",
       "0                  1        12:56007301:G:A;RPS26_RPL41              RPS26   \n",
       "1                  2        12:56007301:G:A;RPS26_RPS27              RPS26   \n",
       "2                  3        12:56007301:G:A;RPS26_RPS29              RPS26   \n",
       "3                  4        12:56042145:C:G;RPS26_RPL41              RPS26   \n",
       "4                  5        12:56042145:C:G;RPS26_RPS27              RPS26   \n",
       "...              ...                                ...                ...   \n",
       "23789512     8693916           11:65885771:C:T;FAU_CTSW           CTSW_FAU   \n",
       "23789513     8693917  18:24001745:T:C;TTC39C_TTC39C-AS1  TTC39C_TTC39C-AS1   \n",
       "23789514     8693918   6:32658495:C:A;HLA-DRB1_HLA-DQB1  HLA-DQB1_HLA-DRB1   \n",
       "23789533     8693937           12:8971336:T:C;A2M_KLRB1          A2M_KLRB1   \n",
       "23789546     8693950           12:8992460:T:C;A2M_KLRB1          A2M_KLRB1   \n",
       "\n",
       "          feature_chromosome  feature_start  feature_end            coeGene  \\\n",
       "0                         12       56041351     56044676              RPL41   \n",
       "1                         12       56041351     56044676              RPS27   \n",
       "2                         12       56041351     56044676              RPS29   \n",
       "3                         12       56041351     56044676              RPL41   \n",
       "4                         12       56041351     56044676              RPS27   \n",
       "...                      ...            ...          ...                ...   \n",
       "23789512                  11       65120628     65122473           CTSW_FAU   \n",
       "23789513                  18       23992773     24135610  TTC39C_TTC39C-AS1   \n",
       "23789514                   6       32578769     32589848  HLA-DQB1_HLA-DRB1   \n",
       "23789533                  12        9067664      9116229          A2M_KLRB1   \n",
       "23789546                  12        9067664      9116229          A2M_KLRB1   \n",
       "\n",
       "                 feature_id           snp_id  snp_position  ...  \\\n",
       "0               RPL41_RPS26  12:56007301:G:A      56007301  ...   \n",
       "1               RPS26_RPS27  12:56007301:G:A      56007301  ...   \n",
       "2               RPS26_RPS29  12:56007301:G:A      56007301  ...   \n",
       "3               RPL41_RPS26  12:56042145:C:G      56042145  ...   \n",
       "4               RPS26_RPS27  12:56042145:C:G      56042145  ...   \n",
       "...                     ...              ...           ...  ...   \n",
       "23789512           CTSW_FAU  11:65885771:C:T      65885771  ...   \n",
       "23789513  TTC39C_TTC39C-AS1  18:24001745:T:C      24001745  ...   \n",
       "23789514  HLA-DQB1_HLA-DRB1   6:32658495:C:A      32658495  ...   \n",
       "23789533          A2M_KLRB1   12:8971336:T:C       8971336  ...   \n",
       "23789546          A2M_KLRB1   12:8992460:T:C       8992460  ...   \n",
       "\n",
       "                                    triplet  is_rb_mt  p_val_mt  \\\n",
       "0               12:56007301:G:A;RPL41_RPS26         1  0.000000   \n",
       "1               12:56007301:G:A;RPS26_RPS27         1  0.000000   \n",
       "2               12:56007301:G:A;RPS26_RPS29         1  0.000000   \n",
       "3               12:56042145:C:G;RPL41_RPS26         1  0.000000   \n",
       "4               12:56042145:C:G;RPS26_RPS27         1  0.000000   \n",
       "...                                     ...       ...       ...   \n",
       "23789512           11:65885771:C:T;CTSW_FAU         1  0.151367   \n",
       "23789513  18:24001745:T:C;TTC39C_TTC39C-AS1         0  0.169862   \n",
       "23789514   6:32658495:C:A;HLA-DQB1_HLA-DRB1         0  0.209483   \n",
       "23789533           12:8971336:T:C;A2M_KLRB1         0  1.558180   \n",
       "23789546           12:8992460:T:C;A2M_KLRB1         0  3.490753   \n",
       "\n",
       "          amount_tests_eGene  p_val_mt_eGene  p_bh_cutoff p_q_cutoff  \\\n",
       "0                     172827        0.000000     0.001380   0.001380   \n",
       "1                     172827        0.000000     0.001380   0.001380   \n",
       "2                     172827        0.000000     0.001380   0.001380   \n",
       "3                     172827        0.000000     0.001380   0.001380   \n",
       "4                     172827        0.000000     0.001380   0.001380   \n",
       "...                      ...             ...          ...        ...   \n",
       "23789512              162076        0.000154     0.002728   0.002728   \n",
       "23789513                8053        0.000009     0.002728   0.002728   \n",
       "23789514             1742214        0.002290     0.002728   0.002728   \n",
       "23789533               97252        0.000951     0.002728   0.002728   \n",
       "23789546               97252        0.002130     0.002728   0.002728   \n",
       "\n",
       "               type  cell_type  q_value  \n",
       "0            unique         NK      NaN  \n",
       "1            unique         NK      NaN  \n",
       "2            unique         NK      NaN  \n",
       "3            unique         NK      NaN  \n",
       "4            unique         NK      NaN  \n",
       "...             ...        ...      ...  \n",
       "23789512  duplicate      CD4_T      NaN  \n",
       "23789513  duplicate      CD4_T      NaN  \n",
       "23789514  duplicate      CD4_T      NaN  \n",
       "23789533  duplicate      CD4_T      NaN  \n",
       "23789546  duplicate      CD4_T      NaN  \n",
       "\n",
       "[91628 rows x 42 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "12b70239-7d2f-47b9-b896-6d180d8652d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if mt_correction  != 'eGene_bf':  # Bonferroni threshold\n",
    "    meta_results = meta_results[meta_results['p_val_mt'] < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "59103fad-5ca8-4157-bcee-a9de3f88f5c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define the sign gene-pair and eGenes for which to save all the tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c50bae8-ff25-479a-8257-28a0528e8444",
   "metadata": {},
   "outputs": [],
   "source": [
    "### All significant gene-pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "383c137f-fc5b-467a-b83a-e69f6c957530",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_gene_pairs = pd.unique(pd.concat([coeqtl_results['feature_id'], meta_results['feature_id']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d754301c-e141-4b78-a637-58db0c0ff2f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### All significant eGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "ab73d469-65df-4ff9-a785-53eca202cb4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_egenes_sc_datasets = coeqtl_results.drop_duplicates(['eGene', 'snp_id'])[['eGene', 'snp_id']]\n",
    "sign_egenes_meta = meta_results.drop_duplicates(['eGene', 'snp_id'])[['eGene', 'snp_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "23b01f0f-91cd-40f7-b107-ac4b06a6a364",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_egenes_sc_datasets['egene_snp_id'] = sign_egenes_sc_datasets['eGene'] + ';' +  sign_egenes_sc_datasets['snp_id']\n",
    "sign_egenes_meta['egene_snp_id'] = sign_egenes_meta['eGene'] + ';' +  sign_egenes_meta['snp_id'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "18e83c13-de8c-4773-af04-79adefaafb38",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_egenes = pd.concat([sign_egenes_sc_datasets, sign_egenes_meta], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "46b0c1b7-b809-4c5f-a87c-44d89c00af03",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_egenes = sign_egenes.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "ff7cfe6f-1894-4ff5-bc00-3622432d70d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGene</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>egene_snp_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPS26</td>\n",
       "      <td>12:55991020:G:A</td>\n",
       "      <td>RPS26;12:55991020:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPS26</td>\n",
       "      <td>12:55995509:T:C</td>\n",
       "      <td>RPS26;12:55995509:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RPS26</td>\n",
       "      <td>12:56001170:A:G</td>\n",
       "      <td>RPS26;12:56001170:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPS26</td>\n",
       "      <td>12:56007301:G:A</td>\n",
       "      <td>RPS26;12:56007301:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPS26</td>\n",
       "      <td>12:56042145:C:G</td>\n",
       "      <td>RPS26;12:56042145:C:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2001</th>\n",
       "      <td>GLIPR1_KRR1</td>\n",
       "      <td>12:75493570:A:G</td>\n",
       "      <td>GLIPR1_KRR1;12:75493570:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2005</th>\n",
       "      <td>CTSW_FAU</td>\n",
       "      <td>11:65885771:C:T</td>\n",
       "      <td>CTSW_FAU;11:65885771:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2007</th>\n",
       "      <td>HLA-DQB1_HLA-DRB1</td>\n",
       "      <td>6:32658495:C:A</td>\n",
       "      <td>HLA-DQB1_HLA-DRB1;6:32658495:C:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2008</th>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>12:8971336:T:C</td>\n",
       "      <td>A2M_KLRB1;12:8971336:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2009</th>\n",
       "      <td>A2M_KLRB1</td>\n",
       "      <td>12:8992460:T:C</td>\n",
       "      <td>A2M_KLRB1;12:8992460:T:C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1161 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  eGene           snp_id                      egene_snp_id\n",
       "0                 RPS26  12:55991020:G:A             RPS26;12:55991020:G:A\n",
       "1                 RPS26  12:55995509:T:C             RPS26;12:55995509:T:C\n",
       "2                 RPS26  12:56001170:A:G             RPS26;12:56001170:A:G\n",
       "3                 RPS26  12:56007301:G:A             RPS26;12:56007301:G:A\n",
       "4                 RPS26  12:56042145:C:G             RPS26;12:56042145:C:G\n",
       "...                 ...              ...                               ...\n",
       "2001        GLIPR1_KRR1  12:75493570:A:G       GLIPR1_KRR1;12:75493570:A:G\n",
       "2005           CTSW_FAU  11:65885771:C:T          CTSW_FAU;11:65885771:C:T\n",
       "2007  HLA-DQB1_HLA-DRB1   6:32658495:C:A  HLA-DQB1_HLA-DRB1;6:32658495:C:A\n",
       "2008          A2M_KLRB1   12:8971336:T:C          A2M_KLRB1;12:8971336:T:C\n",
       "2009          A2M_KLRB1   12:8992460:T:C          A2M_KLRB1;12:8992460:T:C\n",
       "\n",
       "[1161 rows x 3 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sign_egenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "e809e271-e25e-42d6-afa3-8d7b36b0433e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "362"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pd.unique(sign_egenes['eGene']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "04b5f3b0-90b6-4850-950c-620197c457ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1161"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(pd.unique(sign_egenes['egene_snp_id']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c674ae12-c87e-40fd-971f-9467e2cb3da5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get the tests for each sign gene-pair (in the meta-analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "17340d15-7632-480b-b67a-d6348fcf7b63",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = ['meta_analysis']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "23332858-8c80-43ca-a882-fd0b6b662b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = ['CD4_T', 'Mono', 'NK', 'B', 'CD8_T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b7ded4-2f4c-4c8e-9cf1-c44ebb6896b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Initialize an empty DataFrame to store the results\n",
    "executed_tests = pd.DataFrame()\n",
    "\n",
    "\n",
    "# Loop through each dataset and cell type\n",
    "for i in datasets_to_load:\n",
    "    for j in cell_types:\n",
    "        # Construct the file path\n",
    "        file_path = data_path + '/' +coeqtl_results_path  + '/'+ i + '/'  + 'meta_5ds/' + j + '/meta_qtl_results_all.txt.gz'\n",
    "\n",
    "        # Check if the file exists\n",
    "        if os.path.exists(file_path):\n",
    "            # Read the CSV file into a DataFrame\n",
    "            result_file = pd.read_csv(file_path, sep = '\\t')\n",
    "\n",
    "            result_file = result_file[result_file['feature_id'].isin(sign_gene_pairs)]\n",
    "\n",
    "            # Add the dataset and cell_type columns\n",
    "            result_file['dataset'] = i\n",
    "            result_file['cell_type'] = j\n",
    "\n",
    "            # Append the result to the coeqtl_results DataFrame\n",
    "            executed_tests = pd.concat([executed_tests, result_file], ignore_index=True)\n",
    "            del result_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f40b65-edd6-43dd-9c40-180b50f8e9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.unique(executed_tests['cell_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd4f97f-81ee-499c-a6cf-264904a465b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.unique(executed_tests['dataset'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00494dc4-0c2f-4073-97a6-82b724e5e8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1918a4d0-ea2f-4763-89ee-5fc2aedae53f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the file for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02faa68-2bbc-4437-be50-008ef8d76f8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests.to_csv(data_path + coeqtl_results_path + '/tests_sign_pairs/Tests_executed_on_sign_gene_pairs_meta' +  str(n_sample_filter)  + '.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fad56f2-29b4-4d30-8ff8-1af0b74af142",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get the tests for each sign gene-pair (in the single-datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "1973ed5c-c236-4c30-83bf-959f5198e1f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = ['oneK1K', 'Franke_v2'  , 'Franke_v3', 'wijst', 'multiome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "abeb5355-f64e-4b0a-9162-5e1629e238dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = ['CD4_T', 'Mono', 'NK', 'B', 'CD8_T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "8081680c-3f41-4cc7-a334-b2ceb0766446",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Initialize an empty DataFrame to store the results\n",
    "executed_tests = pd.DataFrame()\n",
    "\n",
    "# Loop through each dataset and cell type\n",
    "for i in datasets_to_load:\n",
    "    for j in cell_types:\n",
    "        # Construct the file path\n",
    "        #file_path = data_path + '/' +coeqtl_results_path  + '/'+ i + '/' + j + '/F3_decision_tree_input.csv.gz'\n",
    "        file_path = data_path + '/' +coeqtl_results_path  + '/'+ i + '/' + j + '/qtl_results_all_unqid.csv.gz'  # changed to ensure to have all tests not only those filtered based on n_sample\n",
    "\n",
    "        # Check if the file exists\n",
    "        if os.path.exists(file_path):\n",
    "            # Read the CSV file into a DataFrame\n",
    "            result_file = pd.read_csv(file_path)\n",
    "\n",
    "            result_file = result_file[result_file['feature_id'].isin(sign_gene_pairs)]\n",
    "\n",
    "            # Add the dataset and cell_type columns\n",
    "            result_file['dataset'] = i\n",
    "            result_file['cell_type'] = j\n",
    "\n",
    "            # Append the result to the coeqtl_results DataFrame\n",
    "            executed_tests = pd.concat([executed_tests, result_file], ignore_index=True)\n",
    "            del result_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "3c0fe3ac-2150-4c85-ac69-53ce7ca33a9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CD4_T', 'Mono', 'NK', 'B', 'CD8_T'], dtype=object)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.unique(executed_tests['cell_type'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "76513af5-8ea3-4916-a224-0a7fb6e2ecbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['oneK1K', 'Franke_v2', 'Franke_v3', 'wijst', 'multiome'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.unique(executed_tests['dataset'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "9e7d3eb2-0eed-492e-ab80-879f1d946f38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>p_value</th>\n",
       "      <th>beta</th>\n",
       "      <th>beta_se</th>\n",
       "      <th>empirical_feature_p_value</th>\n",
       "      <th>feature_chromosome</th>\n",
       "      <th>feature_start</th>\n",
       "      <th>feature_end</th>\n",
       "      <th>...</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>closeGenes</th>\n",
       "      <th>z_score</th>\n",
       "      <th>triplet</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>n_e_samples</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>dataset</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42372</td>\n",
       "      <td>A1BG_RPS26</td>\n",
       "      <td>12:55975722:A:G</td>\n",
       "      <td>0.397208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>...</td>\n",
       "      <td>55975722.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-8.469810e-01</td>\n",
       "      <td>A1BG_RPS26;12:55975722:A:G</td>\n",
       "      <td>992</td>\n",
       "      <td>992</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>oneK1K</td>\n",
       "      <td>CD4_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42373</td>\n",
       "      <td>A1BG_RPS26</td>\n",
       "      <td>12:55991020:G:A</td>\n",
       "      <td>0.506004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>...</td>\n",
       "      <td>55991020.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-6.653110e-01</td>\n",
       "      <td>A1BG_RPS26;12:55991020:G:A</td>\n",
       "      <td>992</td>\n",
       "      <td>992</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>oneK1K</td>\n",
       "      <td>CD4_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42374</td>\n",
       "      <td>A1BG_RPS26</td>\n",
       "      <td>12:55991795:C:T</td>\n",
       "      <td>0.850009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>...</td>\n",
       "      <td>55991795.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.891558e-01</td>\n",
       "      <td>A1BG_RPS26;12:55991795:C:T</td>\n",
       "      <td>992</td>\n",
       "      <td>992</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>oneK1K</td>\n",
       "      <td>CD4_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>42375</td>\n",
       "      <td>A1BG_RPS26</td>\n",
       "      <td>12:55995509:T:C</td>\n",
       "      <td>0.512736</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>...</td>\n",
       "      <td>55995509.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-6.548161e-01</td>\n",
       "      <td>A1BG_RPS26;12:55995509:T:C</td>\n",
       "      <td>992</td>\n",
       "      <td>992</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>oneK1K</td>\n",
       "      <td>CD4_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>42376</td>\n",
       "      <td>A1BG_RPS26</td>\n",
       "      <td>12:56001170:A:G</td>\n",
       "      <td>0.458205</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>56041351</td>\n",
       "      <td>56044676</td>\n",
       "      <td>...</td>\n",
       "      <td>56001170.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-7.420924e-01</td>\n",
       "      <td>A1BG_RPS26;12:56001170:A:G</td>\n",
       "      <td>992</td>\n",
       "      <td>992</td>\n",
       "      <td>RPS26</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>oneK1K</td>\n",
       "      <td>CD4_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3634060</th>\n",
       "      <td>86741372</td>\n",
       "      <td>HLA-DRB1_RPS28</td>\n",
       "      <td>6:32457527:A:G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>32578769</td>\n",
       "      <td>32589848</td>\n",
       "      <td>...</td>\n",
       "      <td>32457527.0</td>\n",
       "      <td>False</td>\n",
       "      <td>9.521116e-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>RPS28</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>multiome</td>\n",
       "      <td>CD8_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3634061</th>\n",
       "      <td>86741373</td>\n",
       "      <td>HLA-DRB1_RPS28</td>\n",
       "      <td>6:32459604:G:T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>32578769</td>\n",
       "      <td>32589848</td>\n",
       "      <td>...</td>\n",
       "      <td>32459604.0</td>\n",
       "      <td>False</td>\n",
       "      <td>9.521116e-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>RPS28</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>multiome</td>\n",
       "      <td>CD8_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3634062</th>\n",
       "      <td>86741421</td>\n",
       "      <td>HLA-DRB1_UBA52</td>\n",
       "      <td>6:33072321:C:T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>32578769</td>\n",
       "      <td>32589848</td>\n",
       "      <td>...</td>\n",
       "      <td>33072321.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.760558e-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>UBA52</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>multiome</td>\n",
       "      <td>CD8_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3634063</th>\n",
       "      <td>86741422</td>\n",
       "      <td>HLA-DRB1_UBA52</td>\n",
       "      <td>6:33092341:G:A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>32578769</td>\n",
       "      <td>32589848</td>\n",
       "      <td>...</td>\n",
       "      <td>33092341.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.760558e-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>UBA52</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>multiome</td>\n",
       "      <td>CD8_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3634064</th>\n",
       "      <td>86741423</td>\n",
       "      <td>HLA-DRB1_UBA52</td>\n",
       "      <td>6:33104395:T:G</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>32578769</td>\n",
       "      <td>32589848</td>\n",
       "      <td>...</td>\n",
       "      <td>33104395.0</td>\n",
       "      <td>False</td>\n",
       "      <td>-4.760558e-17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>UBA52</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>multiome</td>\n",
       "      <td>CD8_T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3634065 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Unnamed: 0      feature_id           snp_id   p_value  beta  beta_se  \\\n",
       "0             42372      A1BG_RPS26  12:55975722:A:G  0.397208   NaN      NaN   \n",
       "1             42373      A1BG_RPS26  12:55991020:G:A  0.506004   NaN      NaN   \n",
       "2             42374      A1BG_RPS26  12:55991795:C:T  0.850009   NaN      NaN   \n",
       "3             42375      A1BG_RPS26  12:55995509:T:C  0.512736   NaN      NaN   \n",
       "4             42376      A1BG_RPS26  12:56001170:A:G  0.458205   NaN      NaN   \n",
       "...             ...             ...              ...       ...   ...      ...   \n",
       "3634060    86741372  HLA-DRB1_RPS28   6:32457527:A:G  1.000000   NaN      NaN   \n",
       "3634061    86741373  HLA-DRB1_RPS28   6:32459604:G:T  1.000000   NaN      NaN   \n",
       "3634062    86741421  HLA-DRB1_UBA52   6:33072321:C:T  1.000000   NaN      NaN   \n",
       "3634063    86741422  HLA-DRB1_UBA52   6:33092341:G:A  1.000000   NaN      NaN   \n",
       "3634064    86741423  HLA-DRB1_UBA52   6:33104395:T:G  1.000000   NaN      NaN   \n",
       "\n",
       "         empirical_feature_p_value  feature_chromosome  feature_start  \\\n",
       "0                              NaN                  12       56041351   \n",
       "1                              NaN                  12       56041351   \n",
       "2                              NaN                  12       56041351   \n",
       "3                              NaN                  12       56041351   \n",
       "4                              NaN                  12       56041351   \n",
       "...                            ...                 ...            ...   \n",
       "3634060                        NaN                   6       32578769   \n",
       "3634061                        NaN                   6       32578769   \n",
       "3634062                        NaN                   6       32578769   \n",
       "3634063                        NaN                   6       32578769   \n",
       "3634064                        NaN                   6       32578769   \n",
       "\n",
       "         feature_end  ... snp_position  closeGenes       z_score  \\\n",
       "0           56044676  ...   55975722.0       False -8.469810e-01   \n",
       "1           56044676  ...   55991020.0       False -6.653110e-01   \n",
       "2           56044676  ...   55991795.0       False  1.891558e-01   \n",
       "3           56044676  ...   55995509.0       False -6.548161e-01   \n",
       "4           56044676  ...   56001170.0       False -7.420924e-01   \n",
       "...              ...  ...          ...         ...           ...   \n",
       "3634060     32589848  ...   32457527.0       False  9.521116e-17   \n",
       "3634061     32589848  ...   32459604.0       False  9.521116e-17   \n",
       "3634062     32589848  ...   33072321.0       False -4.760558e-17   \n",
       "3634063     32589848  ...   33092341.0       False -4.760558e-17   \n",
       "3634064     32589848  ...   33104395.0       False -4.760558e-17   \n",
       "\n",
       "                            triplet n_samples n_e_samples  gene1     gene2  \\\n",
       "0        A1BG_RPS26;12:55975722:A:G       992         992  RPS26      A1BG   \n",
       "1        A1BG_RPS26;12:55991020:G:A       992         992  RPS26      A1BG   \n",
       "2        A1BG_RPS26;12:55991795:C:T       992         992  RPS26      A1BG   \n",
       "3        A1BG_RPS26;12:55995509:T:C       992         992  RPS26      A1BG   \n",
       "4        A1BG_RPS26;12:56001170:A:G       992         992  RPS26      A1BG   \n",
       "...                             ...       ...         ...    ...       ...   \n",
       "3634060                         NaN        24          24  RPS28  HLA-DRB1   \n",
       "3634061                         NaN        24          24  RPS28  HLA-DRB1   \n",
       "3634062                         NaN        24          24  UBA52  HLA-DRB1   \n",
       "3634063                         NaN        24          24  UBA52  HLA-DRB1   \n",
       "3634064                         NaN        24          24  UBA52  HLA-DRB1   \n",
       "\n",
       "          dataset cell_type  \n",
       "0          oneK1K     CD4_T  \n",
       "1          oneK1K     CD4_T  \n",
       "2          oneK1K     CD4_T  \n",
       "3          oneK1K     CD4_T  \n",
       "4          oneK1K     CD4_T  \n",
       "...           ...       ...  \n",
       "3634060  multiome     CD8_T  \n",
       "3634061  multiome     CD8_T  \n",
       "3634062  multiome     CD8_T  \n",
       "3634063  multiome     CD8_T  \n",
       "3634064  multiome     CD8_T  \n",
       "\n",
       "[3634065 rows x 26 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "executed_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "0099872c-67f9-44a9-a8b6-bc141c2ec7cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the file for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "a0057230-f9da-4fba-a755-f69adb20b348",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests.to_csv(data_path + coeqtl_results_path + '/tests_sign_pairs/Tests_executed_on_sign_gene_pairs' +   str(n_sample_filter)  +  mt_correction + '_' + snp_filter + '.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b1f6b6-aaf3-455c-b8cc-541b50a3b841",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get tests for each sign egene (in the single datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "dfbb74aa-f877-4900-abd5-28f8d3d31dce",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = ['oneK1K']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "6caa4c33-b33c-4338-8dec-ec08f491051d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = ['CD4_T', 'Mono', 'NK', 'B', 'CD8_T']\n",
    "# ['CD4_T', 'Mono', 'NK', 'B', 'CD8_T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "92adb7da-66ee-490a-828f-6ed65515ef11",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3441, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_1734574/1546546559.py\", line 13, in <module>\n",
      "    result_file = pd.read_csv(file_path)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 1026, in read_csv\n",
      "    return _read(filepath_or_buffer, kwds)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 626, in _read\n",
      "    return parser.read(nrows)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/readers.py\", line 1923, in read\n",
      "    ) = self._engine.read(  # type: ignore[attr-defined]\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py\", line 234, in read\n",
      "    chunks = self._reader.read_low_memory(nrows)\n",
      "  File \"parsers.pyx\", line 838, in pandas._libs.parsers.TextReader.read_low_memory\n",
      "  File \"parsers.pyx\", line 921, in pandas._libs.parsers.TextReader._read_rows\n",
      "  File \"parsers.pyx\", line 1083, in pandas._libs.parsers.TextReader._convert_column_data\n",
      "  File \"parsers.pyx\", line 1456, in pandas._libs.parsers._maybe_upcast\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/numpy/core/multiarray.py\", line 1131, in putmask\n",
      "    @array_function_from_c_func_and_dispatcher(_multiarray_umath.putmask)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'KeyboardInterrupt' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/inspect.py\", line 1541, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/inspect.py\", line 1499, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/inspect.py\", line 709, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/inspect.py\", line 752, in getmodule\n",
      "    f = getabsfile(module)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/inspect.py\", line 721, in getabsfile\n",
      "    _filename = getsourcefile(object) or getfile(object)\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/inspect.py\", line 706, in getsourcefile\n",
      "    if os.path.exists(filename):\n",
      "  File \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1/lib/python3.9/genericpath.py\", line 19, in exists\n",
      "    os.stat(path)\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "object of type 'NoneType' has no len()",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_1734574/1546546559.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m             \u001b[0;31m# Read the CSV file into a DataFrame\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m             \u001b[0mresult_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1025\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1026\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1027\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    625\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 626\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    627\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1922\u001b[0m                     \u001b[0mcol_dict\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1923\u001b[0;31m                 \u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m  \u001b[0;31m# type: ignore[attr-defined]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1924\u001b[0m                     \u001b[0mnrows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m    233\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlow_memory\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 234\u001b[0;31m                 \u001b[0mchunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_low_memory\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    235\u001b[0m                 \u001b[0;31m# destructive to chunks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mparsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.read_low_memory\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mparsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mparsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._convert_column_data\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mparsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers._maybe_upcast\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/numpy/core/multiarray.py\u001b[0m in \u001b[0;36mputmask\u001b[0;34m(a, mask, values)\u001b[0m\n\u001b[1;32m   1130\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1131\u001b[0;31m \u001b[0;34m@\u001b[0m\u001b[0marray_function_from_c_func_and_dispatcher\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_multiarray_umath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mputmask\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1132\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mputmask\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m/\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mshowtraceback\u001b[0;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001b[0m\n\u001b[1;32m   2060\u001b[0m                         \u001b[0;31m# in the engines. This should return a list of strings.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2061\u001b[0;31m                         \u001b[0mstb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_render_traceback_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2062\u001b[0m                     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'KeyboardInterrupt' object has no attribute '_render_traceback_'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mshowtraceback\u001b[0;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001b[0m\n\u001b[1;32m   2061\u001b[0m                         \u001b[0mstb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_render_traceback_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2062\u001b[0m                     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2063\u001b[0;31m                         stb = self.InteractiveTB.structured_traceback(etype,\n\u001b[0m\u001b[1;32m   2064\u001b[0m                                             value, tb, tb_offset=tb_offset)\n\u001b[1;32m   2065\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1365\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1366\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1367\u001b[0;31m         return FormattedTB.structured_traceback(\n\u001b[0m\u001b[1;32m   1368\u001b[0m             self, etype, value, tb, tb_offset, number_of_lines_of_context)\n\u001b[1;32m   1369\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1265\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mverbose_modes\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1266\u001b[0m             \u001b[0;31m# Verbose modes need a full traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1267\u001b[0;31m             return VerboseTB.structured_traceback(\n\u001b[0m\u001b[1;32m   1268\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb_offset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumber_of_lines_of_context\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1269\u001b[0m             )\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1122\u001b[0m         \u001b[0;34m\"\"\"Return a nice text document describing the traceback.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1124\u001b[0;31m         formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n\u001b[0m\u001b[1;32m   1125\u001b[0m                                                                tb_offset)\n\u001b[1;32m   1126\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mformat_exception_as_a_whole\u001b[0;34m(self, etype, evalue, etb, number_of_lines_of_context, tb_offset)\u001b[0m\n\u001b[1;32m   1080\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1081\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1082\u001b[0;31m         \u001b[0mlast_unique\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecursion_repeat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfind_recursion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0morig_etype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mevalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecords\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1083\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1084\u001b[0m         \u001b[0mframes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat_records\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecords\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlast_unique\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecursion_repeat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/scgrn_R_4_1/lib/python3.9/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mfind_recursion\u001b[0;34m(etype, value, records)\u001b[0m\n\u001b[1;32m    380\u001b[0m     \u001b[0;31m# first frame (from in to out) that looks different.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    381\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_recursion_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrecords\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecords\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m     \u001b[0;31m# Select filename, lineno, func_name to track frames with\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: object of type 'NoneType' has no len()"
     ]
    }
   ],
   "source": [
    "# Initialize an empty DataFrame to store the results\n",
    "executed_tests = pd.DataFrame()\n",
    "\n",
    "# Loop through each dataset and cell type\n",
    "for i in datasets_to_load:\n",
    "    for j in cell_types:\n",
    "        # Construct the file path\n",
    "        file_path = data_path + '/' +coeqtl_results_path  + '/'+ i + '/' + j + '/F3_decision_tree_input.csv.gz'\n",
    "\n",
    "        # Check if the file exists\n",
    "        if os.path.exists(file_path):\n",
    "            # Read the CSV file into a DataFrame\n",
    "            result_file = pd.read_csv(file_path)\n",
    "            \n",
    "            \n",
    "            ### Remove unnnecessary columns\n",
    "            del result_file['empirical_feature_p_value']\n",
    "            del result_file ['feature_chromosome']\n",
    "            #del result_file['test1']\n",
    "            #del result_file['test2']\n",
    "            #del result_file['test3']\n",
    "            del result_file['feature_start']\n",
    "            del result_file['feature_end']\n",
    "            #del result_file['biotype']\n",
    "            #del result_file['index']\n",
    "            del result_file[ 'alpha_param']\n",
    "            del result_file['beta_param']\n",
    "\n",
    "           ### Add columns for filtering\n",
    "            result_file['gene1_snp_id'] = result_file['gene1'] + ';' + result_file['snp_id']\n",
    "            result_file['gene2_snp_id'] = result_file['gene2'] + ';' + result_file['snp_id']\n",
    "\n",
    "            ### Filter result file on results for significant egenes\n",
    "            result_file =  result_file[(result_file['gene1_snp_id'].isin(pd.unique(sign_egenes['egene_snp_id']))) | (result_file['gene2_snp_id'].isin(pd.unique(sign_egenes['egene_snp_id'])))  ]\n",
    "\n",
    "            # Add the dataset and cell_type columns\n",
    "            result_file['dataset'] = i\n",
    "            result_file['cell_type'] = j\n",
    "\n",
    "            # Append the result to the coeqtl_results DataFrame\n",
    "            executed_tests = pd.concat([executed_tests, result_file], ignore_index=True)\n",
    "            del result_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f248a4b-63f2-49b2-a772-a8fa0635ef8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the file for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1ce3ab0-88fd-489d-95d2-c530d4a55548",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests.to_csv(data_path + coeqtl_results_path + '/tests_sign_pairs/Tests_executed_on_sign_egenes' +   str(n_sample_filter)  +  mt_correction + '.csv.gz', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3664976b-04f9-4dd0-9083-7fecf683a5c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#executed_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83ab1610-4d73-4c2c-ab8e-4ad1ae403287",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e8169e1f-bbff-46ff-bd7b-548bcbbf36d8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get tests for each sign egene (in the meta-analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "7f7c3f04-d882-48ee-bf12-dfd6a7369afc",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = ['meta_analysis']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "f7a76ef8-1314-4ff3-a4d1-d6510457510f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = [ 'Mono', 'NK', 'B', 'CD8_T', 'CD4_T']\n",
    "# ['CD4_T', 'Mono', 'NK', 'B', 'CD8_T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "ee285624-4732-4500-961d-a6ba580b4881",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CD8_T'"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60070232-b905-4b50-a869-85599ab04905",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Initialize an empty DataFrame to store the results\n",
    "executed_tests = pd.DataFrame()\n",
    "\n",
    "# Loop through each dataset and cell type\n",
    "for i in datasets_to_load:\n",
    "    for j in cell_types:\n",
    "        # Construct the file path\n",
    "        file_path = data_path + '/' +coeqtl_results_path  + '/'+ i + '/'  + 'meta_5ds/' + j + '/meta_qtl_results_all.txt.gz'\n",
    "\n",
    "        # Check if the file exists\n",
    "        if os.path.exists(file_path):\n",
    "            # Read the CSV file into a DataFrame\n",
    "            result_file = pd.read_csv(file_path, sep = '\\t')\n",
    "\n",
    "           ### Add columns for filtering\n",
    "            result_file['eGene_snp_id'] = result_file['eGene'] + ';' + result_file['snp_id']\n",
    "            #result_file['gene2_snp_id'] = result_file['gene2'] + ';' + result_file['snp_id']\n",
    "\n",
    "            ### Filter result file on results for significant egenes\n",
    "            result_file =  result_file[(result_file['eGene_snp_id'].isin(pd.unique(sign_egenes['egene_snp_id'])))  ]\n",
    "            result_file = result_file[['coQTL', 'eGene',  'coeGene', 'feature_id', 'snp_id', 'snp_position', 'closeGenes', 'p_value', 'n_samples']]\n",
    "\n",
    "            # Add the dataset and cell_type columns\n",
    "            result_file['dataset'] = i\n",
    "            result_file['cell_type'] = j\n",
    "\n",
    "            # Append the result to the coeqtl_results DataFrame\n",
    "            executed_tests = pd.concat([executed_tests, result_file], ignore_index=True)\n",
    "            del result_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d4688e-ae01-4672-8569-6c7ff907747f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the file for further analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e27a3a96-fbe2-4e83-9dfe-fc99f18430a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests.to_csv(data_path + coeqtl_results_path + '/tests_sign_pairs/Tests_executed_on_sign_egenes_meta_5ds' +   str(n_sample_filter)  +  mt_correction + '.csv.gz', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94d90774-93e5-4d62-a1e9-82a4c76d216b",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests['ident'] = executed_tests['eGene'] + ';' +  executed_tests['snp_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df8a023-8c25-410b-be18-837ae2ce33d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add to the tests whether significant co-EQTL in meta-analysis or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b6f36f9-64b0-480f-9573-1a43009260a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_results['coeQTL'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c8063be-d1cd-4506-b5f0-7f00bb291396",
   "metadata": {},
   "outputs": [],
   "source": [
    "executed_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f391327-7bf2-454e-aec6-4016d29e1711",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = pd.merge(executed_tests, meta_results[['eGene', 'coeGene', 'snp_id', 'cell_type', 'coeQTL']], on=['eGene', 'coeGene', 'snp_id', 'cell_type'], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22545408-b9f1-4ad6-a12e-348c0e094a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(meta_results['coeQTL'][meta_results['cell_type'] == 'Mono'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f558c820-2220-470e-a02b-2865049f6807",
   "metadata": {},
   "outputs": [],
   "source": [
    "result['coeQTL'] = result['coeQTL'].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0af3765-62f9-4773-b3cd-b86444d7d738",
   "metadata": {},
   "outputs": [],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d1bfd3a-0567-4201-bd83-33639517b97a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(result['coeQTL'][result['cell_type'] == 'B'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c6d1458-2394-4d9b-8b6b-dffce7b80474",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(result['coeQTL'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9cf0669-e13c-4e91-9b05-b27888ad0d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv(data_path + coeqtl_results_path + '/tests_sign_pairs/Tests_executed_on_sign_egenes_meta_5ds_incl_annot' +   str(n_sample_filter)  +  mt_correction + '.csv.gz', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b727214a-f4f4-4730-81bb-a2d91d56e5c8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6522f32-4ec6-47ff-abb5-d3bb187da1d7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09b76a5f-a1b0-41e2-b4a5-dbceeba60b65",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
