{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a1444178-491b-4528-9c7d-b171aa9267c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Script to calculate summary statistics on all single-cell datasets and save them for usage in other scripts\n",
    "# e.g. mean expression, percentage of zero of genes etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dca7889-aa95-41fc-84f2-fc38b8392196",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "49dfe76f-f0c5-476d-8c73-e721c2a6e069",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "This is mgcv 1.9-1. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "corrplot 0.95 loaded\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:genefilter’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "\n",
      "************\n",
      "Welcome to faux. For support and examples visit:\n",
      "https://debruine.github.io/faux/\n",
      "- Get and set global package options with: faux_options()\n",
      "************\n",
      "\n",
      "wCorr v1.9.8\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘sjmisc’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    is_empty\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘DT’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "This is coloc version 5.2.3\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: gsubfn\n",
      "\n",
      "Loading required package: proto\n",
      "\n",
      "Warning message in fun(libname, pkgname):\n",
      "“couldn't connect to display \"localhost:13.0\"”\n",
      "Loading required package: RSQLite\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    replace_na\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dbscan’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    as.dendrogram\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:rtracklayer’:\n",
      "\n",
      "    blocks, path\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    components\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: dynamicTreeCut\n",
      "\n",
      "Loading required package: fastcluster\n",
      "\n",
      "\n",
      "Attaching package: ‘fastcluster’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    hclust\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘WGCNA’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:qvalue’:\n",
      "\n",
      "    qvalue\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    geometry\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: MotifDb\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'XVector'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grid':\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "See system.file(\"LICENSE\", package=\"MotifDb\") for use restrictions.\n",
      "\n",
      "Warning message:\n",
      "\"replacing previous import 'S4Vectors::as.data.frame' by 'motifStack::as.data.frame' when loading 'motifbreakR'\"\n",
      "\n",
      "Attaching package: 'Gviz'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    feature\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f55b4c93-827f-4edf-b237-d399f128ac28",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e8e76471-b038-407f-999a-1735bbb7a186",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to single-cell datasets (oneK1K, Wijst, UMCG2, UMCG3, multiome; stored as .Rds data files)\n",
    "\n",
    "data_path_sc = \"../data/current/wg3/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e814f197-db05-450a-86e2-c24ff80b842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to save calculated summary statistics to\n",
    "\n",
    "output_path_sc = \"../data/current/summary_stats/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64663cd5-8d6a-454a-8d0c-ff44924cee9d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "494ae5c4-500f-4367-9f43-ceb7c2c5667c",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "18b7c1f7-dd8f-44ef-a6aa-de647f7c15d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dd6604-4d42-4bb0-81ed-9555d63ddcb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "45cfaa89-153c-4abe-92f5-868c50e8090b",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6123ec7d-d87b-44df-b708-45126bb587db",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load different sc dataset and get some meta-information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "585d8d96-ba0c-425e-866a-6cb708dc6e37",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Select datasets for which summary statistics should be calculated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60112fac-cddd-4a5d-8a22-2201fc7f7c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "folders = c(#'wg3_wijst2018'#, \n",
    "            #'wg3_Trynka', \n",
    "            'wg3_oneK1k'#,\n",
    "           # 'wg3_NC2020',\n",
    "            #'wg3_Nawijn',\n",
    "            #'wg3_Li',\n",
    "           # 'wg3_Ye',\n",
    "           #  'wg3_okada',\n",
    "           #   'wg3_sawcer',\n",
    "          #  'wg3_Franke_split_v3'#,\n",
    "          # 'wg3_Franke_split_v2'#,\n",
    "          #  'wg3_blokland2020'\n",
    "           )  ## specify folders that contain datasets that should be included"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7de77141-b57b-4e05-a928-e0220deb3b2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d3da6875-238f-44e4-86bc-14af085abdc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "416ec510-f876-405d-af0a-bde54e2237b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get all available datasets overview\n",
    "for(i in folders){\n",
    "\n",
    "    dataset_overview = data.frame( datasets = list.files(paste0(data_path_sc, i, '/input/L1')))\n",
    "    if(nrow(dataset_overview) >= 1){\n",
    "        dataset_overview$name = i\n",
    "        dataset_overview = dataset_overview[!is.na(str_extract(dataset_overview$datasets, '.Rds')),]\n",
    "        overview_data = rbind(overview_data, dataset_overview)\n",
    "      }\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "74b84d59-4da8-40fb-afdc-330ff1964563",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Mono.Qced.Normalized.SCs.Rds </td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_oneK1k</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & datasets & name\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_oneK1k\\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k\\\\\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k\\\\\n",
       "\t4 & Mono.Qced.Normalized.SCs.Rds  & wg3\\_oneK1k\\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_oneK1k\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 2\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_oneK1k |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_oneK1k |\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_oneK1k |\n",
       "| 4 | Mono.Qced.Normalized.SCs.Rds  | wg3_oneK1k |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_oneK1k |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name      \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_oneK1k\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "4 Mono.Qced.Normalized.SCs.Rds  wg3_oneK1k\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_oneK1k"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataset_overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7b52b80c-df6e-4a4f-b974-de354ea40e16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Mono.Qced.Normalized.SCs.Rds </td><td>wg3_oneK1k</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_oneK1k</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & datasets & name\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_oneK1k\\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k\\\\\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k\\\\\n",
       "\t4 & Mono.Qced.Normalized.SCs.Rds  & wg3\\_oneK1k\\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_oneK1k\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 2\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_oneK1k |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_oneK1k |\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_oneK1k |\n",
       "| 4 | Mono.Qced.Normalized.SCs.Rds  | wg3_oneK1k |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_oneK1k |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name      \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_oneK1k\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "4 Mono.Qced.Normalized.SCs.Rds  wg3_oneK1k\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_oneK1k"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(overview_data,20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "57df4271-bfc2-434d-b2c2-a54ed61dc704",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data$datasets_match = paste0(overview_data$name, overview_data$datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d701f5ce-ebb0-4bf0-b56c-5c827e6bc2fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td><td>wg3_oneK1kB.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds & wg3\\_oneK1k & wg3\\_oneK1kB.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds | wg3_oneK1k | wg3_oneK1kB.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                  name       datasets_match                     \n",
       "1 B.Qced.Normalized.SCs.Rds wg3_oneK1k wg3_oneK1kB.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data[1,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7bd88595-715c-4336-ae8c-dc7b799fd8b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1"
      ],
      "text/latex": [
       "1"
      ],
      "text/markdown": [
       "1"
      ],
      "text/plain": [
       "[1] 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(overview_data$name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "09f7a0d3-0f3e-44b8-9d8a-88bf25c72fdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5"
      ],
      "text/latex": [
       "5"
      ],
      "text/markdown": [
       "5"
      ],
      "text/plain": [
       "[1] 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(overview_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "73457506-c230-4e36-8fed-01e99efd6fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load dataset\n",
    "get_dataset = function(path){\n",
    "    data = readRDS(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "37e82f77-d695-47d8-b562-ca2755a554b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get already generated summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7713d4f5-0a3c-4fb8-a29d-4ae4e7c58e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "available_summaries = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2ce6f87a-9ddd-40bd-a4ec-8a16c74bb0aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "available_summaries= data.frame( datasets = list.files(paste0(output_path_sc)))\n",
    "available_summaries = data.frame(datasets = available_summaries[!is.na(str_extract(available_summaries$datasets, 'D1_Sample_cell_statistic')),])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5e54f066-d50e-44d9-8ce2-882e491baead",
   "metadata": {},
   "outputs": [],
   "source": [
    "available_summaries$datasets = str_replace(available_summaries$datasets, 'D1_Sample_cell_statistic', '')\n",
    "available_summaries$datasets = str_replace(available_summaries$datasets, '.csv', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "71bd721b-7727-46df-abea-65cdb37915d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "174"
      ],
      "text/latex": [
       "174"
      ],
      "text/markdown": [
       "174"
      ],
      "text/plain": [
       "[1] 174"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(available_summaries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b25b82c7-0092-434b-a7f2-2b976432b813",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>wg3_blokland2020B.Qced.Normalized.SCs.Rds    </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>wg3_blokland2020CD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 1\n",
       "\\begin{tabular}{r|l}\n",
       "  & datasets\\\\\n",
       "  & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & wg3\\_blokland2020B.Qced.Normalized.SCs.Rds    \\\\\n",
       "\t2 & wg3\\_blokland2020CD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 1\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; |\n",
       "|---|---|\n",
       "| 1 | wg3_blokland2020B.Qced.Normalized.SCs.Rds     |\n",
       "| 2 | wg3_blokland2020CD4_T.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                                     \n",
       "1 wg3_blokland2020B.Qced.Normalized.SCs.Rds    \n",
       "2 wg3_blokland2020CD4_T.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(available_summaries,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "87a91bea-b453-48d9-9163-59fd24a103b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4"
      ],
      "text/latex": [
       "4"
      ],
      "text/markdown": [
       "4"
      ],
      "text/plain": [
       "[1] 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(overview_data[overview_data$datasets_match %in% available_summaries$datasets,]) # already existing summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea54eae-464f-49dd-846a-a29109fbca76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "cfd28d72-fa0e-442c-bd55-b7f382662b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove already existing summaries to not recalculate them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1ff907ff-c789-412c-8889-5e0af088b4ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "calculate_new = TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "3f5f44e5-849e-4306-af7b-e65a6ac40813",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(calculate_new == FALSE){\n",
    "    overview_data = overview_data[!overview_data$datasets_match %in% available_summaries$datasets,]\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e8f861-be62-4ae8-b16a-df792c29fa4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c4f297b9-a2eb-4417-88fe-d9aa19181b36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5"
      ],
      "text/latex": [
       "5"
      ],
      "text/markdown": [
       "5"
      ],
      "text/plain": [
       "[1] 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(overview_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5b62e5e0-4151-4e5f-981a-7458e263b5e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Remove tmp Filtered objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "89dec1a5-327f-4248-8a9e-29140d29c8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data = overview_data[is.na(str_extract(overview_data$datasets, 'tmpFiltered')),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d08ff156-a9dd-4090-b39b-9c7d9edf34d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#overview_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a1d212e6-9b59-44fe-ae43-a361f63a1948",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_oneK1k</td><td>wg3_oneK1kB.Qced.Normalized.SCs.Rds    </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td><td>wg3_oneK1kCD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_oneK1k & wg3\\_oneK1kB.Qced.Normalized.SCs.Rds    \\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k & wg3\\_oneK1kCD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_oneK1k | wg3_oneK1kB.Qced.Normalized.SCs.Rds     |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_oneK1k | wg3_oneK1kCD4_T.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name      \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_oneK1k\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "  datasets_match                         \n",
       "1 wg3_oneK1kB.Qced.Normalized.SCs.Rds    \n",
       "2 wg3_oneK1kCD4_T.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(overview_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "45ca65d5-11a4-459d-82c5-384363ea93fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data = overview_data[overview_data$datasets != 'B.Qced.Normalized.SCs.Rds',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "84a9d387-8b08-4284-a79e-1befbf5f2b4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td><td>wg3_oneK1kCD8_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_oneK1k</td><td>wg3_oneK1kNK.Qced.Normalized.SCs.Rds   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k & wg3\\_oneK1kCD8\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_oneK1k & wg3\\_oneK1kNK.Qced.Normalized.SCs.Rds   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_oneK1k | wg3_oneK1kCD8_T.Qced.Normalized.SCs.Rds |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_oneK1k | wg3_oneK1kNK.Qced.Normalized.SCs.Rds    |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name      \n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_oneK1k\n",
       "  datasets_match                         \n",
       "3 wg3_oneK1kCD8_T.Qced.Normalized.SCs.Rds\n",
       "5 wg3_oneK1kNK.Qced.Normalized.SCs.Rds   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd4f437-a04a-4ea7-a115-99328b6ffb40",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Get summary statistics of input data and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5bb84060-4f15-4630-8539-6831235ac0d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'wg3_oneK1k'"
      ],
      "text/latex": [
       "'wg3\\_oneK1k'"
      ],
      "text/markdown": [
       "'wg3_oneK1k'"
      ],
      "text/plain": [
       "[1] \"wg3_oneK1k\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e16d7275-c5be-47d2-94b0-4ada06565f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data$data = overview_data$datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "baead12d-6bd8-4bdc-8cd6-4890e7434e67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CD8_T.Qced.Normalized.SCs.Rds'</li><li>'NK.Qced.Normalized.SCs.Rds'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CD8\\_T.Qced.Normalized.SCs.Rds'\n",
       "\\item 'NK.Qced.Normalized.SCs.Rds'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CD8_T.Qced.Normalized.SCs.Rds'\n",
       "2. 'NK.Qced.Normalized.SCs.Rds'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"CD8_T.Qced.Normalized.SCs.Rds\" \"NK.Qced.Normalized.SCs.Rds\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(unique(overview_data$data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "95d17ea6-4b38-49dd-8300-9d04b59dd99d",
   "metadata": {},
   "outputs": [],
   "source": [
    "choose_assay = 'data' # alternative: assay = 'data' or 'RNA' --> returns transformed values; whether to use normalized or raw counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "325c4a80-82aa-4eb7-85ce-13b4be0a7e13",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate empty dataframes to store calculated summaries in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "05a0a7cd-7ac4-4edb-b9b2-9b1c6b09bdc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_summary = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3114687c-4b8e-4186-a843-21d97d3ea07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_gene_statistic = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5886cfa5-c26a-4c4a-8a88-a2d2a5c6560f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_cell_statistic = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "0599936d-676e-4f06-aabf-dce8d64a8184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'CD8_T.Qced.Normalized.SCs.Rds'"
      ],
      "text/latex": [
       "'CD8\\_T.Qced.Normalized.SCs.Rds'"
      ],
      "text/markdown": [
       "'CD8_T.Qced.Normalized.SCs.Rds'"
      ],
      "text/plain": [
       "[1] \"CD8_T.Qced.Normalized.SCs.Rds\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data$data[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c4363775-399a-4854-a01f-f956b2573902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th><th scope=col>data</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_oneK1k</td><td>wg3_oneK1kCD8_T.Qced.Normalized.SCs.Rds</td><td>CD8_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_oneK1k</td><td>wg3_oneK1kNK.Qced.Normalized.SCs.Rds   </td><td>NK.Qced.Normalized.SCs.Rds   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & datasets & name & datasets\\_match & data\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_oneK1k & wg3\\_oneK1kCD8\\_T.Qced.Normalized.SCs.Rds & CD8\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_oneK1k & wg3\\_oneK1kNK.Qced.Normalized.SCs.Rds    & NK.Qced.Normalized.SCs.Rds   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 4\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; | data &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_oneK1k | wg3_oneK1kCD8_T.Qced.Normalized.SCs.Rds | CD8_T.Qced.Normalized.SCs.Rds |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_oneK1k | wg3_oneK1kNK.Qced.Normalized.SCs.Rds    | NK.Qced.Normalized.SCs.Rds    |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name      \n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_oneK1k\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_oneK1k\n",
       "  datasets_match                          data                         \n",
       "3 wg3_oneK1kCD8_T.Qced.Normalized.SCs.Rds CD8_T.Qced.Normalized.SCs.Rds\n",
       "5 wg3_oneK1kNK.Qced.Normalized.SCs.Rds    NK.Qced.Normalized.SCs.Rds   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b6ff9e1-bfbc-45b9-be2e-1aa89cd47462",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Calculate summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afee7c2c-e9c8-430f-b92b-684db5a8ba96",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"685\"\n",
      "[1] \"691\"\n",
      "[1] \"689\"\n",
      "[1] \"684\"\n",
      "[1] \"688\"\n",
      "[1] \"690\"\n",
      "[1] \"694\"\n",
      "[1] \"692\"\n",
      "[1] \"686\"\n",
      "[1] \"687\"\n",
      "[1] \"693\"\n",
      "[1] \"683\"\n",
      "[1] \"832\"\n",
      "[1] \"824\"\n",
      "[1] \"831\"\n",
      "[1] \"830\"\n",
      "[1] \"833\"\n",
      "[1] \"825\"\n",
      "[1] \"844\"\n",
      "[1] \"843\"\n",
      "[1] \"845\"\n",
      "[1] \"827\"\n",
      "[1] \"829\"\n",
      "[1] \"826\"\n",
      "[1] \"828\"\n",
      "[1] \"849\"\n",
      "[1] \"835\"\n",
      "[1] \"850\"\n",
      "[1] \"848\"\n",
      "[1] \"991\"\n",
      "[1] \"851\"\n",
      "[1] \"847\"\n",
      "[1] \"990\"\n",
      "[1] \"852\"\n",
      "[1] \"986\"\n",
      "[1] \"998\"\n",
      "[1] \"837\"\n",
      "[1] \"836\"\n",
      "[1] \"846\"\n",
      "[1] \"1000\"\n",
      "[1] \"993\"\n",
      "[1] \"813\"\n",
      "[1] \"838\"\n",
      "[1] \"996\"\n",
      "[1] \"812\"\n",
      "[1] \"994\"\n",
      "[1] \"999\"\n",
      "[1] \"997\"\n",
      "[1] \"995\"\n",
      "[1] \"992\"\n",
      "[1] \"811\"\n",
      "[1] \"853\"\n",
      "[1] \"987\"\n",
      "[1] \"856\"\n",
      "[1] \"988\"\n",
      "[1] \"858\"\n",
      "[1] \"857\"\n",
      "[1] \"1005\"\n",
      "[1] \"1003\"\n",
      "[1] \"855\"\n",
      "[1] \"859\"\n",
      "[1] \"1002\"\n",
      "[1] \"1001\"\n",
      "[1] \"1004\"\n",
      "[1] \"989\"\n",
      "[1] \"860\"\n",
      "[1] \"854\"\n",
      "[1] \"805\"\n",
      "[1] \"1008\"\n",
      "[1] \"862\"\n",
      "[1] \"1007\"\n",
      "[1] \"804\"\n",
      "[1] \"864\"\n",
      "[1] \"892\"\n",
      "[1] \"863\"\n",
      "[1] \"803\"\n",
      "[1] \"861\"\n",
      "[1] \"891\"\n",
      "[1] \"865\"\n",
      "[1] \"1006\"\n",
      "[1] \"806\"\n",
      "[1] \"1011\"\n",
      "[1] \"893\"\n",
      "[1] \"807\"\n",
      "[1] \"1014\"\n",
      "[1] \"1012\"\n",
      "[1] \"1013\"\n",
      "[1] \"809\"\n",
      "[1] \"808\"\n",
      "[1] \"1015\"\n",
      "[1] \"1009\"\n",
      "[1] \"810\"\n",
      "[1] \"894\"\n",
      "[1] \"1010\"\n",
      "[1] \"1017\"\n",
      "[1] \"1019\"\n",
      "[1] \"1022\"\n",
      "[1] \"1023\"\n",
      "[1] \"897\"\n",
      "[1] \"1021\"\n",
      "[1] \"1020\"\n",
      "[1] \"899\"\n",
      "[1] \"895\"\n",
      "[1] \"902\"\n",
      "[1] \"900\"\n",
      "[1] \"901\"\n",
      "[1] \"1016\"\n",
      "[1] \"1026\"\n",
      "[1] \"1028\"\n",
      "[1] \"1025\"\n",
      "[1] \"1027\"\n",
      "[1] \"1031\"\n",
      "[1] \"1034\"\n",
      "[1] \"1032\"\n",
      "[1] \"904\"\n",
      "[1] \"905\"\n",
      "[1] \"903\"\n",
      "[1] \"1029\"\n",
      "[1] \"1030\"\n",
      "[1] \"1033\"\n",
      "[1] \"906\"\n",
      "[1] \"1037\"\n",
      "[1] \"907\"\n",
      "[1] \"1039\"\n",
      "[1] \"909\"\n",
      "[1] \"1041\"\n",
      "[1] \"911\"\n",
      "[1] \"1035\"\n",
      "[1] \"1038\"\n",
      "[1] \"908\"\n",
      "[1] \"913\"\n",
      "[1] \"912\"\n",
      "[1] \"910\"\n",
      "[1] \"1036\"\n",
      "[1] \"1040\"\n",
      "[1] \"915\"\n",
      "[1] \"1043\"\n",
      "[1] \"914\"\n",
      "[1] \"917\"\n",
      "[1] \"918\"\n",
      "[1] \"695\"\n",
      "[1] \"873\"\n",
      "[1] \"1042\"\n",
      "[1] \"868\"\n",
      "[1] \"916\"\n",
      "[1] \"869\"\n",
      "[1] \"872\"\n",
      "[1] \"697\"\n",
      "[1] \"696\"\n",
      "[1] \"698\"\n",
      "[1] \"705\"\n",
      "[1] \"700\"\n",
      "[1] \"665\"\n",
      "[1] \"663\"\n",
      "[1] \"662\"\n",
      "[1] \"702\"\n",
      "[1] \"664\"\n",
      "[1] \"703\"\n",
      "[1] \"699\"\n",
      "[1] \"882\"\n",
      "[1] \"886\"\n",
      "[1] \"884\"\n",
      "[1] \"885\"\n",
      "[1] \"880\"\n",
      "[1] \"881\"\n",
      "[1] \"875\"\n",
      "[1] \"876\"\n",
      "[1] \"879\"\n",
      "[1] \"919\"\n",
      "[1] \"878\"\n",
      "[1] \"877\"\n",
      "[1] \"883\"\n",
      "[1] \"874\"\n",
      "[1] \"890\"\n",
      "[1] \"378\"\n",
      "[1] \"379\"\n",
      "[1] \"375\"\n",
      "[1] \"373\"\n",
      "[1] \"380\"\n",
      "[1] \"887\"\n",
      "[1] \"376\"\n",
      "[1] \"371\"\n",
      "[1] \"889\"\n",
      "[1] \"888\"\n",
      "[1] \"374\"\n",
      "[1] \"377\"\n",
      "[1] \"372\"\n",
      "[1] \"306\"\n",
      "[1] \"304\"\n",
      "[1] \"301\"\n",
      "[1] \"381\"\n",
      "[1] \"300\"\n",
      "[1] \"297\"\n",
      "[1] \"303\"\n",
      "[1] \"295\"\n",
      "[1] \"305\"\n",
      "[1] \"299\"\n",
      "[1] \"302\"\n",
      "[1] \"296\"\n",
      "[1] \"298\"\n",
      "[1] \"294\"\n",
      "[1] \"307\"\n",
      "[1] \"318\"\n",
      "[1] \"317\"\n",
      "[1] \"319\"\n",
      "[1] \"313\"\n",
      "[1] \"315\"\n",
      "[1] \"311\"\n",
      "[1] \"310\"\n",
      "[1] \"312\"\n",
      "[1] \"293\"\n",
      "[1] \"308\"\n",
      "[1] \"314\"\n",
      "[1] \"316\"\n",
      "[1] \"309\"\n",
      "[1] \"326\"\n",
      "[1] \"1044\"\n",
      "[1] \"1046\"\n",
      "[1] \"322\"\n",
      "[1] \"1049\"\n",
      "[1] \"325\"\n",
      "[1] \"1047\"\n",
      "[1] \"324\"\n",
      "[1] \"1045\"\n",
      "[1] \"327\"\n",
      "[1] \"320\"\n",
      "[1] \"323\"\n",
      "[1] \"321\"\n",
      "[1] \"1048\"\n",
      "[1] \"1056\"\n",
      "[1] \"332\"\n",
      "[1] \"1057\"\n",
      "[1] \"1055\"\n",
      "[1] \"1058\"\n",
      "[1] \"1051\"\n",
      "[1] \"333\"\n",
      "[1] \"334\"\n",
      "[1] \"1054\"\n",
      "[1] \"1053\"\n",
      "[1] \"331\"\n",
      "[1] \"1052\"\n",
      "[1] \"1050\"\n",
      "[1] \"328\"\n",
      "[1] \"923\"\n",
      "[1] \"1059\"\n",
      "[1] \"1062\"\n",
      "[1] \"1060\"\n",
      "[1] \"1061\"\n",
      "[1] \"920\"\n",
      "[1] \"925\"\n",
      "[1] \"1064\"\n",
      "[1] \"1065\"\n",
      "[1] \"924\"\n",
      "[1] \"335\"\n",
      "[1] \"922\"\n",
      "[1] \"1066\"\n",
      "[1] \"921\"\n",
      "[1] \"1063\"\n",
      "[1] \"761\"\n",
      "[1] \"762\"\n",
      "[1] \"763\"\n",
      "[1] \"770\"\n",
      "[1] \"755\"\n",
      "[1] \"759\"\n",
      "[1] \"756\"\n",
      "[1] \"760\"\n",
      "[1] \"754\"\n",
      "[1] \"771\"\n",
      "[1] \"757\"\n",
      "[1] \"758\"\n",
      "[1] \"292\"\n",
      "[1] \"764\"\n",
      "[1] \"775\"\n",
      "[1] \"772\"\n",
      "[1] \"768\"\n",
      "[1] \"779\"\n",
      "[1] \"773\"\n",
      "[1] \"766\"\n",
      "[1] \"774\"\n",
      "[1] \"776\"\n",
      "[1] \"778\"\n",
      "[1] \"765\"\n",
      "[1] \"769\"\n",
      "[1] \"767\"\n",
      "[1] \"777\"\n",
      "[1] \"791\"\n",
      "[1] \"783\"\n",
      "[1] \"788\"\n",
      "[1] \"787\"\n",
      "[1] \"792\"\n",
      "[1] \"780\"\n",
      "[1] \"793\"\n",
      "[1] \"785\"\n",
      "[1] \"789\"\n",
      "[1] \"781\"\n",
      "[1] \"782\"\n",
      "[1] \"786\"\n",
      "[1] \"790\"\n",
      "[1] \"784\"\n",
      "[1] \"712\"\n",
      "[1] \"710\"\n",
      "[1] \"709\"\n",
      "[1] \"671\"\n",
      "[1] \"668\"\n",
      "[1] \"707\"\n",
      "[1] \"669\"\n",
      "[1] \"666\"\n",
      "[1] \"667\"\n",
      "[1] \"711\"\n",
      "[1] \"706\"\n",
      "[1] \"670\"\n",
      "[1] \"708\"\n",
      "[1] \"454\"\n",
      "[1] \"407\"\n",
      "[1] \"460\"\n",
      "[1] \"458\"\n",
      "[1] \"455\"\n",
      "[1] \"446\"\n",
      "[1] \"431\"\n",
      "[1] \"457\"\n"
     ]
    }
   ],
   "source": [
    "for(i in 1:nrow(overview_data)){\n",
    "    path = paste0(data_path_sc, overview_data$name[i], '/input/L1/', overview_data$data[i])\n",
    "    data = get_dataset(path)   \n",
    "    #print(data)\n",
    "    #print(unique(data$Assignment))\n",
    "    \n",
    "    ## Get overview on amount samples: sample_summary\n",
    "    sample_data = data[[]] %>% group_by(Assignment) %>% summarise(amount_cells = n())\n",
    "    sample_data$dataset = overview_data$name[i]\n",
    "    sample_data$cell_type = overview_data$data[i]\n",
    "    sample_summary = rbind(sample_summary, sample_data)  # TBD outside of loop!\n",
    "\n",
    "\n",
    "    for(k in unique(data$Assignment)) {     \n",
    "            \n",
    "        ## Get summary for each gene and sample: sample_gene_statistic\n",
    "        sample_data = data[,data$Assignment == k ]\n",
    "        #sample_data_trans = GetAssayData(sample_data ,assay = 'data')\n",
    "        filter_data = GetAssayData(sample_data ,assay = choose_assay)   # need raw counts to determine at least 10 non-zero counts\n",
    "        sample_data = GetAssayData(sample_data ,assay =choose_assay)  # alternative: assay = 'data' --> returns transformed values\n",
    "     \n",
    "            ### Filter genes with at leat 10 non-zero counts\n",
    "            filter_data = filter_data[rowSums(filter_data !=  0) >= 10,]\n",
    "            sample_data = sample_data[rownames(sample_data) %in% rownames(filter_data),]\n",
    "\n",
    "            ### convert format\n",
    "            sample_data = as.matrix(sample_data)\n",
    "            #sample_data_trans = as.matrix(sample_data_trans)\n",
    "\n",
    "            data_long = melt(sample_data)\n",
    "            #data_long_trans = melt(sample_data_trans)\n",
    "\n",
    "            colnames(data_long) = c('gene', 'cell', 'counts')\n",
    "            #colnames(data_long_trans) = c('gene', 'cell', 'trans_counts')\n",
    "\n",
    "            ### get summaries\n",
    "            summary_data = data_long %>% group_by(gene) %>% summarise(\n",
    "                amount_cells = n(),\n",
    "                amount_zero_counts = sum(counts == 0),\n",
    "                mean_counts = mean(counts, na.rm = TRUE),\n",
    "                max_counts = max(counts, na.rm = TRUE),\n",
    "                min_counts = min(counts, na.rm = TRUE),\n",
    "                variance_counts = var(counts, na.rm = TRUE),\n",
    "                #bin_size = as.numeric(try(fitdistr(counts, 'negative binomial')$estimate[1])), \n",
    "                #bin_mu= as.numeric(try( fitdistr(counts, 'negative binomial')$estimate[2]))\n",
    "\n",
    "            )\n",
    "\n",
    "            summary_data$percentage_zero = summary_data$amount_zero_counts / summary_data$amount_cells\n",
    "            summary_data$amount_non_zero_counts = summary_data$amount_cells - summary_data$amount_zero_counts\n",
    "            summary_data$percentage_zero =  summary_data$percentage_zero*100\n",
    "            ## bin different columns\n",
    "            summary_data =  summary_data %>% mutate(amount_cells_bin = cut(amount_cells, breaks=c(0,25, 50, 100,250, 500, 750, 1500)))\n",
    "            summary_data =  summary_data %>% mutate(percentage_zero_bin = cut(percentage_zero, breaks=c(-0.001,5, 10, 20,40, 60, 80,90, 95, 100)))\n",
    "\n",
    "\n",
    "            summary_data$sample = k\n",
    "            summary_data$cell_type = overview_data$data[i]\n",
    "            summary_data$dataset = overview_data$name[i]\n",
    "\n",
    "            sample_gene_statistic = rbind(sample_gene_statistic, summary_data)\n",
    "        #\n",
    "        \n",
    "        # Get summaries for each sample and each cell: sample_cell_statistic\n",
    "        sample_data = data[,data$Assignment == k ]\n",
    "        sample_data = GetAssayData(sample_data ,assay =choose_assay)  # alternative: assay = 'data' --> returns transformed values\n",
    "\n",
    "            ### convert format\n",
    "            sample_data = as.matrix(sample_data)\n",
    "\n",
    "            data_long = melt(sample_data)\n",
    "\n",
    "            colnames(data_long) = c('gene', 'cell', 'counts')\n",
    "\n",
    "            ### get summaries\n",
    "            summary_data = data_long %>% group_by(cell) %>% summarise(\n",
    "                amount_counts = sum(counts),\n",
    "                amount_genes = n(),\n",
    "                amount_zero_counts = sum(counts == 0),\n",
    "                mean_counts = mean(counts,na.rm = TRUE),\n",
    "                max_counts = max(counts,na.rm = TRUE),\n",
    "                min_counts = min(counts,na.rm = TRUE),\n",
    "                variance_counts = var(counts,na.rm = TRUE))\n",
    "\n",
    "            summary_data$percentage_zero = summary_data$amount_zero_counts / summary_data$amount_genes\n",
    "            summary_data$amount_non_zero_counts = summary_data$amount_genes - summary_data$amount_zero_counts\n",
    "\n",
    "\n",
    "            summary_data$sample = k\n",
    "            summary_data$cell_type = overview_data$data[i]\n",
    "            summary_data$dataset = overview_data$name[i]\n",
    "\n",
    "            sample_cell_statistic = rbind(sample_cell_statistic, summary_data)\n",
    "        \n",
    "            write.csv(sample_summary, paste0(output_path_sc, 'D1_Sample_summary', overview_data$name[i], overview_data$data[i], '_', choose_assay, '.csv'))\n",
    "            write.csv(sample_gene_statistic, paste0(output_path_sc, 'D1_Sample_gene_statistic', overview_data$name[i], overview_data$data[i], '_', choose_assay, '.csv'))\n",
    "            write.csv(sample_cell_statistic, paste0(output_path_sc, 'D1_Sample_cell_statistic', overview_data$name[i], overview_data$data[i], '_', choose_assay, '.csv'))\n",
    "            \n",
    "            flush.console()\n",
    "            print(k)\n",
    "            flush.console()\n",
    "                \n",
    "            \n",
    "            \n",
    "        \n",
    "        \n",
    "        }\n",
    "        flush.console()\n",
    "        print(i)\n",
    "        flush.console()\n",
    "    \n",
    "    \n",
    "        sample_summary = data.frame()\n",
    "        sample_gene_statistic = data.frame()\n",
    "        sample_cell_statistic = data.frame()\n",
    "    }\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a3f6d80-deb3-4981-9e62-0d350a3dc206",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Implementation for multiome data; needs to be run differently (multiple observations per subject)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1de7b6f-1f2a-4ebf-beb7-c1cd9cae14fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Path to multiome data\n",
    "path = \"../data/current/wg3/wg3_multiome/input/L1/CD4_T.Qced.Normalized.SCs.Rds\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "68d8eabc-f78a-4413-b288-1a634f167a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "name = 'wg3_multiome'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "84924dea-96a3-4c4b-a83a-462628296e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path_sc  =  \"../data/current/summary_stats/\""
   ]
  },
  {
   "cell_type": "raw",
   "id": "ae19d621-f3e7-4343-977d-9901d750c938",
   "metadata": {},
   "source": [
    "data = get_dataset(path)   \n",
    "    #print(data)\n",
    "    #print(unique(data$Assignment))\n",
    "\n",
    "data$Assignment = data$sample_final\n",
    "\n",
    "## Get overview on amount samples: sample_summary\n",
    "sample_data = data[[]] %>% group_by(Assignment) %>% summarise(amount_cells = n())\n",
    "sample_data$dataset =  str_replace(path, '/groups.*L1/', '')\n",
    "sample_data$cell_type =  str_replace(path, '/groups.*L1/', '')\n",
    "sample_summary = rbind(sample_summary, sample_data)  # TBD outside of loop!\n",
    "\n",
    "\n",
    "for(k in unique(data$Assignment)) {     \n",
    "\n",
    "    ## Get summary for each gene and sample: sample_gene_statistic\n",
    "    sample_data = data[,data$Assignment == k ]\n",
    "    #sample_data_trans = GetAssayData(sample_data ,assay = 'data')\n",
    "    filter_data = GetAssayData(sample_data)   # need raw counts to determine at least 10 non-zero counts\n",
    "    sample_data = GetAssayData(sample_data )  # alternative: assay = 'data' --> returns transformed values\n",
    "\n",
    "        ### Filter genes with at leat 10 non-zero counts\n",
    "        filter_data = filter_data[rowSums(filter_data != 0) >= 10,]\n",
    "        sample_data = sample_data[rownames(sample_data) %in% rownames(filter_data),]\n",
    "\n",
    "        ### convert format\n",
    "        sample_data = as.matrix(sample_data)\n",
    "        #sample_data_trans = as.matrix(sample_data_trans)\n",
    "\n",
    "        data_long = melt(sample_data)\n",
    "        #data_long_trans = melt(sample_data_trans)\n",
    "\n",
    "        colnames(data_long) = c('gene', 'cell', 'counts')\n",
    "        #colnames(data_long_trans) = c('gene', 'cell', 'trans_counts')\n",
    "\n",
    "        ### get summaries\n",
    "        summary_data = data_long %>% group_by(gene) %>% summarise(\n",
    "            amount_cells = n(),\n",
    "            amount_zero_counts = sum(counts == 0),\n",
    "            mean_counts = mean(counts, na.rm = TRUE),\n",
    "            max_counts = max(counts, na.rm = TRUE),\n",
    "            min_counts = min(counts, na.rm = TRUE),\n",
    "            variance_counts = var(counts, na.rm = TRUE),\n",
    "            #bin_size = as.numeric(try(fitdistr(counts, 'negative binomial')$estimate[1])), \n",
    "            #bin_mu= as.numeric(try( fitdistr(counts, 'negative binomial')$estimate[2]))\n",
    "\n",
    "        )\n",
    "\n",
    "        summary_data$percentage_zero = summary_data$amount_zero_counts / summary_data$amount_cells\n",
    "        summary_data$amount_non_zero_counts = summary_data$amount_cells - summary_data$amount_zero_counts\n",
    "        summary_data$percentage_zero =  summary_data$percentage_zero*100\n",
    "        ## bin different columns\n",
    "        summary_data =  summary_data %>% mutate(amount_cells_bin = cut(amount_cells, breaks=c(0,25, 50, 100,250, 500, 750, 1500)))\n",
    "        summary_data =  summary_data %>% mutate(percentage_zero_bin = cut(percentage_zero, breaks=c(-0.001,5, 10, 20,40, 60, 80,90, 95, 100)))\n",
    "\n",
    "\n",
    "        summary_data$sample = k\n",
    "        summary_data$cell_type = str_replace(path, '/groups.*L1/', '')\n",
    "        summary_data$dataset =str_replace(path, '/groups.*L1/', '')\n",
    "\n",
    "        sample_gene_statistic = rbind(sample_gene_statistic, summary_data)\n",
    "    #\n",
    "\n",
    "    # Get summaries for each sample and each cell: sample_cell_statistic\n",
    "    sample_data = data[,data$Assignment == k ]\n",
    "    sample_data = GetAssayData(sample_data)  # alternative: assay = 'data' --> returns transformed values\n",
    "\n",
    "        ### convert format\n",
    "        sample_data = as.matrix(sample_data)\n",
    "\n",
    "        data_long = melt(sample_data)\n",
    "\n",
    "        colnames(data_long) = c('gene', 'cell', 'counts')\n",
    "\n",
    "        ### get summaries\n",
    "        summary_data = data_long %>% group_by(cell) %>% summarise(\n",
    "            amount_counts = sum(counts),\n",
    "            amount_genes = n(),\n",
    "            amount_zero_counts = sum(counts == 0),\n",
    "            mean_counts = mean(counts,na.rm = TRUE),\n",
    "            max_counts = max(counts,na.rm = TRUE),\n",
    "            min_counts = min(counts,na.rm = TRUE),\n",
    "            variance_counts = var(counts,na.rm = TRUE))\n",
    "\n",
    "        summary_data$percentage_zero = summary_data$amount_zero_counts / summary_data$amount_genes\n",
    "        summary_data$amount_non_zero_counts = summary_data$amount_genes - summary_data$amount_zero_counts\n",
    "\n",
    "\n",
    "        summary_data$sample = k\n",
    "        summary_data$cell_type =str_replace(path, '/groups.*L1/', '')\n",
    "        summary_data$dataset = str_replace(path, '/groups.*L1/', '')\n",
    "\n",
    "        sample_cell_statistic = rbind(sample_cell_statistic, summary_data)\n",
    "\n",
    "        write.csv(sample_summary, paste0(output_path_sc, 'D1_Sample_summary',name,  str_replace(path, '/groups.*L1/', '') , '_', choose_assay, '.csv'))\n",
    "        write.csv(sample_gene_statistic, paste0(output_path_sc, 'D1_Sample_gene_statistic', name, str_replace(path, '/groups.*L1/', ''), '_', choose_assay, '.csv'))\n",
    "        write.csv(sample_cell_statistic, paste0(output_path_sc, 'D1_Sample_cell_statistic', name, str_replace(path, '/groups.*L1/', ''), '_', choose_assay, '.csv'))\n",
    "\n",
    "        flush.console()\n",
    "        print(k)\n",
    "        flush.console()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    }\n",
    "    flush.console()\n",
    "    print(i)\n",
    "    flush.console()\n",
    "\n",
    "\n",
    "    sample_summary = data.frame()\n",
    "    sample_gene_statistic = data.frame()\n",
    "    sample_cell_statistic = data.frame()\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc-autonumbering": true,
  "toc-showcode": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
