{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "a1444178-491b-4528-9c7d-b171aa9267c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate the summary statistics for the single-cell datasets similar as in D1\n",
    "## Compared to D1: only calculate for a subset of genes (e.g. random sample, only co-eQTL genes etc + include estimation of negative binomial distribution parameters\n",
    "## Used as input for simulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dca7889-aa95-41fc-84f2-fc38b8392196",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "id": "49dfe76f-f0c5-476d-8c73-e721c2a6e069",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "\"incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f55b4c93-827f-4edf-b237-d399f128ac28",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "id": "e8e76471-b038-407f-999a-1735bbb7a186",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to single-cell datasets (oneK1K, Wijst, UMCG2, UMCG3, multiome; stored as .Rds data files)\n",
    "\n",
    "data_path_sc = \"../data/current/wg3/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "id": "e814f197-db05-450a-86e2-c24ff80b842f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to save calculated summary statistics to\n",
    "output_path_sc = \"../data/current/summary_stats/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f49efb2a-e46e-4e25-a8ce-2c83aa4bc78b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Select cell-type and gene list to calculate the summary stats for"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "id": "d3f4857d-972b-48b4-9307-9561634128e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = 'NK'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "id": "738f7966-e1b9-4e0c-a8b7-0ab3a3550a47",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list_path =  \"/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/coeqtl_mapping/co_qtls_decision_tree/shared_results/5DS_Meta_Analysis_Sign_coeQTLs133p_val_mt_eGene_.csv\"\n",
    "#gene_list_path= paste0('../data/current/coeqtl_mapping/co_qtls_decision_tree/analysis_oneK1K/', cell_type , '/')\n",
    "# '../data/current/coeqtl_mapping/co_qtls_sceqtlgen/oneK1K/CD4_T/'   # path to significant genes from co-eqtl mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64663cd5-8d6a-454a-8d0c-ff44924cee9d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "id": "494ae5c4-500f-4367-9f43-ceb7c2c5667c",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dd6604-4d42-4bb0-81ed-9555d63ddcb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "45cfaa89-153c-4abe-92f5-868c50e8090b",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6123ec7d-d87b-44df-b708-45126bb587db",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load different sc dataset and get some meta-information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c66cc1b5-d586-4860-9428-793a51b5ab98",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Select datasets for which summary statistics should be calculated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "60112fac-cddd-4a5d-8a22-2201fc7f7c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "folders = c('wg3_wijst2018'#, \n",
    "            #'wg3_Trynka', \n",
    "           # 'wg3_oneK1k'#,\n",
    "           # 'wg3_NC2020',\n",
    "            #'wg3_Nawijn',\n",
    "            #'wg3_Li',\n",
    "           # 'wg3_Ye',\n",
    "           #  'wg3_okada',\n",
    "           #   'wg3_sawcer',\n",
    "           # 'wg3_Franke_split_v3',\n",
    "          #  'wg3_Franke_split_v2',\n",
    "          #  'wg3_blokland2020'\n",
    "           )  ## specify folders that contain datasets that should be included"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "7de77141-b57b-4e05-a928-e0220deb3b2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "id": "d3da6875-238f-44e4-86bc-14af085abdc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "id": "416ec510-f876-405d-af0a-bde54e2237b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get all available datasets overview\n",
    "for(i in folders){\n",
    "\n",
    "    dataset_overview = data.frame( datasets = list.files(paste0(data_path_sc, i, '/input/L1')))\n",
    "    if(nrow(dataset_overview) >= 1){\n",
    "        dataset_overview$name = i\n",
    "        dataset_overview = dataset_overview[!is.na(str_extract(dataset_overview$datasets, '.Rds')),]\n",
    "        overview_data = rbind(overview_data, dataset_overview)\n",
    "      }\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "id": "74b84d59-4da8-40fb-afdc-330ff1964563",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Mono.Qced.Normalized.SCs.Rds </td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_wijst2018</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & datasets & name\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_wijst2018\\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\t4 & Mono.Qced.Normalized.SCs.Rds  & wg3\\_wijst2018\\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_wijst2018\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 2\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_wijst2018 |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "| 4 | Mono.Qced.Normalized.SCs.Rds  | wg3_wijst2018 |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_wijst2018 |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name         \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_wijst2018\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "4 Mono.Qced.Normalized.SCs.Rds  wg3_wijst2018\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_wijst2018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dataset_overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "id": "7b52b80c-df6e-4a4f-b974-de354ea40e16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Mono.Qced.Normalized.SCs.Rds </td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_wijst2018</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & datasets & name\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_wijst2018\\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\t4 & Mono.Qced.Normalized.SCs.Rds  & wg3\\_wijst2018\\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_wijst2018\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 2\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_wijst2018 |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "| 4 | Mono.Qced.Normalized.SCs.Rds  | wg3_wijst2018 |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_wijst2018 |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name         \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_wijst2018\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "4 Mono.Qced.Normalized.SCs.Rds  wg3_wijst2018\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_wijst2018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(overview_data,20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "id": "57df4271-bfc2-434d-b2c2-a54ed61dc704",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data$datasets_match = paste0(overview_data$name, overview_data$datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "id": "d701f5ce-ebb0-4bf0-b56c-5c827e6bc2fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td><td>wg3_wijst2018B.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds & wg3\\_wijst2018 & wg3\\_wijst2018B.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds | wg3_wijst2018 | wg3_wijst2018B.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                  name         \n",
       "1 B.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "  datasets_match                        \n",
       "1 wg3_wijst2018B.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data[1,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "id": "7bd88595-715c-4336-ae8c-dc7b799fd8b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1"
      ],
      "text/latex": [
       "1"
      ],
      "text/markdown": [
       "1"
      ],
      "text/plain": [
       "[1] 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(overview_data$name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "id": "09f7a0d3-0f3e-44b8-9d8a-88bf25c72fdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5"
      ],
      "text/latex": [
       "5"
      ],
      "text/markdown": [
       "5"
      ],
      "text/plain": [
       "[1] 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(overview_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "id": "73457506-c230-4e36-8fed-01e99efd6fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load dataset\n",
    "get_dataset = function(path){\n",
    "    data = readRDS(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "id": "37e82f77-d695-47d8-b562-ca2755a554b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get already generated summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "id": "7713d4f5-0a3c-4fb8-a29d-4ae4e7c58e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "available_summaries = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "id": "2ce6f87a-9ddd-40bd-a4ec-8a16c74bb0aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "available_summaries= data.frame( datasets = list.files(paste0(output_path_sc)))\n",
    "available_summaries = data.frame(datasets = available_summaries[!is.na(str_extract(available_summaries$datasets, 'D1_Sample_cell_statistic')),])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "id": "5e54f066-d50e-44d9-8ce2-882e491baead",
   "metadata": {},
   "outputs": [],
   "source": [
    "available_summaries$datasets = str_replace(available_summaries$datasets, 'D1_Sample_cell_statistic', '')\n",
    "available_summaries$datasets = str_replace(available_summaries$datasets, '.csv', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "id": "71bd721b-7727-46df-abea-65cdb37915d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "177"
      ],
      "text/latex": [
       "177"
      ],
      "text/markdown": [
       "177"
      ],
      "text/plain": [
       "[1] 177"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(available_summaries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "id": "b25b82c7-0092-434b-a7f2-2b976432b813",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>wg3_blokland2020B.Qced.Normalized.SCs.Rds    </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>wg3_blokland2020CD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 1\n",
       "\\begin{tabular}{r|l}\n",
       "  & datasets\\\\\n",
       "  & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & wg3\\_blokland2020B.Qced.Normalized.SCs.Rds    \\\\\n",
       "\t2 & wg3\\_blokland2020CD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 1\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; |\n",
       "|---|---|\n",
       "| 1 | wg3_blokland2020B.Qced.Normalized.SCs.Rds     |\n",
       "| 2 | wg3_blokland2020CD4_T.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                                     \n",
       "1 wg3_blokland2020B.Qced.Normalized.SCs.Rds    \n",
       "2 wg3_blokland2020CD4_T.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(available_summaries,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 516,
   "id": "87a91bea-b453-48d9-9163-59fd24a103b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5"
      ],
      "text/latex": [
       "5"
      ],
      "text/markdown": [
       "5"
      ],
      "text/plain": [
       "[1] 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(overview_data[overview_data$datasets_match %in% available_summaries$datasets,]) # already existing summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea54eae-464f-49dd-846a-a29109fbca76",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 517,
   "id": "cfd28d72-fa0e-442c-bd55-b7f382662b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove already existing sommaries from overview data to avoid rerunning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 518,
   "id": "1ff907ff-c789-412c-8889-5e0af088b4ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "calculate_new = TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 519,
   "id": "3f5f44e5-849e-4306-af7b-e65a6ac40813",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(calculate_new == FALSE){\n",
    "    overview_data = overview_data[!overview_data$datasets_match %in% available_summaries$datasets,]\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e8f861-be62-4ae8-b16a-df792c29fa4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 520,
   "id": "c4f297b9-a2eb-4417-88fe-d9aa19181b36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5"
      ],
      "text/latex": [
       "5"
      ],
      "text/markdown": [
       "5"
      ],
      "text/plain": [
       "[1] 5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(overview_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 521,
   "id": "5b62e5e0-4151-4e5f-981a-7458e263b5e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Remove tmp Filtered objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 522,
   "id": "89dec1a5-327f-4248-8a9e-29140d29c8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data = overview_data[is.na(str_extract(overview_data$datasets, 'tmpFiltered')),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 523,
   "id": "d08ff156-a9dd-4090-b39b-9c7d9edf34d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#overview_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "id": "a1d212e6-9b59-44fe-ae43-a361f63a1948",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_wijst2018</td><td>wg3_wijst2018B.Qced.Normalized.SCs.Rds    </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td><td>wg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_wijst2018 & wg3\\_wijst2018B.Qced.Normalized.SCs.Rds    \\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018 & wg3\\_wijst2018CD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_wijst2018 | wg3_wijst2018B.Qced.Normalized.SCs.Rds     |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 | wg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name         \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_wijst2018\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "  datasets_match                            \n",
       "1 wg3_wijst2018B.Qced.Normalized.SCs.Rds    \n",
       "2 wg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(overview_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "id": "84a9d387-8b08-4284-a79e-1befbf5f2b4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>B.Qced.Normalized.SCs.Rds    </td><td>wg3_wijst2018</td><td>wg3_wijst2018B.Qced.Normalized.SCs.Rds    </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td><td>wg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>CD8_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td><td>wg3_wijst2018CD8_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Mono.Qced.Normalized.SCs.Rds </td><td>wg3_wijst2018</td><td>wg3_wijst2018Mono.Qced.Normalized.SCs.Rds </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds   </td><td>wg3_wijst2018</td><td>wg3_wijst2018NK.Qced.Normalized.SCs.Rds   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & B.Qced.Normalized.SCs.Rds     & wg3\\_wijst2018 & wg3\\_wijst2018B.Qced.Normalized.SCs.Rds    \\\\\n",
       "\t2 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018 & wg3\\_wijst2018CD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\t3 & CD8\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018 & wg3\\_wijst2018CD8\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\t4 & Mono.Qced.Normalized.SCs.Rds  & wg3\\_wijst2018 & wg3\\_wijst2018Mono.Qced.Normalized.SCs.Rds \\\\\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds    & wg3\\_wijst2018 & wg3\\_wijst2018NK.Qced.Normalized.SCs.Rds   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | B.Qced.Normalized.SCs.Rds     | wg3_wijst2018 | wg3_wijst2018B.Qced.Normalized.SCs.Rds     |\n",
       "| 2 | CD4_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 | wg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds |\n",
       "| 3 | CD8_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 | wg3_wijst2018CD8_T.Qced.Normalized.SCs.Rds |\n",
       "| 4 | Mono.Qced.Normalized.SCs.Rds  | wg3_wijst2018 | wg3_wijst2018Mono.Qced.Normalized.SCs.Rds  |\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds    | wg3_wijst2018 | wg3_wijst2018NK.Qced.Normalized.SCs.Rds    |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                      name         \n",
       "1 B.Qced.Normalized.SCs.Rds     wg3_wijst2018\n",
       "2 CD4_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "3 CD8_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "4 Mono.Qced.Normalized.SCs.Rds  wg3_wijst2018\n",
       "5 NK.Qced.Normalized.SCs.Rds    wg3_wijst2018\n",
       "  datasets_match                            \n",
       "1 wg3_wijst2018B.Qced.Normalized.SCs.Rds    \n",
       "2 wg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds\n",
       "3 wg3_wijst2018CD8_T.Qced.Normalized.SCs.Rds\n",
       "4 wg3_wijst2018Mono.Qced.Normalized.SCs.Rds \n",
       "5 wg3_wijst2018NK.Qced.Normalized.SCs.Rds   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 526,
   "id": "31da542d-4a55-497e-b1b2-72faf176d69d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on cell-type which was selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 527,
   "id": "eaebb1dd-248f-49bc-b6ad-866c6b75d53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data = overview_data[!is.na(str_extract(overview_data$datasets, cell_type)),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 528,
   "id": "eb33c639-10db-4424-aed5-838527b3cc40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td><td>wg3_wijst2018NK.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & datasets & name & datasets\\_match\\\\\n",
       "  & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds & wg3\\_wijst2018 & wg3\\_wijst2018NK.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 3\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds | wg3_wijst2018 | wg3_wijst2018NK.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                   name         \n",
       "5 NK.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "  datasets_match                         \n",
       "5 wg3_wijst2018NK.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "833e1494-79cb-47d0-8292-d85e724fd32d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1a764e-29eb-4582-ac6d-d41cba6a7578",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4022ace2-c9b8-4d55-9568-020a7c4d5d85",
   "metadata": {},
   "source": [
    "## Load the list of genes for which to calculate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 529,
   "id": "bf38824e-4957-4991-9119-50b304df145c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_selection = 'random' # options: random, coeqtl --> random 2000 random genes will be chosen; coeqtl: significant co-eqtl genes are chosen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "id": "40f746a7-d494-4033-af4e-2d569bb9d5c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Use significant co-eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 531,
   "id": "332e8e0e-d391-4556-8dc7-4b6ef84ece36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/coeqtl_mapping/co_qtls_decision_tree/shared_results/5DS_Meta_Analysis_Sign_coeQTLs133p_val_mt_eGene_.csv'"
      ],
      "text/latex": [
       "'/ictstr01/groups/epigenereg01/workspace/projects/grn\\_dev\\_groningen/coeqtl\\_mapping/co\\_qtls\\_decision\\_tree/shared\\_results/5DS\\_Meta\\_Analysis\\_Sign\\_coeQTLs133p\\_val\\_mt\\_eGene\\_.csv'"
      ],
      "text/markdown": [
       "'/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/coeqtl_mapping/co_qtls_decision_tree/shared_results/5DS_Meta_Analysis_Sign_coeQTLs133p_val_mt_eGene_.csv'"
      ],
      "text/plain": [
       "[1] \"/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/coeqtl_mapping/co_qtls_decision_tree/shared_results/5DS_Meta_Analysis_Sign_coeQTLs133p_val_mt_eGene_.csv\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_list_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 532,
   "id": "4daafe43-6324-4d7d-9b04-ea59497c19c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtls = fread(paste0(gene_list_path), header = TRUE)\n",
    "\n",
    "# 'F3_decision_tree_input.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "id": "6e2bbc3c-8720-49cf-898f-9f5a0d3df1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter which ones to consider as significant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "id": "3466c8a9-c8ec-4187-a8ac-b611e9d67759",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "91562"
      ],
      "text/latex": [
       "91562"
      ],
      "text/markdown": [
       "91562"
      ],
      "text/plain": [
       "[1] 91562"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "id": "236ce97c-301b-4eb4-a6dd-81b58a9349ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "64198"
      ],
      "text/latex": [
       "64198"
      ],
      "text/markdown": [
       "64198"
      ],
      "text/plain": [
       "[1] 64198"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coeqtls$triplet))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "d77b91b5-b121-46f3-94e9-3dc4e822cabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtls = data.frame(coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "id": "fa8e500a-170e-4b83-998f-56f840bb9e6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on the cell-type\n",
    "coeqtls = coeqtls[coeqtls$cell_type == cell_type,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "id": "a7405d40-73ca-49ef-96ec-03edd21aec72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 166</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>eGene</th><th scope=col>coeGene</th><th scope=col>snp_id</th><th scope=col>triplet</th><th scope=col>feature_id</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>n_samples</th><th scope=col>p_val_mt</th><th scope=col>⋯</th><th scope=col>z_score.oneK1K.CD4_T</th><th scope=col>z_score.oneK1K.CD8_T</th><th scope=col>z_score.oneK1K.Mono</th><th scope=col>z_score.oneK1K.NK</th><th scope=col>z_score.wijst.B</th><th scope=col>z_score.wijst.CD4_T</th><th scope=col>z_score.wijst.CD8_T</th><th scope=col>z_score.wijst.Mono</th><th scope=col>z_score.wijst.NK</th><th scope=col>remap_overlap_tf_snp</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>373</th><td>373</td><td>AK5</td><td>CASK </td><td>1:77443300:A:G</td><td>1:77443300:A:G;AK5_CASK </td><td>AK5_CASK </td><td>NK</td><td>meta_5ds</td><td>499</td><td>2.170199e-07</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td> 7.877279</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>374</th><td>374</td><td>AK5</td><td>CD247</td><td>1:77443300:A:G</td><td>1:77443300:A:G;AK5_CD247</td><td>AK5_CD247</td><td>NK</td><td>meta_5ds</td><td>630</td><td>2.838419e-07</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>-7.947331</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 166\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & V1 & eGene & coeGene & snp\\_id & triplet & feature\\_id & cell\\_type & dataset & n\\_samples & p\\_val\\_mt & ⋯ & z\\_score.oneK1K.CD4\\_T & z\\_score.oneK1K.CD8\\_T & z\\_score.oneK1K.Mono & z\\_score.oneK1K.NK & z\\_score.wijst.B & z\\_score.wijst.CD4\\_T & z\\_score.wijst.CD8\\_T & z\\_score.wijst.Mono & z\\_score.wijst.NK & remap\\_overlap\\_tf\\_snp\\\\\n",
       "  & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t373 & 373 & AK5 & CASK  & 1:77443300:A:G & 1:77443300:A:G;AK5\\_CASK  & AK5\\_CASK  & NK & meta\\_5ds & 499 & 2.170199e-07 & ⋯ & NA & NA & NA &  7.877279 & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t374 & 374 & AK5 & CD247 & 1:77443300:A:G & 1:77443300:A:G;AK5\\_CD247 & AK5\\_CD247 & NK & meta\\_5ds & 630 & 2.838419e-07 & ⋯ & NA & NA & NA & -7.947331 & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 166\n",
       "\n",
       "| <!--/--> | V1 &lt;int&gt; | eGene &lt;chr&gt; | coeGene &lt;chr&gt; | snp_id &lt;chr&gt; | triplet &lt;chr&gt; | feature_id &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | n_samples &lt;int&gt; | p_val_mt &lt;dbl&gt; | ⋯ ⋯ | z_score.oneK1K.CD4_T &lt;dbl&gt; | z_score.oneK1K.CD8_T &lt;dbl&gt; | z_score.oneK1K.Mono &lt;dbl&gt; | z_score.oneK1K.NK &lt;dbl&gt; | z_score.wijst.B &lt;dbl&gt; | z_score.wijst.CD4_T &lt;dbl&gt; | z_score.wijst.CD8_T &lt;dbl&gt; | z_score.wijst.Mono &lt;dbl&gt; | z_score.wijst.NK &lt;dbl&gt; | remap_overlap_tf_snp &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 373 | 373 | AK5 | CASK  | 1:77443300:A:G | 1:77443300:A:G;AK5_CASK  | AK5_CASK  | NK | meta_5ds | 499 | 2.170199e-07 | ⋯ | NA | NA | NA |  7.877279 | NA | NA | NA | NA | NA | NA |\n",
       "| 374 | 374 | AK5 | CD247 | 1:77443300:A:G | 1:77443300:A:G;AK5_CD247 | AK5_CD247 | NK | meta_5ds | 630 | 2.838419e-07 | ⋯ | NA | NA | NA | -7.947331 | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "    V1  eGene coeGene snp_id         triplet                  feature_id\n",
       "373 373 AK5   CASK    1:77443300:A:G 1:77443300:A:G;AK5_CASK  AK5_CASK  \n",
       "374 374 AK5   CD247   1:77443300:A:G 1:77443300:A:G;AK5_CD247 AK5_CD247 \n",
       "    cell_type dataset  n_samples p_val_mt     ⋯ z_score.oneK1K.CD4_T\n",
       "373 NK        meta_5ds 499       2.170199e-07 ⋯ NA                  \n",
       "374 NK        meta_5ds 630       2.838419e-07 ⋯ NA                  \n",
       "    z_score.oneK1K.CD8_T z_score.oneK1K.Mono z_score.oneK1K.NK z_score.wijst.B\n",
       "373 NA                   NA                   7.877279         NA             \n",
       "374 NA                   NA                  -7.947331         NA             \n",
       "    z_score.wijst.CD4_T z_score.wijst.CD8_T z_score.wijst.Mono z_score.wijst.NK\n",
       "373 NA                  NA                  NA                 NA              \n",
       "374 NA                  NA                  NA                 NA              \n",
       "    remap_overlap_tf_snp\n",
       "373 NA                  \n",
       "374 NA                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coeqtls,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "id": "d2fbf61b-b4ca-4224-9eb9-603b5b98d3ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Alternative randomly select 2000 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "id": "7920f548-3d65-456f-9fa5-25a21d69bcfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = paste0(data_path_sc, unique(overview_data$name), '/input/L1/', unique(overview_data$datasets))\n",
    "data = get_dataset(path)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "945154a7-71da-4089-b7a9-f06dcd12a1a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "id": "499222d3-91b7-48c8-9122-bee26be604db",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(gene_selection == 'coeqtl'){\n",
    "    genes = unique(c(coeqtls$eGene, coeqtls$coeGene))\n",
    "    }\n",
    "if(gene_selection == 'random'){\n",
    "    set.seed(1)\n",
    "    genes = sample(unique(rownames(data)),5000)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 543,
   "id": "6337da1c-9c72-40e3-a9e7-e806fadb6466",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5000"
      ],
      "text/latex": [
       "5000"
      ],
      "text/markdown": [
       "5000"
      ],
      "text/plain": [
       "[1] 5000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "id": "3a6f912a-0869-4739-9734-2e56d4eede95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "49"
      ],
      "text/latex": [
       "49"
      ],
      "text/markdown": [
       "49"
      ],
      "text/plain": [
       "[1] 49"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coeqtls$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "eaa1f843-0497-41f7-ab39-4d6ee56b23ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "457"
      ],
      "text/latex": [
       "457"
      ],
      "text/markdown": [
       "457"
      ],
      "text/plain": [
       "[1] 457"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coeqtls$coeGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 546,
   "id": "7faedf0b-cb63-4511-aef5-4b7c62786624",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sum(unique(coeqtls$gene1) %in% unique(coeqtls$gene2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd4f437-a04a-4ea7-a115-99328b6ffb40",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Get summary statistics of input data and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "id": "5bb84060-4f15-4630-8539-6831235ac0d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'wg3_wijst2018'"
      ],
      "text/latex": [
       "'wg3\\_wijst2018'"
      ],
      "text/markdown": [
       "'wg3_wijst2018'"
      ],
      "text/plain": [
       "[1] \"wg3_wijst2018\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "id": "e16d7275-c5be-47d2-94b0-4ada06565f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "overview_data$data = overview_data$datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "id": "baead12d-6bd8-4bdc-8cd6-4890e7434e67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'NK.Qced.Normalized.SCs.Rds'"
      ],
      "text/latex": [
       "'NK.Qced.Normalized.SCs.Rds'"
      ],
      "text/markdown": [
       "'NK.Qced.Normalized.SCs.Rds'"
      ],
      "text/plain": [
       "[1] \"NK.Qced.Normalized.SCs.Rds\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(unique(overview_data$data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 550,
   "id": "95d17ea6-4b38-49dd-8300-9d04b59dd99d",
   "metadata": {},
   "outputs": [],
   "source": [
    "choose_assay = 'RNA' # alternative: assay = 'data' --> use 'data' for normalized values; 'RNA' for raw counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f29fcbee-f88a-4a8e-9b65-cd9a8db4efef",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate empty dataframes to store calculated summaries in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "id": "05a0a7cd-7ac4-4edb-b9b2-9b1c6b09bdc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_summary = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "id": "3114687c-4b8e-4186-a843-21d97d3ea07a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_gene_statistic = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 553,
   "id": "5886cfa5-c26a-4c4a-8a88-a2d2a5c6560f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_cell_statistic = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 554,
   "id": "0599936d-676e-4f06-aabf-dce8d64a8184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'NK.Qced.Normalized.SCs.Rds'"
      ],
      "text/latex": [
       "'NK.Qced.Normalized.SCs.Rds'"
      ],
      "text/markdown": [
       "'NK.Qced.Normalized.SCs.Rds'"
      ],
      "text/plain": [
       "[1] \"NK.Qced.Normalized.SCs.Rds\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data$data[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "id": "c4363775-399a-4854-a01f-f956b2573902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>datasets</th><th scope=col>name</th><th scope=col>datasets_match</th><th scope=col>data</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>5</th><td>NK.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td><td>wg3_wijst2018NK.Qced.Normalized.SCs.Rds</td><td>NK.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & datasets & name & datasets\\_match & data\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t5 & NK.Qced.Normalized.SCs.Rds & wg3\\_wijst2018 & wg3\\_wijst2018NK.Qced.Normalized.SCs.Rds & NK.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 4\n",
       "\n",
       "| <!--/--> | datasets &lt;chr&gt; | name &lt;chr&gt; | datasets_match &lt;chr&gt; | data &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 5 | NK.Qced.Normalized.SCs.Rds | wg3_wijst2018 | wg3_wijst2018NK.Qced.Normalized.SCs.Rds | NK.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  datasets                   name         \n",
       "5 NK.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "  datasets_match                          data                      \n",
       "5 wg3_wijst2018NK.Qced.Normalized.SCs.Rds NK.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overview_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b6ff9e1-bfbc-45b9-be2e-1aa89cd47462",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Calculate summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 556,
   "id": "afee7c2c-e9c8-430f-b92b-684db5a8ba96",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Layer counts isn't present in the assay object; returning NULL\"\n",
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 4 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 28: `gene = CCL4`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 3 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL007565\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 66 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 40: `gene = YWHAB`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 65 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL007985\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 12 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 83: `gene = CCL5`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 11 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL002176\"\n",
      "Error in fitdistr(counts, \"negative binomial\") : optimization failed\n",
      "Error in fitdistr(counts, \"negative binomial\") : optimization failed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 8 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 8: `gene = OASL`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 7 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL001060\"\n",
      "Error in solve.default(res$hessian) : \n",
      "  system is computationally singular: reciprocal condition number = 1.73247e-18\n",
      "Error in solve.default(res$hessian) : \n",
      "  system is computationally singular: reciprocal condition number = 1.73247e-18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 2: `gene = HLA-A`.\n",
      "Caused by warning:\n",
      "\u001b[33m!\u001b[39m NAs introduced by coercion\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL008200\"\n",
      "[1] \"LL013471\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 48 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 9: `gene = HP1BP3`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 47 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL011812\"\n",
      "Error in fitdistr(counts, \"negative binomial\") : optimization failed\n",
      "Error in fitdistr(counts, \"negative binomial\") : optimization failed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 8 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 57: `gene = CCL4`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 7 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL004587\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 18 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 27: `gene = BRK1`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 17 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL002215\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 280 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 15: `gene = ATP5F1D`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 279 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL016784\"\n",
      "[1] \"LL003444\"\n",
      "[1] \"LL013148\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 4 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 118: `gene = CCL5`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 3 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL003797\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 16 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 5: `gene = SDCBP`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 15 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL010388\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 62 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 9: `gene = SDCBP`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 61 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL009667\"\n",
      "[1] \"LL012835\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 12 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 14: `gene = REL`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 11 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL003026\"\n",
      "[1] \"LL015738\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 12 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 6: `gene = SDCBP`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 11 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL012786\"\n",
      "[1] \"LL002554\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 16 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 122: `gene = CCL4`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 15 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL009015\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 40: `gene = CCL5`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL002503\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 20 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 4: `gene = SH2D2A`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 19 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL007823\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 328 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 1: `gene = TSPOAP1-AS1`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 327 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL012354\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 96 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 3: `gene = SSR4`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 95 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL004195\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 186 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 11: `gene = SDCBP`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 185 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL005098\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 68 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 6: `gene = EIF3B`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 67 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL009191\"\n",
      "[1] \"LL016428\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 30 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 30: `gene = PRPF4B`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 29 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL014072\"\n",
      "Error in fitdistr(counts, \"negative binomial\") : optimization failed\n",
      "Error in fitdistr(counts, \"negative binomial\") : optimization failed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 9: `gene = DDX5`.\n",
      "Caused by warning:\n",
      "\u001b[33m!\u001b[39m NAs introduced by coercion\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL013965\"\n",
      "Error in fitdistr(counts, \"negative binomial\") : \n",
      "  'x' must be a non-empty numeric vector\n",
      "Error in fitdistr(counts, \"negative binomial\") : \n",
      "  'x' must be a non-empty numeric vector\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 4 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `max_counts = max(counts, na.rm = TRUE)`.\n",
      "Caused by warning in `max()`:\n",
      "\u001b[33m!\u001b[39m no non-missing arguments to max; returning -Inf\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 3 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL007049\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 14 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 12: `gene = RPL27A`.\n",
      "Caused by warning in `sqrt()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 13 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL015552\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 76 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 3: `gene = SPCS1`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 75 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL012075\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 374 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 3: `gene = GTF3C1`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 373 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL014173\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 12 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 14: `gene = CCL3`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 11 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL009083\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 22 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 12: `gene = PPIB`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 21 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL012457\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 72 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 1: `gene = GTF3C1`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 71 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL011807\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 34 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 4: `gene = SH2D2A`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 33 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL016716\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 28: `gene = CCL5`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL008702\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 102 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 7: `gene = SH2D2A`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 101 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL008389\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 10 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `bin_size = as.numeric(try(fitdistr(counts, \"negative\n",
      "  binomial\")$estimate[1]))`.\n",
      "\u001b[36mℹ\u001b[39m In group 64: `gene = CALM2`.\n",
      "Caused by warning in `densfun()`:\n",
      "\u001b[33m!\u001b[39m NaNs produced\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 9 remaining warnings.\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"LL012757\"\n",
      "[1] 1\n"
     ]
    }
   ],
   "source": [
    "for(i in 1:nrow(overview_data)){\n",
    "    path = paste0(data_path_sc, overview_data$name[i], '/input/L1/', overview_data$data[i])\n",
    "    data = get_dataset(path)   \n",
    "    data = data[genes]  # subsample on the selected genes\n",
    "    #print(data)\n",
    "    #print(unique(data$Assignment))\n",
    "    \n",
    "    ## Get overview on amount samples: sample_summary\n",
    "    sample_data = data[[]] %>% group_by(Assignment) %>% summarise(amount_cells = n())\n",
    "    sample_data$dataset = overview_data$name[i]\n",
    "    sample_data$cell_type = overview_data$data[i]\n",
    "    sample_summary = rbind(sample_summary, sample_data)  # TBD outside of loop!\n",
    "\n",
    "\n",
    "    for(k in unique(data$Assignment)) {     \n",
    "            \n",
    "        ## Get summary for each gene and sample: sample_gene_statistic\n",
    "        sample_data = data[,data$Assignment == k ]\n",
    "        #sample_data_trans = GetAssayData(sample_data ,assay = 'data')\n",
    "        sample_data = GetAssayData(sample_data ,assay =choose_assay)  # alternative: assay = 'data' --> returns transformed values\n",
    "\n",
    "            ### Filter genes with at leat 10 non-zero counts\n",
    "            sample_data = sample_data[rowSums(sample_data >= 1) >= 10,]\n",
    "\n",
    "            ### convert format\n",
    "            sample_data = as.matrix(sample_data)\n",
    "            #sample_data_trans = as.matrix(sample_data_trans)\n",
    "\n",
    "            data_long = melt(sample_data)\n",
    "            #data_long_trans = melt(sample_data_trans)\n",
    "\n",
    "            colnames(data_long) = c('gene', 'cell', 'counts')\n",
    "            #colnames(data_long_trans) = c('gene', 'cell', 'trans_counts')\n",
    "\n",
    "            ### get summaries\n",
    "            summary_data = data_long %>% group_by(gene) %>% summarise(\n",
    "                amount_cells = n(),\n",
    "                amount_zero_counts = sum(counts == 0),\n",
    "                mean_counts = mean(counts, na.rm = TRUE),\n",
    "                max_counts = max(counts, na.rm = TRUE),\n",
    "                min_counts = min(counts, na.rm = TRUE),\n",
    "                variance_counts = var(counts, na.rm = TRUE),\n",
    "                bin_size = as.numeric(try(fitdistr(counts, 'negative binomial')$estimate[1])), \n",
    "                bin_mu= as.numeric(try( fitdistr(counts, 'negative binomial')$estimate[2]))\n",
    "\n",
    "            )\n",
    "\n",
    "            summary_data$percentage_zero = summary_data$amount_zero_counts / summary_data$amount_cells\n",
    "            summary_data$amount_non_zero_counts = summary_data$amount_cells - summary_data$amount_zero_counts\n",
    "            summary_data$percentage_zero =  summary_data$percentage_zero*100\n",
    "            ## bin different columns\n",
    "            summary_data =  summary_data %>% mutate(amount_cells_bin = cut(amount_cells, breaks=c(0,25, 50, 100,250, 500, 750, 1500)))\n",
    "            summary_data =  summary_data %>% mutate(percentage_zero_bin = cut(percentage_zero, breaks=c(-0.001,5, 10, 20,40, 60, 80,90, 95, 100)))\n",
    "\n",
    "\n",
    "            summary_data$sample = k\n",
    "            summary_data$cell_type = overview_data$data[i]\n",
    "            summary_data$dataset = overview_data$name[i]\n",
    "\n",
    "            sample_gene_statistic = rbind(sample_gene_statistic, summary_data)\n",
    "        #\n",
    "        \n",
    "        # Get summaries for each sample and each cell: sample_cell_statistic\n",
    "        sample_data = data[,data$Assignment == k ]\n",
    "        sample_data = GetAssayData(sample_data ,assay =choose_assay)  # alternative: assay = 'data' --> returns transformed values\n",
    "\n",
    "            ### convert format\n",
    "            sample_data = as.matrix(sample_data)\n",
    "\n",
    "            data_long = melt(sample_data)\n",
    "\n",
    "            colnames(data_long) = c('gene', 'cell', 'counts')\n",
    "\n",
    "            ### get summaries\n",
    "            summary_data = data_long %>% group_by(cell) %>% summarise(\n",
    "                amount_counts = sum(counts),\n",
    "                amount_genes = n(),\n",
    "                amount_zero_counts = sum(counts == 0),\n",
    "                mean_counts = mean(counts,na.rm = TRUE),\n",
    "                max_counts = max(counts,na.rm = TRUE),\n",
    "                min_counts = min(counts,na.rm = TRUE),\n",
    "                variance_counts = var(counts,na.rm = TRUE))\n",
    "\n",
    "            summary_data$percentage_zero = summary_data$amount_zero_counts / summary_data$amount_genes\n",
    "            summary_data$amount_non_zero_counts = summary_data$amount_genes - summary_data$amount_zero_counts\n",
    "\n",
    "\n",
    "            summary_data$sample = k\n",
    "            summary_data$cell_type = overview_data$data[i]\n",
    "            summary_data$dataset = overview_data$name[i]\n",
    "\n",
    "            sample_cell_statistic = rbind(sample_cell_statistic, summary_data)\n",
    "        \n",
    "            #write.csv(sample_summary, paste0(output_path_sc, 'D1_Sample_summary', overview_data$name[i], overview_data$data[i], '_', choose_assay, '.csv'))\n",
    "            if(gene_selection == 'coeqtl'){\n",
    "            write.csv(sample_gene_statistic, paste0(output_path_sc, 'D1_Sample_gene_statistic_binom', overview_data$name[i], overview_data$data[i], '_', choose_assay, '.csv'))\n",
    "                }\n",
    "            if(gene_selection == 'random'){\n",
    "            write.csv(sample_gene_statistic, paste0(output_path_sc, 'D1_Sample_gene_statistic_binom', overview_data$name[i], overview_data$data[i], '_', choose_assay, '_', gene_selection, '.csv'))  \n",
    "                }\n",
    "            #write.csv(sample_cell_statistic, paste0(output_path_sc, 'D1_Sample_cell_statistic', overview_data$name[i], overview_data$data[i], '_', choose_assay, '.csv'))\n",
    "    \n",
    "                \n",
    "            flush.console()\n",
    "            print(k)\n",
    "            flush.console()\n",
    "\n",
    "        \n",
    "        \n",
    "        }\n",
    "        flush.console()\n",
    "        print(i)\n",
    "        flush.console()\n",
    "    \n",
    "    \n",
    "        sample_summary = data.frame()\n",
    "        sample_gene_statistic = data.frame()\n",
    "        sample_cell_statistic = data.frame()\n",
    "    }\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ce23277-230f-433f-8080-bc18a97137f3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ae3d176-97e9-419f-ab1f-bde9b147bef9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e93e923-fb0b-433c-abf5-adc4b563410b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc-autonumbering": true,
  "toc-showcode": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
