{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "35f293f4-68f4-48bc-bbc8-f4a57d881046",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ------------------------------------------------------------------------------\n",
    "# Calculate the MOTIF Overlaps with SNPs using MOTIFBreakR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "547ec508-6229-43ce-93ff-8474b865afb4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0fd55174-b556-48fc-85af-7a681b949b41",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "81659f25-92e4-4adc-8107-fe9deb906f92",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "“incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Loading required package: nlme\n",
      "\n",
      "This is mgcv 1.9-1. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "corrplot 0.95 loaded\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:genefilter’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "\n",
      "************\n",
      "Welcome to faux. For support and examples visit:\n",
      "https://debruine.github.io/faux/\n",
      "- Get and set global package options with: faux_options()\n",
      "************\n",
      "\n",
      "wCorr v1.9.8\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘sjmisc’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    is_empty\n",
      "\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:limma’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:nlme’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:AnnotationDbi’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:biomaRt’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:genefilter’:\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: ‘DT’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "This is coloc version 5.2.3\n",
      "\n",
      "\n",
      "\n",
      "Loading required package: gsubfn\n",
      "\n",
      "Loading required package: proto\n",
      "\n",
      "Warning message in fun(libname, pkgname):\n",
      "“couldn't connect to display \"localhost:12.0\"”\n",
      "Loading required package: RSQLite\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sjmisc’:\n",
      "\n",
      "    replace_na\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dbscan’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    as.dendrogram\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:rtracklayer’:\n",
      "\n",
      "    blocks, path\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:clusterProfiler’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    normalize, path, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    components\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: dynamicTreeCut\n",
      "\n",
      "Loading required package: fastcluster\n",
      "\n",
      "\n",
      "Attaching package: ‘fastcluster’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    hclust\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘WGCNA’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:qvalue’:\n",
      "\n",
      "    qvalue\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    geometry\n",
      "\n",
      "\n",
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:clusterProfiler':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "Loading required package: MotifDb\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: 'XVector'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    compact\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:grid':\n",
      "\n",
      "    pattern\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "See system.file(\"LICENSE\", package=\"MotifDb\") for use restrictions.\n",
      "\n",
      "Warning message:\n",
      "\"replacing previous import 'S4Vectors::as.data.frame' by 'motifStack::as.data.frame' when loading 'motifbreakR'\"\n",
      "\n",
      "Attaching package: 'Gviz'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    feature\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "\n",
      "Attaching package: 'MOFA2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    predict\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "59c74855-903c-42ea-8bc0-28c5b220b495",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(rtracklayer)\n",
    "library(data.table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "48ca47dd-002e-43ab-8643-83ae5013d733",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(BSgenome.Hsapiens.UCSC.hg38)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "76177215-52cc-44c3-8db6-658c4eae20f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'TFBSTools'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Gviz':\n",
      "\n",
      "    tags\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(TFBSTools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e2d906a8-8f79-4ef2-a2c4-96e7bf26a5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(JASPAR2020)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "692568a5-d98c-476e-bd78-1057526837b9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(Biostrings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7284bda3-94b2-41ab-8826-330d6e5784ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(GenomicRanges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7758a90c-3f87-4c64-a366-7b66f5f4a921",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(stringr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f49112f9-0f54-4e4e-bbd3-7f4a093d6284",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: 'httr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    content\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'magrittr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationFilter':\n",
      "\n",
      "    not\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "The following object is masked from 'package:purrr':\n",
      "\n",
      "    set_names\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source(\"snipe.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "74252c97-e2a5-4dc5-a31b-356997ce8087",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThe `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `linewidth` argument instead.\"\n"
     ]
    }
   ],
   "source": [
    "source('MS4_Plot_Config.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "537a01ba-341c-4cab-8924-23b28f3fe7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4110264f-94c8-4cc9-8e12-a7d5da462841",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] '2.8.0'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "packageVersion('motifbreakR')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b434aad-af45-40f8-bb8e-02931e4710dd",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2a72a804-a531-4632-b165-13feb3496aa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define which co-eQTL mapping results to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c803b08c-a55c-40e7-a081-cb4a97898559",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results_path = 'co_qtls_decision_tree/'   # path to results of full mapping\n",
    "coeqtl_results_path_old = 'co_qtls_sceqtlgen/'   # path to initial mapping (Evaluation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d4533e97-d51d-4ffe-aa6d-8e4ced13c2a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the co-EQTL Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "db0f7377-dcce-4f1f-9413-4d9b8df6d0e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data/current/coeqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3eeee44-8968-455f-acf3-6c20476f8bc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to the eQTL Files (sceQTLGen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e8191c1d-3933-4e7e-8139-0e073f258ca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl = '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9ec49116-f254-4d33-be20-5bd7eea5b764",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_results = 'sc_eqtlgen_final'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e740b6ae-7646-449b-ab11-06368f76ebcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to meta data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "04ff4b5f-7ab3-4fe5-9d7a-bb7663bc8dd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_meta = '../data/current/meta-data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "12e35a77-da69-42dc-9687-73524c936221",
   "metadata": {},
   "outputs": [],
   "source": [
    "### How to identify sign co-eqtls (parameters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "65ecafc7-9a3e-4a37-98d3-059125a33f57",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_val_column = 'p_val_mt_eGene' # p_value # nominal , #p_val_mt_eGene\n",
    "mt_method = 'eGene_bf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "12e25b48-c0a6-4ac5-9977-eb2ebcbbf48e",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_value_sign = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "535d5e27-0f4d-4100-b82a-85ef7cc1f43e",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_sample_filter = 133"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efb7b5db-46b5-4529-9936-e5662b822433",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e5b13d0-4bdf-4a1b-b792-7f906f99706f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f5b9d4cf-84af-4b9a-b1b1-c2206393cd10",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc99fb3-b762-483e-a7ad-2b881b0882eb",
   "metadata": {},
   "source": [
    "## Load significant coeqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c212a90b-3207-457d-b547-17a822b6c5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Annotated file from 00 Data Annotation script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "740082fe-717c-432e-8e39-7b9c72db5555",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "meta_results = fread(paste0( \"../data/current/coeqtl_mapping/\", coeqtl_results_path , \"/shared_results/\", '5DS_Meta_Analysis_Sign_coeQTLs', n_sample_filter, p_val_column  ,'.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "fc61d7fe-7119-47aa-8bcb-bae79871ed2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 161</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>eGene</th><th scope=col>coeGene</th><th scope=col>snp_id</th><th scope=col>triplet</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>feature_id</th><th scope=col>n_samples</th><th scope=col>p_val_mt</th><th scope=col>⋯</th><th scope=col>z_score;oneK1K;CD4_T</th><th scope=col>z_score;oneK1K;CD8_T</th><th scope=col>z_score;oneK1K;Mono</th><th scope=col>z_score;oneK1K;NK</th><th scope=col>z_score;wijst;B</th><th scope=col>z_score;wijst;CD4_T</th><th scope=col>z_score;wijst;CD8_T</th><th scope=col>z_score;wijst;Mono</th><th scope=col>z_score;wijst;NK</th><th scope=col>remap_overlap_tf_snp</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>A2M</td><td>BACH2</td><td>12:8971336:T:C</td><td>12:8971336:T:C;A2M_BACH2</td><td>CD4_T</td><td>meta_5ds</td><td>A2M_BACH2</td><td>977</td><td>0.004284312</td><td>⋯</td><td>-6.281251</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><td>2</td><td>A2M</td><td>BACH2</td><td>12:8992460:T:C</td><td>12:8992460:T:C;A2M_BACH2</td><td>CD4_T</td><td>meta_5ds</td><td>A2M_BACH2</td><td>977</td><td>0.026783812</td><td>⋯</td><td>-5.991429</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 161\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " V1 & eGene & coeGene & snp\\_id & triplet & cell\\_type & dataset & feature\\_id & n\\_samples & p\\_val\\_mt & ⋯ & z\\_score;oneK1K;CD4\\_T & z\\_score;oneK1K;CD8\\_T & z\\_score;oneK1K;Mono & z\\_score;oneK1K;NK & z\\_score;wijst;B & z\\_score;wijst;CD4\\_T & z\\_score;wijst;CD8\\_T & z\\_score;wijst;Mono & z\\_score;wijst;NK & remap\\_overlap\\_tf\\_snp\\\\\n",
       " <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & A2M & BACH2 & 12:8971336:T:C & 12:8971336:T:C;A2M\\_BACH2 & CD4\\_T & meta\\_5ds & A2M\\_BACH2 & 977 & 0.004284312 & ⋯ & -6.281251 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\t 2 & A2M & BACH2 & 12:8992460:T:C & 12:8992460:T:C;A2M\\_BACH2 & CD4\\_T & meta\\_5ds & A2M\\_BACH2 & 977 & 0.026783812 & ⋯ & -5.991429 & NA & NA & NA & NA & NA & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 161\n",
       "\n",
       "| V1 &lt;int&gt; | eGene &lt;chr&gt; | coeGene &lt;chr&gt; | snp_id &lt;chr&gt; | triplet &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | feature_id &lt;chr&gt; | n_samples &lt;int&gt; | p_val_mt &lt;dbl&gt; | ⋯ ⋯ | z_score;oneK1K;CD4_T &lt;dbl&gt; | z_score;oneK1K;CD8_T &lt;dbl&gt; | z_score;oneK1K;Mono &lt;dbl&gt; | z_score;oneK1K;NK &lt;dbl&gt; | z_score;wijst;B &lt;dbl&gt; | z_score;wijst;CD4_T &lt;dbl&gt; | z_score;wijst;CD8_T &lt;dbl&gt; | z_score;wijst;Mono &lt;dbl&gt; | z_score;wijst;NK &lt;dbl&gt; | remap_overlap_tf_snp &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A2M | BACH2 | 12:8971336:T:C | 12:8971336:T:C;A2M_BACH2 | CD4_T | meta_5ds | A2M_BACH2 | 977 | 0.004284312 | ⋯ | -6.281251 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "| 2 | A2M | BACH2 | 12:8992460:T:C | 12:8992460:T:C;A2M_BACH2 | CD4_T | meta_5ds | A2M_BACH2 | 977 | 0.026783812 | ⋯ | -5.991429 | NA | NA | NA | NA | NA | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 eGene coeGene snp_id         triplet                  cell_type dataset \n",
       "1 1  A2M   BACH2   12:8971336:T:C 12:8971336:T:C;A2M_BACH2 CD4_T     meta_5ds\n",
       "2 2  A2M   BACH2   12:8992460:T:C 12:8992460:T:C;A2M_BACH2 CD4_T     meta_5ds\n",
       "  feature_id n_samples p_val_mt    ⋯ z_score;oneK1K;CD4_T z_score;oneK1K;CD8_T\n",
       "1 A2M_BACH2  977       0.004284312 ⋯ -6.281251            NA                  \n",
       "2 A2M_BACH2  977       0.026783812 ⋯ -5.991429            NA                  \n",
       "  z_score;oneK1K;Mono z_score;oneK1K;NK z_score;wijst;B z_score;wijst;CD4_T\n",
       "1 NA                  NA                NA              NA                 \n",
       "2 NA                  NA                NA              NA                 \n",
       "  z_score;wijst;CD8_T z_score;wijst;Mono z_score;wijst;NK remap_overlap_tf_snp\n",
       "1 NA                  NA                 NA               NA                  \n",
       "2 NA                  NA                 NA               NA                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(meta_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb35a1e1-4c1a-4309-939b-65f599c12116",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine non-ribosomal/ non-mitochondrial columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "e9a43a4b-f6ec-41c6-aaf0-e80dad63a442",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_results$n_rb_mt = 0\n",
    "meta_results$n_rb_mt[(meta_results$eGene_rib ==0) & (meta_results$coeGene_rib == 0) & (meta_results$eGene_mt == 0) & (meta_results$coeGene_mt == 0)] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "80048540-e4df-4b51-9915-e39f95801b08",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_results$snp_pos = str_replace(meta_results$snp_id ,\".*:(?=\\\\d)\", \"\")\n",
    "meta_results$snp_pos = str_replace(meta_results$snp_pos , \":.*\", \"\")\n",
    "meta_results$snp_chr = str_replace(meta_results$snp_id , ':.*', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "6d85f549-5901-4fad-a6af-519e7d88d703",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 164</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>eGene</th><th scope=col>coeGene</th><th scope=col>snp_id</th><th scope=col>triplet</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>feature_id</th><th scope=col>n_samples</th><th scope=col>p_val_mt</th><th scope=col>⋯</th><th scope=col>z_score;oneK1K;NK</th><th scope=col>z_score;wijst;B</th><th scope=col>z_score;wijst;CD4_T</th><th scope=col>z_score;wijst;CD8_T</th><th scope=col>z_score;wijst;Mono</th><th scope=col>z_score;wijst;NK</th><th scope=col>remap_overlap_tf_snp</th><th scope=col>n_rb_mt</th><th scope=col>snp_pos</th><th scope=col>snp_chr</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>A2M</td><td>BACH2</td><td>12:8971336:T:C</td><td>12:8971336:T:C;A2M_BACH2</td><td>CD4_T</td><td>meta_5ds</td><td>A2M_BACH2</td><td>977</td><td>0.004284312</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>1</td><td>8971336</td><td>12</td></tr>\n",
       "\t<tr><td>2</td><td>A2M</td><td>BACH2</td><td>12:8992460:T:C</td><td>12:8992460:T:C;A2M_BACH2</td><td>CD4_T</td><td>meta_5ds</td><td>A2M_BACH2</td><td>977</td><td>0.026783812</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>1</td><td>8992460</td><td>12</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 164\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " V1 & eGene & coeGene & snp\\_id & triplet & cell\\_type & dataset & feature\\_id & n\\_samples & p\\_val\\_mt & ⋯ & z\\_score;oneK1K;NK & z\\_score;wijst;B & z\\_score;wijst;CD4\\_T & z\\_score;wijst;CD8\\_T & z\\_score;wijst;Mono & z\\_score;wijst;NK & remap\\_overlap\\_tf\\_snp & n\\_rb\\_mt & snp\\_pos & snp\\_chr\\\\\n",
       " <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & A2M & BACH2 & 12:8971336:T:C & 12:8971336:T:C;A2M\\_BACH2 & CD4\\_T & meta\\_5ds & A2M\\_BACH2 & 977 & 0.004284312 & ⋯ & NA & NA & NA & NA & NA & NA & NA & 1 & 8971336 & 12\\\\\n",
       "\t 2 & A2M & BACH2 & 12:8992460:T:C & 12:8992460:T:C;A2M\\_BACH2 & CD4\\_T & meta\\_5ds & A2M\\_BACH2 & 977 & 0.026783812 & ⋯ & NA & NA & NA & NA & NA & NA & NA & 1 & 8992460 & 12\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 164\n",
       "\n",
       "| V1 &lt;int&gt; | eGene &lt;chr&gt; | coeGene &lt;chr&gt; | snp_id &lt;chr&gt; | triplet &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | feature_id &lt;chr&gt; | n_samples &lt;int&gt; | p_val_mt &lt;dbl&gt; | ⋯ ⋯ | z_score;oneK1K;NK &lt;dbl&gt; | z_score;wijst;B &lt;dbl&gt; | z_score;wijst;CD4_T &lt;dbl&gt; | z_score;wijst;CD8_T &lt;dbl&gt; | z_score;wijst;Mono &lt;dbl&gt; | z_score;wijst;NK &lt;dbl&gt; | remap_overlap_tf_snp &lt;chr&gt; | n_rb_mt &lt;dbl&gt; | snp_pos &lt;chr&gt; | snp_chr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A2M | BACH2 | 12:8971336:T:C | 12:8971336:T:C;A2M_BACH2 | CD4_T | meta_5ds | A2M_BACH2 | 977 | 0.004284312 | ⋯ | NA | NA | NA | NA | NA | NA | NA | 1 | 8971336 | 12 |\n",
       "| 2 | A2M | BACH2 | 12:8992460:T:C | 12:8992460:T:C;A2M_BACH2 | CD4_T | meta_5ds | A2M_BACH2 | 977 | 0.026783812 | ⋯ | NA | NA | NA | NA | NA | NA | NA | 1 | 8992460 | 12 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 eGene coeGene snp_id         triplet                  cell_type dataset \n",
       "1 1  A2M   BACH2   12:8971336:T:C 12:8971336:T:C;A2M_BACH2 CD4_T     meta_5ds\n",
       "2 2  A2M   BACH2   12:8992460:T:C 12:8992460:T:C;A2M_BACH2 CD4_T     meta_5ds\n",
       "  feature_id n_samples p_val_mt    ⋯ z_score;oneK1K;NK z_score;wijst;B\n",
       "1 A2M_BACH2  977       0.004284312 ⋯ NA                NA             \n",
       "2 A2M_BACH2  977       0.026783812 ⋯ NA                NA             \n",
       "  z_score;wijst;CD4_T z_score;wijst;CD8_T z_score;wijst;Mono z_score;wijst;NK\n",
       "1 NA                  NA                  NA                 NA              \n",
       "2 NA                  NA                  NA                 NA              \n",
       "  remap_overlap_tf_snp n_rb_mt snp_pos snp_chr\n",
       "1 NA                   1       8971336 12     \n",
       "2 NA                   1       8992460 12     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(meta_results,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3985a5b-7458-4412-b3cc-051cf8033f41",
   "metadata": {},
   "source": [
    "## Load Finemapped SNPs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bff4af8a-941d-4b72-b44f-7fd1c6691a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get SNPs from co-eQTL Finemapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f51776e-1baa-46ab-8059-073eaadf6a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define dataset (meta-analysis) and cell-types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a2079c71-6f36-43b5-9492-e6e70cf98cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_alt = 'meta_analysis/meta_5ds'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "ef5aa40b-c783-431d-8921-05d4840d1847",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = c('Mono', 'CD8_T', 'NK', 'B', 'CD4_T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65468822-91b7-4c34-95f3-b42646bbeebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define which finemapping version to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c0e74f-157d-4ee0-a4a9-dac3a5134ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapping_version = 'lenient' # lenient or strict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d64cfacc-1d3a-4460-80d2-6315fe946133",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Empty dataframe to save all finemapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "321233d2-7d41-4a4b-948e-a4520851c896",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapping_results = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d78b6769-0f50-45a0-bc14-51b058d53b54",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load finemapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "07cae086-3964-458d-b6fb-36f98cae8f4e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"meta_joint_susie_finemap_B_1_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_11_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_12_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_14_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_15_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_17_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_18_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_19_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_2_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_20_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_22_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_3_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_4_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_6_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_7_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_8_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_B_9_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_1_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:30 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_10_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_11_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_12_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:33 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_13_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_14_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_15_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_16_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_17_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_18_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_19_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_2_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:32 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_20_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_21_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:22 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_22_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:22 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_3_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:33 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_4_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:33 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_5_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:31 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_6_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:35 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_7_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_8_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_9_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_1_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_10_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_11_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_12_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_13_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_14_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_15_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_16_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_17_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_18_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_2_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:32 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_20_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_21_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_22_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_3_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_4_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_5_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_6_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:30 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_7_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_8_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_9_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_1_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_10_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:22 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_11_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_12_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_13_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_16_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_17_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_18_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_19_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:26 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_2_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_20_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_21_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:24 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_22_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_3_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_5_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_6_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:25 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_7_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:22 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_8_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:23 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_9_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:22 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_1_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_10_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_11_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:30 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_12_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:30 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_13_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_14_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_16_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_17_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_18_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_19_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:27 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_2_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:30 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_22_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_3_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_4_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_6_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:31 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_7_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_8_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:29 CEST\"\n",
      "[1] \"meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz\"\n",
      "[1] \"2025-06-04 21:13:28 CEST\"\n"
     ]
    }
   ],
   "source": [
    "if(finemapping_version == 'lenient'){\n",
    "    files = list.files(paste0(data_path,coeqtl_results_path ,dataset_alt, '/Finemaping_lenient/NT/'))\n",
    "    files = files[!is.na(str_extract(files, 'newThreshold'))]\n",
    "\n",
    "    coeqtl_all_files = files\n",
    "\n",
    "\n",
    "    coeqtl_all_effect = data.frame()\n",
    "\n",
    "    for(i in coeqtl_all_files){\n",
    "        print(i)\n",
    "        flush.console()\n",
    "        file = paste0(data_path, coeqtl_results_path ,dataset_alt, '/', '/Finemaping_lenient/NT/', i)\n",
    "        print(file.info(file)$ctime)\n",
    "        data1 = fread(file)\n",
    "        data1 = as.data.frame(data1)\n",
    "        cell_type_var = str_extract(i, 'B|NK|Mono|CD4_T|CD8_T')\n",
    "        data1$cell_type = cell_type_var\n",
    "        data1$file_name = i\n",
    "        data1$SusieRss_CS_G = NULL\n",
    "        #data1$`n_samples;Multiome` = NULL\n",
    "        #data1$`n_e_samples;Multiome` = NULL\n",
    "        #data1$`z_score;Multiome` = NULL\n",
    "        coeqtl_all_effect = rbind(data1, coeqtl_all_effect)\n",
    "    }\n",
    "\n",
    "    finemapping_results = rbind(finemapping_results, coeqtl_all_effect)\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b0db270d-cee8-4da8-99e1-2d9279ecee87",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"meta_joint_susie_finemap_Mono_1_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_10_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:06 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_11_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:07 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_12_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:07 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_13_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:08 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_15_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:08 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_16_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:08 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_17_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:08 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_18_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:08 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_19_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_2_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_20_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_21_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_22_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_3_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:09 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_5_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:10 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_6_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:12 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_7_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:12 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_8_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:12 CET\"\n",
      "[1] \"meta_joint_susie_finemap_Mono_9_formatted.txt\"\n",
      "[1] \"2025-02-26 15:59:12 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_1_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_10_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:55 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_11_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:55 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_12_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:58 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_13_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:58 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_14_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_15_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_16_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_17_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_18_formatted.txt\"\n",
      "[1] \"2025-02-10 10:10:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_2_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:02 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_20_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_21_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_3_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:02 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_4_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:03 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_5_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:03 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_6_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:05 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_7_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:05 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_8_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:05 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD8_T_9_formatted.txt\"\n",
      "[1] \"2025-02-10 10:11:05 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_1_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_10_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:58 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_11_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:58 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_12_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_13_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_14_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_16_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_17_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_18_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_19_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_2_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_22_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_3_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_4_formatted.txt\"\n",
      "[1] \"2025-02-06 13:05:59 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_6_formatted.txt\"\n",
      "[1] \"2025-02-06 13:06:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_7_formatted.txt\"\n",
      "[1] \"2025-02-06 13:06:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_8_formatted.txt\"\n",
      "[1] \"2025-02-06 13:06:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_NK_9_formatted.txt\"\n",
      "[1] \"2025-02-06 13:06:00 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_1_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_11_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:38 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_12_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_14_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_15_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_17_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_18_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_19_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_2_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_20_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_22_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_3_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:39 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_4_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:40 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_6_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:41 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_7_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:41 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_8_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:41 CET\"\n",
      "[1] \"meta_joint_susie_finemap_B_9_formatted.txt\"\n",
      "[1] \"2025-02-26 16:21:41 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_1_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:49 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_10_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:40 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_11_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:41 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_12_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:44 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_13_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:45 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_14_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:46 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_15_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:46 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_16_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:46 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_17_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:48 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_18_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:48 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_19_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:48 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_2_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:52 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_20_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:49 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_21_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:50 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_22_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:50 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_3_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:52 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_4_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:54 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_5_formatted.txt\"\n",
      "[1] \"2025-02-06 13:35:55 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_6_formatted.txt\"\n",
      "[1] \"2025-02-06 13:36:05 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_7_formatted.txt\"\n",
      "[1] \"2025-02-06 13:36:06 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_8_formatted.txt\"\n",
      "[1] \"2025-02-06 13:36:06 CET\"\n",
      "[1] \"meta_joint_susie_finemap_CD4_T_9_formatted.txt\"\n",
      "[1] \"2025-02-06 13:36:07 CET\"\n"
     ]
    }
   ],
   "source": [
    "#### Strict grouping version\n",
    "if(finemapping_version == 'strict'){\n",
    "    for(j in cell_types){\n",
    "        cell_type_var = j\n",
    "        coeqtl_all_files = list.files(paste0(data_path, coeqtl_results_path ,dataset_alt, '/',  cell_type_var , '/Full_Window_Results/'))\n",
    "        coeqtl_all_files = coeqtl_all_files[!is.na(str_extract(coeqtl_all_files, cell_type_var))]\n",
    "        coeqtl_all_files = coeqtl_all_files[!is.na(str_extract(coeqtl_all_files, '.txt'))]\n",
    "\n",
    "\n",
    "        coeqtl_all_effect = data.frame()\n",
    "\n",
    "        for(i in coeqtl_all_files){\n",
    "            print(i)\n",
    "            flush.console()\n",
    "            file = paste0(data_path, coeqtl_results_path ,dataset_alt, '/',  cell_type_var , '/Full_Window_Results/', i)\n",
    "            print(file.info(file)$ctime)\n",
    "            data1 = fread(file)\n",
    "            data1 = as.data.frame(data1)\n",
    "            data1$cell_type = cell_type_var\n",
    "            data1$file_name = i\n",
    "            data1$SusieRss_CS_G = NULL\n",
    "            #data1$`n_samples;Multiome` = NULL\n",
    "            #data1$`n_e_samples;Multiome` = NULL\n",
    "            #data1$`z_score;Multiome` = NULL\n",
    "            coeqtl_all_effect = rbind(data1, coeqtl_all_effect)\n",
    "        }\n",
    "\n",
    "        finemapping_results = rbind(finemapping_results, coeqtl_all_effect)\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "a4a1858e-f78a-4391-a2f8-c746fd2c02bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 45</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coQTL</th><th scope=col>eGene</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>coeGene</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>snp_chromosome</th><th scope=col>snp_position</th><th scope=col>⋯</th><th scope=col>SusieRss_pip</th><th scope=col>SusieRss_CS</th><th scope=col>SusieRss_ResVar</th><th scope=col>lbf_cs_1</th><th scope=col>lbf_cs_2</th><th scope=col>lbf_cs_3</th><th scope=col>lbf_cs_4</th><th scope=col>lbf_cs_5</th><th scope=col>cell_type</th><th scope=col>file_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>9:135985420:T:C;PTGDS_SELL</td><td>PTGDS</td><td>9</td><td>136975092</td><td>136981742</td><td>SELL</td><td>PTGDS_SELL</td><td>9:135985420:T:C</td><td>9</td><td>135985420</td><td>⋯</td><td>0.007615623</td><td>NA</td><td>TRUE</td><td>-1.004602</td><td>-0.3033094</td><td>-0.04461793</td><td>-0.04147275</td><td>-0.03254301</td><td>NK</td><td>meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>9:135986302:G:C;PTGDS_SELL</td><td>PTGDS</td><td>9</td><td>136975092</td><td>136981742</td><td>SELL</td><td>PTGDS_SELL</td><td>9:135986302:G:C</td><td>9</td><td>135986302</td><td>⋯</td><td>0.007823508</td><td>NA</td><td>TRUE</td><td>-0.957815</td><td>-0.2567398</td><td>-0.02082910</td><td>-0.01837484</td><td>-0.01163311</td><td>NK</td><td>meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 45\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & coQTL & eGene & feature\\_chromosome & feature\\_start & feature\\_end & coeGene & feature\\_id & snp\\_id & snp\\_chromosome & snp\\_position & ⋯ & SusieRss\\_pip & SusieRss\\_CS & SusieRss\\_ResVar & lbf\\_cs\\_1 & lbf\\_cs\\_2 & lbf\\_cs\\_3 & lbf\\_cs\\_4 & lbf\\_cs\\_5 & cell\\_type & file\\_name\\\\\n",
       "  & <chr> & <chr> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <dbl> & ⋯ & <dbl> & <int> & <lgl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 9:135985420:T:C;PTGDS\\_SELL & PTGDS & 9 & 136975092 & 136981742 & SELL & PTGDS\\_SELL & 9:135985420:T:C & 9 & 135985420 & ⋯ & 0.007615623 & NA & TRUE & -1.004602 & -0.3033094 & -0.04461793 & -0.04147275 & -0.03254301 & NK & meta\\_joint\\_susie\\_finemap\\_NK\\_9\\_newThreshold\\_formatted.txt.gz\\\\\n",
       "\t2 & 9:135986302:G:C;PTGDS\\_SELL & PTGDS & 9 & 136975092 & 136981742 & SELL & PTGDS\\_SELL & 9:135986302:G:C & 9 & 135986302 & ⋯ & 0.007823508 & NA & TRUE & -0.957815 & -0.2567398 & -0.02082910 & -0.01837484 & -0.01163311 & NK & meta\\_joint\\_susie\\_finemap\\_NK\\_9\\_newThreshold\\_formatted.txt.gz\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 45\n",
       "\n",
       "| <!--/--> | coQTL &lt;chr&gt; | eGene &lt;chr&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | coeGene &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | snp_chromosome &lt;int&gt; | snp_position &lt;dbl&gt; | ⋯ ⋯ | SusieRss_pip &lt;dbl&gt; | SusieRss_CS &lt;int&gt; | SusieRss_ResVar &lt;lgl&gt; | lbf_cs_1 &lt;dbl&gt; | lbf_cs_2 &lt;dbl&gt; | lbf_cs_3 &lt;dbl&gt; | lbf_cs_4 &lt;dbl&gt; | lbf_cs_5 &lt;dbl&gt; | cell_type &lt;chr&gt; | file_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 9:135985420:T:C;PTGDS_SELL | PTGDS | 9 | 136975092 | 136981742 | SELL | PTGDS_SELL | 9:135985420:T:C | 9 | 135985420 | ⋯ | 0.007615623 | NA | TRUE | -1.004602 | -0.3033094 | -0.04461793 | -0.04147275 | -0.03254301 | NK | meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz |\n",
       "| 2 | 9:135986302:G:C;PTGDS_SELL | PTGDS | 9 | 136975092 | 136981742 | SELL | PTGDS_SELL | 9:135986302:G:C | 9 | 135986302 | ⋯ | 0.007823508 | NA | TRUE | -0.957815 | -0.2567398 | -0.02082910 | -0.01837484 | -0.01163311 | NK | meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz |\n",
       "\n"
      ],
      "text/plain": [
       "  coQTL                      eGene feature_chromosome feature_start feature_end\n",
       "1 9:135985420:T:C;PTGDS_SELL PTGDS 9                  136975092     136981742  \n",
       "2 9:135986302:G:C;PTGDS_SELL PTGDS 9                  136975092     136981742  \n",
       "  coeGene feature_id snp_id          snp_chromosome snp_position ⋯ SusieRss_pip\n",
       "1 SELL    PTGDS_SELL 9:135985420:T:C 9              135985420    ⋯ 0.007615623 \n",
       "2 SELL    PTGDS_SELL 9:135986302:G:C 9              135986302    ⋯ 0.007823508 \n",
       "  SusieRss_CS SusieRss_ResVar lbf_cs_1  lbf_cs_2   lbf_cs_3    lbf_cs_4   \n",
       "1 NA          TRUE            -1.004602 -0.3033094 -0.04461793 -0.04147275\n",
       "2 NA          TRUE            -0.957815 -0.2567398 -0.02082910 -0.01837484\n",
       "  lbf_cs_5    cell_type\n",
       "1 -0.03254301 NK       \n",
       "2 -0.01163311 NK       \n",
       "  file_name                                                  \n",
       "1 meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz\n",
       "2 meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(finemapping_results,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f40345-2ffe-409c-b8f4-aa6e936d2a64",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "85e26bcf-b29b-457a-a319-c1cef4c7e27b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'NK'</li><li>'Mono'</li><li>'CD8_T'</li><li>'CD4_T'</li><li>'B'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'NK'\n",
       "\\item 'Mono'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'CD4\\_T'\n",
       "\\item 'B'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'NK'\n",
       "2. 'Mono'\n",
       "3. 'CD8_T'\n",
       "4. 'CD4_T'\n",
       "5. 'B'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"NK\"    \"Mono\"  \"CD8_T\" \"CD4_T\" \"B\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(finemapping_results$cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a728efbe-ad22-4f1d-95da-be2e2732bb42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>&lt;NA&gt;</li><li>1</li><li>2</li><li>3</li><li>5</li><li>4</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item <NA>\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 5\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. &lt;NA&gt;\n",
       "2. 1\n",
       "3. 2\n",
       "4. 3\n",
       "5. 5\n",
       "6. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] NA  1  2  3  5  4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(finemapping_results$SusieRss_CS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e0f4782-3364-4908-a235-e88c2e70f3f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the SNPs from the finemapping within a Credible Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "8663d968-e3ed-4444-8b87-d8a9c2223c60",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapping_snps = finemapping_results[!is.na(finemapping_results$SusieRss_CS),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "9a084320-bae0-4f21-b887-73b97a481014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>3</li><li>5</li><li>4</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 5\n",
       "\\item 4\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 3\n",
       "4. 5\n",
       "5. 4\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1 2 3 5 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(finemapping_snps$SusieRss_CS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "26b97c25-06ec-4667-85b4-fde92c9599df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "9717"
      ],
      "text/latex": [
       "9717"
      ],
      "text/markdown": [
       "9717"
      ],
      "text/plain": [
       "[1] 9717"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(finemapping_snps$snp_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a9ad0ba7-5905-41e2-b2a9-7971b8a11d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapping_snps$snp_pos = str_replace(finemapping_snps$snp_id ,\".*:(?=\\\\d)\", \"\")\n",
    "finemapping_snps$snp_pos = str_replace(finemapping_snps$snp_pos , \":.*\", \"\")\n",
    "finemapping_snps$snp_chr = str_replace(finemapping_snps$snp_id , ':.*', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "8766a6cb-cbe2-4549-a923-2ff6e12292f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 47</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>coQTL</th><th scope=col>eGene</th><th scope=col>feature_chromosome</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>coeGene</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>snp_chromosome</th><th scope=col>snp_position</th><th scope=col>⋯</th><th scope=col>SusieRss_ResVar</th><th scope=col>lbf_cs_1</th><th scope=col>lbf_cs_2</th><th scope=col>lbf_cs_3</th><th scope=col>lbf_cs_4</th><th scope=col>lbf_cs_5</th><th scope=col>cell_type</th><th scope=col>file_name</th><th scope=col>snp_pos</th><th scope=col>snp_chr</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>282</th><td>9:136936281:G:C;PTGDS_SELL</td><td>PTGDS</td><td>9</td><td>136975092</td><td>136981742</td><td>SELL</td><td>PTGDS_SELL</td><td>9:136936281:G:C</td><td>9</td><td>136936281</td><td>⋯</td><td>TRUE</td><td>14.36794</td><td>-0.6699152</td><td>-0.2333248</td><td>-0.2248034</td><td>-0.1989274</td><td>NK</td><td>meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz</td><td>136936281</td><td>9</td></tr>\n",
       "\t<tr><th scope=row>283</th><td>9:136936488:T:C;PTGDS_SELL</td><td>PTGDS</td><td>9</td><td>136975092</td><td>136981742</td><td>SELL</td><td>PTGDS_SELL</td><td>9:136936488:T:C</td><td>9</td><td>136936488</td><td>⋯</td><td>TRUE</td><td>14.07260</td><td>-0.6736336</td><td>-0.2362273</td><td>-0.2276421</td><td>-0.2015498</td><td>NK</td><td>meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz</td><td>136936488</td><td>9</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 47\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & coQTL & eGene & feature\\_chromosome & feature\\_start & feature\\_end & coeGene & feature\\_id & snp\\_id & snp\\_chromosome & snp\\_position & ⋯ & SusieRss\\_ResVar & lbf\\_cs\\_1 & lbf\\_cs\\_2 & lbf\\_cs\\_3 & lbf\\_cs\\_4 & lbf\\_cs\\_5 & cell\\_type & file\\_name & snp\\_pos & snp\\_chr\\\\\n",
       "  & <chr> & <chr> & <int> & <int> & <int> & <chr> & <chr> & <chr> & <int> & <dbl> & ⋯ & <lgl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t282 & 9:136936281:G:C;PTGDS\\_SELL & PTGDS & 9 & 136975092 & 136981742 & SELL & PTGDS\\_SELL & 9:136936281:G:C & 9 & 136936281 & ⋯ & TRUE & 14.36794 & -0.6699152 & -0.2333248 & -0.2248034 & -0.1989274 & NK & meta\\_joint\\_susie\\_finemap\\_NK\\_9\\_newThreshold\\_formatted.txt.gz & 136936281 & 9\\\\\n",
       "\t283 & 9:136936488:T:C;PTGDS\\_SELL & PTGDS & 9 & 136975092 & 136981742 & SELL & PTGDS\\_SELL & 9:136936488:T:C & 9 & 136936488 & ⋯ & TRUE & 14.07260 & -0.6736336 & -0.2362273 & -0.2276421 & -0.2015498 & NK & meta\\_joint\\_susie\\_finemap\\_NK\\_9\\_newThreshold\\_formatted.txt.gz & 136936488 & 9\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 47\n",
       "\n",
       "| <!--/--> | coQTL &lt;chr&gt; | eGene &lt;chr&gt; | feature_chromosome &lt;int&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | coeGene &lt;chr&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | snp_chromosome &lt;int&gt; | snp_position &lt;dbl&gt; | ⋯ ⋯ | SusieRss_ResVar &lt;lgl&gt; | lbf_cs_1 &lt;dbl&gt; | lbf_cs_2 &lt;dbl&gt; | lbf_cs_3 &lt;dbl&gt; | lbf_cs_4 &lt;dbl&gt; | lbf_cs_5 &lt;dbl&gt; | cell_type &lt;chr&gt; | file_name &lt;chr&gt; | snp_pos &lt;chr&gt; | snp_chr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 282 | 9:136936281:G:C;PTGDS_SELL | PTGDS | 9 | 136975092 | 136981742 | SELL | PTGDS_SELL | 9:136936281:G:C | 9 | 136936281 | ⋯ | TRUE | 14.36794 | -0.6699152 | -0.2333248 | -0.2248034 | -0.1989274 | NK | meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz | 136936281 | 9 |\n",
       "| 283 | 9:136936488:T:C;PTGDS_SELL | PTGDS | 9 | 136975092 | 136981742 | SELL | PTGDS_SELL | 9:136936488:T:C | 9 | 136936488 | ⋯ | TRUE | 14.07260 | -0.6736336 | -0.2362273 | -0.2276421 | -0.2015498 | NK | meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz | 136936488 | 9 |\n",
       "\n"
      ],
      "text/plain": [
       "    coQTL                      eGene feature_chromosome feature_start\n",
       "282 9:136936281:G:C;PTGDS_SELL PTGDS 9                  136975092    \n",
       "283 9:136936488:T:C;PTGDS_SELL PTGDS 9                  136975092    \n",
       "    feature_end coeGene feature_id snp_id          snp_chromosome snp_position\n",
       "282 136981742   SELL    PTGDS_SELL 9:136936281:G:C 9              136936281   \n",
       "283 136981742   SELL    PTGDS_SELL 9:136936488:T:C 9              136936488   \n",
       "    ⋯ SusieRss_ResVar lbf_cs_1 lbf_cs_2   lbf_cs_3   lbf_cs_4   lbf_cs_5  \n",
       "282 ⋯ TRUE            14.36794 -0.6699152 -0.2333248 -0.2248034 -0.1989274\n",
       "283 ⋯ TRUE            14.07260 -0.6736336 -0.2362273 -0.2276421 -0.2015498\n",
       "    cell_type file_name                                                  \n",
       "282 NK        meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz\n",
       "283 NK        meta_joint_susie_finemap_NK_9_newThreshold_formatted.txt.gz\n",
       "    snp_pos   snp_chr\n",
       "282 136936281 9      \n",
       "283 136936488 9      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(finemapping_snps,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "953cddde-126e-42da-9b9e-fcc208bd45f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapping_snps = unique(finemapping_snps[,c('snp_id', 'snp_pos', 'snp_chr')])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9274727c-9c36-4d36-9130-8cfed7655e8c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get SNP - eGene combinations used for testing (sceQTLGen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cfae6ed-498c-4167-872c-b0e45b2af355",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get eGene- finemapped SNPs from sceQTLGene used as input for co-eQTL mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "f4407edf-2a3a-474c-8a05-bed94fd73393",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = c('CD4_T', 'Mono', 'CD8_T', 'NK', 'B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "a78f5df5-1935-4c6a-a9b8-90ede275b3c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "egenes = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "4cc67f87-cd23-4e83-9ab0-ffafe067d527",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"CD4_T\"\n",
      "[1] \"Mono\"\n",
      "[1] \"CD8_T\"\n",
      "[1] \"NK\"\n",
      "[1] \"B\"\n"
     ]
    }
   ],
   "source": [
    "for(i in cell_types){\n",
    "    print(i)\n",
    "    snp_gene_mapping = read.csv(paste0(data_path_eqtl, eqtl_results , '/' , i, '_snp_egene_mapping.csv'))\n",
    "    snp_gene_mapping$cell_type = i\n",
    "    egenes = rbind(egenes, snp_gene_mapping)\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "990a09c4-f8b8-43c7-8e2a-8028a482f03b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add the SNP and chr positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "8044884d-e909-4cd4-98dc-2f127ef0f4a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "egenes$snp_pos = str_replace(egenes$snp_id ,\".*:(?=\\\\d)\", \"\")\n",
    "egenes$snp_pos = str_replace(egenes$snp_pos , \":.*\", \"\")\n",
    "egenes$snp_chr = str_replace(egenes$snp_id , ':.*', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "05a0a1c0-bf54-4051-b9be-972c6db19409",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>cell_type</th><th scope=col>snp_pos</th><th scope=col>snp_chr</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>GSTM3</td><td>1:109740350:G:T</td><td>109733932</td><td>109741038</td><td>5.311981e-277</td><td>-1.1703583</td><td>CD4_T</td><td>109740350</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>GSTM3</td><td>1:109731286:G:T</td><td>109733932</td><td>109741038</td><td> 1.872347e-74</td><td>-0.7292481</td><td>CD4_T</td><td>109731286</td><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & X & feature\\_id & snp\\_id & feature\\_start & feature\\_end & p\\_value & beta & cell\\_type & snp\\_pos & snp\\_chr\\\\\n",
       "  & <int> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 0 & GSTM3 & 1:109740350:G:T & 109733932 & 109741038 & 5.311981e-277 & -1.1703583 & CD4\\_T & 109740350 & 1\\\\\n",
       "\t2 & 1 & GSTM3 & 1:109731286:G:T & 109733932 & 109741038 &  1.872347e-74 & -0.7292481 & CD4\\_T & 109731286 & 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 10\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | p_value &lt;dbl&gt; | beta &lt;dbl&gt; | cell_type &lt;chr&gt; | snp_pos &lt;chr&gt; | snp_chr &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 0 | GSTM3 | 1:109740350:G:T | 109733932 | 109741038 | 5.311981e-277 | -1.1703583 | CD4_T | 109740350 | 1 |\n",
       "| 2 | 1 | GSTM3 | 1:109731286:G:T | 109733932 | 109741038 |  1.872347e-74 | -0.7292481 | CD4_T | 109731286 | 1 |\n",
       "\n"
      ],
      "text/plain": [
       "  X feature_id snp_id          feature_start feature_end p_value      \n",
       "1 0 GSTM3      1:109740350:G:T 109733932     109741038   5.311981e-277\n",
       "2 1 GSTM3      1:109731286:G:T 109733932     109741038    1.872347e-74\n",
       "  beta       cell_type snp_pos   snp_chr\n",
       "1 -1.1703583 CD4_T     109740350 1      \n",
       "2 -0.7292481 CD4_T     109731286 1      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(egenes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "acab74a7-5be1-4616-9c25-1261ad5fa44a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CD4_T'</li><li>'Mono'</li><li>'CD8_T'</li><li>'NK'</li><li>'B'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CD4\\_T'\n",
       "\\item 'Mono'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'NK'\n",
       "\\item 'B'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CD4_T'\n",
       "2. 'Mono'\n",
       "3. 'CD8_T'\n",
       "4. 'NK'\n",
       "5. 'B'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"CD4_T\" \"Mono\"  \"CD8_T\" \"NK\"    \"B\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(egenes$cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "b01929c8-418c-434c-b97f-274a44d6677d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13744"
      ],
      "text/latex": [
       "13744"
      ],
      "text/markdown": [
       "13744"
      ],
      "text/plain": [
       "[1] 13744"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(egenes$snp_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "42ee6828-d33e-4f62-9335-e7a01d8d1239",
   "metadata": {},
   "outputs": [],
   "source": [
    "#coegene_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a563e5-bda1-4434-985c-ac2624da8232",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Calculate Motif Overlaps (using MotifbreakR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dfe5451-3ef7-4319-bdb8-0051359a0454",
   "metadata": {},
   "source": [
    "## Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "09c82f23-116a-4778-9b7d-f3e2df6249f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_results = as.data.frame(meta_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "d195de01-0fcd-4d9c-b940-9fc6d5fcb017",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define which snps to run for\n",
    "tested_snps = 'finemapping'   # alternatives: eSNPs, coeqtlSNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "db89f221-7375-4679-a489-49e193a3ee11",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(tested_snps == 'coeqtlSNPs'){\n",
    "    snp_tests = unique(meta_results[,c ('snp_id', 'snp_pos', 'snp_chr'\n",
    "                                       )])\n",
    "    }\n",
    "\n",
    "if(tested_snps == 'finemapping'){\n",
    "    snp_tests = unique(finemapping_snps)\n",
    "    }\n",
    "\n",
    "if(tested_snps == 'eSNPs'){\n",
    "    snp_tests = unique(egenes[,c('snp_id', 'snp_pos', 'snp_chr')])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "5d1ecc2f-b4f9-4bf7-b732-f79e99b029db",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define p-value threshold to use for cutting of MotifBreakR results\n",
    "sel_thres = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "d612690b-13f4-43f9-8a30-9f215c847118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1e-04"
      ],
      "text/latex": [
       "1e-04"
      ],
      "text/markdown": [
       "1e-04"
      ],
      "text/plain": [
       "[1] 1e-04"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sel_thres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "28c314c1-42c3-4073-a84e-d926e9b9ec87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "9717"
      ],
      "text/latex": [
       "9717"
      ],
      "text/markdown": [
       "9717"
      ],
      "text/plain": [
       "[1] 9717"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "6c078b84-900e-4d70-9257-84d5941f07b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#co_egene_tfs$allele = str_replace(co_egene_tfs$snp_id, '.*:', '')\n",
    "snp_tests$allele = str_replace(snp_tests$snp_id, '.*:', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "31aca234-3c2f-4dc6-90fc-011add36f376",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>snp_id</th><th scope=col>snp_pos</th><th scope=col>snp_chr</th><th scope=col>allele</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>282</th><td>9:136936281:G:C</td><td>136936281</td><td>9</td><td>C</td></tr>\n",
       "\t<tr><th scope=row>283</th><td>9:136936488:T:C</td><td>136936488</td><td>9</td><td>C</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & snp\\_id & snp\\_pos & snp\\_chr & allele\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t282 & 9:136936281:G:C & 136936281 & 9 & C\\\\\n",
       "\t283 & 9:136936488:T:C & 136936488 & 9 & C\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 4\n",
       "\n",
       "| <!--/--> | snp_id &lt;chr&gt; | snp_pos &lt;chr&gt; | snp_chr &lt;chr&gt; | allele &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 282 | 9:136936281:G:C | 136936281 | 9 | C |\n",
       "| 283 | 9:136936488:T:C | 136936488 | 9 | C |\n",
       "\n"
      ],
      "text/plain": [
       "    snp_id          snp_pos   snp_chr allele\n",
       "282 9:136936281:G:C 136936281 9       C     \n",
       "283 9:136936488:T:C 136936488 9       C     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_tests,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48c4d8d1-7ede-4fbc-bb5e-37822d81e83f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Calculate the overlap with MOTIFS (motifbreakr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7d8c771b-8f78-49a8-9d76-ddeeb0faa4d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#snp_tests = snp_tests[snp_tests$snp_id %in%  '7:28117268:C:T',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0c2670a9-5bcd-4bdf-a4f2-225e562b55ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Information on  motifbreakr\n",
    "# result intepretetation:\n",
    "# pctRef, pctAlt: display the the score for the motif in the sequence as a percentage of the best score that motif could achieve on an ideal sequence\n",
    "\n",
    "\n",
    "## Scoreing methods:\n",
    "#The first is the standard summation of log probabilities (method='log'). \n",
    "# The second and third are the weighted sum and information content methods (method='default' and method='ic')\n",
    "\n",
    "# For all three methods, motifBreakR scores and reports the reference and alternate alleles of the sequence \n",
    "# (F(sREF,M)nand F(sALT,M) ), and provides the matrix scores psREF and psALT of the SNP (or variant). \n",
    "# The scores are scaled as a fraction of scoring range 0-1 of the motif matrix\n",
    "\n",
    "# invoking filterp=TRUE and setting a threshold at a desired p-value e.g 1e-4, we perform a rough filter on the results by rounding all values in the PWM to two decimal place\n",
    "# with function calculate PValue actual result can b calculated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "325c6d27-646f-4f42-9b66-60a31709562c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "597e149b-d2b4-4d72-8e13-a0015acae12c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'finemapping'"
      ],
      "text/latex": [
       "'finemapping'"
      ],
      "text/markdown": [
       "'finemapping'"
      ],
      "text/plain": [
       "[1] \"finemapping\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tested_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c841630-ea24-4ed9-a879-507a96e022e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get already executed SNPs to not redo again from saved location and/or initialize with empty dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "dab2d824-f688-4b09-aa92-42449905f6fd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "result_data = read.csv(paste0(data_path_meta, '/MOTIF_matches_meta_5ds_results_snps_breakr_280125',tested_snps, '_', sel_thres,  '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "6d300652-61f5-4e92-ae70-280ab7102cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# result_data = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "94653b8d-c901-4a38-834a-c346611881b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "12034"
      ],
      "text/latex": [
       "12034"
      ],
      "text/markdown": [
       "12034"
      ],
      "text/plain": [
       "[1] 12034"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(result_data$SNP_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "e1ec5f1c-337a-466f-80e1-bb408b39f7b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_data$X = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "57ae4b5a-136b-4cd9-b1f1-4863f93e95fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_data$X.1 = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "ac9d3dc3-3f00-4a90-832f-0706c5880d3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "9633"
      ],
      "text/latex": [
       "9633"
      ],
      "text/markdown": [
       "9633"
      ],
      "text/plain": [
       "[1] 9633"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Check SNPs for which already result\n",
    "nrow(snp_tests[snp_tests$snp_id %in% str_replace(unique(result_data$SNP_id), 'chr', ''),])\n",
    "\n",
    "### Only run for the ones not included yet\n",
    "snp_tests = snp_tests[!snp_tests$snp_id %in% str_replace(unique(result_data$SNP_id), 'chr', ''),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "f4103f84-3650-4a19-8d1d-3bba0cc18c43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>snp_id</th><th scope=col>snp_pos</th><th scope=col>snp_chr</th><th scope=col>allele</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2864</th><td>9:126940688:A:G</td><td>126940688</td><td>9</td><td>G</td></tr>\n",
       "\t<tr><th scope=row>124592</th><td>4:118254749:G:A</td><td>118254749</td><td>4</td><td>A</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & snp\\_id & snp\\_pos & snp\\_chr & allele\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t2864 & 9:126940688:A:G & 126940688 & 9 & G\\\\\n",
       "\t124592 & 4:118254749:G:A & 118254749 & 4 & A\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 4\n",
       "\n",
       "| <!--/--> | snp_id &lt;chr&gt; | snp_pos &lt;chr&gt; | snp_chr &lt;chr&gt; | allele &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 2864 | 9:126940688:A:G | 126940688 | 9 | G |\n",
       "| 124592 | 4:118254749:G:A | 118254749 | 4 | A |\n",
       "\n"
      ],
      "text/plain": [
       "       snp_id          snp_pos   snp_chr allele\n",
       "2864   9:126940688:A:G 126940688 9       G     \n",
       "124592 4:118254749:G:A 118254749 4       A     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_tests,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "710f0a47-4d24-4127-9963-f8bd875c6cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNPs for which to execute code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "582fc2ee-f16f-45c0-9dfe-0e5a243d920d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "84"
      ],
      "text/latex": [
       "84"
      ],
      "text/markdown": [
       "84"
      ],
      "text/plain": [
       "[1] 84"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "7873cf17-ef32-48f4-966e-7d9fd051c36d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check sources available for MotifBreakR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "69183019-7133-4847-9e72-bfec6e5154c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sources = data.frame(table(mcols(MotifDb)$organism, mcols(MotifDb)$dataSource))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "3375e0b5-e81c-49dc-b83d-f77d4c90acba",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 21 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Var1</th><th scope=col>Var2</th><th scope=col>Freq</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>19</th><td>Hsapiens</td><td>cisbp_1.02             </td><td>313</td></tr>\n",
       "\t<tr><th scope=row>79</th><td>Hsapiens</td><td>FlyFactorSurvey        </td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>139</th><td>Hsapiens</td><td>HOCOMOCOv10            </td><td>640</td></tr>\n",
       "\t<tr><th scope=row>199</th><td>Hsapiens</td><td>HOCOMOCOv11-core-A     </td><td>181</td></tr>\n",
       "\t<tr><th scope=row>259</th><td>Hsapiens</td><td>HOCOMOCOv11-core-B     </td><td> 84</td></tr>\n",
       "\t<tr><th scope=row>319</th><td>Hsapiens</td><td>HOCOMOCOv11-core-C     </td><td>135</td></tr>\n",
       "\t<tr><th scope=row>379</th><td>Hsapiens</td><td>HOCOMOCOv11-secondary-A</td><td> 46</td></tr>\n",
       "\t<tr><th scope=row>439</th><td>Hsapiens</td><td>HOCOMOCOv11-secondary-B</td><td> 19</td></tr>\n",
       "\t<tr><th scope=row>499</th><td>Hsapiens</td><td>HOCOMOCOv11-secondary-C</td><td> 13</td></tr>\n",
       "\t<tr><th scope=row>559</th><td>Hsapiens</td><td>HOCOMOCOv11-secondary-D</td><td>290</td></tr>\n",
       "\t<tr><th scope=row>619</th><td>Hsapiens</td><td>HOMER                  </td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>679</th><td>Hsapiens</td><td>hPDI                   </td><td>437</td></tr>\n",
       "\t<tr><th scope=row>739</th><td>Hsapiens</td><td>JASPAR_2014            </td><td>117</td></tr>\n",
       "\t<tr><th scope=row>799</th><td>Hsapiens</td><td>JASPAR_CORE            </td><td> 66</td></tr>\n",
       "\t<tr><th scope=row>859</th><td>Hsapiens</td><td>jaspar2016             </td><td>442</td></tr>\n",
       "\t<tr><th scope=row>919</th><td>Hsapiens</td><td>jaspar2018             </td><td>522</td></tr>\n",
       "\t<tr><th scope=row>979</th><td>Hsapiens</td><td>jolma2013              </td><td>710</td></tr>\n",
       "\t<tr><th scope=row>1039</th><td>Hsapiens</td><td>ScerTF                 </td><td>  0</td></tr>\n",
       "\t<tr><th scope=row>1099</th><td>Hsapiens</td><td>stamlab                </td><td>683</td></tr>\n",
       "\t<tr><th scope=row>1159</th><td>Hsapiens</td><td>SwissRegulon           </td><td>684</td></tr>\n",
       "\t<tr><th scope=row>1219</th><td>Hsapiens</td><td>UniPROBE               </td><td>  2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 21 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & Var1 & Var2 & Freq\\\\\n",
       "  & <fct> & <fct> & <int>\\\\\n",
       "\\hline\n",
       "\t19 & Hsapiens & cisbp\\_1.02              & 313\\\\\n",
       "\t79 & Hsapiens & FlyFactorSurvey         &   0\\\\\n",
       "\t139 & Hsapiens & HOCOMOCOv10             & 640\\\\\n",
       "\t199 & Hsapiens & HOCOMOCOv11-core-A      & 181\\\\\n",
       "\t259 & Hsapiens & HOCOMOCOv11-core-B      &  84\\\\\n",
       "\t319 & Hsapiens & HOCOMOCOv11-core-C      & 135\\\\\n",
       "\t379 & Hsapiens & HOCOMOCOv11-secondary-A &  46\\\\\n",
       "\t439 & Hsapiens & HOCOMOCOv11-secondary-B &  19\\\\\n",
       "\t499 & Hsapiens & HOCOMOCOv11-secondary-C &  13\\\\\n",
       "\t559 & Hsapiens & HOCOMOCOv11-secondary-D & 290\\\\\n",
       "\t619 & Hsapiens & HOMER                   &   0\\\\\n",
       "\t679 & Hsapiens & hPDI                    & 437\\\\\n",
       "\t739 & Hsapiens & JASPAR\\_2014             & 117\\\\\n",
       "\t799 & Hsapiens & JASPAR\\_CORE             &  66\\\\\n",
       "\t859 & Hsapiens & jaspar2016              & 442\\\\\n",
       "\t919 & Hsapiens & jaspar2018              & 522\\\\\n",
       "\t979 & Hsapiens & jolma2013               & 710\\\\\n",
       "\t1039 & Hsapiens & ScerTF                  &   0\\\\\n",
       "\t1099 & Hsapiens & stamlab                 & 683\\\\\n",
       "\t1159 & Hsapiens & SwissRegulon            & 684\\\\\n",
       "\t1219 & Hsapiens & UniPROBE                &   2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 21 × 3\n",
       "\n",
       "| <!--/--> | Var1 &lt;fct&gt; | Var2 &lt;fct&gt; | Freq &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| 19 | Hsapiens | cisbp_1.02              | 313 |\n",
       "| 79 | Hsapiens | FlyFactorSurvey         |   0 |\n",
       "| 139 | Hsapiens | HOCOMOCOv10             | 640 |\n",
       "| 199 | Hsapiens | HOCOMOCOv11-core-A      | 181 |\n",
       "| 259 | Hsapiens | HOCOMOCOv11-core-B      |  84 |\n",
       "| 319 | Hsapiens | HOCOMOCOv11-core-C      | 135 |\n",
       "| 379 | Hsapiens | HOCOMOCOv11-secondary-A |  46 |\n",
       "| 439 | Hsapiens | HOCOMOCOv11-secondary-B |  19 |\n",
       "| 499 | Hsapiens | HOCOMOCOv11-secondary-C |  13 |\n",
       "| 559 | Hsapiens | HOCOMOCOv11-secondary-D | 290 |\n",
       "| 619 | Hsapiens | HOMER                   |   0 |\n",
       "| 679 | Hsapiens | hPDI                    | 437 |\n",
       "| 739 | Hsapiens | JASPAR_2014             | 117 |\n",
       "| 799 | Hsapiens | JASPAR_CORE             |  66 |\n",
       "| 859 | Hsapiens | jaspar2016              | 442 |\n",
       "| 919 | Hsapiens | jaspar2018              | 522 |\n",
       "| 979 | Hsapiens | jolma2013               | 710 |\n",
       "| 1039 | Hsapiens | ScerTF                  |   0 |\n",
       "| 1099 | Hsapiens | stamlab                 | 683 |\n",
       "| 1159 | Hsapiens | SwissRegulon            | 684 |\n",
       "| 1219 | Hsapiens | UniPROBE                |   2 |\n",
       "\n"
      ],
      "text/plain": [
       "     Var1     Var2                    Freq\n",
       "19   Hsapiens cisbp_1.02              313 \n",
       "79   Hsapiens FlyFactorSurvey           0 \n",
       "139  Hsapiens HOCOMOCOv10             640 \n",
       "199  Hsapiens HOCOMOCOv11-core-A      181 \n",
       "259  Hsapiens HOCOMOCOv11-core-B       84 \n",
       "319  Hsapiens HOCOMOCOv11-core-C      135 \n",
       "379  Hsapiens HOCOMOCOv11-secondary-A  46 \n",
       "439  Hsapiens HOCOMOCOv11-secondary-B  19 \n",
       "499  Hsapiens HOCOMOCOv11-secondary-C  13 \n",
       "559  Hsapiens HOCOMOCOv11-secondary-D 290 \n",
       "619  Hsapiens HOMER                     0 \n",
       "679  Hsapiens hPDI                    437 \n",
       "739  Hsapiens JASPAR_2014             117 \n",
       "799  Hsapiens JASPAR_CORE              66 \n",
       "859  Hsapiens jaspar2016              442 \n",
       "919  Hsapiens jaspar2018              522 \n",
       "979  Hsapiens jolma2013               710 \n",
       "1039 Hsapiens ScerTF                    0 \n",
       "1099 Hsapiens stamlab                 683 \n",
       "1159 Hsapiens SwissRegulon            684 \n",
       "1219 Hsapiens UniPROBE                  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sources[sources$Var1 == 'Hsapiens',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a486af-9784-49e1-9a05-94b1d0d3d14e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run MotifBreakR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "4415c904-f3dd-41a2-8279-bcab3016f733",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n",
      "[1] \"8:28092228:TTTG:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 2\n",
      "[1] \"6:31331551:G:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 3\n",
      "[1] \"6:31370377:A:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 4\n",
      "[1] \"6:32233370:TC:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 5\n",
      "[1] \"6:166983545:C:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 6\n",
      "[1] \"6:166988662:G:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 7\n",
      "[1] \"6:166994056:G:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 8\n",
      "[1] \"5:102940502:C:CTA\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 9\n",
      "[1] \"5:113465975:C:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 10\n",
      "[1] \"19:18454229:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 11\n",
      "[1] \"19:18498954:AAAC:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 12\n",
      "[1] \"13:31905220:C:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 13\n",
      "[1] \"11:67583528:C:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 14\n",
      "[1] \"10:112311316:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 15\n",
      "[1] \"10:112343866:A:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 16\n",
      "[1] \"10:74728830:CA:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 17\n",
      "[1] \"6:109585126:GCTT:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 18\n",
      "[1] \"6:32711955:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19\n",
      "[1] \"6:33103155:G:GAT\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 20\n",
      "[1] \"6:166961685:G:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 21\n",
      "[1] \"6:45304097:T:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 22\n",
      "[1] \"4:118254749:G:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 23\n",
      "[1] \"4:119199452:A:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 24\n",
      "[1] \"3:108593980:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 25\n",
      "[1] \"21:28725183:G:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 26\n",
      "[1] \"2:119929271:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 27\n",
      "[1] \"2:269679:C:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 28\n",
      "[1] \"18:45685529:T:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 29\n",
      "[1] \"18:45685995:T:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 30\n",
      "[1] \"17:7367035:C:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 31\n",
      "[1] \"12:68003377:C:T\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 32\n",
      "[1] \"12:9172081:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 33\n",
      "[1] \"12:55483277:C:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 34\n",
      "[1] \"12:56432026:C:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 35\n",
      "[1] \"11:17235704:T:TA\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 36\n",
      "[1] \"11:17285014:AATAC:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 37\n",
      "[1] \"11:64248783:C:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 38\n",
      "[1] \"1:20783111:CAAT:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 39\n",
      "[1] \"1:21083487:C:CT\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 40\n",
      "[1] \"1:20979756:C:CT\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 41\n",
      "[1] \"1:21010419:CA:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 42\n",
      "[1] \"1:21010662:C:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 43\n",
      "[1] \"1:20742636:A:C\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 44\n",
      "[1] \"1:20864937:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 45\n",
      "[1] \"1:168549654:T:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 46\n",
      "[1] \"9:126940688:A:G\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 47\n",
      "[1] \"22:42090024:C:CA\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 48\n",
      "[1] \"12:55532911:G:A\"\n",
      "[1] 297504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 49\n",
      "[1] \"1:67383648:A:C\"\n",
      "[1] 297504\n",
      "[1] 50\n",
      "[1] \"8:392733:A:G\"\n",
      "[1] 297552\n",
      "[1] 51\n",
      "[1] \"8:392957:C:T\"\n",
      "[1] 297586\n",
      "[1] 52\n",
      "[1] \"8:392958:A:G\"\n",
      "[1] 297617\n",
      "[1] 53\n",
      "[1] \"8:395326:C:T\"\n",
      "[1] 297751\n",
      "[1] 54\n",
      "[1] \"8:403398:C:G\"\n",
      "[1] 297754\n",
      "[1] 55\n",
      "[1] \"8:403626:G:C\"\n",
      "[1] 297774\n",
      "[1] 56\n",
      "[1] \"8:403936:T:C\"\n",
      "[1] 297786\n",
      "[1] 57\n",
      "[1] \"8:404843:C:T\"\n",
      "[1] 297790\n",
      "[1] 58\n",
      "[1] \"8:405642:C:G\"\n",
      "[1] 297806\n",
      "[1] 59\n",
      "[1] \"8:407362:A:G\"\n",
      "[1] 297809\n",
      "[1] 60\n",
      "[1] \"8:412267:T:A\"\n",
      "[1] 297813\n",
      "[1] 61\n",
      "[1] \"8:412326:A:C\"\n",
      "[1] 297828\n",
      "[1] 62\n",
      "[1] \"8:412816:A:T\"\n",
      "[1] 297848\n",
      "[1] 63\n",
      "[1] \"8:412907:C:G\"\n",
      "[1] 297855\n",
      "[1] 64\n",
      "[1] \"8:416090:T:A\"\n",
      "[1] 297930\n",
      "[1] 65\n",
      "[1] \"8:417409:C:T\"\n",
      "[1] 297960\n",
      "[1] 66\n",
      "[1] \"8:420046:T:A\"\n",
      "[1] 298002\n",
      "[1] 67\n",
      "[1] \"8:432957:T:C\"\n",
      "[1] 298005\n",
      "[1] 68\n",
      "[1] \"8:438327:A:G\"\n",
      "[1] 298018\n",
      "[1] 69\n",
      "[1] \"8:442670:G:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 70\n",
      "[1] \"6:33084000:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 71\n",
      "[1] \"6:32700774:GTA:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 72\n",
      "[1] \"6:32415479:A:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 73\n",
      "[1] \"20:15335874:T:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 74\n",
      "[1] \"18:74258865:T:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 75\n",
      "[1] \"15:83070624:AGT:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 76\n",
      "[1] \"14:105063845:CT:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 77\n",
      "[1] \"14:50320325:C:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 78\n",
      "[1] \"14:50332993:C:T\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 79\n",
      "[1] \"1:157698600:T:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 80\n",
      "[1] \"1:157546450:AAAACC:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 81\n",
      "[1] \"9:33123108:T:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 82\n",
      "[1] \"9:33140978:G:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 83\n",
      "[1] \"9:125832202:C:CT\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 84\n",
      "[1] \"8:89733130:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 85\n",
      "[1] \"6:158788885:T:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 86\n",
      "[1] \"6:139153935:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 87\n",
      "[1] \"6:31491920:T:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 88\n",
      "[1] \"6:31262226:TG:T\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 89\n",
      "[1] \"6:32706353:G:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 90\n",
      "[1] \"6:32342638:C:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 91\n",
      "[1] \"6:32453233:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 92\n",
      "[1] \"6:167020437:G:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 93\n",
      "[1] \"6:44848132:A:AT\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 94\n",
      "[1] \"5:65568005:T:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 95\n",
      "[1] \"5:65593808:G:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 96\n",
      "[1] \"5:65556359:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 97\n",
      "[1] \"5:65759400:CTG:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 98\n",
      "[1] \"5:96913619:G:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 99\n",
      "[1] \"4:185148024:T:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 100\n",
      "[1] \"4:185160368:CA:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 101\n",
      "[1] \"4:185165403:C:T\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 102\n",
      "[1] \"3:32627433:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 103\n",
      "[1] \"3:158624850:GAGA:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 104\n",
      "[1] \"21:36113634:C:CA\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 105\n",
      "[1] \"21:29542032:AGTGT:A\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 106\n",
      "[1] \"16:28518140:T:TA\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 107\n",
      "[1] \"14:55140648:C:T\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 108\n",
      "[1] \"14:55266784:CT:C\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 109\n",
      "[1] \"14:55344609:A:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 110\n",
      "[1] \"14:55376878:G:T\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 111\n",
      "[1] \"13:43023570:C:T\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 112\n",
      "[1] \"12:75513882:C:CAT\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 113\n",
      "[1] \"12:6904912:C:G\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 114\n",
      "[1] \"1:226160095:C:CT\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in prepareVariants(fsnplist = snpList, genome.bsgenome = genome.bsgenome, :\n",
      "\"The only variants included in the input set are indels, but legacy scoring was selected.\n",
      "Legacy scoring is not availble for use with indels and will be disabled.\"\n",
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 115\n",
      "[1] \"1:19868678:C:CT\"\n",
      "[1] 298077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres, :\n",
      "\"No SNP/Motif Interactions reached threshold\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 116\n",
      "[1] \"1:69671928:A:G\"\n",
      "[1] 298077\n"
     ]
    }
   ],
   "source": [
    "for(i in 1:nrow(snp_tests)){\n",
    "    data = snp_tests[i,]\n",
    "    \n",
    "    chr = paste0('chr', data$snp_chr)\n",
    "    snp_position = as.numeric(data$snp_pos)\n",
    "    snp = data$snp_id\n",
    "    \n",
    "    # Split the SNP ID to get individual components\n",
    "    snp_info <- strsplit(snp, \":\")[[1]]\n",
    "    chromosome <- snp_info[1]\n",
    "    position <- as.numeric(snp_info[2])\n",
    "    ref <- snp_info[3]\n",
    "    alt <- snp_info[4]\n",
    "    \n",
    "    # Create a GRanges object\n",
    "    gr <- GRanges(\n",
    "      seqnames = c(paste0('chr', Rle(chromosome))),\n",
    "      ranges = c( IRanges(start = position, end = position  + nchar(ref) -1)),\n",
    "      strand = c( \"*\"),  # SNPs are usually unstranded\n",
    "      SNP_id = c(paste0('chr', snp)),\n",
    "      REF =  DNAStringSet(ref),\n",
    "      ALT =  DNAStringSet(alt)\n",
    "    )\n",
    "    \n",
    "    # Specify the genome build (e.g., hg19 or hg38)\n",
    "    genome(gr) <- \"hg38\"\n",
    "    attr(gr, \"genome.package\") = \"BSgenome.Hsapiens.UCSC.hg38\"\n",
    "    \n",
    "    # Get the break result\n",
    "    breakr_result = try(motifbreakR(snpList = gr, filterp = TRUE, threshold = sel_thres , # defines whether to filter by p-value instead of pct score\n",
    "                       pwmList = subset(MotifDb, \n",
    "                                        dataSource %in% c(\"HOCOMOCOv11-core-A\",\"hPDI\", \"jolma2013\" , \"stamlab\",\"SwissRegulon\", \"JASPAR_CORE\",'JASPAR_2014','jaspar2016', \"jaspar2018\" , 'cisbp_1.02', 'HOMER', 'HOCOMOCOv10' ))))  # previous run:\n",
    "    # \"JASPAR_CORE\",'JASPAR_2014','jaspar2016', \"jaspar2018\" , 'cisbp_1.02', 'HOMER', 'HOCOMOCOv10'\n",
    "    \n",
    "    if(class(breakr_result) != 'try-error'){\n",
    "    \n",
    "    breakr_result_data = as.data.frame(breakr_result )\n",
    "    breakr_result_data$motifPos = NULL\n",
    "    breakr_result_data$altPos = NULL\n",
    "    result_data = rbind(breakr_result_data, result_data) \n",
    "    print(i)\n",
    "    print(snp)\n",
    "    print(nrow(result_data))\n",
    "    flush.console()\n",
    "        }\n",
    "    write.csv(result_data, paste0(data_path_meta, '/MOTIF_matches_meta_5ds_results_snps_breakr_280125', tested_snps, '_', sel_thres , '.csv'))\n",
    "  } \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e272349-4a09-4b64-9e67-14ac8ecdf13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(result_data, paste0(data_path_meta, '/MOTIF_matches_meta_5ds_results_snps_breakr', '.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05dcddac-2505-485c-845d-7d8831c19914",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Calculate the overlap with MOTIFs (searchseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20a870e0-3f55-426e-aa8c-362353f36f6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Use searchseq to check for motifs around SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "288b41f6-bfef-44d6-b139-fb544a9a942e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Window Sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "e08b7556-d194-443a-8178-84e9bb32909a",
   "metadata": {},
   "outputs": [],
   "source": [
    "j  = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1173187e-a483-48ad-b646-1abb6430c02d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Alternative: initialize empty dataframe or load already calculated results to not recalculate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "9b1a70fd-eb3f-4806-89ca-558c3f4b055c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#motif_match = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "258005da-68f1-4d8e-a99d-632fc2886a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_match = read.csv(paste0(data_path_meta, '/MOTIF_matches_meta_5ds_results_snps_', tested_snps, '_', sel_thres, '_', j, '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "482053ef-6856-4b9e-9065-459dc44ce369",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_match$X = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "5371d507-42b2-492f-a77b-7a9116ac2788",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>source</th><th scope=col>feature</th><th scope=col>start</th><th scope=col>end</th><th scope=col>absScore</th><th scope=col>relScore</th><th scope=col>strand</th><th scope=col>ID</th><th scope=col>TF</th><th scope=col>class</th><th scope=col>siteSeqs</th><th scope=col>snp_id</th><th scope=col>sequence</th><th scope=col>window_size</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Wild-type</td><td>TFBS</td><td>TFBS</td><td>11</td><td>19</td><td>3.483203</td><td>0.8007783</td><td>-</td><td>MA0084.1</td><td>SRY    </td><td>High-mobility group (HMG) domain factors</td><td>TTTAGCAAA</td><td>9:134944489:C:A</td><td>ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG</td><td>20</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Wild-type</td><td>TFBS</td><td>TFBS</td><td> 1</td><td> 6</td><td>3.642925</td><td>0.8276184</td><td>+</td><td>MA0130.1</td><td>ZNF354C</td><td>C2H2 zinc finger factors                </td><td>ATCCAT   </td><td>9:134944489:C:A</td><td>ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG</td><td>20</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & seqnames & source & feature & start & end & absScore & relScore & strand & ID & TF & class & siteSeqs & snp\\_id & sequence & window\\_size\\\\\n",
       "  & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & Wild-type & TFBS & TFBS & 11 & 19 & 3.483203 & 0.8007783 & - & MA0084.1 & SRY     & High-mobility group (HMG) domain factors & TTTAGCAAA & 9:134944489:C:A & ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG & 20\\\\\n",
       "\t2 & Wild-type & TFBS & TFBS &  1 &  6 & 3.642925 & 0.8276184 & + & MA0130.1 & ZNF354C & C2H2 zinc finger factors                 & ATCCAT    & 9:134944489:C:A & ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG & 20\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 15\n",
       "\n",
       "| <!--/--> | seqnames &lt;chr&gt; | source &lt;chr&gt; | feature &lt;chr&gt; | start &lt;int&gt; | end &lt;int&gt; | absScore &lt;dbl&gt; | relScore &lt;dbl&gt; | strand &lt;chr&gt; | ID &lt;chr&gt; | TF &lt;chr&gt; | class &lt;chr&gt; | siteSeqs &lt;chr&gt; | snp_id &lt;chr&gt; | sequence &lt;chr&gt; | window_size &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Wild-type | TFBS | TFBS | 11 | 19 | 3.483203 | 0.8007783 | - | MA0084.1 | SRY     | High-mobility group (HMG) domain factors | TTTAGCAAA | 9:134944489:C:A | ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG | 20 |\n",
       "| 2 | Wild-type | TFBS | TFBS |  1 |  6 | 3.642925 | 0.8276184 | + | MA0130.1 | ZNF354C | C2H2 zinc finger factors                 | ATCCAT    | 9:134944489:C:A | ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG | 20 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames  source feature start end absScore relScore  strand ID       TF     \n",
       "1 Wild-type TFBS   TFBS    11    19  3.483203 0.8007783 -      MA0084.1 SRY    \n",
       "2 Wild-type TFBS   TFBS     1     6  3.642925 0.8276184 +      MA0130.1 ZNF354C\n",
       "  class                                    siteSeqs  snp_id         \n",
       "1 High-mobility group (HMG) domain factors TTTAGCAAA 9:134944489:C:A\n",
       "2 C2H2 zinc finger factors                 ATCCAT    9:134944489:C:A\n",
       "  sequence                                  window_size\n",
       "1 ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG 20         \n",
       "2 ATCCATCAAGTTTGCTAAACCACTTTCCCTCTGGAGCCATG 20         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(motif_match,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "5ce56966-ab49-4408-8847-328d3514ce6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13454"
      ],
      "text/latex": [
       "13454"
      ],
      "text/markdown": [
       "13454"
      ],
      "text/plain": [
       "[1] 13454"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "7638e962-a503-48af-8d6a-2ed1894a1bd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13416"
      ],
      "text/latex": [
       "13416"
      ],
      "text/markdown": [
       "13416"
      ],
      "text/plain": [
       "[1] 13416"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(unique(motif_match$snp_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "f184a695-e8bf-49b4-b285-8fc47977914e",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_tests = snp_tests[!snp_tests$snp_id %in% unique(motif_match$snp_id),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "3aa698b5-d84b-4df6-816f-56002156293f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "74"
      ],
      "text/latex": [
       "74"
      ],
      "text/markdown": [
       "74"
      ],
      "text/plain": [
       "[1] 74"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "a59611d2-362d-4706-a599-7c9674b5748f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get human TF binding motifs from JASPAR\n",
    "opts <- list(species = 9606)\n",
    "motifs <- getMatrixSet(JASPAR2020, opts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "98aa3844-de31-43f4-9dbb-94735116d610",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert PFMatrixList to PWMatrixList\n",
    "pwms <- toPWM(motifs)  # motifs is the PFMatrixList, we convert it to PWMatrixList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "42ffbfee-40f5-476b-9d93-8df1f60ccbc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#str(pwms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "eadca983-1505-4c9d-bada-9011b4ff7bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a data frame with Matrix IDs and Gene Symbols (for mapping the tfs)\n",
    "motif_info <- data.frame(\n",
    "  matrix_id = sapply(pwms, ID),  # Get the JASPAR matrix ID\n",
    "  gene_symbol = sapply(pwms, name)  # Get the corresponding gene name\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "a9cf4351-7021-433e-ab18-ff177642c064",
   "metadata": {},
   "outputs": [],
   "source": [
    "##  general parameters for search\n",
    "window_sizes <- c(20) # window around snp position to take into account for motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "cf2fa4e3-3122-49ab-b880-8af99b7f6e6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Iterate over all SNPs and check whether there is motif overlap at the position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "e75bc8e1-f664-42d0-8841-1fa253d055ab",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 1\n",
      "[1] 2\n",
      "[1] 3\n",
      "[1] 4\n",
      "[1] 5\n",
      "[1] 6\n",
      "[1] 7\n",
      "[1] 8\n",
      "[1] 9\n",
      "[1] 10\n",
      "[1] 11\n",
      "[1] 12\n",
      "[1] 13\n",
      "[1] 14\n",
      "[1] 15\n",
      "[1] 16\n",
      "[1] 17\n",
      "[1] 18\n",
      "[1] 19\n",
      "[1] 20\n",
      "[1] 21\n",
      "[1] 22\n",
      "[1] 23\n",
      "[1] 24\n",
      "[1] 25\n",
      "[1] 26\n",
      "[1] 27\n",
      "[1] 28\n",
      "[1] 29\n",
      "[1] 30\n",
      "[1] 31\n",
      "[1] 32\n",
      "[1] 33\n",
      "[1] 34\n",
      "[1] 35\n",
      "[1] 36\n",
      "[1] 37\n",
      "[1] 38\n",
      "[1] 39\n",
      "[1] 40\n",
      "[1] 41\n",
      "[1] 42\n",
      "[1] 43\n",
      "[1] 44\n",
      "[1] 45\n",
      "[1] 46\n",
      "[1] 47\n",
      "[1] 48\n",
      "[1] 49\n",
      "[1] 50\n",
      "[1] 51\n",
      "[1] 52\n",
      "[1] 53\n",
      "[1] 54\n",
      "[1] 55\n",
      "[1] 56\n",
      "[1] 57\n",
      "[1] 58\n",
      "[1] 59\n",
      "[1] 60\n",
      "[1] 61\n",
      "[1] 62\n",
      "[1] 63\n",
      "[1] 64\n",
      "[1] 65\n",
      "[1] 66\n",
      "[1] 67\n",
      "[1] 68\n",
      "[1] 69\n",
      "[1] 70\n",
      "[1] 71\n",
      "[1] 72\n",
      "[1] 73\n",
      "[1] 74\n"
     ]
    }
   ],
   "source": [
    "#for(i in 1:nrow(co_egene_tfs)){\n",
    "for(j in window_sizes){\n",
    "    for(i in 1:nrow(snp_tests)){\n",
    "        #data = co_egene_tfs[i,]\n",
    "        data = snp_tests[i,]\n",
    "\n",
    "        chr = paste0('chr', data$snp_chr)\n",
    "        snp_position = as.numeric(data$snp_pos)\n",
    "        snp = data$snp_id\n",
    "\n",
    "        ## get the sequence of the snp\n",
    "        snp_region <- GRanges(seqnames = chr, ranges = IRanges(start = snp_position - j, end = snp_position + j))\n",
    "        sequence <- getSeq(BSgenome.Hsapiens.UCSC.hg38, snp_region)\n",
    "        data$sequence = as.character(sequence) # save the sequence to the data\n",
    "\n",
    "        ### Create wild-tipe and variant sequence\n",
    "        wild_type_seq <- as.character(sequence)\n",
    "        variant_seq <- wild_type_seq\n",
    "        substr(variant_seq,j + 1,j + 1) <- data$allele ##TBD: check whether allele always the second information as assumed now in extraction\n",
    "\n",
    "        data$variant_seq = variant_seq # save the variant sequence to the data\n",
    "\n",
    "        ##  Convert to format to search for in JASPAR\n",
    "        wild_type_dna <- DNAString(wild_type_seq)\n",
    "        variant_dna <- DNAString(variant_seq)\n",
    "\n",
    "        # Scan for TF binding sites in both sequences\n",
    "\n",
    "        # Scan the wild-type sequence for TF binding sites\n",
    "        wild_sites <- searchSeq(pwms, wild_type_dna, min.score = \"80%\", seqname = \"Wild-type\")\n",
    "        ## TBD: add pvalues(wild_sites) --> to get p-values\n",
    "\n",
    "        # Scan the variant sequence for TF binding sites\n",
    "        variant_sites <- searchSeq(pwms, variant_dna, min.score = \"80%\", seqname = \"Variant\")\n",
    "\n",
    "        # Get the TFS\n",
    "        wild_sites_df <- as.data.frame(wild_sites)   # check whether directly extract the symbol\n",
    "        variant_sites_df <- as.data.frame(variant_sites)\n",
    "\n",
    "        # Replace matrix IDs in the result with gene symbols\n",
    "        #wild_tfs_symbols <- motif_info$gene_symbol[match(wild_tfs, motif_info$matrix_id)]\n",
    "        #variant_tfs_symbols <- motif_info$gene_symbol[match(variant_tfs, motif_info$matrix_id)]\n",
    "\n",
    "        # Get Dataframe of the MOTIF results\n",
    "        #wild_sites_df = writeGFF3(wild_sites)\n",
    "        #variant_sites_df = writeGFF3(variant_sites)\n",
    "\n",
    "        # Add the TF information\n",
    "        #wild_sites_df$TF = wild_tfs_symbols\n",
    "        #variant_sites_df$TF = variant_tfs_symbols\n",
    "\n",
    "        # Add the information about snp\n",
    "        wild_sites_df$snp_id = snp\n",
    "        wild_sites_df$sequence = data$sequence\n",
    "        variant_sites_df$snp_id = snp\n",
    "        variant_sites_df$sequence = data$variant_seq \n",
    "\n",
    "        result_data = rbind(wild_sites_df, variant_sites_df)\n",
    "        result_data$window_size = j\n",
    "\n",
    "        #data$variant_motif_match = 0\n",
    "        #data$variant_motif_match[data$coeGene %in% variant_tfs_symbols] = 1  # do the comparison now later\n",
    "\n",
    "        #data$wild_motif_match = 0\n",
    "        #data$wild_motif_match[data$coeGene %in% wild_tfs_symbols] = 1\n",
    "\n",
    "        #data$motif_tfs = paste0(sort(unique(c(variant_tfs_symbols, wild_tfs_symbols))), collapse = ';')\n",
    "        motif_match = rbind(motif_match, result_data)\n",
    "        print(i)\n",
    "        flush.console()\n",
    "        }\n",
    "\n",
    "    write.csv(motif_match, paste0(data_path_meta, '/MOTIF_matches_meta_5ds_results_snps_191124', tested_snps, '_', sel_thres, '_', j, '.csv'))\n",
    "  }  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8fc57f2-1d1c-4dce-9f0a-3f6f191afa1f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e70310-292c-4db6-9771-6fcce4e44a61",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0715f8ca-2f6f-4458-b994-98ed586a8442",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(motif_match, paste0(data_path_meta, '/MOTIF_matches_meta_5ds_results.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
