{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "253ab18f-d2dc-4dee-a9db-5255571056d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### GWAS enrichment analysis based on coloc results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "802f0d1c-925e-480b-a547-865cc6d997b2",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "d187711e-139c-4eaf-bcaf-5795174c795d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "\"incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "76d160be-a6b2-44d8-b25e-27acd6cecacc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(gwasrapidd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "fb96c2f1-9802-4bf3-b878-a1ac09d21273",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS4_Plot_Config.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "id": "f13113e4-c78d-4078-b0b5-85c87528abdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] '0.99.17'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "packageVersion(\"gwasrapidd\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e59e15e0-23cd-457b-b704-05dbfe404876",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "id": "7ed41574-8e23-442e-9b64-69f4c7a2ad5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Paths to the data: eqtl, co-eQTL mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "c0e8991e-7ce6-4981-b60c-62c9e04a67b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../data/current/coeqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "e29e542a-a68e-414a-8e44-250010340baf",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path_eqtl =  '../data/current/eqtl_mapping/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "105170f5-1178-450d-9577-f081529872b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_results_path = 'co_qtls_decision_tree/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "6f1f80ec-663a-41dc-9527-4e86bd93aedf",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path =  '../data/current//results/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "8902d339-d6f1-4a5d-a10a-4ba0f276a198",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filtering decision"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "2225d863-0eb9-4538-a63d-de63fb431422",
   "metadata": {},
   "outputs": [],
   "source": [
    "only_esnps =TRUE   # only use eQTL as backgrounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "ee95c839-d575-422e-9648-1be2c80d65d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_HLA = TRUE    # remove HLA genes from enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "1fb2ee7e-7dbf-4667-9f55-65806e18a92e",
   "metadata": {},
   "outputs": [],
   "source": [
    "only_dt = TRUE ## only use decision tree genes in background and analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "b7aaf935-bcb3-4b7b-9e82-e60ecbef7f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Which set of significant co-eQTLs to load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "36853f61-4cc3-4b5d-8fd1-ee0dc3ca0bb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### To this meta-dataset all the information should be joined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "719ba303-6981-4f81-8ab7-494f147896cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_val_thres = ''  # '' in case of 0.05; numeric value otherwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "89f2f2c7-9165-4408-9e79-0886335d2166",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_val_column = 'p_val_mt_eGene'  #p_val_mt_eGene; q_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "3ff35de8-ad04-4b5f-b3d3-32b3b70032a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(p_val_thres == ''){\n",
    "sign_coeqtl_path = paste0(data_path, coeqtl_results_path,  \"/shared_results/5DS_Meta_Analysis_Sign_coeQTLs133\", p_val_column, \"_.csv\")\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "baf74bb1-873a-42d5-a9cc-578db7e052a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] \"2025-07-20 12:14:08 CEST\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "file.info(sign_coeqtl_path )$ctime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "cdb64542-e82f-485a-a2d8-a3f1ae6d0ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define the threshold for colocalisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "dc8d7eee-b9b4-4c00-a409-dcae05d82ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_thres =  'relation'   # meani"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "e741fe5b-00d5-4ad5-970f-4119f13c7f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define which traits to filter on"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "id": "ff4a32ab-5108-4661-b62e-5226b36b2b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "use_gwas = 'disease' \n",
    "# Define GWAS subset: all_traits - no filtering; rm_cell_count = remove cell count features; disease = filter on disease specific features; immune = filter on immune specific features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "id": "3a1be7b4-252b-488a-ad35-14c67c0d5a07",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define whether to exclude SNPs without CS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "id": "9c29d4b0-bdd2-44e0-9c63-691c1592a2ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "cs_snps = FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "id": "20c28468-2550-475c-bd1b-031ff0ee35c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_snp_based  = FALSE   # Define whether to combine with SNP annotation from SNP based GWAS enrichment analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96ec5300-7663-481d-b632-a0d23a9f5093",
   "metadata": {},
   "source": [
    "# Data "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1ac7792-06b0-4230-9371-e34c9994c85d",
   "metadata": {},
   "source": [
    "## GWAS enrichment result SNP based"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "767cf470-e354-4e69-9229-eea0df1ec045",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment = fread(paste0('../data/current/meta-data/gwas_enrichment_snp/', 'coeqtl_gwas_enrich_table_all_celltypes.tsv.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "24003353-dbca-433c-8cd3-f9e3a951fe99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>ALL</th><th scope=col>ALL_no_cell_count</th><th scope=col>ALL_disease</th><th scope=col>ALL_immune</th><th scope=col>B_all_traits</th><th scope=col>B_all_traits_no_cell_count</th><th scope=col>B_disease</th><th scope=col>B_immune</th><th scope=col>CD4_T_all_traits</th><th scope=col>⋯</th><th scope=col>CD8_T_disease</th><th scope=col>CD8_T_immune</th><th scope=col>Mono_all_traits</th><th scope=col>Mono_all_traits_no_cell_count</th><th scope=col>Mono_disease</th><th scope=col>Mono_immune</th><th scope=col>NK_all_traits</th><th scope=col>NK_all_traits_no_cell_count</th><th scope=col>NK_disease</th><th scope=col>NK_immune</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>p_value </td><td>2.407397e-08</td><td>2.338709e-08</td><td>0.01870063</td><td>0.0006640809</td><td>0.173391</td><td>0.1731173</td><td>0.1422117</td><td>0.006454817</td><td>2.724444e-06</td><td>⋯</td><td>0.04961523</td><td>0.003217347</td><td>0.1122192</td><td>0.1120808</td><td>0.8767768</td><td>0.2054784</td><td>0.02756775</td><td>0.02746182</td><td>0.4340997</td><td>0.01034244</td></tr>\n",
       "\t<tr><td>estimate</td><td>2.026247e+00</td><td>2.029717e+00</td><td>1.53156342</td><td>2.7811929825</td><td>1.418182</td><td>1.4216715</td><td>1.5984694</td><td>3.626451613</td><td>2.131267e+00</td><td>⋯</td><td>1.63273427</td><td>3.173868313</td><td>1.4624373</td><td>1.4695271</td><td>1.0252781</td><td>1.8622449</td><td>1.63591721</td><td>1.63963964</td><td>1.2548924</td><td>2.98793103</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " V1 & ALL & ALL\\_no\\_cell\\_count & ALL\\_disease & ALL\\_immune & B\\_all\\_traits & B\\_all\\_traits\\_no\\_cell\\_count & B\\_disease & B\\_immune & CD4\\_T\\_all\\_traits & ⋯ & CD8\\_T\\_disease & CD8\\_T\\_immune & Mono\\_all\\_traits & Mono\\_all\\_traits\\_no\\_cell\\_count & Mono\\_disease & Mono\\_immune & NK\\_all\\_traits & NK\\_all\\_traits\\_no\\_cell\\_count & NK\\_disease & NK\\_immune\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t p\\_value  & 2.407397e-08 & 2.338709e-08 & 0.01870063 & 0.0006640809 & 0.173391 & 0.1731173 & 0.1422117 & 0.006454817 & 2.724444e-06 & ⋯ & 0.04961523 & 0.003217347 & 0.1122192 & 0.1120808 & 0.8767768 & 0.2054784 & 0.02756775 & 0.02746182 & 0.4340997 & 0.01034244\\\\\n",
       "\t estimate & 2.026247e+00 & 2.029717e+00 & 1.53156342 & 2.7811929825 & 1.418182 & 1.4216715 & 1.5984694 & 3.626451613 & 2.131267e+00 & ⋯ & 1.63273427 & 3.173868313 & 1.4624373 & 1.4695271 & 1.0252781 & 1.8622449 & 1.63591721 & 1.63963964 & 1.2548924 & 2.98793103\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 25\n",
       "\n",
       "| V1 &lt;chr&gt; | ALL &lt;dbl&gt; | ALL_no_cell_count &lt;dbl&gt; | ALL_disease &lt;dbl&gt; | ALL_immune &lt;dbl&gt; | B_all_traits &lt;dbl&gt; | B_all_traits_no_cell_count &lt;dbl&gt; | B_disease &lt;dbl&gt; | B_immune &lt;dbl&gt; | CD4_T_all_traits &lt;dbl&gt; | ⋯ ⋯ | CD8_T_disease &lt;dbl&gt; | CD8_T_immune &lt;dbl&gt; | Mono_all_traits &lt;dbl&gt; | Mono_all_traits_no_cell_count &lt;dbl&gt; | Mono_disease &lt;dbl&gt; | Mono_immune &lt;dbl&gt; | NK_all_traits &lt;dbl&gt; | NK_all_traits_no_cell_count &lt;dbl&gt; | NK_disease &lt;dbl&gt; | NK_immune &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| p_value  | 2.407397e-08 | 2.338709e-08 | 0.01870063 | 0.0006640809 | 0.173391 | 0.1731173 | 0.1422117 | 0.006454817 | 2.724444e-06 | ⋯ | 0.04961523 | 0.003217347 | 0.1122192 | 0.1120808 | 0.8767768 | 0.2054784 | 0.02756775 | 0.02746182 | 0.4340997 | 0.01034244 |\n",
       "| estimate | 2.026247e+00 | 2.029717e+00 | 1.53156342 | 2.7811929825 | 1.418182 | 1.4216715 | 1.5984694 | 3.626451613 | 2.131267e+00 | ⋯ | 1.63273427 | 3.173868313 | 1.4624373 | 1.4695271 | 1.0252781 | 1.8622449 | 1.63591721 | 1.63963964 | 1.2548924 | 2.98793103 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1       ALL          ALL_no_cell_count ALL_disease ALL_immune   B_all_traits\n",
       "1 p_value  2.407397e-08 2.338709e-08      0.01870063  0.0006640809 0.173391    \n",
       "2 estimate 2.026247e+00 2.029717e+00      1.53156342  2.7811929825 1.418182    \n",
       "  B_all_traits_no_cell_count B_disease B_immune    CD4_T_all_traits ⋯\n",
       "1 0.1731173                  0.1422117 0.006454817 2.724444e-06     ⋯\n",
       "2 1.4216715                  1.5984694 3.626451613 2.131267e+00     ⋯\n",
       "  CD8_T_disease CD8_T_immune Mono_all_traits Mono_all_traits_no_cell_count\n",
       "1 0.04961523    0.003217347  0.1122192       0.1120808                    \n",
       "2 1.63273427    3.173868313  1.4624373       1.4695271                    \n",
       "  Mono_disease Mono_immune NK_all_traits NK_all_traits_no_cell_count NK_disease\n",
       "1 0.8767768    0.2054784   0.02756775    0.02746182                  0.4340997 \n",
       "2 1.0252781    1.8622449   1.63591721    1.63963964                  1.2548924 \n",
       "  NK_immune \n",
       "1 0.01034244\n",
       "2 2.98793103"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_enrichment,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "37743d82-d450-4367-98a5-ebf1265d1c0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Transform  format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "2af4e58f-dc77-4b72-91f3-10b90c82e5ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(snp_enrichment) = snp_enrichment$V1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "2116ddae-c537-4208-88a5-1aae773a8021",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment = as.matrix(snp_enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "33799d56-9098-4a3e-93f9-54ad7551987d",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment = t(snp_enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "ea2848b3-699e-416d-a60e-10224d12aa3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(snp_enrichment) = snp_enrichment['V1',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "4a1b76d2-f9f2-4841-8073-20e313c8b885",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment <- snp_enrichment[-1, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "f75684ad-75d8-47ad-a895-b6b0e2cbceaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment = data.frame(snp_enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "0c7bd6cd-2460-480c-9a21-57703662f5c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment$cell_type = rownames(snp_enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "9815abf9-2d91-4035-9d46-5a4020bd18b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment$type = str_replace(snp_enrichment$cell_type, 'ALL|CD4_T|CD8_T|Mono|B|NK', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "id": "b42b1acb-c0ae-4221-be33-f65f7b549dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment$cell_type[snp_enrichment$type != ''] = str_replace(snp_enrichment$cell_type[snp_enrichment$type != ''], snp_enrichment$type[snp_enrichment$type != ''], '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "id": "8c4f62ae-e39d-4d96-963e-d5c358e9dd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment$type = str_replace(snp_enrichment$type, '_', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "id": "aca693c6-0141-4f94-a796-4fa00ed79011",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ALL</th><td>2.407397e-08</td><td>2.026247e+00</td><td>2.617472e+00</td><td>1.569775e+00</td><td>ALL</td><td>             </td></tr>\n",
       "\t<tr><th scope=row>ALL_no_cell_count</th><td>2.338709e-08</td><td>2.029717e+00</td><td>2.621962e+00</td><td>1.572458e+00</td><td>ALL</td><td>no_cell_count</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tALL & 2.407397e-08 & 2.026247e+00 & 2.617472e+00 & 1.569775e+00 & ALL &              \\\\\n",
       "\tALL\\_no\\_cell\\_count & 2.338709e-08 & 2.029717e+00 & 2.621962e+00 & 1.572458e+00 & ALL & no\\_cell\\_count\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 6\n",
       "\n",
       "| <!--/--> | p_value &lt;chr&gt; | estimate &lt;chr&gt; | conf_upper &lt;chr&gt; | conf_lower &lt;chr&gt; | cell_type &lt;chr&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| ALL | 2.407397e-08 | 2.026247e+00 | 2.617472e+00 | 1.569775e+00 | ALL | <!----> |\n",
       "| ALL_no_cell_count | 2.338709e-08 | 2.029717e+00 | 2.621962e+00 | 1.572458e+00 | ALL | no_cell_count |\n",
       "\n"
      ],
      "text/plain": [
       "                  p_value      estimate     conf_upper   conf_lower   cell_type\n",
       "ALL               2.407397e-08 2.026247e+00 2.617472e+00 1.569775e+00 ALL      \n",
       "ALL_no_cell_count 2.338709e-08 2.029717e+00 2.621962e+00 1.572458e+00 ALL      \n",
       "                  type         \n",
       "ALL                            \n",
       "ALL_no_cell_count no_cell_count"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_enrichment,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "id": "f2bab068-9914-4086-9967-0f420ce5ce98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>''</li><li>'no_cell_count'</li><li>'disease'</li><li>'immune'</li><li>'all_traits'</li><li>'all_traits_no_cell_count'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item ''\n",
       "\\item 'no\\_cell\\_count'\n",
       "\\item 'disease'\n",
       "\\item 'immune'\n",
       "\\item 'all\\_traits'\n",
       "\\item 'all\\_traits\\_no\\_cell\\_count'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. ''\n",
       "2. 'no_cell_count'\n",
       "3. 'disease'\n",
       "4. 'immune'\n",
       "5. 'all_traits'\n",
       "6. 'all_traits_no_cell_count'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"\"                         \"no_cell_count\"           \n",
       "[3] \"disease\"                  \"immune\"                  \n",
       "[5] \"all_traits\"               \"all_traits_no_cell_count\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(snp_enrichment$type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "id": "37d45e13-7dcd-4879-999f-4c1495cf80be",
   "metadata": {},
   "outputs": [],
   "source": [
    "### replace categories to match selection\n",
    "snp_enrichment$type = str_replace(snp_enrichment$type, 'all_traits_no_cell_count', 'rm_cell_count')\n",
    "snp_enrichment$type[snp_enrichment$type == 'no_cell_count'] = 'rm_cell_count'\n",
    "snp_enrichment$type[snp_enrichment$type == ''] = 'all_traits'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "id": "5a4f0ba7-c66b-4b89-a3b8-7c2363eaea98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ALL'</li><li>'B'</li><li>'CD4_T'</li><li>'CD8_T'</li><li>'Mono'</li><li>'NK'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ALL'\n",
       "\\item 'B'\n",
       "\\item 'CD4\\_T'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'Mono'\n",
       "\\item 'NK'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ALL'\n",
       "2. 'B'\n",
       "3. 'CD4_T'\n",
       "4. 'CD8_T'\n",
       "5. 'Mono'\n",
       "6. 'NK'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ALL\"   \"B\"     \"CD4_T\" \"CD8_T\" \"Mono\"  \"NK\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(snp_enrichment$cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "id": "0852b938-2d5c-44f3-96f4-5a625b230003",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ALL</th><td>2.407397e-08</td><td>2.026247e+00</td><td>2.617472e+00</td><td>1.569775e+00</td><td>ALL</td><td>all_traits   </td></tr>\n",
       "\t<tr><th scope=row>ALL_no_cell_count</th><td>2.338709e-08</td><td>2.029717e+00</td><td>2.621962e+00</td><td>1.572458e+00</td><td>ALL</td><td>rm_cell_count</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tALL & 2.407397e-08 & 2.026247e+00 & 2.617472e+00 & 1.569775e+00 & ALL & all\\_traits   \\\\\n",
       "\tALL\\_no\\_cell\\_count & 2.338709e-08 & 2.029717e+00 & 2.621962e+00 & 1.572458e+00 & ALL & rm\\_cell\\_count\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 6\n",
       "\n",
       "| <!--/--> | p_value &lt;chr&gt; | estimate &lt;chr&gt; | conf_upper &lt;chr&gt; | conf_lower &lt;chr&gt; | cell_type &lt;chr&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| ALL | 2.407397e-08 | 2.026247e+00 | 2.617472e+00 | 1.569775e+00 | ALL | all_traits    |\n",
       "| ALL_no_cell_count | 2.338709e-08 | 2.029717e+00 | 2.621962e+00 | 1.572458e+00 | ALL | rm_cell_count |\n",
       "\n"
      ],
      "text/plain": [
       "                  p_value      estimate     conf_upper   conf_lower   cell_type\n",
       "ALL               2.407397e-08 2.026247e+00 2.617472e+00 1.569775e+00 ALL      \n",
       "ALL_no_cell_count 2.338709e-08 2.029717e+00 2.621962e+00 1.572458e+00 ALL      \n",
       "                  type         \n",
       "ALL               all_traits   \n",
       "ALL_no_cell_count rm_cell_count"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_enrichment,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "id": "acb44871-20b7-48c8-a110-f71246e10625",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment$p_value = as.numeric(snp_enrichment$p_value)\n",
    "snp_enrichment$estimate = as.numeric(snp_enrichment$estimate)\n",
    "snp_enrichment$conf_upper = as.numeric(snp_enrichment$conf_upper)\n",
    "snp_enrichment$conf_lower = as.numeric(snp_enrichment$conf_lower)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b2f338e-49e3-4e81-9365-c3bc9b2d4c8f",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Information about genes passing Decision Tree Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "id": "15f7ceb7-d41f-4954-8104-9a29a32c8f77",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = c( 'B', 'NK', 'Mono', 'CD8_T' , 'CD4_T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "6aeb8ee0-2733-4c8f-b196-5e1688afe97d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "id": "b92cdf6b-fd54-47cf-89e3-faaabe49d836",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_load = c('Franke_v2', 'Franke_v3', 'wijst', 'oneK1K', 'multiome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "id": "8e0446b1-6e32-4ade-b3d5-a2c8fb166735",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(j in datasets_to_load){\n",
    "    for(i in cell_types){\n",
    "        gene_list_input = fread(paste0(data_path, coeqtl_results_path , '/correlation_input/', 'F11_Decision_Tree_Geneswg3_',j,'_' , i, '.Qced.Normalized.SCs.Rds.tsv'), header = FALSE)\n",
    "        gene_list_input$cell_type = i\n",
    "        gene_list_input$dataset = j\n",
    "        gene_list = rbind(gene_list, gene_list_input)\n",
    "        }\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "id": "e1e70f38-9e7c-48fb-a86a-8e77a2cf3e7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>cell_type</th><th scope=col>dataset</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A2M    </td><td>CD4_T</td><td>multiome</td></tr>\n",
       "\t<tr><td>A2M-AS1</td><td>CD4_T</td><td>multiome</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " V1 & cell\\_type & dataset\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t A2M     & CD4\\_T & multiome\\\\\n",
       "\t A2M-AS1 & CD4\\_T & multiome\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 3\n",
       "\n",
       "| V1 &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| A2M     | CD4_T | multiome |\n",
       "| A2M-AS1 | CD4_T | multiome |\n",
       "\n"
      ],
      "text/plain": [
       "  V1      cell_type dataset \n",
       "1 A2M     CD4_T     multiome\n",
       "2 A2M-AS1 CD4_T     multiome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_list_input,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "id": "aaa68e47-e156-4bc3-a6a3-32f60bfbe5fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### unique gene list per cell-type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "id": "1f04f3f5-28e7-4cfe-bc87-c166928e7a5a",
   "metadata": {},
   "outputs": [],
   "source": [
    " gene_list= unique( gene_list[,c('cell_type', 'V1')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "id": "79aec27d-e499-4ddf-8cee-cf6a6e52ba79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>V1</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B</td><td>x         </td></tr>\n",
       "\t<tr><td>B</td><td>AC004687.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " cell\\_type & V1\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t B & x         \\\\\n",
       "\t B & AC004687.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 2\n",
       "\n",
       "| cell_type &lt;chr&gt; | V1 &lt;chr&gt; |\n",
       "|---|---|\n",
       "| B | x          |\n",
       "| B | AC004687.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type V1        \n",
       "1 B         x         \n",
       "2 B         AC004687.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head( gene_list,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "id": "5bde790c-1ab3-4f76-993f-0164d8b530a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames( gene_list) = c('cell_type', 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "id": "9b879865-7457-4d5b-90fc-fafa50dac12c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13680"
      ],
      "text/latex": [
       "13680"
      ],
      "text/markdown": [
       "13680"
      ],
      "text/plain": [
       "[1] 13680"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(gene_list$gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e92aebf-d96a-4490-a07d-1f4b56ef6f35",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "394ce449-73a6-496c-85b4-8d07096c681b",
   "metadata": {},
   "source": [
    "## Load significant co-eQTLs (from meta-analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "id": "8f7117cc-afd4-413c-a917-26656462ead2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_coeqtls =fread(paste0(sign_coeqtl_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "id": "be03633c-08d4-4144-bac7-29244cf81aef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "91628"
      ],
      "text/latex": [
       "91628"
      ],
      "text/markdown": [
       "91628"
      ],
      "text/plain": [
       "[1] 91628"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "id": "d10f67ee-9046-400e-aeae-8df70235386a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_coeqtls = as.data.frame(sign_coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "id": "fd38b970-0897-4730-84f4-9e42f9e7f116",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_coeqtls = sign_coeqtls[sign_coeqtls$type == 'unique',] ## only clearly identifiable eGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "id": "406e5c00-ec84-4ce4-8d22-c03c76c8419b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 176</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>eGene</th><th scope=col>coeGene</th><th scope=col>snp_id</th><th scope=col>cell_type</th><th scope=col>feature_id</th><th scope=col>triplet</th><th scope=col>n_samples</th><th scope=col>p_value</th><th scope=col>z_score</th><th scope=col>⋯</th><th scope=col>eQTL_remap2022_1</th><th scope=col>eQTL_remap2022_2</th><th scope=col>eQTL_remap2022_3</th><th scope=col>eQTL_HPA</th><th scope=col>eQTL_MIRNA</th><th scope=col>eQTL_TF</th><th scope=col>coeGene_is_enriched_TF</th><th scope=col>cctqtl</th><th scope=col>simulation_score</th><th scope=col>ccInteraction</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>A2M</td><td>BACH2</td><td>12:8971336:T:C</td><td>CD4_T</td><td>A2M_BACH2</td><td>12:8971336:T:C;A2M_BACH2</td><td>977</td><td>2.687871e-11</td><td>-6.662730</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td></td><td>0</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>A2M</td><td>BACH2</td><td>12:8992460:T:C</td><td>CD4_T</td><td>A2M_BACH2</td><td>12:8992460:T:C;A2M_BACH2</td><td>977</td><td>1.680350e-10</td><td>-6.388031</td><td>⋯</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td>NA</td><td></td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 176\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & V1 & eGene & coeGene & snp\\_id & cell\\_type & feature\\_id & triplet & n\\_samples & p\\_value & z\\_score & ⋯ & eQTL\\_remap2022\\_1 & eQTL\\_remap2022\\_2 & eQTL\\_remap2022\\_3 & eQTL\\_HPA & eQTL\\_MIRNA & eQTL\\_TF & coeGene\\_is\\_enriched\\_TF & cctqtl & simulation\\_score & ccInteraction\\\\\n",
       "  & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & ⋯ & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <int> & <chr> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & A2M & BACH2 & 12:8971336:T:C & CD4\\_T & A2M\\_BACH2 & 12:8971336:T:C;A2M\\_BACH2 & 977 & 2.687871e-11 & -6.662730 & ⋯ & NA & NA & NA & NA & NA & NA & NA &  & 0 & 0\\\\\n",
       "\t2 & 2 & A2M & BACH2 & 12:8992460:T:C & CD4\\_T & A2M\\_BACH2 & 12:8992460:T:C;A2M\\_BACH2 & 977 & 1.680350e-10 & -6.388031 & ⋯ & NA & NA & NA & NA & NA & NA & NA &  & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 176\n",
       "\n",
       "| <!--/--> | V1 &lt;int&gt; | eGene &lt;chr&gt; | coeGene &lt;chr&gt; | snp_id &lt;chr&gt; | cell_type &lt;chr&gt; | feature_id &lt;chr&gt; | triplet &lt;chr&gt; | n_samples &lt;int&gt; | p_value &lt;dbl&gt; | z_score &lt;dbl&gt; | ⋯ ⋯ | eQTL_remap2022_1 &lt;lgl&gt; | eQTL_remap2022_2 &lt;lgl&gt; | eQTL_remap2022_3 &lt;lgl&gt; | eQTL_HPA &lt;lgl&gt; | eQTL_MIRNA &lt;lgl&gt; | eQTL_TF &lt;lgl&gt; | coeGene_is_enriched_TF &lt;int&gt; | cctqtl &lt;chr&gt; | simulation_score &lt;dbl&gt; | ccInteraction &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | A2M | BACH2 | 12:8971336:T:C | CD4_T | A2M_BACH2 | 12:8971336:T:C;A2M_BACH2 | 977 | 2.687871e-11 | -6.662730 | ⋯ | NA | NA | NA | NA | NA | NA | NA | <!----> | 0 | 0 |\n",
       "| 2 | 2 | A2M | BACH2 | 12:8992460:T:C | CD4_T | A2M_BACH2 | 12:8992460:T:C;A2M_BACH2 | 977 | 1.680350e-10 | -6.388031 | ⋯ | NA | NA | NA | NA | NA | NA | NA | <!----> | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 eGene coeGene snp_id         cell_type feature_id triplet                 \n",
       "1 1  A2M   BACH2   12:8971336:T:C CD4_T     A2M_BACH2  12:8971336:T:C;A2M_BACH2\n",
       "2 2  A2M   BACH2   12:8992460:T:C CD4_T     A2M_BACH2  12:8992460:T:C;A2M_BACH2\n",
       "  n_samples p_value      z_score   ⋯ eQTL_remap2022_1 eQTL_remap2022_2\n",
       "1 977       2.687871e-11 -6.662730 ⋯ NA               NA              \n",
       "2 977       1.680350e-10 -6.388031 ⋯ NA               NA              \n",
       "  eQTL_remap2022_3 eQTL_HPA eQTL_MIRNA eQTL_TF coeGene_is_enriched_TF cctqtl\n",
       "1 NA               NA       NA         NA      NA                           \n",
       "2 NA               NA       NA         NA      NA                           \n",
       "  simulation_score ccInteraction\n",
       "1 0                0            \n",
       "2 0                0            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sign_coeqtls,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "e49d310f-3f74-43a3-8166-22e68dc02a42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "301"
      ],
      "text/latex": [
       "301"
      ],
      "text/markdown": [
       "301"
      ],
      "text/plain": [
       "[1] 301"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "id": "f4b1671b-2d9d-41d9-9c9b-33a0c2e6261b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "168"
      ],
      "text/latex": [
       "168"
      ],
      "text/markdown": [
       "168"
      ],
      "text/plain": [
       "[1] 168"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene[sign_coeqtls$cell_type == 'CD4_T']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "id": "bcbff887-46a7-4fc4-895f-0737b7d3b3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extract the position from the SNP-ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "75d3a360-132e-4f33-a50e-4f93097d2173",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_coeqtls$pos = str_extract(sign_coeqtls$snp_id, ':\\\\d.*\\\\d:')\n",
    "sign_coeqtls$pos = str_replace_all(sign_coeqtls$pos, ':', '')\n",
    "#sign_coeqtls$pos = str_replace(sign_coeqtls$pos, '.*:', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "id": "bb4578f1-7210-4d17-ba3b-597b3173bf78",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on only eSNPs (remove co-eQTLs caused by added GWAS SNPs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "id": "e91cc40a-17f2-41bc-b985-e768d6d6d9da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "only_esnps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "id": "1ccab610-fa0c-4b6b-a311-7cc475b8d524",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(only_esnps == TRUE){\n",
    "    sign_coeqtls = sign_coeqtls[sign_coeqtls$eqtl_snp == 1,]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "id": "60ea0e70-c0aa-4078-8678-096c1bd544f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "19338"
      ],
      "text/latex": [
       "19338"
      ],
      "text/markdown": [
       "19338"
      ],
      "text/plain": [
       "[1] 19338"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "id": "65aeb0e3-38c6-4aed-a46c-debade4ea933",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "280"
      ],
      "text/latex": [
       "280"
      ],
      "text/markdown": [
       "280"
      ],
      "text/plain": [
       "[1] 280"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "id": "b0239ce7-f6bf-45be-a850-bd5b21a85c99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "157"
      ],
      "text/latex": [
       "157"
      ],
      "text/markdown": [
       "157"
      ],
      "text/plain": [
       "[1] 157"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene[sign_coeqtls$cell_type == 'CD4_T']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "id": "d30e24ae-5704-4a9f-ab08-9b5ed14ca4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter: remove HLA- genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "id": "51010b1e-ceb4-4eda-8ba8-5819a5bf1b84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "remove_HLA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "id": "63cce539-b725-438b-9803-9ebe4848905c",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(remove_HLA == TRUE){\n",
    "    sign_coeqtls = sign_coeqtls[is.na(str_extract(sign_coeqtls$eGene, 'HLA')),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "id": "371d4bd3-2adb-4be5-be27-21bcdd68dd44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "18663"
      ],
      "text/latex": [
       "18663"
      ],
      "text/markdown": [
       "18663"
      ],
      "text/plain": [
       "[1] 18663"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_coeqtls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 394,
   "id": "73082de5-55ab-4847-a4c4-2162f1d0d779",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "273"
      ],
      "text/latex": [
       "273"
      ],
      "text/markdown": [
       "273"
      ],
      "text/plain": [
       "[1] 273"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "id": "e8e99e88-e2d6-4592-8c40-9a6607dffc0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "152"
      ],
      "text/latex": [
       "152"
      ],
      "text/markdown": [
       "152"
      ],
      "text/plain": [
       "[1] 152"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene[sign_coeqtls$cell_type == 'CD4_T']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "id": "8fd2a2a3-1c89-48f2-9880-f3e5bc5330c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate amount of significant eGenes per cell-type (after filters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "id": "00f3f124-3610-42b3-8ad4-857a2984e482",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_egenes = sign_coeqtls %>% group_by(cell_type) %>% dplyr::summarise(amount_coeqtl_egenes  = length(unique(eGene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "id": "7e1635a6-20ea-4910-af68-66e355a26b3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>amount_coeqtl_egenes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td> 32</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>152</td></tr>\n",
       "\t<tr><td>CD8_T</td><td> 76</td></tr>\n",
       "\t<tr><td>Mono </td><td> 50</td></tr>\n",
       "\t<tr><td>NK   </td><td> 40</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " cell\\_type & amount\\_coeqtl\\_egenes\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t B     &  32\\\\\n",
       "\t CD4\\_T & 152\\\\\n",
       "\t CD8\\_T &  76\\\\\n",
       "\t Mono  &  50\\\\\n",
       "\t NK    &  40\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 5 × 2\n",
       "\n",
       "| cell_type &lt;chr&gt; | amount_coeqtl_egenes &lt;int&gt; |\n",
       "|---|---|\n",
       "| B     |  32 |\n",
       "| CD4_T | 152 |\n",
       "| CD8_T |  76 |\n",
       "| Mono  |  50 |\n",
       "| NK    |  40 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type amount_coeqtl_egenes\n",
       "1 B          32                 \n",
       "2 CD4_T     152                 \n",
       "3 CD8_T      76                 \n",
       "4 Mono       50                 \n",
       "5 NK         40                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coeqtl_egenes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc4fe58a-7f35-4763-9a04-f51b27587cfd",
   "metadata": {},
   "source": [
    "## sceQTLGen: eGene- SNP mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "id": "8faa5f39-fd64-4b29-8139-fd98a4fcb92d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get all eQTLs and SNPs from sc-eQTL Gen that were tested in the co-eQTL mapping as backgroun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "id": "f48c6858-1136-4f2e-841b-227e37a009af",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_gene_mapping_all = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "id": "6d94a22c-c733-4575-a288-c0524c94ffe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = c('B', 'CD4_T', 'CD8_T', 'NK', 'Mono')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "id": "1a136ad2-57d6-41fb-9f7e-3054d8972dd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-02-12 14:31:31 CET\"\n",
      "[1] \"2024-12-11 09:52:00 CET\"\n",
      "[1] \"2024-12-11 09:51:32 CET\"\n",
      "[1] \"2024-12-11 09:51:16 CET\"\n",
      "[1] \"2025-02-12 14:30:19 CET\"\n"
     ]
    }
   ],
   "source": [
    "for(i in cell_types){\n",
    "    print(file.info(paste0(data_path_eqtl, 'sc_eqtlgen_final/'  , i, '_snp_egene_mapping.csv'))$ctime)\n",
    "    snp_gene_mapping = read.csv(paste0(data_path_eqtl, 'sc_eqtlgen_final/'  , i, '_snp_egene_mapping.csv'))\n",
    "    snp_gene_mapping$cell_type = i\n",
    "    \n",
    "    snp_gene_mapping_all = rbind(snp_gene_mapping, snp_gene_mapping_all)\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5e6a0c8-7116-4ab4-a8f2-fa1f1a185d4f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "id": "493ea5a1-1ea3-4f41-b2f5-81ca0dcc97fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>amount_egenes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td>2017</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>4988</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>2392</td></tr>\n",
       "\t<tr><td>Mono </td><td>2029</td></tr>\n",
       "\t<tr><td>NK   </td><td>2170</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " cell\\_type & amount\\_egenes\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t B     & 2017\\\\\n",
       "\t CD4\\_T & 4988\\\\\n",
       "\t CD8\\_T & 2392\\\\\n",
       "\t Mono  & 2029\\\\\n",
       "\t NK    & 2170\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 5 × 2\n",
       "\n",
       "| cell_type &lt;chr&gt; | amount_egenes &lt;int&gt; |\n",
       "|---|---|\n",
       "| B     | 2017 |\n",
       "| CD4_T | 4988 |\n",
       "| CD8_T | 2392 |\n",
       "| Mono  | 2029 |\n",
       "| NK    | 2170 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type amount_egenes\n",
       "1 B         2017         \n",
       "2 CD4_T     4988         \n",
       "3 CD8_T     2392         \n",
       "4 Mono      2029         \n",
       "5 NK        2170         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "snp_gene_mapping_all %>% group_by(cell_type) %>% summarise(amount_egenes = length(unique(feature_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 404,
   "id": "5032fa23-e5b3-4522-bed0-10b713812586",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>feature_id</th><th scope=col>snp_id</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>p_value</th><th scope=col>beta</th><th scope=col>cell_type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>CD52</td><td>1:26306290:C:T</td><td>26317957</td><td>26320523</td><td>1.903053e-214</td><td>-1.0420481</td><td>Mono</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>CD52</td><td>1:26358699:C:T</td><td>26317957</td><td>26320523</td><td> 2.438329e-19</td><td> 0.3022236</td><td>Mono</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & X & feature\\_id & snp\\_id & feature\\_start & feature\\_end & p\\_value & beta & cell\\_type\\\\\n",
       "  & <int> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & CD52 & 1:26306290:C:T & 26317957 & 26320523 & 1.903053e-214 & -1.0420481 & Mono\\\\\n",
       "\t2 & 2 & CD52 & 1:26358699:C:T & 26317957 & 26320523 &  2.438329e-19 &  0.3022236 & Mono\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 8\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | feature_id &lt;chr&gt; | snp_id &lt;chr&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | p_value &lt;dbl&gt; | beta &lt;dbl&gt; | cell_type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | CD52 | 1:26306290:C:T | 26317957 | 26320523 | 1.903053e-214 | -1.0420481 | Mono |\n",
       "| 2 | 2 | CD52 | 1:26358699:C:T | 26317957 | 26320523 |  2.438329e-19 |  0.3022236 | Mono |\n",
       "\n"
      ],
      "text/plain": [
       "  X feature_id snp_id         feature_start feature_end p_value      \n",
       "1 1 CD52       1:26306290:C:T 26317957      26320523    1.903053e-214\n",
       "2 2 CD52       1:26358699:C:T 26317957      26320523     2.438329e-19\n",
       "  beta       cell_type\n",
       "1 -1.0420481 Mono     \n",
       "2  0.3022236 Mono     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping_all,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "id": "9fb6a945-add3-4815-9356-e57e4ba6fb6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "14767"
      ],
      "text/latex": [
       "14767"
      ],
      "text/markdown": [
       "14767"
      ],
      "text/plain": [
       "[1] 14767"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(unique(snp_gene_mapping_all[, c('feature_id', 'snp_id')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "id": "1bb9df1e-535b-476f-a900-430cd82cd1a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "13744"
      ],
      "text/latex": [
       "13744"
      ],
      "text/markdown": [
       "13744"
      ],
      "text/plain": [
       "[1] 13744"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(snp_gene_mapping_all$snp_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "id": "8f00a6af-9a00-4861-8bfd-9acf737ea646",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "14767"
      ],
      "text/latex": [
       "14767"
      ],
      "text/markdown": [
       "14767"
      ],
      "text/plain": [
       "[1] 14767"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(unique(snp_gene_mapping_all[,c('feature_id', 'snp_id')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "id": "fb2a4a63-04df-4869-9826-3e9b2de55c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on only genes passing decision tree filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "id": "459d3c72-af37-4c52-9bd1-dfa126629013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>gene</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B</td><td>x         </td></tr>\n",
       "\t<tr><td>B</td><td>AC004687.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " cell\\_type & gene\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t B & x         \\\\\n",
       "\t B & AC004687.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 2\n",
       "\n",
       "| cell_type &lt;chr&gt; | gene &lt;chr&gt; |\n",
       "|---|---|\n",
       "| B | x          |\n",
       "| B | AC004687.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type gene      \n",
       "1 B         x         \n",
       "2 B         AC004687.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_list,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "id": "5d46196f-af09-4958-945d-a85a89d4cd2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "only_dt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "id": "f91bdeb7-70fa-4944-b71d-ffdd8be7094d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(only_dt == TRUE){\n",
    "snp_gene_mapping_all = merge(snp_gene_mapping_all, gene_list, by.x = c('feature_id', 'cell_type'),\n",
    "                             by.y = c('gene', 'cell_type'))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 412,
   "id": "67d1de10-505f-4993-a325-34165c1c0fd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5309"
      ],
      "text/latex": [
       "5309"
      ],
      "text/markdown": [
       "5309"
      ],
      "text/plain": [
       "[1] 5309"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(snp_gene_mapping_all$feature_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "id": "8bda941e-f7aa-4d01-afb2-81623140e8ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "12199"
      ],
      "text/latex": [
       "12199"
      ],
      "text/markdown": [
       "12199"
      ],
      "text/plain": [
       "[1] 12199"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(unique(snp_gene_mapping_all[,c('feature_id', 'snp_id')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "id": "a04d9e1c-d564-40f0-b3f4-f41a6b91f1a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter out HLA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "id": "a88251e7-3945-4c4a-88a8-c26721d6d4af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>feature_id</th><th scope=col>cell_type</th><th scope=col>X</th><th scope=col>snp_id</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>p_value</th><th scope=col>beta</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG</td><td>B    </td><td>1348</td><td>19:58406208:G:A</td><td>58345178</td><td>58353499</td><td>4.333125e-07</td><td>-0.1954896</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A1BG</td><td>CD4_T</td><td>3178</td><td>19:58497199:G:A</td><td>58345178</td><td>58353499</td><td>2.302583e-13</td><td>-0.2558016</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & feature\\_id & cell\\_type & X & snp\\_id & feature\\_start & feature\\_end & p\\_value & beta\\\\\n",
       "  & <chr> & <chr> & <int> & <chr> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG & B     & 1348 & 19:58406208:G:A & 58345178 & 58353499 & 4.333125e-07 & -0.1954896\\\\\n",
       "\t2 & A1BG & CD4\\_T & 3178 & 19:58497199:G:A & 58345178 & 58353499 & 2.302583e-13 & -0.2558016\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 8\n",
       "\n",
       "| <!--/--> | feature_id &lt;chr&gt; | cell_type &lt;chr&gt; | X &lt;int&gt; | snp_id &lt;chr&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | p_value &lt;dbl&gt; | beta &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG | B     | 1348 | 19:58406208:G:A | 58345178 | 58353499 | 4.333125e-07 | -0.1954896 |\n",
       "| 2 | A1BG | CD4_T | 3178 | 19:58497199:G:A | 58345178 | 58353499 | 2.302583e-13 | -0.2558016 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id cell_type X    snp_id          feature_start feature_end\n",
       "1 A1BG       B         1348 19:58406208:G:A 58345178      58353499   \n",
       "2 A1BG       CD4_T     3178 19:58497199:G:A 58345178      58353499   \n",
       "  p_value      beta      \n",
       "1 4.333125e-07 -0.1954896\n",
       "2 2.302583e-13 -0.2558016"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping_all,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "id": "375376cf-7691-4b3d-abd3-1a0aafe45b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(remove_HLA == TRUE){\n",
    "snp_gene_mapping_all =  snp_gene_mapping_all[is.na(str_extract(snp_gene_mapping_all$feature_id, 'HLA')),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "id": "945c0522-7696-432b-963b-181b1c27ec5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5293"
      ],
      "text/latex": [
       "5293"
      ],
      "text/markdown": [
       "5293"
      ],
      "text/plain": [
       "[1] 5293"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(snp_gene_mapping_all$feature_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "id": "853fa20b-4be4-4d38-ba1e-43fdc46e1fe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### OPT: filter on only eGenes with CS (only for testing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "id": "857ffd78-81b5-4f18-819e-46ea5abae26f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "FALSE"
      ],
      "text/latex": [
       "FALSE"
      ],
      "text/markdown": [
       "FALSE"
      ],
      "text/plain": [
       "[1] FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cs_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "439e2cc4-e3a3-4fee-a386-4aed73d194c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### CS mapping indicating whether an eGene has a CS or not\n",
    "cs_mapping = fread( paste0('/ictstr01/groups/epigenereg01/workspace/projects/grn_dev_groningen/eqtl_mapping/sc_eqtlgen_final/Finemapping/', 'Top_CS_per_eGene.csv'), header = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "id": "8709dc2a-0fbf-4ab3-a652-d47544691b33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>feature_id</th><th scope=col>cell_type</th><th scope=col>SusieRss_CS</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>A1BG</td><td>B    </td><td>-Inf</td></tr>\n",
       "\t<tr><td>2</td><td>A1BG</td><td>CD4_T</td><td>   3</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 4\n",
       "\\begin{tabular}{llll}\n",
       " V1 & feature\\_id & cell\\_type & SusieRss\\_CS\\\\\n",
       " <int> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1 & A1BG & B     & -Inf\\\\\n",
       "\t 2 & A1BG & CD4\\_T &    3\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 4\n",
       "\n",
       "| V1 &lt;int&gt; | feature_id &lt;chr&gt; | cell_type &lt;chr&gt; | SusieRss_CS &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | A1BG | B     | -Inf |\n",
       "| 2 | A1BG | CD4_T |    3 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 feature_id cell_type SusieRss_CS\n",
       "1 1  A1BG       B         -Inf       \n",
       "2 2  A1BG       CD4_T        3       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cs_mapping,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "id": "fb051de0-f964-448f-ab30-23525f53fa6b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>-Inf</li><li>3</li><li>1</li><li>4</li><li>2</li><li>5</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item -Inf\n",
       "\\item 3\n",
       "\\item 1\n",
       "\\item 4\n",
       "\\item 2\n",
       "\\item 5\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. -Inf\n",
       "2. 3\n",
       "3. 1\n",
       "4. 4\n",
       "5. 2\n",
       "6. 5\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] -Inf    3    1    4    2    5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(cs_mapping$SusieRss_CS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "936d39eb-64d5-429a-88de-10652017b7f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "id": "2f929a1b-97a9-4815-b23b-33a1d5ace4ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(cs_snps == TRUE){\n",
    "    print(nrow(snp_gene_mapping_all))\n",
    "    cs_mapping = cs_mapping[!is.infinite(cs_mapping$SusieRss_CS),]\n",
    "    snp_gene_mapping_all = merge(snp_gene_mapping_all, cs_mapping, by = c('feature_id', 'cell_type'))\n",
    "    print(nrow(snp_gene_mapping_all))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3f0f6e-a85f-485c-ae86-08677b58a74f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b83ccb1-1b83-446a-a354-8d1a24ee74a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "id": "67a91256-1de7-422d-998a-74a25b4f1fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the amount of eGenes per cell-type "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "id": "6d1b2146-8c51-43c6-bb08-bec225b60d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "egenes_cell_type = snp_gene_mapping_all %>% group_by(cell_type) %>% summarise(amount_egenes = length(unique(feature_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "id": "5aa5e534-54e7-449a-b1aa-a575b8e95de3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>amount_egenes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td>1679</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>3676</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>2019</td></tr>\n",
       "\t<tr><td>Mono </td><td>1748</td></tr>\n",
       "\t<tr><td>NK   </td><td>1914</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " cell\\_type & amount\\_egenes\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t B     & 1679\\\\\n",
       "\t CD4\\_T & 3676\\\\\n",
       "\t CD8\\_T & 2019\\\\\n",
       "\t Mono  & 1748\\\\\n",
       "\t NK    & 1914\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 5 × 2\n",
       "\n",
       "| cell_type &lt;chr&gt; | amount_egenes &lt;int&gt; |\n",
       "|---|---|\n",
       "| B     | 1679 |\n",
       "| CD4_T | 3676 |\n",
       "| CD8_T | 2019 |\n",
       "| Mono  | 1748 |\n",
       "| NK    | 1914 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type amount_egenes\n",
       "1 B         1679         \n",
       "2 CD4_T     3676         \n",
       "3 CD8_T     2019         \n",
       "4 Mono      1748         \n",
       "5 NK        1914         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "egenes_cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "id": "4787d437-6d02-4a28-9aca-30e38802bb0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the amount of unique eGenes across all cell-types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "id": "8fc41817-d93d-4a23-a6e2-07a83b4f2013",
   "metadata": {},
   "outputs": [],
   "source": [
    "egenes_amount = snp_gene_mapping_all  %>% summarise(amount_egenes = length(unique(feature_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "id": "ed6c3867-410f-4722-b2cf-6f749f61039c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 1 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>amount_egenes</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>5293</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 1 × 1\n",
       "\\begin{tabular}{l}\n",
       " amount\\_egenes\\\\\n",
       " <int>\\\\\n",
       "\\hline\n",
       "\t 5293\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 1 × 1\n",
       "\n",
       "| amount_egenes &lt;int&gt; |\n",
       "|---|\n",
       "| 5293 |\n",
       "\n"
      ],
      "text/plain": [
       "  amount_egenes\n",
       "1 5293         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "egenes_amount "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "id": "14539c30-f08f-4ba6-868a-099c03facaae",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Select alternative for doing across cell-type analysis (how to select an effect when it appears in multiple cell-types for the eGene)\n",
    "alt = 'p_value' # ALTERNATIVES: minimum p-value (p_value); random (random)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 431,
   "id": "b984d708-498f-46f6-b564-4279736cc959",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>feature_id</th><th scope=col>cell_type</th><th scope=col>X</th><th scope=col>snp_id</th><th scope=col>feature_start</th><th scope=col>feature_end</th><th scope=col>p_value</th><th scope=col>beta</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A1BG</td><td>B    </td><td>1348</td><td>19:58406208:G:A</td><td>58345178</td><td>58353499</td><td>4.333125e-07</td><td>-0.1954896</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A1BG</td><td>CD4_T</td><td>3178</td><td>19:58497199:G:A</td><td>58345178</td><td>58353499</td><td>2.302583e-13</td><td>-0.2558016</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & feature\\_id & cell\\_type & X & snp\\_id & feature\\_start & feature\\_end & p\\_value & beta\\\\\n",
       "  & <chr> & <chr> & <int> & <chr> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A1BG & B     & 1348 & 19:58406208:G:A & 58345178 & 58353499 & 4.333125e-07 & -0.1954896\\\\\n",
       "\t2 & A1BG & CD4\\_T & 3178 & 19:58497199:G:A & 58345178 & 58353499 & 2.302583e-13 & -0.2558016\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 8\n",
       "\n",
       "| <!--/--> | feature_id &lt;chr&gt; | cell_type &lt;chr&gt; | X &lt;int&gt; | snp_id &lt;chr&gt; | feature_start &lt;int&gt; | feature_end &lt;int&gt; | p_value &lt;dbl&gt; | beta &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG | B     | 1348 | 19:58406208:G:A | 58345178 | 58353499 | 4.333125e-07 | -0.1954896 |\n",
       "| 2 | A1BG | CD4_T | 3178 | 19:58497199:G:A | 58345178 | 58353499 | 2.302583e-13 | -0.2558016 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id cell_type X    snp_id          feature_start feature_end\n",
       "1 A1BG       B         1348 19:58406208:G:A 58345178      58353499   \n",
       "2 A1BG       CD4_T     3178 19:58497199:G:A 58345178      58353499   \n",
       "  p_value      beta      \n",
       "1 4.333125e-07 -0.1954896\n",
       "2 2.302583e-13 -0.2558016"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping_all,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "id": "e2ae09e0-b230-4ff0-8de4-569bde5b9b97",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(alt == 'random'){\n",
    "    set.seed(10)\n",
    "    snp_gene_mapping_all_subset = unique(snp_gene_mapping_all[,c('feature_id', 'cell_type')]) %>% group_by(feature_id) %>% slice_sample(n = 1) \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "id": "605abaa2-08b2-4a1a-9346-47788a2161ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(alt == 'p_value'){\n",
    "    set.seed(1)\n",
    "    snp_gene_mapping_all_subset = unique(snp_gene_mapping_all[,c('feature_id', 'cell_type', 'p_value')]) %>% group_by(feature_id) %>% dplyr::filter(p_value == min(p_value))  %>% slice_sample(n = 1) \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "id": "5241caad-bcf6-408f-b740-33f58567f8e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5293"
      ],
      "text/latex": [
       "5293"
      ],
      "text/markdown": [
       "5293"
      ],
      "text/plain": [
       "[1] 5293"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_gene_mapping_all_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "id": "db335c4c-88bc-4581-ac38-00c0e00ce1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get the total amount of coeqtls across all cell-types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "id": "bac8a8be-84ea-4e63-b90e-37acf8f1aa6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "amount_coeqtls = length(unique(sign_coeqtls$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "id": "ed568b04-f343-4cca-b8af-418997272f53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "273"
      ],
      "text/latex": [
       "273"
      ],
      "text/markdown": [
       "273"
      ],
      "text/plain": [
       "[1] 273"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "amount_coeqtls "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "id": "0abc72b6-912a-418a-b4d2-b462b4e5d1ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "## For each eGene summarize in how many cell-types it is an eGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "id": "6e3cd2c5-2aa8-4499-b2d2-69d592523d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types_per_egene =  snp_gene_mapping_all %>% group_by(feature_id) %>% summarise(amount_cell_types = length(unique(cell_type)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "id": "505b8596-b1a2-414e-be10-3e99e0047f2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>amount_cell_types</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ABRACL</td><td>5</td></tr>\n",
       "\t<tr><td>ACYP2 </td><td>5</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " feature\\_id & amount\\_cell\\_types\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t ABRACL & 5\\\\\n",
       "\t ACYP2  & 5\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| feature_id &lt;chr&gt; | amount_cell_types &lt;int&gt; |\n",
       "|---|---|\n",
       "| ABRACL | 5 |\n",
       "| ACYP2  | 5 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id amount_cell_types\n",
       "1 ABRACL     5                \n",
       "2 ACYP2      5                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_types_per_egene  %>% arrange(-amount_cell_types),2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd116338-cd42-4c89-b966-638bf6255be2",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load GWAS coloc results (from OTAR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "id": "ca5fe0aa-341a-4490-be5b-2833d9e3a92a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### All COLOC Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "id": "e769205a-e8f8-424e-8f68-b0111d3e070f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-06-11 12:02:01 CEST\"\n"
     ]
    }
   ],
   "source": [
    "#coloc_all = fread(paste0(data_path_eqtl, 'sc_eqtlgen_final/GWAS_coloc/', 'coloc_w_study_h4_0.8.txt'))\n",
    "\n",
    "#coloc_all = fread(paste0(data_path_eqtl, 'sc_eqtlgen_final/GWAS_coloc/', 'coloc_w_study.txt'))  # OpenTarget complete file\n",
    "path = paste0(data_path_eqtl, 'sc_eqtlgen_final/GWAS_coloc/', 'FDR0.05_annotated_coloc_w_study_080425.txt')\n",
    "coloc_all= fread(path)  # alternative file filtered by MJ for CS in his finemapping\n",
    "print(file.info(path)$ctime)\n",
    "\n",
    "## old file: FDR0.05_annotated_coloc_w_study.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "id": "68f05e34-08e3-472f-af13-091c1f5519ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>eqtlTraitFromSource</th><th scope=col>snp_id</th><th scope=col>eQTL_CT</th><th scope=col>V1.1</th><th scope=col>leftStudyLocusId</th><th scope=col>leftStudyId</th><th scope=col>gwasTraitFromSource</th><th scope=col>gwasEFO</th><th scope=col>rightStudyLocusId</th><th scope=col>⋯</th><th scope=col>h1</th><th scope=col>h2</th><th scope=col>h3</th><th scope=col>h4</th><th scope=col>h3h4</th><th scope=col>H4/H3</th><th scope=col>Alt_ct_gene_cs_id</th><th scope=col>Observed</th><th scope=col>gwasStudyId</th><th scope=col>SusieRss_CS</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>A1BG</td><td>19:57346753:C:T</td><td>B</td><td>1</td><td>bea15e5399c770d864f175c159b892ba</td><td>GCST004621  </td><td>Red cell distribution width</td><td>['EFO_0009188']</td><td>84ed1e402f40b5bcd7217236864923eb</td><td>⋯</td><td>0.8442357</td><td>2.199654e-07</td><td>0.07133792</td><td>0.08442357</td><td>0.1557615</td><td>1.183432</td><td>B:A1BG:19_57346753_C_T</td><td>FALSE</td><td>GCST004621  </td><td>NA</td></tr>\n",
       "\t<tr><td>2</td><td>A1BG</td><td>19:57346753:C:T</td><td>B</td><td>2</td><td>1bf2bcb5ce6934553fcd44f7f18d1c45</td><td>GCST90002396</td><td>Mean reticulocyte volume   </td><td>['EFO_0010701']</td><td>84ed1e402f40b5bcd7217236864923eb</td><td>⋯</td><td>0.8444519</td><td>3.674213e-10</td><td>0.07110285</td><td>0.08444519</td><td>0.1555480</td><td>1.187648</td><td>B:A1BG:19_57346753_C_T</td><td>FALSE</td><td>GCST90002396</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 31\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " V1 & eqtlTraitFromSource & snp\\_id & eQTL\\_CT & V1.1 & leftStudyLocusId & leftStudyId & gwasTraitFromSource & gwasEFO & rightStudyLocusId & ⋯ & h1 & h2 & h3 & h4 & h3h4 & H4/H3 & Alt\\_ct\\_gene\\_cs\\_id & Observed & gwasStudyId & SusieRss\\_CS\\\\\n",
       " <int> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <lgl> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t 1 & A1BG & 19:57346753:C:T & B & 1 & bea15e5399c770d864f175c159b892ba & GCST004621   & Red cell distribution width & {[}'EFO\\_0009188'{]} & 84ed1e402f40b5bcd7217236864923eb & ⋯ & 0.8442357 & 2.199654e-07 & 0.07133792 & 0.08442357 & 0.1557615 & 1.183432 & B:A1BG:19\\_57346753\\_C\\_T & FALSE & GCST004621   & NA\\\\\n",
       "\t 2 & A1BG & 19:57346753:C:T & B & 2 & 1bf2bcb5ce6934553fcd44f7f18d1c45 & GCST90002396 & Mean reticulocyte volume    & {[}'EFO\\_0010701'{]} & 84ed1e402f40b5bcd7217236864923eb & ⋯ & 0.8444519 & 3.674213e-10 & 0.07110285 & 0.08444519 & 0.1555480 & 1.187648 & B:A1BG:19\\_57346753\\_C\\_T & FALSE & GCST90002396 & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 31\n",
       "\n",
       "| V1 &lt;int&gt; | eqtlTraitFromSource &lt;chr&gt; | snp_id &lt;chr&gt; | eQTL_CT &lt;chr&gt; | V1.1 &lt;int&gt; | leftStudyLocusId &lt;chr&gt; | leftStudyId &lt;chr&gt; | gwasTraitFromSource &lt;chr&gt; | gwasEFO &lt;chr&gt; | rightStudyLocusId &lt;chr&gt; | ⋯ ⋯ | h1 &lt;dbl&gt; | h2 &lt;dbl&gt; | h3 &lt;dbl&gt; | h4 &lt;dbl&gt; | h3h4 &lt;dbl&gt; | H4/H3 &lt;dbl&gt; | Alt_ct_gene_cs_id &lt;chr&gt; | Observed &lt;lgl&gt; | gwasStudyId &lt;chr&gt; | SusieRss_CS &lt;int&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A1BG | 19:57346753:C:T | B | 1 | bea15e5399c770d864f175c159b892ba | GCST004621   | Red cell distribution width | ['EFO_0009188'] | 84ed1e402f40b5bcd7217236864923eb | ⋯ | 0.8442357 | 2.199654e-07 | 0.07133792 | 0.08442357 | 0.1557615 | 1.183432 | B:A1BG:19_57346753_C_T | FALSE | GCST004621   | NA |\n",
       "| 2 | A1BG | 19:57346753:C:T | B | 2 | 1bf2bcb5ce6934553fcd44f7f18d1c45 | GCST90002396 | Mean reticulocyte volume    | ['EFO_0010701'] | 84ed1e402f40b5bcd7217236864923eb | ⋯ | 0.8444519 | 3.674213e-10 | 0.07110285 | 0.08444519 | 0.1555480 | 1.187648 | B:A1BG:19_57346753_C_T | FALSE | GCST90002396 | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 eqtlTraitFromSource snp_id          eQTL_CT V1.1\n",
       "1 1  A1BG                19:57346753:C:T B       1   \n",
       "2 2  A1BG                19:57346753:C:T B       2   \n",
       "  leftStudyLocusId                 leftStudyId  gwasTraitFromSource        \n",
       "1 bea15e5399c770d864f175c159b892ba GCST004621   Red cell distribution width\n",
       "2 1bf2bcb5ce6934553fcd44f7f18d1c45 GCST90002396 Mean reticulocyte volume   \n",
       "  gwasEFO         rightStudyLocusId                ⋯ h1        h2          \n",
       "1 ['EFO_0009188'] 84ed1e402f40b5bcd7217236864923eb ⋯ 0.8442357 2.199654e-07\n",
       "2 ['EFO_0010701'] 84ed1e402f40b5bcd7217236864923eb ⋯ 0.8444519 3.674213e-10\n",
       "  h3         h4         h3h4      H4/H3    Alt_ct_gene_cs_id      Observed\n",
       "1 0.07133792 0.08442357 0.1557615 1.183432 B:A1BG:19_57346753_C_T FALSE   \n",
       "2 0.07110285 0.08444519 0.1555480 1.187648 B:A1BG:19_57346753_C_T FALSE   \n",
       "  gwasStudyId  SusieRss_CS\n",
       "1 GCST004621   NA         \n",
       "2 GCST90002396 NA         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coloc_all,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "id": "77ee77b1-810f-42dc-973f-1b7a6f8e57c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2089899"
      ],
      "text/latex": [
       "2089899"
      ],
      "text/markdown": [
       "2089899"
      ],
      "text/plain": [
       "[1] 2089899"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coloc_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "id": "be9732a1-9a3d-48cd-936f-eb886707f0f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter out GWAS results that are not based on finemapped SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "id": "9f089635-8c68-4bb5-bfd8-63fb2dcf851b",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all = coloc_all[coloc_all$Observed == TRUE,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "id": "aa26cde8-6b1e-4cbe-b99a-b6ec1f6b88e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "86292"
      ],
      "text/latex": [
       "86292"
      ],
      "text/markdown": [
       "86292"
      ],
      "text/plain": [
       "[1] 86292"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coloc_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "id": "8d1f45e0-7f0e-4def-926c-89d83c0ff341",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all = coloc_all[coloc_all$sceqtl_log10BF >0.8685  ,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "id": "021e0d8c-807f-4e36-bdf0-94c76e82484e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "83339"
      ],
      "text/latex": [
       "83339"
      ],
      "text/markdown": [
       "83339"
      ],
      "text/plain": [
       "[1] 83339"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coloc_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "id": "972cb0f0-ee19-4279-8887-4fd903d530f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add gwasStudyId (new data format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "id": "ca12ea8b-c577-4c1c-aaef-ffc5a81a6e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$gwasStudyId = coloc_all$leftStudyId"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "id": "af9d9a51-f9f9-421a-8492-5861ee55f91b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5390"
      ],
      "text/latex": [
       "5390"
      ],
      "text/markdown": [
       "5390"
      ],
      "text/plain": [
       "[1] 5390"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coloc_all$gwasStudyId))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "id": "20906c13-1746-4c51-85fe-5ce36dfa95c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4251"
      ],
      "text/latex": [
       "4251"
      ],
      "text/markdown": [
       "4251"
      ],
      "text/plain": [
       "[1] 4251"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coloc_all$gwasTraitFromSource))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "id": "973bee2e-fd7d-4029-a6f9-9c3eb6a04925",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extract EFO Id (new data fromat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "id": "21e66b39-140a-45d7-9437-b3e041e95603",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$gwasEFO = str_replace_all(coloc_all$gwasEFO, '\\\\[|\\\\]', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "id": "eddf2d6a-c251-4183-9217-30368a4c4391",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$gwasEFO = str_replace_all(coloc_all$gwasEFO, \"'\", '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "id": "a675589e-bd58-4004-a8c5-ce1b98380ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$gwasEFO <- trimws(coloc_all$gwasEFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "7c2770d8-95dd-4d1a-8050-ccecae45f204",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$V1= NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "id": "11748001-1b2e-44b3-a844-fc94f102036d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Split based on gwas EFO term\n",
    "coloc_all =coloc_all %>%  separate_rows(gwasEFO, sep = \",\")\n",
    "coloc_all = unique(coloc_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "id": "c682f5ca-ac87-4e5d-b5f0-fdb8892a9077",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$efo_id = coloc_all$gwasEFO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "id": "e8ceda0c-63ba-4fbc-a5e0-8bcbcd30feb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all$efo_id <- trimws(coloc_all$efo_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "id": "197bcfdb-2e50-4ab4-a9c5-4c0d95858274",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter coloc results based on defined threshold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "id": "be567c3b-a1b4-4dfd-b4a7-5c32943dbe44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>eqtlTraitFromSource</th><th scope=col>snp_id</th><th scope=col>eQTL_CT</th><th scope=col>V1</th><th scope=col>leftStudyLocusId</th><th scope=col>leftStudyId</th><th scope=col>gwasTraitFromSource</th><th scope=col>gwasEFO</th><th scope=col>rightStudyLocusId</th><th scope=col>rightStudyId</th><th scope=col>⋯</th><th scope=col>h2</th><th scope=col>h3</th><th scope=col>h4</th><th scope=col>h3h4</th><th scope=col>H4/H3</th><th scope=col>Alt_ct_gene_cs_id</th><th scope=col>Observed</th><th scope=col>gwasStudyId</th><th scope=col>SusieRss_CS</th><th scope=col>efo_id</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG</td><td>19:58497199:G:A</td><td>CD4_T</td><td> 68</td><td>392db299e55fb8e6a525cd86172932f8</td><td>FINNGEN_R12_HEIGHT_IRN</td><td>Height, inverse-rank normalized</td><td>EFO_0004339</td><td>b82585de2a37b39493a88db9af1da33d</td><td>eQTLGen_CD4_T_ENSG00000121410</td><td>⋯</td><td>2.093031e-02</td><td>0.9729671</td><td>0.006098111</td><td>0.9790652</td><td>0.006267541</td><td>CD4_T:A1BG:19_58497199_G_A</td><td>TRUE</td><td>FINNGEN_R12_HEIGHT_IRN</td><td>2</td><td>EFO_0004339</td></tr>\n",
       "\t<tr><td>A2M </td><td>12:8972169:C:T </td><td>CD4_T</td><td>149</td><td>c0b3b6d627347cc40ac2252bcd27d7f1</td><td>GCST90018949          </td><td>Weight                         </td><td>EFO_0004338</td><td>ea74fe3c38c7b9c4d67a414a08b7f310</td><td>eQTLGen_CD4_T_ENSG00000175899</td><td>⋯</td><td>4.417051e-07</td><td>0.3921881</td><td>0.607811400</td><td>0.9999995</td><td>1.549796000</td><td>CD4_T:A2M:12_8972169_C_T  </td><td>TRUE</td><td>GCST90018949          </td><td>2</td><td>EFO_0004338</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 31\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " eqtlTraitFromSource & snp\\_id & eQTL\\_CT & V1 & leftStudyLocusId & leftStudyId & gwasTraitFromSource & gwasEFO & rightStudyLocusId & rightStudyId & ⋯ & h2 & h3 & h4 & h3h4 & H4/H3 & Alt\\_ct\\_gene\\_cs\\_id & Observed & gwasStudyId & SusieRss\\_CS & efo\\_id\\\\\n",
       " <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <lgl> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t A1BG & 19:58497199:G:A & CD4\\_T &  68 & 392db299e55fb8e6a525cd86172932f8 & FINNGEN\\_R12\\_HEIGHT\\_IRN & Height, inverse-rank normalized & EFO\\_0004339 & b82585de2a37b39493a88db9af1da33d & eQTLGen\\_CD4\\_T\\_ENSG00000121410 & ⋯ & 2.093031e-02 & 0.9729671 & 0.006098111 & 0.9790652 & 0.006267541 & CD4\\_T:A1BG:19\\_58497199\\_G\\_A & TRUE & FINNGEN\\_R12\\_HEIGHT\\_IRN & 2 & EFO\\_0004339\\\\\n",
       "\t A2M  & 12:8972169:C:T  & CD4\\_T & 149 & c0b3b6d627347cc40ac2252bcd27d7f1 & GCST90018949           & Weight                          & EFO\\_0004338 & ea74fe3c38c7b9c4d67a414a08b7f310 & eQTLGen\\_CD4\\_T\\_ENSG00000175899 & ⋯ & 4.417051e-07 & 0.3921881 & 0.607811400 & 0.9999995 & 1.549796000 & CD4\\_T:A2M:12\\_8972169\\_C\\_T   & TRUE & GCST90018949           & 2 & EFO\\_0004338\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 31\n",
       "\n",
       "| eqtlTraitFromSource &lt;chr&gt; | snp_id &lt;chr&gt; | eQTL_CT &lt;chr&gt; | V1 &lt;int&gt; | leftStudyLocusId &lt;chr&gt; | leftStudyId &lt;chr&gt; | gwasTraitFromSource &lt;chr&gt; | gwasEFO &lt;chr&gt; | rightStudyLocusId &lt;chr&gt; | rightStudyId &lt;chr&gt; | ⋯ ⋯ | h2 &lt;dbl&gt; | h3 &lt;dbl&gt; | h4 &lt;dbl&gt; | h3h4 &lt;dbl&gt; | H4/H3 &lt;dbl&gt; | Alt_ct_gene_cs_id &lt;chr&gt; | Observed &lt;lgl&gt; | gwasStudyId &lt;chr&gt; | SusieRss_CS &lt;int&gt; | efo_id &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| A1BG | 19:58497199:G:A | CD4_T |  68 | 392db299e55fb8e6a525cd86172932f8 | FINNGEN_R12_HEIGHT_IRN | Height, inverse-rank normalized | EFO_0004339 | b82585de2a37b39493a88db9af1da33d | eQTLGen_CD4_T_ENSG00000121410 | ⋯ | 2.093031e-02 | 0.9729671 | 0.006098111 | 0.9790652 | 0.006267541 | CD4_T:A1BG:19_58497199_G_A | TRUE | FINNGEN_R12_HEIGHT_IRN | 2 | EFO_0004339 |\n",
       "| A2M  | 12:8972169:C:T  | CD4_T | 149 | c0b3b6d627347cc40ac2252bcd27d7f1 | GCST90018949           | Weight                          | EFO_0004338 | ea74fe3c38c7b9c4d67a414a08b7f310 | eQTLGen_CD4_T_ENSG00000175899 | ⋯ | 4.417051e-07 | 0.3921881 | 0.607811400 | 0.9999995 | 1.549796000 | CD4_T:A2M:12_8972169_C_T   | TRUE | GCST90018949           | 2 | EFO_0004338 |\n",
       "\n"
      ],
      "text/plain": [
       "  eqtlTraitFromSource snp_id          eQTL_CT V1 \n",
       "1 A1BG                19:58497199:G:A CD4_T    68\n",
       "2 A2M                 12:8972169:C:T  CD4_T   149\n",
       "  leftStudyLocusId                 leftStudyId           \n",
       "1 392db299e55fb8e6a525cd86172932f8 FINNGEN_R12_HEIGHT_IRN\n",
       "2 c0b3b6d627347cc40ac2252bcd27d7f1 GCST90018949          \n",
       "  gwasTraitFromSource             gwasEFO     rightStudyLocusId               \n",
       "1 Height, inverse-rank normalized EFO_0004339 b82585de2a37b39493a88db9af1da33d\n",
       "2 Weight                          EFO_0004338 ea74fe3c38c7b9c4d67a414a08b7f310\n",
       "  rightStudyId                  ⋯ h2           h3        h4          h3h4     \n",
       "1 eQTLGen_CD4_T_ENSG00000121410 ⋯ 2.093031e-02 0.9729671 0.006098111 0.9790652\n",
       "2 eQTLGen_CD4_T_ENSG00000175899 ⋯ 4.417051e-07 0.3921881 0.607811400 0.9999995\n",
       "  H4/H3       Alt_ct_gene_cs_id          Observed gwasStudyId           \n",
       "1 0.006267541 CD4_T:A1BG:19_58497199_G_A TRUE     FINNGEN_R12_HEIGHT_IRN\n",
       "2 1.549796000 CD4_T:A2M:12_8972169_C_T   TRUE     GCST90018949          \n",
       "  SusieRss_CS efo_id     \n",
       "1 2           EFO_0004339\n",
       "2 2           EFO_0004338"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(coloc_all,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "id": "f74f0e19-351d-41a1-9ef5-cc8d6c1298f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'relation'"
      ],
      "text/latex": [
       "'relation'"
      ],
      "text/markdown": [
       "'relation'"
      ],
      "text/plain": [
       "[1] \"relation\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_thres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "id": "b87a649b-fe6c-4887-a82f-b39fd1e12265",
   "metadata": {},
   "outputs": [],
   "source": [
    "## OPT1: based on threshold 0.6\n",
    "if(coloc_thres != 'relation'){\n",
    "    coloc_all = coloc_all[coloc_all$h4 > coloc_thres,]\n",
    "}\n",
    "\n",
    "## OPT2: new combination of thres:\n",
    "if(coloc_thres == 'relation'){\n",
    "    coloc_all = coloc_all[((coloc_all$h4 + coloc_all$h3) > 0.9) & ((coloc_all$h4 / coloc_all$h3) > 3),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "id": "7265f779-d2b0-4688-8a16-71f5b19f1c9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.6798902"
      ],
      "text/latex": [
       "0.6798902"
      ],
      "text/markdown": [
       "0.6798902"
      ],
      "text/plain": [
       "[1] 0.6798902"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "min(coloc_all$h4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 467,
   "id": "ab84de9c-804f-49e1-be21-e5bda366d287",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amount of unique studies and traits in coloc result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "id": "4b0bc8a7-99c3-47bd-a542-57a43bac493b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4987"
      ],
      "text/latex": [
       "4987"
      ],
      "text/markdown": [
       "4987"
      ],
      "text/plain": [
       "[1] 4987"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coloc_all$gwasStudyId))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "id": "ed25335c-a1f9-4cef-ab95-c69c6fd911db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "3902"
      ],
      "text/latex": [
       "3902"
      ],
      "text/markdown": [
       "3902"
      ],
      "text/plain": [
       "[1] 3902"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(coloc_all$gwasTraitFromSource))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "id": "1a1a40d2-3994-43e5-88b1-5f305b3f6562",
   "metadata": {},
   "outputs": [],
   "source": [
    "### OPTIONAL: Remove FINNGEN studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "id": "6fb75538-9b24-4c62-af8a-7f8e1d0df5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_finngen = FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "id": "0f408366-d04d-4e3e-ab36-8228b7e16281",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "68232"
      ],
      "text/latex": [
       "68232"
      ],
      "text/markdown": [
       "68232"
      ],
      "text/plain": [
       "[1] 68232"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coloc_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "id": "938e0431-0611-40e3-b7a8-51060854c8f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if( remove_finngen == TRUE){\n",
    "    coloc_all = coloc_all[(is.na(str_extract(coloc_all$gwasStudyId, 'FINNGEN'))),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "id": "1ad7aa6b-f89e-42ee-871c-499f97277601",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Map classification of traits based on GWAS catalogue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "id": "d927a00d-8313-48ac-b3b8-34b0f0c980fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_trait_class = fread(paste0('../data/current/', 'meta-data/GWAS/gwas_catalog_trait-mappings_r2024-11-20.tsv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "id": "5ae5187a-ef03-4d72-a048-d9c52d0ef016",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Disease trait</th><th scope=col>EFO term</th><th scope=col>EFO URI</th><th scope=col>Parent term</th><th scope=col>Parent URI</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Cervical cancer                </td><td>cervical cancer</td><td>http://purl.obolibrary.org/obo/MONDO_0002974</td><td>Cancer</td><td>http://www.ebi.ac.uk/efo/EFO_0000616</td></tr>\n",
       "\t<tr><td>Cervical cancer (PheCode 180.1)</td><td>cervical cancer</td><td>http://purl.obolibrary.org/obo/MONDO_0002974</td><td>Cancer</td><td>http://www.ebi.ac.uk/efo/EFO_0000616</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " Disease trait & EFO term & EFO URI & Parent term & Parent URI\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t Cervical cancer                 & cervical cancer & http://purl.obolibrary.org/obo/MONDO\\_0002974 & Cancer & http://www.ebi.ac.uk/efo/EFO\\_0000616\\\\\n",
       "\t Cervical cancer (PheCode 180.1) & cervical cancer & http://purl.obolibrary.org/obo/MONDO\\_0002974 & Cancer & http://www.ebi.ac.uk/efo/EFO\\_0000616\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 5\n",
       "\n",
       "| Disease trait &lt;chr&gt; | EFO term &lt;chr&gt; | EFO URI &lt;chr&gt; | Parent term &lt;chr&gt; | Parent URI &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| Cervical cancer                 | cervical cancer | http://purl.obolibrary.org/obo/MONDO_0002974 | Cancer | http://www.ebi.ac.uk/efo/EFO_0000616 |\n",
       "| Cervical cancer (PheCode 180.1) | cervical cancer | http://purl.obolibrary.org/obo/MONDO_0002974 | Cancer | http://www.ebi.ac.uk/efo/EFO_0000616 |\n",
       "\n"
      ],
      "text/plain": [
       "  Disease trait                   EFO term       \n",
       "1 Cervical cancer                 cervical cancer\n",
       "2 Cervical cancer (PheCode 180.1) cervical cancer\n",
       "  EFO URI                                      Parent term\n",
       "1 http://purl.obolibrary.org/obo/MONDO_0002974 Cancer     \n",
       "2 http://purl.obolibrary.org/obo/MONDO_0002974 Cancer     \n",
       "  Parent URI                          \n",
       "1 http://www.ebi.ac.uk/efo/EFO_0000616\n",
       "2 http://www.ebi.ac.uk/efo/EFO_0000616"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gwas_trait_class,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "id": "920a773d-9fbc-4e6e-8f1f-33c0ab63f412",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Classify into immune and disease related terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "id": "bc1a576e-83b9-47b1-9325-f4efb852ec95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "15579"
      ],
      "text/latex": [
       "15579"
      ],
      "text/markdown": [
       "15579"
      ],
      "text/plain": [
       "[1] 15579"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "17"
      ],
      "text/latex": [
       "17"
      ],
      "text/markdown": [
       "17"
      ],
      "text/plain": [
       "[1] 17"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(gwas_trait_class$`EFO URI`))\n",
    "length(unique(gwas_trait_class$`Parent URI`))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "id": "2541aa5d-aef3-4849-a78b-c6c6ec2610e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Cancer'</li><li>'Other trait'</li><li>'Immune system disorder'</li><li>'Other disease'</li><li>'Cardiovascular disease'</li><li>'Neurological disorder'</li><li>'Other measurement'</li><li>'Biological process'</li><li>'Lipid or lipoprotein measurement'</li><li>'Body measurement'</li><li>'Cardiovascular measurement'</li><li>'Metabolic disorder'</li><li>'Inflammatory measurement'</li><li>'Digestive system disorder'</li><li>'Hematological measurement'</li><li>'Liver enzyme measurement'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Cancer'\n",
       "\\item 'Other trait'\n",
       "\\item 'Immune system disorder'\n",
       "\\item 'Other disease'\n",
       "\\item 'Cardiovascular disease'\n",
       "\\item 'Neurological disorder'\n",
       "\\item 'Other measurement'\n",
       "\\item 'Biological process'\n",
       "\\item 'Lipid or lipoprotein measurement'\n",
       "\\item 'Body measurement'\n",
       "\\item 'Cardiovascular measurement'\n",
       "\\item 'Metabolic disorder'\n",
       "\\item 'Inflammatory measurement'\n",
       "\\item 'Digestive system disorder'\n",
       "\\item 'Hematological measurement'\n",
       "\\item 'Liver enzyme measurement'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Cancer'\n",
       "2. 'Other trait'\n",
       "3. 'Immune system disorder'\n",
       "4. 'Other disease'\n",
       "5. 'Cardiovascular disease'\n",
       "6. 'Neurological disorder'\n",
       "7. 'Other measurement'\n",
       "8. 'Biological process'\n",
       "9. 'Lipid or lipoprotein measurement'\n",
       "10. 'Body measurement'\n",
       "11. 'Cardiovascular measurement'\n",
       "12. 'Metabolic disorder'\n",
       "13. 'Inflammatory measurement'\n",
       "14. 'Digestive system disorder'\n",
       "15. 'Hematological measurement'\n",
       "16. 'Liver enzyme measurement'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Cancer\"                           \"Other trait\"                     \n",
       " [3] \"Immune system disorder\"           \"Other disease\"                   \n",
       " [5] \"Cardiovascular disease\"           \"Neurological disorder\"           \n",
       " [7] \"Other measurement\"                \"Biological process\"              \n",
       " [9] \"Lipid or lipoprotein measurement\" \"Body measurement\"                \n",
       "[11] \"Cardiovascular measurement\"       \"Metabolic disorder\"              \n",
       "[13] \"Inflammatory measurement\"         \"Digestive system disorder\"       \n",
       "[15] \"Hematological measurement\"        \"Liver enzyme measurement\"        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Categories in GWAS catalogue\n",
    "unique(gwas_trait_class$`Parent term`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "id": "2a25c79a-af0d-499f-90ab-d251935fb7a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define disease and immune traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "id": "ee9c856b-86aa-4fc1-bae1-4785d6e3f0f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_trait_class$gwas_disease_class = 0\n",
    "gwas_trait_class$gwas_disease_class[!gwas_trait_class$`Parent term` %in% c('Other trait', 'Other measurement', 'Biological process', 'Lipid or lipoprotein measurement',\n",
    "                                                                          'Body measurement','Cardiovascular measurement', 'Inflammatory measurement', 'Hematological measurement', 'Liver enzyme measurement' )]= 1\n",
    "\n",
    "gwas_trait_class$gwas_immune_class = 0\n",
    "gwas_trait_class$gwas_immune_class[gwas_trait_class$`Parent term` %in% c('Immune system disorder')]= 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "id": "c9d57906-2be5-4638-a364-1ce2dfc8869b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Disease trait</th><th scope=col>EFO term</th><th scope=col>EFO URI</th><th scope=col>Parent term</th><th scope=col>Parent URI</th><th scope=col>gwas_disease_class</th><th scope=col>gwas_immune_class</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Cervical cancer                </td><td>cervical cancer</td><td>http://purl.obolibrary.org/obo/MONDO_0002974</td><td>Cancer</td><td>http://www.ebi.ac.uk/efo/EFO_0000616</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><td>Cervical cancer (PheCode 180.1)</td><td>cervical cancer</td><td>http://purl.obolibrary.org/obo/MONDO_0002974</td><td>Cancer</td><td>http://www.ebi.ac.uk/efo/EFO_0000616</td><td>1</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 7\n",
       "\\begin{tabular}{lllllll}\n",
       " Disease trait & EFO term & EFO URI & Parent term & Parent URI & gwas\\_disease\\_class & gwas\\_immune\\_class\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Cervical cancer                 & cervical cancer & http://purl.obolibrary.org/obo/MONDO\\_0002974 & Cancer & http://www.ebi.ac.uk/efo/EFO\\_0000616 & 1 & 0\\\\\n",
       "\t Cervical cancer (PheCode 180.1) & cervical cancer & http://purl.obolibrary.org/obo/MONDO\\_0002974 & Cancer & http://www.ebi.ac.uk/efo/EFO\\_0000616 & 1 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 7\n",
       "\n",
       "| Disease trait &lt;chr&gt; | EFO term &lt;chr&gt; | EFO URI &lt;chr&gt; | Parent term &lt;chr&gt; | Parent URI &lt;chr&gt; | gwas_disease_class &lt;dbl&gt; | gwas_immune_class &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| Cervical cancer                 | cervical cancer | http://purl.obolibrary.org/obo/MONDO_0002974 | Cancer | http://www.ebi.ac.uk/efo/EFO_0000616 | 1 | 0 |\n",
       "| Cervical cancer (PheCode 180.1) | cervical cancer | http://purl.obolibrary.org/obo/MONDO_0002974 | Cancer | http://www.ebi.ac.uk/efo/EFO_0000616 | 1 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Disease trait                   EFO term       \n",
       "1 Cervical cancer                 cervical cancer\n",
       "2 Cervical cancer (PheCode 180.1) cervical cancer\n",
       "  EFO URI                                      Parent term\n",
       "1 http://purl.obolibrary.org/obo/MONDO_0002974 Cancer     \n",
       "2 http://purl.obolibrary.org/obo/MONDO_0002974 Cancer     \n",
       "  Parent URI                           gwas_disease_class gwas_immune_class\n",
       "1 http://www.ebi.ac.uk/efo/EFO_0000616 1                  0                \n",
       "2 http://www.ebi.ac.uk/efo/EFO_0000616 1                  0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gwas_trait_class,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "id": "348e6bef-d75e-4bf6-ac71-c135171c365c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extact EFO- ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "id": "cdd8edba-eb0d-42de-8ecb-486c50939289",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_trait_class$EFO_ID = str_extract(gwas_trait_class$`EFO URI`, \"[^/]+$\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5948ce96-147f-4ece-8ff6-8ed4b54c16dd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "id": "4dbfe611-a4b7-4f32-b0a8-bc73c91ab97f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge the classification from GWAS catalogue to coloc_all results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "5738d648-458d-47d4-83e6-d80b876fde61",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all_merged = coloc_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "id": "fb9fb9e9-0250-4627-b80c-ade2ad535a85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "68232"
      ],
      "text/latex": [
       "68232"
      ],
      "text/markdown": [
       "68232"
      ],
      "text/plain": [
       "[1] 68232"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coloc_all_merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "id": "d29cd180-027a-4d18-a919-14a6f99ac78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(coloc_all_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "id": "2c231fc9-ef55-4cf2-828a-c04d2e6f2c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_all_merged = merge(coloc_all_merged, unique(gwas_trait_class[,c('EFO URI', 'gwas_disease_class', 'gwas_immune_class', 'Parent term', 'EFO_ID')]), by.x = 'efo_id', by.y = 'EFO_ID',  all.x = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "9a7d968f-c235-4a3f-9ad2-acb5d30bdfa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "68232"
      ],
      "text/latex": [
       "68232"
      ],
      "text/markdown": [
       "68232"
      ],
      "text/plain": [
       "[1] 68232"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(coloc_all_merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "id": "bd6f2a40-0efc-4314-8552-50eba470e218",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on the defined GWAS subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "id": "205403ac-dbde-4c39-8d08-0ebafeb50220",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if(use_gwas == 'rm_cell_count'){\n",
    "    coloc_all_merged = coloc_all_merged[coloc_all_merged$`Parent term` %in% unique(coloc_all_merged$`Parent term`[(coloc_all_merged$`Parent term` != 'Hematological measurement') & (!is.na(coloc_all_merged$`Parent term`))]),]  # everything except: 'Hematological measurement'\n",
    "    coloc_all_merged = coloc_all_merged[!coloc_all_merged$gwasTraitFromSource %in% c( 'basophil count', 'basophil percentage of leukocytes', 'basophil percentage of granulocytes') ,]\n",
    "    #coloc_all_merged = coloc_all_merged[coloc_all_merged$`Parent term` != 'Inflammatory measurement',]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "id": "51d6c36c-1724-4479-91c1-0d7395416e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(use_gwas == 'disease'){\n",
    "    coloc_all_merged = coloc_all_merged[(coloc_all_merged$gwas_disease_class == 1) & (!is.na(coloc_all_merged$gwas_disease_class)),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "id": "6186c515-85de-4938-87ed-f4b434972de8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(use_gwas == 'immune'){\n",
    "    coloc_all_merged = coloc_all_merged[(coloc_all_merged$gwas_immune_class == 1) & (!is.na(coloc_all_merged$gwas_immune_class)),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "id": "8aab1a05-c95d-4251-b7e1-52228fe306fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1"
      ],
      "text/latex": [
       "1"
      ],
      "text/markdown": [
       "1"
      ],
      "text/plain": [
       "[1] 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(coloc_all_merged$gwas_disease_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "id": "c57067d3-40a5-4efc-8181-ae2d1de8f406",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>1</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 1\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 1\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(coloc_all_merged$gwas_immune_class)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c5abed-1baa-42e9-a46f-8caf9abd0306",
   "metadata": {},
   "source": [
    "# Prepare Analysis data: Contingency table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "id": "8deb565d-da39-4c7e-8ae7-5e58cc2780df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Get information about colocalisation of eGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "id": "6edc5e9f-ac12-4677-96dc-ba18dba660f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data = coloc_all_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "b261ecd9-be18-4952-a4d7-94e54d16bb0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Amount of studies used for enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "d44c31f5-da7a-4fd6-a1ce-eaf7431d931d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "577"
      ],
      "text/latex": [
       "577"
      ],
      "text/markdown": [
       "577"
      ],
      "text/plain": [
       "[1] 577"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(analysis_data$gwasStudyId))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "id": "dddc10e1-a579-4f09-8dd7-1796815212ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 35</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>efo_id</th><th scope=col>eqtlTraitFromSource</th><th scope=col>snp_id</th><th scope=col>eQTL_CT</th><th scope=col>V1</th><th scope=col>leftStudyLocusId</th><th scope=col>leftStudyId</th><th scope=col>gwasTraitFromSource</th><th scope=col>gwasEFO</th><th scope=col>rightStudyLocusId</th><th scope=col>⋯</th><th scope=col>h3h4</th><th scope=col>H4/H3</th><th scope=col>Alt_ct_gene_cs_id</th><th scope=col>Observed</th><th scope=col>gwasStudyId</th><th scope=col>SusieRss_CS</th><th scope=col>EFO URI</th><th scope=col>gwas_disease_class</th><th scope=col>gwas_immune_class</th><th scope=col>Parent term</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>EFO_0000178</td><td>TTC33</td><td>5:40756406:A:T </td><td>CD4_T</td><td>1923931</td><td>409a70434d077ba59a65dfb0d882105a</td><td>GCST90018629</td><td>Gastric cancer</td><td>EFO_0000178</td><td>468413dbbccd4b21d3116f75bc2aa268</td><td>⋯</td><td>1.0000000</td><td> 12.57807</td><td>CD4_T:TTC33:5_40756406_A_T</td><td>TRUE</td><td>GCST90018629</td><td>1</td><td>http://www.ebi.ac.uk/efo/EFO_0000178</td><td>1</td><td>0</td><td>Cancer</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>EFO_0000178</td><td>GBA  </td><td>1:155206101:G:A</td><td>CD4_T</td><td> 682817</td><td>1cd53b3adcc6af1ffcae0818f5979bd2</td><td>GCST90018849</td><td>Gastric cancer</td><td>EFO_0000178</td><td>1bba5edbeb0aa0c39d568f48dc3319ba</td><td>⋯</td><td>0.9999577</td><td>132.63780</td><td>CD4_T:GBA:1_155206101_G_A </td><td>TRUE</td><td>GCST90018849</td><td>1</td><td>http://www.ebi.ac.uk/efo/EFO_0000178</td><td>1</td><td>0</td><td>Cancer</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 35\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & efo\\_id & eqtlTraitFromSource & snp\\_id & eQTL\\_CT & V1 & leftStudyLocusId & leftStudyId & gwasTraitFromSource & gwasEFO & rightStudyLocusId & ⋯ & h3h4 & H4/H3 & Alt\\_ct\\_gene\\_cs\\_id & Observed & gwasStudyId & SusieRss\\_CS & EFO URI & gwas\\_disease\\_class & gwas\\_immune\\_class & Parent term\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <chr> & <lgl> & <chr> & <int> & <chr> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & EFO\\_0000178 & TTC33 & 5:40756406:A:T  & CD4\\_T & 1923931 & 409a70434d077ba59a65dfb0d882105a & GCST90018629 & Gastric cancer & EFO\\_0000178 & 468413dbbccd4b21d3116f75bc2aa268 & ⋯ & 1.0000000 &  12.57807 & CD4\\_T:TTC33:5\\_40756406\\_A\\_T & TRUE & GCST90018629 & 1 & http://www.ebi.ac.uk/efo/EFO\\_0000178 & 1 & 0 & Cancer\\\\\n",
       "\t2 & EFO\\_0000178 & GBA   & 1:155206101:G:A & CD4\\_T &  682817 & 1cd53b3adcc6af1ffcae0818f5979bd2 & GCST90018849 & Gastric cancer & EFO\\_0000178 & 1bba5edbeb0aa0c39d568f48dc3319ba & ⋯ & 0.9999577 & 132.63780 & CD4\\_T:GBA:1\\_155206101\\_G\\_A  & TRUE & GCST90018849 & 1 & http://www.ebi.ac.uk/efo/EFO\\_0000178 & 1 & 0 & Cancer\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 35\n",
       "\n",
       "| <!--/--> | efo_id &lt;chr&gt; | eqtlTraitFromSource &lt;chr&gt; | snp_id &lt;chr&gt; | eQTL_CT &lt;chr&gt; | V1 &lt;int&gt; | leftStudyLocusId &lt;chr&gt; | leftStudyId &lt;chr&gt; | gwasTraitFromSource &lt;chr&gt; | gwasEFO &lt;chr&gt; | rightStudyLocusId &lt;chr&gt; | ⋯ ⋯ | h3h4 &lt;dbl&gt; | H4/H3 &lt;dbl&gt; | Alt_ct_gene_cs_id &lt;chr&gt; | Observed &lt;lgl&gt; | gwasStudyId &lt;chr&gt; | SusieRss_CS &lt;int&gt; | EFO URI &lt;chr&gt; | gwas_disease_class &lt;dbl&gt; | gwas_immune_class &lt;dbl&gt; | Parent term &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | EFO_0000178 | TTC33 | 5:40756406:A:T  | CD4_T | 1923931 | 409a70434d077ba59a65dfb0d882105a | GCST90018629 | Gastric cancer | EFO_0000178 | 468413dbbccd4b21d3116f75bc2aa268 | ⋯ | 1.0000000 |  12.57807 | CD4_T:TTC33:5_40756406_A_T | TRUE | GCST90018629 | 1 | http://www.ebi.ac.uk/efo/EFO_0000178 | 1 | 0 | Cancer |\n",
       "| 2 | EFO_0000178 | GBA   | 1:155206101:G:A | CD4_T |  682817 | 1cd53b3adcc6af1ffcae0818f5979bd2 | GCST90018849 | Gastric cancer | EFO_0000178 | 1bba5edbeb0aa0c39d568f48dc3319ba | ⋯ | 0.9999577 | 132.63780 | CD4_T:GBA:1_155206101_G_A  | TRUE | GCST90018849 | 1 | http://www.ebi.ac.uk/efo/EFO_0000178 | 1 | 0 | Cancer |\n",
       "\n"
      ],
      "text/plain": [
       "  efo_id      eqtlTraitFromSource snp_id          eQTL_CT V1     \n",
       "1 EFO_0000178 TTC33               5:40756406:A:T  CD4_T   1923931\n",
       "2 EFO_0000178 GBA                 1:155206101:G:A CD4_T    682817\n",
       "  leftStudyLocusId                 leftStudyId  gwasTraitFromSource gwasEFO    \n",
       "1 409a70434d077ba59a65dfb0d882105a GCST90018629 Gastric cancer      EFO_0000178\n",
       "2 1cd53b3adcc6af1ffcae0818f5979bd2 GCST90018849 Gastric cancer      EFO_0000178\n",
       "  rightStudyLocusId                ⋯ h3h4      H4/H3    \n",
       "1 468413dbbccd4b21d3116f75bc2aa268 ⋯ 1.0000000  12.57807\n",
       "2 1bba5edbeb0aa0c39d568f48dc3319ba ⋯ 0.9999577 132.63780\n",
       "  Alt_ct_gene_cs_id          Observed gwasStudyId  SusieRss_CS\n",
       "1 CD4_T:TTC33:5_40756406_A_T TRUE     GCST90018629 1          \n",
       "2 CD4_T:GBA:1_155206101_G_A  TRUE     GCST90018849 1          \n",
       "  EFO URI                              gwas_disease_class gwas_immune_class\n",
       "1 http://www.ebi.ac.uk/efo/EFO_0000178 1                  0                \n",
       "2 http://www.ebi.ac.uk/efo/EFO_0000178 1                  0                \n",
       "  Parent term\n",
       "1 Cancer     \n",
       "2 Cancer     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(analysis_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "id": "4e33fddd-10ea-4ba1-983b-278b33b774f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract cell-type from coloc results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "id": "4f6bee43-8e4a-4197-ae84-ea26f1fc03bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data$eQTL_CT = str_extract(analysis_data$rightStudyId, 'CD4_T|CD8_T|Mono|NK|B|DC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "id": "2e977fd6-a72b-4bd3-9712-dace31a49e40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'CD4_T'</li><li>'B'</li><li>'CD8_T'</li><li>'Mono'</li><li>'NK'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'CD4\\_T'\n",
       "\\item 'B'\n",
       "\\item 'CD8\\_T'\n",
       "\\item 'Mono'\n",
       "\\item 'NK'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'CD4_T'\n",
       "2. 'B'\n",
       "3. 'CD8_T'\n",
       "4. 'Mono'\n",
       "5. 'NK'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"CD4_T\" \"B\"     \"CD8_T\" \"Mono\"  \"NK\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(analysis_data$eQTL_CT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "id": "3ffc92a1-5eb9-470f-a70a-686f921eb9e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter coloc results on only decision tree eGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "id": "8bd817b5-1486-45a3-894a-4a906863fe34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>gene</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B</td><td>x         </td></tr>\n",
       "\t<tr><td>B</td><td>AC004687.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " cell\\_type & gene\\\\\n",
       " <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t B & x         \\\\\n",
       "\t B & AC004687.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 2\n",
       "\n",
       "| cell_type &lt;chr&gt; | gene &lt;chr&gt; |\n",
       "|---|---|\n",
       "| B | x          |\n",
       "| B | AC004687.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type gene      \n",
       "1 B         x         \n",
       "2 B         AC004687.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_list,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "id": "eec28366-18cc-44c7-ac26-a77e4cadfffa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "819"
      ],
      "text/latex": [
       "819"
      ],
      "text/markdown": [
       "819"
      ],
      "text/plain": [
       "[1] 819"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(analysis_data$eqtlTraitFromSource))   # amount of eGenes before filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "id": "db070dec-1595-45f3-a53f-589b7773c393",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "only_dt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "id": "0417b4be-51d2-4062-a84e-e1004049babf",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(only_dt == TRUE){\n",
    "    analysis_data = merge(analysis_data, gene_list, by.x = c('eqtlTraitFromSource', 'eQTL_CT'),\n",
    "                     by.y = c('gene', 'cell_type'))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "id": "914f7236-032c-4676-a8ae-82239cc57a5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "695"
      ],
      "text/latex": [
       "695"
      ],
      "text/markdown": [
       "695"
      ],
      "text/plain": [
       "[1] 695"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(analysis_data$eqtlTraitFromSource)) # amount of colocalising eGenes after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "id": "942513c8-bac2-49ea-865b-8bcca5fa6972",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove HLA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "id": "7bed1566-edc8-4323-b95c-1d48fa43544d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(remove_HLA == TRUE){\n",
    "    analysis_data = analysis_data[is.na(str_extract(analysis_data$eqtlTraitFromSource, 'HLA')),]\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "id": "ab26eb6f-d01d-47ff-b57f-9fda1dbed094",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort(unique(coloc_all$eqtlTraitFromSource))[700:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "id": "89413b61-6309-4a23-add1-a35216fcaa34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "695"
      ],
      "text/latex": [
       "695"
      ],
      "text/markdown": [
       "695"
      ],
      "text/plain": [
       "[1] 695"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(analysis_data$eqtlTraitFromSource))   # amount of colocalising eGenes after removing HLA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "id": "672431ad-5259-4459-b061-0bdcb14afc40",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove eGenes that were not tested for eQTL/ co-eQTL  (affects B and Mono as there coloc results include more eGenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 516,
   "id": "2cd20b30-e919-4207-8868-789b160d6f56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4452"
      ],
      "text/latex": [
       "4452"
      ],
      "text/markdown": [
       "4452"
      ],
      "text/plain": [
       "[1] 4452"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(analysis_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 517,
   "id": "d458432a-b38c-488b-bea4-8560d9615556",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data = merge(analysis_data, unique(snp_gene_mapping_all[,c('feature_id', 'cell_type')]), by.x = c('eqtlTraitFromSource', 'eQTL_CT'), by.y = c('feature_id', 'cell_type'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 518,
   "id": "b39a8ae7-542c-4bbe-b3e8-b48dcd0e221d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>eQTL_CT</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td> 597</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>1460</td></tr>\n",
       "\t<tr><td>CD8_T</td><td> 791</td></tr>\n",
       "\t<tr><td>Mono </td><td> 684</td></tr>\n",
       "\t<tr><td>NK   </td><td> 920</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " eQTL\\_CT & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t B     &  597\\\\\n",
       "\t CD4\\_T & 1460\\\\\n",
       "\t CD8\\_T &  791\\\\\n",
       "\t Mono  &  684\\\\\n",
       "\t NK    &  920\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 5 × 2\n",
       "\n",
       "| eQTL_CT &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| B     |  597 |\n",
       "| CD4_T | 1460 |\n",
       "| CD8_T |  791 |\n",
       "| Mono  |  684 |\n",
       "| NK    |  920 |\n",
       "\n"
      ],
      "text/plain": [
       "  eQTL_CT n   \n",
       "1 B        597\n",
       "2 CD4_T   1460\n",
       "3 CD8_T    791\n",
       "4 Mono     684\n",
       "5 NK       920"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "analysis_data %>% group_by(eQTL_CT) %>% dplyr::count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 519,
   "id": "a0f13248-677a-450e-ad11-1a30bb97d405",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4452"
      ],
      "text/latex": [
       "4452"
      ],
      "text/markdown": [
       "4452"
      ],
      "text/plain": [
       "[1] 4452"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(analysis_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 520,
   "id": "9951463f-2635-4210-bf82-20ee31952512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 35</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>eqtlTraitFromSource</th><th scope=col>eQTL_CT</th><th scope=col>efo_id</th><th scope=col>snp_id</th><th scope=col>V1</th><th scope=col>leftStudyLocusId</th><th scope=col>leftStudyId</th><th scope=col>gwasTraitFromSource</th><th scope=col>gwasEFO</th><th scope=col>rightStudyLocusId</th><th scope=col>⋯</th><th scope=col>h3h4</th><th scope=col>H4/H3</th><th scope=col>Alt_ct_gene_cs_id</th><th scope=col>Observed</th><th scope=col>gwasStudyId</th><th scope=col>SusieRss_CS</th><th scope=col>EFO URI</th><th scope=col>gwas_disease_class</th><th scope=col>gwas_immune_class</th><th scope=col>Parent term</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A2M    </td><td>CD4_T</td><td>MONDO_0005147</td><td>12:8972169:C:T</td><td>156</td><td>a95199d25d084b697b9446af089bd878</td><td>FINNGEN_R12_T1D</td><td>Type 1 diabetes, definitions combined</td><td>MONDO_0005147</td><td>ea74fe3c38c7b9c4d67a414a08b7f310</td><td>⋯</td><td>0.9999990</td><td>14.961340</td><td>CD4_T:A2M:12_8972169_C_T    </td><td>TRUE</td><td>FINNGEN_R12_T1D</td><td>2</td><td>http://purl.obolibrary.org/obo/MONDO_0005147</td><td>1</td><td>1</td><td>Immune system disorder</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2M-AS1</td><td>CD4_T</td><td>MONDO_0005147</td><td>12:8959409:C:T</td><td>307</td><td>a95199d25d084b697b9446af089bd878</td><td>FINNGEN_R12_T1D</td><td>Type 1 diabetes, definitions combined</td><td>MONDO_0005147</td><td>5753696279250a3cd12034956f8c20d1</td><td>⋯</td><td>0.9999984</td><td> 8.645992</td><td>CD4_T:A2M-AS1:12_8959409_C_T</td><td>TRUE</td><td>FINNGEN_R12_T1D</td><td>2</td><td>http://purl.obolibrary.org/obo/MONDO_0005147</td><td>1</td><td>1</td><td>Immune system disorder</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 35\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & eqtlTraitFromSource & eQTL\\_CT & efo\\_id & snp\\_id & V1 & leftStudyLocusId & leftStudyId & gwasTraitFromSource & gwasEFO & rightStudyLocusId & ⋯ & h3h4 & H4/H3 & Alt\\_ct\\_gene\\_cs\\_id & Observed & gwasStudyId & SusieRss\\_CS & EFO URI & gwas\\_disease\\_class & gwas\\_immune\\_class & Parent term\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <chr> & <lgl> & <chr> & <int> & <chr> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & A2M     & CD4\\_T & MONDO\\_0005147 & 12:8972169:C:T & 156 & a95199d25d084b697b9446af089bd878 & FINNGEN\\_R12\\_T1D & Type 1 diabetes, definitions combined & MONDO\\_0005147 & ea74fe3c38c7b9c4d67a414a08b7f310 & ⋯ & 0.9999990 & 14.961340 & CD4\\_T:A2M:12\\_8972169\\_C\\_T     & TRUE & FINNGEN\\_R12\\_T1D & 2 & http://purl.obolibrary.org/obo/MONDO\\_0005147 & 1 & 1 & Immune system disorder\\\\\n",
       "\t2 & A2M-AS1 & CD4\\_T & MONDO\\_0005147 & 12:8959409:C:T & 307 & a95199d25d084b697b9446af089bd878 & FINNGEN\\_R12\\_T1D & Type 1 diabetes, definitions combined & MONDO\\_0005147 & 5753696279250a3cd12034956f8c20d1 & ⋯ & 0.9999984 &  8.645992 & CD4\\_T:A2M-AS1:12\\_8959409\\_C\\_T & TRUE & FINNGEN\\_R12\\_T1D & 2 & http://purl.obolibrary.org/obo/MONDO\\_0005147 & 1 & 1 & Immune system disorder\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 35\n",
       "\n",
       "| <!--/--> | eqtlTraitFromSource &lt;chr&gt; | eQTL_CT &lt;chr&gt; | efo_id &lt;chr&gt; | snp_id &lt;chr&gt; | V1 &lt;int&gt; | leftStudyLocusId &lt;chr&gt; | leftStudyId &lt;chr&gt; | gwasTraitFromSource &lt;chr&gt; | gwasEFO &lt;chr&gt; | rightStudyLocusId &lt;chr&gt; | ⋯ ⋯ | h3h4 &lt;dbl&gt; | H4/H3 &lt;dbl&gt; | Alt_ct_gene_cs_id &lt;chr&gt; | Observed &lt;lgl&gt; | gwasStudyId &lt;chr&gt; | SusieRss_CS &lt;int&gt; | EFO URI &lt;chr&gt; | gwas_disease_class &lt;dbl&gt; | gwas_immune_class &lt;dbl&gt; | Parent term &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A2M     | CD4_T | MONDO_0005147 | 12:8972169:C:T | 156 | a95199d25d084b697b9446af089bd878 | FINNGEN_R12_T1D | Type 1 diabetes, definitions combined | MONDO_0005147 | ea74fe3c38c7b9c4d67a414a08b7f310 | ⋯ | 0.9999990 | 14.961340 | CD4_T:A2M:12_8972169_C_T     | TRUE | FINNGEN_R12_T1D | 2 | http://purl.obolibrary.org/obo/MONDO_0005147 | 1 | 1 | Immune system disorder |\n",
       "| 2 | A2M-AS1 | CD4_T | MONDO_0005147 | 12:8959409:C:T | 307 | a95199d25d084b697b9446af089bd878 | FINNGEN_R12_T1D | Type 1 diabetes, definitions combined | MONDO_0005147 | 5753696279250a3cd12034956f8c20d1 | ⋯ | 0.9999984 |  8.645992 | CD4_T:A2M-AS1:12_8959409_C_T | TRUE | FINNGEN_R12_T1D | 2 | http://purl.obolibrary.org/obo/MONDO_0005147 | 1 | 1 | Immune system disorder |\n",
       "\n"
      ],
      "text/plain": [
       "  eqtlTraitFromSource eQTL_CT efo_id        snp_id         V1 \n",
       "1 A2M                 CD4_T   MONDO_0005147 12:8972169:C:T 156\n",
       "2 A2M-AS1             CD4_T   MONDO_0005147 12:8959409:C:T 307\n",
       "  leftStudyLocusId                 leftStudyId    \n",
       "1 a95199d25d084b697b9446af089bd878 FINNGEN_R12_T1D\n",
       "2 a95199d25d084b697b9446af089bd878 FINNGEN_R12_T1D\n",
       "  gwasTraitFromSource                   gwasEFO      \n",
       "1 Type 1 diabetes, definitions combined MONDO_0005147\n",
       "2 Type 1 diabetes, definitions combined MONDO_0005147\n",
       "  rightStudyLocusId                ⋯ h3h4      H4/H3    \n",
       "1 ea74fe3c38c7b9c4d67a414a08b7f310 ⋯ 0.9999990 14.961340\n",
       "2 5753696279250a3cd12034956f8c20d1 ⋯ 0.9999984  8.645992\n",
       "  Alt_ct_gene_cs_id            Observed gwasStudyId     SusieRss_CS\n",
       "1 CD4_T:A2M:12_8972169_C_T     TRUE     FINNGEN_R12_T1D 2          \n",
       "2 CD4_T:A2M-AS1:12_8959409_C_T TRUE     FINNGEN_R12_T1D 2          \n",
       "  EFO URI                                      gwas_disease_class\n",
       "1 http://purl.obolibrary.org/obo/MONDO_0005147 1                 \n",
       "2 http://purl.obolibrary.org/obo/MONDO_0005147 1                 \n",
       "  gwas_immune_class Parent term           \n",
       "1 1                 Immune system disorder\n",
       "2 1                 Immune system disorder"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(analysis_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d525f54-1a26-4251-9b52-f1f2f7e8a78e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 521,
   "id": "9feb4cd5-3fe1-46c9-a4b6-9be67161cd68",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge Information about whether eGene is a co-eQTL or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 522,
   "id": "d33ebd96-1b4b-4864-bc12-247e2dd16804",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 35</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>eqtlTraitFromSource</th><th scope=col>eQTL_CT</th><th scope=col>efo_id</th><th scope=col>snp_id</th><th scope=col>V1</th><th scope=col>leftStudyLocusId</th><th scope=col>leftStudyId</th><th scope=col>gwasTraitFromSource</th><th scope=col>gwasEFO</th><th scope=col>rightStudyLocusId</th><th scope=col>⋯</th><th scope=col>h3h4</th><th scope=col>H4/H3</th><th scope=col>Alt_ct_gene_cs_id</th><th scope=col>Observed</th><th scope=col>gwasStudyId</th><th scope=col>SusieRss_CS</th><th scope=col>EFO URI</th><th scope=col>gwas_disease_class</th><th scope=col>gwas_immune_class</th><th scope=col>Parent term</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>4451</th><td>ZSWIM7</td><td>Mono</td><td>MONDO_0004979</td><td>17:16138478:A:G</td><td>2089679</td><td>43f7914feab94d31f5f12a641542e0a0</td><td>GCST90399686</td><td>Asthma</td><td>MONDO_0004979</td><td>2ffe954b450599ddcc24bb1ab6b7ceba</td><td>⋯</td><td>0.9999979</td><td>3.300380</td><td>Mono:ZSWIM7:17_16138478_A_G</td><td>TRUE</td><td>GCST90399686</td><td>1</td><td>http://purl.obolibrary.org/obo/MONDO_0004979</td><td>1</td><td>0</td><td>Other disease</td></tr>\n",
       "\t<tr><th scope=row>4452</th><td>ZSWIM7</td><td>NK  </td><td>MONDO_0004979</td><td>17:16100945:A:G</td><td>2089584</td><td>43f7914feab94d31f5f12a641542e0a0</td><td>GCST90399686</td><td>Asthma</td><td>MONDO_0004979</td><td>eaa76075e7dc6cf5475d0c4336d2920c</td><td>⋯</td><td>0.9999896</td><td>3.851711</td><td>NK:ZSWIM7:17_16100945_A_G  </td><td>TRUE</td><td>GCST90399686</td><td>1</td><td>http://purl.obolibrary.org/obo/MONDO_0004979</td><td>1</td><td>0</td><td>Other disease</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 35\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & eqtlTraitFromSource & eQTL\\_CT & efo\\_id & snp\\_id & V1 & leftStudyLocusId & leftStudyId & gwasTraitFromSource & gwasEFO & rightStudyLocusId & ⋯ & h3h4 & H4/H3 & Alt\\_ct\\_gene\\_cs\\_id & Observed & gwasStudyId & SusieRss\\_CS & EFO URI & gwas\\_disease\\_class & gwas\\_immune\\_class & Parent term\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <dbl> & <chr> & <lgl> & <chr> & <int> & <chr> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t4451 & ZSWIM7 & Mono & MONDO\\_0004979 & 17:16138478:A:G & 2089679 & 43f7914feab94d31f5f12a641542e0a0 & GCST90399686 & Asthma & MONDO\\_0004979 & 2ffe954b450599ddcc24bb1ab6b7ceba & ⋯ & 0.9999979 & 3.300380 & Mono:ZSWIM7:17\\_16138478\\_A\\_G & TRUE & GCST90399686 & 1 & http://purl.obolibrary.org/obo/MONDO\\_0004979 & 1 & 0 & Other disease\\\\\n",
       "\t4452 & ZSWIM7 & NK   & MONDO\\_0004979 & 17:16100945:A:G & 2089584 & 43f7914feab94d31f5f12a641542e0a0 & GCST90399686 & Asthma & MONDO\\_0004979 & eaa76075e7dc6cf5475d0c4336d2920c & ⋯ & 0.9999896 & 3.851711 & NK:ZSWIM7:17\\_16100945\\_A\\_G   & TRUE & GCST90399686 & 1 & http://purl.obolibrary.org/obo/MONDO\\_0004979 & 1 & 0 & Other disease\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 35\n",
       "\n",
       "| <!--/--> | eqtlTraitFromSource &lt;chr&gt; | eQTL_CT &lt;chr&gt; | efo_id &lt;chr&gt; | snp_id &lt;chr&gt; | V1 &lt;int&gt; | leftStudyLocusId &lt;chr&gt; | leftStudyId &lt;chr&gt; | gwasTraitFromSource &lt;chr&gt; | gwasEFO &lt;chr&gt; | rightStudyLocusId &lt;chr&gt; | ⋯ ⋯ | h3h4 &lt;dbl&gt; | H4/H3 &lt;dbl&gt; | Alt_ct_gene_cs_id &lt;chr&gt; | Observed &lt;lgl&gt; | gwasStudyId &lt;chr&gt; | SusieRss_CS &lt;int&gt; | EFO URI &lt;chr&gt; | gwas_disease_class &lt;dbl&gt; | gwas_immune_class &lt;dbl&gt; | Parent term &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 4451 | ZSWIM7 | Mono | MONDO_0004979 | 17:16138478:A:G | 2089679 | 43f7914feab94d31f5f12a641542e0a0 | GCST90399686 | Asthma | MONDO_0004979 | 2ffe954b450599ddcc24bb1ab6b7ceba | ⋯ | 0.9999979 | 3.300380 | Mono:ZSWIM7:17_16138478_A_G | TRUE | GCST90399686 | 1 | http://purl.obolibrary.org/obo/MONDO_0004979 | 1 | 0 | Other disease |\n",
       "| 4452 | ZSWIM7 | NK   | MONDO_0004979 | 17:16100945:A:G | 2089584 | 43f7914feab94d31f5f12a641542e0a0 | GCST90399686 | Asthma | MONDO_0004979 | eaa76075e7dc6cf5475d0c4336d2920c | ⋯ | 0.9999896 | 3.851711 | NK:ZSWIM7:17_16100945_A_G   | TRUE | GCST90399686 | 1 | http://purl.obolibrary.org/obo/MONDO_0004979 | 1 | 0 | Other disease |\n",
       "\n"
      ],
      "text/plain": [
       "     eqtlTraitFromSource eQTL_CT efo_id        snp_id          V1     \n",
       "4451 ZSWIM7              Mono    MONDO_0004979 17:16138478:A:G 2089679\n",
       "4452 ZSWIM7              NK      MONDO_0004979 17:16100945:A:G 2089584\n",
       "     leftStudyLocusId                 leftStudyId  gwasTraitFromSource\n",
       "4451 43f7914feab94d31f5f12a641542e0a0 GCST90399686 Asthma             \n",
       "4452 43f7914feab94d31f5f12a641542e0a0 GCST90399686 Asthma             \n",
       "     gwasEFO       rightStudyLocusId                ⋯ h3h4      H4/H3   \n",
       "4451 MONDO_0004979 2ffe954b450599ddcc24bb1ab6b7ceba ⋯ 0.9999979 3.300380\n",
       "4452 MONDO_0004979 eaa76075e7dc6cf5475d0c4336d2920c ⋯ 0.9999896 3.851711\n",
       "     Alt_ct_gene_cs_id           Observed gwasStudyId  SusieRss_CS\n",
       "4451 Mono:ZSWIM7:17_16138478_A_G TRUE     GCST90399686 1          \n",
       "4452 NK:ZSWIM7:17_16100945_A_G   TRUE     GCST90399686 1          \n",
       "     EFO URI                                      gwas_disease_class\n",
       "4451 http://purl.obolibrary.org/obo/MONDO_0004979 1                 \n",
       "4452 http://purl.obolibrary.org/obo/MONDO_0004979 1                 \n",
       "     gwas_immune_class Parent term  \n",
       "4451 0                 Other disease\n",
       "4452 0                 Other disease"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tail(analysis_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 523,
   "id": "d0806ac5-7b3e-45b2-a78c-b6e03d89d0ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_coeqtls$is_coeqtl = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "id": "9f53374b-6eb0-4a61-ad81-61e653cd587e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4452"
      ],
      "text/latex": [
       "4452"
      ],
      "text/markdown": [
       "4452"
      ],
      "text/plain": [
       "[1] 4452"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(analysis_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "id": "b480e319-de10-40e2-ac15-639e1f55e733",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data= merge(analysis_data, unique(sign_coeqtls[,c('eGene', 'cell_type', 'is_coeqtl')]), \n",
    "                     by.x = c('eqtlTraitFromSource', 'eQTL_CT'), by.y = c('eGene', 'cell_type'), all.x = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 526,
   "id": "8234d8c0-b4fa-4cac-ad79-3684a7612e70",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data$is_coeqtl[is.na(analysis_data$is_coeqtl)] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 527,
   "id": "e96bf5ae-b2d4-4595-b06a-cf8e7e211876",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4452"
      ],
      "text/latex": [
       "4452"
      ],
      "text/markdown": [
       "4452"
      ],
      "text/plain": [
       "[1] 4452"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(analysis_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 528,
   "id": "8c7b97dc-4ee6-47e4-9a68-c4a69cb0fb3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "47"
      ],
      "text/latex": [
       "47"
      ],
      "text/markdown": [
       "47"
      ],
      "text/plain": [
       "[1] 47"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(analysis_data$eqtlTraitFromSource[analysis_data$is_coeqtl == 1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 529,
   "id": "0302623c-060e-4054-a9c7-b6add11f7ea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Summarise amount of colocalising eGenes (by whether it is a co-eqtl or not)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "id": "e1f8821e-930a-45fd-a7ae-2c409c1b0614",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'eQTL_CT'. You can override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "coloc_summary = analysis_data %>% group_by(eQTL_CT, is_coeqtl) %>% dplyr::summarise(amount_coloc = length(unique(eqtlTraitFromSource)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 531,
   "id": "a561ced0-001c-4cee-bfd9-ce681e7e32e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 10 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>eQTL_CT</th><th scope=col>is_coeqtl</th><th scope=col>amount_coloc</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td>0</td><td>147</td></tr>\n",
       "\t<tr><td>B    </td><td>1</td><td>  4</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>0</td><td>399</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>1</td><td> 27</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>0</td><td>218</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>1</td><td> 17</td></tr>\n",
       "\t<tr><td>Mono </td><td>0</td><td>194</td></tr>\n",
       "\t<tr><td>Mono </td><td>1</td><td>  7</td></tr>\n",
       "\t<tr><td>NK   </td><td>0</td><td>216</td></tr>\n",
       "\t<tr><td>NK   </td><td>1</td><td>  7</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 10 × 3\n",
       "\\begin{tabular}{lll}\n",
       " eQTL\\_CT & is\\_coeqtl & amount\\_coloc\\\\\n",
       " <chr> & <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t B     & 0 & 147\\\\\n",
       "\t B     & 1 &   4\\\\\n",
       "\t CD4\\_T & 0 & 399\\\\\n",
       "\t CD4\\_T & 1 &  27\\\\\n",
       "\t CD8\\_T & 0 & 218\\\\\n",
       "\t CD8\\_T & 1 &  17\\\\\n",
       "\t Mono  & 0 & 194\\\\\n",
       "\t Mono  & 1 &   7\\\\\n",
       "\t NK    & 0 & 216\\\\\n",
       "\t NK    & 1 &   7\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 10 × 3\n",
       "\n",
       "| eQTL_CT &lt;chr&gt; | is_coeqtl &lt;dbl&gt; | amount_coloc &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| B     | 0 | 147 |\n",
       "| B     | 1 |   4 |\n",
       "| CD4_T | 0 | 399 |\n",
       "| CD4_T | 1 |  27 |\n",
       "| CD8_T | 0 | 218 |\n",
       "| CD8_T | 1 |  17 |\n",
       "| Mono  | 0 | 194 |\n",
       "| Mono  | 1 |   7 |\n",
       "| NK    | 0 | 216 |\n",
       "| NK    | 1 |   7 |\n",
       "\n"
      ],
      "text/plain": [
       "   eQTL_CT is_coeqtl amount_coloc\n",
       "1  B       0         147         \n",
       "2  B       1           4         \n",
       "3  CD4_T   0         399         \n",
       "4  CD4_T   1          27         \n",
       "5  CD8_T   0         218         \n",
       "6  CD8_T   1          17         \n",
       "7  Mono    0         194         \n",
       "8  Mono    1           7         \n",
       "9  NK      0         216         \n",
       "10 NK      1           7         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 532,
   "id": "8348523d-839a-468c-9b57-be10cc73c9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge information about total amounts of eGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "id": "89fce804-270b-4317-be68-14eb3c7d44ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_amount_info = merge(coeqtl_egenes, egenes_cell_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 534,
   "id": "3035bab4-37d7-4040-a1b0-5885d437855f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>amount_coeqtl_egenes</th><th scope=col>amount_egenes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td> 32</td><td>1679</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>152</td><td>3676</td></tr>\n",
       "\t<tr><td>CD8_T</td><td> 76</td><td>2019</td></tr>\n",
       "\t<tr><td>Mono </td><td> 50</td><td>1748</td></tr>\n",
       "\t<tr><td>NK   </td><td> 40</td><td>1914</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 3\n",
       "\\begin{tabular}{lll}\n",
       " cell\\_type & amount\\_coeqtl\\_egenes & amount\\_egenes\\\\\n",
       " <chr> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t B     &  32 & 1679\\\\\n",
       "\t CD4\\_T & 152 & 3676\\\\\n",
       "\t CD8\\_T &  76 & 2019\\\\\n",
       "\t Mono  &  50 & 1748\\\\\n",
       "\t NK    &  40 & 1914\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 3\n",
       "\n",
       "| cell_type &lt;chr&gt; | amount_coeqtl_egenes &lt;int&gt; | amount_egenes &lt;int&gt; |\n",
       "|---|---|---|\n",
       "| B     |  32 | 1679 |\n",
       "| CD4_T | 152 | 3676 |\n",
       "| CD8_T |  76 | 2019 |\n",
       "| Mono  |  50 | 1748 |\n",
       "| NK    |  40 | 1914 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type amount_coeqtl_egenes amount_egenes\n",
       "1 B          32                  1679         \n",
       "2 CD4_T     152                  3676         \n",
       "3 CD8_T      76                  2019         \n",
       "4 Mono       50                  1748         \n",
       "5 NK         40                  1914         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "total_amount_info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "id": "41cb86fc-8090-484d-a3c8-fce3c0e6c6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_amount_info$amount_non_coeqtl_egenes = total_amount_info$amount_egenes - total_amount_info$amount_coeqtl_egenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "id": "9c0bb24f-fb4e-40b8-8597-1c50480079ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_amount_info$amount_egenes = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "87750843-67fb-4997-bbb4-7cb22364d8be",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using cell_type as id variables\n",
      "\n"
     ]
    }
   ],
   "source": [
    "total_amount_info = melt(total_amount_info )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "id": "228ffc49-e8d9-40c1-bf1f-e71769304227",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_amount_info$is_coeqtl = 0\n",
    "total_amount_info$is_coeqtl[total_amount_info$variable == 'amount_coeqtl_egenes'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "id": "27f208fc-3a55-4684-a0b1-d1919a6d1454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>is_coeqtl</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td>amount_coeqtl_egenes    </td><td>  32</td><td>1</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>amount_coeqtl_egenes    </td><td> 152</td><td>1</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>amount_coeqtl_egenes    </td><td>  76</td><td>1</td></tr>\n",
       "\t<tr><td>Mono </td><td>amount_coeqtl_egenes    </td><td>  50</td><td>1</td></tr>\n",
       "\t<tr><td>NK   </td><td>amount_coeqtl_egenes    </td><td>  40</td><td>1</td></tr>\n",
       "\t<tr><td>B    </td><td>amount_non_coeqtl_egenes</td><td>1647</td><td>0</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>amount_non_coeqtl_egenes</td><td>3524</td><td>0</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>amount_non_coeqtl_egenes</td><td>1943</td><td>0</td></tr>\n",
       "\t<tr><td>Mono </td><td>amount_non_coeqtl_egenes</td><td>1698</td><td>0</td></tr>\n",
       "\t<tr><td>NK   </td><td>amount_non_coeqtl_egenes</td><td>1874</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 4\n",
       "\\begin{tabular}{llll}\n",
       " cell\\_type & variable & value & is\\_coeqtl\\\\\n",
       " <chr> & <fct> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t B     & amount\\_coeqtl\\_egenes     &   32 & 1\\\\\n",
       "\t CD4\\_T & amount\\_coeqtl\\_egenes     &  152 & 1\\\\\n",
       "\t CD8\\_T & amount\\_coeqtl\\_egenes     &   76 & 1\\\\\n",
       "\t Mono  & amount\\_coeqtl\\_egenes     &   50 & 1\\\\\n",
       "\t NK    & amount\\_coeqtl\\_egenes     &   40 & 1\\\\\n",
       "\t B     & amount\\_non\\_coeqtl\\_egenes & 1647 & 0\\\\\n",
       "\t CD4\\_T & amount\\_non\\_coeqtl\\_egenes & 3524 & 0\\\\\n",
       "\t CD8\\_T & amount\\_non\\_coeqtl\\_egenes & 1943 & 0\\\\\n",
       "\t Mono  & amount\\_non\\_coeqtl\\_egenes & 1698 & 0\\\\\n",
       "\t NK    & amount\\_non\\_coeqtl\\_egenes & 1874 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 4\n",
       "\n",
       "| cell_type &lt;chr&gt; | variable &lt;fct&gt; | value &lt;int&gt; | is_coeqtl &lt;dbl&gt; |\n",
       "|---|---|---|---|\n",
       "| B     | amount_coeqtl_egenes     |   32 | 1 |\n",
       "| CD4_T | amount_coeqtl_egenes     |  152 | 1 |\n",
       "| CD8_T | amount_coeqtl_egenes     |   76 | 1 |\n",
       "| Mono  | amount_coeqtl_egenes     |   50 | 1 |\n",
       "| NK    | amount_coeqtl_egenes     |   40 | 1 |\n",
       "| B     | amount_non_coeqtl_egenes | 1647 | 0 |\n",
       "| CD4_T | amount_non_coeqtl_egenes | 3524 | 0 |\n",
       "| CD8_T | amount_non_coeqtl_egenes | 1943 | 0 |\n",
       "| Mono  | amount_non_coeqtl_egenes | 1698 | 0 |\n",
       "| NK    | amount_non_coeqtl_egenes | 1874 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "   cell_type variable                 value is_coeqtl\n",
       "1  B         amount_coeqtl_egenes       32  1        \n",
       "2  CD4_T     amount_coeqtl_egenes      152  1        \n",
       "3  CD8_T     amount_coeqtl_egenes       76  1        \n",
       "4  Mono      amount_coeqtl_egenes       50  1        \n",
       "5  NK        amount_coeqtl_egenes       40  1        \n",
       "6  B         amount_non_coeqtl_egenes 1647  0        \n",
       "7  CD4_T     amount_non_coeqtl_egenes 3524  0        \n",
       "8  CD8_T     amount_non_coeqtl_egenes 1943  0        \n",
       "9  Mono      amount_non_coeqtl_egenes 1698  0        \n",
       "10 NK        amount_non_coeqtl_egenes 1874  0        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "total_amount_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "id": "24fc2d94-662c-4ae7-b875-f4ac5adad0e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge to coloc summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "id": "def73f30-e1dd-4969-9d1f-6049fea3c85d",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary = merge(coloc_summary, total_amount_info, by.x = c('eQTL_CT', 'is_coeqtl'), \n",
    "                     by.y = c('cell_type', 'is_coeqtl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 542,
   "id": "9c8f55be-d575-45d6-88a4-dbb1aaa8811c",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary$amount_non_coloc = coloc_summary$value - coloc_summary$amount_coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 543,
   "id": "061ebd7d-5729-4b91-91b0-e3455c14aab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary = coloc_summary %>% arrange(-is_coeqtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "id": "6b1a4a62-42bd-4760-9365-406427bc31f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>eQTL_CT</th><th scope=col>is_coeqtl</th><th scope=col>amount_coloc</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>amount_non_coloc</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td>1</td><td>  4</td><td>amount_coeqtl_egenes    </td><td>  32</td><td>  28</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>1</td><td> 27</td><td>amount_coeqtl_egenes    </td><td> 152</td><td> 125</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>1</td><td> 17</td><td>amount_coeqtl_egenes    </td><td>  76</td><td>  59</td></tr>\n",
       "\t<tr><td>Mono </td><td>1</td><td>  7</td><td>amount_coeqtl_egenes    </td><td>  50</td><td>  43</td></tr>\n",
       "\t<tr><td>NK   </td><td>1</td><td>  7</td><td>amount_coeqtl_egenes    </td><td>  40</td><td>  33</td></tr>\n",
       "\t<tr><td>B    </td><td>0</td><td>147</td><td>amount_non_coeqtl_egenes</td><td>1647</td><td>1500</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>0</td><td>399</td><td>amount_non_coeqtl_egenes</td><td>3524</td><td>3125</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>0</td><td>218</td><td>amount_non_coeqtl_egenes</td><td>1943</td><td>1725</td></tr>\n",
       "\t<tr><td>Mono </td><td>0</td><td>194</td><td>amount_non_coeqtl_egenes</td><td>1698</td><td>1504</td></tr>\n",
       "\t<tr><td>NK   </td><td>0</td><td>216</td><td>amount_non_coeqtl_egenes</td><td>1874</td><td>1658</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " eQTL\\_CT & is\\_coeqtl & amount\\_coloc & variable & value & amount\\_non\\_coloc\\\\\n",
       " <chr> & <dbl> & <int> & <fct> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t B     & 1 &   4 & amount\\_coeqtl\\_egenes     &   32 &   28\\\\\n",
       "\t CD4\\_T & 1 &  27 & amount\\_coeqtl\\_egenes     &  152 &  125\\\\\n",
       "\t CD8\\_T & 1 &  17 & amount\\_coeqtl\\_egenes     &   76 &   59\\\\\n",
       "\t Mono  & 1 &   7 & amount\\_coeqtl\\_egenes     &   50 &   43\\\\\n",
       "\t NK    & 1 &   7 & amount\\_coeqtl\\_egenes     &   40 &   33\\\\\n",
       "\t B     & 0 & 147 & amount\\_non\\_coeqtl\\_egenes & 1647 & 1500\\\\\n",
       "\t CD4\\_T & 0 & 399 & amount\\_non\\_coeqtl\\_egenes & 3524 & 3125\\\\\n",
       "\t CD8\\_T & 0 & 218 & amount\\_non\\_coeqtl\\_egenes & 1943 & 1725\\\\\n",
       "\t Mono  & 0 & 194 & amount\\_non\\_coeqtl\\_egenes & 1698 & 1504\\\\\n",
       "\t NK    & 0 & 216 & amount\\_non\\_coeqtl\\_egenes & 1874 & 1658\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 6\n",
       "\n",
       "| eQTL_CT &lt;chr&gt; | is_coeqtl &lt;dbl&gt; | amount_coloc &lt;int&gt; | variable &lt;fct&gt; | value &lt;int&gt; | amount_non_coloc &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| B     | 1 |   4 | amount_coeqtl_egenes     |   32 |   28 |\n",
       "| CD4_T | 1 |  27 | amount_coeqtl_egenes     |  152 |  125 |\n",
       "| CD8_T | 1 |  17 | amount_coeqtl_egenes     |   76 |   59 |\n",
       "| Mono  | 1 |   7 | amount_coeqtl_egenes     |   50 |   43 |\n",
       "| NK    | 1 |   7 | amount_coeqtl_egenes     |   40 |   33 |\n",
       "| B     | 0 | 147 | amount_non_coeqtl_egenes | 1647 | 1500 |\n",
       "| CD4_T | 0 | 399 | amount_non_coeqtl_egenes | 3524 | 3125 |\n",
       "| CD8_T | 0 | 218 | amount_non_coeqtl_egenes | 1943 | 1725 |\n",
       "| Mono  | 0 | 194 | amount_non_coeqtl_egenes | 1698 | 1504 |\n",
       "| NK    | 0 | 216 | amount_non_coeqtl_egenes | 1874 | 1658 |\n",
       "\n"
      ],
      "text/plain": [
       "   eQTL_CT is_coeqtl amount_coloc variable                 value\n",
       "1  B       1           4          amount_coeqtl_egenes       32 \n",
       "2  CD4_T   1          27          amount_coeqtl_egenes      152 \n",
       "3  CD8_T   1          17          amount_coeqtl_egenes       76 \n",
       "4  Mono    1           7          amount_coeqtl_egenes       50 \n",
       "5  NK      1           7          amount_coeqtl_egenes       40 \n",
       "6  B       0         147          amount_non_coeqtl_egenes 1647 \n",
       "7  CD4_T   0         399          amount_non_coeqtl_egenes 3524 \n",
       "8  CD8_T   0         218          amount_non_coeqtl_egenes 1943 \n",
       "9  Mono    0         194          amount_non_coeqtl_egenes 1698 \n",
       "10 NK      0         216          amount_non_coeqtl_egenes 1874 \n",
       "   amount_non_coloc\n",
       "1    28            \n",
       "2   125            \n",
       "3    59            \n",
       "4    43            \n",
       "5    33            \n",
       "6  1500            \n",
       "7  3125            \n",
       "8  1725            \n",
       "9  1504            \n",
       "10 1658            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_summary "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "81aeaaf3-6aac-417d-84ec-ad3c3f816ff7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 10 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>eQTL_CT</th><th scope=col>is_coeqtl</th><th scope=col>amount_coloc</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>amount_non_coloc</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>B    </td><td>1</td><td>  4</td><td>amount_coeqtl_egenes    </td><td>  32</td><td>  28</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>1</td><td> 27</td><td>amount_coeqtl_egenes    </td><td> 152</td><td> 125</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>1</td><td> 17</td><td>amount_coeqtl_egenes    </td><td>  76</td><td>  59</td></tr>\n",
       "\t<tr><td>Mono </td><td>1</td><td>  7</td><td>amount_coeqtl_egenes    </td><td>  50</td><td>  43</td></tr>\n",
       "\t<tr><td>NK   </td><td>1</td><td>  7</td><td>amount_coeqtl_egenes    </td><td>  40</td><td>  33</td></tr>\n",
       "\t<tr><td>B    </td><td>0</td><td>147</td><td>amount_non_coeqtl_egenes</td><td>1647</td><td>1500</td></tr>\n",
       "\t<tr><td>CD4_T</td><td>0</td><td>399</td><td>amount_non_coeqtl_egenes</td><td>3524</td><td>3125</td></tr>\n",
       "\t<tr><td>CD8_T</td><td>0</td><td>218</td><td>amount_non_coeqtl_egenes</td><td>1943</td><td>1725</td></tr>\n",
       "\t<tr><td>Mono </td><td>0</td><td>194</td><td>amount_non_coeqtl_egenes</td><td>1698</td><td>1504</td></tr>\n",
       "\t<tr><td>NK   </td><td>0</td><td>216</td><td>amount_non_coeqtl_egenes</td><td>1874</td><td>1658</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 10 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " eQTL\\_CT & is\\_coeqtl & amount\\_coloc & variable & value & amount\\_non\\_coloc\\\\\n",
       " <chr> & <dbl> & <int> & <fct> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t B     & 1 &   4 & amount\\_coeqtl\\_egenes     &   32 &   28\\\\\n",
       "\t CD4\\_T & 1 &  27 & amount\\_coeqtl\\_egenes     &  152 &  125\\\\\n",
       "\t CD8\\_T & 1 &  17 & amount\\_coeqtl\\_egenes     &   76 &   59\\\\\n",
       "\t Mono  & 1 &   7 & amount\\_coeqtl\\_egenes     &   50 &   43\\\\\n",
       "\t NK    & 1 &   7 & amount\\_coeqtl\\_egenes     &   40 &   33\\\\\n",
       "\t B     & 0 & 147 & amount\\_non\\_coeqtl\\_egenes & 1647 & 1500\\\\\n",
       "\t CD4\\_T & 0 & 399 & amount\\_non\\_coeqtl\\_egenes & 3524 & 3125\\\\\n",
       "\t CD8\\_T & 0 & 218 & amount\\_non\\_coeqtl\\_egenes & 1943 & 1725\\\\\n",
       "\t Mono  & 0 & 194 & amount\\_non\\_coeqtl\\_egenes & 1698 & 1504\\\\\n",
       "\t NK    & 0 & 216 & amount\\_non\\_coeqtl\\_egenes & 1874 & 1658\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 10 × 6\n",
       "\n",
       "| eQTL_CT &lt;chr&gt; | is_coeqtl &lt;dbl&gt; | amount_coloc &lt;int&gt; | variable &lt;fct&gt; | value &lt;int&gt; | amount_non_coloc &lt;int&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| B     | 1 |   4 | amount_coeqtl_egenes     |   32 |   28 |\n",
       "| CD4_T | 1 |  27 | amount_coeqtl_egenes     |  152 |  125 |\n",
       "| CD8_T | 1 |  17 | amount_coeqtl_egenes     |   76 |   59 |\n",
       "| Mono  | 1 |   7 | amount_coeqtl_egenes     |   50 |   43 |\n",
       "| NK    | 1 |   7 | amount_coeqtl_egenes     |   40 |   33 |\n",
       "| B     | 0 | 147 | amount_non_coeqtl_egenes | 1647 | 1500 |\n",
       "| CD4_T | 0 | 399 | amount_non_coeqtl_egenes | 3524 | 3125 |\n",
       "| CD8_T | 0 | 218 | amount_non_coeqtl_egenes | 1943 | 1725 |\n",
       "| Mono  | 0 | 194 | amount_non_coeqtl_egenes | 1698 | 1504 |\n",
       "| NK    | 0 | 216 | amount_non_coeqtl_egenes | 1874 | 1658 |\n",
       "\n"
      ],
      "text/plain": [
       "   eQTL_CT is_coeqtl amount_coloc variable                 value\n",
       "1  B       1           4          amount_coeqtl_egenes       32 \n",
       "2  CD4_T   1          27          amount_coeqtl_egenes      152 \n",
       "3  CD8_T   1          17          amount_coeqtl_egenes       76 \n",
       "4  Mono    1           7          amount_coeqtl_egenes       50 \n",
       "5  NK      1           7          amount_coeqtl_egenes       40 \n",
       "6  B       0         147          amount_non_coeqtl_egenes 1647 \n",
       "7  CD4_T   0         399          amount_non_coeqtl_egenes 3524 \n",
       "8  CD8_T   0         218          amount_non_coeqtl_egenes 1943 \n",
       "9  Mono    0         194          amount_non_coeqtl_egenes 1698 \n",
       "10 NK      0         216          amount_non_coeqtl_egenes 1874 \n",
       "   amount_non_coloc\n",
       "1    28            \n",
       "2   125            \n",
       "3    59            \n",
       "4    43            \n",
       "5    33            \n",
       "6  1500            \n",
       "7  3125            \n",
       "8  1725            \n",
       "9  1504            \n",
       "10 1658            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_summary "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30cca428-8dcc-4b9e-bad4-d868284952a1",
   "metadata": {},
   "source": [
    "# Check Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 546,
   "id": "b4435c97-fdb9-4772-807b-eddbf559f090",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform fisher test for enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "id": "7f7f16d9-2541-436f-b2b9-2d87592c46fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "contingency_table_data = data.frame()\n",
    "odds_ratio_result = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "id": "972537d5-aba6-436a-89df-df458d6f131c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#str(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "id": "a1344453-970f-4982-8750-05475341e763",
   "metadata": {},
   "outputs": [],
   "source": [
    "#contingency_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 550,
   "id": "6e7e4ac1-4eaf-4e9b-9e46-c33176a9f709",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  amount_coloc amount_non_coloc\n",
      "1            4               28\n",
      "6          147             1500\n",
      "[1] \"B\"\n",
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  contingency_table\n",
      "p-value = 0.5246\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 0.3664286 4.2480515\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  1.457338 \n",
      "\n",
      "  amount_coloc amount_non_coloc\n",
      "2           27              125\n",
      "7          399             3125\n",
      "[1] \"CD4_T\"\n",
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  contingency_table\n",
      "p-value = 0.01952\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 1.058339 2.619116\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  1.691435 \n",
      "\n",
      "  amount_coloc amount_non_coloc\n",
      "3           17               59\n",
      "8          218             1725\n",
      "[1] \"CD8_T\"\n",
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  contingency_table\n",
      "p-value = 0.00572\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 1.222246 4.049781\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  2.278742 \n",
      "\n",
      "  amount_coloc amount_non_coloc\n",
      "4            7               43\n",
      "9          194             1504\n",
      "[1] \"Mono\"\n",
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  contingency_table\n",
      "p-value = 0.5044\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 0.4722654 2.8821123\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  1.261864 \n",
      "\n",
      "   amount_coloc amount_non_coloc\n",
      "5             7               33\n",
      "10          216             1658\n",
      "[1] \"NK\"\n",
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  contingency_table\n",
      "p-value = 0.219\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 0.6002015 3.8023157\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  1.627725 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for(i in unique(coloc_summary$eQTL_CT)){\n",
    "    contingency_table = coloc_summary[coloc_summary$eQTL_CT == i, c('amount_coloc', 'amount_non_coloc', 'eQTL_CT')]\n",
    "    if(nrow(contingency_table) == 2){\n",
    "    contingency_table$eQTL_CT = NULL\n",
    "    print(contingency_table)\n",
    "    contingency_table = as.matrix (contingency_table)\n",
    "    \n",
    "    result = fisher.test(contingency_table)\n",
    "    print(i)\n",
    "    print(result)\n",
    "        \n",
    "    ## save the data\n",
    "    contingency_table = as.data.frame(contingency_table)\n",
    "    contingency_table$is_coeqtl = c(1,0)\n",
    "    contingency_table$cell_type = i\n",
    "    contingency_table$trait_filter = use_gwas\n",
    "    contingency_table$only_dt = only_dt\n",
    "    contingency_table$remove_HLA = remove_HLA\n",
    "    contingency_table$only_esnps = only_esnps\n",
    "    contingency_table$coloc_thres = coloc_thres\n",
    "    contingency_table_data = rbind(contingency_table_data, contingency_table)\n",
    "        \n",
    "    or_result = data.frame(p_value = result$p.value,\n",
    "                           estimate = result$estimate,\n",
    "                           conf_upper = result$conf.int[2],\n",
    "                           conf_lower = result$conf.int[1])\n",
    "    or_result$cell_type = i\n",
    "    or_result$trait_filter =use_gwas\n",
    "    or_result$only_dt = only_dt\n",
    "    or_result$remove_HLA = remove_HLA\n",
    "    or_result$only_esnps = only_esnps\n",
    "    or_result$coloc_thres = coloc_thres\n",
    "    odds_ratio_result = rbind(odds_ratio_result, or_result)\n",
    "        \n",
    "    \n",
    "    \n",
    "    }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "id": "dbc45c80-6990-4e31-8827-2e65e595c07f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>cell_type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>odds ratio2</th><td>0.005719577</td><td>2.278742</td><td>CD8_T</td></tr>\n",
       "\t<tr><th scope=row>odds ratio1</th><td>0.019517062</td><td>1.691435</td><td>CD4_T</td></tr>\n",
       "\t<tr><th scope=row>odds ratio4</th><td>0.218986969</td><td>1.627725</td><td>NK   </td></tr>\n",
       "\t<tr><th scope=row>odds ratio3</th><td>0.504399605</td><td>1.261864</td><td>Mono </td></tr>\n",
       "\t<tr><th scope=row>odds ratio</th><td>0.524644885</td><td>1.457338</td><td>B    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & p\\_value & estimate & cell\\_type\\\\\n",
       "  & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\todds ratio2 & 0.005719577 & 2.278742 & CD8\\_T\\\\\n",
       "\todds ratio1 & 0.019517062 & 1.691435 & CD4\\_T\\\\\n",
       "\todds ratio4 & 0.218986969 & 1.627725 & NK   \\\\\n",
       "\todds ratio3 & 0.504399605 & 1.261864 & Mono \\\\\n",
       "\todds ratio & 0.524644885 & 1.457338 & B    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 3\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | cell_type &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| odds ratio2 | 0.005719577 | 2.278742 | CD8_T |\n",
       "| odds ratio1 | 0.019517062 | 1.691435 | CD4_T |\n",
       "| odds ratio4 | 0.218986969 | 1.627725 | NK    |\n",
       "| odds ratio3 | 0.504399605 | 1.261864 | Mono  |\n",
       "| odds ratio | 0.524644885 | 1.457338 | B     |\n",
       "\n"
      ],
      "text/plain": [
       "            p_value     estimate cell_type\n",
       "odds ratio2 0.005719577 2.278742 CD8_T    \n",
       "odds ratio1 0.019517062 1.691435 CD4_T    \n",
       "odds ratio4 0.218986969 1.627725 NK       \n",
       "odds ratio3 0.504399605 1.261864 Mono     \n",
       "odds ratio  0.524644885 1.457338 B        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    " odds_ratio_result[,c('p_value', 'estimate', 'cell_type')] %>% arrange(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "id": "ac5daf39-f1ae-4a16-b777-dc49a05717d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(odds_ratio_result, paste0(result_path, '/G2/Odds_Ratio_result',use_gwas , only_dt, remove_HLA, only_esnps , coloc_thres, '.csv'))\n",
    "write.csv(    contingency_table_data , paste0(result_path, '/G2/Contingency_Table', use_gwas, only_dt, remove_HLA, only_esnps , coloc_thres, '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 553,
   "id": "8f5459a1-1d3e-4973-87a6-80737c9c75e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check enrichment across all cell-types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 554,
   "id": "b13f0229-735f-48e6-8db0-31367691ba1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Subset colocalisation data to eGene with lowes p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "id": "5732c347-5b5e-4bb3-89bf-e9bbb4660852",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 36</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>eqtlTraitFromSource</th><th scope=col>eQTL_CT</th><th scope=col>efo_id</th><th scope=col>snp_id</th><th scope=col>V1</th><th scope=col>leftStudyLocusId</th><th scope=col>leftStudyId</th><th scope=col>gwasTraitFromSource</th><th scope=col>gwasEFO</th><th scope=col>rightStudyLocusId</th><th scope=col>⋯</th><th scope=col>H4/H3</th><th scope=col>Alt_ct_gene_cs_id</th><th scope=col>Observed</th><th scope=col>gwasStudyId</th><th scope=col>SusieRss_CS</th><th scope=col>EFO URI</th><th scope=col>gwas_disease_class</th><th scope=col>gwas_immune_class</th><th scope=col>Parent term</th><th scope=col>is_coeqtl</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A2M    </td><td>CD4_T</td><td>MONDO_0005147</td><td>12:8972169:C:T</td><td>156</td><td>a95199d25d084b697b9446af089bd878</td><td>FINNGEN_R12_T1D</td><td>Type 1 diabetes, definitions combined</td><td>MONDO_0005147</td><td>ea74fe3c38c7b9c4d67a414a08b7f310</td><td>⋯</td><td>14.961340</td><td>CD4_T:A2M:12_8972169_C_T    </td><td>TRUE</td><td>FINNGEN_R12_T1D</td><td>2</td><td>http://purl.obolibrary.org/obo/MONDO_0005147</td><td>1</td><td>1</td><td>Immune system disorder</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A2M-AS1</td><td>CD4_T</td><td>MONDO_0005147</td><td>12:8959409:C:T</td><td>307</td><td>a95199d25d084b697b9446af089bd878</td><td>FINNGEN_R12_T1D</td><td>Type 1 diabetes, definitions combined</td><td>MONDO_0005147</td><td>5753696279250a3cd12034956f8c20d1</td><td>⋯</td><td> 8.645992</td><td>CD4_T:A2M-AS1:12_8959409_C_T</td><td>TRUE</td><td>FINNGEN_R12_T1D</td><td>2</td><td>http://purl.obolibrary.org/obo/MONDO_0005147</td><td>1</td><td>1</td><td>Immune system disorder</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 36\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & eqtlTraitFromSource & eQTL\\_CT & efo\\_id & snp\\_id & V1 & leftStudyLocusId & leftStudyId & gwasTraitFromSource & gwasEFO & rightStudyLocusId & ⋯ & H4/H3 & Alt\\_ct\\_gene\\_cs\\_id & Observed & gwasStudyId & SusieRss\\_CS & EFO URI & gwas\\_disease\\_class & gwas\\_immune\\_class & Parent term & is\\_coeqtl\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & ⋯ & <dbl> & <chr> & <lgl> & <chr> & <int> & <chr> & <dbl> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & A2M     & CD4\\_T & MONDO\\_0005147 & 12:8972169:C:T & 156 & a95199d25d084b697b9446af089bd878 & FINNGEN\\_R12\\_T1D & Type 1 diabetes, definitions combined & MONDO\\_0005147 & ea74fe3c38c7b9c4d67a414a08b7f310 & ⋯ & 14.961340 & CD4\\_T:A2M:12\\_8972169\\_C\\_T     & TRUE & FINNGEN\\_R12\\_T1D & 2 & http://purl.obolibrary.org/obo/MONDO\\_0005147 & 1 & 1 & Immune system disorder & 1\\\\\n",
       "\t2 & A2M-AS1 & CD4\\_T & MONDO\\_0005147 & 12:8959409:C:T & 307 & a95199d25d084b697b9446af089bd878 & FINNGEN\\_R12\\_T1D & Type 1 diabetes, definitions combined & MONDO\\_0005147 & 5753696279250a3cd12034956f8c20d1 & ⋯ &  8.645992 & CD4\\_T:A2M-AS1:12\\_8959409\\_C\\_T & TRUE & FINNGEN\\_R12\\_T1D & 2 & http://purl.obolibrary.org/obo/MONDO\\_0005147 & 1 & 1 & Immune system disorder & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 36\n",
       "\n",
       "| <!--/--> | eqtlTraitFromSource &lt;chr&gt; | eQTL_CT &lt;chr&gt; | efo_id &lt;chr&gt; | snp_id &lt;chr&gt; | V1 &lt;int&gt; | leftStudyLocusId &lt;chr&gt; | leftStudyId &lt;chr&gt; | gwasTraitFromSource &lt;chr&gt; | gwasEFO &lt;chr&gt; | rightStudyLocusId &lt;chr&gt; | ⋯ ⋯ | H4/H3 &lt;dbl&gt; | Alt_ct_gene_cs_id &lt;chr&gt; | Observed &lt;lgl&gt; | gwasStudyId &lt;chr&gt; | SusieRss_CS &lt;int&gt; | EFO URI &lt;chr&gt; | gwas_disease_class &lt;dbl&gt; | gwas_immune_class &lt;dbl&gt; | Parent term &lt;chr&gt; | is_coeqtl &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | A2M     | CD4_T | MONDO_0005147 | 12:8972169:C:T | 156 | a95199d25d084b697b9446af089bd878 | FINNGEN_R12_T1D | Type 1 diabetes, definitions combined | MONDO_0005147 | ea74fe3c38c7b9c4d67a414a08b7f310 | ⋯ | 14.961340 | CD4_T:A2M:12_8972169_C_T     | TRUE | FINNGEN_R12_T1D | 2 | http://purl.obolibrary.org/obo/MONDO_0005147 | 1 | 1 | Immune system disorder | 1 |\n",
       "| 2 | A2M-AS1 | CD4_T | MONDO_0005147 | 12:8959409:C:T | 307 | a95199d25d084b697b9446af089bd878 | FINNGEN_R12_T1D | Type 1 diabetes, definitions combined | MONDO_0005147 | 5753696279250a3cd12034956f8c20d1 | ⋯ |  8.645992 | CD4_T:A2M-AS1:12_8959409_C_T | TRUE | FINNGEN_R12_T1D | 2 | http://purl.obolibrary.org/obo/MONDO_0005147 | 1 | 1 | Immune system disorder | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  eqtlTraitFromSource eQTL_CT efo_id        snp_id         V1 \n",
       "1 A2M                 CD4_T   MONDO_0005147 12:8972169:C:T 156\n",
       "2 A2M-AS1             CD4_T   MONDO_0005147 12:8959409:C:T 307\n",
       "  leftStudyLocusId                 leftStudyId    \n",
       "1 a95199d25d084b697b9446af089bd878 FINNGEN_R12_T1D\n",
       "2 a95199d25d084b697b9446af089bd878 FINNGEN_R12_T1D\n",
       "  gwasTraitFromSource                   gwasEFO      \n",
       "1 Type 1 diabetes, definitions combined MONDO_0005147\n",
       "2 Type 1 diabetes, definitions combined MONDO_0005147\n",
       "  rightStudyLocusId                ⋯ H4/H3     Alt_ct_gene_cs_id           \n",
       "1 ea74fe3c38c7b9c4d67a414a08b7f310 ⋯ 14.961340 CD4_T:A2M:12_8972169_C_T    \n",
       "2 5753696279250a3cd12034956f8c20d1 ⋯  8.645992 CD4_T:A2M-AS1:12_8959409_C_T\n",
       "  Observed gwasStudyId     SusieRss_CS\n",
       "1 TRUE     FINNGEN_R12_T1D 2          \n",
       "2 TRUE     FINNGEN_R12_T1D 2          \n",
       "  EFO URI                                      gwas_disease_class\n",
       "1 http://purl.obolibrary.org/obo/MONDO_0005147 1                 \n",
       "2 http://purl.obolibrary.org/obo/MONDO_0005147 1                 \n",
       "  gwas_immune_class Parent term            is_coeqtl\n",
       "1 1                 Immune system disorder 1        \n",
       "2 1                 Immune system disorder 0        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(analysis_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 556,
   "id": "3aba09ce-c4fe-44bd-9011-2b523f2ed5a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>cell_type</th><th scope=col>p_value</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG</td><td>CD4_T</td><td>2.793323e-17</td></tr>\n",
       "\t<tr><td>A2M </td><td>CD4_T</td><td>5.356633e-48</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " feature\\_id & cell\\_type & p\\_value\\\\\n",
       " <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t A1BG & CD4\\_T & 2.793323e-17\\\\\n",
       "\t A2M  & CD4\\_T & 5.356633e-48\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 3\n",
       "\n",
       "| feature_id &lt;chr&gt; | cell_type &lt;chr&gt; | p_value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| A1BG | CD4_T | 2.793323e-17 |\n",
       "| A2M  | CD4_T | 5.356633e-48 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id cell_type p_value     \n",
       "1 A1BG       CD4_T     2.793323e-17\n",
       "2 A2M        CD4_T     5.356633e-48"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping_all_subset,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 557,
   "id": "934bb30d-52b0-4c74-a373-ed5bcf2b7574",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5293"
      ],
      "text/latex": [
       "5293"
      ],
      "text/markdown": [
       "5293"
      ],
      "text/plain": [
       "[1] 5293"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_gene_mapping_all_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 558,
   "id": "b25ac2c2-1733-478b-b625-1e0c0fb441bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "4452"
      ],
      "text/latex": [
       "4452"
      ],
      "text/markdown": [
       "4452"
      ],
      "text/plain": [
       "[1] 4452"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(analysis_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 559,
   "id": "0af2ac50-8d46-4fa0-95d7-6c9541d930c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_data_subset = merge(analysis_data, snp_gene_mapping_all_subset, by.x = c('eqtlTraitFromSource', 'eQTL_CT'), by.y = c('feature_id', 'cell_type'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 560,
   "id": "9e4ed29c-298a-4691-8ab2-c0767da01b23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2100"
      ],
      "text/latex": [
       "2100"
      ],
      "text/markdown": [
       "2100"
      ],
      "text/plain": [
       "[1] 2100"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(analysis_data_subset )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "284bd5d9-2db1-41f2-94a3-12caa6e52e51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 561,
   "id": "58b1e9d5-291f-4246-aa0e-0512dff443ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate amount of coloc for eGenes with and without co-eQTL effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 562,
   "id": "230c959a-2ed5-4275-9b63-23c16087dd08",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary_subset = analysis_data_subset %>% group_by( is_coeqtl) %>% dplyr::summarise(amount_coloc = length(unique(eqtlTraitFromSource)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 563,
   "id": "89d951a6-7ec8-411f-89e1-237ac5080ce6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>is_coeqtl</th><th scope=col>amount_coloc</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>0</td><td>541</td></tr>\n",
       "\t<tr><td>1</td><td> 40</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 2\n",
       "\\begin{tabular}{ll}\n",
       " is\\_coeqtl & amount\\_coloc\\\\\n",
       " <dbl> & <int>\\\\\n",
       "\\hline\n",
       "\t 0 & 541\\\\\n",
       "\t 1 &  40\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 2\n",
       "\n",
       "| is_coeqtl &lt;dbl&gt; | amount_coloc &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 | 541 |\n",
       "| 1 |  40 |\n",
       "\n"
      ],
      "text/plain": [
       "  is_coeqtl amount_coloc\n",
       "1 0         541         \n",
       "2 1          40         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_summary_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "id": "a3f71426-0ec5-4141-bfc9-2d588277d4d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate the amount of coeqtl_eGenes with subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "id": "cf2f9f91-d302-47c7-a895-c2ac130784bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>feature_id</th><th scope=col>cell_type</th><th scope=col>p_value</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>A1BG</td><td>CD4_T</td><td>2.793323e-17</td></tr>\n",
       "\t<tr><td>A2M </td><td>CD4_T</td><td>5.356633e-48</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 3\n",
       "\\begin{tabular}{lll}\n",
       " feature\\_id & cell\\_type & p\\_value\\\\\n",
       " <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t A1BG & CD4\\_T & 2.793323e-17\\\\\n",
       "\t A2M  & CD4\\_T & 5.356633e-48\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 3\n",
       "\n",
       "| feature_id &lt;chr&gt; | cell_type &lt;chr&gt; | p_value &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| A1BG | CD4_T | 2.793323e-17 |\n",
       "| A2M  | CD4_T | 5.356633e-48 |\n",
       "\n"
      ],
      "text/plain": [
       "  feature_id cell_type p_value     \n",
       "1 A1BG       CD4_T     2.793323e-17\n",
       "2 A2M        CD4_T     5.356633e-48"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_gene_mapping_all_subset,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "id": "ddb004ab-0094-4c62-a48f-1beb72761571",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5293"
      ],
      "text/latex": [
       "5293"
      ],
      "text/markdown": [
       "5293"
      ],
      "text/plain": [
       "[1] 5293"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(snp_gene_mapping_all_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "id": "364e25ab-0aa5-4155-be36-ab0217f812cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "273"
      ],
      "text/latex": [
       "273"
      ],
      "text/markdown": [
       "273"
      ],
      "text/plain": [
       "[1] 273"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "id": "05d4b2cf-ff9e-4c3b-8dc1-93c630aea765",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sign_coeqtls[sign_coeqtls$eGene == 'ADCY9',1:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 569,
   "id": "cd8a823a-6101-4358-b1a5-dfd4de5c8fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sign_coeqtls_subset = merge(sign_coeqtls, snp_gene_mapping_all_subset, by.x = c('eGene', 'cell_type'), by.y = c('feature_id', 'cell_type'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 570,
   "id": "554c9309-c596-451c-b690-d8ca5e7beed1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "213"
      ],
      "text/latex": [
       "213"
      ],
      "text/markdown": [
       "213"
      ],
      "text/plain": [
       "[1] 213"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls_subset$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 571,
   "id": "7591ee5f-80e4-45de-807d-944935e335fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "18663"
      ],
      "text/latex": [
       "18663"
      ],
      "text/markdown": [
       "18663"
      ],
      "text/plain": [
       "[1] 18663"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "8406"
      ],
      "text/latex": [
       "8406"
      ],
      "text/markdown": [
       "8406"
      ],
      "text/plain": [
       "[1] 8406"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(sign_coeqtls)\n",
    "nrow(sign_coeqtls_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 572,
   "id": "24171f9b-b005-45f9-a595-5bd65f419bfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "213"
      ],
      "text/latex": [
       "213"
      ],
      "text/markdown": [
       "213"
      ],
      "text/plain": [
       "[1] 213"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(sign_coeqtls_subset$eGene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "id": "75a1d459-8e4f-4f60-94b1-1c81416c927d",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary_subset$total_amount = c(nrow(snp_gene_mapping_all_subset) - length(unique(sign_coeqtls_subset$eGene)), length(unique(sign_coeqtls_subset$eGene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 574,
   "id": "e4a61c84-c535-4500-ba55-4f8b6881015a",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary_subset$amount_non_coloc = coloc_summary_subset$total_amount - coloc_summary_subset$amount_coloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 575,
   "id": "37404e1b-0128-4b71-ace6-1c7706450aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "coloc_summary_subset = coloc_summary_subset %>% arrange(-is_coeqtl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 576,
   "id": "5d19e2e2-9dd4-46dd-8853-b25327a3d7a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 2 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>is_coeqtl</th><th scope=col>amount_coloc</th><th scope=col>total_amount</th><th scope=col>amount_non_coloc</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td> 40</td><td> 213</td><td> 173</td></tr>\n",
       "\t<tr><td>0</td><td>541</td><td>5080</td><td>4539</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 2 × 4\n",
       "\\begin{tabular}{llll}\n",
       " is\\_coeqtl & amount\\_coloc & total\\_amount & amount\\_non\\_coloc\\\\\n",
       " <dbl> & <int> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t 1 &  40 &  213 &  173\\\\\n",
       "\t 0 & 541 & 5080 & 4539\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 2 × 4\n",
       "\n",
       "| is_coeqtl &lt;dbl&gt; | amount_coloc &lt;int&gt; | total_amount &lt;int&gt; | amount_non_coloc &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 |  40 |  213 |  173 |\n",
       "| 0 | 541 | 5080 | 4539 |\n",
       "\n"
      ],
      "text/plain": [
       "  is_coeqtl amount_coloc total_amount amount_non_coloc\n",
       "1 1          40           213          173            \n",
       "2 0         541          5080         4539            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coloc_summary_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 577,
   "id": "31a0916b-7c9e-4a48-a237-34121df60827",
   "metadata": {},
   "outputs": [],
   "source": [
    "contingency_table = as.matrix(coloc_summary_subset[,c('amount_coloc', 'amount_non_coloc')] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 578,
   "id": "fd493a71-ee3d-493c-aa40-4b174b1c68fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 2 of type int</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>amount_coloc</th><th scope=col>amount_non_coloc</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td> 40</td><td> 173</td></tr>\n",
       "\t<tr><td>541</td><td>4539</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 2 of type int\n",
       "\\begin{tabular}{ll}\n",
       " amount\\_coloc & amount\\_non\\_coloc\\\\\n",
       "\\hline\n",
       "\t  40 &  173\\\\\n",
       "\t 541 & 4539\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 2 of type int\n",
       "\n",
       "| amount_coloc | amount_non_coloc |\n",
       "|---|---|\n",
       "|  40 |  173 |\n",
       "| 541 | 4539 |\n",
       "\n"
      ],
      "text/plain": [
       "     amount_coloc amount_non_coloc\n",
       "[1,]  40           173            \n",
       "[2,] 541          4539            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "contingency_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 579,
   "id": "42129ee0-0175-46b9-90c6-f666484bd27d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tFisher's Exact Test for Count Data\n",
      "\n",
      "data:  contingency_table\n",
      "p-value = 0.0006863\n",
      "alternative hypothesis: true odds ratio is not equal to 1\n",
      "95 percent confidence interval:\n",
      " 1.323877 2.784903\n",
      "sample estimates:\n",
      "odds ratio \n",
      "  1.939638 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "result = fisher.test(contingency_table)\n",
    "print(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 580,
   "id": "a909f996-ab61-4609-aec4-cbb469af5e3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 581,
   "id": "19de54a2-f44c-4338-8e5b-aed11214182a",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot = rbind(odds_ratio_result[,c('p_value', 'estimate', 'conf_upper', 'conf_lower', 'cell_type')],\n",
    "                         data.frame(p_value = result$p.value,\n",
    "                           estimate = result$estimate,\n",
    "                           conf_upper = result$conf.int[2],\n",
    "                           conf_lower = result$conf.int[1], cell_type = 'ALL'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 582,
   "id": "0ef7808a-f9e9-4468-9a88-dd057583a5a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot$cell_type = factor(odds_ratio_plot$cell_type, levels = c('ALL', 'CD4_T', 'CD8_T', 'NK', 'B', 'Mono'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "id": "ce7f6fc8-876d-43ff-9bad-ac1704dbf7e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create the plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 4)\n",
    "or_plot = ggplot(odds_ratio_plot, aes(x = cell_type, y = estimate, color = cell_type )) +\n",
    "  geom_point(size = 1) +  cell_type_color +   geom_text(aes(y = conf_upper +0.5, label =paste0('OR: ', round(estimate, 2), \"\\n\", ' (p = ', signif(p_value, digits = 2), ')')  ), size = 2)  +                           # Plot points\n",
    "  geom_errorbar(aes(ymin = conf_lower, \n",
    "                    ymax = conf_upper), \n",
    "                width = 0.1, size = 0.5) +  plot_config  +  geom_hline(yintercept=1) + ylim(0,max(odds_ratio_plot$conf_upper) + 1) + plot_config + theme(axis.title.y = element_blank(), axis.title.x = element_blank())\n",
    "\n",
    "#or_plot\n",
    "\n",
    "pdf(paste0('figures/G2/G2_OR_Plot', use_gwas, remove_HLA, only_dt, only_esnps, coloc_thres,use_gwas ,cs_snps,  '.pdf'), width =3, height =2)\n",
    "#or_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e940b36-97de-476f-81d2-09c85342e2a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 584,
   "id": "a261dd28-9c97-445f-b85c-4c54b8f1a7e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save all result contingency table\n",
    "contingency_table = data.frame(contingency_table)\n",
    "contingency_table$is_coeqtl = c(1,0)\n",
    "contingency_table$cell_type = 'ALL'\n",
    "contingency_table$trait_filter = use_gwas\n",
    "contingency_table$remove_HLA = remove_HLA\n",
    "contingency_table$only_dt = only_dt\n",
    "contingency_table$only_esnps = only_esnps\n",
    "contingency_table$coloc_thres = coloc_thres\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 585,
   "id": "f248e228-c9b1-4994-9bf5-00f8c3a3df8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>amount_coloc</th><th scope=col>amount_non_coloc</th><th scope=col>is_coeqtl</th><th scope=col>cell_type</th><th scope=col>trait_filter</th><th scope=col>remove_HLA</th><th scope=col>only_dt</th><th scope=col>only_esnps</th><th scope=col>coloc_thres</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td> 40</td><td> 173</td><td>1</td><td>ALL</td><td>disease</td><td>TRUE</td><td>TRUE</td><td>TRUE</td><td>relation</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>541</td><td>4539</td><td>0</td><td>ALL</td><td>disease</td><td>TRUE</td><td>TRUE</td><td>TRUE</td><td>relation</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & amount\\_coloc & amount\\_non\\_coloc & is\\_coeqtl & cell\\_type & trait\\_filter & remove\\_HLA & only\\_dt & only\\_esnps & coloc\\_thres\\\\\n",
       "  & <int> & <int> & <dbl> & <chr> & <chr> & <lgl> & <lgl> & <lgl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 &  40 &  173 & 1 & ALL & disease & TRUE & TRUE & TRUE & relation\\\\\n",
       "\t2 & 541 & 4539 & 0 & ALL & disease & TRUE & TRUE & TRUE & relation\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| <!--/--> | amount_coloc &lt;int&gt; | amount_non_coloc &lt;int&gt; | is_coeqtl &lt;dbl&gt; | cell_type &lt;chr&gt; | trait_filter &lt;chr&gt; | remove_HLA &lt;lgl&gt; | only_dt &lt;lgl&gt; | only_esnps &lt;lgl&gt; | coloc_thres &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 |  40 |  173 | 1 | ALL | disease | TRUE | TRUE | TRUE | relation |\n",
       "| 2 | 541 | 4539 | 0 | ALL | disease | TRUE | TRUE | TRUE | relation |\n",
       "\n"
      ],
      "text/plain": [
       "  amount_coloc amount_non_coloc is_coeqtl cell_type trait_filter remove_HLA\n",
       "1  40           173             1         ALL       disease      TRUE      \n",
       "2 541          4539             0         ALL       disease      TRUE      \n",
       "  only_dt only_esnps coloc_thres\n",
       "1 TRUE    TRUE       relation   \n",
       "2 TRUE    TRUE       relation   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(contingency_table,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "id": "5f2c9e0f-8b66-4a7a-b2c2-5c2c63dd9619",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>amount_coloc</th><th scope=col>amount_non_coloc</th><th scope=col>is_coeqtl</th><th scope=col>cell_type</th><th scope=col>trait_filter</th><th scope=col>only_dt</th><th scope=col>remove_HLA</th><th scope=col>only_esnps</th><th scope=col>coloc_thres</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>  4</td><td>  28</td><td>1</td><td>B</td><td>disease</td><td>TRUE</td><td>TRUE</td><td>TRUE</td><td>relation</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>147</td><td>1500</td><td>0</td><td>B</td><td>disease</td><td>TRUE</td><td>TRUE</td><td>TRUE</td><td>relation</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & amount\\_coloc & amount\\_non\\_coloc & is\\_coeqtl & cell\\_type & trait\\_filter & only\\_dt & remove\\_HLA & only\\_esnps & coloc\\_thres\\\\\n",
       "  & <int> & <int> & <dbl> & <chr> & <chr> & <lgl> & <lgl> & <lgl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 &   4 &   28 & 1 & B & disease & TRUE & TRUE & TRUE & relation\\\\\n",
       "\t6 & 147 & 1500 & 0 & B & disease & TRUE & TRUE & TRUE & relation\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 9\n",
       "\n",
       "| <!--/--> | amount_coloc &lt;int&gt; | amount_non_coloc &lt;int&gt; | is_coeqtl &lt;dbl&gt; | cell_type &lt;chr&gt; | trait_filter &lt;chr&gt; | only_dt &lt;lgl&gt; | remove_HLA &lt;lgl&gt; | only_esnps &lt;lgl&gt; | coloc_thres &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 |   4 |   28 | 1 | B | disease | TRUE | TRUE | TRUE | relation |\n",
       "| 6 | 147 | 1500 | 0 | B | disease | TRUE | TRUE | TRUE | relation |\n",
       "\n"
      ],
      "text/plain": [
       "  amount_coloc amount_non_coloc is_coeqtl cell_type trait_filter only_dt\n",
       "1   4            28             1         B         disease      TRUE   \n",
       "6 147          1500             0         B         disease      TRUE   \n",
       "  remove_HLA only_esnps coloc_thres\n",
       "1 TRUE       TRUE       relation   \n",
       "6 TRUE       TRUE       relation   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(contingency_table_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 587,
   "id": "265b956f-5d37-40ac-b750-a54debbb79bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "contingency_table_data = rbind( contingency_table_data, contingency_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 588,
   "id": "cfd8c226-ae9e-44f7-8909-bca39d5b7982",
   "metadata": {},
   "outputs": [],
   "source": [
    "contingency_table_data$excluded_genes_without_cs = cs_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 589,
   "id": "7809154d-fac1-4de6-8158-5f13af9c2ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#write.csv(odds_ratio_plot, paste0(result_path, '/G2/Odds_Ratio_result', use_gwas, only_dt, remove_HLA, only_esnps , coloc_thres,use_gwas,cs_snps, '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 590,
   "id": "eb065759-57bf-4af1-bf28-c38911bb2a8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(    contingency_table_data , paste0('figures/G2/', 'Contingency_Table', use_gwas, only_dt, remove_HLA, only_esnps , coloc_thres, use_gwas, cs_snps,'.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 591,
   "id": "d27aadfa-4f9c-4357-acca-643dff7f7459",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create Plot including SNP based GWAS enrichment results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 592,
   "id": "4efe2801-2712-49c2-9622-143835c4ba9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>odds ratio</th><td>0.52464488</td><td>1.457338</td><td>4.248052</td><td>0.3664286</td><td>B    </td></tr>\n",
       "\t<tr><th scope=row>odds ratio1</th><td>0.01951706</td><td>1.691435</td><td>2.619116</td><td>1.0583392</td><td>CD4_T</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <fct>\\\\\n",
       "\\hline\n",
       "\todds ratio & 0.52464488 & 1.457338 & 4.248052 & 0.3664286 & B    \\\\\n",
       "\todds ratio1 & 0.01951706 & 1.691435 & 2.619116 & 1.0583392 & CD4\\_T\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 5\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | conf_upper &lt;dbl&gt; | conf_lower &lt;dbl&gt; | cell_type &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| odds ratio | 0.52464488 | 1.457338 | 4.248052 | 0.3664286 | B     |\n",
       "| odds ratio1 | 0.01951706 | 1.691435 | 2.619116 | 1.0583392 | CD4_T |\n",
       "\n"
      ],
      "text/plain": [
       "            p_value    estimate conf_upper conf_lower cell_type\n",
       "odds ratio  0.52464488 1.457338 4.248052   0.3664286  B        \n",
       "odds ratio1 0.01951706 1.691435 2.619116   1.0583392  CD4_T    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(odds_ratio_plot,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 593,
   "id": "07b7c017-0880-4e9d-bb90-a590a0500888",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot$type = 'eGene_based'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 594,
   "id": "727c0668-f2b1-45b5-a0ac-52b96b7d5e90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'all_traits'</li><li>'rm_cell_count'</li><li>'disease'</li><li>'immune'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'all\\_traits'\n",
       "\\item 'rm\\_cell\\_count'\n",
       "\\item 'disease'\n",
       "\\item 'immune'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'all_traits'\n",
       "2. 'rm_cell_count'\n",
       "3. 'disease'\n",
       "4. 'immune'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"all_traits\"    \"rm_cell_count\" \"disease\"       \"immune\"       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(snp_enrichment$type )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 595,
   "id": "77efd8cf-9a8a-4027-8e86-d3a8f315ea34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use_gwas = 'all_traits'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 596,
   "id": "8a49ced6-026b-49d8-8f19-d9a2e8492a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter on relevant GWAS traits\n",
    "snp_enrichment = snp_enrichment[snp_enrichment$type == use_gwas,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 597,
   "id": "4358fe88-1ff4-4151-81e1-02f882642df7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ALL_disease</th><td>0.01870063</td><td>1.531563</td><td>2.180496</td><td>1.0527763</td><td>ALL</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>B_disease</th><td>0.14221170</td><td>1.598469</td><td>3.021836</td><td>0.7877197</td><td>B  </td><td>disease</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tALL\\_disease & 0.01870063 & 1.531563 & 2.180496 & 1.0527763 & ALL & disease\\\\\n",
       "\tB\\_disease & 0.14221170 & 1.598469 & 3.021836 & 0.7877197 & B   & disease\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 6\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | conf_upper &lt;dbl&gt; | conf_lower &lt;dbl&gt; | cell_type &lt;chr&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| ALL_disease | 0.01870063 | 1.531563 | 2.180496 | 1.0527763 | ALL | disease |\n",
       "| B_disease | 0.14221170 | 1.598469 | 3.021836 | 0.7877197 | B   | disease |\n",
       "\n"
      ],
      "text/plain": [
       "            p_value    estimate conf_upper conf_lower cell_type type   \n",
       "ALL_disease 0.01870063 1.531563 2.180496   1.0527763  ALL       disease\n",
       "B_disease   0.14221170 1.598469 3.021836   0.7877197  B         disease"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(snp_enrichment,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "id": "a7da8d5a-1cfb-43bc-a658-67b5ae2696ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_enrichment$type = 'snp_based'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "id": "b47f209a-d79c-448a-b922-5295b9498759",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot = rbind(odds_ratio_plot, snp_enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "id": "f6c13f99-6793-47ef-bacf-2ab6e5272d4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 12 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>odds ratio</th><td>0.5246448846</td><td>1.457338</td><td>4.248052</td><td>0.3664286</td><td>B    </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio1</th><td>0.0195170622</td><td>1.691435</td><td>2.619116</td><td>1.0583392</td><td>CD4_T</td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio2</th><td>0.0057195772</td><td>2.278742</td><td>4.049781</td><td>1.2222459</td><td>CD8_T</td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio3</th><td>0.5043996051</td><td>1.261864</td><td>2.882112</td><td>0.4722654</td><td>Mono </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio4</th><td>0.2189869690</td><td>1.627725</td><td>3.802316</td><td>0.6002015</td><td>NK   </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio5</th><td>0.0006863434</td><td>1.939638</td><td>2.784903</td><td>1.3238773</td><td>ALL  </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>ALL_disease</th><td>0.0187006300</td><td>1.531563</td><td>2.180496</td><td>1.0527763</td><td>ALL  </td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>B_disease</th><td>0.1422117000</td><td>1.598469</td><td>3.021836</td><td>0.7877197</td><td>B    </td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>CD4_T_disease</th><td>0.0037729260</td><td>1.862487</td><td>2.812239</td><td>1.2007029</td><td>CD4_T</td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>CD8_T_disease</th><td>0.0496152300</td><td>1.632734</td><td>2.655485</td><td>0.9681380</td><td>CD8_T</td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>Mono_disease</th><td>0.8767768000</td><td>1.025278</td><td>1.905349</td><td>0.5121500</td><td>Mono </td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>NK_disease</th><td>0.4340997000</td><td>1.254892</td><td>2.250417</td><td>0.6577192</td><td>NK   </td><td>snp_based  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 12 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr>\\\\\n",
       "\\hline\n",
       "\todds ratio & 0.5246448846 & 1.457338 & 4.248052 & 0.3664286 & B     & eGene\\_based\\\\\n",
       "\todds ratio1 & 0.0195170622 & 1.691435 & 2.619116 & 1.0583392 & CD4\\_T & eGene\\_based\\\\\n",
       "\todds ratio2 & 0.0057195772 & 2.278742 & 4.049781 & 1.2222459 & CD8\\_T & eGene\\_based\\\\\n",
       "\todds ratio3 & 0.5043996051 & 1.261864 & 2.882112 & 0.4722654 & Mono  & eGene\\_based\\\\\n",
       "\todds ratio4 & 0.2189869690 & 1.627725 & 3.802316 & 0.6002015 & NK    & eGene\\_based\\\\\n",
       "\todds ratio5 & 0.0006863434 & 1.939638 & 2.784903 & 1.3238773 & ALL   & eGene\\_based\\\\\n",
       "\tALL\\_disease & 0.0187006300 & 1.531563 & 2.180496 & 1.0527763 & ALL   & snp\\_based  \\\\\n",
       "\tB\\_disease & 0.1422117000 & 1.598469 & 3.021836 & 0.7877197 & B     & snp\\_based  \\\\\n",
       "\tCD4\\_T\\_disease & 0.0037729260 & 1.862487 & 2.812239 & 1.2007029 & CD4\\_T & snp\\_based  \\\\\n",
       "\tCD8\\_T\\_disease & 0.0496152300 & 1.632734 & 2.655485 & 0.9681380 & CD8\\_T & snp\\_based  \\\\\n",
       "\tMono\\_disease & 0.8767768000 & 1.025278 & 1.905349 & 0.5121500 & Mono  & snp\\_based  \\\\\n",
       "\tNK\\_disease & 0.4340997000 & 1.254892 & 2.250417 & 0.6577192 & NK    & snp\\_based  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 12 × 6\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | conf_upper &lt;dbl&gt; | conf_lower &lt;dbl&gt; | cell_type &lt;fct&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| odds ratio | 0.5246448846 | 1.457338 | 4.248052 | 0.3664286 | B     | eGene_based |\n",
       "| odds ratio1 | 0.0195170622 | 1.691435 | 2.619116 | 1.0583392 | CD4_T | eGene_based |\n",
       "| odds ratio2 | 0.0057195772 | 2.278742 | 4.049781 | 1.2222459 | CD8_T | eGene_based |\n",
       "| odds ratio3 | 0.5043996051 | 1.261864 | 2.882112 | 0.4722654 | Mono  | eGene_based |\n",
       "| odds ratio4 | 0.2189869690 | 1.627725 | 3.802316 | 0.6002015 | NK    | eGene_based |\n",
       "| odds ratio5 | 0.0006863434 | 1.939638 | 2.784903 | 1.3238773 | ALL   | eGene_based |\n",
       "| ALL_disease | 0.0187006300 | 1.531563 | 2.180496 | 1.0527763 | ALL   | snp_based   |\n",
       "| B_disease | 0.1422117000 | 1.598469 | 3.021836 | 0.7877197 | B     | snp_based   |\n",
       "| CD4_T_disease | 0.0037729260 | 1.862487 | 2.812239 | 1.2007029 | CD4_T | snp_based   |\n",
       "| CD8_T_disease | 0.0496152300 | 1.632734 | 2.655485 | 0.9681380 | CD8_T | snp_based   |\n",
       "| Mono_disease | 0.8767768000 | 1.025278 | 1.905349 | 0.5121500 | Mono  | snp_based   |\n",
       "| NK_disease | 0.4340997000 | 1.254892 | 2.250417 | 0.6577192 | NK    | snp_based   |\n",
       "\n"
      ],
      "text/plain": [
       "              p_value      estimate conf_upper conf_lower cell_type type       \n",
       "odds ratio    0.5246448846 1.457338 4.248052   0.3664286  B         eGene_based\n",
       "odds ratio1   0.0195170622 1.691435 2.619116   1.0583392  CD4_T     eGene_based\n",
       "odds ratio2   0.0057195772 2.278742 4.049781   1.2222459  CD8_T     eGene_based\n",
       "odds ratio3   0.5043996051 1.261864 2.882112   0.4722654  Mono      eGene_based\n",
       "odds ratio4   0.2189869690 1.627725 3.802316   0.6002015  NK        eGene_based\n",
       "odds ratio5   0.0006863434 1.939638 2.784903   1.3238773  ALL       eGene_based\n",
       "ALL_disease   0.0187006300 1.531563 2.180496   1.0527763  ALL       snp_based  \n",
       "B_disease     0.1422117000 1.598469 3.021836   0.7877197  B         snp_based  \n",
       "CD4_T_disease 0.0037729260 1.862487 2.812239   1.2007029  CD4_T     snp_based  \n",
       "CD8_T_disease 0.0496152300 1.632734 2.655485   0.9681380  CD8_T     snp_based  \n",
       "Mono_disease  0.8767768000 1.025278 1.905349   0.5121500  Mono      snp_based  \n",
       "NK_disease    0.4340997000 1.254892 2.250417   0.6577192  NK        snp_based  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "odds_ratio_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "id": "aa811456-3dc5-432c-9f25-6b2eb3f28bd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create the plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 4)\n",
    "or_plot = ggplot(odds_ratio_plot, aes(x = cell_type, y = estimate, color = cell_type, linetype = type )) +\n",
    "  geom_point(size = 1, position = position_dodge(width = 1)) +  cell_type_color +   geom_text(aes(y = conf_upper +0.5, label =paste0('OR: ', round(estimate, 2), \"\\n\", ' (p = ', signif(p_value, digits = 2), ')')  ), size = 2, position = position_dodge(width = 1))  +                           # Plot points\n",
    "  geom_errorbar(aes(ymin = conf_lower, \n",
    "                    ymax = conf_upper), position = position_dodge(width = 1), \n",
    "                width = 0.1, size = 0.5) +  plot_config  +  geom_hline(yintercept=1) + ylim(0,max(odds_ratio_plot$conf_upper)+0.5) + plot_config + theme(axis.title.y = element_blank(), axis.title.x = element_blank())\n",
    "\n",
    "#or_plot\n",
    "\n",
    "pdf(paste0('figures/G2/G2_OR_Plot_combined', use_gwas, remove_HLA, only_dt, only_esnps, coloc_thres,use_gwas , cs_snps,  '.pdf'), width =5.5, height =3)\n",
    "#or_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 602,
   "id": "38b36dec-b3c7-4907-bbce-e6da61f2fd20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 12 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>odds ratio</th><td>0.5246448846</td><td>1.457338</td><td>4.248052</td><td>0.3664286</td><td>B    </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio1</th><td>0.0195170622</td><td>1.691435</td><td>2.619116</td><td>1.0583392</td><td>CD4_T</td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio2</th><td>0.0057195772</td><td>2.278742</td><td>4.049781</td><td>1.2222459</td><td>CD8_T</td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio3</th><td>0.5043996051</td><td>1.261864</td><td>2.882112</td><td>0.4722654</td><td>Mono </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio4</th><td>0.2189869690</td><td>1.627725</td><td>3.802316</td><td>0.6002015</td><td>NK   </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>odds ratio5</th><td>0.0006863434</td><td>1.939638</td><td>2.784903</td><td>1.3238773</td><td>ALL  </td><td>eGene_based</td></tr>\n",
       "\t<tr><th scope=row>ALL_disease</th><td>0.0187006300</td><td>1.531563</td><td>2.180496</td><td>1.0527763</td><td>ALL  </td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>B_disease</th><td>0.1422117000</td><td>1.598469</td><td>3.021836</td><td>0.7877197</td><td>B    </td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>CD4_T_disease</th><td>0.0037729260</td><td>1.862487</td><td>2.812239</td><td>1.2007029</td><td>CD4_T</td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>CD8_T_disease</th><td>0.0496152300</td><td>1.632734</td><td>2.655485</td><td>0.9681380</td><td>CD8_T</td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>Mono_disease</th><td>0.8767768000</td><td>1.025278</td><td>1.905349</td><td>0.5121500</td><td>Mono </td><td>snp_based  </td></tr>\n",
       "\t<tr><th scope=row>NK_disease</th><td>0.4340997000</td><td>1.254892</td><td>2.250417</td><td>0.6577192</td><td>NK   </td><td>snp_based  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 12 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr>\\\\\n",
       "\\hline\n",
       "\todds ratio & 0.5246448846 & 1.457338 & 4.248052 & 0.3664286 & B     & eGene\\_based\\\\\n",
       "\todds ratio1 & 0.0195170622 & 1.691435 & 2.619116 & 1.0583392 & CD4\\_T & eGene\\_based\\\\\n",
       "\todds ratio2 & 0.0057195772 & 2.278742 & 4.049781 & 1.2222459 & CD8\\_T & eGene\\_based\\\\\n",
       "\todds ratio3 & 0.5043996051 & 1.261864 & 2.882112 & 0.4722654 & Mono  & eGene\\_based\\\\\n",
       "\todds ratio4 & 0.2189869690 & 1.627725 & 3.802316 & 0.6002015 & NK    & eGene\\_based\\\\\n",
       "\todds ratio5 & 0.0006863434 & 1.939638 & 2.784903 & 1.3238773 & ALL   & eGene\\_based\\\\\n",
       "\tALL\\_disease & 0.0187006300 & 1.531563 & 2.180496 & 1.0527763 & ALL   & snp\\_based  \\\\\n",
       "\tB\\_disease & 0.1422117000 & 1.598469 & 3.021836 & 0.7877197 & B     & snp\\_based  \\\\\n",
       "\tCD4\\_T\\_disease & 0.0037729260 & 1.862487 & 2.812239 & 1.2007029 & CD4\\_T & snp\\_based  \\\\\n",
       "\tCD8\\_T\\_disease & 0.0496152300 & 1.632734 & 2.655485 & 0.9681380 & CD8\\_T & snp\\_based  \\\\\n",
       "\tMono\\_disease & 0.8767768000 & 1.025278 & 1.905349 & 0.5121500 & Mono  & snp\\_based  \\\\\n",
       "\tNK\\_disease & 0.4340997000 & 1.254892 & 2.250417 & 0.6577192 & NK    & snp\\_based  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 12 × 6\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | conf_upper &lt;dbl&gt; | conf_lower &lt;dbl&gt; | cell_type &lt;fct&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| odds ratio | 0.5246448846 | 1.457338 | 4.248052 | 0.3664286 | B     | eGene_based |\n",
       "| odds ratio1 | 0.0195170622 | 1.691435 | 2.619116 | 1.0583392 | CD4_T | eGene_based |\n",
       "| odds ratio2 | 0.0057195772 | 2.278742 | 4.049781 | 1.2222459 | CD8_T | eGene_based |\n",
       "| odds ratio3 | 0.5043996051 | 1.261864 | 2.882112 | 0.4722654 | Mono  | eGene_based |\n",
       "| odds ratio4 | 0.2189869690 | 1.627725 | 3.802316 | 0.6002015 | NK    | eGene_based |\n",
       "| odds ratio5 | 0.0006863434 | 1.939638 | 2.784903 | 1.3238773 | ALL   | eGene_based |\n",
       "| ALL_disease | 0.0187006300 | 1.531563 | 2.180496 | 1.0527763 | ALL   | snp_based   |\n",
       "| B_disease | 0.1422117000 | 1.598469 | 3.021836 | 0.7877197 | B     | snp_based   |\n",
       "| CD4_T_disease | 0.0037729260 | 1.862487 | 2.812239 | 1.2007029 | CD4_T | snp_based   |\n",
       "| CD8_T_disease | 0.0496152300 | 1.632734 | 2.655485 | 0.9681380 | CD8_T | snp_based   |\n",
       "| Mono_disease | 0.8767768000 | 1.025278 | 1.905349 | 0.5121500 | Mono  | snp_based   |\n",
       "| NK_disease | 0.4340997000 | 1.254892 | 2.250417 | 0.6577192 | NK    | snp_based   |\n",
       "\n"
      ],
      "text/plain": [
       "              p_value      estimate conf_upper conf_lower cell_type type       \n",
       "odds ratio    0.5246448846 1.457338 4.248052   0.3664286  B         eGene_based\n",
       "odds ratio1   0.0195170622 1.691435 2.619116   1.0583392  CD4_T     eGene_based\n",
       "odds ratio2   0.0057195772 2.278742 4.049781   1.2222459  CD8_T     eGene_based\n",
       "odds ratio3   0.5043996051 1.261864 2.882112   0.4722654  Mono      eGene_based\n",
       "odds ratio4   0.2189869690 1.627725 3.802316   0.6002015  NK        eGene_based\n",
       "odds ratio5   0.0006863434 1.939638 2.784903   1.3238773  ALL       eGene_based\n",
       "ALL_disease   0.0187006300 1.531563 2.180496   1.0527763  ALL       snp_based  \n",
       "B_disease     0.1422117000 1.598469 3.021836   0.7877197  B         snp_based  \n",
       "CD4_T_disease 0.0037729260 1.862487 2.812239   1.2007029  CD4_T     snp_based  \n",
       "CD8_T_disease 0.0496152300 1.632734 2.655485   0.9681380  CD8_T     snp_based  \n",
       "Mono_disease  0.8767768000 1.025278 1.905349   0.5121500  Mono      snp_based  \n",
       "NK_disease    0.4340997000 1.254892 2.250417   0.6577192  NK        snp_based  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "odds_ratio_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 603,
   "id": "3271761c-10ad-4327-bef7-8d560bafd7f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot$excluded_genes_without_cs = cs_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 604,
   "id": "247a2b97-e996-49bb-9078-087cecb9a771",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot$trait_filter = use_gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "id": "362ea10b-936b-4e54-9628-5073b4410d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(odds_ratio_plot) = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "id": "99b09c97-8321-4a0c-8542-5ad142bc158a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save with plot data for sharing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "id": "15db6dd2-c2a1-4fb3-98d3-e994987dbbfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(odds_ratio_plot, paste0('figures/G2/G2_OR_Plot', use_gwas, remove_HLA, only_dt, only_esnps, coloc_thres,use_gwas , cs_snps ,'_data.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bb1cffb-db64-453d-b20b-c4c68ebba8ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "id": "dc63a8c3-1a0e-471b-8806-094bcc0c1171",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlgAAAHgCAIAAAD2dYQOAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdd3wT9RsH8Oey0yRN924ZLVAoGwplypZREBAUUVEREQQV3JshIioKKOLGH4iC\ngoIyZFmZsmcplFIolO6djjT7fn+kQEsnbdokd5/3iz8ud9fjCQ355L65574My7IEAADAVwJ7\nFwAAAGBPCEIAAOA1BCEAAPAaghAAAHgNQQgAALyGIAQAAF5DEAIAAK8hCAEAgNdsEIR///13\nUlJSw48DAADQ9GwQhCUlJUajseHHAQAAaHoYGgUAAF5DEAIAAK8hCAEAgNcQhAAAwGsIQgAA\n4DUEIQAA8BqCEAAAeA1BCAAAvIYgBAAAXkMQAgAAryEIAQCA1xCEAADAawhCAADgNQQhAADw\nGoIQAAB4DUEIAAC8hiAEAABeQxACAACvIQgBAIDXEIQAAMBrCEIAAOA1BCEAAPAaghAAAHgN\nQQgAALyGIAQAAF5DEAIAAK8hCAEAgNcQhAAAwGsIQgAA4DUEIQAA8BqCEAAAeA1BCAAAvIYg\nBAAAXkMQAgAAryEIAQCA1xCEAADAawhCAADgNQQhAADwGoIQAAB4DUEIAAC8hiAEAABeQxAC\nAACvIQgBAIDXEIQAAMBrCEIAAOA1BCEAAPAaghAAAHgNQQgAALyGIAQAAF5DEAIAAK8hCAEA\ngNcQhAAAwGsIQgAA4DUEIQAA8BqCEAAAeA1BCAAAvIYgBAAAXkMQAgAAryEIAQCA1xCEAADA\nawhCAADgNQQhAADwGoIQAAB4DUEIAAC8hiAEAABeQxACAACvIQgBAIDXEIQAAMBrCEIAAOA1\nBCEAAPAaghAAAHgNQQgAALyGIAQAAF5DEAIAAK/VJQjNFz/qLWf85hxq9GoAAACaWO1BaL64\ndPa/Hca3boJiAAAAmlptQWi+uGRWzIRVs1swTVIOAABA06o5CE2xS2YdnLBqZkvEIAAAcFNN\nQWiK/WjWgfFfPReKGAQAAK4S1bDt+vbNB3efail4oexxPybxx6JtTyqboi4AAIAmUdMZYdgb\nJ9kysW+38X3xIIsUBAAAjqnpjLAGsbGxmZmZ1uUrV6507NjRdiUBAAA0nToGYftF8RnlH3t6\neopEZT+bkJBg66oAAACaSD3PCAMCAgICAqzLcXFxtqsHAACgSeEWawAAwGsIQgAA4DUEIQAA\n8BqCEAAAeA1BCAAAvIYgBAAAXkMQAgAAryEIAQCA1xCEAADAawhCAADgNQQhAADwGoIQAAB4\nDUEIAAC8hiAEAABeQxACAACvIQgBAIDXEIQAAMBrCEIAAOA1BCEAAPAaghAAAHgNQQgAALyG\nIAQAAF5DEAIAAK8hCAEAgNcQhAAAwGsIQgAA4DUEIQAA8BqCEAAAeE1k7wIAwHGZtLo1pwv/\nzjIWCISt/ZUvdFGGS4hY7esb8g4SCYQCPw/5jCj3Icoqf1r/6ebs33VEatf1I12bVdps1uq+\nP6nZkWXMNgrGDgx4za+xnw1A1RCEAFANi/G7fXnngzw+jZR6W4wHzufPOcz8NFDhTUQCl08e\n9uhlNh8/kz3vom5ID1lVPy99eVzQy0VFUw6yVW01bTice9Dd44sxshAJ06jPA6BmGBoFgGpk\nFf8hUL3bUdZMyrjIJcMj3QYUFO8ounsvmcgaY6YfdqSvyqvzwQu1f2vkz3aRIwXB7nBGCABV\nKyw2G1xlAbcfC0TNlObkEiIlkUX76notEYnVyhVDpUREJHp6pP89HF1rzlIxsYfS52ey7p4u\nz0e53aewbfkAdYUzQgComqtCKC40pd1+bDHdKBb6WeNK4PLJI0H/POD9uKT093RLfY4uFao1\nJnVH350P+r7rZfggVlevowDYAIIQAKrhqxhvLnr/vO6GntWWGnaeKNjnphyhurNd7iKdGqlI\nOl90qcovAWvmJutV7hSQwfgo2A+CEACqIZA8M9Cjq6bwpW1pY3bm/UWK5X0UPnftolZOddd+\nc81c1XeE5l/3pvTepknUFD6yPmXGZTMREek/3JS1SUfESKb1kJ0/kjn898zFudL3OsjwZgT2\nwrBsPT7LVbBp06aOHTu2bt3aJgUBAAA0JXwIAwAAXkMQAgAAryEIAQCA1xCEAADAawhCAADg\nNQQhANxi1L0fU3itoReS24hZ98E/misOUgxwGoIQAMqkXCksaKZsaYve9twMzZxtaUN+z3jh\njC737o36Tzen9F6f0ntH4Y3yq82Gb3en9t9bXLa/UPaIj351Mm44A40OQQgAVqZd15l+QbZ4\nTzDrVh7RhUX6bhnl3jYrf2XKXad10pfHBf0XrQ6rsJI9ea5I0FzhXm5Vy2DZjaTSYhsUBFAT\n3HQbAIiIyKi/YBRPs95AW1c8I8bQ2d24PdUs9VC81ltdcZ4l/YebsrcaK/z0/f2C5gXdepCn\nO65S/OQrVJLwsTDxw+kGCpLW/JcXpBRsVarneZRsTS63Vi0OzdVdJEWPBj41gBohCAGAiIgM\nFo1YcGeG3UKjoavXb5EUdyH3/Vj9lkhpuRFT6ZsTgt6s4VB6i0YmsJ7bqWSCEr3ZUuPoE6st\n+TJN+kIPkSCn4gZGoCRzQc0/DNBgCMKmZtKeXH1q9basG/kC7/CAcXO7jGknYYj99+X1i/YR\nCYQqf4/7ZveaM0xZ5Rc1Fz7648XfdETqKb+PeqL53Vt1cfEfLbp44rpJ1sxv4tu9Hu4gJrP2\nxNcnvt2WdSPLKHtw0F9vYRJwqI5EoDYa74xDKmUj/YRyou4tZIoTxgKSlhu0rO2MUCpQ6yz5\nRB5ERTqLi1RYc5BdvVq8/apx+9WyG5WO3sX+fr/Kn4hYSzEJ3ZCC0MgQhE3LcuWrfxefC3r1\n8x6dfSw3/z23ZNYh4a+Don2ISDB4+aS3+phzj55++e2Lp4b16F7Vz7d/ffw/rxetf+SAoYqN\n+j0rYnUP3v/bGHn+tiOzl10dtDrc9POhBfvdX/v6gd7NJPhdQ43E0vZibaKeIqyjmMW6vzOV\nz3jRxes6rVrtVmHX2s4IPWQ9CjW/ZMqfUpvXJRp6Rkhq/pvDOvj+14GIiHI0D5wVrh6i9LRu\n0Bivekra1fcJAdQRPms1rcytGwUTFnTq1Vwqd5G3HtVj9qCCLdvunvKblQmtX8gkf7N98hd1\nn/Jb6OElISo7k3TxcXGhwphtml6zuvZHCkIdiIY1Zw+k3LpK01UsvpYzcXPmB7nStzpI7+1K\nUqFsVi9ZwonMMdvz4nw8Zgdaf/rWvBNVz0pRhaQUXUhzubK6zQA2gvfHJlVYkmFQ9wi8/VjQ\nrIUy80bZlN//zPnlHyISqx9cNaw9ERGFPDvql3s4uijqmXY7nt02+kOW8fB//ttgBWVmZaok\n5w5OfjtT6+E5eG6v2QMVmPYNqhXcytXtYPHVlq6hRMSIont5zKjvobz81Z9Hqyuuk745wTqJ\nk/DhIUEPV/1j6j+H3Fo26zZkSqdG4MM6NDq8yJqUq8JXrElOvf3YciOp2Ne/bMrvwcsn7z04\n9rMnJYd+S9fU5+gF616O95o3ZvuRh9YvVP31amwiSdVuhelunb6OmfDjQq+L82OPoScLaiKW\nvTvINdRBPiwJZW8OVrd2kGKA0xCETcs3eqJ547xzR6/rddrShO3HV8a4jR11Z8pvxsWl0/TI\nEdfO/RJXjxtqGIqtg6wMEUsWjaGE3Hr0cbmzXUB4U4G6kSm/HuEaVPt+AFyAIGxagjYzB77Z\nrXDN81sfvP/vxVuYMV/2HeV71y5uD053/3fV1dyqviPM+2XP4G5bv0/QrH3wl8FT47OJiOj0\n+xtf/FVH5PPw24Fpn+4c3/+P6UsLe7zbriMx7Wf07H7uv6mDf39qQV7HhR174PcNAHAXzFAP\nAAC8hjMEAADgNQQhAADwGoIQAAB4DUEIAAC8hiAEAABew51lGofx6LwDl6YMespRepNrZTqx\nYP/5SYOebtPggtn8uKJftpRcTjcL3SUdB7k9NECiYMhyImfat6VEJFGImnVXT3vUxbu6vykx\nf85HxYVEgaP93h8jvmtj/gXN2t9LLqeZdVLlK5+7tzMY/9uYv/m4QaeQRI33mNxd5Cz/4ADg\nMHBG2ChuXvk5v/k4W6Qgm5v+w3Nbx/fb9PjMM8dyat9aeY3+atLn0/58oO+mx6ef3J9W1iyj\ni4tf8Mgf0b1+mzDpwK+xRiJR5GM+p7670eA5UM3JhZ9+r/OP9lr8aeD8aQp2X/bXB8tuJSnq\n6bX6u+DP57sHX87fcbH6Q4S5L/8ueMn4uyOQiCi7aNXXpR7RPktXBa/+3L0dkSW+NLG521sf\nBS6aKrmypvAc7pwDAPcMQdgYkndcF98X5Fb7jrUyH1t+5EjrHt/9Hf16u8xPlqeU1rK18hr2\nTEx26xeHrt876vWI7KXfpJmJbs1TMeS3/eNWPEQbll3NJqLQoMjrSQfuvgH4vYrfVywa6jG+\nk1gpZdTNFFMmyW/GlGRU3IdlGYl1goMMzbtvaK7X+eAZx0vSOqkf7iZyEZatEXR0ndJH4i4l\nRsgIFQIVzgcB4J5haLQRGC/FGsJmlM30vWXqP5e7elz7KyVD5jHyrT7PRlWY6fv0+xtf3VJh\nXjfBiP57Ft2+t1XuqaOqkb/6eirJ84mw0HHplyio6519K2+VVN4/6pkeRMQadCKRwNVNKiCq\nap4KInJrFZZ7JI5GRjXgqVtyctmADndeVSJ/sXeeKZfIh8h0LGfqMSJiAod4vxVGRER+6veX\n3MPR8/LNvmL9qjdy40uFzaPcnpkk92SIzuROW6W1iERRT/m0RBACwD1DEDYCQ6FGrCg30/cV\nQ7cP/4wUxMa+My/29I7IruXerbu+O/Gfd2s4lF5TIGtrnRDVVeau1WkqTNZdeWvV+x9+c/17\nu1lRYPPXv/NkiKqap4KIGKWKNA2cDVzg6Ulp6SbqUvbCMqcbsz1E1snlRD29vn1alnel8JvV\nxRdGSSPvfXYdpVKQdkX45DuBfhbDX5/nbIkNfLojURfP77/1KEorWfN53r7W3gNtcSIOAHyC\nodFGIHFVG0vKzfTda7Sft1zk2aNlb4Xmen6FXU+/v3Fwt1/K/xn6Tkq57VK1m67A+iOFuny5\nVF3hF1Z5a9X79/nwkb1HHvx6ivnbeYk5RFXNU0FEbHERqRs6G3h4f6VxT94f543FelZzo2TN\nhtLgQQq/25sZxqO1ekof4x87DPX4Oi+ovczDepiyg1FSTEHMFVOpmVgLsUaLodqZ7QAAqoMz\nwkYgbttBsjtBT+3LZvo+ti1jzEwvQVzSEa16mnuFXWs7I/To1rPwh7WZ9z2jTl+TmNgrIryW\nrR6Gu9fkbv40L2xy83B31sKyRp3JRFTVPBVEpElM9IyIaOBzFzVXv/R04fotOW9mmEXuko4D\nvWf0E961T+AQ15D3NAeHed+n07y7nJ5eom5efrOm6MNXCq4QEWVM3Sp95GOfoRl5s/6SLH5d\nqQ5TPxWR9908TbZZGNbb4+n25JovPbk2+7UEs8Bd0nmcx0DPBpYPADyEm243iuSLzy+Vvv95\nqFvZd4Tu1/5KTZd6jHr77u8Ia8Vmp3333qm/L+gV7VrOWtSllzdz+v2Na1qPXvGwrMqtldaw\nWfvOf7Yi8XwO49U2aNJr3UeGCYkoZ9/pjz5Piks3S/29hs2JmtHfhbka+9wq1UefNlfVVhIA\nAKcgCBuH8ei8A5ceH/RUmH7L1H+LF4x4LNjeFdXCdOL9/eceGjSt4X2EAADOBUOjjUMctWBw\nQ66+bHKiyHcHR9q7CAAAO0AQNjLZ2NUj7F0DAABUD1eNAgAAryEIAQCA1xCEYDu683kfbzE2\n9OorWzBcyPt4sxF3HgWAOsB3hGAzppgd5qhZYlvcatwctyVv3QF9sVza7xGPiR2ETM2bLue/\nsLTYegsDaW/vr56SSdqr2vypOX6/V5RLw6sBAG5DEIKtZJQcEclft0UboiG24Ltz4mfnewZm\nFy3/quDMYs+u0lo2CSK9Vk+XlzuGuGuEaeMpS1Q/DHoA2JzJYhIJuBMfNb9LGG/u/mBityBX\nhUfznpOWH9E0UVHgjHTXDKXBEusNRDUxWe/+qPl+YdqsF9M/2aQrLL/fpbxZz9ycWv7Ps7mn\nKx7q+kW9bz9VW7XANUx1X7DuwrXaN1nO5c16LvWlhbl/J5YNiAY1EyddNTTWswXgqVNZ5x7b\nObPXryP6/Ra98NjSAn1Dc6Ho/OpnB4T7u6qDukxYcjCbTVkexTAMI5K5BXUe+cLaCyV3dk1a\n0Uemmraz3A/r1kUzFYS9cbIeNdQc6Sc2bBQ+8b9T/wujS188NnDq0qGX3m/oLbiAq0qKLXL5\nnSHM9BvspBf8Hxfo/1yZt+VSwJS2tza09fjyO4+aD1VcbHZtab0xm0DtyhYV17apjfvnX7qz\nJkvqqfwvvta0+Ng9XECMXEDFFhNGPQBsJiH/6qx/XjOTxcJatKbSP6/+faPo5neDlwmY+o68\nFG6bGb1M8dHPp6NbU9LeL9ZsvTmXqM+y9EOzVVnXjvzy+tMPvN7iwsp+ciLS/P7+Qf/BFW+j\nKHtsG/sYke5/0e0TFyUu6lzPKmp+k+j96ne9iciizxFLRC5+fup6/i3AAwqloDT/zoUy3p0U\nER4MkaxvJ8HaNDO1vXXH0Ut5sz4rqTCvosBl9jee5WaXIqVSWFhkJhISWTSFjEpZp02MSBDU\nU9X1j4KbGgp3J7bUQkoODd4ANJ29yfv3JO+3Lg8NuW9IyH3W5aWnvjSyZqI7/9HPZl04kx2b\nryuocv9a6bat3jPsk5RHuomJqOO4xZ8SpSwnIiKRwqf1kDlr5+8JXX9kZb9BZDyxZI3f689J\nJm6yzVOsoNb3iZSlUcGvHiNhwIhV/8wIqm1v4C9ZS4n8iKGYykZHs2NLLg5QhwkM/523BIwv\nd9/tOpwRNm8nzdhcdKm7a2B20f5kWXSLOm1iTZa0U0WnTeIn1UREKcnGFqESmz09AB65qrm+\n91awtVQ3G0JlwXajKKV8Clpd01zP0xVUuX+tUm/eDA4LE1e7XRUWJt6coiXK/fqjrGk/RIpe\nv6fnUVe1BmHQK0fZOSUpx758etwD73S/9GFXXHsAVfNT9DLmny5S9lcREfkHM4e/SP8yh2nW\n32NG29p+tiJJB7dpV/LWzE8rkUv6PuHZRUZ06dYEFJU3EV3ZkPbhP2ZGJHAPlA2ZqW4rICLj\n6Qui3nPxam0aJgtbaip7fxQJGLkIt6x1bqHq5rfP6kLLzQ4ToPDJKc29a+cAhb+71K3K/WsV\nGBx8MybRSNVlYVFiojHoAZfCv17c1/+d39W0t+6Hvhd1vum2Yf0Dsm9GZO2b4XX3Ftx0G8ro\nzud9fk316lhxYUzWZ6UeC0bZbWDSEJe3/LLylfESJGGTOJyqe/VAjnW5T6Dsk/6V3iaAEw6n\nHXtx31sswxDLEhFDpJD77xz9vVx0b7Pq3FG47bH2byo+WrMguhUl7f1iTf6zcwsfmkRbDj3v\nmp10bMPrU5cHrL3wBfNmUL8Vqbd+JvDlIylLK9zLuaHfEdb4PnHl+znzNsemF5cWXIv56PuD\nLXpG4uUNNZB19HhtrC36CBtMEuHxGlLQPk5m6pedKrB3FdAo+gT0fL3z6NsfcN1EogFdFtU/\nBYnINXrV1jnsN5M7BwZ1f/IX17Gjg4no8Fx/RuIedt+cHQELtizpJ6e+y1NYlmVZds+zzZ7+\n+64UtIEazwgtOYeWzXlpxda4QmWLnuPe+OyTxyLKOrUSExPz8vKsy4cOHYqOjsYZIQBvFegt\nV/KNRPTNec3FXMOQZi4Le9fyRTA4L51Jn1R4QyVRBij8BYwjfPJtqBrHrgRefV9ed/zlKra4\nuLgYjUbrskzWgI8DAOD83KSCSD8pEW24jNNw7pOJpG097Hfmk7Gyr//zh2898nx2T87XQxp4\nyHp+iRMQEBAQEGBdjouLa2ARAAAAdeI3+xA727aHRJsVANjMWz3d9SZWLsJ5ITgTBCEA2Iyn\nTFj7TgAOBh/cAACA1xCEAADAaxgaBQCbWXws/3SWvk+AbG43N3vXAlBXOCMEAJvJ1ZnTik35\neou9CwG4BwhCAADgNQQhAADwGr4jBACbQR8hOCMEIcA9s2iyz2+9nHitWCeUebZpETmqmZc0\n7Z+3ziQTMWKxMtC/20PtW3hUdw/GvKOLjlwqJvJtNX5u6wqTXVv0afviThzJLmJlvj3a9R3m\nLSciyv1v4dHLWiIikgQNW9gpsHGfXIOgjxCcEYIQ4B6ZNWd+OJvZvuPQcZ4ulpLkXWd3/8KM\nfUpEosAhizoHGXVpO47u25fTYrx3NT/vEfXOqKjcq3+uNd+9JT31mjmg3wsdlfqs/747d771\nkJ7NiYhIEDB4SZeQRnxKALyGEQyAe3Qt+ZKoRb9hvmqFSKxSh46LaJZ+PTGn3A4sicTWE6Pi\nM8tiTqbU+ciBLfsO9fNQiSRe7p4qkUR++4CZB+b9vW7xwX/35xls9jQAoAzOCB1dvr4gIf+a\nddldqm7tHmrfekCfrzX7eKtuPxYq3TxKNflEptS9b6QSkcC3xf0zrZMQKbvMHXTvf4MpO+b8\nzZYdhvpaH3r2fnd4b9aiy0w/vvb0qaDBvUIdd+Ib9BGCM0IQOrpz2XEvH3jPutwnoOdHfd9r\n0DSY0GBSN7kwq7iIqCwLzcUFeXKVO5WKAocs6uSnybvwy7n4K2F+HST1OTqrT91+JlbUavBI\nT3H59YxA5hcYHnH5RLqeQh33BYA+QnBGGBp1JofTjn14Yrm9q+C90GbhxqSDu7M0WrOxSHN1\nc9wN/+ahXtZtjFjt2Xl8cMGuqzn1yAJTScL6E3GqtoOHe0rv2sRadJmp8XEWD/+7twBAAznc\nGaG58OrBnf+cu56jFbr6hfW4f1hkgDRuw8JN8UQCsVztHzF43Kj27tUNDSXvWLr6eDGRz4DZ\nzw3wqrDp+raP/3ey7NK7zo+9NTaMiMiYdW739n9j0/UuAR0Gjx0R4eZ4Y049/br9NWYdEb13\nZMnZ7Av2LgeIhOquT3c+t/Xy7qXFeqHMs03zYZNDFJR2ezvj26JzwP7TJ1sM62E6s+y4eeKg\n7kHlf14f99Xe4zeIiP5444pP9OBRfYsPz0/weLl328xrJ85rDOcPrfubiCjk4RGDuwjo+sUN\nXyeVMgKJqyqwT9duDjwuCuCkGJZlG3iITZs2dezYsXVrW0xYbE7f+/VPN9s+MDqqhcqcEx+z\neWdh1MxHpTsWJ3R9Z3wrY9HV3Wt+t4x4fXSN35PlHfp6g2lC5SBcejLslQnh5ddZEvbvKAnt\n386Xydz3y8/FA14f18ZxT5Gfi3ntWMapUS2GLuz1hr1rAahars5s7SN0lznufyWAuzjYi/X6\nyROiXmMHtfFykUhVAZ2iR7TNOH4ut/werFhs/eokZ9+q5XvTqjxKNSyXN3+46MNPv/7tcHKp\ndY2g9X3RXYJcJYxAKBDIXVwc+rO2i1juKlHJhBgZA8flKRMGKEVIQXAujjU0WlpQYPIKc7/9\nWOjt7VGQU0BkOv/L/PNEJPSJmjLV2k/lNeC5Ofdy7ObRr70dzZq0WRd3bvgtJuClUS2s/1fj\nNy3ccMEidO849qkghw7Cpf0W2LsEAABi9XpjerrAxUXk7U2MQ79t1pFjfXCTq9WinJz824/N\n2dl5bm5uRKKOk+fPe2vuk31l8cevauv/FzAiF9+OkeHCzMziW6vCJ7w3751Xp/cx7t18srBB\n1QMAcF1xzL+pc17OXPhB+htvZy5435ia2sADFp1f/eyAcH9XdVCXCUsOZrMpy6MYhmFEMreg\nziNfWHuhhIiIMre/3L+52kUV1HPmpht3vs/TrYtmKgh742Q9anCsIKQW3bsbj2yJScjRGvRF\naee2/X3Jr0cnT+s2RqJuPiC6S3bMwdR6X5vNmrRZ50/Em319lUREqcd2nrhRoDOzLMuyJqMR\n13wDNMjiY/kTtmYsO1Vg70KgUZSeP5//83rWUHZfB8PN1JwVX7B6ff2PWLhtZvQyevbn06kp\nJ9Y8Wrhl600i6rMsndXlJsQsHXbz3QdeP1hKdPGb1/7p9VtS3s3to8+8tuLE7R+XPbaNZVm2\n9MdRoW+fYVmWTVzSvR5VONbQKAkDBj0+7sDOf9d9kVMqdPVr1ePxCV1dKe72dsYn6j6/lTFn\nej/eTb9v1TrT2DlDAsr/fPHR1Ut3JhMRrZy/L2T4S1Ojcv5ass9v9tQeeTuXrj5azIjkrj6h\nPR8aZB0X9QtvfnHrTytuFAlcA9oMHh3p0C3AWlOpyWISC8ToIwSHhT5CbtCeOFl68pR1Wd69\nm0tkWboU/rWdIbpziSVrMeXk6hOuWHS6KvevlW7b6j3DPkl5pJuYiDqOW/wpUYq1Q0yk8Gk9\nZM7a+XtC1x9Z2a+tr6+07L4iIn//6u5eWG8OFoREQnWrgQ+3GlhhXftJ77S/tSyNmPhyBBGR\nqqrvCJVRU+dHVVjjOuaNlkREyuGvzB9e+S8LH/pY+FCbFN7oXjkwD1eN2pk+6+Da/PbT2lTb\nv9NkjNmHfsxrN62Nh4ON6QA3GNPStCfKxhhFAf6315vz8io3Gpiyc8xFhVXuX6vUmzeDw8LE\n1W5XhYWJN6doadCj7/brNcDz41JJ6IzNx1rU/S+oG4cLQgCHVXg0Ude5u01SsPTK5QN/Jmdr\nRV7dI/qP8KnigmWT5vQ3R2MFrR+a2UJO5oLTl/7bk5arF3t1Du832l8p9o5omXDmfMtBnat/\nDwGoL3FAwO2zOnHAnWE3kY+PuUBDVCELxYEBgkJllfvXKjA4+GZMopGqy8KixERj0AMulv3P\nPxv3zOmCZwNzNjwxYs72wT+Nsunl8whCgDoqvnpaEDK9XjdOu4sx68SvmR6P9B/oWxL74+mT\nFwf0j7jrf6IpfWci0zlQdp6IiCy512+oI2eEuwsKz/544ky8X7+2jHuEl+bvDEPnYFsUZDOY\nj5AbXCK7Vzm86f7oI5kLFrIW5nYWCoNbSVuFkUBQ9+HQ8mTRU4e+8ers9cPByTIAACAASURB\nVD4LoltR0t4v1uQ/O9e6xazNTjq24fWF8Q+v7UXs7oJCImKIWAtbqNFUmrilgRCETgN9hHam\nL8jWuXZREBFR0fXt32v8AguvxJWKgoJ7PdI2UFl+15zD848l6MqvYUIfH9k/4tajlOxUr+Cx\noVIJSTv0dP09oYAiKtz+QRcXn+AR3j8oOcEahAKfzuOIiDUXCxihWKpgiIj81O4pWdkU7FDT\nE2I+Qm4TBwd5zxqb8+1fFr2ZEYpEKoF04nMkaMDnHtfoVVszX3lxcudn00Vhw55fsSqYfqbD\nc/2ZlyWufuG9xy/YsqSfnGjkm4t+e3xQ4OuFLq1GvLfmQRfbPSUiQhA6EQ73EabNfYU1Ga3L\nAcs+ZUQO+bLUGvQy0Z3Tr2yNOTpywlgme+/JA3tyHh5XPsm8+swf1afmQyndrZMsSZQSY4mB\nJbozOKpJOXnZo8d4BZNc/mcy9r11KskiUHXpNCKYiIgYsYQMejMRogeakLTLyMCvRrJmMyO0\nzSvPtdPT3+57+ts7K+YcZStd/yFuN+3nU9OqPYbsyW2JDanB3u84+oTNG1L6TBnkY/erDyoz\nJv75842eUwb7YZinkVlKSliTybqc8/lK75fu6VYJTcVFItXp7kwH6OHbqpVcROTf3VeyuVhH\nXuWu5a3tjNBFIi3WlxLJiQzFBpFCUv7ln3/86pXjxVeOn7E+3PCFZeLzoUryG7B4ZL+S4it/\nnDh4zGN4lIxYo4EkUqQg2IOtUrA+Mlb29X/+8K1Hns/uyfl6SAMPaecgzDtxUNthki1SkC2+\nGrN5x6kUrSywy4hxQ1upmJo3Vb49d1HCni27ztwsEqhbREaPva+ZXBzWq/m//8b1ebgD2hUa\nl3r8WNbCGpKSSk+d1sdftnc51ZC6ectS8rTkbR2WyctKTGzWpRmTcyrT6BNe8SVS2xlhkHdg\n9uULVwM6+2pjj2kCB1Xo23Efet9T1iuZky/9ul02ZmYLecr14zdd23Vxk7EssWaTkSUiyijM\nD3K3+XXkDYT5CKHR+c0+xM627SHte7KTc/68sE24wgZHMl7Z/UeCb/TMubPHBlz/c89lQ22b\nQka+Mn/+C0P8bu3FXtu/OdF79MxXX5s1oXny1pgkExH5hIfmnLuou/svsw+tqbTQUFRqcpBy\nbEk1/H7XkcPlHdrXvqs9KUO7WpLjbr20vF0Fp05sWrTvYIpnn6FeNf5gJWKf7g955/5x4NfP\nzma37BjZTkSUc3j+f5eKqtnf39evIHHXx7s3LDudpAjvEyUnooK4bNcufg51pQyhjxCck13P\nCPUpKXq/AdbP18XHV69Na+afcSa+QBzQNfrBYaF3rj649teStacrvP8LOk56b3z5iSTSrl71\n7DKzhUpGqj7dfL9ITKHwlrVvKo9lqex7GoaoKDExk1oEkq+/T2pCKnV1hFnh0Udod65RYbK1\nSfnd27gTkUDR6uFO3ep7KJc24cNfLf8K9uozv1KahrR9eCYREQnlISN6hIwot8mYHXfNs/Mg\n9E4A2IBdg7BUq5XK7owp5aSb7p/8wihByv71W/Zdmxvd8tboZssxb8wfU/OhtNpSRZD11FKu\nVBq02jtXH9SwqRymZf8xLTZvWfVxsTg4MtS7pKSEiIiRyUirxfUITULWoYP3Ky859D18pT79\nnvGxdxFERCT27jPd0YZFAZyVXYNQ7uKiL7xzpufeunOoWkLUsnMb6bZsLbW8NWZahzNCFxd5\nSUkJkZKotLhYUn5GpRo2lceo2w5/su1wIjJf+fPTK2o1ERGr05GLC1KwSQjd3YTuTvLFkqr5\nqBftXYNDQh8hOCO7BqE0KEh6LlNLQdbR0fwr565FDgxmUs9d1nsPKdcoUoczwoDQ0Nx//kuK\nuM8n//CpjNB+gXXaVJlFl5e4/5+Lbp2n+xARZWVkBTpKnxb6CMHxoY8QnJF9rxr16tDRtDO+\npFtXBRGRl5/w7C8rfi0QBXR94MGW9zZAJm41dOyNzVu/+rRUGth5zPhwKdG1stttKytvoqpu\nz12wY+nq41qJyiuk04MP9/JkiCjrUqJnx8cc5JpRDvcRAgDYEVP5Dqr3atOmTR07dmzdunV9\nflifsHlDSu8pg3xLjq9epxs7o79HA6uxIWPiXz9fj5wyxB/DPE1CF3uhaNduEgq952LYEQCa\njr0b6qWtxz1RrwRtAuKwMU+G2bsIHjHn5+suXnLQ28pA3aCPEJyRw7zpKHtMnWHvGhwb5iME\nx4c+QnBGGPVzGq8cmDdw09gPTyy3dyEAzk+3u/RsZEGMMn9feOGFL8tu1sNuKN7F5O1i8vZ6\nFhyfYSitIc4Paf9l8nYxeYfnV5oIgbXkvlV00Ds/Jqzo8g4LEZHWnDar8IB7fkxY0aWNloZ+\nGQW25zBnhMB7TtBHCNzAntaeetTks1rVbqCAvWy4MqXonFjdfbqAiASPqoauE5vTjAmDiq/t\nlUQMq+YQfV0Gsi7aJZozle70ZN6hPb9N2Om8Unm19NQEbVai0nu/oaC7S48lQuZc6emRpTnj\nFN5443Us+H2Ao3CmPkKohnP0EeZ9rRfMVbcaLSAi6iZt97lp/1x9yXR5+cl9WJYRWh/Hlx4e\nzna47uJat4MX7jYpnnb18GfIXx7cqSDnGPmMkrcjIgsZJAzjSWJ0mDgcBKHTQB8hOD6n6CNk\nS6+zihF3slrQVuhyw6wjciGy/Fy062ciIuUc1559iYgoXN7n+j0c3ZBjkURZD85IfFlDNhER\nbSnePc7ASgT+/3NVY8zD4SAInQb6CAFsgpE1Y0ouWWhcWRayl8zaZgLrJWiCR1VD14p0B3Xn\nntDlvK30u8fbqRORxEtgyLIQCYhYQyYjsR5hrHKYmQxx+oujSlL6qxzlLh1QxrFHMIBPdLEX\nspd+lr1shb0LAa7zeFZq/qz4yjazsYTVn9LHvWBQzZbemQVHwMjuk0dMNV/50FSPC1tch4pK\nVuvy0lnD4dKbZ8WePUnzhTb5oMVkZMlCrJ41G234VMAmcEYIjgJ9hBzgHH2Egu4u3X8qjX+3\n6OAkVhAk9Jqt6vTM3ecEyjly17alqS+rggqr+o4wQ3fMX1tARKTZtUAUftO1WXzJP/OFfQ/J\npKNcOhwsietQYFSLAr9X+KiIHhBlTi86cMDCBAl9PlAEN2+q5wl1hTcdp4E+QnB8TtNHKBsh\n7zxCftdKZpJy6KRbD1wlnVIlREQBVX1H6CfryVb8fxikGGydJ50ReH2k6vdRuU0hktY7JY56\n4xAgDI06EfQRAgA0BpwRgqNAHyGAU8hN1WSnFEjlYv9QL5lCYu9ybABBCI4CfYQc4Bx9hFBv\nLB36/fyVkzetj8Qy0X0Pdw5u61v/A6Ysjwpe0PL367+MV5Npw9j2FxbFL2qfsjxqEm05NMdP\ne27pA08cfXzThilhjZpVeL06DfQRguPzlAkDlCJ3Gd5YuCn+2I3bKUhEJr1p/69ndcWGBh20\ndUj8x5/HV75AV3tu6QNTjj7a6ClIOCN0IugjBICmcT02Pel8unW5RUf/5h38rcsJx2+W341l\nyagzZVzPI5atcv868X/i7Q5rF/354rrocitLz3/6wI/XJm/69clGT0HCGSE4DvQRQlMwbS85\n8a7ZEe58bd5ZcuIth6ikkvzMouux6dY/+ZlFt9eXlugr76zXGqrbv24k970xI/2jLxLKr0vc\n/EPKgFceatU052oIQnAU1j5C/aV4excC9bf4WP6ErRnLThXYu5BqmW9+YPF/QWiDS7IqzzJx\nW4npxgTNPlX+Xj/NmY9NpmpWCofL3PeWZjjiv5W7r6p5B3/rH3df1e31Hn6qyjt7BblVt39d\nBU59p+vGRX+VS9AOb26affmpsZ9frCJ5bQ9Do04DfYTg+By+jzDekCaVRHrb4EiVZ5nwUZZt\nMvykTdTKolKl0kx9bC9t6mTXZkFVrhT63m9O2MT6T3O0S6WtkVZ5fd8JnTYvjdHrLQwREcMS\n6x4e7Bng6hngem/DoXeTDHxj+sIHv8ym28Ojknaz/lhXcP+Dj7j9vXFK80a+hS3OCJ0G+ggB\nGsp01GTqLLRe8K9fWXj4ydLYLgX/uBeceNVY4YKPvSX/MHm7yv8RFWdVPFThbpPiaZmHPyPp\nKw/uZMw5dmeT0F8gti4xRK4CqaralcquQs1/pkZ5qo1CrpKOndHeXSkkhhG6yP28xYEDOtjm\n0MFPv9MtJ67CKpfub2/9KvTz0c9syWjkAWScEYKjQB8hNDpjLisqN/tDyWk2fLu6ndB0dUxJ\n4j9u7Qbf2jBEMZhVVHmE26qeZYKIiISj5C2+KzqsLmEFAq/PVb7qalcybgzlsBZnOiVx8Q8e\n+06wzQ4XNOfoPuuSdPBXKeztlYesS24DPjkZa7O/rFoIQnAU6CPkAEfvIxR7MqaUO2cX8mip\nZzBDJA4Yw1yMs9DgW3XvLflnqL7CmZpQ0sWk9Cm3oupZJoiIqGhR8Y1ARf9isTjfmDC8+Gof\ndVjnqleyBSx5MY76r8UfCEKngT5CcHyOPh+hKEokWms2ksg6Slm6Q5/7nNxNaErbyiqXlMuj\nOpwRug4Vlbyty3tIrrxWevOsuGXPO5tM+eydsLWwxoJqVxafMat7413Y7vArcBroIwRosHBJ\ngE6bmS0N8iYiUnRm0kZrziaR6zOKToNr+9mKhJVnmdhbNgGF+2suPk9oj/iYzXKh2xOK9vcR\nEVW10py1UxiwC18G2B3Dsg39FnLTpk0dO3Zs3Ro3V4cG0cVeKNq1m4RC77kv2rsW4C7T9pIz\nR2Xd3xcaVhaeKlD0fsdu57DmXSWn98m6f2iLXg5oEJwRgqPg8HyEN+OzCm51GQeF+9Sn0cpJ\nOMF8hKJRishR9i6CiIiE9ysi77d3EUCEIHQi6CN0XtfPpyeeTrEu67WG9v1CZUou3LO/Mofv\nIyxHOtu1t71rAMeAy5WcBvoInZdMIVF5uCjd5UQUu/9a5o08e1cEAHfgjBAcBYf7CCNHtY0c\n1daoN62bt8vetQDA3RCE4CjQR8gBjt5HCFAVBKHTQB+hsxMImIAwLyKSKzn7S3T0PkKAqjh9\n+0RsbGxmZqZ1uX379n5+fnYpAwAAnJTTnxEmJCTExpbdi85oNPbt21el4uy16dyGPkIAsAun\nD0KFQuHu7s6ybEFBwbFjx0JDQxGETorDfYT84QR9hACVOP2bzvDhw4cPH15aWvrRRx/Zu5bG\nhT5CZ8da2PRruUTk4eeKPkIAx4GLu5wG+gidnclo3vX9sV3fH0MfIYBDcfozQiuhUBgREUFE\nGBd1XhzuIwQAR8aRIJRIJBMnTrR3FdAg6CPkAPQRgjPiSBDyAfoInR36CAEck9P3EQIAADRE\njSMYbP6p72cNbuOtVAd0HP32ztSGRmbjMZvNcXFxcXFxRUVF9q4F6kkXeyF76WfZy1bYuxAA\n4Jcah0aTNn59vP07WxdE+mRvmTXywdntMjc/qmyqyu6JwWDYuHEjEU2ePBnXyzgp9BFyAPoI\nOc9kYXckaS/mGlQSwYAgeYQXFxqBajwjbDn9u29nDmztpXRrO3lSfzY1FRd925HWVFpoKCo1\n6exdCNQTa2HTEnPSEnN0xQZ719JY0EfIbXozO2139pLj+VuvlfxyqWj6nqxf4hs2CJeyPIph\nGIYRq4Kinv0t2U7DjnW7uMty88dv/uk6eVxII1cDNUAfobNDHyE4u58uFl3JNxARy5I1s1ad\n1aQVmxp00D7L0lmLNvnPwafe+Oq0DYqshzoMQ7EZf82IXu6zcveLrRq/nnpCHyEHoI8QwEHE\nJJfG3NRalwcFuwwKkVuX/00uJYbo1nkbS8SydC7bEJ9nrHL/OrP+t3dTuza89vqoLQjNyb9N\nHbFQ9MHuH8b6O/AbFPoIOQB9hByAPkJuSNIYY5JLrcstXMVEZcGmM1fdZ1Dd/nVyeK4/M5dI\n4D/2h6NhDam6/moMQlPSukfvX+L6yZ5vH3DkFOQJ9BE6O/QRgrNooRbfPqtroRbfXt/WQ5Je\nUmEglBEwnbwlUiFT5f510mdZ+qE53tqULc8Nmvm/6O1PeTWo9HqpMQgv/DD/tytXaWzA90RE\nFPjikZTlUURElJaWptForHtlZWU1bo1ARERL+y2wdwnQIEKx8P5pPe1dBUDtBoXIqxzefLeX\ne0qeLqHEwjDEssQw1LmVOkApClCK7n04tAKGISJdTk4RkaMFYedFieyiKrdotdrbHXs6nf2v\nYzSbzfHx8UQUEhKCrwmdFOYjBHBwUiHzw1CPuXvyzpayLgpJV4nRP6TBLXWH5/ozcwVyj1b9\nZn31WQtblHnP6tmzFRZ2Zyg3OTnZRsXUH/oIOQB9hByAPkLOE8pln48JsNnhguYcZefY7Gj1\nhTcdp4H5CJ0d5iMEcEy4uMtpoI/Q2aGPEMAxceSMEH2EHIA+QgCwC44EIfoIOQB9hByAPkJw\nRhwJQj5AH6GzQx+hU7OUJt2+pYpA3oIIQxfcgfkIAQBql79XyZpLrMvuQ4oZocK+9YANceSM\nkNt9hCyx25P2bLi8Ob0kI0QV9ES7SQOC+ti7KNtDHyE4i4L9Ie6Dcu1dBdgMR4KQ232E6+P/\n+PT0KgExFmILDZdePvDe4j7v3N9soL3rsjH0EXIAh/sIlV23Ems25uzWXf+ENRXYuxywJXyn\n7egsrGXVudUMMRZiicjCsgzDfHH2O3vXBfcM8xE6NbHHQLHnEKGynb0LAdtDEDq6lOL0UrOO\npTtf5bIsm1GSVWLU2rEqqAf0EQI4Jo4MQ3G4j9BL5i5gGEvFa5pkIqlc1KBb3Dog9BGC45P6\nT5b4PGDvKsDGOBKEHO4jdBG79PLv8V/a8XInhczw5oMFnAsM9BFyAPf7CAUSRsDN2+PxGXdf\nrxwyP+rVLj4dbj/sFxj1UteZdqwH6sfaRxgQ5sXtPsIApchdhjcWcCboI3Qal/MT00sym7kG\nt3ANsXctAADcwZGhUW73EVq1cQ9r4x5W+35OC32E4PgM6etLLs0iErgPyrF3LWAzHAlCbvcR\n8gT6CDmAw32EVqxFzxrzicHYL6fgTcdpYD5CZ4f5CAEcEz7XOA3MR+js0EcI4Jg4ckbI4T5C\n/kAfITg+9BFyEkeCkMN9hPyBPkIOQB8hOCPuvl45B/MROjv0EQIHWHSpJRdnFh7pWnRisD55\nFbHmBh0uZXkUI33gJ+tNzM27nvFhgucctUWd94QjZ4R8sLTfAnuXAA0iFAvvn9bT3lUA1J9F\nn174XyeLMY+IGIYx5sWYCk8o2v/YoIN2iMje8lfh41NcTf9uvBbc1R5TO3Pkg5vZbI6Li4uL\niysqKrJ3LVBPutgL2Us/y162wt6FAFTLkL4+P8YjP8bL3oU0LkPGxuJzD1n/GDI23l5fcmGq\nxZRLxBKxLGshIn3q/8yFZ6rbv048xo42/rm1kEwxG5MGTmxlXcle/+3ZPsFqlVfb0R8fLaSz\n73Qe/eKr/YPd3MMnrb1699YG40gQWvsIN27cmJ6ebu9aoJ6sfYT6S/H2LgTqb/Gx/AlbM5ad\n4ux0fdY+QtaUb+9CGpe5OM6QsdH6x1wcV259LFW6F5mp8ER1+9eNx7gxpi1b8/7ZeGXQxHDr\nKt2f77yW+uju5IzzX4T9NHN5PFFxbEHvtZdSjs/IWLr6RKWtDYShUaeBPkJnhz5CcBZCZYTE\nb+Lt5dvrGYkP6dOo4o05GbGnUORe5f515DVujH7aS6s1Axa2oVgiIkq9fDniwa/aqhU05JGh\nRcsTqKVL/wfHNVcS9e0u+ObM3VspvL7P1ApB6DReOTDvWMapUS2GLuz1hr1raRRmjYaIiGUt\npaUCOdcmmaJbfYRENOjxbs0i/OxdDkC1JH4TbwdbebKgZ0suzqiwSugqdu/PSLyr3L+uPMdG\nl8z4csC5NpRgXRHYps3F1T/Hj5/iemL9HmWb6WQUCG6PX7IBd29tKI4MjVr7CCMiItBH6JRY\nNn/tOs2Wv4iINZszXn9Ld/GSvWsCqILUf7L7oDz3gTy90ag0eLo0YOidx4xAEvYLI/Fu8IE9\np+82nnuv3e3HsjELP/RdMzjYp8PshEe/fPGuE76at9YDZp9wGs/FvMbVM8Liffvz166785gR\nCFzk/ksWCxQu9ivK9ox607p5u4jTZ4S5OrO1jxAdFBxmLjpn0hxjRG5iz8GM2NPe5dgAhkad\nBof7CEvPnCVGQOytL5ZYi6WkRJ+YKO/U0a512Zi1j5CIuNpHqDezZ7L0GcXmIJWob6BMJMBN\ngrhJqOokVHWydxW2hCB0GhzuI7QUFlKlkQlLoQ2uinYo3O4jvFlkevHfnIwSk/VhsEq0crC3\nt9weTWEA94gjwxfoI3Rq4ubNqNLJg6R5czuUAvW18EhelvbOTUZSik0fHudgjwFP+gj5hiNB\niD5Cp6aOjhbIZeWzUNGntzg4yH4Vwb0pNlgu5hos5U7rWZZOZuiNloZeguBoeNJHyDccCUI+\n0JpKCw1FpSadvQuxPaGnh+/896ShoUREDOM2eZLHU0/YuyjbYy1sWmJOWmKOrthg71psrMRU\nxUV3ZgurN3MtCIGTEIROg9vzEYq8vBR9+xARIxSqhgwmAQdfmRyej9DXReghEzLlptASMEyI\nq1gp5uDvEbiHCxfLWCyWs2fPKpVKlmXj4+NDQkJkMtx7xflgPkKn9lqk21uHcsudALKvRnJw\nUi3MR8hJXAjCTZs2Xbx40fpp9MyZM9evX58xY4ZEws1bWHEY5iN0av2D5N8N81l3sSijxByk\nEj0RoWqpFtu7qEaA+Qi5yOmDMCkp6eLFi0R0+84A+fn5R48e7d+/v13rsj0O9xHyBOf7CNt6\nSD7oy4X2auAbpw/C1NTUOq50dhzuI+QJbvcRAjgvpw9CqfTuD9cMw1ReCY5PF3uhaNduEgq9\n575o71oAqmZIX19yaRaRwH0QT283yklOH4QtW7YUCoUWi+X20KjFYsGNT52RdT5CRuT0r0k+\nW3ws/3SWvk+AbG43bn7da+0jJAZXw3KK0/86PT09o6OjhcKyOzkxDNOjR4/27dvbt6rGwOE+\nQp7gcB/hbZiPEJyR0wchEXXp0uWZZ56xLg8fPnzkyJH2raeRcLuPkA843EcI4NQ4Mgzl7u4e\nERFBRCEhIfauBeoJfYTg+NBHyEkcCUKJRDJxYgPmRwYHgD5CDnirp7t1PkJ7F9Jo0EfIRRwJ\nQj5AH6Gz43wfIRF5yjDvEjgfzFAPAAC8xpEzQrPZHB8fT0QhISEqlcre5UB9oI8QHB/6CDmJ\nI0FonY+QiCZPnowgdFLoI+QA9BGCM8KbjtPQmkpNFpNYIJaLMLeGU2ItbPq1XCLy8HOVKbl5\nwQX6CMEZ4XON00AfobNDHyGAY+LIGaFQKLT2EWJc1HmhjxAcH/oIOYkjQYg+Qg5AHyEHoI8Q\nnBFHgpAP0Efo7NBHCOCY0EcIAAC8xpEzQvQRcgD6CMHxoY+QkzgShOgj5AD0EXIA+gjBGeFN\nx2mgj9DZoY8QwDHhc43TQB+hszPoTNY+wrTEbHvXAgB3cOSMEH2EHMDtPsKsG/n7Npy2Lh/a\ndN6oN7fpibkznQ/6CDmJI0GIPkIO4HAfoa7EsHftCYPWaH1oNluObLng6qXwD/W0b2E2hz5C\ncEbcfb1yDvoInVdqQra+xHinU4klIrp2NtV+FTUWT5kwQClyl+GNBZwJR84I+WBpvwX2LgHq\nqaSgtPLKovwqVgJA0+NIEKKPkAM43Efo7lf5Ncl6VLESHB36CDmJI0GIPkIO4HAfYWBrH68g\nt5zUAuugKMOQSCpq16eFveuyPfQRgjPi4JsOxxQZilOK04hIZ9bLRbJmqmD0ETodgZAZ+lTk\nqZ2Xk2LTLCaLd4hH1Oi2Sne5veuyPfQRgjNCEDq6U1nnXj7w3u2Ho1oMXdjrDTvWA/UjU0h6\nRLdNOJFMRO36NHP3d7V3RQBQhiNByJM+wgFBfTicgtzuIwRuQB8hJ3EkCDncR9jTr9tfY9ZZ\nl7k9KMrhPkL+QB8hOCOOBCGHyUWyQKW/vasAG8B8hACOCUEI0ESEYuH903rauwoAuBtHghB9\nhBzA4T5C4Az0EXISR4IQfYQcwOE+Qv5AHyE4I7zpADQRzEcI4JjwuQagiZiMZut8hJk38uxd\nCwDcwZEzQp70EXIb+gjB8aGPkJNqDsLrS7q3ePMUEQW+fCRlaVTTlFQfHO4j5A/0EXIA+gjB\nGdX8em3+xkmWZc+8HtpE1QBwmLWPMCDMi9t9hJiPEJyO0w+Nnj59OjW1bILTrl27BgYG2rce\ngOqgj9DZsaZC3Y3lJs1xRuQq8Z0o8R1n74rANpw+CJOSkmJjY63LzZs3RxA6L/QRgiNjTQWa\nw50tuhvWh4b09bLmL7u0WWrfqsAmnD4IfXx8WrZsaV1WKpX2LQYaAn2EHMDhPsLSqx9YdMnl\n1+iufyYNnCpUtrNXSWArTv+m069fv379+tm7CoDaoY/QqZkK/mOIYa1zK5dhTQVHEIQcgO+0\nAZoI+gidGiNUsJV6exihwi7FgG0xLMvWsHnLY7JxP+vLHggf3mzaMJaIiNLS0jQajXX133//\nvWfPHjTwQQOxBqNFV0rECF25+VqyWNjkuAwi8mnm7uLKzRm1zmXrc0stvgphhCfXTnnNJZfN\nRefvPGaISCj2GsEI5fYqaeLEiWgbs4lahkbHrtOx66pYr9Vqi4qKrMslJSU7d+60eWUAnBVv\n7wIaWS7RRXvX0CTMRNvs+Ne3a9cOQWgT9fyOMCws7PZyfHz88OHDcUYI9WbKyaFbIxMiLy+u\n3lyGtbCZN/KJyN1XKXXh2gkTT1gMmawhlwQSgdTf7uOi1ttpQcPZ4GIZFxeXFStWtG7duuGH\nAn5KmT6TNZmsy0HffoULRwGgKeFiGbA/gUIhULhY/9i7FoCaGNLX58d45Md42bsQsCV89Ab7\nC1iGrmSO4HAfoRXmI+QkBCFAE0EfIYBjwucagCaCPkIAx4QzQgCACvYVfAAAC8FJREFUusJ8\nhJyEIAQAm8F8hOCMEIQATcQ6HyERcXs+QnuXAHDPEIQATQTzEQI4Ju6OYAAA2Br6CDkJZ4QA\nYDPoIwRnhCAEaCLoIwRwTPhcA9BE0EcI4JhwRggAUFfoI+QkBCEANJTJwpaaWCJ6oat6Vme1\nu5S7TRToI+QiBCFAE+FwH+GxdP2rB3Ksy30CZZ/0x0WV4EwQhABNBH2EAI4JQQgADRXhJVkx\n0Nu67Cbl8iV4hvT1JZdmEQncB+XYuxawGQQhADSUm1QQ6ce18d4qoY+QkxCEAE2ED32EAM4I\nn2sAmgj6CAEcE84IAQDqCn2EnIQgBACoM/QRchGCEKCJcLiPEMCpIQgBmgj6CAEcEy6WAWh0\nJoPZUGq0/rGYMTODE8N8hJyEM0KARndky4XE0ynW5X4PdQrrGmTfeqDe0EfISfh1AgAAr+GM\nEKDRhfdqFtzWx7rsFcTNqdsBnBeCEKDReQe7eQcj/7gAfYSchCAEAKgz9BFyEb4jBAAAXkMQ\nAgAAryEIAQDqCn2EnITvCAEA6gp9hJyEXycAAPAaghAAAHgNQ6MAAHWFPkJOQhACANSuOPZx\nsuity8oOP9m3GLAtBCEAQO2MmZtZc4l1mW3/I0OYVJI7EIQAALUTeQwki67sASO0ay1gYwhC\nAIDaqbputXcJ0Fhw1SgAAPAaghAAAHgNQQgAALyGIAQAAF5DEAIAAK8hCAEAgNcQhAAAwGsI\nQgAA4DUEIQAA8BqCEAAAeA1BCAAAvIYgBAAAXkMQAgAAryEIAQCA1xCEAADAawhCAADgNQQh\nAADwGoIQAAB4DUEIAAC8hiAEAABeQxACAACvIQgBAIDXEIQAAMBrCEIAAOA1BCEAAPAaghAA\nAHgNQQgAALyGIAQAAF5DEAIAAK8hCAEAgNcQhAAAwGsIQgAA4DUEIQAA8FotQai/9L8newQq\npIqgntN+vmJsmpoAAACaTM1BmPzl0y9cHv7L1cyrP98fN2v6t+lNVBUAAEATqTEIb/7+67nh\nr751n5+b331vvTrs+PrfM5qqLgAAgCZRYxAmJycHhoXJiIhIFhYWmJyc3CRFAQAANJXaLpZh\nGKZsiWXZxi4GAACgqdUYhCEhISlXruiIiEh39WpaSEhIkxQFAADQVGoMwuDxEzvs/GTx/oyC\njP2LP9kVOWm8X1PVBQAA0CRENW5tNvuH5eefeKTlEo1754dW/jQ94NaGvLy80tJS67JGo2nU\nEgEAABpPzUFIsoipa09OXVtpfWpqamZmpnU5KyurEQoDAABoCrUEYXU6dOjQoUMH63JBQYHt\n6gEAAGhSuMUaAADwGoIQAAB4DUEIAAC8hiAEAABeQxACAACvIQgBAIDXEIQAAMBrCEIAAOA1\nBCEAAPAaghAAAHgNQQgAALyGIAQAAF5DEAIAAK8hCAEAgNcQhAAAwGsIQgAA4DUEIQAA8BqC\nEAAAeI1hWbaBh0hKSlq5cqVYLLZJQfVmMBiKi4s9PDzsW0ajysjI8PPzs3cVjSgrK8vLy0sg\n4OznM6PRWFRUhFepU8vOzvb09HSEV+nixYsdoQwOsEEQOoi0tLTz588PHz7c3oU0otWrV0+d\nOtXeVTSin3/+eeLEiRKJxN6FNJasrKzjx49HR0fbu5BGxPlX6YYNG8aMGePi4mLvQsBm8GkC\nAAB4DUEIAAC8xp0gFIlEnB+scHV1tXcJjUulUtm7hMaFVykHKJVKfDPHMdz5jhAAAKAe8LkG\nAAB4DUEIAAC8hiAEAABec9ogzP3xfikTMueQpezxvhle4fMvlN+j8hrHY7z+13sP9WjhIZe7\nNes88uV1F4qIiPbN8GIYhhGIZGr/1r0ffOnHk/nlfsZ88aPecsZvzqFKR9Oti2YqGbCyqZ5M\nNap4jtU9QfO1jXMGt/FRyBTebQbN2XTNfNehHPMJ1mTfDC9G1O2Dy2VfxB+a7dd+UTxVeHFq\nzy27P6TdtD9SLNUfxgmU/U4ZRuTiGdrv6dUXDfau6N7N8GKY4OcP33rZZXw9RMIwA1Zm2LUo\naArOGoQ3162OHTmh+a+r9zrh/7dbcjc/1ffJXf5zfj2TlpN8eN2cZru/3Zpr3dRmXixr0eUk\nHP55TrvYt/sPXny+7H+n+eLS2f92GN+6quPJHtvGsizLsmfeDvWdddC6vG92Uz2dKlX3HKt6\ngqbt7zyxNWzpiSxN1omPQ/+a8s52U8WDOeITrJV/UO6Kd37Lq3qj5sj8YaN+ivh+/3fjg5z1\n/+JtbebFsiyrzz73bacj09/eWGLveuohol+L7V9v0xIRUeL33xVH9ZHauSJoEk76ny9xzY83\nHnntm6e7b139V7G9i6mvK98t2Nhy0e8rJvdo5q5wDWo/7IW1n0/2LLeDSOnbMvKh9//4ctzN\nT5btNBOR+eKSWTETVs1uwdir6HtU83O86wkKA4IDhEQMMcQQCQOCA0X2rN1GAp5ZOOn4ux+e\nNN29gc3aO2fAxL1Dfv33s2HezvILrQuGiFw9POx8z8V6kY2aMfbYNxvyiCzHvlnnO/3J5tb1\nlpvb3xjRzk+l8m496PmNidZP35smMMPfWPZQpwC1OjBq1paUanYDJ+CUQcieXv0/3ZSnenlM\neHLI3tW/5df+E47IfOb0+bCBA4Nq3VF1331dCuLiUskUu2TWwQmrZrZ0mnfNuj3HW0+Q6f7e\n2sfOTGmmlCqbP3nu8Z/e69Y0VTYyaf935oWtefOH1IqrU798aNzhMX/unt9HbZ+6bO7ygg4M\nw4iUQSO2tV8xb4RT3idP2HvG4xnf/u+6/u+vt/SYMansnrAZa6Y9uiP80yOpaWe/jjo+7bHP\nr5XtfvBQ/nPbE9Ivfdd55wufHa92N3B0zhiElgM//uT25JPtiRSjn4g+ufrn1Np/xgHdQwcn\nyxKRKfajWQfGf/VcqNPEYN2fo3W3ou0zH/y58w+XNVpN/Hcd142fud1pz/Yr8nniwxmpixbE\nlJZf6dl7ZNfkDV/FpHOmkdc6NGrWph+emffKQ0sT7F1P/bSe9oxy9Sdvf/3f8GdHld36wHxk\n774esxeMaOGqChy04LVhp/fsL7JuGTjjrQFBSpeAkaN751xNOFDdbuDonDAI9btW/5Jy6s1W\nDMMw0hFrsg+vXuOU/+VEXbt1TNy3r/YUL9y3/4xb+/am7ZsP7n6hpYBhmA4fXM5c0Y+J/p+j\nB0XdnmPZEwy6tH+faOTT41u7yl3bTJg2Urj/wKWmKbPRCTu9+kH/rW9+Uf6FKuv62o5tTyY8\nPXD6Nu5kIREJ5H6RT42POOa0KeD/2Izwnz69PnlGH+GtVRU+z7EsMbc+jMpkMuuCUCgwm03V\n7QaOzvmCsOjPHzeHf5bCljHtfz7jx9Vn7V1VfbSa9u74K+9MeGnDyeR8bWHqhd2fP/HCL7nl\ndjCVZCWd+O29B2dvCXn1xftbv3Hy1pOOfbuN74sH2W1PKu1WfB3V/BwrPkFBaERE3o4ftyQW\nlRZe2bx6R0H79qH2rN2mVOPef4357P3tFU4KFZFv7tg65fLTA6dvz+BMFlp0WafWbDkdEB7u\nrLfLU03YVMjGzou4vULYe+iA4yvn/Z1UVJz2//bu4BWCMIzjuG1TDg52OKzdclgXB5oiGrGK\n5rS1F+vC2LakVQ4uTsbF5uJA68BtIqyLPYqUlDb5CxQOapKz5G5zWHugIWVqZjzfz23e3t7e\n5/D2631neueysHHeq4841fbLbvChwAXh89HuSf/0ZLz+HE7mjLf9nUq1/oqiRrdeGpxafKUt\ns3dljT2uZ9R2pWPQKNp6Pv3xIcl9oScUalI6taniTfdq5cJUwz+P5Vff1ehUYGt2+yD1sDQU\nU+LDpp0ubRn/6bd9ifm1iar9+qW1ecA8Pc7ezYzOBT8La8utMdI1Xo4sl1eSXs/HPdGcdZi6\nXdRiUTV/3WeVFhJ/6Qb/4a5RAIBogdsRAgDgJoIwsJ42tc9XrLTMnnk9J7dJqBGA1zgaBQCI\nxo4QACAaQQgAEI0gBACIRhACAEQjCAEAohGEAADR3gEnYFFiXpjPvgAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 240,
       "width": 300
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Plot with only snp_based\n",
    "# Create the plot\n",
    "options(repr.plot.width = 5, repr.plot.height = 4)\n",
    "or_plot = ggplot(odds_ratio_plot[odds_ratio_plot$type == 'snp_based',], aes(x = cell_type, y = estimate, color = cell_type, linetype = 'dashed' )) +\n",
    "  geom_point(size = 1, position = position_dodge(width = 1)) +  cell_type_color +   geom_text(aes(y = conf_upper +0.5, label =paste0('OR: ', round(estimate, 2), \"\\n\", ' (p = ', signif(p_value, digits = 2), ')')  ), size = 2, position = position_dodge(width = 1))  +                           # Plot points\n",
    "  geom_errorbar(aes(ymin = conf_lower, \n",
    "                    ymax = conf_upper), position = position_dodge(width = 1), \n",
    "                width = 0.1, size = 0.5, linetype = 'dashed') +  plot_config  +  geom_hline(yintercept=1) + ylim(0,max(odds_ratio_plot$conf_upper)+0.5) + plot_config + theme(axis.title.y = element_blank(), axis.title.x = element_blank())\n",
    "\n",
    "or_plot\n",
    "\n",
    "pdf(paste0('figures/G2/G2_OR_Plot_snp', use_gwas, remove_HLA, only_dt, only_esnps, coloc_thres,use_gwas , cs_snps,  '.pdf'), width =3, height =2)\n",
    "or_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1be8eca-a255-4b53-894a-fe5524f2f121",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4088c4b-e580-420c-9b45-c08380f12590",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "id": "7774c80b-2ff4-4883-98d3-985aeffe6760",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plot with only ALL for eGene based approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "id": "a457d6ad-af23-4a67-849b-92b1136a4936",
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_ratio_plot = odds_ratio_plot[( odds_ratio_plot$type == 'snp_based') | ((odds_ratio_plot$type == 'eGene_based') & (odds_ratio_plot$cell_type == 'ALL')),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "id": "291e457b-346d-4ae9-9466-0ebcb83a55a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 7 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th><th scope=col>excluded_genes_without_cs</th><th scope=col>trait_filter</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>6</th><td>0.0006863434</td><td>1.939638</td><td>2.784903</td><td>1.3238773</td><td>ALL  </td><td>eGene_based</td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>0.0187006300</td><td>1.531563</td><td>2.180496</td><td>1.0527763</td><td>ALL  </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>0.1422117000</td><td>1.598469</td><td>3.021836</td><td>0.7877197</td><td>B    </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>0.0037729260</td><td>1.862487</td><td>2.812239</td><td>1.2007029</td><td>CD4_T</td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>0.0496152300</td><td>1.632734</td><td>2.655485</td><td>0.9681380</td><td>CD8_T</td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>0.8767768000</td><td>1.025278</td><td>1.905349</td><td>0.5121500</td><td>Mono </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>0.4340997000</td><td>1.254892</td><td>2.250417</td><td>0.6577192</td><td>NK   </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 7 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type & excluded\\_genes\\_without\\_cs & trait\\_filter\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr> & <lgl> & <chr>\\\\\n",
       "\\hline\n",
       "\t6 & 0.0006863434 & 1.939638 & 2.784903 & 1.3238773 & ALL   & eGene\\_based & FALSE & disease\\\\\n",
       "\t7 & 0.0187006300 & 1.531563 & 2.180496 & 1.0527763 & ALL   & snp\\_based   & FALSE & disease\\\\\n",
       "\t8 & 0.1422117000 & 1.598469 & 3.021836 & 0.7877197 & B     & snp\\_based   & FALSE & disease\\\\\n",
       "\t9 & 0.0037729260 & 1.862487 & 2.812239 & 1.2007029 & CD4\\_T & snp\\_based   & FALSE & disease\\\\\n",
       "\t10 & 0.0496152300 & 1.632734 & 2.655485 & 0.9681380 & CD8\\_T & snp\\_based   & FALSE & disease\\\\\n",
       "\t11 & 0.8767768000 & 1.025278 & 1.905349 & 0.5121500 & Mono  & snp\\_based   & FALSE & disease\\\\\n",
       "\t12 & 0.4340997000 & 1.254892 & 2.250417 & 0.6577192 & NK    & snp\\_based   & FALSE & disease\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 7 × 8\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | conf_upper &lt;dbl&gt; | conf_lower &lt;dbl&gt; | cell_type &lt;fct&gt; | type &lt;chr&gt; | excluded_genes_without_cs &lt;lgl&gt; | trait_filter &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 6 | 0.0006863434 | 1.939638 | 2.784903 | 1.3238773 | ALL   | eGene_based | FALSE | disease |\n",
       "| 7 | 0.0187006300 | 1.531563 | 2.180496 | 1.0527763 | ALL   | snp_based   | FALSE | disease |\n",
       "| 8 | 0.1422117000 | 1.598469 | 3.021836 | 0.7877197 | B     | snp_based   | FALSE | disease |\n",
       "| 9 | 0.0037729260 | 1.862487 | 2.812239 | 1.2007029 | CD4_T | snp_based   | FALSE | disease |\n",
       "| 10 | 0.0496152300 | 1.632734 | 2.655485 | 0.9681380 | CD8_T | snp_based   | FALSE | disease |\n",
       "| 11 | 0.8767768000 | 1.025278 | 1.905349 | 0.5121500 | Mono  | snp_based   | FALSE | disease |\n",
       "| 12 | 0.4340997000 | 1.254892 | 2.250417 | 0.6577192 | NK    | snp_based   | FALSE | disease |\n",
       "\n"
      ],
      "text/plain": [
       "   p_value      estimate conf_upper conf_lower cell_type type       \n",
       "6  0.0006863434 1.939638 2.784903   1.3238773  ALL       eGene_based\n",
       "7  0.0187006300 1.531563 2.180496   1.0527763  ALL       snp_based  \n",
       "8  0.1422117000 1.598469 3.021836   0.7877197  B         snp_based  \n",
       "9  0.0037729260 1.862487 2.812239   1.2007029  CD4_T     snp_based  \n",
       "10 0.0496152300 1.632734 2.655485   0.9681380  CD8_T     snp_based  \n",
       "11 0.8767768000 1.025278 1.905349   0.5121500  Mono      snp_based  \n",
       "12 0.4340997000 1.254892 2.250417   0.6577192  NK        snp_based  \n",
       "   excluded_genes_without_cs trait_filter\n",
       "6  FALSE                     disease     \n",
       "7  FALSE                     disease     \n",
       "8  FALSE                     disease     \n",
       "9  FALSE                     disease     \n",
       "10 FALSE                     disease     \n",
       "11 FALSE                     disease     \n",
       "12 FALSE                     disease     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "odds_ratio_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "id": "3d2ae354-5985-4efa-a1ab-c8f977352fbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdeVwU5R8H8O/sveyyy3KfcoooCiiIeN/mgYa3llmpaZal/jrssNS0stIyMysr\nK0szJe8r7/tGRERFEFCR+1hg72t+fyzicsPA7g7r9/3ij2Hm2dlngfky+5lnnyFIkgSEEEL0\nw7B2BxBCCNUNCzRCCNEUFmiEEKIpLNAIIURTWKARQoimsEAjhBBNYYFGCCGawgKNEEI01QoF\n+uDBg5mZmS3fD0IIIVOtUKDlcrlWq235fhBCCJnCiAMhhGgKCzRCCNEUFmiEEKIpLNAIIURT\nWKARQoimsEAjhBBNYYFGCCGawgKNEEI0xbJ2BxAyF51C9ce18oMFWimDGewhfLOrMIQDQCoW\nbS05A8BgMtwd+a/GSIYI63y0evXOwn9VAGLR3yNFvrU26xWqX66WHSjQFmoZcQM933U396tB\nTyMs0MhGGbQ/nyy54e24ujvXxaA9faN0wTniz4ECFwBg2H012bGnXn85sXDJLdWQaF5dj+e+\nNdb7rYqK6WfqvGmnbuu54jMSx+/G8NpxCLO+DvQ0w4gD2agC2Q6G/UdhPF8uYcfnDO/uMEAq\nO1BRsxWPZSyvul8P5K4vafLOyxUHy/hzuvKxOiOzwjNoZJvKZXqNiOdZ9T2D5SvUP5ADCAEM\ninf+VgAAWyz8digXAABYM0d6NGPvCn2BPZF8NndpPilxsnsjxqG/oHW7jxAAnkEjWyUSMNnl\nupyq7w26+zKmu7GMMuy+mup97FmXFzjKf3MNVPbOZYrLdOIwt0Pj3T5y1nyarKK0F4QagQUa\n2Sg3wTh9xfIbqvtqUqHUHLoiPekgHGH/ZDvfjjujuyDzRsXtOkPmhjnwepqcMhOYcyDzwAKN\nbBSD88pAx25l5f/blzPmUMkeEKzpLXCt0UQsnCFR/JShryuD1v9zNLvXvrL0svKpf2e/mqoH\nUH8eXxCvAgAAgjMrmnfjQv7wf/M/K+Z+3IWHBxIyB4IkKZw/VBMfHx8WFhYcHNwqHUIIIWSE\n//gRQoimsEAjhBBNYYFGCCGawgKNEEI0hQUaIYRoCgs0shVa1fLj5RktHZTUGvSqT4+VpdGh\nJ6iNwwKNbER2WrnUVxjQGp8ZKc4rW7AvZ8i/eW8mqoprblSv3pnd6+/sXgfK75uu1ms2HH7U\n76isGACYvKmu6o0P8NOFqKWwQCPboPsvi+jr3Rp/z3rVuguqoO5uu0ZJOhaUrsuucSbMfWus\n9/lYcVC1leTVpAqGn0Dy+PsAH979TKWsFXqDnmo4WRKyCVr1TS17lnHiI5Xs1eOaCIl2/yM9\n11Hwbi9x9flE1Z/HF+7VVnv0M329l3g//qZEddle8KcbUwjMaUHsybka8OY2/OTSbOleoXiJ\no3zvg8erxOzAYtUtEES3/KWhpxgWaGQTNIYyNuPJzPvlWk03523dIeVm8fJk9a7uXJPkg/v+\nBO/3G9iV2lDGYxjPhe15DLlab2jwnSapkH+fw30zmsUoMllLMISglzb8SIQagwUa2QQOQ6zV\nPokUhLyR7kw+QJQ/T3BFKwWu5EnTxs6guQyxylAK4AhQoTLYcZkN19h792T772n336ucyGP0\nf+S/z9h7kAYZMB2wOqOWwQKNbAKb25mtSFdDqDGNkKkO5gtfcYZbWSqFWOxQrWljZ9COvOjy\nsi35/JfF+r/SNT1COQ0/c1AXt/NdAACgqOzZ68yNQ4ROAFCmvefE6dSCF4QQ4BswZCtYw/zI\n09mPB06I2OyMook78z8t5n7Qhdu8kR1M3us9eXev5I/ZX5Li6jjPi6g2j10ds9zVITNb1c6P\nX/fNDhFqMpzNDtkKrWr5Gc1zA0WBatmrJwyLR4i8G3+MeehVn59Ujx8kDsZ5olHLYMSBbAWb\n99GgOm//anFM3vuD6dET1MZhgUY2hyf8cYS1+4BQa3jaC7ROcXVjwsZ9BfdLGS4hnmMXdh3T\niUMAeeKtv1ecBGAw7T0c+8/ruWCYsL43qze/2DF/mwpAPP3fUS/6VdukSrnzxYpbV7J0PF/3\niR/2nNyFDXrFlR+vbNhXcL9Ayxs/aM8H7uZ+fQihNuzpLtCGtB9OfJbk/c7a6AhXw8MTSStf\nP8v8Z1CsKwAwBq+Z8kFvffHFa299eCthWHRUPbvovGjcsUUVf089ram5RX3k22TV+Ge2jeGX\n7rsw75t7gzaG6DafXXZK8u6Pz/by5TzdP3mEUBM83aM48vduZ0xYFt7Tj8u34wePip43SLpr\nX0WNRiSPacwTH/y0/7nvSmrvpR5MR2cOQOWZt52rnR2UH99X1vP1bv2wOiOEmqIVKsXu3buv\nXr3as2fPYcOG8fl8g8HwySefAACbzfbw8Bg7dqxEIqnvsatWrZLJZK6urq+99lqNTUeOHElM\nTGQwGN27d+/fv3/Vep1O99tvvzGZzBkzZrSw5+XyPI042qvqe4avvzD/vhxACGA4tmDLMQBg\ni8evH9YZAADazRm1pRl7Z8W80unAnH2jPycJR483NvgIIL8g356TdOa5D/MVjk6DF/acN1CA\nl/kRQvVqhTPo4cOHT548GQDi4+ONa1gs1tKlS9999103N7ezZ8828Ni33377zTffrL0+IyMj\nPT197ty5r7/++oMHDzIzM6s2HT16NCwsrOXdBgCRwI1d9uBR1feG+5kyNw8BAAAwBq957uiZ\nuK9f4pzdlltGZe/Sv96647xkzP4Lk/7+xH7PO8npwBU7lOc6hP94fMJvnzjfWpp8Cac7Qwg1\noBUKNJfLFQgEsbGxeXl5xcU1Z2dks9nGhfXr1+fk5DRxnzVGZ6enpxsXbt++LZFIvLy86npQ\n87nFTtRvX5J0MUutUijv7r+87rhD3Cj7qs2EnV347O4jMpK2pFAYLK6RGcMSAoAEQ5lGDg7R\nve2ebGcAnj4jhBrSamEok8l0dHSUSqUSiUSn0y1duhQAXF1dq4KI2iFGAwICAvz9/devX89m\nswMDA+VyOQCUlZWlp6ePHj06Ozu7dTrN6DB34Pu/XPvtjb3LpQyXEK+47/uMcqvRxGH8bMlL\n6+9N+j5I/tP+9zQ9t7zhaLq9ZMuRiasLAQDGb9kU3m3rxpCHy7f/ETz628mukz/0+mL1oXG5\neq6H87CPYsKAIF7tEbXs/Iyf5Tpn52c+iYl+uq8AIIQa0WqfJAwMDPz6669nzJghkUg+++yz\nxYsXl5WV/fvvvz169AgNDW3g4SUlJVu3bm2gfO/evVskEg0cOPDEiROnTp2qWu/p6Tl79uwW\ndh4hhGirFc6g1Wq1XC7ft2+fu7u7k5OTwVCZrIrF4tjY2H/++adjx44MBpWTRYPBkJ6efuvW\nLWMhHjhw4MCBAwEgOzv78OHDLb9IiBBCdNYKb7IPHjy4detWAJgwYUKNTa6uru7u7omJiVBP\nBr1x48a1a9cWFBQsXbr04sWLALBy5UqZTAYAq1atWrFixdGjR8ePH+/k5NTyfiKEUNuCkyUh\nhBBN4WUqhBCiKSzQCCFEU09HgdZeXHLst3stzXIsSHdl2bFfU9tQhxFCZvBUFOiHaZtL/cYG\ntsLnQsji3F9f2zuub/wLcxMvFTW+tc726nuZa2ftfrZP/Auzr57KIQFAlXJn2dQdsT23TZhy\n+p9kLQCr+zTXhJ/vy2o+A0LoafI0FOgHB7LY/b0dGm/YKP2lNRcuBEf/fDB2Uaf8r9ZkKxvZ\nWmd7MvF4YfD8oX8fHbUotHDVTzn6ynnvhmw7NfbbSbD1m3uFABDo3T0r83TNiZsQQk8TG59W\nTavVbt788w2HwLnGe4mqds04ltrNMWNPdh7PceQHvefEVLvxxbXl29/ZVe1+z4wR/Y6sqLp1\nUnHCRfuR/7g5CcHpxaDAsbm3wbvbk7a1t3Lqak/EvBINAKRGxWIxRA5cRh3z3gGAQ/ug4gsp\nMDLGPD8ZhBD9WaJA37t378CBAwqFomvXrkOHDiUIouGtdbavvbK8vHzfvn33799Xq9XTp08P\nCAioPQEem8328xHskyqe3L6zPE0T+fnu7ozk5MVLkq8d6N7NpDfdPpp47KMGXom6TMrraJya\nT8STKFRlBtP3ILW31tv+3Pt/f3yYZHn5LfrZiQCi1rx3AEAI7aFMang63uQghOpi9gKt1Wp3\n7NgxYcIEV1fXzZs3p6amhoSENLA1MDCwdvs6d7Jt2zYPD4+FCxfyeDwwmQCPxWLFx8dnZmb6\n+/sDQEiHAMPps0/SXGHP0e4ufIDogF6CK1ml0M1kao3GzqC5YgeVtBTAEaBcVcrniqsVz9pb\n623f+/OpR5eps/Zcfn9JetiPLvuN895151YkXH/vneTQ+LAgIGVyFavdD5/mpubqmRJO2CCH\nSQM4AgIMV4pmbVACAEfA8o0Sz3rezqXOcD29dMEXsnIAr9Huy8ewa28vvVm26V95ao5exRW+\nvVbSSaM9v71052WNSsCJGef4XBQLp3JCyNpaoUDv2rXr1KlTph/2mzJlSlUVzsnJcXJyMtbK\nyMjI9PR00wJdeyufz6/dvnYzZ2fnoqKiF198sWq2vNoT4Bnbu3n1EGq331VDZ2PKIbu0L2/M\nXGdGSuYFhXhW9ZmqGzuDdozsUf7rpvz+r4hz/0hP7xka0shWR00d7Yt3ri4Jes4vREIaSFKr\n0unqmPcOAEqSghMKJX1nO78YwtLnKf7dWPgj0/2tfkwAYPVw3jCLr5Gqtq0uPnDL7sU6pzoJ\nkqz5WVJwMO87bV1bCyvW/6hs97LrqgiWHRMAwHBHme7n8MF4NiO77Jtvy5O6OUbgqTtCVtYK\nBTouLq6BTxIqFAqBwDjFMgiFQoVC0fDWOtvXXlleXu7o6Lh9+/bMzExPT8+xY8fWOQEeABCE\nbzuF/sSD4nHtnQAARIHsjMVjzuRyHUd93Ltb804TWT3m97zx8eVZw9WCTgFvr/A2Th56rXL6\nOl6trURd7SW9Ix9+/druG0WEc0fvmYuD3YFZa947gHsXbvTlD3cZF84CAPAVTJ+ifvsfeV4/\nkatJf0iS4Bj/7eSVfbQGZq4U+zXtheRdlueEi9+JZHEer2GEiaYDAEmWMwmmgGGP588IWZ/Z\nz6Dt7OyqaqVMJrOzszN9bO2tdbavs1lhYeGYMWOcnJxOnTp18uTJuLi44cOHDx8+HAB2794t\nFouN7UmStK8IYD7Ykx70chAAMLye7bVoHsXXSrh4zv7Bs8YUet0+mtitnq11tWe4DohYOSDC\ndJXzgG5fDTC53Ai6K/8+sp/q5/fk18PyYLuU6IoBXAF0l4pmXAIAwmuIywdBAADgLl6+shmv\no6RU78ZWr3+v+I6S6Rfj8MoUvhMBkFg8a73CwGLFvOwagAUaIesz+xm0p6dncXFxZmamq6tr\nQkJC3759G95aZ/vaK11dXY0l2HgJserCY40J8ACgoKDAyzNi2uBpLX+llsPq/m7v29nxuTro\nWvkb0udqCx1Zxn+CrB7OG2byStLKf9oouzmK213YwJ7qJhQyctKYLy32cjdo9qwt2pXsNTMM\noKvTLxscK3Lkf6wtORnsMrA1BiYihFrA7BcJ2Wx2XFzc3r17lUplRERE1Zn1ypUr582bJxQK\na2wlCKJ2+9o7IQhizJgxO3bsKC0t9fHxiYuLA4BVq1YpFApnZ2fTCfBu37795BZZvLiNI8z9\niltJSD+h9tuSHd6SYR1Y+jzFv1uVPoPd3QEqJ3MlCMdg8fTeeesOaCIncZobF3t35jmmATwe\n20cQkHlcmukj7OnPJA1Aag0afWu+FoQQJTY+m51Wq928efP06dOpTUhtXWTxzfK/dylS8/Qs\nCSdsoMOkgZWjOF5NEmyYxQcAUCp++Fje6UOX/qpaGXRZxedvS9Mqv+FO/dJ1qATgdsnrezif\nLRKKwZB2oGTTMVWhnhnUy2HmRL6oVLljk/T0XT1DwokYLnm+L5tTR5cQQpZk4wUaIYTarrZ3\nXokQQk8JLNAIIURTWKARQoimsEAjhBBNYYFGCCGawgJNS6obJV/u0tLhjiqamyVf7tQarN0N\nhJ5KNj4fdBulO35AH/M6uzVuAaNP2VXy12m1jM/tO9VxYhcm0cCm1NI3V8mM8/5xe7n88DIP\nADid7TvsLrv8jHOMXT1PgRAyFyzQNJQnv8DiL7JvhT1pkqU/J7HnLHXyKqxY84M08TOnbtz6\nNwEwujtvnM2vvg92t1Dd9gRDTF98t4WQhWGBph9VhkbpIxYClB0vWHWf6/tQnlhM+PWVzJnA\nE5m2u13y+tfyanfdYtjN+8nJdNKlrFtqt75uHcUMENv398m9mQHdOta/iQGGpJLXXwOuO2/o\nc5IRQZUV2duXnZmkgb7Vbj6DEDI/LND0I5cZ+PzKKCL3PjnlTY8XGOrd60p23fac3tGkXUfH\n7392rHMPVWQyvSiACQAADLGIrJA1uKm7ZO33ElJneJRQ+t2PZf5fSkIYAAAEnwEygw7/WBCy\nNDzm6EcgZChLKy8QuoQLQh0JAF6fcMamHD10ZD5p14QzaKGQWV6hB2ACGMrKCXth45sIFsO7\nh323HdKHZRAiAQAglQYQMvAvBSGLw8OOfngBHP4FjQw4AFCYLL81QBzE0Jy/YfAcx6zWrgln\n0H6duHk7K25HibwKK0494MX6N76J1BlyEiqu6dgvVU6oDdkPtP6BOHUSQpbX8IUf7cPDn06M\n9BYJHP16TFlzocxCnXrauQt6apXXKgAAPHyIc9/lLvyoJKODY1zHxh5YC6eLw6zO2j+W5ny4\nUd3hRYeuPIDbJa9/ISura1Pa1pwZrzyc9UbuN0dg4Fxxx8q/De21m6xekXiFECHLa3g2u/Nf\nvXK645svDwyC299NG/hbzMXby2vf/w5ns2t9qhslazPsZ4tKv1E6Lhtlzbc5mpSSNanCt8c1\ne8pphFCLNXzs93rn514AYFAXsTksO3d3cYOtUavhhTm+GwZlx63dDwBOqOO7dd6UFiFkdo2e\nnGWvivF55xIwPUesP/aqtyW6hB4TD3JdZu0+2LoreWoA6O7ObWAZIWtp9I2r99sXSa3s4an5\n+sXPLr6GH/lFtmX+icL5JwobXkbIWpqULLIE3r0XvNQz7cLlEnP3ByHrSCpUGxeOP1BqDXSY\nBgWhhgt02i8LluxMzpUppRnHv/jljH+P7s6W6hdClvVHSoVxYfG5Yjkt5qlCqOEMOjBugmjB\nzB4vpZQL/XuMXbtnaeTjLUqlUqvVGpfVarV5+4iQ2UzrVDnnSbgLN1euA4BABzaHQVStR8iK\nKN409vLly9nZ2cbl5OTkqVOn4jA7hBBqXRSH2EZHR0dHRxuXDQa8dIjavKRCdbGy8i+5rzeP\nzWiFyV4RaiH8qDdCAAB/pFRczFUZlw+M83TgYoFG1ocFGiEAgHAXrt3jWyRw8PQZ0QMWaIQA\nAF4MxauCiHawQCMEgBk0oiUs0AgBYAaNaAkLNEIAmEEjWsICjRAAZtCIlrBAIwSAGTSiJSzQ\nCAFgBo1oCQs0QgCYQSNawgKNEABm0IiWsEAjBIAZNKIlLNAIAWAGjWgJCzRCAJhBI1rCAo0Q\nAGbQiJawQCMEgBk0oiUs0IgKQ1nhjb2p6RkyFZPn1MG/+yhfZz6AIefYB4kPAAg2W+jlETmp\ns79jnWWu5OKKC7dlAG7txy0MFtfctTrnZMqVC4UVJM8tulOfYS58KD7/ycVUBQAAcLyHfRLu\nZYZXhBk0oiEs0M0gVZflyvMduGIPgVt9y9buo0XoyxJ/vZ7fOWzoWCc7g/zBf9cPbyHiZraz\nAwCW15AVEd5aVc6BiydPFvmPc6nr8Y4xi0fFFN/bvUlfx8bcRxl6z75vhgnVBed/TroRPKSH\nHwDDc/DKru3M+Zowg0Y01OBdvVF1J7PPTTs095ebfzWw/FTIeHCb5d93mJtYwGLbiwPHhvrm\nZqUXVW9DAovNBAAAWeI3x69mN3nnXgF9hro72rM4zhInexaHb9xb/uklB//67MyJUyWaVnsZ\n1bwYar+it5Pxq6pSI2RdeAZNhVqvvl1yFwByZLlJRSnW7o6lqUsVeleXJ9fUmEIHR2VZKYAj\ngO7R0fceAQDDzf+ZuY4AACDsunBQ859EV3j8xsOALkPdAMCp10fDe5EGVX7u5U3XErwH9wxs\n/QKKGTSiISsUaK1Wq9Ppqr5lMpkcDsfy3WgJqbo8Pm0vAFzOT7S7ZdfXK8baPbIorgOfWSCr\nAKis0XqZtIRvLwEAY8QR7l5WcnNL0p20IPculH6zpPrR/sRkVvvBI53YVSsJBs/dKyQ09Uqu\nGgJ5LX8VNWAGjWjICgX6xIkT58+fr/o2IiIiLi7O8t2gwE/UblxQbIRLZx6TG+UWYcydw11C\nq9Zbu4OWEugbor185rCodx8nO73swX8p9z38n3UGqDwBJdhip4hxPrv/vFcU2tG5uSmaTn53\nW2KWV5fB/cVs0/WkQVWQeyfF4Die20ovoxrMoBENYcTRDBEunauq8E+DV9fYZI0eWQlT3G1m\nRNLe1MOrZGomz6mD37Dn2gmqNyHc/CM8T1276j8sWpf4zWX9xEFR3lUb1Sk/HL18HwBgx3tp\nrrGDR/WRnVt61/GtXh3tAbIyrtwo09w4+9dBAIB2k0cMltzZ+mOmkmBwRPZevbtFmiHfABwH\njWiJIEmyhbuIj48PCwsLDg5uYvvDhw+30TPoKmq9+syji8ZlFzvncOdQ6/YHtRxm0IiGrHAG\n7ejoyGaztVotADAYDBeXypFY+vJ7Zw4dS8oqUjBF7kHRzwzr7sknwHBz6yfxdwAYbL7YI3Tw\n2FGdJfUdOQ8OrNp4WQbgOmDeawOcq23K2vfl71crB9JGTPsgLggAQFuQdHj/ieRctZ1nl8Fx\nI0IdmnxIStXli85+Ylwe4N17db9PmvcjQPSDGTSiISsU6KioqPPnz5eUlAAAh8Pp3bs3AIA+\n98Sf/z7s+Ozzsf72+qI7x3f+Gc+c+0KkCABYYc8tHtdeW3Hv8B//ns3oPDqwnh23G/n20pEl\nZ3/cqqtjI6PzlKUTQkzXGDJvP/QcOmOIG5F/csvmE3c7ju3Q1LiUx+QOadffuBzq1KGJj0J0\nhhk0oiHaZNBZV6+wer46qIMEAMAzPHbE/a//SyqO7CsxbUSy2cbrRkUn1/+li1swxLPJ+zek\n7vx8BXCcA2NGju7djg8AjOD+sQBA6uRMBoNvZ9eMY1LMFX3R5+Omt0f0hxk0oiG6FGilVKpz\nDnpSjZkuLo7SIimABEB3Y8vSGwDAdI2ZPsP4aTLnAa8taM7u/WLf/TCW1CkKbh3auu245/9G\n+RtPlu/Ef7L1poEpCYt72bsZBRozaNuDGTSiIboUaL5YzCoqKgWorNH6wsISB4kDABgjjrF+\nZffP/bvr8r1+vqF2FJ+DYNm5hXUPOXYkXwb+IgAACJnw8ZI4ecH1fVt2XvV9sbuoiXvCDNr2\nYAaNaIg2H/X2j4rSXth1/G6RQqOuyEnad/C2e3S4U9VmgiP2GxDbtfD4mUcGqk9B6hQFN67c\n0bu5CQEAHl06dOW+VKUnSZIkdVptM/ZrzKCNX+EuT9Pps7rgzM+ppS0d+NNKtIVnN6SWUP57\nqC7chTuoHd/4hRk0ogm6nEED03PQC2NPHzrx13dFSqbIvX30CxO61TihJVxj+ruvO57Y64VI\ndV0ZtOzixlWHHgAArFt6st3w/82IKdqz8qT7vBnRJYdWbbwoI1h8kWtgj0mDjPmGe4jfrb1/\nfnu/giHy7DB4dHeHpnf2qc2gyy+mqyKi6h1H0xzKtNTTux8UKljOUaH9RrjWcQlAV3btp4vJ\njOBJc/35oJdeu33+SE6xmu0cEdJ3tIeQAGC7hAbcTbwRMCiCXccTNBNm0IiGrDAOGgDWrl1r\nHMXB4/Hee++9FnbA8p7WDFqW+M1Nu9kxHQSNN22EtuD0F3f4U3uEu8mTf7smHzSgX2iNcwVd\n7r6kPAn37g3BmLn+fEPB9d1qr0EeEkb59d+uKIcO69uRAADITd150G7UDJ+WzxWAGTRqxPXF\nEWtirv8e2+IdnZ0XFb/u6pqmNKXNGXSb8pRm0GppoUrU1VidK7L2/1Lm7lWelqJkefv0nNrR\nS2jatOjc0kt3VaZriMAXRvar+keWXfjI2ScukMsBbpceon/vSiG02sh1Vcqdu44h/bwf3L0B\nAAAM14ixAEDqZQyCyeYKHldPd7Eku6AQfFo+QzRm0DapLOnXdxeu3pfwUOvebezb3656JaJt\nvVHCAk3FUzoOWqFR81hPzlULy/Sx3SfEEYVHr54+UjR5rGmFde69dFTvhncllBhnEuUIOVq5\nhgR4UhHLsq+mOkaPExAPTB+Td/KDhEwDw75r+Aifx+sINgc0aj0As4WvDcdBm0tCQkJCQkIL\ndxIZGRkZGdnsh5XtmRv7rfirP6+MDOEWJO78asvBSRGTat4hgtawQFPxlGbQdhyuSvVkOmZH\nt/bt+SwAjyg3zk6ZCpxNpphr7AzajsOVqZUAfACNTMMScEwrYunle2mXZWmXE43fbv3OMPGN\nQCG4D/hsZF+5LG3HlTOXHIfH8AAASK0GONwWV2fADNps9u7du2zZshbuZMmSJU0o0Pn/vT/9\njQ3nCoTdZvzwz9cjPVT7fz82/KtHUyJZACDqM+unPjXbeH4f8VHF0LIdPycLhn+7f+v0wNo7\nqfYMhQcX9Jzze5p42KfbN88JZeet6+Pxxjmm0DNy6nfxG+JUv08d+86e20UqYvJO3bf8Gvsx\n3PvzpQn/25PbbvIkX7KpozOwQFNBwwxam5unLy1le7gzJZL6llv6HFwHF152iQJcjAMdSwrS\n0327+hJFCfla15DqE4A2dgbt7eJVmHrznmeEmyL5UpnXoGpXaCVD+788FAAAHtz+Zz9vzFx/\nfnbW5YeiTl0deCQJpF6nfXzhJK+81FtS511bmgsz6LZOc+jDd7Om7cvY41m8/eUJn58fudbt\n4UOfwEAWABgrKbjN/TD2nGmbjSJZsrTXydtLtL/Ejt+YPP3TDrV20svkKSYuEvAAACAASURB\nVGRJxTGn7ywn984cuuifyfumuc87S84DvTxr2+zJ35yPbvdLxqTjJYu78BmgOTSre/X9dN39\n8VLZvPMPJnOOLxh4QFjfq6iBNsPs2hRjBm382nTrH2t3BwBAduRo4aqvlUk3GlhuMWFgN8OD\nlMfn0C4iRsKV+BUnz2Q79R7q3OADa2G7Rk1yKd5x+p+vrxcGhHXvZDxRKDq39Pztirrae7i5\nS9P/+/Lw1m+uZQpCescY0xGQphSKurq3ymzif6RULD5XbPySa2kykBA1w6Pk5ORtL3dytBMF\nvRCfmJpaAV4+Pg/v3dMBgPu8syS583le7TZ2/caP9RMK2/eJYhQU1LUTU6ze4ycH2NsHTp0Q\ndS/1PkDunjd7+0sEIv/ntlzOyHCbsnTU7YXDBo14cfmhKzdq7ifn7t3OY59rLxL6xk0a0OR3\na1Y+g9ZqdAUkuNLwZEWbvnvz/R7TB7vX9S+Mzhm0vlSqKygEAE3WfW1ubuvuXBQTxNuUWRrV\nQQIADEH7yeHNzwUr2XUIGf5OSPV1zr2XVi/07TpOngsAAEx+uxHR7UZUb64tTMlwihjUCmPs\nADNosxk9erSnZ9PnZKhbUwJoz9DQ8Ofe3PrtxGDHx4nZyJcGL3rnja3OH43q7Kh9mFlg8Ooe\nGi4yaXN98beMqiOcJOvciQnd+Z3bM0eMhH07EgJjfaH4n++ORW5IORgl3z237z86vfuQpX8P\nWaLOWD10xI5hvWvsRxkcnPLX1vS4SdyT8acqmvppOysXaJLJbI3qTMruHd95ICFbwfPqOmLs\n0Pb2RMObas97V3H3yK7/Eh9WMMT+3WPj+vvy2UE9/U6cSOk9uUsdN++gcwatycxU3boFAPLT\nZ1hOTo22bx6ua99XXFt5n5SxXXrPbpV4AwAzaLOheH2v+bgjP195fkZch9l3ixQG37euZK2K\nchjzw97CdxdOi5r9UOMSOnTmb18v6nxjiUmbXTWP7jp2YrJVGOZwbmqH2WniYZ9tn+wArKEv\nhH49MXiL24Dno9wBri8O6flpKoPnHDL8oy3vTB62vPp+nv1kyc4JPX3ebzdpcniTjyErj4Nm\nc3kfvt/icdDauzvWHBNOmNbPtfTs5m3l/d4cF8JpfFPJ2R+36ia8NsAZgMzY9+V/rMnTBnuz\nSq7Eby/p8+oofxbkHf/+qMPMad1qV2gaZtDKa4majAx+t25AENK//zHIZUyJRDhwABCEcT0n\nwN/afaQ1zKARDVn5DJowHgWyyxs35fh65CXekbI9u8WOHxb4JEPP2LNy07Vq4wEYYVM+Hmf6\n5jjn3j2nrnP97Xlg3zvS7bv0bAgJaHyTKZKEymFeBEBFeno++HuBm4fro7uPoFvt6U1pOA6a\n360rv1tX47LrB4tqbLJGj9oYHAeNaqq8tFjJac6Roh+HWLgL1i7QVUtFubpnnntzFCP71N+7\nTmYsjA14vClgzHtLxzS8F4VCKfA2fn6CLxRqFIong2ob2GTajYB+Y/x37lr/pYzt0z3QRS6X\nAwAQPB4oFHWNsaV5Bq1OTzcus7292B4eDbdHRphBo5qMgzSsytoZdNWSJDgiUMwBCIjowN1X\nqICAxx8nbsIZtJ0dXy6XAwgBlDIZx3Rq5wY2mSLEHYe/1HE4AOjTdq9OE4sBAEiVCuzs6hpj\nS/MMuviHn4zL4rFx7NGjrNuftgIzaERD1i7QVRW6NC0po/tAH+JRUqraZYjJNc4mnEF7BgYW\nHzufGdrftfRcQl5gX68mbarNoCpJP3XslkPEbFcAgIK8Aq+6P0NMwwy6ClMiseteeWGD7Ymn\nz02FGTSiISsXaL3hcYV2dmde3/LtP1KWZ7dnxwc07+Bgtx8ad3/n3h9WK7leEWPGhXABMirn\nsRPW3gR1zXsnPbBq42UFx965Xfj4yT2dCAAouJ3uFDatjjEctMygq3D8/ZzmzrF2L9oezKAR\nDVk7g9br80lwAwCGY9dxcYOp7sa+/dDp7YearAkY815APZsAQBgzY2lMtTWikW8vHWm6Qpt+\nMcu/fz9+nc+HGbTtwQwa0ZCVCzSbw3Kj57HADhrzUlB9GzGDtj2YQSMaosdcHMLoGa9auw/N\ngRm07cEMGtEQPQp0W4MZtO3BDBrREBZoKjCDtj2YQSMawgJNBWbQtgczaERDWKCpwAza9mAG\njWgICzQVmEHbHsygEQ1hgaYCM2jbgxk0oiEs0FRgBm17MINGNIQFmgrMoG0PZtCIhrBAU4EZ\ntO2xRAatOqy886G65DbJ8GY6v2HX4TUWmwByq+zwVA0AMB0Zool2XdZz+PXdKfSs4kRflQZA\nuETce2n1aRZJQ/GH8ls/67RiltdaQYeRDAAAhT7nHXn6Fr3OieXxuSBkIv7TaXOwQFOBGbTt\nMXsGTV5TJDyvc91o32kgg0zVpE2vSGKLo2YzAIDxvP3Qv9j6HO3dQbKMo5zQYfXsoo/dQNJO\nsbIsUVVzi/6A4sY+ZvgNofCeMmGCoiBd6CoE8oRGGmUXvZJJJCmvjVQWjRW44PHexuAvjArM\noG2P2TPokh/VjIXi9qMZAACR3E5rdacWquWz+aa3DyVJgmn8/o7y3HCyS5adqGk7Lz+sE8wU\nOXoQ4MH3CZcWXQLXwUCM4ncCAANoOAThBOy6pjZH9IYFmgrMoG2PuTNoUplFCkY8CS8YHZl2\n9/UqADsAw+aK/zYDAAgXiHr0AQCAEH7vrGbsXVNk4MQYd05w3EhN4eMNu2SHx2pIDsPjd5EY\nA462Bws0FZhB2x5zZ9AEz5eQ3zbA2MoaTd7WK3wZxunGGc/bD93EUp1RJb2oKvpQ6O7c7L1z\nnBmaAgMAA4DU5BOcqj3ECYfpQZOivjVKnt3Pvu77TyCA64sj1sRc/z22ZXs5Oy8qft3VNa3T\nJQDAAk0NZtC2x+wZtOMcrn64LC1M4DeQYbijSXtTYz9fLKi67RuD4PXnh84oS/xc57aa1dzn\nFw1lyT9UlUziCzOUD6+zA3oAAJR9pyiL4HlGE2AAUk3qta38ipD5YYGmAjNo22P2DJoRZRf1\np/LORxVnppAMb6bzPPvwV2oO1xAu4Is6Kh+9Ze9dXlcGnae65KGQAgCU/beMFfJQ5HtHfmwp\ns89ZHneUXZcz8pQuUq2Y5fWLwNUeAED0LCt/dsXp0wbCm+n6qcDHz7wvsB4JCQkJCQmRkZGR\nkZEtWbZK562uvgE9qCFqvfrog1PGr6SiFGt3pxpjBm38wgy66ZIK1ccfKI1fWgPZ+AMo4I3g\nR1x1GCSTDLgj6jyPZTxhJ6YIh/7Frmwh4oQ/svf2NGbQta4QuvN6kI7PVH6JfL0BhggGn+Vx\nAYBgOH9h37dIMuiefYfRlW8AiHac4EPiQQrJwLui0FlMK10j3Lt375w5c/bu3dvC5UYZ0n6f\n3NmFzyQI1pRdAHB9ccTo+e/083GQhEzZdA+uL44Y9fqCnl4Ozp0m/ZRSx7uJwoNPtuat60MQ\nBMveq8fsHQ9r7Rny/3v/mWAnoYNvv/8dyAXDvT+nd3VxcI+csy2ntf9w8AyaCsygbQ/OxdHW\n5ez/JWPS8ZLFXarGkcuSpb1O3l6i/SV2/MbkTYQsqTjm9J3l5N6ZQxf9M3nfNIdqD6+x9Sw5\nD/TyrG2zJ39zPrpdtT1rDn34bta0fRl7PIu3vzzh82Pq4qWyeecfTOYcXzDwgLB1XxQWaCow\ng7Y9OBdHW+c9ZemohQuHDWIHDJ7/+eLhngB2/caP9RMC9Ili/FQAbqze4ycH2BMwdULU8uT7\nANULdLWtF/e8uXz+nwkPpGoDPDvC7d1qe1YnJydv+73TtpcMpIEkh3VIKOk89pf2Ij7ETRrw\nwYHWfVFYoKnADNr24FwcZjJ69GhPT09jiNyS5ca5D1n695Al6ozVQ0esG5X6GQMYjKoIlyQB\ndOd3bs8cMRL27UgIjPWt+WjTrT0jv9sRuSHlYJR899y+/+j01fe8JDQ0/Lk3t347MdiRQwAo\ndz7f5a+t6XGTuCfjT1XY1dxvy2CBpgLHQdsenIvDTEwv8bVkuVHXF4d0/TSVwXMOGf7RllCA\n2zUbCMMczk3tMDtNPOyz7ZMdGto6d2jCuRETg7e4DXg+yh3g+uKQniZ75nb7fOX5GXEdZt8t\nUhh837qS8cmSnRN6+rzfbtLkcNdm/WwaR5BkS2Pt+Pj4sLCw4ODgpj9k7dq1JSUlAMDj8d57\n770WdsDy8hWFI3dNMS7TLYNG1PzvZFH1DBqvn9uW1hnpbGl4Bk0FZtC2x7wZtG6/PPEiL2o5\n0+rn5fpD8muneVGfWb8n1pW3ro/HG+cef+c050jRj0Os2Z96YIGmAjNo22PWDFr/8FODx+7W\nqIl1zlpnJNfdf1Ge+Z9BJ2A4/U/Q5V0Wq86Vw3mSj5V57wo9ar3Jt20RK67/bvKt+7yz5Dxr\n9aXpsEBTgRm07TFnBn1Hk8PldHdphT3VOWudkeZPRbqCF/OIy81XJ/dUPHpO5Otd50qm2zP6\nu/Gkx6yn/By6TcACTQWOg7Y9ZhwHrbuo00XwOQAAoF5XfvUqW5SkLsgC0SxB+FdsTlW7o/Jj\nQ9U600cyOV11QtPrTnXOWlfZ1oNR+XEXAkDE4NrXu1LYjVm2Vwez2IDoDgs0FZhB2x4zZtDa\nYpJlMpWc/BoZsl/ciam7N0aefsyh0+MKC0MEg0lBw7uqd9Y6AOYovv/PFefEcpLBcF5r7yau\ndyXhQEARacDPEbcBWKCpwAza9pgxg2Y7EbrsJ4Ol+LFcJx8CgO05hriVYoDBj+tkE86g6521\nDqBihey+l6CfjM0u1d4dLrvXWxwUUfdKUkqCM4HVuS3AAk0FZtC2x4wZNCuGxdqk1wLLmCko\nD6iLX+M7MHU5e0nhSpM62YQz6DpnrTPSlZqMmDWQWmm9K2WJenEvPPLbBPw1UYEZtO0x51wc\nIRxPlSK/kOvtAgAgiCByRpddzwTRK4LwwY09tjpm7VnrjlZOaCd51871RcUFV72ez3R4UdC5\nPwBAXSv1BYeYnv/hFcI2AQs0FZhB2x6zjoNm+nzISFyr91rOBAAiiNPldz7FPRlnrfvCZM0Q\nwWDjAF4vToejnJp/jLVW6v9TlQzmB0goPj+yLCzQVGAGbXvMOxcHa5SgOz1+EcxnBN2fsXYn\nUFNhgaYCM2jbY6G5OLjzRL3MsmNkk7BAU4EZtO3B+aARDWGBpgIzaNuD80EjGsICTQVm0M1y\n+p/r0gJZv8kRDq7C+pat3UecDxrRERZoKjCDbpayQlnxozKdRt/AstXhfNCIhrBAU4EZNDVJ\nx9OLc8oB4NDPFwe/QK/7NGMGjWgICzQVmEFTo9fpSQMJAFq1zmCmO2dThRk0oiEs0FRgBt0s\no+f1MS5I3ISd+wYYl1kc1uh5rTEDZyvBDBrREBZoKjCDpobJZjLZTGv3om6YQSMawgJNBWbQ\n1CQdT795JsO4POj5bh5Bzg23tyTMoBENYYGmAjNoavQ6vUapNS5jBo1Qo9p8gT527Ni9e/cG\nDRoUFBR07Ngxe3v76Ohocz8pZtDUhA8MMs2grduZGjCDRjREr4OEAqlUmpOTo1QqjcuWeVLM\noKnBDBqhZmnzBdpo7969Bw4cIAgiMtISo2sxg6YGM2iEmsVGCrRGowEALpdrmafDDJoazKAR\napY2X6DHjx8/atQokiQBgMFgWKZGYwZNDWbQCDULvQ4Sang8noWfETNoajCDRqhZbKFAr169\nWqfTAYCTk9OsWbMs8IyYQVODGTRCzWILBVqlUmm1WgBQq9WWeUbMoKnBDBqhZrGFAv3WW29V\nZdCWeUbMoKnBDBqhZqHXQUINZtCmMIOmBjNoREO2UKAxgzaFGTQ1mEEjGrKFAo0ZtCnMoKnB\nDBrRkC0UaMygTWEGTQ1m0IiG6HWQUIMZtCnMoKnBDBrRkC0UaMygTWEGTQ1m0IiGGizQZGnC\nr4vf/WrbpTx2QL+Xv/xxxXAvOv7VYgZtCjNoajCDRjTUYIHO3P7j5c6L9y7r7lq46/WR4+d1\nyt/5vNBSPWs6zKBNYQZNDWbQiIYaPEgCZv+8wbjk/NyUfrOXPyoBoGGBxgzaFGbQ1GAGjWio\naWcxhoe//XSs23OL25m5N9RgBm0KM2hqMINGNNSEAk3m7Xk1do3rusPz25u/P1RgBm0KM2hq\nMINGNNRYgdY/2DZjxCesTw//GudB1z9azKBNYQZNDWbQiIYaPEh0mX89/8xK0VdHNjxL2+oM\nmEFXhxk0NZhBIxpqsEDf/HXptrR7EOf5CwAAeM2/kL0mBgAATp8+nZmZaWyVmpoaFhbW9KfU\naDTGO1QBgF6vV6lULaywmEGbwgyaGsygEQ01WKAjVqSTK+rc0q9fv379+hmX4+Pjm/hkGo3m\n1KlTly5dMtZTANBqtV9++WVUVNSgQYMol2nMoE1hBk0NZtCIhghjetsS8fHxYWFhwcHBDTeT\nyWR//vlnfn4+QdR8UoIgJBLJ9OnTHRwcKHRApVJZ+J6EdKa8lli0br1xWTw2TkSnDFqv1et1\nlTECi8NiMLEOItQQC12oMRgMW7duLSgoAIDa/xJIkiwtLd28efPs2bPZbHZzd44ZtCnMoKnB\nDBrRkIUK9PXr17OzsxtoQJJkYWHh5cuXe/fu3dydYwZtCjNoajCDRjRkoQJ99erV2slGDQRB\nXLlyhUKBxgzaFGbQ1GAGjWjIEgVao9Hk5uY2GnaTJCmVSqVSaXOTaBwHbQrHQVOD46ARDVni\nIJHJZE2/FFlRUdHcAo0ZtCnMoKnBDBrRkCUKNIfDaXpjCsMwMIM2hRk0NZhBIxqyRIEWCAQC\ngUChUDR6Hs1isSQSSXP3b+EM2kCSd0rS/ETt1Ho1k2CSQObJC9wFrhZ46qbADJoaumXQpK5c\nV3aZYIlY4uj6lq3dR2R2lijQBEGEhoZevny50WYhISEUhtlZMoO+Xnjz08vfZJRlAQADGCRB\nZstyzjy6OC5o1Pyuc+xYfHN3oFGYQVNDtwxar7hbcXUoSxQl6nmlvmVr97GlcuR5F3Ov5iuK\n2Aymv9g3xj1KwLazdqfoxUIHSZ8+fRITE/V6vcFgqLMBQRAEQQwYMIDCzi2WQR95cHLx+c+q\nTv0MYAASAMBAGuLT9l0vTPl5yNcijpWPc8ygqaFzBq0tOQ0ApLZUW3iA4NLlvVpLPJLlfX1t\n/cnsc6YruUzO8yETXukyncNo9lmarbJQgRaJROPGjdu+fTuDwahdowmCAIDY2FhnZyqhpGUy\n6DRpxkfnVxpIQ2VVrolMl2Z8cG7FuoFfmKkDTYQZNDV0zqBVGcsBQK+8J7/9ujBiu7W701I3\ni2/PO/GeTCuvsV5j0GxM2XI5P3HdgJX2HBreGsQKLPc2s2PHjlOnTt21a5dcXvMXw+VyR48e\nHRpKcSyEZTLobxN/0pM6Q4Mx+oXcq+dyLvf2tGY4iBk0NXTLoBkcd57/IgbPGwDYTs/oK5KA\nyWc59DRd3xYVKYvnn/xQrq3jipRxRUrx7Q/Pf752wKdW6Bz9WG4uDiONRpOYmHj06FFjSWUw\nGEOGDOnatSufTz29tcBcHEXK4hG7JjdcnQGAIBhDfPqt7POROfrQRDgXB6KzFZe/3pW+v9Gi\ns6b/p329Ypq784obG99+88s913KZgUPnrf1hkf/mnj4LLwGTK3YP6TXuf19+Pr2zoLJp5re9\nOy7uuKvil+Emj1f9Fct/Yb/JisBFV9JXRjW3G63I0hdqOBxOjx49Ll26VFJSYvy2V69eLdyn\nBTLo5KLbjVZnACBJw/XCZHN3pmGYQVND5wxaU7AHDGoAIJgCtstIa3eneW6X3L1dcte4rDPo\n9mYcavRAIgC+T/q1UFlUtaajY3BHx8ZOAcv3zY39RvDF5muxwZB59Ls/9j5cCND7m9yz8+wL\nMi5sWTTz2UX+N9f15QNA2b/Lz3gMdqq5B960feQ0ANXvsZ3TV6SviGjWCzULel1Jp8YCGXSp\nWtrEllJ1mTk60HSYQVND5wxanjKT1BQBAIPv5+CSae3uNM/pRxc2JG9q1kNIgDRpxqeXv6la\nM7vL9EYLtGrfxiPDvsqeGskGgLCxn60GyF4DAAAsgWvwkAWblh4J/PvCur6DQHtl5R/ui17j\nTGzqRMnWYwsF2gIZtIgjamJLe2uP4sAMmhq6ZdCmOC5jSH0FABAcF2v3hb4ePXzoExRU//gP\n+6Ag9s5sBUDxj18UzPq1O2uRBTtHlS0UaAuMg+7k1KSEnQCii3NHM/WhiXAcNDV0GwdtStD5\nV2t3oQ3w8vF5eDxdC/XV6Ir0dK33s3ble+af7Lf4XzEctWz3qKHXQUKNBTJoT4F7uEtoctFt\nA1n3OG4jEsiRfkPM3ZmGYQZNDWbQZtLPq6cLvzLtlWnla69vaPRqDkGAu53rjNDnq9Y0HkAD\n8GJnDH3vnXl/uy6LbQ+ZR7/7o3TOQuMWvaIw89LWRZ/cmbypJ9x4/+COb38h5gMAwE/eDhey\nVzX7YqTl2EKBtsw46Pld58w8Mp8Agqx7HDQwgAh1Dhncrp+ZOtBEmEFTgxm0mdS4vnfm0cXr\nhTcbOdEhYUrwuHFBsc17JlHs+r35b89/LmJOLito2BvfrveBzXBuoQfxFkfkHtJr3LJdK/vy\nAdZkk2sAAI6+6rc1LovO1Rlso0BbZhx0uHPo25Gvr7r6PUEwyFp/XgQQLnbOX/ZZSoCVD2zM\noKnBDNoy5oXPnHV0QQMNGMBwsXOa0H4MhZ2LwmduODlzw5MVCy6S9T7XkB+z6nm3y3tpXzqF\nZzcHWyjQFpuLY0rwWBe+0xdX1harShmE8RgmjOcCfbxiPu7xtiOPyj0VWxdm0NRgBm0Z4S6d\n54W/svb6hjrv4MEgGBwGZ3W/T3gsi9xcNG9dH483qj5v7jTnSNGPVo4oa6DXQUKNJeeDHuzT\nr7dHj8MPju/PPFKmrmAzWH5i36kdxnZypMt9VTCDpgYzaIt5sdNkCU/81dV1Cp2SQRDGTxgY\nF3ztvT/v81F7hwALdcV93llynoWeixJbKNAWng+ax+L2cI9adnGV8VtXOxf6VGfADJoqzKAt\naUzA8P5evfZm/HcxLyFHlsdhsQNEvgO8+wxp149BWOi+SG2CLRRovCehKcygqcEM2sLEXNG0\njhOndZxo7Y7Qmi0UaLwnoSnMoKnBDBrREL0OEmrwnoSmMIOmBjNoREO2UKDxnoSmMIOmBjNo\nREO2UKAxgzaFGTQ1mEEjGrKFAo0ZtCnMoKnBDBrREL0OEmowgzaFGTQ1mEEjGrKFAo0ZtCnM\noKnBDBrRkC0UaMygTWEGTQ1m0IiGbKFAYwZtCjNoajCDRjREr4OEGsygTTHs7XmdOpI6HcFk\nspxr3XXNqjCDpgYz6KeWLRRozKCN9FJp+Z598gsXycdRj/pumuJaonhcHE2CDsygqcEM+qll\nCwUaM2gAUKelF61dZ5DLTVeSer3yWqIq6YbjjJfsYnpYq29VMIOmBjPop5YtFGjMoLW5uUXf\nfGuo8/8TSZJ6ffEvGxn2Ql6olaMYzKCpwQz6qUWvg4QazKBLN202qNVQ373eSJIgoPS3Te6f\nryDY9d/12PzonEEfypKnlep0BtKBy5gcIrRj0WjSS5vMoEmdTpV0Q3UzRVdaSjCZbE9PftcI\nToC/tftFL7ZQoJ/yDFr74KE6NbXhNqSB1JWUKBOuWTfooGcGfeS+4ufk8uwKXdWaP2/LxrcX\nvBhqL2TTokzbXgatSrlV+sefuqIiIAggCABSeT2pfP8BXudQx5dfYkqsf2cimrCFAv2UZ9DK\nmylNakcQqpsp1i3QdMug9SR8fqnkQKaiRuas1hk236449VD59QBnb3vrHyM2lkErLl4q/vnX\nyrt3kqTpOz9Vyq38ZctdP1jEcnW1Wv/oxPp/fC33lGfQ+uLiOm/vVpuuqNgC/WkA3TLo7xKl\nBzIVAFDjn4Xxu0dy3f9OFm0c7mr182hbyqC12Y9Kfv2NAKh952UAAJLUV1QUrf3ebdnHBJOm\naZgl0eIdXAvxeDw+n8/n87lci9xoEkCtVx99cMr4lVTUtBNYsyGYzKZUZwIIgmXlv3gmm8nh\ns41fDKaVR0qklmq3p8oaaECSkC3T/ZFSYbEu1UdTsEeTt12Tt11beMDafWmpsh07SYOhob9Y\nktTm5CjOnqew84obG+cMCPEQib27Tlh5ppDMXhNDEATB4jl4R4x8c9NN4xin/P1v9fMT29l7\n95gbf79aP1R/xRLVBL13lUI3WpH1z2Ja7inPoFnubk1rSDa5pbnQKoOOv9tQdTYiAHakyWeH\niaz7uZU2nUFrsu5r7t83LpNqtTLpRr1Xs6sQRPmh/0h40ozj68vx823kUeX75sZ+I/hi87XY\nYMg8+t0fex8uBOj9Te7ZefYFGRe2LJr57CL/m+v6Zv707rGe2zIPBz9cNXzst1cmfB1dtQfe\ntH3kNADV77Gd01ekr4ig9opbky0U6Kc8g+aFh8GWrWCo6w2jCZIk+V2t/BdHqwz6cq6q0R6Q\nAEqd4VaxJtzFQm/O6tSmM2hlUlL57r3NewxJ6vLzS//4s2qF6NnRjRZo1b6NR4Z9lT01kg0A\nYWM/Ww2QvQYAAFgC1+AhCzYtPRL494V1fTu6uXErzxJYHh40/4HaQoF+yjNolpOToGeM/FyD\nbwkJguPny+vUyVKdqht9MmgDCcUqfRMb58n14VY9jm0pgzafRw8f+gQF1T+M1D4oiL0zWwGD\nnv+ob88BTl8qOYGv7rxE82F9bbtAkySZmZmZlpYmlUqZTKZEIunQoYO3t7e5n5du46AdpkxS\n303TFRXV/eaRwWBwuU6vzALCyrEvfcZBMwhgEYSmCdk9AHCtHZfb5DjoVufl4/PweLoW6qvR\nFenpWu9n7Qyn3piT8so16Ryvoq0vjliwf/Cfo6z57qgRbbhA5+bm7t69Oy8vDwAIggAAkiTP\nnDnj7+8/ZswYiURivqemVQYNAAyBwPW9d4q++16Tdb/aiA6CAJJkbWVbDwAAIABJREFUOTg4\nz59n9QAaaJZB+4pY96TaRoIhAADwE1v5MGnTGTQ/PJzpUDmumdRqpVu3NRrHAUGw3N3thw2p\nWsHxbSyABuDFzhj63jvz/nZdFtseMo9+90fpnIXGLXpFYealrYs+uTN5U08gD0vLAYAAIA1k\neVlZU99HWUdbLdAZGRlbtmwxPP5Nm14UzsrK2rBhw0svveTmZq6SRKsM2ogpkbgt/kB+7rzs\n5GlNZuUxzHJ3FfbtKxw0kOBwrNs9I1pl0AN8+GlSbcNtGAThLWT5iaz58Uto4xk0x6/a9T1N\n+j3FlauNXCckSfHoUc0esy+KXb83/+35z0XMyWUFDXvj2/U+sBnOLfQg3uKI3EN6jVu2a2Vf\nPsDI91dse2GQ16Jyu/YjPv5jvB2lV2UpTRo/27D4+PiwsLDg4OCmP2Tt2rUlJSUAwOPx3nvv\nveY+Y3l5+fr169VqdX2dZzAYIpHo9ddfZ1v1k81WobyWWLRuvXFZPDZORKf5oPVavV5X+T+V\nxWFZd6SdTGMYvzdPpiUbPgSW93Ya3I5vsV7ZPF1BQd7Hy0itDuocBw0ABMHx93P78H2rJ3J0\n0CbHQZ8+fVqlUjVwXBkMBqlUeunSJTN1gFbjoGsw3pPQ+EXDexLSZxy0kMNY1ssRgGxgBF1s\ngIAO1dmWxkGzXF2dXp1NMBl111+CYDk5Oc97zULVOW9dH5NRz86vHrXEkzZH24s49Hp9cnJy\no80Igrh+/XqfPn3M0Qe6ZdCm8J6ETRfjwVvd33nJ+ZIKTc2zOQYBkzvYvx4htkrHamjTGXRt\n/Ihw1/ffLf39T83DhwBAEAQJJJAABGHXI1ry/FSGQGChrrjPO0vOs9BzUdL2CnRxcXFTxjuT\nJFlUVKRWq83x8UIaZtBV8J6EzRLjwYsf7R5/V7YjTV6q1pMAHAYxqJ3dlA7C9hK65GNtOoOu\nE8ff323pR+rUu6pbt3TFJQSLxfZw53ftynLDKTiqaXsFWqFQNL2xUqk0R4Gm1TjoGvCehM1l\nz2G83Fn0cmcRABhIoNlk/QC2Og6aILghHbgh9Dq/oRu6HCRNZ2fXjMuuzWrcdHQbB23KmEEb\nl2mYQdNkHHRteE9CRENtr0A7OTnxeDyVStVwM4IgnJ2dOeYZXoYZNDV0y6BN4T0JEQ21vQLN\nZDK7dOly5cqVhpuRJBkRYa6pJzCDpoaGGXQVvCchoqG2V6ABoF+/fjdv3mxgpB1BEBKJJDo6\nus6tLYcZNDX0zKCN8J6EiIbodZA0kb29/eTJkzdv3qzX6w21PjZKEASfz586dar5PqWCGTQ1\nmEFTgxn0U6tNFmgA8PPzmzlz5p49e3JycsBkLg4ACAwMHD16tFhsxhGsmEFTgxk0NZhBP7Xa\naoEGAHd391deeeX+/fubNm0ynkfb2dlNmzbN09PT3E+NGTQ1mEFTgxn0U6tNzsVRQ1UYzWAw\nLHbXK9rCuTgQshlt+Ay6Co/Hs/AzYgZNDWbQ1GAG/dSyhQL9lN+TsAbMoKnBDBrRkC0U6Kf8\nnoQ1YAZNDWbQiIZsoUA/5fckrAHHQVOD46ARDdHrIKEGM2hTmEFTgxk0oiFbKNCYQZvCDJoa\nzKAtTC5V3rn0IDu1UFGmZDAZEjf7dp3c2kd50/ZfuFXYQoHGDNoUZtDUYAZtSTdPZyT8l2rQ\nG6rucaysUD9KK0w6kT7w+W6uvma843PbguOgbQ2Og0Y0l3Dozo2T9+rcZLz11DOzot0DnCzc\nK3qyhTNozKBNYQZNDWbQlvHobmF91Rkez9ZwfHPCxHcHsbm2UJ1ayBZ+BJhBm8IMmhrMoC3j\n2uG7VbFGnUiSVMu1t85lhQ8Kat6us9fE+CwL+Ddryzgx6LbGdb654s6KztlrYqbArrML3BVJ\nq5598eIL8VunB7WhqteGulovzKBNYQZNDWbQZlL8qKzoUZlxWa3QFmVLG38MQdy+dJ8nfHK3\nDWcvsZNXE6Y/C25358u1d8Z+FFJjvSJp1bPTLz7/b9uqzmAbBRrHQZvCcdDU4DhoM3lwO//6\n0bTmPYYklWWq8zuSq1ZEDGnfpALt8eKHXTat2D3/r1iTlcobq5/9LeO5+H9eamPVGWyjQGMG\nbQozaGowg7YJnP7vvbp+wnd3YwOfrEvf+avn9P2T2rfFYmeFPt+6dUsulxuXtVrt9evXW3hv\nKsygTWEGTQ1m0DbCa8bibtEr9rz+ZE2X9+OnHn85bu2OfW92amuDvKxQoLOzs6vCYr1en5WV\n1cICjRm0KcygqcEM2kzadXSzE1W+xzUYyMv7bhn0Bmj4l08QTp6iDj3aVa1wbkq+UYkz8L3Z\nn4z/vhCqYg5Op9d3/CV9ZvxUh4Pbp/vR9C1c3driWX9NmEGbwgyaGsygzcSp+vW9iiJ5yrnG\n3gSQZPeRHT0CqQ6F9pm5OPKzISdMV9lFfbj3h3eGjH5FdOTXOHd6/f9tCL0OEmowgzZFzwza\nYCAzEh9lXM8pflSmUel49hx3f8fg7u2oH4StDTNoywgf1D7rZq68TA31j7QL7OpF5Q/De8HF\nk8Yl7uAfssmqlWeNSw4DvrqaXOcDacwKnyQsLy+XyWRV39rZ2Tk4OLSkA5bPoPMVhSN3TTEu\n0y2DpqHyYvmxTVel+TIAAh6/uTUu+XXx6DMxjE2Ds+n/nSyqnkFb6N1YU5SecHmSQfdr8xl0\naX7F4V8vKco1UCPpIAggSa9gl8EvRNL2YrKFWeHAEIlEIpGoFXeIGbQpumXQslLlvvXnNAod\nAJgekMalrJu5ijLV8NkxTJaVCyJm0BYjcbMf82bfa/+lpidkm16K4PJZ4YPad+rlR9Ds529F\n1pmLo3XhXBym6DUXBwn7fzhf+FDa8J9Z2MCgyGfo9X8OWYBGqc29VyyTKplMhthV6ObvyMDS\nXJ3131q2HGbQpmiVQT9KKyx4UNpwGwLg5pnMzn0DuHZsy/SqTphBWx6Hz/bt7G7tXtCaLRRo\nHAdtilbjoO/fzDPNnetEApA6fXZqQWBXL4t1rDYcB41oyBYKNGbQpmiVQUsLZQTRwOV6k5YF\nssYbmRNm0IiGbKFA4zhoU7QaB63T6JtUngnQa/Xm705DcBw0oiFbKNCYQZuiVQYtEPNKcsob\nr9EkCMSW/iXWgBk0oiFbKNCYQZuiVQbtGeT84FZ+E1uauzMNwwwa0ZAtFGjMoE3RKoMO7OqV\ncDhVp9Y1dA5NEO6+EolHaw6NpwAzaERDOA66GdKkGedyLrd3COjtGV3fsvmevYnoNQ4aIO3q\nw7PxN+obyUEQBINFjH69j8SdvhEwQtZiC2fQFsugU4rvfHf957jAkVFu4fFpe+LT9nZ36yrk\nCDLL7hvX06FA0yqDBoD2UT6KcvW1I6kEAaSh2iaCACabMWhaJB2qM2bQiIZsoUBbPoM2kKRS\npwIArUGr1Cot8IxNR6sM2ih8UJBrO4crB+8UP771EQAQBLQLdY8aESJyElixb1Uwg0Y0ZAsF\n2vIZNJ/F6+Yatj/ziJ+oXYxH1K57ByzzvE1Bqwy6ikeQ85g3+lzYffPOhfvGNf2ndvMPo0Xf\njDCDRjSEGXQzPJLl3S5J9RS6d3LsUN+y+Z69ieiWQZvSa/V6XWWMwOKwGEx61UGE6MYWzqAt\nlkF7Cd29hO4NL1sd3TJoU3hPQmowg35q2UKBtnwGTWc0zKCr4D0JqbHhDFpnIEtUBjaDkPBo\nNAE3fdhCgbZ8Bk1n9MygjfCehNTYZAZ9IVe1LVWWkK/WGUgAsOcw+nvzp3ey97a3haLUWjCD\ntjWYQSOaU+vJTy+VHr2vYBBg+m+aIAgmAQsjHcYG0WJgDx3Ywj8ry8/FQWeYQVODGbRlGEj4\n+HzJmWylcdkUSZJ6Er66UsokYEwg1mgA2yjQmEGbwgyaGsygLWN/htxYnetEAhAErL4q7eHB\nc7Oj6f9yS7KFAo0ZtCnMoKnBDNoyfkupYABhqP8eDiQJWpL8+07Fgm7NvJd09poYn4WXAFhC\nr8jnvt7246R29Po1UoEZtK3BDBrRyp0STWpJ5X/lIqX+15vlTXmUmMN8NfzJ/FkdHNkhjpxG\nHpO9JmYK7Do736n02tKhEw0/ZXweSbXTtGELZ9CYQZvCDJoazKDN5NwjVROLsqkyjf6LK0/u\nZjmzs6jxAl2JIAAAHMRWnh6xddhCgcYM2hRm0NRgBm0Lzi30IBYCMDzifr0YZO3OtAZbKNCY\nQZvCDJoa2mbQurIrTJ6PgSkAIAium77iOtM+wtqdoqve3+SeXeCiyN712qC5v8fuf5lGZwDU\n2EKBtvw9CemMVvckrCF8YFDnvgHGZRaHXn97NLwnoV5+V3FrtrbkFAAAMAAAlFll57uynZ8R\ndPqJwfe1au+aqrcXz5lfmWuRJKy7XqbUNfLPmQAizIUz3M+uak0HR3bTn5EgAEBVVFQBYOMF\nOmtllP/7CQDg9daF7FUxlulS8yQnJ+fn53fu3Nnd3T05OZnP5wcF2cR7G6owg6aGbhm0ruxy\nxdVhpK7i8Yonc2nrio+UXYgUdT/BtO9ilb41S4gjxzQ+VuoN3yWWNdAeAAiCXBTt4CdqRlGu\ndG6hB7GQ8f/27jwuqnp/A/j3OzsMiwIiAuIuJuCa5jUlt1xI+yUKLe5LXkrzelt8lZpm+2Iu\n/bAsDVIxtyxT8ap5C7cWdyVzJYhAEQHZGWBmzv1jAEFhgHHmnA8zz/vlH6dzzgwfk3k4PHPO\nHCePTgNnf7a8XaMfTk9DzuI4+2rH0fq4OgNa2rM4duzYkZiYqFKp5HI557x3795Dhw6VZBKo\nF+UO+sWErJodtJS/jQnl2XlHg4SyWwIz1r4Hl8nUrd0H/M7lLuKOdr/KjcLs/966kFVmJnee\nDXGbFmwX7/HdN1q/ZlosODjYz8/v3LlzUg8iPXTQliHVQZckf2AsM3uzXcFo1P2lS1np1GGR\nWENZh1LGPwz1evVI9vlbpTLOjdUOEGWcCQKb2NV1KtK5kp0EdLt27UJCQlJSUqQeRHrooC1D\nqIMWjGXp6xmr6z6OVXhpekyTC2jGWDO1LHqI166koi2XCtMK9aaVnLOe3uoZwW49vB39Uobq\naL1ILDBw4MCePXt6e3ublhWKJv83uk/ooC1Dp4M26lKMZZkN2FEwliQbyzJlKm+bz2RtChkP\n7+QS3sklo8hws1ivkPEAV4WrCm/y363Jx5m3t7cpnU3L0g5DAc6Dtgyd86CNZVkN31koy2JN\nMKCr+GjlPlqiP7MpaPIBDXdBB20ZOh00V3o0YmeVp+0mAcnVcxbHzomasZsqr/6QP/mdfssT\n9+yzYcOGzZs3u7qSqfAcW42A9vMj1XIIRqHq+43LZJzQ5SCkCGWZu5ixrN79uNxZ2cI67zFE\nRERERERY5anAiuo5gn4iTifE1bL+8OHDyckVl5yePXt23759Vp8MrCAFlwXbt2LGtlvlibp2\n7YqAJsjCiiM0NDQ0NNS0zDkfOXIkjqChXrmZhflZRaZl74BmGhdC79fTeZOQMSYYdeVZ+5mx\nvM4TOTjnXK1sMZLxxl/NUZugoCCrPA9YlxU6aGdn51WrVkn4caPAGMuP31ty6rTbY2Gq9u1z\nt20v+zPZqVdPbeiAktNnTOudeveSekZ2+sDlcz9W1C/Dp/f160zo043XXyi4mlvRKizo61HV\nR0ulPOenwlNhglDGhHuuVeEyLnN27XNQ4f6QFKOBePAmoZ0wZOeUpfxlKCgoS04u/u04Y6xg\n/wGZs7Ph9m3TeqkHZAznQTeG0mOw60PHihKnGQrPM8YZlzGBMWZkTFC4PagN+UqufUDqGcHm\naL1I4P5punfTDni46OixZk9GugwZlLt5q9QT3YHzoBtF4dbLvf+Z8uwfdKnRxtIbjDG5prW6\nzQtKj8GMST8eiAABbW+4XM4VCsYYV6u40joFpbXgPOhG4zKl14jCxImmz4MWyrNdPIZIPROI\nBwFtJ5pPnth88kTzy5LDedCWsad7EkKj2MM9CaGpwD0JARoFR9AgHnTQlmnS9ySE+4GABvGg\ng7YM7knosBDQIB500Jax2w5aMOoLzhp1f3OZSq59QObUVuqByEEHDeJBBw0mgqFYl7K89K9V\nxvI7H90ndwlx6rhU1XKshINRgyNoEA86aMvYWQdt1KUVnAozFCaymh+XZSi6UHg2XO0/U9t1\nDeNEv09EhoAG8aCDtow9ddCCoajgVJix8AJjjN3167tgZIyVpq3jCjfnwI+lmI4c3MIAxGPq\noE1/CHbQQwKcTH8IdtAqnwiVT4QdHD7rUpYZChPrvBluxT4r9PmnGv3UaSv7cfX/bcxljDFm\n2P+sN28971cLx6QCR9AgHnwWh2W0wV9KPYLl9PmnDFVpKxh1yR/Ve7tFzlnxxTlqv2lVa+Ru\nvRVuvev/YiFBt3buyp802U3/0/Y/W/dq+jUJrRcJ2Dd00JZp0h10eebukqSljXqIIAj63F/1\nuXcOf506LGlQQHs8Mab8+935k5/8dXvy4IhOSRmMMSakbIua8NKW8yW+g+bHbpqv+bDH6wWP\n5n27NlE7clX8lknyGlv70bqhOAIaxIMO2jL21EHbmMfYx/Wv787xSrg65OVRiesyGGO67xfN\nT59wIPWA+4lFI55b+XgsK0zM7Z9wcUn5utHjYk64/VVj65nFXaT+O1SHgAbx4Dxoy9jtedA2\n4DX28dKZL8bkDXozkCUyxhhLv3w5aNxnD7hr2bCnHy1YeYW1dw4dN7atC2MDHpR9fuburQwB\nDQ4KHbRlmnQHrfQeI9P4mpaNpTdLri2u/zGcy5QtnDq9VbVC3pB+w8TzidFFUasHnQtkV0wr\n/AID/4jZdCl8stuJzT+4BM5i5TJZ1ckRgu/dW2mh9SIB+4YO2jJNuoNW1Hx/r+zG14aiK8zs\nWRxMEFR+09T+lqWl56wD5bMYY5UBrXn8zffiJwxt/bLO95FXvtzUhS2rvvc9W2nBlYQgHsod\n9IsJWTU7aEJnoN7+qcWdDjq0aXfQZRlbC889ZXYXGZc7uYcmyVQtRZqJMBxBg3gIdtDXC/WM\nMV8XRfcWatNBs0bBVTJetV7a8UzsqYNW+TypvrW39PqGOrbLGBO0wbFIZxMcQYN4CH4WR//N\naYyxn5/2N7MMViaUF1/6ty71U8Z5tVvicsYELnPWhsSqfCKlHI8SEgcI4CAod9DZOoNp4Xqh\n3kdL63XRpDvoWnCl8wPRqlYTdamflGftF8pzGONyp/ZKn3GaNv+WqX2kno8QWt+IYN8od9Ce\nmoqfHERqjers8jxoRbN+Ls36McYEQzGXKRmndf9MIsh9L4IdI9hBDwlwqndZcvbUQd+Ly52l\nHoEudNAgHoIdNABlOIIG8VDuoCmztw4aGgwBDeKh3EFTZpcdNDQEAhrEQ7CDbhLsu4MGMxDQ\nIB7Kn8VBWZP+LA64H3iRgHjQQVsGHbTDQkCDeNBBWwYdtMNCQIN40EFbBh20w0JAg3jQQVsG\nHbTDwosExIMO2jLooB0WAhpsLjez0FBucPd2uXAkOfFQEmOMy/iQCb2c3DSm9QqktlnooB0W\nAhps7si2s1lpeWPmDDDoDeWletNKo1GoWu/l7y7thMShg3ZYCGgQT/fBHdMuZWZfzx/5bL+W\nbT0Yuyz1RE0DOmiHhYAG8ciVci7jjDGlGp+U1AjooB0WAhpsrk2Qj6evu0arMrMMZqCDdlgI\naLC5boM71rsMZqCDdlgIaADq0EE7LAQ0AHXooB0WAhqAOnTQDgsBDUAdOmiHhYAGoA4dtMNC\nQANQhw7aYSGgAahDB+2wENAA1KGDdlgIaADq0EE7LAQ0AHXooB0WAhqAOnTQDgsBDUCRUZem\nS42Wafw1AXOUHkP1Bee43EnR7B/V10s9I9icTOoBAKAWxrIMXfIHZenrGWOadi8biy5xJtc+\nsLr6erB7CGgAAKIQ0AAARKGDBqBI7tzZ9cEfuMLNzDLYPQQ0AEVc4ab0HGZ+GeweKg4AAKIQ\n0AAARCGgAQCIQkADABCFgAYAIAoBDQBAFAIaAIAoBDQAAFEIaAAAohDQAABEIaABAIhCQAMA\nEIWABgAgCgENAEAUAhoAgCgENAAAUQhoAACiENAAAEQhoAEAiEJAAwAQhYAGACAKAQ0AQBQC\nGgCAKAQ0AABRCGgAAKIQ0AAARCGgAQCIQkADABCFgAYAIAoBDQBAFAIaAIAoBDQAAFEIaAAA\nohDQAABEIaABAIhCQAMAEIWABgAgCgENAEAUAhoAgCgENAAAUQhoAACiENAAAEQhoAEAiEJA\nAwAQhYAGACAKAQ0AQBQCGgCAqHoCuvTiV1P7+mnVWv+HZm66Wi7OTAAAwOoL6NTVM+ZeHvl1\n0s2kTSMuzJ71xQ2RpgIAAPMB/feOredGvrLgEZ9mPo8seGX48c07MsSaCwDA4ZkN6NTUVL+O\nHTWMMcY0HTv6paamijIUAADU/yYh57xiSRAEWw8DAAB3mA3ogICAtKtXdYwxxnRJSdcDAgJE\nGQoAAOoJ6NbhESH7Pnr3UEZuxqF3P9rf56lwH7HmAgBweAqzW9vM+XLl+SlPt38/r3mPyOiN\ns3wrN+Tk5JSUlJiW8/LybDoiAIBjMh/QTBM0fcPJ6RvuWZ+enn7z5k3TcmZmpg0GAwBwdPUE\ndF1CQkJCQkJMy7m5udabBwAAKuBSbwAAohDQAABEIaABAIhCQAMAEIWABgAgCgENAEAUAhoA\ngCgENAAAUQhoAACiENAAAEQhoAEAiEJAAwAQhYAGACAKAQ0AQBQCGgCAKAQ0AABRCGgAAKIQ\n0AAARHFBEO7zKZKTk6Ojo5VKpVUGuktBQQHn3MXFxRZPfp8yMzNbtGjBOZd6kFpkZGT4+BC9\nBzvZ2QwGw+3bt728vKQepBYFBQUymUyr1dro+ZcuXapWq2305GAxKwS0TZ06dUqpVHbr1k3q\nQWoRFxcXGRmpUqmkHqQWMTEx06dPl3qK2pGdrbCwcO/evZGRkVIPUovjx487OzsHBwdLPQiI\nChUHAABRCGgAAKIUUg9QD7VarVAQHdLV1ZVmAc0Yc3Nzk3qEOpGdTSaT0Xy3gzGm0Wholmlg\nU9Q7aAAAh4WKAwCAKAQ0AABRCGgAAKIoBXR27Ag1D5h31HhnVUKUV5c3fq++071rrKo8Zdfi\nyL7tPJycmrXpEfZS3O8FLCHKi3POZQqNe6vO/ce9GHvydrUHGP74oL8T95l39J6n0sWN5vcY\nFG29wRirczbDn9vnDQ301mq0LQKHzPvmT4NNZzMnIcqLK3q/c7ninY6jc3yC377Eavw7Fp9b\nMSKg68xv04x1P41tVfxv5Fzh7Nlh4IyYP8rE/dKtXzhW+S+UsWaYivNB0RmiTQCUEQrov+Ni\nEsPGt90ac1C8l8ddsr+bNmDq/lbztp65npV6LG5emwNf7M5mjAUuSRSMuqwrxzbN65q4MHTo\nu+crXk+GP5bN+SkkvHNtT6aZuEcQBEEQzizs0HL2EdNywhyrDlbHbPr4RVN2d1x2IjMv88SH\nHXZNXhSvt91s9Wrln71q0bac2jfm/fLG8Mc2Bq07tDbcX8pvxsAliYIglN4690X3X2Yt3F4k\n4pcOGtgufs2eYsYYY9fWrS3s9zCu6IMKdAL62vrYv56e//mMB3fH7CqUZoSra5dub//2jlXP\n9G3TXOvmHzx87oZPnvGs3Kpwadm+T+Rb364e+/dHK/YZGGOGP96f/eP4T+e0s/W5duYHu3c2\nuW9rXzljnHHGGZP7tvaT8kxF32fffOr46++d1N+9Qcg8OG9QxMFhW39aPrwFkfMVOWPMzcPD\nJp9bUAfNY1FP/Pb5lhzGjL99Htdy1tS2pvXGv+NfHdXVx9W1RechL2y/Zjps+WY8H/nqisju\nvu7ufv1m70yre0+wB1QCWjgd85Vu8rR/eIyfOuxgzLbb9T/C+gxnTp/vOHiwv/m9XB95pGfu\nhQvpTJ/4/uwj4z99rr3No6Vhg1WbjT+4eMPEM5PbuKhd2k49N2nj4t62HtEsdeiiJR3Xv/Zl\nes3V6asjxx57/PsDbzzsLs1c1V1eGsI5V7j4j9oTvGrJKFHPOJb3j5qU8cVXKaX/WbOzb9RT\nHqa1GetnTtjb5eNf0q+fXdPv+MyJn/xZsfuRo7efj79y4+LaHvvmLj9ubk9o8ogEtPFw7MZm\nU6cGM6YdM2X0yZhN6fU/xuoaekq4IDDG9IkfzD4c/tnzHUQ48mvEueqmPQvinxu3qceXl/OK\n8y6t7RYX/ly8RL+TVPKe8l5U+ttLfyypvtKzf1iv1C2f/XiDwpn4porDUHzj2HM5L0cuuyLu\nV+8881mXmI8Wrvl55D8fczatMvxyMKHvnKWj2rm5+g1ZOn/46R8OFZi2DI5aMMjfxdk3bEz/\nrKSkcjN7QpNHI6BL98d8nXbqtU6cc64etf7WsZj1Ir9CGGNM0at3t2sJCfX8bMhPOHSmWXCw\nPv67IwfmtpdxzkPeuXxz1UA++itbpWDDBrszm//FQwmKsBnhnd2c3ALHzwyTHzp80UajNZS8\n+yvvhO5+7f+r/7Nqes3fu2fqlRmDZ+0hkdGMMZmTT59p4UG/iR5xrSZGddn4ccozUQ/LK1fV\n+KksCKzqolWNRmNakMtlBoPBzJ7Q5JEI6ILvY7/rsjxNqKA/9EJGbMxZ8efoNPP18KuLxr+4\n5WTq7eL89N8PfDJl7tfZlVv1RZnJJ7YtHjdnZ8Ar/xrR+dWTlfMmLgxs+a8jwp6pNrtK2Pxg\n98wm6xAUlLM3due1gpL8q9/F7M0NDu5gq9EazHXsW/P58rfiaxxEa/u8tnf35MszBs+Kz6CQ\n0UZd5qn1O0/7duniKvJXdh3/Tb6QuCSoaoW8/6ODjkcv+U9yQeH1hKUf/9BrWGgdMzV8T2hy\nKAR0zrbY+D4Tn/ar/G/5wCkTDBtiDpvOlDCVgybD1uXWusZavMatP7puSOqycd1beQT8Y8KK\nlGGzxnhWfkWNR4d+z6z4Pfitw/9d0F1e73NZVV2D1TGb56Th2U0UAAAApklEQVTVG8OSXnvY\n18NvwIKUMXHREzzEnbdW7Z9/f7wxJf+utS59F+zdPenS9MH/lDSjTd9UyuZdwrc3X7j9jYHS\nTVLJZ8q6TWEXX+rn69N91s+918XNbX//e0JTg8/iAAAgisIRNAAA1AIBbT1pK/vVvDSv2cx9\nUs9UifJsAFAHVBwAAEThCBoAgCgENAAAUQhoAACiENAAAEQhoAEAiEJAAwAQ9T9E2EgZsFOq\nFgAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 240,
       "width": 240
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Create the plot\n",
    "options(repr.plot.width = 4, repr.plot.height = 4)\n",
    "or_plot = ggplot(odds_ratio_plot, aes(x = cell_type, y = estimate, color = cell_type, linetype = type )) +\n",
    "  geom_point(size = 3, position = position_dodge(width = 1)) +  cell_type_color +   geom_text(aes(y = conf_upper +0.5, label =paste0('OR: ', round(estimate, 2), \"\\n\", ' (p = ', signif(p_value, digits = 2), ')')  ), size = 2, position = position_dodge(width = 1))  +                           # Plot points\n",
    "  geom_errorbar(aes(ymin = conf_lower, \n",
    "                    ymax = conf_upper), position = position_dodge(width = 1), \n",
    "                width = 0.2, size = 0.8) +  plot_config  +  geom_hline(yintercept=1) + ylim(0,max(odds_ratio_plot$conf_upper)+0.5) + plot_config + theme(axis.title.y = element_blank(), axis.title.x = element_blank())\n",
    "\n",
    "or_plot\n",
    "\n",
    "pdf(paste0('figures/G2/G2_OR_Plot_combined_reduced', use_gwas, remove_HLA, only_dt, only_esnps, coloc_thres,use_gwas , cs_snps,  '.pdf'), width =4.5, height =3)\n",
    "or_plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "id": "fd7378c9-f9c6-4426-8870-4b709ebe9467",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 7 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>estimate</th><th scope=col>conf_upper</th><th scope=col>conf_lower</th><th scope=col>cell_type</th><th scope=col>type</th><th scope=col>excluded_genes_without_cs</th><th scope=col>trait_filter</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>6</th><td>0.0006863434</td><td>1.939638</td><td>2.784903</td><td>1.3238773</td><td>ALL  </td><td>eGene_based</td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>0.0187006300</td><td>1.531563</td><td>2.180496</td><td>1.0527763</td><td>ALL  </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>0.1422117000</td><td>1.598469</td><td>3.021836</td><td>0.7877197</td><td>B    </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>0.0037729260</td><td>1.862487</td><td>2.812239</td><td>1.2007029</td><td>CD4_T</td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>10</th><td>0.0496152300</td><td>1.632734</td><td>2.655485</td><td>0.9681380</td><td>CD8_T</td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>0.8767768000</td><td>1.025278</td><td>1.905349</td><td>0.5121500</td><td>Mono </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "\t<tr><th scope=row>12</th><td>0.4340997000</td><td>1.254892</td><td>2.250417</td><td>0.6577192</td><td>NK   </td><td>snp_based  </td><td>FALSE</td><td>disease</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 7 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & p\\_value & estimate & conf\\_upper & conf\\_lower & cell\\_type & type & excluded\\_genes\\_without\\_cs & trait\\_filter\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr> & <lgl> & <chr>\\\\\n",
       "\\hline\n",
       "\t6 & 0.0006863434 & 1.939638 & 2.784903 & 1.3238773 & ALL   & eGene\\_based & FALSE & disease\\\\\n",
       "\t7 & 0.0187006300 & 1.531563 & 2.180496 & 1.0527763 & ALL   & snp\\_based   & FALSE & disease\\\\\n",
       "\t8 & 0.1422117000 & 1.598469 & 3.021836 & 0.7877197 & B     & snp\\_based   & FALSE & disease\\\\\n",
       "\t9 & 0.0037729260 & 1.862487 & 2.812239 & 1.2007029 & CD4\\_T & snp\\_based   & FALSE & disease\\\\\n",
       "\t10 & 0.0496152300 & 1.632734 & 2.655485 & 0.9681380 & CD8\\_T & snp\\_based   & FALSE & disease\\\\\n",
       "\t11 & 0.8767768000 & 1.025278 & 1.905349 & 0.5121500 & Mono  & snp\\_based   & FALSE & disease\\\\\n",
       "\t12 & 0.4340997000 & 1.254892 & 2.250417 & 0.6577192 & NK    & snp\\_based   & FALSE & disease\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 7 × 8\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | estimate &lt;dbl&gt; | conf_upper &lt;dbl&gt; | conf_lower &lt;dbl&gt; | cell_type &lt;fct&gt; | type &lt;chr&gt; | excluded_genes_without_cs &lt;lgl&gt; | trait_filter &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 6 | 0.0006863434 | 1.939638 | 2.784903 | 1.3238773 | ALL   | eGene_based | FALSE | disease |\n",
       "| 7 | 0.0187006300 | 1.531563 | 2.180496 | 1.0527763 | ALL   | snp_based   | FALSE | disease |\n",
       "| 8 | 0.1422117000 | 1.598469 | 3.021836 | 0.7877197 | B     | snp_based   | FALSE | disease |\n",
       "| 9 | 0.0037729260 | 1.862487 | 2.812239 | 1.2007029 | CD4_T | snp_based   | FALSE | disease |\n",
       "| 10 | 0.0496152300 | 1.632734 | 2.655485 | 0.9681380 | CD8_T | snp_based   | FALSE | disease |\n",
       "| 11 | 0.8767768000 | 1.025278 | 1.905349 | 0.5121500 | Mono  | snp_based   | FALSE | disease |\n",
       "| 12 | 0.4340997000 | 1.254892 | 2.250417 | 0.6577192 | NK    | snp_based   | FALSE | disease |\n",
       "\n"
      ],
      "text/plain": [
       "   p_value      estimate conf_upper conf_lower cell_type type       \n",
       "6  0.0006863434 1.939638 2.784903   1.3238773  ALL       eGene_based\n",
       "7  0.0187006300 1.531563 2.180496   1.0527763  ALL       snp_based  \n",
       "8  0.1422117000 1.598469 3.021836   0.7877197  B         snp_based  \n",
       "9  0.0037729260 1.862487 2.812239   1.2007029  CD4_T     snp_based  \n",
       "10 0.0496152300 1.632734 2.655485   0.9681380  CD8_T     snp_based  \n",
       "11 0.8767768000 1.025278 1.905349   0.5121500  Mono      snp_based  \n",
       "12 0.4340997000 1.254892 2.250417   0.6577192  NK        snp_based  \n",
       "   excluded_genes_without_cs trait_filter\n",
       "6  FALSE                     disease     \n",
       "7  FALSE                     disease     \n",
       "8  FALSE                     disease     \n",
       "9  FALSE                     disease     \n",
       "10 FALSE                     disease     \n",
       "11 FALSE                     disease     \n",
       "12 FALSE                     disease     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "odds_ratio_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "f50c57f6-2806-4b90-89cc-75ce7f5da62c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(odds_ratio_plot) = NULL\n",
    "odds_ratio_plot$excluded_eGenes_without_credible_set = cs_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "8f4f0568-3098-429c-b640-b516711db2b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save with plot data for sharing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "24c99f29-cd9c-4558-ae49-90a9fac9fa57",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(odds_ratio_plot, paste0('figures/G2/G2_OR_Plot_reduced', use_gwas, remove_HLA, only_dt, only_esnps, coloc_thres,use_gwas , cs_snps ,'_data.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
