{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a1444178-491b-4528-9c7d-b171aa9267c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Script to calculate correlations for simulated counts from previous script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dca7889-aa95-41fc-84f2-fc38b8392196",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e9c0c86f-b63b-4c31-850c-97805d5b9368",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "\"incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f55b4c93-827f-4edf-b237-d399f128ac28",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "e8e76471-b038-407f-999a-1735bbb7a186",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to saved simulation results from previous script\n",
    "\n",
    "data_path_sim = \"../results/current/A1/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e7287b00-af61-4e6d-b410-8f1f10d97c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for simulated co-EQTL data (names of the datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ab72f079-74ff-40d1-b70f-0361d155eb13",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_counts = 'A1_4_Simulated_Counts'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "1c636d12-d4f5-4e89-af76-832c8eb88663",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_true_cor = 'A1_4_Simulation_Specified_Correlation'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a26159d1-096f-4f44-bd65-d090ad12124a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_gene_summary = 'A1_4_Gene_Summary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "387f2cd0-d4ac-4689-8bd5-45531dfc37bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_cell_summary = 'A1_4_Cell_Summary'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e5496af-a015-48c2-84d8-391677f35ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Simulated variant to load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "bfc7b8b5-fb34-49a8-bd0b-708b8e4b20b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant = 'wijstCD4_T4' # alternatives: scaling, no scaling, 'limited_scaling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "917352f9-6a66-4036-bf50-deb3b08d93bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_variant = 'scaling'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64663cd5-8d6a-454a-8d0c-ff44924cee9d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5dd6604-4d42-4bb0-81ed-9555d63ddcb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45cfaa89-153c-4abe-92f5-868c50e8090b",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6123ec7d-d87b-44df-b708-45126bb587db",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Define Links to individual datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "783625e4-edd1-406f-aa81-53efa05670d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load specified correlations\n",
    "cor_data = function( ){\n",
    "    path =  paste0(data_path_sim,  sim_data_name_true_cor,  variant, '_', coeqtl_variant, '.csv')\n",
    "    print(file.info(path)$mtime)\n",
    "    fread(path)\n",
    "    \n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "417a9965-2891-49bc-9b32-807906b701ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "bbd56fbb-eb4d-4554-abeb-54d6c9b02b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load specified counts\n",
    "count_data = function(){\n",
    "    path = paste0(data_path_sim,  sim_data_name_counts ,  variant, coeqtl_variant, '.csv.gz')\n",
    "    print(file.info(path)$mtime)\n",
    "    fread(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "1c882ea7-4edc-4a05-ab0e-d7f487d6676a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load  specified gene summary){\n",
    "get_gene_summary = function( ){\n",
    "    path = paste0(data_path_sim,  sim_data_name_gene_summary  , '_', variant, '_',  coeqtl_variant, '.csv.gz')\n",
    "    print(file.info(path)$mtime)\n",
    "    fread(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "30e4dcaf-9024-45dc-b464-c6664113eb96",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load  specified cell summary\n",
    "get_cell_summary = function(){\n",
    "    path = paste0(data_path_sim, sim_data_name_cell_summary, '_', variant, '_',  coeqtl_variant, '.csv.gz')\n",
    "    print(file.info(path)$mtime)\n",
    "    fread(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ced91b7d-8d86-496e-af7d-970c0d05a7ce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "18f87e5c-ce3d-4021-b2ce-b6d9068b6b81",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Get Specification of true correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "8ee45782-fbea-488a-b320-c2728d46c53d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the specified correlation for the simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "850b3a2a-04a9-486b-b1ee-a57bdc9e2816",
   "metadata": {},
   "outputs": [],
   "source": [
    "group1_data = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "0f533ad8-66d2-4f1b-b277-9578bc85bc33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-01-31 10:46:37 CET\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in set(x, j = name, value = value):\n",
      "\"Tried to assign NULL to column 'X', but this column does not exist to remove\"\n"
     ]
    }
   ],
   "source": [
    "group1_data = cor_data()\n",
    "group1_data$X = NULL\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "20b85921-d925-4926-8dc5-9014e537c19c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>gene1_mu</th><th scope=col>gene2_mu</th><th scope=col>gene1_size</th><th scope=col>gene2_size</th><th scope=col>genotype</th><th scope=col>correlation</th><th scope=col>gene_pair</th><th scope=col>ind</th><th scope=col>ident</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>20001</td><td>gene1_1</td><td>gene2_1</td><td>0.05645162</td><td>0.02941176</td><td>100</td><td>100</td><td>XX</td><td>0</td><td>gene1_1;gene2_1</td><td>1</td><td>XX;1</td></tr>\n",
       "\t<tr><td>20004</td><td>gene1_4</td><td>gene2_4</td><td>0.16272190</td><td>0.23465706</td><td>100</td><td>100</td><td>XX</td><td>0</td><td>gene1_4;gene2_4</td><td>1</td><td>XX;1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 12\n",
       "\\begin{tabular}{llllllllllll}\n",
       " V1 & gene1 & gene2 & gene1\\_mu & gene2\\_mu & gene1\\_size & gene2\\_size & genotype & correlation & gene\\_pair & ind & ident\\\\\n",
       " <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 20001 & gene1\\_1 & gene2\\_1 & 0.05645162 & 0.02941176 & 100 & 100 & XX & 0 & gene1\\_1;gene2\\_1 & 1 & XX;1\\\\\n",
       "\t 20004 & gene1\\_4 & gene2\\_4 & 0.16272190 & 0.23465706 & 100 & 100 & XX & 0 & gene1\\_4;gene2\\_4 & 1 & XX;1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 12\n",
       "\n",
       "| V1 &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | gene1_mu &lt;dbl&gt; | gene2_mu &lt;dbl&gt; | gene1_size &lt;dbl&gt; | gene2_size &lt;dbl&gt; | genotype &lt;chr&gt; | correlation &lt;dbl&gt; | gene_pair &lt;chr&gt; | ind &lt;int&gt; | ident &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 20001 | gene1_1 | gene2_1 | 0.05645162 | 0.02941176 | 100 | 100 | XX | 0 | gene1_1;gene2_1 | 1 | XX;1 |\n",
       "| 20004 | gene1_4 | gene2_4 | 0.16272190 | 0.23465706 | 100 | 100 | XX | 0 | gene1_4;gene2_4 | 1 | XX;1 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1    gene1   gene2   gene1_mu   gene2_mu   gene1_size gene2_size genotype\n",
       "1 20001 gene1_1 gene2_1 0.05645162 0.02941176 100        100        XX      \n",
       "2 20004 gene1_4 gene2_4 0.16272190 0.23465706 100        100        XX      \n",
       "  correlation gene_pair       ind ident\n",
       "1 0           gene1_1;gene2_1 1   XX;1 \n",
       "2 0           gene1_4;gene2_4 1   XX;1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(group1_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "671a6086-b1b8-4b57-88cf-8423257eb16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>0.05</li><li>0.1</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 0.05\n",
       "\\item 0.1\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 0.05\n",
       "3. 0.1\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.00 0.05 0.10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Different correlations in the data\n",
    "unique(group1_data$correlation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "3e08369e-7a8c-4c9e-b451-16326abaa32a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "450000"
      ],
      "text/latex": [
       "450000"
      ],
      "text/markdown": [
       "450000"
      ],
      "text/plain": [
       "[1] 450000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(group1_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4532b7bd-3c7e-4203-8b0c-eec804b52c62",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_pairs = unique(group1_data[,c('gene1', 'gene2')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "0504afad-7fe7-4622-8a4e-093f22da4f1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "10000"
      ],
      "text/latex": [
       "10000"
      ],
      "text/markdown": [
       "10000"
      ],
      "text/plain": [
       "[1] 10000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(correlation_pairs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cc69567-d751-4074-b13a-22857fc4a94a",
   "metadata": {},
   "source": [
    "# Calculate the correlations for the simulated count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "992169fe-94fe-4dc0-a839-f7dc89d8a7c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Empty data frame to save correlations\n",
    "calc_correlations_all = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e75202e9-d43c-4d93-9421-05525ec9cc2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "bfe8e3ec-ee43-45a1-b743-8d92348cf1a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-01-31 13:28:09 CET\"\n"
     ]
    }
   ],
   "source": [
    "count_data = count_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "faae657c-d208-4418-b46d-b4983971a4f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>cell_ident</th><th scope=col>value</th><th scope=col>scaling</th><th scope=col>mu</th><th scope=col>size</th><th scope=col>counts</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td>cell_XX;1_149_1 </td><td>0.9802071</td><td>0.9754055</td><td>0.05645162</td><td>100</td><td>1</td><td>0.05645162</td><td>149</td><td>XX;1</td></tr>\n",
       "\t<tr><td>2</td><td>gene1_1</td><td>cell_XX;1_149_39</td><td>0.3085414</td><td>0.7945828</td><td>0.05645162</td><td>100</td><td>0</td><td>0.05645162</td><td>149</td><td>XX;1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " V1 & variable & cell\\_ident & value & scaling & mu & size & counts & basemean & ncells & ind\\\\\n",
       " <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 & cell\\_XX;1\\_149\\_1  & 0.9802071 & 0.9754055 & 0.05645162 & 100 & 1 & 0.05645162 & 149 & XX;1\\\\\n",
       "\t 2 & gene1\\_1 & cell\\_XX;1\\_149\\_39 & 0.3085414 & 0.7945828 & 0.05645162 & 100 & 0 & 0.05645162 & 149 & XX;1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 11\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | cell_ident &lt;chr&gt; | value &lt;dbl&gt; | scaling &lt;dbl&gt; | mu &lt;dbl&gt; | size &lt;dbl&gt; | counts &lt;int&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 | cell_XX;1_149_1  | 0.9802071 | 0.9754055 | 0.05645162 | 100 | 1 | 0.05645162 | 149 | XX;1 |\n",
       "| 2 | gene1_1 | cell_XX;1_149_39 | 0.3085414 | 0.7945828 | 0.05645162 | 100 | 0 | 0.05645162 | 149 | XX;1 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable cell_ident       value     scaling   mu         size counts\n",
       "1 1  gene1_1  cell_XX;1_149_1  0.9802071 0.9754055 0.05645162 100  1     \n",
       "2 2  gene1_1  cell_XX;1_149_39 0.3085414 0.7945828 0.05645162 100  0     \n",
       "  basemean   ncells ind \n",
       "1 0.05645162 149    XX;1\n",
       "2 0.05645162 149    XX;1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(count_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "7166a90a-d73d-46be-b846-8cc3c852e817",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_data$V1 = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "f109ce9d-7848-454b-9e4c-3a9681521670",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get cell summary for merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "07d4adaf-93c0-4241-98a8-8a5ee42105c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-01-31 12:38:40 CET\"\n"
     ]
    }
   ],
   "source": [
    "cell_summary  =get_cell_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2eb9d767-a53c-45eb-858a-34286cc7276b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>cell_ident</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>var_counts</th><th scope=col>percentage_zero</th><th scope=col>total_mean_counts</th><th scope=col>scaling</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>149</td><td>XX;1</td><td>cell_XX;1_149_1 </td><td>7989</td><td>20000</td><td>17892</td><td>0.39945</td><td>74</td><td>0</td><td>7.101445</td><td>0.89460</td><td>0.5335534</td><td>1.335720</td></tr>\n",
       "\t<tr><td>2</td><td>149</td><td>XX;1</td><td>cell_XX;1_149_10</td><td>9841</td><td>20000</td><td>16807</td><td>0.49205</td><td>88</td><td>0</td><td>8.085241</td><td>0.84035</td><td>0.5335534</td><td>1.084348</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & ncells & ind & cell\\_ident & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & var\\_counts & percentage\\_zero & total\\_mean\\_counts & scaling\\\\\n",
       " <int> & <int> & <chr> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1 & 149 & XX;1 & cell\\_XX;1\\_149\\_1  & 7989 & 20000 & 17892 & 0.39945 & 74 & 0 & 7.101445 & 0.89460 & 0.5335534 & 1.335720\\\\\n",
       "\t 2 & 149 & XX;1 & cell\\_XX;1\\_149\\_10 & 9841 & 20000 & 16807 & 0.49205 & 88 & 0 & 8.085241 & 0.84035 & 0.5335534 & 1.084348\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | cell_ident &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | var_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | total_mean_counts &lt;dbl&gt; | scaling &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 149 | XX;1 | cell_XX;1_149_1  | 7989 | 20000 | 17892 | 0.39945 | 74 | 0 | 7.101445 | 0.89460 | 0.5335534 | 1.335720 |\n",
       "| 2 | 149 | XX;1 | cell_XX;1_149_10 | 9841 | 20000 | 16807 | 0.49205 | 88 | 0 | 8.085241 | 0.84035 | 0.5335534 | 1.084348 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 ncells ind  cell_ident       amount_counts amount_genes amount_zero_counts\n",
       "1 1  149    XX;1 cell_XX;1_149_1  7989          20000        17892             \n",
       "2 2  149    XX;1 cell_XX;1_149_10 9841          20000        16807             \n",
       "  mean_counts max_counts min_counts var_counts percentage_zero\n",
       "1 0.39945     74         0          7.101445   0.89460        \n",
       "2 0.49205     88         0          8.085241   0.84035        \n",
       "  total_mean_counts scaling \n",
       "1 0.5335534         1.335720\n",
       "2 0.5335534         1.084348"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "5eac6fe0-8978-4343-b4b2-0ff9cbe3abc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "amount_cells_per_ind = unique(cell_summary[,c('ncells', 'ind')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "48e45635-8ab5-434c-acb6-7998bb600699",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ncells</th><th scope=col>ind</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>149</td><td>XX;1</td></tr>\n",
       "\t<tr><td>281</td><td>XX;2</td></tr>\n",
       "\t<tr><td>271</td><td>XX;3</td></tr>\n",
       "\t<tr><td>511</td><td>XX;4</td></tr>\n",
       "\t<tr><td>231</td><td>XX;5</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 5 × 2\n",
       "\\begin{tabular}{ll}\n",
       " ncells & ind\\\\\n",
       " <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 149 & XX;1\\\\\n",
       "\t 281 & XX;2\\\\\n",
       "\t 271 & XX;3\\\\\n",
       "\t 511 & XX;4\\\\\n",
       "\t 231 & XX;5\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 5 × 2\n",
       "\n",
       "| ncells &lt;int&gt; | ind &lt;chr&gt; |\n",
       "|---|---|\n",
       "| 149 | XX;1 |\n",
       "| 281 | XX;2 |\n",
       "| 271 | XX;3 |\n",
       "| 511 | XX;4 |\n",
       "| 231 | XX;5 |\n",
       "\n"
      ],
      "text/plain": [
       "  ncells ind \n",
       "1 149    XX;1\n",
       "2 281    XX;2\n",
       "3 271    XX;3\n",
       "4 511    XX;4\n",
       "5 231    XX;5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(amount_cells_per_ind, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46f33cf3-a257-41e8-85cf-87f897836b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate different correlations on the simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e3cb2e9e-a2e7-4697-b2ff-4cabce7f66ba",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 14 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 2958: `gene1 = \"gene1_366\"`, `gene2 = \"gene2_366\"`, `ncells = 149`,\n",
      "  `ind = \"XX;1\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 13 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;2\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 8881: `gene1 = \"gene1_8991\"`, `gene2 = \"gene2_8991\"`, `ncells =\n",
      "  281`, `ind = \"XX;2\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;3\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 9926: `gene1 = \"gene1_9931\"`, `gene2 = \"gene2_9931\"`, `ncells =\n",
      "  271`, `ind = \"XX;3\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;4\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 8 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 3758: `gene1 = \"gene1_438\"`, `gene2 = \"gene2_438\"`, `ncells = 231`,\n",
      "  `ind = \"XX;5\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 7 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;6\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 200 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 12: `gene1 = \"gene1_1007\"`, `gene2 = \"gene2_1007\"`, `ncells = 91`,\n",
      "  `ind = \"XX;6\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 199 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;7\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 34 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 1544: `gene1 = \"gene1_2387\"`, `gene2 = \"gene2_2387\"`, `ncells =\n",
      "  148`, `ind = \"XX;8\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 33 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;10\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 7480: `gene1 = \"gene1_773\"`, `gene2 = \"gene2_773\"`, `ncells = 283`,\n",
      "  `ind = \"XX;10\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;11\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 78 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 195: `gene1 = \"gene1_1172\"`, `gene2 = \"gene2_1172\"`, `ncells = 101`,\n",
      "  `ind = \"XX;11\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 77 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;12\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;13\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 68 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 12: `gene1 = \"gene1_1007\"`, `gene2 = \"gene2_1007\"`, `ncells = 128`,\n",
      "  `ind = \"XX;13\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 67 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;14\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;15\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 4 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 3496: `gene1 = \"gene1_4143\"`, `gene2 = \"gene2_4143\"`, `ncells =\n",
      "  224`, `ind = \"XX;15\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 3 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;16\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 166 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 56: `gene1 = \"gene1_1047\"`, `gene2 = \"gene2_1047\"`, `ncells = 86`,\n",
      "  `ind = \"XX;16\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 165 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;17\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 48: `gene1 = \"gene1_104\"`, `gene2 = \"gene2_104\"`, `ncells = 270`,\n",
      "  `ind = \"XX;17\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;18\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;19\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XX;20\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;2\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 8 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 3957: `gene1 = \"gene1_4559\"`, `gene2 = \"gene2_4559\"`, `ncells =\n",
      "  194`, `ind = \"XY;2\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 7 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;3\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;4\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 9802 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 1: `gene1 = \"gene1_1\"`, `gene2 = \"gene2_1\"`, `ncells = 10`, `ind =\n",
      "  \"XY;4\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 9801 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;6\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 14 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 1308: `gene1 = \"gene1_2174\"`, `gene2 = \"gene2_2174\"`, `ncells =\n",
      "  185`, `ind = \"XY;6\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 13 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;7\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 6 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 56: `gene1 = \"gene1_1047\"`, `gene2 = \"gene2_1047\"`, `ncells = 226`,\n",
      "  `ind = \"XY;9\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 5 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;10\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 10 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 2715: `gene1 = \"gene1_3440\"`, `gene2 = \"gene2_3440\"`, `ncells =\n",
      "  194`, `ind = \"XY;10\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 9 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;11\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 8 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 1376: `gene1 = \"gene1_2235\"`, `gene2 = \"gene2_2235\"`, `ncells =\n",
      "  186`, `ind = \"XY;11\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 7 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;12\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;13\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 24 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 298: `gene1 = \"gene1_1265\"`, `gene2 = \"gene2_1265\"`, `ncells = 156`,\n",
      "  `ind = \"XY;13\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 23 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;14\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"XY;15\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;2\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 6 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 1376: `gene1 = \"gene1_2235\"`, `gene2 = \"gene2_2235\"`, `ncells =\n",
      "  197`, `ind = \"YY;2\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 5 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;3\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 4 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 6147: `gene1 = \"gene1_653\"`, `gene2 = \"gene2_653\"`, `ncells = 216`,\n",
      "  `ind = \"YY;3\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 3 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;4\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 50 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 1: `gene1 = \"gene1_1\"`, `gene2 = \"gene2_1\"`, `ncells = 135`, `ind =\n",
      "  \"YY;5\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 49 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;6\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;7\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 134 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 298: `gene1 = \"gene1_1265\"`, `gene2 = \"gene2_1265\"`, `ncells = 94`,\n",
      "  `ind = \"YY;7\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 133 remaining warnings.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 6295: `gene1 = \"gene1_6663\"`, `gene2 = \"gene2_6663\"`, `ncells =\n",
      "  299`, `ind = \"YY;8\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;9\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"YY;10\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mThere were 2 warnings in `summarise()`.\n",
      "The first warning was:\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In argument: `spearman_raw = cor(counts.x, counts.y, method = \"spearman\")`.\n",
      "\u001b[36mℹ\u001b[39m In group 7973: `gene1 = \"gene1_8173\"`, `gene2 = \"gene2_8173\"`, `ncells =\n",
      "  256`, `ind = \"YY;10\"`.\n",
      "Caused by warning in `stats::cor()`:\n",
      "\u001b[33m!\u001b[39m the standard deviation is zero\n",
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Run `dplyr::last_dplyr_warnings()` to see the 1 remaining warning.\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene1', 'gene2', 'ncells'. You can\n",
      "override using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "for(i in unique(count_data$ind)){\n",
    "    print(i)\n",
    "    flush.console()\n",
    "    ind = i\n",
    "    ncells = amount_cells_per_ind$ncells[amount_cells_per_ind$ind == ind]\n",
    "    # get countdata\n",
    "    countdata1 = count_data[count_data$ind == i,]\n",
    "    #countdata1$X = NULL\n",
    "    #countdata1$variable = str_replace(countdata1$variable, '__.*', '')\n",
    "    \n",
    "    ### Merge cor data\n",
    "    \n",
    "    cor_data = merge(correlation_pairs, countdata1, by.x = c('gene1'), by.y = c('variable'))\n",
    "    cor_data = merge(cor_data, countdata1, by.x = c('gene2', 'cell_ident', 'ncells', 'ind'), by.y = c('variable', 'cell_ident', 'ncells', 'ind'))\n",
    "    \n",
    "    ### Add cell_summary info\n",
    "    cor_data = merge(cor_data, cell_summary, by = c('ncells', 'ind', 'cell_ident') ) \n",
    "    \n",
    "    ## Calculate diff correlations\n",
    "\n",
    "    calc_correlations = cor_data  %>% group_by(gene1, gene2, ncells, ind) %>% summarise(\n",
    "\n",
    "        spearman_raw = cor(counts.x, counts.y, method = 'spearman'),\n",
    "\n",
    "        weighted_spearman = weightedCorr(counts.x * scaling, counts.y * scaling, method = 'Spearman', weights = 1/scaling, ML = FALSE) ,\n",
    "        \n",
    "        weighted_spearman_non_zero_counts = weightedCorr(counts.x * scaling, counts.y * scaling, method = 'Spearman', weights = (amount_counts - amount_zero_counts), ML = FALSE),\n",
    "        \n",
    "        weighted_spearman_non_zero_genes = weightedCorr(counts.x * scaling, counts.y * scaling, method = 'Spearman', weights = (amount_genes - amount_zero_counts), ML = FALSE),\n",
    "        \n",
    "        weighted_spearman_ranks = weightedCorr(dplyr::dense_rank(counts.x * scaling), dplyr::dense_rank(counts.y * scaling), method = 'Pearson', weights = 1/scaling, ML = FALSE) ,\n",
    "\n",
    "        pearson_raw = cor(counts.x, counts.y, method = 'pearson')  ,\n",
    "\n",
    "        weighted_pearson = weightedCorr(counts.x * scaling, counts.y * scaling, method = 'Pearson', weights =1/scaling, ML = FALSE),\n",
    "                                                  \n",
    "        weighted_pearson_non_zero_counts = weightedCorr(counts.x * scaling, counts.y * scaling, method = 'Pearson', weights = (amount_counts - amount_zero_counts), ML = FALSE),\n",
    "        \n",
    "        weighted_pearson_non_zero_genes = weightedCorr(counts.x * scaling, counts.y * scaling, method = 'Pearson', weights = (amount_genes - amount_zero_counts), ML = FALSE),\n",
    "\n",
    "        spearman =  cor(counts.x * scaling, counts.y * scaling, method = 'spearman'),\n",
    "        \n",
    "        scaled_spearman_ranks =  cor(dplyr::dense_rank(counts.x * scaling), dplyr::dense_rank(counts.y * scaling), method = 'pearson'),\n",
    "\n",
    "        pearson =  cor((counts.x * scaling), (counts.y *scaling), method = 'pearson') ,\n",
    "        \n",
    "\n",
    "    )\n",
    "    \n",
    "    ## Save calculated correlations\n",
    "    # V1 seperate per individual --> use if datasizes get to big\n",
    "    # write.csv(calc_correlations, paste0(output_path_sc,'coeQTL_sim/', 'A1_5_Calc_Cor', ind,'_', ncells, variant, 'TRUE', coeqtl_variant, '.csv'))\n",
    "    # V2: combined for all individuals\n",
    "    \n",
    "    calc_correlations_all = rbind(calc_correlations_all, calc_correlations) \n",
    "   \n",
    "}\n",
    "# write.csv(calc_correlations_all, paste0(output_path_sc,'coeQTL_sim/', 'A1_5_Calc_Cor', '_', ncells_mean ,'_', genotype, '_',  variant, coeqtl_variant, '.csv'))\n",
    " write.csv(calc_correlations_all, paste0(data_path_sim, 'A1_5_Calc_Cor_Ext',  '_',  variant, coeqtl_variant, '.csv'))   # include new versions to calculate correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af061835-d0b2-4e11-81bb-73f203d57c2a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc98e38c-b8ed-4d1f-bc54-bed89a88a927",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27e9a361-326b-4f0b-9348-e9da55b6bf47",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5eafc680-7f48-43da-ab3f-9fa4a92b268f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "315f8643-2c80-4001-b84d-3dd88778b365",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e359120-fea8-47f1-afba-cef86e5f70ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee28ca6d-a787-44c2-ad2c-d9431849ca80",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0a1f3cd-fca1-4ffd-976f-d7e37166b50c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abc8e118-6be1-40c9-b973-ea4c809636d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "723b4acf-3843-4dd0-896a-c3a36c196adf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a223cfed-1625-46a8-ab75-e62b31770dd6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f56d265-1d59-4476-8747-4927f220a610",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0e4d6fd-9d60-4801-93e1-2220f909ad40",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c0c2527-6cff-4f0d-b86e-86bc7f6b035e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
