{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1627,
   "id": "a1444178-491b-4528-9c7d-b171aa9267c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Script to estimate linear model on correlations from simulated data (co-eQTL effect estimation)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dca7889-aa95-41fc-84f2-fc38b8392196",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1628,
   "id": "24b28cc5-2ad8-42af-b06c-3dd3dfb31ba0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "\"incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f55b4c93-827f-4edf-b237-d399f128ac28",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1629,
   "id": "e7287b00-af61-4e6d-b410-8f1f10d97c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for co-EQTL data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1630,
   "id": "c324e253-0781-40e7-9460-dc350b4e517e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to saved simulation results from previous script\n",
    "data_path_sim = \"../results/current/A1/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f50d42e6-104f-4bcc-90e7-60357607ddfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for simulated co-EQTL data (names of the datasets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1631,
   "id": "ab72f079-74ff-40d1-b70f-0361d155eb13",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_counts = 'A1_4_Simulated_Counts'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1632,
   "id": "1c636d12-d4f5-4e89-af76-832c8eb88663",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_true_cor = 'A1_4_Simulation_Specified_Correlation'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1633,
   "id": "1784b9de-e160-45cb-bb76-778371e2a4e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_name = 'A1_5_Calc_Cor_Ext_'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63c8aea3-50e2-4074-a87c-07db20004e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Simulated variant to load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1634,
   "id": "bfc7b8b5-fb34-49a8-bd0b-708b8e4b20b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant = 'wijstB3' # alternatives: scaling, no scaling, 'limited_scaling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1635,
   "id": "917352f9-6a66-4036-bf50-deb3b08d93bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_variant = 'scaling'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1636,
   "id": "442ce2e9-4b6a-4221-bb6b-943f65b222a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_data_name_gene_summary = 'A1_4_Gene_Summary'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64663cd5-8d6a-454a-8d0c-ff44924cee9d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6aa5cc2-cc5c-4eaa-872b-5e56cc042cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45cfaa89-153c-4abe-92f5-868c50e8090b",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0a07384-1c2a-4225-baba-392acace940d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Define Links to individual datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1639,
   "id": "783625e4-edd1-406f-aa81-53efa05670d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load specified correlations\n",
    "cor_data = function(){\n",
    "    path = paste0(data_path_sim,  sim_data_name_true_cor,variant, '_', coeqtl_variant, '.csv')\n",
    "    print(file.info(path)$mtime)\n",
    "    read.csv(path)\n",
    "    \n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1640,
   "id": "bbd56fbb-eb4d-4554-abeb-54d6c9b02b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load  specified counts\n",
    "count_data = function(){\n",
    "    path = paste0(data_path_sim,  sim_data_name_counts ,  variant, coeqtl_variant, '.csv.gz')\n",
    "    print(file.info(path)$mtime)\n",
    "    fread(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1641,
   "id": "200c9fd1-ef09-44ce-81eb-0584707de172",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load calculated correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1642,
   "id": "22e79db9-283f-42ef-9310-eb0eed49be63",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_calc = function(){\n",
    "    path = paste0(data_path_sim,  cor_data_name , variant, coeqtl_variant, '.csv')\n",
    "    print(file.info(path)$mtime)\n",
    "    read.csv(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1643,
   "id": "3f9501ed-3822-4fdf-8dc3-d494336c099c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load gene summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1644,
   "id": "05213353-c2d3-4160-9929-13a5e80a5938",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function to load  specified gene summary){\n",
    "get_gene_summary = function( ){\n",
    "    path = paste0(data_path_sim,  sim_data_name_gene_summary  , '_', variant, '_',  coeqtl_variant, '.csv.gz')\n",
    "    print(file.info(path)$mtime)\n",
    "    fread(path)\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18f87e5c-ce3d-4021-b2ce-b6d9068b6b81",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get specification of simulation ( true values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1645,
   "id": "8ee45782-fbea-488a-b320-c2728d46c53d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the specification of the correlation and expression of the different gene-pairs before simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1646,
   "id": "850b3a2a-04a9-486b-b1ee-a57bdc9e2816",
   "metadata": {},
   "outputs": [],
   "source": [
    "group1_data = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1647,
   "id": "0f533ad8-66d2-4f1b-b277-9578bc85bc33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-01-22 10:19:06 CET\"\n"
     ]
    }
   ],
   "source": [
    "group1_data = cor_data()\n",
    "group1_data$X = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1648,
   "id": "f6925679-b348-409c-a6a4-f67ba2590d2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>gene1_mu</th><th scope=col>gene2_mu</th><th scope=col>gene1_size</th><th scope=col>gene2_size</th><th scope=col>genotype</th><th scope=col>correlation</th><th scope=col>gene_pair</th><th scope=col>ind</th><th scope=col>ident</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene1_1</td><td>gene2_1</td><td> 1</td><td>16</td><td>12</td><td>12</td><td>XX</td><td>0</td><td>gene1_1;gene2_1</td><td>1</td><td>XX;1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene1_4</td><td>gene2_4</td><td>24</td><td> 9</td><td>12</td><td>12</td><td>XX</td><td>0</td><td>gene1_4;gene2_4</td><td>1</td><td>XX;1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & gene1 & gene2 & gene1\\_mu & gene2\\_mu & gene1\\_size & gene2\\_size & genotype & correlation & gene\\_pair & ind & ident\\\\\n",
       "  & <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <chr> & <dbl> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & gene1\\_1 & gene2\\_1 &  1 & 16 & 12 & 12 & XX & 0 & gene1\\_1;gene2\\_1 & 1 & XX;1\\\\\n",
       "\t2 & gene1\\_4 & gene2\\_4 & 24 &  9 & 12 & 12 & XX & 0 & gene1\\_4;gene2\\_4 & 1 & XX;1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 11\n",
       "\n",
       "| <!--/--> | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | gene1_mu &lt;dbl&gt; | gene2_mu &lt;dbl&gt; | gene1_size &lt;int&gt; | gene2_size &lt;int&gt; | genotype &lt;chr&gt; | correlation &lt;dbl&gt; | gene_pair &lt;chr&gt; | ind &lt;int&gt; | ident &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 | gene2_1 |  1 | 16 | 12 | 12 | XX | 0 | gene1_1;gene2_1 | 1 | XX;1 |\n",
       "| 2 | gene1_4 | gene2_4 | 24 |  9 | 12 | 12 | XX | 0 | gene1_4;gene2_4 | 1 | XX;1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene1   gene2   gene1_mu gene2_mu gene1_size gene2_size genotype correlation\n",
       "1 gene1_1 gene2_1  1       16       12         12         XX       0          \n",
       "2 gene1_4 gene2_4 24        9       12         12         XX       0          \n",
       "  gene_pair       ind ident\n",
       "1 gene1_1;gene2_1 1   XX;1 \n",
       "2 gene1_4;gene2_4 1   XX;1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(group1_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1649,
   "id": "17cd53d4-ca7e-4f8d-a110-729df7549d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extract the unique gene-pairs that have been specified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1650,
   "id": "ad23901a-8faf-43c8-9f30-f2b66ddc5455",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlation_pairs = unique(group1_data[,c('gene1', 'gene2')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1651,
   "id": "6823d0ca-fdcc-4df7-adf5-d7be9b04d0d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene1_1</td><td>gene2_1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene1_4</td><td>gene2_4</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & gene1 & gene2\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & gene1\\_1 & gene2\\_1\\\\\n",
       "\t2 & gene1\\_4 & gene2\\_4\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 2\n",
       "\n",
       "| <!--/--> | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | gene1_1 | gene2_1 |\n",
       "| 2 | gene1_4 | gene2_4 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene1   gene2  \n",
       "1 gene1_1 gene2_1\n",
       "2 gene1_4 gene2_4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(correlation_pairs,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1652,
   "id": "61807b68-63dc-4915-a1f9-cf72e5c59345",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extract the correlation that has been specified for each gene-pair for each genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1653,
   "id": "e6dfa06a-895a-432e-86cf-9ecff1d6cafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "true_correlations = unique(group1_data[,c('gene1', 'gene2', 'correlation', 'genotype')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1654,
   "id": "a51370a9-08b9-4572-b0e0-eeffd02e9b60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "30000"
      ],
      "text/latex": [
       "30000"
      ],
      "text/markdown": [
       "30000"
      ],
      "text/plain": [
       "[1] 30000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(true_correlations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1655,
   "id": "0becea8e-df88-41bf-b6b8-e6da0607d5f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0</li><li>0.05</li><li>0.1</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0\n",
       "\\item 0.05\n",
       "\\item 0.1\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0\n",
       "2. 0.05\n",
       "3. 0.1\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 0.00 0.05 0.10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(group1_data$correlation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1656,
   "id": "3e08369e-7a8c-4c9e-b451-16326abaa32a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "450000"
      ],
      "text/latex": [
       "450000"
      ],
      "text/markdown": [
       "450000"
      ],
      "text/plain": [
       "[1] 450000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(group1_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "120aa72a-dcd6-453e-b0b0-dedf22ce9edf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "230cda59-0bd6-4ea9-b657-c0cc36b17c83",
   "metadata": {},
   "source": [
    "## Get calculated correlations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1657,
   "id": "d048ec85-28c5-4b93-93dc-660039a9bea7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get the correlations that have been calculated on the simulated counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1658,
   "id": "7aeece9a-176c-4f2c-a4c2-d94ddb57b947",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-01-22 11:01:14 CET\"\n"
     ]
    }
   ],
   "source": [
    "cor_data1 = cor_data_calc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1659,
   "id": "049ebf92-9cfa-482b-86f7-8d8f613f5b2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 17</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>spearman_raw</th><th scope=col>weighted_spearman</th><th scope=col>weighted_spearman_non_zero_counts</th><th scope=col>weighted_spearman_non_zero_genes</th><th scope=col>weighted_spearman_ranks</th><th scope=col>pearson_raw</th><th scope=col>weighted_pearson</th><th scope=col>weighted_pearson_non_zero_counts</th><th scope=col>weighted_pearson_non_zero_genes</th><th scope=col>spearman</th><th scope=col>scaled_spearman_ranks</th><th scope=col>pearson</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>gene1_1 </td><td>gene2_1 </td><td>10</td><td>XX;1</td><td>0.1376790</td><td>-0.02444693</td><td>-0.03085977</td><td>0.04700206</td><td>-0.02905645</td><td>0.08543577</td><td>-0.02495656</td><td>-0.03241479</td><td>0.04931058</td><td>0.05521888</td><td>0.05611031</td><td>0.05650416</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>gene1_10</td><td>gene2_10</td><td>10</td><td>XX;1</td><td>0.4699383</td><td> 0.42961983</td><td> 0.44380475</td><td>0.31078641</td><td> 0.37536750</td><td>0.58172802</td><td> 0.29048743</td><td> 0.29985118</td><td>0.21190062</td><td>0.29012943</td><td>0.29012943</td><td>0.19952510</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 17\n",
       "\\begin{tabular}{r|lllllllllllllllll}\n",
       "  & X & gene1 & gene2 & ncells & ind & spearman\\_raw & weighted\\_spearman & weighted\\_spearman\\_non\\_zero\\_counts & weighted\\_spearman\\_non\\_zero\\_genes & weighted\\_spearman\\_ranks & pearson\\_raw & weighted\\_pearson & weighted\\_pearson\\_non\\_zero\\_counts & weighted\\_pearson\\_non\\_zero\\_genes & spearman & scaled\\_spearman\\_ranks & pearson\\\\\n",
       "  & <int> & <chr> & <chr> & <int> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & gene1\\_1  & gene2\\_1  & 10 & XX;1 & 0.1376790 & -0.02444693 & -0.03085977 & 0.04700206 & -0.02905645 & 0.08543577 & -0.02495656 & -0.03241479 & 0.04931058 & 0.05521888 & 0.05611031 & 0.05650416\\\\\n",
       "\t2 & 2 & gene1\\_10 & gene2\\_10 & 10 & XX;1 & 0.4699383 &  0.42961983 &  0.44380475 & 0.31078641 &  0.37536750 & 0.58172802 &  0.29048743 &  0.29985118 & 0.21190062 & 0.29012943 & 0.29012943 & 0.19952510\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 17\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | spearman_raw &lt;dbl&gt; | weighted_spearman &lt;dbl&gt; | weighted_spearman_non_zero_counts &lt;dbl&gt; | weighted_spearman_non_zero_genes &lt;dbl&gt; | weighted_spearman_ranks &lt;dbl&gt; | pearson_raw &lt;dbl&gt; | weighted_pearson &lt;dbl&gt; | weighted_pearson_non_zero_counts &lt;dbl&gt; | weighted_pearson_non_zero_genes &lt;dbl&gt; | spearman &lt;dbl&gt; | scaled_spearman_ranks &lt;dbl&gt; | pearson &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | gene1_1  | gene2_1  | 10 | XX;1 | 0.1376790 | -0.02444693 | -0.03085977 | 0.04700206 | -0.02905645 | 0.08543577 | -0.02495656 | -0.03241479 | 0.04931058 | 0.05521888 | 0.05611031 | 0.05650416 |\n",
       "| 2 | 2 | gene1_10 | gene2_10 | 10 | XX;1 | 0.4699383 |  0.42961983 |  0.44380475 | 0.31078641 |  0.37536750 | 0.58172802 |  0.29048743 |  0.29985118 | 0.21190062 | 0.29012943 | 0.29012943 | 0.19952510 |\n",
       "\n"
      ],
      "text/plain": [
       "  X gene1    gene2    ncells ind  spearman_raw weighted_spearman\n",
       "1 1 gene1_1  gene2_1  10     XX;1 0.1376790    -0.02444693      \n",
       "2 2 gene1_10 gene2_10 10     XX;1 0.4699383     0.42961983      \n",
       "  weighted_spearman_non_zero_counts weighted_spearman_non_zero_genes\n",
       "1 -0.03085977                       0.04700206                      \n",
       "2  0.44380475                       0.31078641                      \n",
       "  weighted_spearman_ranks pearson_raw weighted_pearson\n",
       "1 -0.02905645             0.08543577  -0.02495656     \n",
       "2  0.37536750             0.58172802   0.29048743     \n",
       "  weighted_pearson_non_zero_counts weighted_pearson_non_zero_genes spearman  \n",
       "1 -0.03241479                      0.04931058                      0.05521888\n",
       "2  0.29985118                      0.21190062                      0.29012943\n",
       "  scaled_spearman_ranks pearson   \n",
       "1 0.05611031            0.05650416\n",
       "2 0.29012943            0.19952510"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1660,
   "id": "0519fe64-6145-4d6b-ade8-17b97096c31d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XX;1'</li><li>'XX;10'</li><li>'XX;11'</li><li>'XX;12'</li><li>'XX;13'</li><li>'XX;14'</li><li>'XX;15'</li><li>'XX;16'</li><li>'XX;17'</li><li>'XX;18'</li><li>'XX;19'</li><li>'XX;2'</li><li>'XX;20'</li><li>'XX;3'</li><li>'XX;4'</li><li>'XX;5'</li><li>'XX;6'</li><li>'XX;7'</li><li>'XX;8'</li><li>'XX;9'</li><li>'XY;1'</li><li>'XY;10'</li><li>'XY;11'</li><li>'XY;12'</li><li>'XY;13'</li><li>'XY;14'</li><li>'XY;15'</li><li>'XY;2'</li><li>'XY;3'</li><li>'XY;4'</li><li>'XY;5'</li><li>'XY;6'</li><li>'XY;7'</li><li>'XY;8'</li><li>'XY;9'</li><li>'YY;1'</li><li>'YY;10'</li><li>'YY;2'</li><li>'YY;3'</li><li>'YY;4'</li><li>'YY;5'</li><li>'YY;6'</li><li>'YY;7'</li><li>'YY;8'</li><li>'YY;9'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XX;1'\n",
       "\\item 'XX;10'\n",
       "\\item 'XX;11'\n",
       "\\item 'XX;12'\n",
       "\\item 'XX;13'\n",
       "\\item 'XX;14'\n",
       "\\item 'XX;15'\n",
       "\\item 'XX;16'\n",
       "\\item 'XX;17'\n",
       "\\item 'XX;18'\n",
       "\\item 'XX;19'\n",
       "\\item 'XX;2'\n",
       "\\item 'XX;20'\n",
       "\\item 'XX;3'\n",
       "\\item 'XX;4'\n",
       "\\item 'XX;5'\n",
       "\\item 'XX;6'\n",
       "\\item 'XX;7'\n",
       "\\item 'XX;8'\n",
       "\\item 'XX;9'\n",
       "\\item 'XY;1'\n",
       "\\item 'XY;10'\n",
       "\\item 'XY;11'\n",
       "\\item 'XY;12'\n",
       "\\item 'XY;13'\n",
       "\\item 'XY;14'\n",
       "\\item 'XY;15'\n",
       "\\item 'XY;2'\n",
       "\\item 'XY;3'\n",
       "\\item 'XY;4'\n",
       "\\item 'XY;5'\n",
       "\\item 'XY;6'\n",
       "\\item 'XY;7'\n",
       "\\item 'XY;8'\n",
       "\\item 'XY;9'\n",
       "\\item 'YY;1'\n",
       "\\item 'YY;10'\n",
       "\\item 'YY;2'\n",
       "\\item 'YY;3'\n",
       "\\item 'YY;4'\n",
       "\\item 'YY;5'\n",
       "\\item 'YY;6'\n",
       "\\item 'YY;7'\n",
       "\\item 'YY;8'\n",
       "\\item 'YY;9'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XX;1'\n",
       "2. 'XX;10'\n",
       "3. 'XX;11'\n",
       "4. 'XX;12'\n",
       "5. 'XX;13'\n",
       "6. 'XX;14'\n",
       "7. 'XX;15'\n",
       "8. 'XX;16'\n",
       "9. 'XX;17'\n",
       "10. 'XX;18'\n",
       "11. 'XX;19'\n",
       "12. 'XX;2'\n",
       "13. 'XX;20'\n",
       "14. 'XX;3'\n",
       "15. 'XX;4'\n",
       "16. 'XX;5'\n",
       "17. 'XX;6'\n",
       "18. 'XX;7'\n",
       "19. 'XX;8'\n",
       "20. 'XX;9'\n",
       "21. 'XY;1'\n",
       "22. 'XY;10'\n",
       "23. 'XY;11'\n",
       "24. 'XY;12'\n",
       "25. 'XY;13'\n",
       "26. 'XY;14'\n",
       "27. 'XY;15'\n",
       "28. 'XY;2'\n",
       "29. 'XY;3'\n",
       "30. 'XY;4'\n",
       "31. 'XY;5'\n",
       "32. 'XY;6'\n",
       "33. 'XY;7'\n",
       "34. 'XY;8'\n",
       "35. 'XY;9'\n",
       "36. 'YY;1'\n",
       "37. 'YY;10'\n",
       "38. 'YY;2'\n",
       "39. 'YY;3'\n",
       "40. 'YY;4'\n",
       "41. 'YY;5'\n",
       "42. 'YY;6'\n",
       "43. 'YY;7'\n",
       "44. 'YY;8'\n",
       "45. 'YY;9'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"XX;1\"  \"XX;10\" \"XX;11\" \"XX;12\" \"XX;13\" \"XX;14\" \"XX;15\" \"XX;16\" \"XX;17\"\n",
       "[10] \"XX;18\" \"XX;19\" \"XX;2\"  \"XX;20\" \"XX;3\"  \"XX;4\"  \"XX;5\"  \"XX;6\"  \"XX;7\" \n",
       "[19] \"XX;8\"  \"XX;9\"  \"XY;1\"  \"XY;10\" \"XY;11\" \"XY;12\" \"XY;13\" \"XY;14\" \"XY;15\"\n",
       "[28] \"XY;2\"  \"XY;3\"  \"XY;4\"  \"XY;5\"  \"XY;6\"  \"XY;7\"  \"XY;8\"  \"XY;9\"  \"YY;1\" \n",
       "[37] \"YY;10\" \"YY;2\"  \"YY;3\"  \"YY;4\"  \"YY;5\"  \"YY;6\"  \"YY;7\"  \"YY;8\"  \"YY;9\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sort(unique(cor_data1$ind))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8351b970-2b93-4fcf-afdc-63a640b2e578",
   "metadata": {},
   "source": [
    "## Get gene summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1661,
   "id": "575d538f-9bbd-4abf-9113-b9d267d3a205",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get the expression summary stats for all the genes per individual of the simulated counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1662,
   "id": "b97a7100-a53b-4f04-b251-d6da57bb7178",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"2025-01-22 10:42:20 CET\"\n"
     ]
    }
   ],
   "source": [
    "gene_summary1 = get_gene_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1663,
   "id": "c163593b-822b-42ff-8fce-e2e5ab650e71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>amount_zero_counts</th><th scope=col>amount_cells</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>percentage_zero_bin</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td> 1</td><td>10</td><td>XX;1</td><td>3</td><td>10</td><td> 1.0</td><td> 3</td><td>0</td><td> 0.8888889</td><td>30</td><td>(20,40]   </td></tr>\n",
       "\t<tr><td>2</td><td>gene2_1</td><td>16</td><td>10</td><td>XX;1</td><td>0</td><td>10</td><td>14.5</td><td>33</td><td>6</td><td>68.5000000</td><td> 0</td><td>(-0.001,5]</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " V1 & variable & basemean & ncells & ind & amount\\_zero\\_counts & amount\\_cells & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & percentage\\_zero\\_bin\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 &  1 & 10 & XX;1 & 3 & 10 &  1.0 &  3 & 0 &  0.8888889 & 30 & (20,40{]}   \\\\\n",
       "\t 2 & gene2\\_1 & 16 & 10 & XX;1 & 0 & 10 & 14.5 & 33 & 6 & 68.5000000 &  0 & (-0.001,5{]}\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 13\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | amount_zero_counts &lt;int&gt; | amount_cells &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | percentage_zero_bin &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 |  1 | 10 | XX;1 | 3 | 10 |  1.0 |  3 | 0 |  0.8888889 | 30 | (20,40]    |\n",
       "| 2 | gene2_1 | 16 | 10 | XX;1 | 0 | 10 | 14.5 | 33 | 6 | 68.5000000 |  0 | (-0.001,5] |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable basemean ncells ind  amount_zero_counts amount_cells mean_counts\n",
       "1 1  gene1_1   1       10     XX;1 3                  10            1.0       \n",
       "2 2  gene2_1  16       10     XX;1 0                  10           14.5       \n",
       "  max_counts min_counts variance_counts percentage_zero percentage_zero_bin\n",
       "1  3         0           0.8888889      30              (20,40]            \n",
       "2 33         6          68.5000000       0              (-0.001,5]         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_summary1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1664,
   "id": "34535bd4-13e2-4c7a-8b32-129ef778f885",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract the genotype information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1665,
   "id": "7a07d5d3-f8ea-4566-a806-7f03c698aa90",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_summary1$genotype = str_replace(gene_summary1$ind, ';.*','')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc3663db-47c4-4f95-a5fc-9f1ff37a517c",
   "metadata": {},
   "source": [
    "# Prepare Data for Linear Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1666,
   "id": "e9218fa4-d42a-40c4-b252-145bd95ff5ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Prepare data to estimate effect sizes on correlations with linear model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1667,
   "id": "13de44ee-b504-4c52-affd-e412c0e96952",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "450000"
      ],
      "text/latex": [
       "450000"
      ],
      "text/markdown": [
       "450000"
      ],
      "text/plain": [
       "[1] 450000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1668,
   "id": "0b5f8de0-650b-4f67-997b-035897b04670",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 17</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>spearman_raw</th><th scope=col>weighted_spearman</th><th scope=col>weighted_spearman_non_zero_counts</th><th scope=col>weighted_spearman_non_zero_genes</th><th scope=col>weighted_spearman_ranks</th><th scope=col>pearson_raw</th><th scope=col>weighted_pearson</th><th scope=col>weighted_pearson_non_zero_counts</th><th scope=col>weighted_pearson_non_zero_genes</th><th scope=col>spearman</th><th scope=col>scaled_spearman_ranks</th><th scope=col>pearson</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>gene1_1 </td><td>gene2_1 </td><td>10</td><td>XX;1</td><td>0.1376790</td><td>-0.02444693</td><td>-0.03085977</td><td>0.04700206</td><td>-0.02905645</td><td>0.08543577</td><td>-0.02495656</td><td>-0.03241479</td><td>0.04931058</td><td>0.05521888</td><td>0.05611031</td><td>0.05650416</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>gene1_10</td><td>gene2_10</td><td>10</td><td>XX;1</td><td>0.4699383</td><td> 0.42961983</td><td> 0.44380475</td><td>0.31078641</td><td> 0.37536750</td><td>0.58172802</td><td> 0.29048743</td><td> 0.29985118</td><td>0.21190062</td><td>0.29012943</td><td>0.29012943</td><td>0.19952510</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 17\n",
       "\\begin{tabular}{r|lllllllllllllllll}\n",
       "  & X & gene1 & gene2 & ncells & ind & spearman\\_raw & weighted\\_spearman & weighted\\_spearman\\_non\\_zero\\_counts & weighted\\_spearman\\_non\\_zero\\_genes & weighted\\_spearman\\_ranks & pearson\\_raw & weighted\\_pearson & weighted\\_pearson\\_non\\_zero\\_counts & weighted\\_pearson\\_non\\_zero\\_genes & spearman & scaled\\_spearman\\_ranks & pearson\\\\\n",
       "  & <int> & <chr> & <chr> & <int> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & gene1\\_1  & gene2\\_1  & 10 & XX;1 & 0.1376790 & -0.02444693 & -0.03085977 & 0.04700206 & -0.02905645 & 0.08543577 & -0.02495656 & -0.03241479 & 0.04931058 & 0.05521888 & 0.05611031 & 0.05650416\\\\\n",
       "\t2 & 2 & gene1\\_10 & gene2\\_10 & 10 & XX;1 & 0.4699383 &  0.42961983 &  0.44380475 & 0.31078641 &  0.37536750 & 0.58172802 &  0.29048743 &  0.29985118 & 0.21190062 & 0.29012943 & 0.29012943 & 0.19952510\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 17\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | spearman_raw &lt;dbl&gt; | weighted_spearman &lt;dbl&gt; | weighted_spearman_non_zero_counts &lt;dbl&gt; | weighted_spearman_non_zero_genes &lt;dbl&gt; | weighted_spearman_ranks &lt;dbl&gt; | pearson_raw &lt;dbl&gt; | weighted_pearson &lt;dbl&gt; | weighted_pearson_non_zero_counts &lt;dbl&gt; | weighted_pearson_non_zero_genes &lt;dbl&gt; | spearman &lt;dbl&gt; | scaled_spearman_ranks &lt;dbl&gt; | pearson &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | gene1_1  | gene2_1  | 10 | XX;1 | 0.1376790 | -0.02444693 | -0.03085977 | 0.04700206 | -0.02905645 | 0.08543577 | -0.02495656 | -0.03241479 | 0.04931058 | 0.05521888 | 0.05611031 | 0.05650416 |\n",
       "| 2 | 2 | gene1_10 | gene2_10 | 10 | XX;1 | 0.4699383 |  0.42961983 |  0.44380475 | 0.31078641 |  0.37536750 | 0.58172802 |  0.29048743 |  0.29985118 | 0.21190062 | 0.29012943 | 0.29012943 | 0.19952510 |\n",
       "\n"
      ],
      "text/plain": [
       "  X gene1    gene2    ncells ind  spearman_raw weighted_spearman\n",
       "1 1 gene1_1  gene2_1  10     XX;1 0.1376790    -0.02444693      \n",
       "2 2 gene1_10 gene2_10 10     XX;1 0.4699383     0.42961983      \n",
       "  weighted_spearman_non_zero_counts weighted_spearman_non_zero_genes\n",
       "1 -0.03085977                       0.04700206                      \n",
       "2  0.44380475                       0.31078641                      \n",
       "  weighted_spearman_ranks pearson_raw weighted_pearson\n",
       "1 -0.02905645             0.08543577  -0.02495656     \n",
       "2  0.37536750             0.58172802   0.29048743     \n",
       "  weighted_pearson_non_zero_counts weighted_pearson_non_zero_genes spearman  \n",
       "1 -0.03241479                      0.04931058                      0.05521888\n",
       "2  0.29985118                      0.21190062                      0.29012943\n",
       "  scaled_spearman_ranks pearson   \n",
       "1 0.05611031            0.05650416\n",
       "2 0.29012943            0.19952510"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1669,
   "id": "856c4184-ddad-47b8-a8b0-10cf2836e976",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data1$X = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1670,
   "id": "c5a80feb-39ae-4f80-85ed-e86f3faccb9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extract genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1671,
   "id": "cd9dd133-128b-4e91-8e9c-206f26dec37d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data1$genotype = str_replace(cor_data1$ind, ';.*', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1672,
   "id": "bf625e9c-5671-4dd0-9c8b-85f4b608eb8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Transform to long format per correlation measure\n",
    "cor_data1_long = melt(cor_data1, id = c('gene1', 'gene2', 'ncells', 'ind', 'genotype'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1673,
   "id": "491b9a3c-8bf4-41e3-85fe-7335e97cecca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>spearman_raw</li><li>weighted_spearman</li><li>weighted_spearman_non_zero_counts</li><li>weighted_spearman_non_zero_genes</li><li>weighted_spearman_ranks</li><li>pearson_raw</li><li>weighted_pearson</li><li>weighted_pearson_non_zero_counts</li><li>weighted_pearson_non_zero_genes</li><li>spearman</li><li>scaled_spearman_ranks</li><li>pearson</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'spearman_raw'</li><li>'weighted_spearman'</li><li>'weighted_spearman_non_zero_counts'</li><li>'weighted_spearman_non_zero_genes'</li><li>'weighted_spearman_ranks'</li><li>'pearson_raw'</li><li>'weighted_pearson'</li><li>'weighted_pearson_non_zero_counts'</li><li>'weighted_pearson_non_zero_genes'</li><li>'spearman'</li><li>'scaled_spearman_ranks'</li><li>'pearson'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item spearman\\_raw\n",
       "\\item weighted\\_spearman\n",
       "\\item weighted\\_spearman\\_non\\_zero\\_counts\n",
       "\\item weighted\\_spearman\\_non\\_zero\\_genes\n",
       "\\item weighted\\_spearman\\_ranks\n",
       "\\item pearson\\_raw\n",
       "\\item weighted\\_pearson\n",
       "\\item weighted\\_pearson\\_non\\_zero\\_counts\n",
       "\\item weighted\\_pearson\\_non\\_zero\\_genes\n",
       "\\item spearman\n",
       "\\item scaled\\_spearman\\_ranks\n",
       "\\item pearson\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'spearman\\_raw'\n",
       "\\item 'weighted\\_spearman'\n",
       "\\item 'weighted\\_spearman\\_non\\_zero\\_counts'\n",
       "\\item 'weighted\\_spearman\\_non\\_zero\\_genes'\n",
       "\\item 'weighted\\_spearman\\_ranks'\n",
       "\\item 'pearson\\_raw'\n",
       "\\item 'weighted\\_pearson'\n",
       "\\item 'weighted\\_pearson\\_non\\_zero\\_counts'\n",
       "\\item 'weighted\\_pearson\\_non\\_zero\\_genes'\n",
       "\\item 'spearman'\n",
       "\\item 'scaled\\_spearman\\_ranks'\n",
       "\\item 'pearson'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. spearman_raw\n",
       "2. weighted_spearman\n",
       "3. weighted_spearman_non_zero_counts\n",
       "4. weighted_spearman_non_zero_genes\n",
       "5. weighted_spearman_ranks\n",
       "6. pearson_raw\n",
       "7. weighted_pearson\n",
       "8. weighted_pearson_non_zero_counts\n",
       "9. weighted_pearson_non_zero_genes\n",
       "10. spearman\n",
       "11. scaled_spearman_ranks\n",
       "12. pearson\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'spearman_raw'\n",
       "2. 'weighted_spearman'\n",
       "3. 'weighted_spearman_non_zero_counts'\n",
       "4. 'weighted_spearman_non_zero_genes'\n",
       "5. 'weighted_spearman_ranks'\n",
       "6. 'pearson_raw'\n",
       "7. 'weighted_pearson'\n",
       "8. 'weighted_pearson_non_zero_counts'\n",
       "9. 'weighted_pearson_non_zero_genes'\n",
       "10. 'spearman'\n",
       "11. 'scaled_spearman_ranks'\n",
       "12. 'pearson'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] spearman_raw                      weighted_spearman                \n",
       " [3] weighted_spearman_non_zero_counts weighted_spearman_non_zero_genes \n",
       " [5] weighted_spearman_ranks           pearson_raw                      \n",
       " [7] weighted_pearson                  weighted_pearson_non_zero_counts \n",
       " [9] weighted_pearson_non_zero_genes   spearman                         \n",
       "[11] scaled_spearman_ranks             pearson                          \n",
       "12 Levels: spearman_raw weighted_spearman ... pearson"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Available correlation measures\n",
    "unique(cor_data1_long$variable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1674,
   "id": "5a94ea4b-93b7-4ded-b6aa-4b59e70cf125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XX'</li><li>'XY'</li><li>'YY'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XX'\n",
       "\\item 'XY'\n",
       "\\item 'YY'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XX'\n",
       "2. 'XY'\n",
       "3. 'YY'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"XX\" \"XY\" \"YY\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(cor_data1$genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1675,
   "id": "b57f943d-a942-47d2-8338-57c919259dd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add gene summary statistics to the correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1676,
   "id": "542906ef-dcbb-4048-a25b-71bc175ba530",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>genotype</th><th scope=col>variable</th><th scope=col>value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene1_1 </td><td>gene2_1 </td><td>10</td><td>XX;1</td><td>XX</td><td>spearman_raw</td><td>0.1376790</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene1_10</td><td>gene2_10</td><td>10</td><td>XX;1</td><td>XX</td><td>spearman_raw</td><td>0.4699383</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & gene1 & gene2 & ncells & ind & genotype & variable & value\\\\\n",
       "  & <chr> & <chr> & <int> & <chr> & <chr> & <fct> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene1\\_1  & gene2\\_1  & 10 & XX;1 & XX & spearman\\_raw & 0.1376790\\\\\n",
       "\t2 & gene1\\_10 & gene2\\_10 & 10 & XX;1 & XX & spearman\\_raw & 0.4699383\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 7\n",
       "\n",
       "| <!--/--> | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | genotype &lt;chr&gt; | variable &lt;fct&gt; | value &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1  | gene2_1  | 10 | XX;1 | XX | spearman_raw | 0.1376790 |\n",
       "| 2 | gene1_10 | gene2_10 | 10 | XX;1 | XX | spearman_raw | 0.4699383 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene1    gene2    ncells ind  genotype variable     value    \n",
       "1 gene1_1  gene2_1  10     XX;1 XX       spearman_raw 0.1376790\n",
       "2 gene1_10 gene2_10 10     XX;1 XX       spearman_raw 0.4699383"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data1_long,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1677,
   "id": "3dafc001-b574-4d85-a623-bf64db02c57d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5400000"
      ],
      "text/latex": [
       "5400000"
      ],
      "text/markdown": [
       "5400000"
      ],
      "text/plain": [
       "[1] 5400000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data1_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1678,
   "id": "6e75f597-806a-41c6-b601-2e88afde544b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>amount_zero_counts</th><th scope=col>amount_cells</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>percentage_zero_bin</th><th scope=col>genotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td> 1</td><td>10</td><td>XX;1</td><td>3</td><td>10</td><td> 1.0</td><td> 3</td><td>0</td><td> 0.8888889</td><td>30</td><td>(20,40]   </td><td>XX</td></tr>\n",
       "\t<tr><td>2</td><td>gene2_1</td><td>16</td><td>10</td><td>XX;1</td><td>0</td><td>10</td><td>14.5</td><td>33</td><td>6</td><td>68.5000000</td><td> 0</td><td>(-0.001,5]</td><td>XX</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & variable & basemean & ncells & ind & amount\\_zero\\_counts & amount\\_cells & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & percentage\\_zero\\_bin & genotype\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 &  1 & 10 & XX;1 & 3 & 10 &  1.0 &  3 & 0 &  0.8888889 & 30 & (20,40{]}    & XX\\\\\n",
       "\t 2 & gene2\\_1 & 16 & 10 & XX;1 & 0 & 10 & 14.5 & 33 & 6 & 68.5000000 &  0 & (-0.001,5{]} & XX\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | amount_zero_counts &lt;int&gt; | amount_cells &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | percentage_zero_bin &lt;chr&gt; | genotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 |  1 | 10 | XX;1 | 3 | 10 |  1.0 |  3 | 0 |  0.8888889 | 30 | (20,40]    | XX |\n",
       "| 2 | gene2_1 | 16 | 10 | XX;1 | 0 | 10 | 14.5 | 33 | 6 | 68.5000000 |  0 | (-0.001,5] | XX |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable basemean ncells ind  amount_zero_counts amount_cells mean_counts\n",
       "1 1  gene1_1   1       10     XX;1 3                  10            1.0       \n",
       "2 2  gene2_1  16       10     XX;1 0                  10           14.5       \n",
       "  max_counts min_counts variance_counts percentage_zero percentage_zero_bin\n",
       "1  3         0           0.8888889      30              (20,40]            \n",
       "2 33         6          68.5000000       0              (-0.001,5]         \n",
       "  genotype\n",
       "1 XX      \n",
       "2 XX      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_summary1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1679,
   "id": "aa0119a6-f18a-40b4-b105-3c43d620660f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add summary stats for gene 1\n",
    "cor_data1_long = merge(cor_data1_long, gene_summary1[,c('basemean', 'variable', 'ncells', 'ind', 'amount_zero_counts', 'mean_counts', 'percentage_zero', 'genotype')],\n",
    "                       by.x = c('gene1', 'ncells', 'ind', 'genotype'), by.y = c('variable', 'ncells', 'ind' ,'genotype'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1680,
   "id": "b3257405-bd2e-4b21-9562-9ad65bab30a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5400000"
      ],
      "text/latex": [
       "5400000"
      ],
      "text/markdown": [
       "5400000"
      ],
      "text/plain": [
       "[1] 5400000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data1_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1681,
   "id": "91f669dd-2dbd-423b-9568-5cec9149deef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add summary stats for gene 2\n",
    "cor_data1_long = merge(cor_data1_long, gene_summary1[,c('basemean', 'variable', 'ncells', 'ind', 'amount_zero_counts', 'mean_counts', 'percentage_zero', 'genotype')],\n",
    "                       by.x = c('gene2', 'ncells', 'ind', 'genotype'), by.y = c('variable', 'ncells', 'ind', 'genotype'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1682,
   "id": "694bbf89-bf3b-4a45-8b90-cd70138f7c39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5400000"
      ],
      "text/latex": [
       "5400000"
      ],
      "text/markdown": [
       "5400000"
      ],
      "text/plain": [
       "[1] 5400000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data1_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1683,
   "id": "35312c5c-df76-4aa2-ad97-5f9ad22a44b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>genotype</th><th scope=col>gene1</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>basemean.x</th><th scope=col>amount_zero_counts.x</th><th scope=col>mean_counts.x</th><th scope=col>percentage_zero.x</th><th scope=col>basemean.y</th><th scope=col>amount_zero_counts.y</th><th scope=col>mean_counts.y</th><th scope=col>percentage_zero.y</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene2_1</td><td>10</td><td>XX;1</td><td>XX</td><td>gene1_1</td><td>spearman_raw                    </td><td>0.13767902</td><td>1</td><td>3</td><td>1</td><td>30</td><td>16</td><td>0</td><td>14.5</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene2_1</td><td>10</td><td>XX;1</td><td>XX</td><td>gene1_1</td><td>weighted_spearman_non_zero_genes</td><td>0.04700206</td><td>1</td><td>3</td><td>1</td><td>30</td><td>16</td><td>0</td><td>14.5</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & gene2 & ncells & ind & genotype & gene1 & variable & value & basemean.x & amount\\_zero\\_counts.x & mean\\_counts.x & percentage\\_zero.x & basemean.y & amount\\_zero\\_counts.y & mean\\_counts.y & percentage\\_zero.y\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <fct> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene2\\_1 & 10 & XX;1 & XX & gene1\\_1 & spearman\\_raw                     & 0.13767902 & 1 & 3 & 1 & 30 & 16 & 0 & 14.5 & 0\\\\\n",
       "\t2 & gene2\\_1 & 10 & XX;1 & XX & gene1\\_1 & weighted\\_spearman\\_non\\_zero\\_genes & 0.04700206 & 1 & 3 & 1 & 30 & 16 & 0 & 14.5 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 15\n",
       "\n",
       "| <!--/--> | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | genotype &lt;chr&gt; | gene1 &lt;chr&gt; | variable &lt;fct&gt; | value &lt;dbl&gt; | basemean.x &lt;dbl&gt; | amount_zero_counts.x &lt;int&gt; | mean_counts.x &lt;dbl&gt; | percentage_zero.x &lt;dbl&gt; | basemean.y &lt;dbl&gt; | amount_zero_counts.y &lt;int&gt; | mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene2_1 | 10 | XX;1 | XX | gene1_1 | spearman_raw                     | 0.13767902 | 1 | 3 | 1 | 30 | 16 | 0 | 14.5 | 0 |\n",
       "| 2 | gene2_1 | 10 | XX;1 | XX | gene1_1 | weighted_spearman_non_zero_genes | 0.04700206 | 1 | 3 | 1 | 30 | 16 | 0 | 14.5 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene2   ncells ind  genotype gene1   variable                        \n",
       "1 gene2_1 10     XX;1 XX       gene1_1 spearman_raw                    \n",
       "2 gene2_1 10     XX;1 XX       gene1_1 weighted_spearman_non_zero_genes\n",
       "  value      basemean.x amount_zero_counts.x mean_counts.x percentage_zero.x\n",
       "1 0.13767902 1          3                    1             30               \n",
       "2 0.04700206 1          3                    1             30               \n",
       "  basemean.y amount_zero_counts.y mean_counts.y percentage_zero.y\n",
       "1 16         0                    14.5          0                \n",
       "2 16         0                    14.5          0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data1_long,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1684,
   "id": "f88b04ba-173b-4742-be01-3b2290cb2ec7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Rename dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1685,
   "id": "c89877f2-237e-4273-bfb7-1ab6c357eacb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cor_data_long = cor_data1_long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1686,
   "id": "be4b03ad-e965-49b7-810e-aec35f32ff2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5400000"
      ],
      "text/latex": [
       "5400000"
      ],
      "text/markdown": [
       "5400000"
      ],
      "text/plain": [
       "[1] 5400000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b89b34b-10b9-4827-9a25-0ac352c5820b",
   "metadata": {},
   "source": [
    "# co-eQTL Testing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b56de401-2da7-469e-9ed4-bfb4ff58ce32",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Linear Model on correlation coeficient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1687,
   "id": "bd9ccb6a-e40c-4697-a9bc-8704284fd7c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Estimate effect sizes for simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1688,
   "id": "6551a2a1-1729-4dc2-99ca-5ed1e847a7c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>genotype</th><th scope=col>gene1</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>basemean.x</th><th scope=col>amount_zero_counts.x</th><th scope=col>mean_counts.x</th><th scope=col>percentage_zero.x</th><th scope=col>basemean.y</th><th scope=col>amount_zero_counts.y</th><th scope=col>mean_counts.y</th><th scope=col>percentage_zero.y</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene2_1</td><td>10</td><td>XX;1</td><td>XX</td><td>gene1_1</td><td>spearman_raw                    </td><td>0.13767902</td><td>1</td><td>3</td><td>1</td><td>30</td><td>16</td><td>0</td><td>14.5</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene2_1</td><td>10</td><td>XX;1</td><td>XX</td><td>gene1_1</td><td>weighted_spearman_non_zero_genes</td><td>0.04700206</td><td>1</td><td>3</td><td>1</td><td>30</td><td>16</td><td>0</td><td>14.5</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & gene2 & ncells & ind & genotype & gene1 & variable & value & basemean.x & amount\\_zero\\_counts.x & mean\\_counts.x & percentage\\_zero.x & basemean.y & amount\\_zero\\_counts.y & mean\\_counts.y & percentage\\_zero.y\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <fct> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene2\\_1 & 10 & XX;1 & XX & gene1\\_1 & spearman\\_raw                     & 0.13767902 & 1 & 3 & 1 & 30 & 16 & 0 & 14.5 & 0\\\\\n",
       "\t2 & gene2\\_1 & 10 & XX;1 & XX & gene1\\_1 & weighted\\_spearman\\_non\\_zero\\_genes & 0.04700206 & 1 & 3 & 1 & 30 & 16 & 0 & 14.5 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 15\n",
       "\n",
       "| <!--/--> | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | genotype &lt;chr&gt; | gene1 &lt;chr&gt; | variable &lt;fct&gt; | value &lt;dbl&gt; | basemean.x &lt;dbl&gt; | amount_zero_counts.x &lt;int&gt; | mean_counts.x &lt;dbl&gt; | percentage_zero.x &lt;dbl&gt; | basemean.y &lt;dbl&gt; | amount_zero_counts.y &lt;int&gt; | mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene2_1 | 10 | XX;1 | XX | gene1_1 | spearman_raw                     | 0.13767902 | 1 | 3 | 1 | 30 | 16 | 0 | 14.5 | 0 |\n",
       "| 2 | gene2_1 | 10 | XX;1 | XX | gene1_1 | weighted_spearman_non_zero_genes | 0.04700206 | 1 | 3 | 1 | 30 | 16 | 0 | 14.5 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene2   ncells ind  genotype gene1   variable                        \n",
       "1 gene2_1 10     XX;1 XX       gene1_1 spearman_raw                    \n",
       "2 gene2_1 10     XX;1 XX       gene1_1 weighted_spearman_non_zero_genes\n",
       "  value      basemean.x amount_zero_counts.x mean_counts.x percentage_zero.x\n",
       "1 0.13767902 1          3                    1             30               \n",
       "2 0.04700206 1          3                    1             30               \n",
       "  basemean.y amount_zero_counts.y mean_counts.y percentage_zero.y\n",
       "1 16         0                    14.5          0                \n",
       "2 16         0                    14.5          0                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data_long,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1689,
   "id": "63cb4613-e55b-4850-8aa6-b9f50653d5c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate the amount of non-zero counts per gene (gene1 and gene2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1690,
   "id": "bd59ae2e-30a0-4218-b558-255e788a6f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long$non_zero_counts.x = cor_data_long$ncells - cor_data_long$amount_zero_counts.x\n",
    "cor_data_long$non_zero_counts.y = cor_data_long$ncells - cor_data_long$amount_zero_counts.y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1691,
   "id": "b6e576aa-1bc5-4737-a502-1bde2aecde03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "min(cor_data_long$non_zero_counts.x)\n",
    "min(cor_data_long$non_zero_counts.y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1692,
   "id": "97fe12d3-6d2e-4a3f-aa23-e9cf23754c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "## OPTIONAL: Filtering on at least 10 non-zero counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1693,
   "id": "cc165409-0e39-4f46-8103-75363947cd1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "5400000"
      ],
      "text/latex": [
       "5400000"
      ],
      "text/markdown": [
       "5400000"
      ],
      "text/plain": [
       "[1] 5400000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1694,
   "id": "68070e7e-acdb-4284-83d8-78cdd79b76d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_zero_counts_filter = 10\n",
    "cor_data_long = cor_data_long[(cor_data_long$non_zero_counts.x  > non_zero_counts_filter) & (cor_data_long$non_zero_counts.y  > non_zero_counts_filter),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1695,
   "id": "5abae737-a4ba-47be-ac84-cc703d8e7eb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1292964"
      ],
      "text/latex": [
       "1292964"
      ],
      "text/markdown": [
       "1292964"
      ],
      "text/plain": [
       "[1] 1292964"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1696,
   "id": "ab04a2d5-2854-4b0d-b4ed-9fef2c25994c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get for each individuum genotype and amount of cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1697,
   "id": "ef2fba2d-ccc0-41e1-b894-fabd09e87e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ind_genotypes = unique(cor_data_long[,c('ind', 'genotype', 'ncells')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1698,
   "id": "3a3ada29-5243-47e2-b7d9-c230926fd7af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ind</th><th scope=col>genotype</th><th scope=col>ncells</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>385</th><td>XX;5</td><td>XX</td><td>20</td></tr>\n",
       "\t<tr><th scope=row>397</th><td>XY;3</td><td>XY</td><td>20</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & ind & genotype & ncells\\\\\n",
       "  & <chr> & <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t385 & XX;5 & XX & 20\\\\\n",
       "\t397 & XY;3 & XY & 20\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 3\n",
       "\n",
       "| <!--/--> | ind &lt;chr&gt; | genotype &lt;chr&gt; | ncells &lt;int&gt; |\n",
       "|---|---|---|---|\n",
       "| 385 | XX;5 | XX | 20 |\n",
       "| 397 | XY;3 | XY | 20 |\n",
       "\n"
      ],
      "text/plain": [
       "    ind  genotype ncells\n",
       "385 XX;5 XX       20    \n",
       "397 XY;3 XY       20    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(ind_genotypes,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1699,
   "id": "949ab348-c0a7-4c63-984b-28b0f4a2dd06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "29"
      ],
      "text/latex": [
       "29"
      ],
      "text/markdown": [
       "29"
      ],
      "text/plain": [
       "[1] 29"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(ind_genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1700,
   "id": "c7b72948-d4bc-484a-bfa8-c1adc2b88f26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XX'</li><li>'XY'</li><li>'YY'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XX'\n",
       "\\item 'XY'\n",
       "\\item 'YY'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XX'\n",
       "2. 'XY'\n",
       "3. 'YY'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"XX\" \"XY\" \"YY\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(ind_genotypes$genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1701,
   "id": "7357a2fb-3160-48b5-a1bf-73b0079d60d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XX;5'</li><li>'XY;3'</li><li>'YY;6'</li><li>'XX;12'</li><li>'XX;9'</li><li>'XY;13'</li><li>'XX;17'</li><li>'XX;20'</li><li>'XX;6'</li><li>'XX;4'</li><li>'XY;6'</li><li>'YY;9'</li><li>'YY;7'</li><li>'YY;4'</li><li>'XY;12'</li><li>'XY;8'</li><li>'YY;2'</li><li>'YY;1'</li><li>'YY;3'</li><li>'XX;14'</li><li>'XX;8'</li><li>'XX;10'</li><li>'XY;4'</li><li>'XY;10'</li><li>'XY;14'</li><li>'XY;7'</li><li>'XY;11'</li><li>'XY;9'</li><li>'XY;15'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XX;5'\n",
       "\\item 'XY;3'\n",
       "\\item 'YY;6'\n",
       "\\item 'XX;12'\n",
       "\\item 'XX;9'\n",
       "\\item 'XY;13'\n",
       "\\item 'XX;17'\n",
       "\\item 'XX;20'\n",
       "\\item 'XX;6'\n",
       "\\item 'XX;4'\n",
       "\\item 'XY;6'\n",
       "\\item 'YY;9'\n",
       "\\item 'YY;7'\n",
       "\\item 'YY;4'\n",
       "\\item 'XY;12'\n",
       "\\item 'XY;8'\n",
       "\\item 'YY;2'\n",
       "\\item 'YY;1'\n",
       "\\item 'YY;3'\n",
       "\\item 'XX;14'\n",
       "\\item 'XX;8'\n",
       "\\item 'XX;10'\n",
       "\\item 'XY;4'\n",
       "\\item 'XY;10'\n",
       "\\item 'XY;14'\n",
       "\\item 'XY;7'\n",
       "\\item 'XY;11'\n",
       "\\item 'XY;9'\n",
       "\\item 'XY;15'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XX;5'\n",
       "2. 'XY;3'\n",
       "3. 'YY;6'\n",
       "4. 'XX;12'\n",
       "5. 'XX;9'\n",
       "6. 'XY;13'\n",
       "7. 'XX;17'\n",
       "8. 'XX;20'\n",
       "9. 'XX;6'\n",
       "10. 'XX;4'\n",
       "11. 'XY;6'\n",
       "12. 'YY;9'\n",
       "13. 'YY;7'\n",
       "14. 'YY;4'\n",
       "15. 'XY;12'\n",
       "16. 'XY;8'\n",
       "17. 'YY;2'\n",
       "18. 'YY;1'\n",
       "19. 'YY;3'\n",
       "20. 'XX;14'\n",
       "21. 'XX;8'\n",
       "22. 'XX;10'\n",
       "23. 'XY;4'\n",
       "24. 'XY;10'\n",
       "25. 'XY;14'\n",
       "26. 'XY;7'\n",
       "27. 'XY;11'\n",
       "28. 'XY;9'\n",
       "29. 'XY;15'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"XX;5\"  \"XY;3\"  \"YY;6\"  \"XX;12\" \"XX;9\"  \"XY;13\" \"XX;17\" \"XX;20\" \"XX;6\" \n",
       "[10] \"XX;4\"  \"XY;6\"  \"YY;9\"  \"YY;7\"  \"YY;4\"  \"XY;12\" \"XY;8\"  \"YY;2\"  \"YY;1\" \n",
       "[19] \"YY;3\"  \"XX;14\" \"XX;8\"  \"XX;10\" \"XY;4\"  \"XY;10\" \"XY;14\" \"XY;7\"  \"XY;11\"\n",
       "[28] \"XY;9\"  \"XY;15\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(ind_genotypes$ind)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1702,
   "id": "d70a5bb2-294d-4376-b6d5-7e739f039c4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Check the amount of NA correlations in the data \n",
    "nrow(cor_data_long[is.na(cor_data_long$value),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1703,
   "id": "47e46b4e-ccbd-4fe5-838e-37926dc79aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out NA (correlatin could not be calculated) \n",
    "cor_data_long = cor_data_long[!is.na(cor_data_long$value),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1704,
   "id": "0062f68d-67da-48db-a594-77e6834928f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1292964"
      ],
      "text/latex": [
       "1292964"
      ],
      "text/markdown": [
       "1292964"
      ],
      "text/plain": [
       "[1] 1292964"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1705,
   "id": "c6749e2f-15fd-44c3-87fa-5170dc398d01",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Filter out gene-pairs where correlation could not be calculated for sufficuent amount of individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1706,
   "id": "5e2d0aa8-d107-4575-bd73-19ccfbdd6b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Get the gene-pairs where the correlation could not be calculated in at least 20 individuals\n",
    "#filter_genes_to_low_samples = cor_data_long %>% group_by( variable, gene1, gene2) %>% dplyr::count() %>% dplyr::filter(n<=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1707,
   "id": "2c75c9e1-577a-4435-9f41-f04c6197280b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2424"
      ],
      "text/latex": [
       "2424"
      ],
      "text/markdown": [
       "2424"
      ],
      "text/plain": [
       "[1] 2424"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes_to_low_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1708,
   "id": "fe5a9bde-b44c-4c46-b139-8bc6a0859076",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the gene-pairs where the correlation has onyl been calculated for one individual of the genotype\n",
    "filter_genes_only_one_estimate = cor_data_long %>% group_by( variable, gene1, gene2, genotype) %>% dplyr::count() %>% dplyr::filter(n==1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1709,
   "id": "ac801be2-a107-44f4-97a7-55c63cfb8069",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "17844"
      ],
      "text/latex": [
       "17844"
      ],
      "text/markdown": [
       "17844"
      ],
      "text/plain": [
       "[1] 17844"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes_only_one_estimate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1710,
   "id": "24b90673-2780-4ee6-8ef7-4a9053af3bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the gene-pairs where correlation has onyl been calculated for one specific genotype \n",
    "filter_genes_only_one_group = unique(cor_data_long[,c( 'variable', 'gene1' , 'gene2', 'genotype')])  %>% group_by( variable, gene1, gene2) %>% dplyr::count() %>% dplyr::filter(n ==1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1711,
   "id": "f23ab483-cc72-4bdf-b1f5-299b4944dbe8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "11124"
      ],
      "text/latex": [
       "11124"
      ],
      "text/markdown": [
       "11124"
      ],
      "text/plain": [
       "[1] 11124"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes_only_one_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1712,
   "id": "ddcbf350-2007-4c7c-a1be-7fc9785ad694",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine the two types of filtering out gene-pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1713,
   "id": "8e4558dc-ee22-44cc-8959-4846a23db9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes = unique(rbind(filter_genes_only_one_estimate[, c( 'variable', 'gene1', 'gene2')],\n",
    "                            filter_genes_only_one_group[, c( 'variable', 'gene1', 'gene2')]#,\n",
    "                           # filter_genes_to_low_samples[, c( 'variable', 'gene1', 'gene2')]                           \n",
    "                           ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1714,
   "id": "8926243e-e6fc-48e0-8394-ea58ca294c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes = unique(filter_genes[, c( 'variable', 'gene1', 'gene2')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1715,
   "id": "9613c3ae-3835-478d-8b39-5b7fe806ae50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "24108"
      ],
      "text/latex": [
       "24108"
      ],
      "text/markdown": [
       "24108"
      ],
      "text/plain": [
       "[1] 24108"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1716,
   "id": "0a7d5beb-a6b6-46ec-b71b-1bb57469e815",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove the gene-pairs from the correlation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1717,
   "id": "fe852696-4adb-4526-99b5-742035a345bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long = anti_join(cor_data_long, filter_genes, by = c( 'variable', 'gene1', 'gene2')  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1718,
   "id": "feaaa592-7779-4715-a191-2a404767048c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1145436"
      ],
      "text/latex": [
       "1145436"
      ],
      "text/markdown": [
       "1145436"
      ],
      "text/plain": [
       "[1] 1145436"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1719,
   "id": "143da9a6-24b2-454e-82d2-3620c338da6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate the z-score of the correlation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1720,
   "id": "2bfa17f7-35dd-4859-a75c-012fd9305ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long$z_value = corr_to_z(cor_data_long$value, cor_data_long$ncells) # z-score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1721,
   "id": "dac31aff-3c3b-4f3c-8678-dab1119d4295",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long$z_value_non_zero = corr_to_z(cor_data_long$value, (cor_data_long$non_zero_counts.x + cor_data_long$non_zero_counts.y)/2)   # z-score based on ncells with  non-zero counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1722,
   "id": "e7d3185b-b2a8-42cd-a2ae-1d14dd7391d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>genotype</th><th scope=col>gene1</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>basemean.x</th><th scope=col>amount_zero_counts.x</th><th scope=col>mean_counts.x</th><th scope=col>percentage_zero.x</th><th scope=col>basemean.y</th><th scope=col>amount_zero_counts.y</th><th scope=col>mean_counts.y</th><th scope=col>percentage_zero.y</th><th scope=col>non_zero_counts.x</th><th scope=col>non_zero_counts.y</th><th scope=col>z_value</th><th scope=col>z_value_non_zero</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene2_10</td><td>11</td><td>YY;9</td><td>YY</td><td>gene1_10</td><td>pearson_raw</td><td>0.78447571</td><td>9</td><td>0</td><td>10.27273</td><td>0</td><td>12</td><td>0</td><td>13.63636</td><td>0</td><td>11</td><td>11</td><td>2.85893170</td><td>2.85893170</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene2_10</td><td>11</td><td>YY;9</td><td>YY</td><td>gene1_10</td><td>pearson    </td><td>0.01650075</td><td>9</td><td>0</td><td>10.27273</td><td>0</td><td>12</td><td>0</td><td>13.63636</td><td>0</td><td>11</td><td>11</td><td>0.04815215</td><td>0.04815215</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 19\n",
       "\\begin{tabular}{r|lllllllllllllllllll}\n",
       "  & gene2 & ncells & ind & genotype & gene1 & variable & value & basemean.x & amount\\_zero\\_counts.x & mean\\_counts.x & percentage\\_zero.x & basemean.y & amount\\_zero\\_counts.y & mean\\_counts.y & percentage\\_zero.y & non\\_zero\\_counts.x & non\\_zero\\_counts.y & z\\_value & z\\_value\\_non\\_zero\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <fct> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene2\\_10 & 11 & YY;9 & YY & gene1\\_10 & pearson\\_raw & 0.78447571 & 9 & 0 & 10.27273 & 0 & 12 & 0 & 13.63636 & 0 & 11 & 11 & 2.85893170 & 2.85893170\\\\\n",
       "\t2 & gene2\\_10 & 11 & YY;9 & YY & gene1\\_10 & pearson     & 0.01650075 & 9 & 0 & 10.27273 & 0 & 12 & 0 & 13.63636 & 0 & 11 & 11 & 0.04815215 & 0.04815215\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 19\n",
       "\n",
       "| <!--/--> | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | genotype &lt;chr&gt; | gene1 &lt;chr&gt; | variable &lt;fct&gt; | value &lt;dbl&gt; | basemean.x &lt;dbl&gt; | amount_zero_counts.x &lt;int&gt; | mean_counts.x &lt;dbl&gt; | percentage_zero.x &lt;dbl&gt; | basemean.y &lt;dbl&gt; | amount_zero_counts.y &lt;int&gt; | mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; | non_zero_counts.x &lt;int&gt; | non_zero_counts.y &lt;int&gt; | z_value &lt;dbl&gt; | z_value_non_zero &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene2_10 | 11 | YY;9 | YY | gene1_10 | pearson_raw | 0.78447571 | 9 | 0 | 10.27273 | 0 | 12 | 0 | 13.63636 | 0 | 11 | 11 | 2.85893170 | 2.85893170 |\n",
       "| 2 | gene2_10 | 11 | YY;9 | YY | gene1_10 | pearson     | 0.01650075 | 9 | 0 | 10.27273 | 0 | 12 | 0 | 13.63636 | 0 | 11 | 11 | 0.04815215 | 0.04815215 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene2    ncells ind  genotype gene1    variable    value      basemean.x\n",
       "1 gene2_10 11     YY;9 YY       gene1_10 pearson_raw 0.78447571 9         \n",
       "2 gene2_10 11     YY;9 YY       gene1_10 pearson     0.01650075 9         \n",
       "  amount_zero_counts.x mean_counts.x percentage_zero.x basemean.y\n",
       "1 0                    10.27273      0                 12        \n",
       "2 0                    10.27273      0                 12        \n",
       "  amount_zero_counts.y mean_counts.y percentage_zero.y non_zero_counts.x\n",
       "1 0                    13.63636      0                 11               \n",
       "2 0                    13.63636      0                 11               \n",
       "  non_zero_counts.y z_value    z_value_non_zero\n",
       "1 11                2.85893170 2.85893170      \n",
       "2 11                0.04815215 0.04815215      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data_long,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1723,
   "id": "52b2e755-7b32-4ed8-9005-e5dcf316adfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check for missing and infinite z-scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1724,
   "id": "473caf2b-697a-469b-825c-c8f2825d00aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long[is.na(cor_data_long$z_value),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1725,
   "id": "f88904c1-fb21-4115-b8bc-44314feb38a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cor_data_long[is.infinite(cor_data_long$z_value),])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1726,
   "id": "867dc55f-cd1b-4acf-851c-43a1436816f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.970322429013271"
      ],
      "text/latex": [
       "0.970322429013271"
      ],
      "text/markdown": [
       "0.970322429013271"
      ],
      "text/plain": [
       "[1] 0.9703224"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "max(cor_data_long$value[!is.infinite(cor_data_long$z_value)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1727,
   "id": "3ab172e2-3283-4aa3-af7a-c4b5b4ee64dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get the amount of individuals left per genotype after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1728,
   "id": "f59169eb-0599-41c8-8391-308b78d84470",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 3 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>genotype</th><th scope=col>n</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>XX</td><td>10</td></tr>\n",
       "\t<tr><td>XY</td><td>12</td></tr>\n",
       "\t<tr><td>YY</td><td> 7</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 3 × 2\n",
       "\\begin{tabular}{ll}\n",
       " genotype & n\\\\\n",
       " <chr> & <int>\\\\\n",
       "\\hline\n",
       "\t XX & 10\\\\\n",
       "\t XY & 12\\\\\n",
       "\t YY &  7\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 3 × 2\n",
       "\n",
       "| genotype &lt;chr&gt; | n &lt;int&gt; |\n",
       "|---|---|\n",
       "| XX | 10 |\n",
       "| XY | 12 |\n",
       "| YY |  7 |\n",
       "\n"
      ],
      "text/plain": [
       "  genotype n \n",
       "1 XX       10\n",
       "2 XY       12\n",
       "3 YY        7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(cor_data_long[,c('ind', 'genotype')]) %>% group_by(genotype) %>% dplyr::count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1729,
   "id": "b07b0533-3c0d-4e29-8edb-cba5dc709389",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Test the co-eQTL effect with a linear model (correlation =y; genotype = x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1730,
   "id": "4f3df5b8-69a2-4153-9bd2-4abcbeff8073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 19</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene2</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>genotype</th><th scope=col>gene1</th><th scope=col>variable</th><th scope=col>value</th><th scope=col>basemean.x</th><th scope=col>amount_zero_counts.x</th><th scope=col>mean_counts.x</th><th scope=col>percentage_zero.x</th><th scope=col>basemean.y</th><th scope=col>amount_zero_counts.y</th><th scope=col>mean_counts.y</th><th scope=col>percentage_zero.y</th><th scope=col>non_zero_counts.x</th><th scope=col>non_zero_counts.y</th><th scope=col>z_value</th><th scope=col>z_value_non_zero</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene2_10</td><td>11</td><td>YY;9</td><td>YY</td><td>gene1_10</td><td>pearson_raw</td><td>0.78447571</td><td>9</td><td>0</td><td>10.27273</td><td>0</td><td>12</td><td>0</td><td>13.63636</td><td>0</td><td>11</td><td>11</td><td>2.85893170</td><td>2.85893170</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene2_10</td><td>11</td><td>YY;9</td><td>YY</td><td>gene1_10</td><td>pearson    </td><td>0.01650075</td><td>9</td><td>0</td><td>10.27273</td><td>0</td><td>12</td><td>0</td><td>13.63636</td><td>0</td><td>11</td><td>11</td><td>0.04815215</td><td>0.04815215</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 19\n",
       "\\begin{tabular}{r|lllllllllllllllllll}\n",
       "  & gene2 & ncells & ind & genotype & gene1 & variable & value & basemean.x & amount\\_zero\\_counts.x & mean\\_counts.x & percentage\\_zero.x & basemean.y & amount\\_zero\\_counts.y & mean\\_counts.y & percentage\\_zero.y & non\\_zero\\_counts.x & non\\_zero\\_counts.y & z\\_value & z\\_value\\_non\\_zero\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <fct> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene2\\_10 & 11 & YY;9 & YY & gene1\\_10 & pearson\\_raw & 0.78447571 & 9 & 0 & 10.27273 & 0 & 12 & 0 & 13.63636 & 0 & 11 & 11 & 2.85893170 & 2.85893170\\\\\n",
       "\t2 & gene2\\_10 & 11 & YY;9 & YY & gene1\\_10 & pearson     & 0.01650075 & 9 & 0 & 10.27273 & 0 & 12 & 0 & 13.63636 & 0 & 11 & 11 & 0.04815215 & 0.04815215\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 19\n",
       "\n",
       "| <!--/--> | gene2 &lt;chr&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | genotype &lt;chr&gt; | gene1 &lt;chr&gt; | variable &lt;fct&gt; | value &lt;dbl&gt; | basemean.x &lt;dbl&gt; | amount_zero_counts.x &lt;int&gt; | mean_counts.x &lt;dbl&gt; | percentage_zero.x &lt;dbl&gt; | basemean.y &lt;dbl&gt; | amount_zero_counts.y &lt;int&gt; | mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; | non_zero_counts.x &lt;int&gt; | non_zero_counts.y &lt;int&gt; | z_value &lt;dbl&gt; | z_value_non_zero &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene2_10 | 11 | YY;9 | YY | gene1_10 | pearson_raw | 0.78447571 | 9 | 0 | 10.27273 | 0 | 12 | 0 | 13.63636 | 0 | 11 | 11 | 2.85893170 | 2.85893170 |\n",
       "| 2 | gene2_10 | 11 | YY;9 | YY | gene1_10 | pearson     | 0.01650075 | 9 | 0 | 10.27273 | 0 | 12 | 0 | 13.63636 | 0 | 11 | 11 | 0.04815215 | 0.04815215 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene2    ncells ind  genotype gene1    variable    value      basemean.x\n",
       "1 gene2_10 11     YY;9 YY       gene1_10 pearson_raw 0.78447571 9         \n",
       "2 gene2_10 11     YY;9 YY       gene1_10 pearson     0.01650075 9         \n",
       "  amount_zero_counts.x mean_counts.x percentage_zero.x basemean.y\n",
       "1 0                    10.27273      0                 12        \n",
       "2 0                    10.27273      0                 12        \n",
       "  amount_zero_counts.y mean_counts.y percentage_zero.y non_zero_counts.x\n",
       "1 0                    13.63636      0                 11               \n",
       "2 0                    13.63636      0                 11               \n",
       "  non_zero_counts.y z_value    z_value_non_zero\n",
       "1 11                2.85893170 2.85893170      \n",
       "2 11                0.04815215 0.04815215      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cor_data_long,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1731,
   "id": "4de1b0f5-7748-4784-b4e0-4993143afd9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long$genotype = factor(cor_data_long$genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1732,
   "id": "0a908b74-14f1-420c-8a57-140834055e4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>YY</li><li>XY</li><li>XX</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'XX'</li><li>'XY'</li><li>'YY'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item YY\n",
       "\\item XY\n",
       "\\item XX\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'XX'\n",
       "\\item 'XY'\n",
       "\\item 'YY'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. YY\n",
       "2. XY\n",
       "3. XX\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'XX'\n",
       "2. 'XY'\n",
       "3. 'YY'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] YY XY XX\n",
       "Levels: XX XY YY"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(cor_data_long$genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1733,
   "id": "39da2b7b-773b-4515-810d-05c254bcf075",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>3</li><li>2</li><li>1</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 3\n",
       "\\item 2\n",
       "\\item 1\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 3\n",
       "2. 2\n",
       "3. 1\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 3 2 1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(as.numeric(cor_data_long$genotype))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1734,
   "id": "d50615ae-0bc2-45dd-9526-b683e0db2d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#as.numeric(levels(cor_data_long$genotype))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1735,
   "id": "02a952d6-2c9e-40d7-a026-0de74b12d325",
   "metadata": {},
   "outputs": [],
   "source": [
    "lm_result = cor_data_long %>% group_by( variable, gene1, gene2) %>% group_modify(~ broom::tidy(lm(value ~ as.numeric(genotype), data = .x, na.action = 'na.omit')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1736,
   "id": "c4b0f55c-c0d5-49ef-8d32-ee56a4a275a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lm_result_backup = lm_result   # save the result of the linear model as backup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1737,
   "id": "a05d06fb-5202-4416-a8fd-4bcead9fe1af",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check the structure of the result of the linear model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1738,
   "id": "5adb5e73-908c-42ed-8ba3-20bdb03818fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 4 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>variable</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std.error</th><th scope=col>statistic</th><th scope=col>p.value</th></tr>\n",
       "\t<tr><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>spearman_raw</td><td>gene1_10  </td><td>gene2_10  </td><td>(Intercept)         </td><td>-0.61835725</td><td>0.16078955</td><td>-3.845755</td><td>0.0032345513</td></tr>\n",
       "\t<tr><td>spearman_raw</td><td>gene1_10  </td><td>gene2_10  </td><td>as.numeric(genotype)</td><td> 0.34705695</td><td>0.06113774</td><td> 5.676640</td><td>0.0002047867</td></tr>\n",
       "\t<tr><td>spearman_raw</td><td>gene1_1002</td><td>gene2_1002</td><td>(Intercept)         </td><td> 0.29214202</td><td>0.12469905</td><td> 2.342777</td><td>0.0295938863</td></tr>\n",
       "\t<tr><td>spearman_raw</td><td>gene1_1002</td><td>gene2_1002</td><td>as.numeric(genotype)</td><td> 0.07548951</td><td>0.07679987</td><td> 0.982938</td><td>0.3373801584</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 4 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " variable & gene1 & gene2 & term & estimate & std.error & statistic & p.value\\\\\n",
       " <fct> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t spearman\\_raw & gene1\\_10   & gene2\\_10   & (Intercept)          & -0.61835725 & 0.16078955 & -3.845755 & 0.0032345513\\\\\n",
       "\t spearman\\_raw & gene1\\_10   & gene2\\_10   & as.numeric(genotype) &  0.34705695 & 0.06113774 &  5.676640 & 0.0002047867\\\\\n",
       "\t spearman\\_raw & gene1\\_1002 & gene2\\_1002 & (Intercept)          &  0.29214202 & 0.12469905 &  2.342777 & 0.0295938863\\\\\n",
       "\t spearman\\_raw & gene1\\_1002 & gene2\\_1002 & as.numeric(genotype) &  0.07548951 & 0.07679987 &  0.982938 & 0.3373801584\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 4 × 8\n",
       "\n",
       "| variable &lt;fct&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std.error &lt;dbl&gt; | statistic &lt;dbl&gt; | p.value &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| spearman_raw | gene1_10   | gene2_10   | (Intercept)          | -0.61835725 | 0.16078955 | -3.845755 | 0.0032345513 |\n",
       "| spearman_raw | gene1_10   | gene2_10   | as.numeric(genotype) |  0.34705695 | 0.06113774 |  5.676640 | 0.0002047867 |\n",
       "| spearman_raw | gene1_1002 | gene2_1002 | (Intercept)          |  0.29214202 | 0.12469905 |  2.342777 | 0.0295938863 |\n",
       "| spearman_raw | gene1_1002 | gene2_1002 | as.numeric(genotype) |  0.07548951 | 0.07679987 |  0.982938 | 0.3373801584 |\n",
       "\n"
      ],
      "text/plain": [
       "  variable     gene1      gene2      term                 estimate   \n",
       "1 spearman_raw gene1_10   gene2_10   (Intercept)          -0.61835725\n",
       "2 spearman_raw gene1_10   gene2_10   as.numeric(genotype)  0.34705695\n",
       "3 spearman_raw gene1_1002 gene2_1002 (Intercept)           0.29214202\n",
       "4 spearman_raw gene1_1002 gene2_1002 as.numeric(genotype)  0.07548951\n",
       "  std.error  statistic p.value     \n",
       "1 0.16078955 -3.845755 0.0032345513\n",
       "2 0.06113774  5.676640 0.0002047867\n",
       "3 0.12469905  2.342777 0.0295938863\n",
       "4 0.07679987  0.982938 0.3373801584"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(lm_result_backup,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1739,
   "id": "3fc929cd-f7f1-4736-a6f5-30296042f20d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "109680"
      ],
      "text/latex": [
       "109680"
      ],
      "text/markdown": [
       "109680"
      ],
      "text/plain": [
       "[1] 109680"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(lm_result_backup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1740,
   "id": "0a9c101b-5900-460c-80cd-6c12b244cb32",
   "metadata": {},
   "outputs": [],
   "source": [
    "## filter out Intercept information to keep only beta\n",
    "lm_result = lm_result_backup[lm_result_backup$term != '(Intercept)',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1741,
   "id": "9a675df4-835c-445d-a502-961278545aa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "54840"
      ],
      "text/latex": [
       "54840"
      ],
      "text/markdown": [
       "54840"
      ],
      "text/plain": [
       "[1] 54840"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(lm_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1742,
   "id": "af5f9d46-9b10-4615-8ef1-ea6435649e27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'as.numeric(genotype)'"
      ],
      "text/latex": [
       "'as.numeric(genotype)'"
      ],
      "text/markdown": [
       "'as.numeric(genotype)'"
      ],
      "text/plain": [
       "[1] \"as.numeric(genotype)\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(lm_result$term)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3de953f-eed8-4ecd-b6cd-3eee33fa0e26",
   "metadata": {},
   "source": [
    "## Linear model on z-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1743,
   "id": "8f681cd9-c8d8-4dd8-b0ea-d8a44f8b5a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Estimate the linear model on the z-value transformed correlations instead of the correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1744,
   "id": "6ff8b04a-cdba-4fd0-a35d-9b68f875202b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Define how to handle infinite z-value\n",
    "# 1) Set to NA: ''\n",
    "# 2) Replace by max: 'max'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1745,
   "id": "c5ef78d8-8531-4f35-8153-9d6753c841b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "z_value_handle = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1746,
   "id": "fcc71899-e376-4fdf-8306-fc2ea35cabe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(z_value_handle == ''){\n",
    "    cor_data_long$z_value[is.infinite(cor_data_long$z_value)] = NA\n",
    "    cor_data_long$z_value_non_zero[is.infinite(cor_data_long$z_value_non_zero)] = NA\n",
    "    }\n",
    "\n",
    "if(z_value_handle == 'max'){\n",
    "    cor_data_long$z_value[is.infinite(cor_data_long$z_value)]  = max(cor_data_long$z_value[!is.infinite(cor_data_long$z_value)])\n",
    "    cor_data_long$z_value_non_zero[is.infinite(cor_data_long$z_value_non_zero)]  = max(cor_data_long$z_value[!is.infinite(cor_data_long$z_value_non_zero)])\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1747,
   "id": "ed1b3002-85e8-4a43-a179-cd66f9354f97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "7.08315441942221"
      ],
      "text/latex": [
       "7.08315441942221"
      ],
      "text/markdown": [
       "7.08315441942221"
      ],
      "text/plain": [
       "[1] 7.083154"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "max(cor_data_long$z_value[!is.na(cor_data_long$z_value)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1748,
   "id": "44c0b97f-683f-497e-b1e4-131d69b53e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove the NA z-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1749,
   "id": "5b1d0d30-8e95-4f48-9805-2ec3d24eeaaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long = cor_data_long[!is.na(cor_data_long$z_value),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1750,
   "id": "127513cc-a4a0-436d-b9ee-18efb38a7a37",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Define gene-pairs where z-score is only calculated for one specific individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1751,
   "id": "5a92ba69-2861-4b00-915d-95b72e4b693c",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes_only_one_estimate = cor_data_long %>% group_by( variable, gene1, gene2, genotype) %>% dplyr::count() %>% dplyr::filter(n==1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1752,
   "id": "0c2d65b8-64af-4228-bb26-d2fe90ba4aff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes_only_one_estimate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1753,
   "id": "1f9ae0dd-0f56-4a20-a2fa-9cb076790766",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define gene-pairs where z-score is only calculated for one specific genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1754,
   "id": "370b82d7-c03a-421c-a98d-4828b17f66e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes_only_one_group = unique(cor_data_long[,c( 'variable', 'gene1' , 'gene2', 'genotype')])  %>% group_by( variable, gene1, gene2) %>% dplyr::count() %>% dplyr::filter(n ==1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1755,
   "id": "5fa9d4ea-e7ad-4a1d-87f0-b162fdb59096",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes_only_one_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1756,
   "id": "0df16fa0-28aa-4465-9f19-924f8e7ae6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Combine the two types of filtering out gene-pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1757,
   "id": "1d5aff0a-0229-4e4e-a316-45fb8e4b2b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes = unique(rbind(filter_genes_only_one_estimate[, c( 'variable', 'gene1', 'gene2')],filter_genes_only_one_group[, c( 'variable', 'gene1', 'gene2')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1758,
   "id": "7e8582f7-9a06-4cee-a237-e787650705a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_genes = unique(filter_genes[, c( 'variable', 'gene1', 'gene2')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1759,
   "id": "65a543f1-caab-48c7-80c2-97c5895a67b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(filter_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1760,
   "id": "f81175d4-01c8-4de9-982f-872dda05cc10",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove the gene-pairs from the correlation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1761,
   "id": "77eac3ec-6976-4b38-9e06-af52b0585e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "cor_data_long = anti_join(cor_data_long, filter_genes, by = c( 'variable', 'gene1', 'gene2')  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1762,
   "id": "73bdaff2-147c-4b39-bd30-31e0a51127de",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate the linear model for the z-score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1763,
   "id": "05232e1d-8ad1-4b2c-80cf-661dbb53118d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "lm_result_z = cor_data_long[!is.na(cor_data_long$z_value),] %>% group_by(variable, gene1, gene2) %>% group_modify(~ broom::tidy(lm(z_value ~ as.numeric(genotype), data = .x, na.action = 'na.omit')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1764,
   "id": "9630ac26-ad2b-45fe-819e-790b7d276336",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Filter out the Intercept information\n",
    "lm_result_z = lm_result_z[lm_result_z$term != '(Intercept)',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1765,
   "id": "2b57bca3-79a7-4b2c-a109-ad8b528e01ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Rename the columns to specify they are based on z-scores\n",
    "colnames(lm_result_z) = paste0(colnames(lm_result_z), '_z')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1766,
   "id": "d9ef051a-ec8e-44f6-bcc3-c1b798bd3f4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "54840"
      ],
      "text/latex": [
       "54840"
      ],
      "text/markdown": [
       "54840"
      ],
      "text/plain": [
       "[1] 54840"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(lm_result_z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1767,
   "id": "ed43028e-02e5-44dc-9bc1-18408040ace5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'as.numeric(genotype)'"
      ],
      "text/latex": [
       "'as.numeric(genotype)'"
      ],
      "text/markdown": [
       "'as.numeric(genotype)'"
      ],
      "text/plain": [
       "[1] \"as.numeric(genotype)\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(lm_result_z$term_z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1768,
   "id": "e68e8555-e94d-4de6-939b-5662af92880c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merge the z-score based result to the result based on correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1769,
   "id": "64156833-8d03-4a80-a7d3-39ee53912c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "lm_result = merge(lm_result, lm_result_z, by.x = c('gene1', 'gene2', 'variable'), by.y =  c('gene1_z', 'gene2_z','variable_z'), all.x = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59ce3bc8-b6e7-4f38-963e-c95c7607438b",
   "metadata": {},
   "source": [
    "## Add meta- information to linear model result and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1770,
   "id": "6064afbc-d901-4b15-823d-c881748d54aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add summary statistics like mean expression per gene etc. to the model results and save the model results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1771,
   "id": "ffbbb79e-96c4-4cdc-8a15-8f2b302a2aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Get amount of tests and correct the p-value for multiple testing (correlation based p-value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1772,
   "id": "b2a3d718-de48-4afa-aae4-cbf04ed9e9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "amount_tests = lm_result %>% group_by(variable) %>% summarise(amount_tests = n())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1773,
   "id": "8750e366-765d-432d-8f59-a7038ce187b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lm_result = merge(lm_result, amount_tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1774,
   "id": "b10a6415-1ba4-425f-a9df-c17a8a793fa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "lm_result$p.value_adj = lm_result$p.value * lm_result$amount_tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1775,
   "id": "e4763719-4992-4697-8b92-2aee6aacfd42",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add information about mean percentage of zero of gene across individuals "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1776,
   "id": "bfd6c576-bc22-4cd5-b1bf-2e54a239c4a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_summary = gene_summary1 # combine info from all individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1777,
   "id": "e272b301-1a46-4281-b28c-0e73d608d560",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>amount_zero_counts</th><th scope=col>amount_cells</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>percentage_zero_bin</th><th scope=col>genotype</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td> 1</td><td>10</td><td>XX;1</td><td>3</td><td>10</td><td> 1.0</td><td> 3</td><td>0</td><td> 0.8888889</td><td>30</td><td>(20,40]   </td><td>XX</td></tr>\n",
       "\t<tr><td>2</td><td>gene2_1</td><td>16</td><td>10</td><td>XX;1</td><td>0</td><td>10</td><td>14.5</td><td>33</td><td>6</td><td>68.5000000</td><td> 0</td><td>(-0.001,5]</td><td>XX</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & variable & basemean & ncells & ind & amount\\_zero\\_counts & amount\\_cells & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & percentage\\_zero\\_bin & genotype\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 &  1 & 10 & XX;1 & 3 & 10 &  1.0 &  3 & 0 &  0.8888889 & 30 & (20,40{]}    & XX\\\\\n",
       "\t 2 & gene2\\_1 & 16 & 10 & XX;1 & 0 & 10 & 14.5 & 33 & 6 & 68.5000000 &  0 & (-0.001,5{]} & XX\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | amount_zero_counts &lt;int&gt; | amount_cells &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | percentage_zero_bin &lt;chr&gt; | genotype &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 |  1 | 10 | XX;1 | 3 | 10 |  1.0 |  3 | 0 |  0.8888889 | 30 | (20,40]    | XX |\n",
       "| 2 | gene2_1 | 16 | 10 | XX;1 | 0 | 10 | 14.5 | 33 | 6 | 68.5000000 |  0 | (-0.001,5] | XX |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable basemean ncells ind  amount_zero_counts amount_cells mean_counts\n",
       "1 1  gene1_1   1       10     XX;1 3                  10            1.0       \n",
       "2 2  gene2_1  16       10     XX;1 0                  10           14.5       \n",
       "  max_counts min_counts variance_counts percentage_zero percentage_zero_bin\n",
       "1  3         0           0.8888889      30              (20,40]            \n",
       "2 33         6          68.5000000       0              (-0.001,5]         \n",
       "  genotype\n",
       "1 XX      \n",
       "2 XX      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1778,
   "id": "387e739d-896a-4d64-b045-f7c511637af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# summarise the gene-statistics across both genotypes\n",
    "gene_summary_sum_all = gene_summary %>% group_by(variable) %>% summarise(\n",
    "    amount_zero_counts = mean(amount_zero_counts,na.rm = TRUE),\n",
    "    var_amount_zero_counts = var(amount_zero_counts, na.rm = TRUE),\n",
    "    mean_counts = mean(mean_counts, na.rm = TRUE),\n",
    "    var_mean_counts = var(mean_counts, na.rm = TRUE),\n",
    "    percentage_zero = mean(percentage_zero, na.rm = TRUE),\n",
    "    var_percentage_zero = var(percentage_zero, na.rm = TRUE),\n",
    "    mean_ncells = mean(ncells),\n",
    "    sd_ncells =sd(ncells)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1779,
   "id": "9cd6efea-7ec4-4af9-80dd-b88fa572435e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'variable', 'basemean'. You can override\n",
      "using the `.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "# summarise the gene-statistics per genotype and basemean\n",
    "gene_summary_sum = gene_summary %>% group_by(variable, basemean, genotype) %>% summarise(\n",
    "    amount_zero_counts = mean(amount_zero_counts,na.rm = TRUE),\n",
    "    var_amount_zero_counts = var(amount_zero_counts, na.rm = TRUE),\n",
    "    mean_counts = mean(mean_counts, na.rm = TRUE),\n",
    "    var_mean_counts = var(mean_counts, na.rm = TRUE),\n",
    "    percentage_zero = mean(percentage_zero, na.rm = TRUE),\n",
    "    var_percentage_zero = var(percentage_zero, na.rm = TRUE),\n",
    "    mean_ncells = mean(ncells),\n",
    "    sd_ncells =sd(ncells)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1780,
   "id": "358c47c7-10cd-404f-a893-e3f8c5e931fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>variable</th><th scope=col>basemean</th><th scope=col>genotype</th><th scope=col>amount_zero_counts</th><th scope=col>var_amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>var_mean_counts</th><th scope=col>percentage_zero</th><th scope=col>var_percentage_zero</th><th scope=col>mean_ncells</th><th scope=col>sd_ncells</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>gene1_1</td><td>1</td><td>XX</td><td>6.95</td><td>NA</td><td>1.0276534</td><td>NA</td><td>43.02716</td><td>NA</td><td>15.95000</td><td>7.082038</td></tr>\n",
       "\t<tr><td>gene1_1</td><td>1</td><td>XY</td><td>7.20</td><td>NA</td><td>0.9940176</td><td>NA</td><td>42.13823</td><td>NA</td><td>16.53333</td><td>7.623897</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 2 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " variable & basemean & genotype & amount\\_zero\\_counts & var\\_amount\\_zero\\_counts & mean\\_counts & var\\_mean\\_counts & percentage\\_zero & var\\_percentage\\_zero & mean\\_ncells & sd\\_ncells\\\\\n",
       " <chr> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t gene1\\_1 & 1 & XX & 6.95 & NA & 1.0276534 & NA & 43.02716 & NA & 15.95000 & 7.082038\\\\\n",
       "\t gene1\\_1 & 1 & XY & 7.20 & NA & 0.9940176 & NA & 42.13823 & NA & 16.53333 & 7.623897\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 2 × 11\n",
       "\n",
       "| variable &lt;chr&gt; | basemean &lt;dbl&gt; | genotype &lt;chr&gt; | amount_zero_counts &lt;dbl&gt; | var_amount_zero_counts &lt;dbl&gt; | mean_counts &lt;dbl&gt; | var_mean_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | var_percentage_zero &lt;dbl&gt; | mean_ncells &lt;dbl&gt; | sd_ncells &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| gene1_1 | 1 | XX | 6.95 | NA | 1.0276534 | NA | 43.02716 | NA | 15.95000 | 7.082038 |\n",
       "| gene1_1 | 1 | XY | 7.20 | NA | 0.9940176 | NA | 42.13823 | NA | 16.53333 | 7.623897 |\n",
       "\n"
      ],
      "text/plain": [
       "  variable basemean genotype amount_zero_counts var_amount_zero_counts\n",
       "1 gene1_1  1        XX       6.95               NA                    \n",
       "2 gene1_1  1        XY       7.20               NA                    \n",
       "  mean_counts var_mean_counts percentage_zero var_percentage_zero mean_ncells\n",
       "1 1.0276534   NA              43.02716        NA                  15.95000   \n",
       "2 0.9940176   NA              42.13823        NA                  16.53333   \n",
       "  sd_ncells\n",
       "1 7.082038 \n",
       "2 7.623897 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_summary_sum,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1781,
   "id": "e37a9ca3-0dda-4c3e-b1e8-6b03ecda293e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene_summary_sum1 = gene_summary_sum[gene_summary_sum$genotype == genotype_perm,]\n",
    "#colnames(gene_summary_sum1) = paste0(genotype_perm, '_',  colnames(gene_summary_sum1 ))\n",
    "#gene_summary_sum2 = gene_summary_sum[gene_summary_sum$genotype == genotype_orig,]\n",
    "#colnames(gene_summary_sum2) = paste0(genotype_orig, '_',  colnames(gene_summary_sum2 ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1782,
   "id": "fb1b331f-7786-4c7a-b4a3-6f67c18f8cda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'XX'</li><li>'XY'</li><li>'YY'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'XX'\n",
       "\\item 'XY'\n",
       "\\item 'YY'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'XX'\n",
       "2. 'XY'\n",
       "3. 'YY'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"XX\" \"XY\" \"YY\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(gene_summary_sum$genotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1783,
   "id": "f3465aec-6063-449b-bb47-de9e8ec0fa1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>variable</th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std.error</th><th scope=col>statistic</th><th scope=col>p.value</th><th scope=col>term_z</th><th scope=col>estimate_z</th><th scope=col>std.error_z</th><th scope=col>statistic_z</th><th scope=col>p.value_z</th><th scope=col>amount_tests</th><th scope=col>p.value_adj</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>pearson</td><td>gene1_10  </td><td>gene2_10  </td><td>as.numeric(genotype)</td><td>0.12779867</td><td>0.04387965</td><td>2.9124813</td><td>0.01549873</td><td>as.numeric(genotype)</td><td>0.53349520</td><td>0.18399799</td><td>2.8994621</td><td>0.01584819</td><td>4570</td><td>  70.8292</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>pearson</td><td>gene1_6952</td><td>gene2_6952</td><td>as.numeric(genotype)</td><td>0.01561604</td><td>0.02580120</td><td>0.6052446</td><td>0.55007033</td><td>as.numeric(genotype)</td><td>0.06402256</td><td>0.09795724</td><td>0.6535766</td><td>0.51891244</td><td>4570</td><td>2513.8214</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 15\n",
       "\\begin{tabular}{r|lllllllllllllll}\n",
       "  & variable & gene1 & gene2 & term & estimate & std.error & statistic & p.value & term\\_z & estimate\\_z & std.error\\_z & statistic\\_z & p.value\\_z & amount\\_tests & p.value\\_adj\\\\\n",
       "  & <fct> & <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & pearson & gene1\\_10   & gene2\\_10   & as.numeric(genotype) & 0.12779867 & 0.04387965 & 2.9124813 & 0.01549873 & as.numeric(genotype) & 0.53349520 & 0.18399799 & 2.8994621 & 0.01584819 & 4570 &   70.8292\\\\\n",
       "\t2 & pearson & gene1\\_6952 & gene2\\_6952 & as.numeric(genotype) & 0.01561604 & 0.02580120 & 0.6052446 & 0.55007033 & as.numeric(genotype) & 0.06402256 & 0.09795724 & 0.6535766 & 0.51891244 & 4570 & 2513.8214\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 15\n",
       "\n",
       "| <!--/--> | variable &lt;fct&gt; | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std.error &lt;dbl&gt; | statistic &lt;dbl&gt; | p.value &lt;dbl&gt; | term_z &lt;chr&gt; | estimate_z &lt;dbl&gt; | std.error_z &lt;dbl&gt; | statistic_z &lt;dbl&gt; | p.value_z &lt;dbl&gt; | amount_tests &lt;int&gt; | p.value_adj &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | pearson | gene1_10   | gene2_10   | as.numeric(genotype) | 0.12779867 | 0.04387965 | 2.9124813 | 0.01549873 | as.numeric(genotype) | 0.53349520 | 0.18399799 | 2.8994621 | 0.01584819 | 4570 |   70.8292 |\n",
       "| 2 | pearson | gene1_6952 | gene2_6952 | as.numeric(genotype) | 0.01561604 | 0.02580120 | 0.6052446 | 0.55007033 | as.numeric(genotype) | 0.06402256 | 0.09795724 | 0.6535766 | 0.51891244 | 4570 | 2513.8214 |\n",
       "\n"
      ],
      "text/plain": [
       "  variable gene1      gene2      term                 estimate   std.error \n",
       "1 pearson  gene1_10   gene2_10   as.numeric(genotype) 0.12779867 0.04387965\n",
       "2 pearson  gene1_6952 gene2_6952 as.numeric(genotype) 0.01561604 0.02580120\n",
       "  statistic p.value    term_z               estimate_z std.error_z statistic_z\n",
       "1 2.9124813 0.01549873 as.numeric(genotype) 0.53349520 0.18399799  2.8994621  \n",
       "2 0.6052446 0.55007033 as.numeric(genotype) 0.06402256 0.09795724  0.6535766  \n",
       "  p.value_z  amount_tests p.value_adj\n",
       "1 0.01584819 4570           70.8292  \n",
       "2 0.51891244 4570         2513.8214  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(lm_result,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1784,
   "id": "92392c1f-b660-48c4-8a57-b50318ed56ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "## add for gene 1\n",
    "#lm_result = merge(lm_result, gene_summary_sum, by.x = c('gene1', 'genotype'), by.y = c('variable', 'genotyp'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1785,
   "id": "56630458-0ce8-4fce-9a28-d531c9ed188e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## add for gene2 \n",
    "#lm_result = merge(lm_result, gene_summary_sum1, by.x = c('gene2'), by.y = c(paste0(genotype_perm, '_variable')))\n",
    "#lm_result = merge(lm_result, gene_summary_sum2, by.x = c('gene2'), by.y = c(paste0(genotype_orig, '_variable')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1786,
   "id": "b6599897-a1ad-488a-badd-1205fe92b83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## add summary across both genotypes\n",
    "lm_result = merge(lm_result, gene_summary_sum_all, by.x = c('gene1'), by.y = c('variable'))\n",
    "lm_result = merge(lm_result, gene_summary_sum_all, by.x = c('gene2'), by.y = c('variable'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1787,
   "id": "34ead53b-bf57-46a2-a6f9-ab25f5660a8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene2</th><th scope=col>gene1</th><th scope=col>variable</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std.error</th><th scope=col>statistic</th><th scope=col>p.value</th><th scope=col>term_z</th><th scope=col>estimate_z</th><th scope=col>⋯</th><th scope=col>mean_ncells.x</th><th scope=col>sd_ncells.x</th><th scope=col>amount_zero_counts.y</th><th scope=col>var_amount_zero_counts.y</th><th scope=col>mean_counts.y</th><th scope=col>var_mean_counts.y</th><th scope=col>percentage_zero.y</th><th scope=col>var_percentage_zero.y</th><th scope=col>mean_ncells.y</th><th scope=col>sd_ncells.y</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene2_10</td><td>gene1_10</td><td>pearson              </td><td>as.numeric(genotype)</td><td>0.1277987</td><td>0.04387965</td><td>2.9124813</td><td>0.01549873</td><td>as.numeric(genotype)</td><td>0.53349520</td><td>⋯</td><td>15.93333</td><td>6.668788</td><td>0</td><td>NA</td><td>12.19637</td><td>NA</td><td>0</td><td>NA</td><td>15.93333</td><td>6.668788</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene2_10</td><td>gene1_10</td><td>scaled_spearman_ranks</td><td>as.numeric(genotype)</td><td>0.0216004</td><td>0.05444663</td><td>0.3967261</td><td>0.69990385</td><td>as.numeric(genotype)</td><td>0.04768307</td><td>⋯</td><td>15.93333</td><td>6.668788</td><td>0</td><td>NA</td><td>12.19637</td><td>NA</td><td>0</td><td>NA</td><td>15.93333</td><td>6.668788</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 31\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & gene2 & gene1 & variable & term & estimate & std.error & statistic & p.value & term\\_z & estimate\\_z & ⋯ & mean\\_ncells.x & sd\\_ncells.x & amount\\_zero\\_counts.y & var\\_amount\\_zero\\_counts.y & mean\\_counts.y & var\\_mean\\_counts.y & percentage\\_zero.y & var\\_percentage\\_zero.y & mean\\_ncells.y & sd\\_ncells.y\\\\\n",
       "  & <chr> & <chr> & <fct> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene2\\_10 & gene1\\_10 & pearson               & as.numeric(genotype) & 0.1277987 & 0.04387965 & 2.9124813 & 0.01549873 & as.numeric(genotype) & 0.53349520 & ⋯ & 15.93333 & 6.668788 & 0 & NA & 12.19637 & NA & 0 & NA & 15.93333 & 6.668788\\\\\n",
       "\t2 & gene2\\_10 & gene1\\_10 & scaled\\_spearman\\_ranks & as.numeric(genotype) & 0.0216004 & 0.05444663 & 0.3967261 & 0.69990385 & as.numeric(genotype) & 0.04768307 & ⋯ & 15.93333 & 6.668788 & 0 & NA & 12.19637 & NA & 0 & NA & 15.93333 & 6.668788\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 31\n",
       "\n",
       "| <!--/--> | gene2 &lt;chr&gt; | gene1 &lt;chr&gt; | variable &lt;fct&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std.error &lt;dbl&gt; | statistic &lt;dbl&gt; | p.value &lt;dbl&gt; | term_z &lt;chr&gt; | estimate_z &lt;dbl&gt; | ⋯ ⋯ | mean_ncells.x &lt;dbl&gt; | sd_ncells.x &lt;dbl&gt; | amount_zero_counts.y &lt;dbl&gt; | var_amount_zero_counts.y &lt;dbl&gt; | mean_counts.y &lt;dbl&gt; | var_mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; | var_percentage_zero.y &lt;dbl&gt; | mean_ncells.y &lt;dbl&gt; | sd_ncells.y &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene2_10 | gene1_10 | pearson               | as.numeric(genotype) | 0.1277987 | 0.04387965 | 2.9124813 | 0.01549873 | as.numeric(genotype) | 0.53349520 | ⋯ | 15.93333 | 6.668788 | 0 | NA | 12.19637 | NA | 0 | NA | 15.93333 | 6.668788 |\n",
       "| 2 | gene2_10 | gene1_10 | scaled_spearman_ranks | as.numeric(genotype) | 0.0216004 | 0.05444663 | 0.3967261 | 0.69990385 | as.numeric(genotype) | 0.04768307 | ⋯ | 15.93333 | 6.668788 | 0 | NA | 12.19637 | NA | 0 | NA | 15.93333 | 6.668788 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene2    gene1    variable              term                 estimate \n",
       "1 gene2_10 gene1_10 pearson               as.numeric(genotype) 0.1277987\n",
       "2 gene2_10 gene1_10 scaled_spearman_ranks as.numeric(genotype) 0.0216004\n",
       "  std.error  statistic p.value    term_z               estimate_z ⋯\n",
       "1 0.04387965 2.9124813 0.01549873 as.numeric(genotype) 0.53349520 ⋯\n",
       "2 0.05444663 0.3967261 0.69990385 as.numeric(genotype) 0.04768307 ⋯\n",
       "  mean_ncells.x sd_ncells.x amount_zero_counts.y var_amount_zero_counts.y\n",
       "1 15.93333      6.668788    0                    NA                      \n",
       "2 15.93333      6.668788    0                    NA                      \n",
       "  mean_counts.y var_mean_counts.y percentage_zero.y var_percentage_zero.y\n",
       "1 12.19637      NA                0                 NA                   \n",
       "2 12.19637      NA                0                 NA                   \n",
       "  mean_ncells.y sd_ncells.y\n",
       "1 15.93333      6.668788   \n",
       "2 15.93333      6.668788   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(lm_result,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1788,
   "id": "dfdb65ea-16dd-4a69-bcb7-5e64dd89b82c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "54840"
      ],
      "text/latex": [
       "54840"
      ],
      "text/markdown": [
       "54840"
      ],
      "text/plain": [
       "[1] 54840"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(lm_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1789,
   "id": "32529cc1-55b5-4b66-a1c6-4c727fcaf1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add information about true correlation difference from the specifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1790,
   "id": "4b7d2041-f4d0-4409-8785-75675b6926ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene2</th><th scope=col>gene1</th><th scope=col>variable</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std.error</th><th scope=col>statistic</th><th scope=col>p.value</th><th scope=col>term_z</th><th scope=col>estimate_z</th><th scope=col>⋯</th><th scope=col>mean_ncells.x</th><th scope=col>sd_ncells.x</th><th scope=col>amount_zero_counts.y</th><th scope=col>var_amount_zero_counts.y</th><th scope=col>mean_counts.y</th><th scope=col>var_mean_counts.y</th><th scope=col>percentage_zero.y</th><th scope=col>var_percentage_zero.y</th><th scope=col>mean_ncells.y</th><th scope=col>sd_ncells.y</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene2_10</td><td>gene1_10</td><td>pearson              </td><td>as.numeric(genotype)</td><td>0.1277987</td><td>0.04387965</td><td>2.9124813</td><td>0.01549873</td><td>as.numeric(genotype)</td><td>0.53349520</td><td>⋯</td><td>15.93333</td><td>6.668788</td><td>0</td><td>NA</td><td>12.19637</td><td>NA</td><td>0</td><td>NA</td><td>15.93333</td><td>6.668788</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene2_10</td><td>gene1_10</td><td>scaled_spearman_ranks</td><td>as.numeric(genotype)</td><td>0.0216004</td><td>0.05444663</td><td>0.3967261</td><td>0.69990385</td><td>as.numeric(genotype)</td><td>0.04768307</td><td>⋯</td><td>15.93333</td><td>6.668788</td><td>0</td><td>NA</td><td>12.19637</td><td>NA</td><td>0</td><td>NA</td><td>15.93333</td><td>6.668788</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 31\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & gene2 & gene1 & variable & term & estimate & std.error & statistic & p.value & term\\_z & estimate\\_z & ⋯ & mean\\_ncells.x & sd\\_ncells.x & amount\\_zero\\_counts.y & var\\_amount\\_zero\\_counts.y & mean\\_counts.y & var\\_mean\\_counts.y & percentage\\_zero.y & var\\_percentage\\_zero.y & mean\\_ncells.y & sd\\_ncells.y\\\\\n",
       "  & <chr> & <chr> & <fct> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene2\\_10 & gene1\\_10 & pearson               & as.numeric(genotype) & 0.1277987 & 0.04387965 & 2.9124813 & 0.01549873 & as.numeric(genotype) & 0.53349520 & ⋯ & 15.93333 & 6.668788 & 0 & NA & 12.19637 & NA & 0 & NA & 15.93333 & 6.668788\\\\\n",
       "\t2 & gene2\\_10 & gene1\\_10 & scaled\\_spearman\\_ranks & as.numeric(genotype) & 0.0216004 & 0.05444663 & 0.3967261 & 0.69990385 & as.numeric(genotype) & 0.04768307 & ⋯ & 15.93333 & 6.668788 & 0 & NA & 12.19637 & NA & 0 & NA & 15.93333 & 6.668788\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 31\n",
       "\n",
       "| <!--/--> | gene2 &lt;chr&gt; | gene1 &lt;chr&gt; | variable &lt;fct&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std.error &lt;dbl&gt; | statistic &lt;dbl&gt; | p.value &lt;dbl&gt; | term_z &lt;chr&gt; | estimate_z &lt;dbl&gt; | ⋯ ⋯ | mean_ncells.x &lt;dbl&gt; | sd_ncells.x &lt;dbl&gt; | amount_zero_counts.y &lt;dbl&gt; | var_amount_zero_counts.y &lt;dbl&gt; | mean_counts.y &lt;dbl&gt; | var_mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; | var_percentage_zero.y &lt;dbl&gt; | mean_ncells.y &lt;dbl&gt; | sd_ncells.y &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene2_10 | gene1_10 | pearson               | as.numeric(genotype) | 0.1277987 | 0.04387965 | 2.9124813 | 0.01549873 | as.numeric(genotype) | 0.53349520 | ⋯ | 15.93333 | 6.668788 | 0 | NA | 12.19637 | NA | 0 | NA | 15.93333 | 6.668788 |\n",
       "| 2 | gene2_10 | gene1_10 | scaled_spearman_ranks | as.numeric(genotype) | 0.0216004 | 0.05444663 | 0.3967261 | 0.69990385 | as.numeric(genotype) | 0.04768307 | ⋯ | 15.93333 | 6.668788 | 0 | NA | 12.19637 | NA | 0 | NA | 15.93333 | 6.668788 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene2    gene1    variable              term                 estimate \n",
       "1 gene2_10 gene1_10 pearson               as.numeric(genotype) 0.1277987\n",
       "2 gene2_10 gene1_10 scaled_spearman_ranks as.numeric(genotype) 0.0216004\n",
       "  std.error  statistic p.value    term_z               estimate_z ⋯\n",
       "1 0.04387965 2.9124813 0.01549873 as.numeric(genotype) 0.53349520 ⋯\n",
       "2 0.05444663 0.3967261 0.69990385 as.numeric(genotype) 0.04768307 ⋯\n",
       "  mean_ncells.x sd_ncells.x amount_zero_counts.y var_amount_zero_counts.y\n",
       "1 15.93333      6.668788    0                    NA                      \n",
       "2 15.93333      6.668788    0                    NA                      \n",
       "  mean_counts.y var_mean_counts.y percentage_zero.y var_percentage_zero.y\n",
       "1 12.19637      NA                0                 NA                   \n",
       "2 12.19637      NA                0                 NA                   \n",
       "  mean_ncells.y sd_ncells.y\n",
       "1 15.93333      6.668788   \n",
       "2 15.93333      6.668788   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(lm_result,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1791,
   "id": "371c5106-e75e-47b9-8113-ecab19450977",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Format the information about the true correlations from the specification to calculate the specified difference in correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1792,
   "id": "c41327b4-b8dd-4fb2-92ea-0f8005206726",
   "metadata": {},
   "outputs": [],
   "source": [
    "true_correlations = true_correlations %>% dcast(gene1 + gene2 ~ genotype, value.var = 'correlation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1793,
   "id": "e3d77c45-67ae-4944-8698-fe67567b72dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>XX</th><th scope=col>XY</th><th scope=col>YY</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene1_1 </td><td>gene2_1 </td><td>0</td><td>0.05</td><td>0.1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene1_10</td><td>gene2_10</td><td>0</td><td>0.05</td><td>0.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & gene1 & gene2 & XX & XY & YY\\\\\n",
       "  & <chr> & <chr> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene1\\_1  & gene2\\_1  & 0 & 0.05 & 0.1\\\\\n",
       "\t2 & gene1\\_10 & gene2\\_10 & 0 & 0.05 & 0.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 5\n",
       "\n",
       "| <!--/--> | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | XX &lt;dbl&gt; | XY &lt;dbl&gt; | YY &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | gene1_1  | gene2_1  | 0 | 0.05 | 0.1 |\n",
       "| 2 | gene1_10 | gene2_10 | 0 | 0.05 | 0.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene1    gene2    XX XY   YY \n",
       "1 gene1_1  gene2_1  0  0.05 0.1\n",
       "2 gene1_10 gene2_10 0  0.05 0.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(true_correlations,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1794,
   "id": "bf91e0be-9d25-4997-aefb-850b8f0cef09",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(ncol(true_correlations) == 4){\n",
    "    colnames(true_correlations) = c('gene1', 'gene2', 'group1_correlation', 'group2_correlation')\n",
    "    true_correlations$cor_diff = true_correlations$group2_correlation - true_correlations$group1_correlation\n",
    "    }\n",
    "if(ncol(true_correlations) == 5){\n",
    "    colnames(true_correlations) = c('gene1', 'gene2', 'group1_correlation', 'group2_correlation', 'group3_correlation')\n",
    "    true_correlations$cor_diff = true_correlations$group3_correlation - true_correlations$group1_correlation\n",
    "    }\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1795,
   "id": "3db87d5e-baf0-4551-bbac-a4e26fa5f745",
   "metadata": {},
   "outputs": [],
   "source": [
    "#colnames(true_correlations) = c('gene1', 'gene2', 'group1_correlation', 'group2_correlation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1796,
   "id": "691413fd-57cb-43bc-b7f4-6494c900f275",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>group1_correlation</th><th scope=col>group2_correlation</th><th scope=col>group3_correlation</th><th scope=col>cor_diff</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene1_1 </td><td>gene2_1 </td><td>0</td><td>0.05</td><td>0.1</td><td>0.1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene1_10</td><td>gene2_10</td><td>0</td><td>0.05</td><td>0.1</td><td>0.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & gene1 & gene2 & group1\\_correlation & group2\\_correlation & group3\\_correlation & cor\\_diff\\\\\n",
       "  & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene1\\_1  & gene2\\_1  & 0 & 0.05 & 0.1 & 0.1\\\\\n",
       "\t2 & gene1\\_10 & gene2\\_10 & 0 & 0.05 & 0.1 & 0.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 6\n",
       "\n",
       "| <!--/--> | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | group1_correlation &lt;dbl&gt; | group2_correlation &lt;dbl&gt; | group3_correlation &lt;dbl&gt; | cor_diff &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1  | gene2_1  | 0 | 0.05 | 0.1 | 0.1 |\n",
       "| 2 | gene1_10 | gene2_10 | 0 | 0.05 | 0.1 | 0.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene1    gene2    group1_correlation group2_correlation group3_correlation\n",
       "1 gene1_1  gene2_1  0                  0.05               0.1               \n",
       "2 gene1_10 gene2_10 0                  0.05               0.1               \n",
       "  cor_diff\n",
       "1 0.1     \n",
       "2 0.1     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(true_correlations,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1797,
   "id": "9bb08331-111c-4cde-8845-935ab4dd6cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#true_correlations$cor_diff = true_correlations$group2_correlation - true_correlations$group1_correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1798,
   "id": "1037d2cd-36c3-42f3-9b12-1885d162e740",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "10000"
      ],
      "text/latex": [
       "10000"
      ],
      "text/markdown": [
       "10000"
      ],
      "text/plain": [
       "[1] 10000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(true_correlations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1799,
   "id": "a4effb8e-d1f0-4427-80ee-8b407f01ec0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Add the calculated true correlation differences to the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1800,
   "id": "1b4fa6f2-2574-4918-9921-15e41930b49a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(ncol(true_correlations) == 5){\n",
    "    lm_result = merge(lm_result, true_correlations[,c('gene1', 'gene2',  'group1_correlation' , 'group2_correlation',  'cor_diff')],\n",
    "          by.x = c('gene1', 'gene2'),\n",
    "          by.y = c('gene1', 'gene2'))\n",
    "    }\n",
    "if(ncol(true_correlations) == 6){\n",
    "    lm_result = merge(lm_result, true_correlations[,c('gene1', 'gene2',  'group1_correlation' , 'group2_correlation',  'group3_correlation', 'cor_diff')],\n",
    "          by.x = c('gene1', 'gene2'),\n",
    "          by.y = c('gene1', 'gene2'))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1801,
   "id": "da066df6-ae29-4244-a94c-2016d182af47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 35</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene1</th><th scope=col>gene2</th><th scope=col>variable</th><th scope=col>term</th><th scope=col>estimate</th><th scope=col>std.error</th><th scope=col>statistic</th><th scope=col>p.value</th><th scope=col>term_z</th><th scope=col>estimate_z</th><th scope=col>⋯</th><th scope=col>mean_counts.y</th><th scope=col>var_mean_counts.y</th><th scope=col>percentage_zero.y</th><th scope=col>var_percentage_zero.y</th><th scope=col>mean_ncells.y</th><th scope=col>sd_ncells.y</th><th scope=col>group1_correlation</th><th scope=col>group2_correlation</th><th scope=col>group3_correlation</th><th scope=col>cor_diff</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>gene1_10</td><td>gene2_10</td><td>pearson              </td><td>as.numeric(genotype)</td><td>0.1277987</td><td>0.04387965</td><td>2.9124813</td><td>0.01549873</td><td>as.numeric(genotype)</td><td>0.53349520</td><td>⋯</td><td>12.19637</td><td>NA</td><td>0</td><td>NA</td><td>15.93333</td><td>6.668788</td><td>0</td><td>0.05</td><td>0.1</td><td>0.1</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>gene1_10</td><td>gene2_10</td><td>scaled_spearman_ranks</td><td>as.numeric(genotype)</td><td>0.0216004</td><td>0.05444663</td><td>0.3967261</td><td>0.69990385</td><td>as.numeric(genotype)</td><td>0.04768307</td><td>⋯</td><td>12.19637</td><td>NA</td><td>0</td><td>NA</td><td>15.93333</td><td>6.668788</td><td>0</td><td>0.05</td><td>0.1</td><td>0.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 35\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & gene1 & gene2 & variable & term & estimate & std.error & statistic & p.value & term\\_z & estimate\\_z & ⋯ & mean\\_counts.y & var\\_mean\\_counts.y & percentage\\_zero.y & var\\_percentage\\_zero.y & mean\\_ncells.y & sd\\_ncells.y & group1\\_correlation & group2\\_correlation & group3\\_correlation & cor\\_diff\\\\\n",
       "  & <chr> & <chr> & <fct> & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & gene1\\_10 & gene2\\_10 & pearson               & as.numeric(genotype) & 0.1277987 & 0.04387965 & 2.9124813 & 0.01549873 & as.numeric(genotype) & 0.53349520 & ⋯ & 12.19637 & NA & 0 & NA & 15.93333 & 6.668788 & 0 & 0.05 & 0.1 & 0.1\\\\\n",
       "\t2 & gene1\\_10 & gene2\\_10 & scaled\\_spearman\\_ranks & as.numeric(genotype) & 0.0216004 & 0.05444663 & 0.3967261 & 0.69990385 & as.numeric(genotype) & 0.04768307 & ⋯ & 12.19637 & NA & 0 & NA & 15.93333 & 6.668788 & 0 & 0.05 & 0.1 & 0.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 35\n",
       "\n",
       "| <!--/--> | gene1 &lt;chr&gt; | gene2 &lt;chr&gt; | variable &lt;fct&gt; | term &lt;chr&gt; | estimate &lt;dbl&gt; | std.error &lt;dbl&gt; | statistic &lt;dbl&gt; | p.value &lt;dbl&gt; | term_z &lt;chr&gt; | estimate_z &lt;dbl&gt; | ⋯ ⋯ | mean_counts.y &lt;dbl&gt; | var_mean_counts.y &lt;dbl&gt; | percentage_zero.y &lt;dbl&gt; | var_percentage_zero.y &lt;dbl&gt; | mean_ncells.y &lt;dbl&gt; | sd_ncells.y &lt;dbl&gt; | group1_correlation &lt;dbl&gt; | group2_correlation &lt;dbl&gt; | group3_correlation &lt;dbl&gt; | cor_diff &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_10 | gene2_10 | pearson               | as.numeric(genotype) | 0.1277987 | 0.04387965 | 2.9124813 | 0.01549873 | as.numeric(genotype) | 0.53349520 | ⋯ | 12.19637 | NA | 0 | NA | 15.93333 | 6.668788 | 0 | 0.05 | 0.1 | 0.1 |\n",
       "| 2 | gene1_10 | gene2_10 | scaled_spearman_ranks | as.numeric(genotype) | 0.0216004 | 0.05444663 | 0.3967261 | 0.69990385 | as.numeric(genotype) | 0.04768307 | ⋯ | 12.19637 | NA | 0 | NA | 15.93333 | 6.668788 | 0 | 0.05 | 0.1 | 0.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  gene1    gene2    variable              term                 estimate \n",
       "1 gene1_10 gene2_10 pearson               as.numeric(genotype) 0.1277987\n",
       "2 gene1_10 gene2_10 scaled_spearman_ranks as.numeric(genotype) 0.0216004\n",
       "  std.error  statistic p.value    term_z               estimate_z ⋯\n",
       "1 0.04387965 2.9124813 0.01549873 as.numeric(genotype) 0.53349520 ⋯\n",
       "2 0.05444663 0.3967261 0.69990385 as.numeric(genotype) 0.04768307 ⋯\n",
       "  mean_counts.y var_mean_counts.y percentage_zero.y var_percentage_zero.y\n",
       "1 12.19637      NA                0                 NA                   \n",
       "2 12.19637      NA                0                 NA                   \n",
       "  mean_ncells.y sd_ncells.y group1_correlation group2_correlation\n",
       "1 15.93333      6.668788    0                  0.05              \n",
       "2 15.93333      6.668788    0                  0.05              \n",
       "  group3_correlation cor_diff\n",
       "1 0.1                0.1     \n",
       "2 0.1                0.1     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(lm_result,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1802,
   "id": "23ab9581-a330-4aa6-8fda-59f70e38bc7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "54840"
      ],
      "text/latex": [
       "54840"
      ],
      "text/markdown": [
       "54840"
      ],
      "text/plain": [
       "[1] 54840"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(lm_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1803,
   "id": "339dc829-ba6c-46dd-a924-3e7ec09b240e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'gene1'</li><li>'gene2'</li><li>'variable'</li><li>'term'</li><li>'estimate'</li><li>'std.error'</li><li>'statistic'</li><li>'p.value'</li><li>'term_z'</li><li>'estimate_z'</li><li>'std.error_z'</li><li>'statistic_z'</li><li>'p.value_z'</li><li>'amount_tests'</li><li>'p.value_adj'</li><li>'amount_zero_counts.x'</li><li>'var_amount_zero_counts.x'</li><li>'mean_counts.x'</li><li>'var_mean_counts.x'</li><li>'percentage_zero.x'</li><li>'var_percentage_zero.x'</li><li>'mean_ncells.x'</li><li>'sd_ncells.x'</li><li>'amount_zero_counts.y'</li><li>'var_amount_zero_counts.y'</li><li>'mean_counts.y'</li><li>'var_mean_counts.y'</li><li>'percentage_zero.y'</li><li>'var_percentage_zero.y'</li><li>'mean_ncells.y'</li><li>'sd_ncells.y'</li><li>'group1_correlation'</li><li>'group2_correlation'</li><li>'group3_correlation'</li><li>'cor_diff'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'gene1'\n",
       "\\item 'gene2'\n",
       "\\item 'variable'\n",
       "\\item 'term'\n",
       "\\item 'estimate'\n",
       "\\item 'std.error'\n",
       "\\item 'statistic'\n",
       "\\item 'p.value'\n",
       "\\item 'term\\_z'\n",
       "\\item 'estimate\\_z'\n",
       "\\item 'std.error\\_z'\n",
       "\\item 'statistic\\_z'\n",
       "\\item 'p.value\\_z'\n",
       "\\item 'amount\\_tests'\n",
       "\\item 'p.value\\_adj'\n",
       "\\item 'amount\\_zero\\_counts.x'\n",
       "\\item 'var\\_amount\\_zero\\_counts.x'\n",
       "\\item 'mean\\_counts.x'\n",
       "\\item 'var\\_mean\\_counts.x'\n",
       "\\item 'percentage\\_zero.x'\n",
       "\\item 'var\\_percentage\\_zero.x'\n",
       "\\item 'mean\\_ncells.x'\n",
       "\\item 'sd\\_ncells.x'\n",
       "\\item 'amount\\_zero\\_counts.y'\n",
       "\\item 'var\\_amount\\_zero\\_counts.y'\n",
       "\\item 'mean\\_counts.y'\n",
       "\\item 'var\\_mean\\_counts.y'\n",
       "\\item 'percentage\\_zero.y'\n",
       "\\item 'var\\_percentage\\_zero.y'\n",
       "\\item 'mean\\_ncells.y'\n",
       "\\item 'sd\\_ncells.y'\n",
       "\\item 'group1\\_correlation'\n",
       "\\item 'group2\\_correlation'\n",
       "\\item 'group3\\_correlation'\n",
       "\\item 'cor\\_diff'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'gene1'\n",
       "2. 'gene2'\n",
       "3. 'variable'\n",
       "4. 'term'\n",
       "5. 'estimate'\n",
       "6. 'std.error'\n",
       "7. 'statistic'\n",
       "8. 'p.value'\n",
       "9. 'term_z'\n",
       "10. 'estimate_z'\n",
       "11. 'std.error_z'\n",
       "12. 'statistic_z'\n",
       "13. 'p.value_z'\n",
       "14. 'amount_tests'\n",
       "15. 'p.value_adj'\n",
       "16. 'amount_zero_counts.x'\n",
       "17. 'var_amount_zero_counts.x'\n",
       "18. 'mean_counts.x'\n",
       "19. 'var_mean_counts.x'\n",
       "20. 'percentage_zero.x'\n",
       "21. 'var_percentage_zero.x'\n",
       "22. 'mean_ncells.x'\n",
       "23. 'sd_ncells.x'\n",
       "24. 'amount_zero_counts.y'\n",
       "25. 'var_amount_zero_counts.y'\n",
       "26. 'mean_counts.y'\n",
       "27. 'var_mean_counts.y'\n",
       "28. 'percentage_zero.y'\n",
       "29. 'var_percentage_zero.y'\n",
       "30. 'mean_ncells.y'\n",
       "31. 'sd_ncells.y'\n",
       "32. 'group1_correlation'\n",
       "33. 'group2_correlation'\n",
       "34. 'group3_correlation'\n",
       "35. 'cor_diff'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"gene1\"                    \"gene2\"                   \n",
       " [3] \"variable\"                 \"term\"                    \n",
       " [5] \"estimate\"                 \"std.error\"               \n",
       " [7] \"statistic\"                \"p.value\"                 \n",
       " [9] \"term_z\"                   \"estimate_z\"              \n",
       "[11] \"std.error_z\"              \"statistic_z\"             \n",
       "[13] \"p.value_z\"                \"amount_tests\"            \n",
       "[15] \"p.value_adj\"              \"amount_zero_counts.x\"    \n",
       "[17] \"var_amount_zero_counts.x\" \"mean_counts.x\"           \n",
       "[19] \"var_mean_counts.x\"        \"percentage_zero.x\"       \n",
       "[21] \"var_percentage_zero.x\"    \"mean_ncells.x\"           \n",
       "[23] \"sd_ncells.x\"              \"amount_zero_counts.y\"    \n",
       "[25] \"var_amount_zero_counts.y\" \"mean_counts.y\"           \n",
       "[27] \"var_mean_counts.y\"        \"percentage_zero.y\"       \n",
       "[29] \"var_percentage_zero.y\"    \"mean_ncells.y\"           \n",
       "[31] \"sd_ncells.y\"              \"group1_correlation\"      \n",
       "[33] \"group2_correlation\"       \"group3_correlation\"      \n",
       "[35] \"cor_diff\"                "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(lm_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1804,
   "id": "61bc8517-7a9e-496b-b436-34c304dd542a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate the true z-value difference based on the true correlation difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1805,
   "id": "b0bea11c-cea4-4213-badd-38a163e92f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "lm_result$cor_diff_z = corr_to_z(as.numeric(lm_result$cor_diff), (lm_result$mean_ncells.x + lm_result$mean_ncells.y) /2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1806,
   "id": "b5a45e85-6c9b-426a-ab85-03167bbc38d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1807,
   "id": "500d9980-3414-4195-a61d-26f9903f8cc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(lm_result, paste0(data_path_sim, 'A1_6_Calc_LM', '_', variant, coeqtl_variant, z_value_handle, non_zero_counts_filter,  '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "652a4433-5cc0-4b74-a6bd-fea1da422dfc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fbd6a15-295a-490a-8187-b30e4d4206df",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b832fb5-44f0-473f-96ac-06856c4e6604",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a31c872-6bba-4bf1-aabc-1810aad39d81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5780e380-90db-434d-9589-6ad58a8e5b4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8196344c-0cdc-444f-83a5-c63138fb9f3f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb1ed651-b31f-4d8f-8cd4-8c9095889b51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc6cf2ba-ed1f-4925-8a59-bcd5b7cfeae9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
