{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e0aefede-4889-440b-b522-3199487552c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Script to compare distributions in simulated data to distributions in real data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af6708bc-de11-4765-9607-26a0a31dca4d",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "253f83f7-8152-4cf3-b7b5-8b39b4394c51",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n",
      "\"incomplete final line found by readTableHeader on '../conda_environment/Environment_Configs.csv'\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/home/icb/corinna.losert/miniconda3/envs/scgrn_R_4_1//lib/R/library\"\n"
     ]
    }
   ],
   "source": [
    "source('MS0_Libraries.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c781ef21-de0f-4a28-8129-89e0490e6f96",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS4_Plot_Config.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c495a8af-a40e-4cbf-b733-3adc15f58c41",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17daea84-1149-4899-9cb9-ca88840dd42b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Parameters for co-EQTL data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "bd085e7d-9b26-450d-83a0-d2904b8f6822",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Path to saved simulation results from previous script\n",
    "output_path_sc = \"../results/current/A1/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0eead86a-582c-46cb-b3bd-ede8c94a5775",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Path to summary statistics of real single-cell input datasets\n",
    "data_path_summary_stats = '../data/current/summary_stats/'\n",
    "\n",
    "## Files: Sample_gene_statistic  (info about genes on sample level)\n",
    "#         Sample_summary (information about samples)\n",
    "#         Sample_cell statistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "688cae21-bbe9-4d6f-9168-9227deb42a9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Simulated variant to load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "52fe050a-b16b-45f0-9ecf-125d12f27151",
   "metadata": {},
   "outputs": [],
   "source": [
    "coeqtl_variant = 'scaling'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "51403d2f-92d8-4440-9fe2-2aed5143225c",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant = 'wijstCD4_T3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "id": "974a08e6-2b2a-4f96-854c-2a4d0c5f7d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlations_var = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd90cc2-c6dd-44da-984b-df34267f7f5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Specify which real dataset to load (should match simulated dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "id": "d4c74b37-0819-4e97-b177-e520b19a796d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = 'wijst2018'\n",
    "cell_type = 'CD4_T'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "933b9886-19f2-4cf6-ab82-952085cf0792",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "id": "92df0c00-720e-48bf-b471-fe1ec69fd6ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "source('MS1_Functions.r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "938a7487-efe7-45f6-a400-34e503426322",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "443aa5c0-720c-47ff-862e-87e9dbf1204c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load summaries from real single-cell input datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "602f2cd9-2b1e-49b1-b1e5-976db9938d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "id": "e9ccec80-fa76-4e60-9f40-59d7679289d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types =  cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "11d82fc1-fceb-463a-b692-13c1816f32dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## get for each sample and gene of dataset summary information, like mean expression, percentage zero, ..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "597569a8-9d6d-4b0c-91f1-39bd16dcbc1e",
   "metadata": {},
   "source": [
    "### Sample summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "ae2fb73a-23cc-44a5-9ff0-b49c74176c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_summaries = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "e524ac60-3218-431f-9109-57750a9fc2bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in datasets){\n",
    "    for (j in cell_types){\n",
    "    data = read.csv(paste0(data_path_summary_stats, 'D1_Sample_summarywg3_', i ,  j, '.Qced.Normalized.SCs.Rds.csv'))\n",
    "    data$dataset = i\n",
    "    sample_summaries = rbind(data, sample_summaries)\n",
    "    }\n",
    "    }\n",
    "\n",
    "#D1_Sample_summarywg3_wijst2018CD4_T.Qced.Normalized.SCs.Rds.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "6b252750-6dd3-49fa-8db1-f88b123624f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>Assignment</th><th scope=col>amount_cells</th><th scope=col>dataset</th><th scope=col>cell_type</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>LL001060</td><td>221</td><td>wijst2018</td><td>CD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>LL002176</td><td>338</td><td>wijst2018</td><td>CD4_T.Qced.Normalized.SCs.Rds</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & X & Assignment & amount\\_cells & dataset & cell\\_type\\\\\n",
       "  & <int> & <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & LL001060 & 221 & wijst2018 & CD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\t2 & 2 & LL002176 & 338 & wijst2018 & CD4\\_T.Qced.Normalized.SCs.Rds\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 5\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | Assignment &lt;chr&gt; | amount_cells &lt;int&gt; | dataset &lt;chr&gt; | cell_type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | 1 | LL001060 | 221 | wijst2018 | CD4_T.Qced.Normalized.SCs.Rds |\n",
       "| 2 | 2 | LL002176 | 338 | wijst2018 | CD4_T.Qced.Normalized.SCs.Rds |\n",
       "\n"
      ],
      "text/plain": [
       "  X Assignment amount_cells dataset   cell_type                    \n",
       "1 1 LL001060   221          wijst2018 CD4_T.Qced.Normalized.SCs.Rds\n",
       "2 2 LL002176   338          wijst2018 CD4_T.Qced.Normalized.SCs.Rds"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sample_summaries,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "724da779-d708-4bf0-9007-1e358c3fd8ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample_summaries$cell_type_short = str_replace(sample_summaries$cell_type, '\\\\..*', '')\n",
    "\n",
    "sample_summaries$cell_type_short = str_replace(sample_summaries$cell_type, 'Qced.Normalized.SCs.Rds', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "89b4ef8c-f2c5-4781-91e9-97b5fd7aa988",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'CD4_T.'"
      ],
      "text/latex": [
       "'CD4\\_T.'"
      ],
      "text/markdown": [
       "'CD4_T.'"
      ],
      "text/plain": [
       "[1] \"CD4_T.\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(sample_summaries$cell_type_short)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "479d4d95-2cbb-40fc-a2b5-638ec164a9b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>X</th><th scope=col>Assignment</th><th scope=col>amount_cells</th><th scope=col>dataset</th><th scope=col>cell_type</th><th scope=col>cell_type_short</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>LL001060</td><td>221</td><td>wijst2018</td><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>CD4_T.</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>2</td><td>LL002176</td><td>338</td><td>wijst2018</td><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>CD4_T.</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & X & Assignment & amount\\_cells & dataset & cell\\_type & cell\\_type\\_short\\\\\n",
       "  & <int> & <chr> & <int> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & LL001060 & 221 & wijst2018 & CD4\\_T.Qced.Normalized.SCs.Rds & CD4\\_T.\\\\\n",
       "\t2 & 2 & LL002176 & 338 & wijst2018 & CD4\\_T.Qced.Normalized.SCs.Rds & CD4\\_T.\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 6\n",
       "\n",
       "| <!--/--> | X &lt;int&gt; | Assignment &lt;chr&gt; | amount_cells &lt;int&gt; | dataset &lt;chr&gt; | cell_type &lt;chr&gt; | cell_type_short &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | LL001060 | 221 | wijst2018 | CD4_T.Qced.Normalized.SCs.Rds | CD4_T. |\n",
       "| 2 | 2 | LL002176 | 338 | wijst2018 | CD4_T.Qced.Normalized.SCs.Rds | CD4_T. |\n",
       "\n"
      ],
      "text/plain": [
       "  X Assignment amount_cells dataset   cell_type                    \n",
       "1 1 LL001060   221          wijst2018 CD4_T.Qced.Normalized.SCs.Rds\n",
       "2 2 LL002176   338          wijst2018 CD4_T.Qced.Normalized.SCs.Rds\n",
       "  cell_type_short\n",
       "1 CD4_T.         \n",
       "2 CD4_T.         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Check for uniqueness\n",
    "head(sample_summaries,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "id": "b33bd166-9b74-4eac-9bc9-e926ca80bf3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_summaries$Assignment = str_replace(sample_summaries$Assignment, ';;.*', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "id": "5bda75f9-991e-4380-aabe-85e1a346b4d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Summarise the amount of cells per dataset and cell-type as table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "0dbf14bd-8a4e-4aae-8676-8f1d3975e227",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'cell_type_short'. You can override using\n",
      "the `.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 1 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cell_type_short</th><th scope=col>dataset</th><th scope=col>amount_individuals</th><th scope=col>mean_amount_cells</th><th scope=col>sd_amount_cells</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>CD4_T.</td><td>wijst2018</td><td>41</td><td>262.2195</td><td>125.6834</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 1 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " cell\\_type\\_short & dataset & amount\\_individuals & mean\\_amount\\_cells & sd\\_amount\\_cells\\\\\n",
       " <chr> & <chr> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t CD4\\_T. & wijst2018 & 41 & 262.2195 & 125.6834\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 1 × 5\n",
       "\n",
       "| cell_type_short &lt;chr&gt; | dataset &lt;chr&gt; | amount_individuals &lt;int&gt; | mean_amount_cells &lt;dbl&gt; | sd_amount_cells &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "| CD4_T. | wijst2018 | 41 | 262.2195 | 125.6834 |\n",
       "\n"
      ],
      "text/plain": [
       "  cell_type_short dataset   amount_individuals mean_amount_cells\n",
       "1 CD4_T.          wijst2018 41                 262.2195         \n",
       "  sd_amount_cells\n",
       "1 125.6834       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(sample_summaries[,c('Assignment', 'amount_cells', 'dataset', 'cell_type', 'cell_type_short')]) %>% group_by(cell_type_short, dataset) %>% summarise(amount_individuals = length(unique(Assignment)),\n",
    "                                                                      mean_amount_cells = mean(amount_cells),\n",
    "                                                                      sd_amount_cells = sd(amount_cells)) %>% arrange(mean_amount_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e636d0b-61bd-422b-8531-607494ac0b5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b1ce2c1-f8cd-412f-b88e-a57e432c9560",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e02476b2-6f6e-458e-98e5-eb2ec9776067",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2b9418c2-71cc-47a4-8ab5-8480f423f3ce",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Sample Gene statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "d69294cc-c1cf-460e-893e-c27137d3c159",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_gene_statistic = data.frame()\n",
    "sample_gene_statistic_trans = data.frame()  # tBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "5c9aaee9-e166-4c03-8031-f8a2f88406de",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in datasets){\n",
    "    for (j in cell_types){\n",
    "        summary_stats_name = paste0(\"D1_Sample_gene_statisticwg3_\",i ,  j, \".Qced.Normalized.SCs.Rds.csv\")   # without data for raw count based / for current threshold doesn't matter\n",
    "        if(file.exists(paste0(data_path_summary_stats, summary_stats_name))){\n",
    "            gene_sample_stats = read.csv(paste0(data_path_summary_stats, summary_stats_name))\n",
    "            gene_sample_stats = setDT(gene_sample_stats)\n",
    "            sample_gene_statistic = rbind(gene_sample_stats, sample_gene_statistic)             \n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "1b6ac289-d7ee-474d-8d30-6b5aebb4c82c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>X</th><th scope=col>gene</th><th scope=col>amount_cells</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>amount_non_zero_counts</th><th scope=col>amount_cells_bin</th><th scope=col>percentage_zero_bin</th><th scope=col>sample</th><th scope=col>cell_type</th><th scope=col>dataset</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>NOC2L</td><td>221</td><td>189</td><td>0.14932127</td><td>2</td><td>0</td><td>0.13669272</td><td>85.52036</td><td>32</td><td>(100,250]</td><td>(80,90]</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs_Dan.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><td>2</td><td>HES4 </td><td>221</td><td>209</td><td>0.06334842</td><td>2</td><td>0</td><td>0.07778692</td><td>94.57014</td><td>12</td><td>(100,250]</td><td>(90,95]</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs_Dan.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " X & gene & amount\\_cells & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & amount\\_non\\_zero\\_counts & amount\\_cells\\_bin & percentage\\_zero\\_bin & sample & cell\\_type & dataset\\\\\n",
       " <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <int> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & NOC2L & 221 & 189 & 0.14932127 & 2 & 0 & 0.13669272 & 85.52036 & 32 & (100,250{]} & (80,90{]} & LL001060 & CD4\\_T.Qced.Normalized.SCs\\_Dan.Rds & wg3\\_wijst2018\\\\\n",
       "\t 2 & HES4  & 221 & 209 & 0.06334842 & 2 & 0 & 0.07778692 & 94.57014 & 12 & (100,250{]} & (90,95{]} & LL001060 & CD4\\_T.Qced.Normalized.SCs\\_Dan.Rds & wg3\\_wijst2018\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 15\n",
       "\n",
       "| X &lt;int&gt; | gene &lt;chr&gt; | amount_cells &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | amount_non_zero_counts &lt;int&gt; | amount_cells_bin &lt;chr&gt; | percentage_zero_bin &lt;chr&gt; | sample &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | NOC2L | 221 | 189 | 0.14932127 | 2 | 0 | 0.13669272 | 85.52036 | 32 | (100,250] | (80,90] | LL001060 | CD4_T.Qced.Normalized.SCs_Dan.Rds | wg3_wijst2018 |\n",
       "| 2 | HES4  | 221 | 209 | 0.06334842 | 2 | 0 | 0.07778692 | 94.57014 | 12 | (100,250] | (90,95] | LL001060 | CD4_T.Qced.Normalized.SCs_Dan.Rds | wg3_wijst2018 |\n",
       "\n"
      ],
      "text/plain": [
       "  X gene  amount_cells amount_zero_counts mean_counts max_counts min_counts\n",
       "1 1 NOC2L 221          189                0.14932127  2          0         \n",
       "2 2 HES4  221          209                0.06334842  2          0         \n",
       "  variance_counts percentage_zero amount_non_zero_counts amount_cells_bin\n",
       "1 0.13669272      85.52036        32                     (100,250]       \n",
       "2 0.07778692      94.57014        12                     (100,250]       \n",
       "  percentage_zero_bin sample   cell_type                         dataset      \n",
       "1 (80,90]             LL001060 CD4_T.Qced.Normalized.SCs_Dan.Rds wg3_wijst2018\n",
       "2 (90,95]             LL001060 CD4_T.Qced.Normalized.SCs_Dan.Rds wg3_wijst2018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sample_gene_statistic,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec9ccd67-024c-4edb-ab75-09e23b810fa0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Sample_cell_statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "5d9e80df-938b-4d5b-9b10-44ab551b6fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_cell_statistic = data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "d0428adb-4e88-466d-9ec4-ad971203e825",
   "metadata": {},
   "outputs": [],
   "source": [
    "for(i in datasets){\n",
    "    for (j in cell_types){\n",
    "        summary_stats_name = paste0(\"D1_Sample_cell_statisticwg3_\",i ,  j, \".Qced.Normalized.SCs.Rds.csv\")   # without data for raw count based / for current threshold doesn't matter\n",
    "        if(file.exists(paste0(data_path_summary_stats, summary_stats_name))){\n",
    "            gene_sample_stats = read.csv(paste0(data_path_summary_stats, summary_stats_name))\n",
    "            gene_sample_stats = setDT(gene_sample_stats)\n",
    "            sample_cell_statistic = rbind(gene_sample_stats,  sample_cell_statistic)             \n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "7cf0e997-2384-48cf-832e-0f562ea32064",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>X</th><th scope=col>cell</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>amount_non_zero_counts</th><th scope=col>sample</th><th scope=col>cell_type</th><th scope=col>dataset</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>AAACCTGGTGCCTTGG_pilot3_lane1</td><td>2078</td><td>19422</td><td>18516</td><td>0.10699207</td><td>180</td><td>0</td><td>2.4715344</td><td>0.9533519</td><td>906</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><td>2</td><td>AAAGCAACACATCCGG_pilot3_lane1</td><td>1303</td><td>19422</td><td>18973</td><td>0.06708887</td><td> 85</td><td>0</td><td>0.9660462</td><td>0.9768819</td><td>449</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " X & cell & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & amount\\_non\\_zero\\_counts & sample & cell\\_type & dataset\\\\\n",
       " <int> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <int> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & AAACCTGGTGCCTTGG\\_pilot3\\_lane1 & 2078 & 19422 & 18516 & 0.10699207 & 180 & 0 & 2.4715344 & 0.9533519 & 906 & LL001060 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\t 2 & AAAGCAACACATCCGG\\_pilot3\\_lane1 & 1303 & 19422 & 18973 & 0.06708887 &  85 & 0 & 0.9660462 & 0.9768819 & 449 & LL001060 & CD4\\_T.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| X &lt;int&gt; | cell &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | amount_non_zero_counts &lt;int&gt; | sample &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | AAACCTGGTGCCTTGG_pilot3_lane1 | 2078 | 19422 | 18516 | 0.10699207 | 180 | 0 | 2.4715344 | 0.9533519 | 906 | LL001060 | CD4_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "| 2 | AAAGCAACACATCCGG_pilot3_lane1 | 1303 | 19422 | 18973 | 0.06708887 |  85 | 0 | 0.9660462 | 0.9768819 | 449 | LL001060 | CD4_T.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "\n"
      ],
      "text/plain": [
       "  X cell                          amount_counts amount_genes amount_zero_counts\n",
       "1 1 AAACCTGGTGCCTTGG_pilot3_lane1 2078          19422        18516             \n",
       "2 2 AAAGCAACACATCCGG_pilot3_lane1 1303          19422        18973             \n",
       "  mean_counts max_counts min_counts variance_counts percentage_zero\n",
       "1 0.10699207  180        0          2.4715344       0.9533519      \n",
       "2 0.06708887   85        0          0.9660462       0.9768819      \n",
       "  amount_non_zero_counts sample   cell_type                     dataset      \n",
       "1 906                    LL001060 CD4_T.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "2 449                    LL001060 CD4_T.Qced.Normalized.SCs.Rds wg3_wijst2018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sample_cell_statistic,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0ba4319-318a-4e6f-9711-dc136e4fca64",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "e125299a-9f21-4d3b-ab4a-59f1670b15d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### counts\n",
    "count_data = fread(paste0(output_path_sc, 'A1_4_Simulated_Counts' , variant, coeqtl_variant , '.csv.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "bbf41432-1247-419d-a9df-0aa4e2928162",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>cell_ident</th><th scope=col>value</th><th scope=col>scaling</th><th scope=col>mu</th><th scope=col>size</th><th scope=col>counts</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td>cell_XX;1_149_1 </td><td>0.9802071</td><td>0.9754055</td><td>1</td><td>12</td><td>4</td><td>1</td><td>149</td><td>XX;1</td></tr>\n",
       "\t<tr><td>2</td><td>gene1_1</td><td>cell_XX;1_149_39</td><td>0.3085414</td><td>0.7945828</td><td>1</td><td>12</td><td>1</td><td>1</td><td>149</td><td>XX;1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 11\n",
       "\\begin{tabular}{lllllllllll}\n",
       " V1 & variable & cell\\_ident & value & scaling & mu & size & counts & basemean & ncells & ind\\\\\n",
       " <int> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <int> & <int> & <dbl> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 & cell\\_XX;1\\_149\\_1  & 0.9802071 & 0.9754055 & 1 & 12 & 4 & 1 & 149 & XX;1\\\\\n",
       "\t 2 & gene1\\_1 & cell\\_XX;1\\_149\\_39 & 0.3085414 & 0.7945828 & 1 & 12 & 1 & 1 & 149 & XX;1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 11\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | cell_ident &lt;chr&gt; | value &lt;dbl&gt; | scaling &lt;dbl&gt; | mu &lt;dbl&gt; | size &lt;int&gt; | counts &lt;int&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 | cell_XX;1_149_1  | 0.9802071 | 0.9754055 | 1 | 12 | 4 | 1 | 149 | XX;1 |\n",
       "| 2 | gene1_1 | cell_XX;1_149_39 | 0.3085414 | 0.7945828 | 1 | 12 | 1 | 1 | 149 | XX;1 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable cell_ident       value     scaling   mu size counts basemean\n",
       "1 1  gene1_1  cell_XX;1_149_1  0.9802071 0.9754055 1  12   4      1       \n",
       "2 2  gene1_1  cell_XX;1_149_39 0.3085414 0.7945828 1  12   1      1       \n",
       "  ncells ind \n",
       "1 149    XX;1\n",
       "2 149    XX;1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(count_data,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "49475590-8e23-4626-af36-ef19c72968d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Summary stats of simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "5fe1b2d7-d9ce-4d00-8265-a76b2ba53df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_summary = fread(paste0(output_path_sc, 'A1_4_Gene_Summary_' ,variant, '_', coeqtl_variant , '.csv.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "id": "a2debf5c-8e56-4eee-b918-6bc250d50d7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 13</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>amount_zero_counts</th><th scope=col>amount_cells</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>percentage_zero_bin</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td> 1</td><td>149</td><td>XX;1</td><td>53</td><td>149</td><td> 1.006711</td><td> 5</td><td>0</td><td> 1.033738</td><td>35.57047</td><td>(20,40]   </td></tr>\n",
       "\t<tr><td>2</td><td>gene2_1</td><td>16</td><td>149</td><td>XX;1</td><td> 0</td><td>149</td><td>16.127517</td><td>42</td><td>3</td><td>55.760657</td><td> 0.00000</td><td>(-0.001,5]</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 13\n",
       "\\begin{tabular}{lllllllllllll}\n",
       " V1 & variable & basemean & ncells & ind & amount\\_zero\\_counts & amount\\_cells & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & percentage\\_zero\\_bin\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 &  1 & 149 & XX;1 & 53 & 149 &  1.006711 &  5 & 0 &  1.033738 & 35.57047 & (20,40{]}   \\\\\n",
       "\t 2 & gene2\\_1 & 16 & 149 & XX;1 &  0 & 149 & 16.127517 & 42 & 3 & 55.760657 &  0.00000 & (-0.001,5{]}\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 13\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | amount_zero_counts &lt;int&gt; | amount_cells &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | percentage_zero_bin &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 |  1 | 149 | XX;1 | 53 | 149 |  1.006711 |  5 | 0 |  1.033738 | 35.57047 | (20,40]    |\n",
       "| 2 | gene2_1 | 16 | 149 | XX;1 |  0 | 149 | 16.127517 | 42 | 3 | 55.760657 |  0.00000 | (-0.001,5] |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable basemean ncells ind  amount_zero_counts amount_cells mean_counts\n",
       "1 1  gene1_1   1       149    XX;1 53                 149           1.006711  \n",
       "2 2  gene2_1  16       149    XX;1  0                 149          16.127517  \n",
       "  max_counts min_counts variance_counts percentage_zero percentage_zero_bin\n",
       "1  5         0           1.033738       35.57047        (20,40]            \n",
       "2 42         3          55.760657        0.00000        (-0.001,5]         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "dcc72bec-eb4e-4ae5-be24-8f85c1fd2170",
   "metadata": {},
   "outputs": [],
   "source": [
    "## binarize amount of cells\n",
    "gene_summary = gene_summary %>% mutate(amount_cells_bin = cut(amount_cells, breaks=c(0,25, 50, 100,250, 500, 750, 1500)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "29c62fd5-9710-48da-ac64-0ddf75e9d49c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_summary = fread(paste0(output_path_sc, 'A1_4_Cell_Summary_' ,variant, '_', coeqtl_variant , '.csv.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "5b299363-ea84-4b62-9346-263b77862d90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>cell_ident</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>var_counts</th><th scope=col>percentage_zero</th><th scope=col>total_mean_counts</th><th scope=col>scaling</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>149</td><td>XX;1</td><td>cell_XX;1_149_1 </td><td>101937</td><td>20000</td><td>7948</td><td>5.09685</td><td>61</td><td>0</td><td>53.19423</td><td>0.3974</td><td>5.865534</td><td>1.150815</td></tr>\n",
       "\t<tr><td>2</td><td>149</td><td>XX;1</td><td>cell_XX;1_149_10</td><td>111749</td><td>20000</td><td>7262</td><td>5.58745</td><td>55</td><td>0</td><td>61.15941</td><td>0.3631</td><td>5.865534</td><td>1.049769</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & ncells & ind & cell\\_ident & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & var\\_counts & percentage\\_zero & total\\_mean\\_counts & scaling\\\\\n",
       " <int> & <int> & <chr> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1 & 149 & XX;1 & cell\\_XX;1\\_149\\_1  & 101937 & 20000 & 7948 & 5.09685 & 61 & 0 & 53.19423 & 0.3974 & 5.865534 & 1.150815\\\\\n",
       "\t 2 & 149 & XX;1 & cell\\_XX;1\\_149\\_10 & 111749 & 20000 & 7262 & 5.58745 & 55 & 0 & 61.15941 & 0.3631 & 5.865534 & 1.049769\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | cell_ident &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | var_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | total_mean_counts &lt;dbl&gt; | scaling &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 149 | XX;1 | cell_XX;1_149_1  | 101937 | 20000 | 7948 | 5.09685 | 61 | 0 | 53.19423 | 0.3974 | 5.865534 | 1.150815 |\n",
       "| 2 | 149 | XX;1 | cell_XX;1_149_10 | 111749 | 20000 | 7262 | 5.58745 | 55 | 0 | 61.15941 | 0.3631 | 5.865534 | 1.049769 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 ncells ind  cell_ident       amount_counts amount_genes amount_zero_counts\n",
       "1 1  149    XX;1 cell_XX;1_149_1  101937        20000        7948              \n",
       "2 2  149    XX;1 cell_XX;1_149_10 111749        20000        7262              \n",
       "  mean_counts max_counts min_counts var_counts percentage_zero\n",
       "1 5.09685     61         0          53.19423   0.3974         \n",
       "2 5.58745     55         0          61.15941   0.3631         \n",
       "  total_mean_counts scaling \n",
       "1 5.865534          1.150815\n",
       "2 5.865534          1.049769"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_summary,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57e5470a-205d-4bb0-a469-f27f128160b9",
   "metadata": {},
   "source": [
    "# Compare distributions/ investigate relationships"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9378d794-dd47-494c-80ee-15a9675da3ea",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Gene level relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "ddf8bea1-7d10-4caf-9fa2-1755b152f9ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Combine with real data for comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "id": "a7b16520-dd5e-4c10-88a6-957afe9bf1f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>X</th><th scope=col>gene</th><th scope=col>amount_cells</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>amount_non_zero_counts</th><th scope=col>amount_cells_bin</th><th scope=col>percentage_zero_bin</th><th scope=col>sample</th><th scope=col>cell_type</th><th scope=col>dataset</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>NOC2L</td><td>221</td><td>189</td><td>0.14932127</td><td>2</td><td>0</td><td>0.13669272</td><td>85.52036</td><td>32</td><td>(100,250]</td><td>(80,90]</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs_Dan.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><td>2</td><td>HES4 </td><td>221</td><td>209</td><td>0.06334842</td><td>2</td><td>0</td><td>0.07778692</td><td>94.57014</td><td>12</td><td>(100,250]</td><td>(90,95]</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs_Dan.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " X & gene & amount\\_cells & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & amount\\_non\\_zero\\_counts & amount\\_cells\\_bin & percentage\\_zero\\_bin & sample & cell\\_type & dataset\\\\\n",
       " <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <int> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & NOC2L & 221 & 189 & 0.14932127 & 2 & 0 & 0.13669272 & 85.52036 & 32 & (100,250{]} & (80,90{]} & LL001060 & CD4\\_T.Qced.Normalized.SCs\\_Dan.Rds & wg3\\_wijst2018\\\\\n",
       "\t 2 & HES4  & 221 & 209 & 0.06334842 & 2 & 0 & 0.07778692 & 94.57014 & 12 & (100,250{]} & (90,95{]} & LL001060 & CD4\\_T.Qced.Normalized.SCs\\_Dan.Rds & wg3\\_wijst2018\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 15\n",
       "\n",
       "| X &lt;int&gt; | gene &lt;chr&gt; | amount_cells &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | amount_non_zero_counts &lt;int&gt; | amount_cells_bin &lt;chr&gt; | percentage_zero_bin &lt;chr&gt; | sample &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | NOC2L | 221 | 189 | 0.14932127 | 2 | 0 | 0.13669272 | 85.52036 | 32 | (100,250] | (80,90] | LL001060 | CD4_T.Qced.Normalized.SCs_Dan.Rds | wg3_wijst2018 |\n",
       "| 2 | HES4  | 221 | 209 | 0.06334842 | 2 | 0 | 0.07778692 | 94.57014 | 12 | (100,250] | (90,95] | LL001060 | CD4_T.Qced.Normalized.SCs_Dan.Rds | wg3_wijst2018 |\n",
       "\n"
      ],
      "text/plain": [
       "  X gene  amount_cells amount_zero_counts mean_counts max_counts min_counts\n",
       "1 1 NOC2L 221          189                0.14932127  2          0         \n",
       "2 2 HES4  221          209                0.06334842  2          0         \n",
       "  variance_counts percentage_zero amount_non_zero_counts amount_cells_bin\n",
       "1 0.13669272      85.52036        32                     (100,250]       \n",
       "2 0.07778692      94.57014        12                     (100,250]       \n",
       "  percentage_zero_bin sample   cell_type                         dataset      \n",
       "1 (80,90]             LL001060 CD4_T.Qced.Normalized.SCs_Dan.Rds wg3_wijst2018\n",
       "2 (90,95]             LL001060 CD4_T.Qced.Normalized.SCs_Dan.Rds wg3_wijst2018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sample_gene_statistic,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "id": "50227136-6197-423e-8f62-2db56e72c56b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulated data summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "id": "ed545992-f69d-4be3-afe1-ea898558c437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>variable</th><th scope=col>basemean</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>amount_zero_counts</th><th scope=col>amount_cells</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>percentage_zero_bin</th><th scope=col>amount_cells_bin</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>gene1_1</td><td> 1</td><td>149</td><td>XX;1</td><td>53</td><td>149</td><td> 1.006711</td><td> 5</td><td>0</td><td> 1.033738</td><td>35.57047</td><td>(20,40]   </td><td>(100,250]</td></tr>\n",
       "\t<tr><td>2</td><td>gene2_1</td><td>16</td><td>149</td><td>XX;1</td><td> 0</td><td>149</td><td>16.127517</td><td>42</td><td>3</td><td>55.760657</td><td> 0.00000</td><td>(-0.001,5]</td><td>(100,250]</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & variable & basemean & ncells & ind & amount\\_zero\\_counts & amount\\_cells & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & percentage\\_zero\\_bin & amount\\_cells\\_bin\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr> & <fct>\\\\\n",
       "\\hline\n",
       "\t 1 & gene1\\_1 &  1 & 149 & XX;1 & 53 & 149 &  1.006711 &  5 & 0 &  1.033738 & 35.57047 & (20,40{]}    & (100,250{]}\\\\\n",
       "\t 2 & gene2\\_1 & 16 & 149 & XX;1 &  0 & 149 & 16.127517 & 42 & 3 & 55.760657 &  0.00000 & (-0.001,5{]} & (100,250{]}\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | variable &lt;chr&gt; | basemean &lt;dbl&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | amount_zero_counts &lt;int&gt; | amount_cells &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | percentage_zero_bin &lt;chr&gt; | amount_cells_bin &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | gene1_1 |  1 | 149 | XX;1 | 53 | 149 |  1.006711 |  5 | 0 |  1.033738 | 35.57047 | (20,40]    | (100,250] |\n",
       "| 2 | gene2_1 | 16 | 149 | XX;1 |  0 | 149 | 16.127517 | 42 | 3 | 55.760657 |  0.00000 | (-0.001,5] | (100,250] |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 variable basemean ncells ind  amount_zero_counts amount_cells mean_counts\n",
       "1 1  gene1_1   1       149    XX;1 53                 149           1.006711  \n",
       "2 2  gene2_1  16       149    XX;1  0                 149          16.127517  \n",
       "  max_counts min_counts variance_counts percentage_zero percentage_zero_bin\n",
       "1  5         0           1.033738       35.57047        (20,40]            \n",
       "2 42         3          55.760657        0.00000        (-0.001,5]         \n",
       "  amount_cells_bin\n",
       "1 (100,250]       \n",
       "2 (100,250]       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "id": "55db1cf0-2800-4ef3-985f-9789c8cde5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_summary$type = 'simulated'\n",
    "sample_gene_statistic$type = 'real'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "id": "6ab7a447-5569-4b1b-b12d-c66aa3de3bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_sample_subset = unique(sample_gene_statistic$sample) # decide whether one or many samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "44d763d9-05e3-4277-800c-2c1dcc8cca69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 16</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>X</th><th scope=col>gene</th><th scope=col>amount_cells</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>amount_non_zero_counts</th><th scope=col>amount_cells_bin</th><th scope=col>percentage_zero_bin</th><th scope=col>sample</th><th scope=col>cell_type</th><th scope=col>dataset</th><th scope=col>type</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>NOC2L</td><td>221</td><td>189</td><td>0.14932127</td><td>2</td><td>0</td><td>0.13669272</td><td>85.52036</td><td>32</td><td>(100,250]</td><td>(80,90]</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs_Dan.Rds</td><td>wg3_wijst2018</td><td>real</td></tr>\n",
       "\t<tr><td>2</td><td>HES4 </td><td>221</td><td>209</td><td>0.06334842</td><td>2</td><td>0</td><td>0.07778692</td><td>94.57014</td><td>12</td><td>(100,250]</td><td>(90,95]</td><td>LL001060</td><td>CD4_T.Qced.Normalized.SCs_Dan.Rds</td><td>wg3_wijst2018</td><td>real</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 16\n",
       "\\begin{tabular}{llllllllllllllll}\n",
       " X & gene & amount\\_cells & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & amount\\_non\\_zero\\_counts & amount\\_cells\\_bin & percentage\\_zero\\_bin & sample & cell\\_type & dataset & type\\\\\n",
       " <int> & <chr> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <int> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & NOC2L & 221 & 189 & 0.14932127 & 2 & 0 & 0.13669272 & 85.52036 & 32 & (100,250{]} & (80,90{]} & LL001060 & CD4\\_T.Qced.Normalized.SCs\\_Dan.Rds & wg3\\_wijst2018 & real\\\\\n",
       "\t 2 & HES4  & 221 & 209 & 0.06334842 & 2 & 0 & 0.07778692 & 94.57014 & 12 & (100,250{]} & (90,95{]} & LL001060 & CD4\\_T.Qced.Normalized.SCs\\_Dan.Rds & wg3\\_wijst2018 & real\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 16\n",
       "\n",
       "| X &lt;int&gt; | gene &lt;chr&gt; | amount_cells &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | amount_non_zero_counts &lt;int&gt; | amount_cells_bin &lt;chr&gt; | percentage_zero_bin &lt;chr&gt; | sample &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; | type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | NOC2L | 221 | 189 | 0.14932127 | 2 | 0 | 0.13669272 | 85.52036 | 32 | (100,250] | (80,90] | LL001060 | CD4_T.Qced.Normalized.SCs_Dan.Rds | wg3_wijst2018 | real |\n",
       "| 2 | HES4  | 221 | 209 | 0.06334842 | 2 | 0 | 0.07778692 | 94.57014 | 12 | (100,250] | (90,95] | LL001060 | CD4_T.Qced.Normalized.SCs_Dan.Rds | wg3_wijst2018 | real |\n",
       "\n"
      ],
      "text/plain": [
       "  X gene  amount_cells amount_zero_counts mean_counts max_counts min_counts\n",
       "1 1 NOC2L 221          189                0.14932127  2          0         \n",
       "2 2 HES4  221          209                0.06334842  2          0         \n",
       "  variance_counts percentage_zero amount_non_zero_counts amount_cells_bin\n",
       "1 0.13669272      85.52036        32                     (100,250]       \n",
       "2 0.07778692      94.57014        12                     (100,250]       \n",
       "  percentage_zero_bin sample   cell_type                         dataset      \n",
       "1 (80,90]             LL001060 CD4_T.Qced.Normalized.SCs_Dan.Rds wg3_wijst2018\n",
       "2 (90,95]             LL001060 CD4_T.Qced.Normalized.SCs_Dan.Rds wg3_wijst2018\n",
       "  type\n",
       "1 real\n",
       "2 real"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sample_gene_statistic,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "5b301ad0-ee0b-426e-abf4-306f0f45e5a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "### combine simulated + real dataset\n",
    "plot_data = rbind(gene_summary[,c('mean_counts', 'variance_counts','amount_cells_bin' , 'type', 'percentage_zero', 'max_counts')],\n",
    "                  sample_gene_statistic[sample_gene_statistic$sample %in% random_sample_subset[1:3],c('mean_counts', 'variance_counts','amount_cells_bin',  'type' , 'percentage_zero', 'max_counts')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "id": "6661a8ca-a016-48d6-9fc4-c07e4f525309",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Investigate mean variance relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "id": "955cebec-1a88-4a9d-b676-a29475b6eed1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mSaving 7 x 7 in image\n"
     ]
    }
   ],
   "source": [
    "### Investigate mean variance relationship combined\n",
    "options(repr.plot.width=10, repr.plot.height=10)\n",
    "pdf(paste0('figures/A1_10/Mean_Variance_Combined_Fixed_Parameters', variant, coeqtl_variant, '.pdf'), width =2, height =2)\n",
    "p1 = ggplot(plot_data, aes(x=log(mean_counts+1), y=log(variance_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5)  +ylim(0,15) + xlim(0,10) + plot_config + theme(legend.position = 'None')#  facet_wrap( amount_cells_bin~.  )\n",
    "\n",
    "print(p1)\n",
    "\n",
    "p2 = ggplot(plot_data, aes(x=log(mean_counts+1), y=log(variance_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5)  +ylim(0,15) + xlim(0,10) + facet_wrap( .~ type  ) + plot_config \n",
    "\n",
    "print(p2)\n",
    "dev.off()\n",
    "\n",
    "ggsave(plot = p1, filename= paste0('figures/A1_10/Mean_Variance_Combined_Fixed_Parameters', variant, coeqtl_variant, '.tiff'), dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "7b77cfe9-5d19-4bd0-91c1-f4c5c1ae4846",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Mean zero relationship on simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "id": "c00fd6e8-f489-421b-bc2c-4224b6d1475e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#### Mean zero relationship combined\n",
    "\n",
    "pdf(paste0('figures/A1_10/Mean_Zero_Combined_Fixed_Parameters',variant, coeqtl_variant, '.pdf'), width =2, height =2)\n",
    "p= ggplot(plot_data, aes(x=percentage_zero, y=log(mean_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5)  + ylim(0,10) + plot_config + theme(legend.position = 'None')# +  facet_wrap( amount_cells_bin~ . ) \n",
    "\n",
    "print(p)\n",
    "\n",
    "p= ggplot(plot_data, aes(x=percentage_zero, y=log(mean_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5) + ylim(0,10) +facet_wrap( .~ type  ) + plot_config\n",
    "\n",
    "print(p)\n",
    "\n",
    "dev.off()\n",
    "\n",
    "# take twice log to see curve as in splatter package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "id": "72910193-d0d1-427a-8f24-2de4b503fc63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Max zero relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "id": "889bf323-f2a1-496f-933b-728ba90125a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#### Max zero relationship combined\n",
    "\n",
    "pdf(paste0('figures/A1_10/Max_Counts_Zero_Combined_Fixed_Parameters', variant, coeqtl_variant, '.pdf'), width =2, height =2)\n",
    "p = ggplot(plot_data, aes(x=percentage_zero, y=log(max_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5) + ylim(0,10) + plot_config+ theme(legend.position = 'None') #  +  facet_wrap( amount_cells_bin~ . ) \n",
    "print(p)\n",
    "\n",
    "p = ggplot(plot_data, aes(x=percentage_zero, y=log(max_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5)  + ylim(0,10)  +    facet_wrap( .~ type ) + plot_config\n",
    "print(p)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "a6c5711a-ad51-4776-97aa-24adf01a690a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Variance zero relationship simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "d13d3221-5982-4d18-a23a-bb489fd42ef8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#### Variance zero relationship combined\n",
    "options(repr.plot.width=30, repr.plot.height=30)\n",
    "pdf(paste0('figures/A1_10/Variance_Zero_Combined_Fixed_Parameters',  variant, coeqtl_variant,  '.pdf'), width =2, height =2)\n",
    "p= ggplot(plot_data, aes(x=percentage_zero, y=log(variance_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5)  + ylim(0,15) + plot_config + theme(legend.position = 'None')# +   facet_wrap(amount_cells_bin~ .)\n",
    "print(p)\n",
    "\n",
    "\n",
    "p= ggplot(plot_data, aes(x=percentage_zero, y=log(variance_counts+1), color = type)) + \n",
    "    geom_point(size = 0.1, alpha = 0.5)  + ylim(0,15)  +   facet_wrap(.~ type ) + plot_config\n",
    "print(p)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e23a2049-8a93-420c-a484-1426e36df8cc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e6f6f94e-ef43-4240-8f3a-0762dbffb7eb",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Cell level relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "79c87179-5968-461d-9ffa-887e6992c0c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>cell_ident</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>var_counts</th><th scope=col>percentage_zero</th><th scope=col>total_mean_counts</th><th scope=col>scaling</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>14</td><td>XX;1</td><td>cell_XX;1_14_1 </td><td> 88060</td><td>20000</td><td>8177</td><td>4.40300</td><td>47</td><td>0</td><td>40.38871</td><td>0.40885</td><td>5.377257</td><td>1.2212712</td></tr>\n",
       "\t<tr><td>2</td><td>14</td><td>XX;1</td><td>cell_XX;1_14_10</td><td>115329</td><td>20000</td><td>7162</td><td>5.76645</td><td>50</td><td>0</td><td>63.06186</td><td>0.35810</td><td>5.377257</td><td>0.9325074</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & ncells & ind & cell\\_ident & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & var\\_counts & percentage\\_zero & total\\_mean\\_counts & scaling\\\\\n",
       " <int> & <int> & <chr> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1 & 14 & XX;1 & cell\\_XX;1\\_14\\_1  &  88060 & 20000 & 8177 & 4.40300 & 47 & 0 & 40.38871 & 0.40885 & 5.377257 & 1.2212712\\\\\n",
       "\t 2 & 14 & XX;1 & cell\\_XX;1\\_14\\_10 & 115329 & 20000 & 7162 & 5.76645 & 50 & 0 & 63.06186 & 0.35810 & 5.377257 & 0.9325074\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | cell_ident &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | var_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | total_mean_counts &lt;dbl&gt; | scaling &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 14 | XX;1 | cell_XX;1_14_1  |  88060 | 20000 | 8177 | 4.40300 | 47 | 0 | 40.38871 | 0.40885 | 5.377257 | 1.2212712 |\n",
       "| 2 | 14 | XX;1 | cell_XX;1_14_10 | 115329 | 20000 | 7162 | 5.76645 | 50 | 0 | 63.06186 | 0.35810 | 5.377257 | 0.9325074 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 ncells ind  cell_ident      amount_counts amount_genes amount_zero_counts\n",
       "1 1  14     XX;1 cell_XX;1_14_1   88060        20000        8177              \n",
       "2 2  14     XX;1 cell_XX;1_14_10 115329        20000        7162              \n",
       "  mean_counts max_counts min_counts var_counts percentage_zero\n",
       "1 4.40300     47         0          40.38871   0.40885        \n",
       "2 5.76645     50         0          63.06186   0.35810        \n",
       "  total_mean_counts scaling  \n",
       "1 5.377257          1.2212712\n",
       "2 5.377257          0.9325074"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_summary,2) # simulated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "310fd661-3d2e-4445-b92c-b560eb753f3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1279"
      ],
      "text/latex": [
       "1279"
      ],
      "text/markdown": [
       "1279"
      ],
      "text/plain": [
       "[1] 1279"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "1273"
      ],
      "text/latex": [
       "1273"
      ],
      "text/markdown": [
       "1273"
      ],
      "text/plain": [
       "[1] 1273"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.995308835027365"
      ],
      "text/latex": [
       "0.995308835027365"
      ],
      "text/markdown": [
       "0.995308835027365"
      ],
      "text/plain": [
       "[1] 0.9953088"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### test filtering out cells based on scaling factor in simulated data\n",
    "nrow(cell_summary)\n",
    "nrow(cell_summary[(cell_summary$scaling > 0.5) & (cell_summary$scaling < 2),])\n",
    "\n",
    "nrow(cell_summary[(cell_summary$scaling > 0.5) & (cell_summary$scaling < 2),])/ nrow(cell_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "cab64b44-8831-4d48-856a-da6d419d4230",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "663"
      ],
      "text/latex": [
       "663"
      ],
      "text/markdown": [
       "663"
      ],
      "text/plain": [
       "[1] 663"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### test filtering out cells based on scaling fator in real data\n",
    "nrow(sample_cell_statistic)\n",
    "nrow(sample_cell_statistic[(sample_cell_statistic$scaling_factor < 2 ) & (sample_cell_statistic$scaling_factor > 0.5 ),])\n",
    "nrow(sample_cell_statistic[(sample_cell_statistic$scaling_factor < 2 ) & (sample_cell_statistic$scaling_factor > 0.5 ),]) /nrow(sample_cell_statistic)\n",
    "# filter out around 7% of cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "608d43d5-a897-4116-9124-139a2311dfdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>X</th><th scope=col>cell</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>variance_counts</th><th scope=col>percentage_zero</th><th scope=col>amount_non_zero_counts</th><th scope=col>sample</th><th scope=col>cell_type</th><th scope=col>dataset</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>AAAGCAAGTCGCATCG_pilot3_lane1</td><td>5216</td><td>14699</td><td>13237</td><td>0.3548541</td><td>322</td><td>0</td><td>16.401625</td><td>0.9005375</td><td>1462</td><td>LL007985</td><td>B.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "\t<tr><td>2</td><td>AACTTTCTCACAATGC_pilot3_lane1</td><td>3848</td><td>14699</td><td>13536</td><td>0.2617865</td><td>217</td><td>0</td><td> 8.229191</td><td>0.9208790</td><td>1163</td><td>LL007985</td><td>B.Qced.Normalized.SCs.Rds</td><td>wg3_wijst2018</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " X & cell & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & variance\\_counts & percentage\\_zero & amount\\_non\\_zero\\_counts & sample & cell\\_type & dataset\\\\\n",
       " <int> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <int> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1 & AAAGCAAGTCGCATCG\\_pilot3\\_lane1 & 5216 & 14699 & 13237 & 0.3548541 & 322 & 0 & 16.401625 & 0.9005375 & 1462 & LL007985 & B.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\t 2 & AACTTTCTCACAATGC\\_pilot3\\_lane1 & 3848 & 14699 & 13536 & 0.2617865 & 217 & 0 &  8.229191 & 0.9208790 & 1163 & LL007985 & B.Qced.Normalized.SCs.Rds & wg3\\_wijst2018\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| X &lt;int&gt; | cell &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | variance_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | amount_non_zero_counts &lt;int&gt; | sample &lt;chr&gt; | cell_type &lt;chr&gt; | dataset &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | AAAGCAAGTCGCATCG_pilot3_lane1 | 5216 | 14699 | 13237 | 0.3548541 | 322 | 0 | 16.401625 | 0.9005375 | 1462 | LL007985 | B.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "| 2 | AACTTTCTCACAATGC_pilot3_lane1 | 3848 | 14699 | 13536 | 0.2617865 | 217 | 0 |  8.229191 | 0.9208790 | 1163 | LL007985 | B.Qced.Normalized.SCs.Rds | wg3_wijst2018 |\n",
       "\n"
      ],
      "text/plain": [
       "  X cell                          amount_counts amount_genes amount_zero_counts\n",
       "1 1 AAAGCAAGTCGCATCG_pilot3_lane1 5216          14699        13237             \n",
       "2 2 AACTTTCTCACAATGC_pilot3_lane1 3848          14699        13536             \n",
       "  mean_counts max_counts min_counts variance_counts percentage_zero\n",
       "1 0.3548541   322        0          16.401625       0.9005375      \n",
       "2 0.2617865   217        0           8.229191       0.9208790      \n",
       "  amount_non_zero_counts sample   cell_type                 dataset      \n",
       "1 1462                   LL007985 B.Qced.Normalized.SCs.Rds wg3_wijst2018\n",
       "2 1163                   LL007985 B.Qced.Normalized.SCs.Rds wg3_wijst2018"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sample_cell_statistic,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "82ff5428-b8fd-4458-a4f7-4be56f8e871a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>cell_ident</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>var_counts</th><th scope=col>percentage_zero</th><th scope=col>total_mean_counts</th><th scope=col>scaling</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1</td><td>10</td><td>XX;1</td><td>cell_XX;1_10_1 </td><td>117295</td><td>20000</td><td>8459</td><td>5.86475</td><td>613</td><td>0</td><td>332.1105</td><td>0.42295</td><td>6.408285</td><td>1.092678</td></tr>\n",
       "\t<tr><td>2</td><td>10</td><td>XX;1</td><td>cell_XX;1_10_10</td><td> 94422</td><td>20000</td><td>7796</td><td>4.72110</td><td>403</td><td>0</td><td>186.6399</td><td>0.38980</td><td>6.408285</td><td>1.357371</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 14\n",
       "\\begin{tabular}{llllllllllllll}\n",
       " V1 & ncells & ind & cell\\_ident & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & var\\_counts & percentage\\_zero & total\\_mean\\_counts & scaling\\\\\n",
       " <int> & <int> & <chr> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 1 & 10 & XX;1 & cell\\_XX;1\\_10\\_1  & 117295 & 20000 & 8459 & 5.86475 & 613 & 0 & 332.1105 & 0.42295 & 6.408285 & 1.092678\\\\\n",
       "\t 2 & 10 & XX;1 & cell\\_XX;1\\_10\\_10 &  94422 & 20000 & 7796 & 4.72110 & 403 & 0 & 186.6399 & 0.38980 & 6.408285 & 1.357371\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 14\n",
       "\n",
       "| V1 &lt;int&gt; | ncells &lt;int&gt; | ind &lt;chr&gt; | cell_ident &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | var_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | total_mean_counts &lt;dbl&gt; | scaling &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 10 | XX;1 | cell_XX;1_10_1  | 117295 | 20000 | 8459 | 5.86475 | 613 | 0 | 332.1105 | 0.42295 | 6.408285 | 1.092678 |\n",
       "| 2 | 10 | XX;1 | cell_XX;1_10_10 |  94422 | 20000 | 7796 | 4.72110 | 403 | 0 | 186.6399 | 0.38980 | 6.408285 | 1.357371 |\n",
       "\n"
      ],
      "text/plain": [
       "  V1 ncells ind  cell_ident      amount_counts amount_genes amount_zero_counts\n",
       "1 1  10     XX;1 cell_XX;1_10_1  117295        20000        8459              \n",
       "2 2  10     XX;1 cell_XX;1_10_10  94422        20000        7796              \n",
       "  mean_counts max_counts min_counts var_counts percentage_zero\n",
       "1 5.86475     613        0          332.1105   0.42295        \n",
       "2 4.72110     403        0          186.6399   0.38980        \n",
       "  total_mean_counts scaling \n",
       "1 6.408285          1.092678\n",
       "2 6.408285          1.357371"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "9594a6e7-b76c-4546-9890-b450cde99cfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "567109"
      ],
      "text/latex": [
       "567109"
      ],
      "text/markdown": [
       "567109"
      ],
      "text/plain": [
       "[1] 567109"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "max(cell_summary$amount_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e1adb98a-4b44-4cbc-933e-299776332c5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n",
      "Warning message:\n",
      "\"\u001b[1m\u001b[22mRemoved 90 rows containing missing values or values outside the scale range\n",
      "(`geom_bar()`).\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Distribution of counts per sample and cell-type\n",
    "options(repr.plot.width=30, repr.plot.height=30)\n",
    "pdf(paste0('figures/A1_10/Distribution_Counts_Simulation',  variant, coeqtl_variant, '.pdf'), width =10, height =10)\n",
    "p = ggplot(cell_summary, aes(x=amount_counts)) + \n",
    "     geom_histogram() + facet_wrap(ncells + ind ~ ., scales = 'free') + theme(text = element_text(size =20)) + xlim(0,max(cell_summary$amount_counts)) + plot_config\n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "f02ec602-e692-421f-b6ed-132f6d11ec94",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Correlation calc + defined scaling factor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d7e0df15-3d7e-49c7-a298-ecfeedd55b31",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'ncells'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "total_mean = cell_summary %>% group_by(ncells, ind) %>% summarise(total_mean_counts = mean(mean_counts))\n",
    "cell_summary = merge(cell_summary, total_mean)\n",
    "cell_summary$scaling_calc = cell_summary$total_mean_counts / cell_summary$mean_counts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "1a6feed9-1f56-45f1-b6d2-a2cac28166b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>total_mean_counts</th><th scope=col>V1</th><th scope=col>cell_ident</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>var_counts</th><th scope=col>percentage_zero</th><th scope=col>scaling</th><th scope=col>scaling_calc</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>10</td><td>XX;1</td><td>6.408285</td><td>1</td><td>cell_XX;1_10_1 </td><td>117295</td><td>20000</td><td>8459</td><td>5.86475</td><td>613</td><td>0</td><td>332.1105</td><td>0.42295</td><td>1.092678</td><td>1.092678</td></tr>\n",
       "\t<tr><td>10</td><td>XX;1</td><td>6.408285</td><td>2</td><td>cell_XX;1_10_10</td><td> 94422</td><td>20000</td><td>7796</td><td>4.72110</td><td>403</td><td>0</td><td>186.6399</td><td>0.38980</td><td>1.357371</td><td>1.357371</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " ncells & ind & total\\_mean\\_counts & V1 & cell\\_ident & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & var\\_counts & percentage\\_zero & scaling & scaling\\_calc\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 10 & XX;1 & 6.408285 & 1 & cell\\_XX;1\\_10\\_1  & 117295 & 20000 & 8459 & 5.86475 & 613 & 0 & 332.1105 & 0.42295 & 1.092678 & 1.092678\\\\\n",
       "\t 10 & XX;1 & 6.408285 & 2 & cell\\_XX;1\\_10\\_10 &  94422 & 20000 & 7796 & 4.72110 & 403 & 0 & 186.6399 & 0.38980 & 1.357371 & 1.357371\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 15\n",
       "\n",
       "| ncells &lt;int&gt; | ind &lt;chr&gt; | total_mean_counts &lt;dbl&gt; | V1 &lt;int&gt; | cell_ident &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | var_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | scaling &lt;dbl&gt; | scaling_calc &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 10 | XX;1 | 6.408285 | 1 | cell_XX;1_10_1  | 117295 | 20000 | 8459 | 5.86475 | 613 | 0 | 332.1105 | 0.42295 | 1.092678 | 1.092678 |\n",
       "| 10 | XX;1 | 6.408285 | 2 | cell_XX;1_10_10 |  94422 | 20000 | 7796 | 4.72110 | 403 | 0 | 186.6399 | 0.38980 | 1.357371 | 1.357371 |\n",
       "\n"
      ],
      "text/plain": [
       "  ncells ind  total_mean_counts V1 cell_ident      amount_counts amount_genes\n",
       "1 10     XX;1 6.408285          1  cell_XX;1_10_1  117295        20000       \n",
       "2 10     XX;1 6.408285          2  cell_XX;1_10_10  94422        20000       \n",
       "  amount_zero_counts mean_counts max_counts min_counts var_counts\n",
       "1 8459               5.86475     613        0          332.1105  \n",
       "2 7796               4.72110     403        0          186.6399  \n",
       "  percentage_zero scaling  scaling_calc\n",
       "1 0.42295         1.092678 1.092678    \n",
       "2 0.38980         1.357371 1.357371    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "276fc889-ffdc-4d13-af62-6f7be6d69516",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 1 × 1 of type dbl</caption>\n",
       "<tbody>\n",
       "\t<tr><td>1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 1 × 1 of type dbl\n",
       "\\begin{tabular}{l}\n",
       "\t 1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 1 × 1 of type dbl\n",
       "\n",
       "| 1 |\n",
       "\n"
      ],
      "text/plain": [
       "     [,1]\n",
       "[1,] 1   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cor(cell_summary$scaling, cell_summary$scaling_calc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "035479e1-340f-465a-939e-60df09c3f11e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "120d2c91-8c97-4abe-a4fb-02c66a8cf34b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Relationship Count and Amount Zero Relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b4900ff4-6eb6-4f11-8341-57b098147b89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/A1_10/Cell_Zero_Count_Relationship_Simulation', variant, coeqtl_variant, '.pdf'), width =10, height =5)\n",
    "p =ggplot(cell_summary , aes(x=amount_counts, y=amount_zero_counts, color= ncells)) + \n",
    "    geom_point(size = 0.1) + theme(text = element_text(size = 20)) + facet_wrap(ncells ~ ., scales = 'free')  + plot_config\n",
    "print(p)\n",
    "dev.off()\n",
    "# consider: different amount of genes per cell-type in data!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f2b03adb-bb7c-42a6-ba5c-512c162e0a5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/A1_10/Cell_Zero_Percentage_Count_Relationship_Simulation',  variant, coeqtl_variant, '.pdf'), width =10, height =5)\n",
    "p =ggplot(cell_summary , aes(x=amount_counts, y=percentage_zero, color= ncells)) + \n",
    "    geom_point(size = 0.1) + theme(text = element_text(size = 20)) + facet_wrap(ncells ~ ., scales = 'free')  + ylim(0,1) + plot_config\n",
    "print(p)\n",
    "dev.off()\n",
    "# consider: different amount of genes per cell-type in data!\n",
    "# here: scaling factor needs to be calculated per cell-type as differing amount of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a3de0e89-7d0f-4417-9db5-234cf348b3d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Relationship Scaling Factor / Counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "590ca958-3dc4-43ca-9e21-a7a90456722c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 15</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>ncells</th><th scope=col>ind</th><th scope=col>total_mean_counts</th><th scope=col>V1</th><th scope=col>cell_ident</th><th scope=col>amount_counts</th><th scope=col>amount_genes</th><th scope=col>amount_zero_counts</th><th scope=col>mean_counts</th><th scope=col>max_counts</th><th scope=col>min_counts</th><th scope=col>var_counts</th><th scope=col>percentage_zero</th><th scope=col>scaling</th><th scope=col>scaling_calc</th></tr>\n",
       "\t<tr><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>10</td><td>XX;1</td><td>6.408285</td><td>1</td><td>cell_XX;1_10_1 </td><td>117295</td><td>20000</td><td>8459</td><td>5.86475</td><td>613</td><td>0</td><td>332.1105</td><td>0.42295</td><td>1.092678</td><td>1.092678</td></tr>\n",
       "\t<tr><td>10</td><td>XX;1</td><td>6.408285</td><td>2</td><td>cell_XX;1_10_10</td><td> 94422</td><td>20000</td><td>7796</td><td>4.72110</td><td>403</td><td>0</td><td>186.6399</td><td>0.38980</td><td>1.357371</td><td>1.357371</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 15\n",
       "\\begin{tabular}{lllllllllllllll}\n",
       " ncells & ind & total\\_mean\\_counts & V1 & cell\\_ident & amount\\_counts & amount\\_genes & amount\\_zero\\_counts & mean\\_counts & max\\_counts & min\\_counts & var\\_counts & percentage\\_zero & scaling & scaling\\_calc\\\\\n",
       " <int> & <chr> & <dbl> & <int> & <chr> & <int> & <int> & <int> & <dbl> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 10 & XX;1 & 6.408285 & 1 & cell\\_XX;1\\_10\\_1  & 117295 & 20000 & 8459 & 5.86475 & 613 & 0 & 332.1105 & 0.42295 & 1.092678 & 1.092678\\\\\n",
       "\t 10 & XX;1 & 6.408285 & 2 & cell\\_XX;1\\_10\\_10 &  94422 & 20000 & 7796 & 4.72110 & 403 & 0 & 186.6399 & 0.38980 & 1.357371 & 1.357371\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 15\n",
       "\n",
       "| ncells &lt;int&gt; | ind &lt;chr&gt; | total_mean_counts &lt;dbl&gt; | V1 &lt;int&gt; | cell_ident &lt;chr&gt; | amount_counts &lt;int&gt; | amount_genes &lt;int&gt; | amount_zero_counts &lt;int&gt; | mean_counts &lt;dbl&gt; | max_counts &lt;int&gt; | min_counts &lt;int&gt; | var_counts &lt;dbl&gt; | percentage_zero &lt;dbl&gt; | scaling &lt;dbl&gt; | scaling_calc &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 10 | XX;1 | 6.408285 | 1 | cell_XX;1_10_1  | 117295 | 20000 | 8459 | 5.86475 | 613 | 0 | 332.1105 | 0.42295 | 1.092678 | 1.092678 |\n",
       "| 10 | XX;1 | 6.408285 | 2 | cell_XX;1_10_10 |  94422 | 20000 | 7796 | 4.72110 | 403 | 0 | 186.6399 | 0.38980 | 1.357371 | 1.357371 |\n",
       "\n"
      ],
      "text/plain": [
       "  ncells ind  total_mean_counts V1 cell_ident      amount_counts amount_genes\n",
       "1 10     XX;1 6.408285          1  cell_XX;1_10_1  117295        20000       \n",
       "2 10     XX;1 6.408285          2  cell_XX;1_10_10  94422        20000       \n",
       "  amount_zero_counts mean_counts max_counts min_counts var_counts\n",
       "1 8459               5.86475     613        0          332.1105  \n",
       "2 7796               4.72110     403        0          186.6399  \n",
       "  percentage_zero scaling  scaling_calc\n",
       "1 0.42295         1.092678 1.092678    \n",
       "2 0.38980         1.357371 1.357371    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(cell_summary,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "ea65e304-f41b-4296-89ea-d8902f222771",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/A1_10/Cell_Scaling_Factor_Count_Relationship_Simulation',  variant, coeqtl_variant, '.pdf'), width =10, height =5)\n",
    "p =ggplot(cell_summary , aes(x=amount_counts, y=scaling, color=ncells)) + \n",
    "    geom_point(size = 0.1) + theme(text = element_text(size = 20)) + facet_wrap(ncells ~ ., scales = 'free') + plot_config\n",
    "print(p)\n",
    "p =ggplot(cell_summary , aes(x=amount_counts, y=scaling_calc, color=ncells)) + \n",
    "    geom_point(size = 0.1) + theme(text = element_text(size = 20)) + facet_wrap(ncells ~ ., scales = 'free') + ggtitle('calculated_scaling') + plot_config\n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "084a24bc-952e-4af6-886f-70c666e53a0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Relationship Scaling Factor / Percentage Zero"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4f4dc2ff-1d6d-4454-baad-6be9e0a99518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/A1_10/Cell_Scaling_Factor_Percentage_Zero_Relationship_Simulation',  variant, coeqtl_variant, '.pdf'), width =10, height =5)\n",
    "p =ggplot(cell_summary , aes(x=percentage_zero, y=scaling, color=ncells)) + \n",
    "    geom_point(size = 0.1) + theme(text = element_text(size = 20)) + facet_wrap(ncells~ ., scales = 'free')  + plot_config\n",
    "print(p)\n",
    "p =ggplot(cell_summary , aes(x=percentage_zero, y=scaling_calc, color=ncells)) + \n",
    "    geom_point(size = 0.1) + theme(text = element_text(size = 20)) + facet_wrap(ncells~ ., scales = 'free') + ggtitle('calculated scaling') + plot_config\n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "997e38a0-24a6-4228-b0e3-7666144dda54",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Distribution mean counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "7a4d7c5e-7003-4b99-9dea-5ffb07455b10",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0('figures/A1_10/Distribution_Mean_Counts_Simulation',  variant, coeqtl_variant, '.pdf'), width =10, height =5)\n",
    "p = ggplot(cell_summary , aes(x=mean_counts)) + \n",
    "     geom_histogram() + facet_wrap(ncells + ind ~ ., scales = 'free') + theme(text = element_text(size =20))  + plot_config\n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d80c8fdc-f55f-43e3-be56-b3c07e4a8096",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b27e14ba-9dd0-474c-8253-7c336421e2d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
